---
title: "Antarctic Toothfish: Analysis of methylation output from Bismark"
author: "Floriaan Devloo-Delva"
date: '2025-02-05'
output: pdf_document
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Packages
```{r}
library(methylKit)
# library(GeneDMRs)
library(GenomicRanges)
# library(methyAnalysis)
library(ggplot2)
# library(DSS)
library(GenomicFeatures)
library(IRanges)
library(GenomicRanges)
library(genomation)
library(vegan)
library(data.table)
library(ggbio)
library(poppr) # also loads adegenet
library(ggplot2)
library(ggsci)
library(ggforce)
library(dplyr)
library(ggrepel)

# library(radiator)
library(vcfR)
library(dartRverse)
library(adegenet)

library(ape)
library(apex)
library(pegas)
library(haplotypes)


# source("Filtering funtions.R")

# colours2 <- c("#adcf9f", "#08306B")
# names(colours2) <- c("58.4.2-2024", "88.2-2024")

colours2 <- c("#adcf9f", "seagreen", "#08306B")
names(colours2) <- c("58.4.2-2024_50L", "58.4.2-2024_70L", "88.2-2024")

colours4 <- c("#52B334","#4d6b40", "#F21A00", "#78B7C5")
names(colours4) <- c("58.4.1-2018", "58.4.2-2018", "58.5.2-2017", "88.1-2018")

colours5 <- c("#E1AF00","#52B334","#4d6b40", "#F21A00", "#78B7C5")
names(colours5) <- c("48.2-2017","58.4.1-2018", "58.4.2-2018", "58.5.2-2017", "88.1-2018")


colours6m <- c("#52B334","#4d6b40","#adcf9f","seagreen", "#F21A00", "#78B7C5", "#08306B")
names(colours6m) <- c("58.4.1-2018", "58.4.2-2018", "58.4.2-2024_50L", "58.4.2-2024_70L", "58.5.2-2017",
                     "88.1-2018", "88.2-2024")

colours3f <- c("#E1AF00", "#adcf9f","seagreen","#08306B")
names(colours3f) <- c("48.2-2017", "58.4.2-2024_50L", "58.4.2-2024_70L", "88.2-2024")

colours.all <- c("#E1AF00","#52B334","#4d6b40","#adcf9f", "seagreen","#F21A00", "#78B7C5", "#08306B")
names(colours.all) <- c("48.2-2017", "58.4.1-2018", "58.4.2-2018", "58.4.2-2024_50L", "58.4.2-2024_70L",
                        "58.5.2-2017","88.1-2018", "88.2-2024")
```

# Working directory

```{r}
setwd("/datasets/work/ncmi-toa-rrbs/work/1.Toothfish/0.R-analysis/")
figure.path <- "/datasets/work/ncmi-toa-rrbs/work/1.Toothfish/0.R-analysis/Figures/"
rdata.path <- "/datasets/work/ncmi-toa-rrbs/work/1.Toothfish/0.R-analysis/Rdata/"
```

# 1_Metadata 

```{r}
metadata <- readr::read_csv("/datasets/work/ncmi-toa-rrbs/work/1.Toothfish/0.R-analysis/1.Toothfish_ALL_RRBS_metadata.csv")
```

## No replicated info

```{r}
metadata <- dplyr::filter(metadata, !STRATA3 %in% c("POOL-FIN", "POOL-MUSCLE")) %>%
  dplyr::filter(!grepl(".*_REP", x = Genotype_names)) %>%
  dplyr::filter(!grepl(".*58-4-2.*_fin_.*", x = Genotype_names)) %>%
  dplyr::filter(!grepl(".*88-2.*_fin_.*", x = Genotype_names))

metadata$Maturity_full <- factor(metadata$Maturity_full, 
                                 levels = c("Immature","Resting/Developing",
                                            "Developed","Gravid","Spent"))
```

## Visual Sex

```{r}
summ <- dplyr::count(metadata, Sex, STRATA1, Year)
Sex.plot <- ggplot2::ggplot(summ, 
                            ggplot2::aes(x = STRATA1, y = n, fill = Sex)) + 
  ggplot2::geom_bar( stat = "identity") + 
  ggplot2::labs(y = "Number of fish") +
  ggplot2::facet_wrap(~Year) +
  ggplot2::theme(
    axis.text.x = ggplot2::element_text(face = "bold", color = "black", 
                                        size = 15, angle = 90, hjust = 1, vjust = 0),
    axis.text.y = ggplot2::element_text(size = 15),
    axis.title.y = ggplot2::element_text(size = 15),
    axis.title.x = ggplot2::element_blank(),
    panel.background = ggplot2::element_rect(fill = "white"),
    axis.line = ggplot2::element_line(colour = "black", linewidth = 0.5, 
                                      linetype = 1, lineend = "butt")) +
  ggplot2::scale_fill_manual(values = c("#E69F00", "#56B4E9","#999999"), 
                             na.value = "#999999")
print(Sex.plot)
ggplot2::ggsave(plot = Sex.plot, filename = paste0(figure.path,"0.summary_visual_Sex_NEW.png"),
                width = 30, height = 15, units = "cm")
```

## Total length

```{r}
length.plot <- ggplot2::ggplot(metadata, 
                               ggplot2::aes(x = STRATA1, y = Total_Length)) +
  ggplot2::geom_boxplot() +
  ggplot2::geom_jitter(ggplot2::aes(colour = Sex), alpha = 0.8, size = 2) +
  ggplot2::labs(y = "Total length (cm)") +
  ggplot2::ylim(c(0,210)) +
  ggplot2::scale_colour_manual(values = c("#E69F00", "#56B4E9","#999999"), 
                             na.value = "#999999") +
  ggplot2::facet_wrap(~Year) +
  ggplot2::theme(
    axis.title.x = ggplot2::element_blank(),
    axis.text.x = ggplot2::element_text(color = "black", 
                                        size = 15, angle = 90, hjust = 1, vjust = 0),
    axis.text.y = ggplot2::element_text(size = 15),
    axis.title.y = ggplot2::element_text(size = 15),
    panel.background = ggplot2::element_rect(fill = "white"),
    panel.grid.major = ggplot2::element_line(size = 0.5, linetype = 'solid',
                                             colour = "grey"), 
    panel.grid.minor = ggplot2::element_line(size = 0.1, linetype = 'solid',
                                             colour = "grey"),
    axis.line = ggplot2::element_line(colour = "black", size = 0.5, 
                                      linetype = 1, lineend = "butt")
  )
print(length.plot)
ggplot2::ggsave(plot = length.plot, filename = paste0(figure.path,"0.summary_TL_Sex_NEW.png"),
                width = 30, height = 15, units = "cm")


length.plot2 <- ggplot2::ggplot(metadata, 
                               ggplot2::aes(x = STRATA1, y = Total_Length)) +
  ggplot2::geom_boxplot() +
  ggplot2::geom_jitter(ggplot2::aes(colour = Maturity_full), alpha = 0.9, size = 2) +
  ggplot2::labs(y = "Total length (cm)") +
  ggplot2::ylim(c(0,210)) +
  ggplot2::scale_colour_manual(values = c("#adcf9f", "#78B7C5", "#08306B",
                                          "#E1AF00","#F21A00"), 
                             na.value = "#999999") +
  ggplot2::theme(
    axis.title.x = ggplot2::element_blank(),
    axis.text.x = ggplot2::element_text(color = "black", 
                                        size = 15, angle = 90, hjust = 1, vjust = 0),
    axis.text.y = ggplot2::element_text(size = 15),
    axis.title.y = ggplot2::element_text(size = 15),
    panel.background = ggplot2::element_rect(fill = "white"),
    panel.grid.major = ggplot2::element_line(size = 0.5, linetype = 'solid',
                                             colour = "grey"), 
    panel.grid.minor = ggplot2::element_line(size = 0.1, linetype = 'solid',
                                             colour = "grey"),
    axis.line = ggplot2::element_line(colour = "black", size = 0.5, 
                                      linetype = 1, lineend = "butt")
  )
print(length.plot2)
ggplot2::ggsave(plot = length.plot2, filename = paste0(figure.path,"0.summary_TL_mat_NEW.png"),
                width = 30, height = 15, units = "cm")
```


## Aproximate Length-at-age

Dunn, A., P. L. Horn., & S. M. Hanchet, 2006. Revised estimates of the biological parameters for Antarctic toothfish (Dissostichus mawsoni) in the Ross Sea. Document SCCAMLR-WG-FSA-SAM-06/08. Hobart, TAS: CCAMLR.

```{r, echo=FALSE, out.width = '100%'}
#Female
k1 <- 0.090
Linf1 <- 180.20
t01 <- 0.021
age1 <- c(t01, 0, seq(1,50))
length1 <- Linf1*(1 - exp(-k1 * (age1 - t01)))

#Male
k2 <- 0.093
Linf2 <- 169.07
t02 <- -0.256
age2 <- c(t02, 0, seq(1,50))
length2 <- Linf2*(1 - exp(-k2 * (age2 - t02)))

df <- data.frame(age1, length1, age2, length2)

plot <- ggplot2::ggplot(df) +
  ggplot2::geom_point(ggplot2::aes(x = age1, y = length1),
                      size = 4, colour = "#E69F00") +
  ggplot2::geom_point(ggplot2::aes(x = age2, y = length2),
                      size = 4, colour = "#56B4E9") +
  ggplot2::geom_hline(yintercept = Linf1, colour = "#E69F00") +
  ggplot2::geom_hline(yintercept = Linf2, colour = "#56B4E9") +
  ggplot2::geom_vline(xintercept = t01, colour = "#E69F00") +
  ggplot2::geom_vline(xintercept = t02, colour = "#56B4E9") +
  ggplot2::ylim(0,200) +
  ggplot2::xlab("Age (years)") +
  ggplot2::ylab("Length (mm)") +
  ggplot2::theme_minimal() +
  ggplot2::theme(axis.text = ggplot2::element_text(size = 5),
                 axis.title.x = ggplot2::element_text(size = 10),
                 axis.title.y = ggplot2::element_text(size = 10),
                 legend.text = ggplot2::element_text(size = 7))
print(plot)
ggplot2::ggsave(plot = plot, filename = paste0(figure.path,"0.Length_at_age_curve.png"),
                width = 30, height = 15, units = "cm", bg = "white")
```

### Catch date

```{r, echo=FALSE, out.width = '100%'}
metadata$Date <- as.Date(metadata$Date, "%d/%m/%Y")

date.plot <- ggplot2::ggplot(metadata, ggplot2::aes(x = STRATA1, y = Date)) +
  ggplot2::geom_boxplot() +
  ggplot2::geom_jitter(alpha = 0.1, height = 0) +
  ggplot2::labs(y = "Catch date") +
  ggplot2::theme(
    axis.title.x = ggplot2::element_blank(),
    axis.text.x = ggplot2::element_text(color = "black", size = 10, angle = 90, 
                                        hjust = 1, vjust = 0),
    panel.background = ggplot2::element_rect(fill = "white"),
    panel.grid.major = ggplot2::element_line(size = 0.5, linetype = 'solid',
                                             colour = "grey"),
    panel.grid.minor = ggplot2::element_line(size = 0.1, linetype = 'solid',
                                             colour = "grey"),
    axis.line = ggplot2::element_line(colour = "black", size = 0.5,
                                      linetype = 1, lineend = "butt")
  )
print(date.plot)

date2.plot <- ggplot2::ggplot(metadata, 
                              ggplot2::aes(x = STRATA1, y = Year)) +
  ggplot2::geom_boxplot() +
  ggplot2::geom_jitter(alpha = 0.1, height = 0) +
  ggplot2::labs(y = "Catch year") +
  ggplot2::theme(
    axis.title.x = ggplot2::element_blank(),
    axis.text.x = ggplot2::element_text(color = "black", 
                                        size = 10, angle = 90, 
                                        hjust = 1, vjust = 0),
    panel.background = ggplot2::element_rect(fill = "white"),
    panel.grid.major = ggplot2::element_line(size = 0.5, linetype = 'solid',
                                             colour = "grey"), 
    panel.grid.minor = ggplot2::element_line(size = 0.1, linetype = 'solid',
                                             colour = "grey"),
    axis.line = ggplot2::element_line(colour = "black", size = 0.5, 
                                      linetype = 1, lineend = "butt")
  )
print(date2.plot)

ggplot2::ggsave(plot = date2.plot, filename = paste0(figure.path,"0.Summary_Catch_date_NEW.png"),
                width = 30, height = 15, units = "cm", bg = "white")
```

### Age

Dunn, A., P. L. Horn., & S. M. Hanchet, 2006. Revised estimates of the biological parameters for Antarctic toothfish (Dissostichus mawsoni) in the Ross Sea. Document SCCAMLR-WG-FSA-SAM-06/08. Hobart, TAS: CCAMLR.

```{r, echo=FALSE, out.width = '100%'}
#Female
k1 <- 0.090
Linf1 <- 180.20
t01 <- 0.021

#Male
k2 <- 0.093
Linf2 <- 169.07
t02 <- -0.256


metadata[["Age"]] <- NA
if (class(metadata$Total_Length) == "character") {
  metadata$Total_Length <- as.numeric(metadata$Total_Length)
}
summary(metadata$Total_Length, na.rm = TRUE)
sum(metadata$Total_Length[metadata$Sex == "F"] > 180, na.rm = TRUE) ## 3
sum(metadata$Total_Length[metadata$Sex == "M"] > 169, na.rm = TRUE) ## 1

metadata$Sex[is.na(metadata$Sex)] <- "F" #only 1

metadata$Age[metadata$Sex == "F"] <- (1/k1)*log(Linf1/(Linf1 - metadata$Total_Length[metadata$Sex == "F"])) + t01
metadata$Age[metadata$Sex != "F"] <- (1/k2)*log(Linf2/(Linf2 - metadata$Total_Length[metadata$Sex != "F"])) + t02


metadata$Age[metadata$Age < 0 & !is.na(metadata$Age)] <- 0
age.plot <- ggplot2::ggplot(metadata, ggplot2::aes(x = STRATA1, y = Age)) +
  ggplot2::geom_boxplot() +
  ggplot2::geom_jitter(ggplot2::aes(colour = Maturity_full), alpha = 0.8, size = 2) +
  ggplot2::labs(y = "Age (years)") +
  ggplot2::ylim(c(0,50)) +
  ggplot2::scale_colour_manual(values = c("#adcf9f", "#78B7C5", "#08306B","#E1AF00","#F21A00"),
                             na.value = "#999999") +
  ggplot2::facet_wrap(~Year) +
  ggplot2::theme(
    axis.title.x = ggplot2::element_blank(),
    axis.text.x = ggplot2::element_text(color = "black", 
                                        size = 15, angle = 90, 
                                        hjust = 1, vjust = 0),
    axis.text.y = ggplot2::element_text(size = 15),
    axis.title.y = ggplot2::element_text(size = 15),
    panel.background = ggplot2::element_rect(fill = "white"),
    panel.grid.major = ggplot2::element_line(size = 0.5, linetype = 'solid',
                                             colour = "grey"), 
    panel.grid.minor = ggplot2::element_line(size = 0.1, linetype = 'solid',
                                             colour = "grey"),
    axis.line = ggplot2::element_line(colour = "black", size = 0.5, 
                                      linetype = 1, lineend = "butt")
  )
print(age.plot)
ggplot2::ggsave(plot = age.plot, filename = paste0(figure.path,"0.summary_age_NEW.png"),
                width = 30, height = 15, units = "cm")
```

### Cohort

```{r, echo=FALSE, out.width = '100%'}
metadata[["Cohort"]] <- metadata$Date - metadata$Age*365

cohort.plot <- ggplot2::ggplot(metadata, ggplot2::aes(x = STRATA1, y = Cohort)) +
  ggplot2::geom_boxplot() +
  ggplot2::geom_jitter(ggplot2::aes(colour = Sex ), alpha = 0.8) +
  ggplot2::scale_colour_manual(values = c("#E69F00", "#56B4E9","#999999"), 
                             na.value = "#999999") +
  ggplot2::labs(y = "Year of birth") +
  ggplot2::facet_wrap(~Year) +
  ggplot2::theme(
    axis.title.x = ggplot2::element_blank(),
        axis.text.x = ggplot2::element_text(color = "black", 
                                            size = 15, angle = 90, 
                                            hjust = 1, vjust = 0),
    axis.text.y = ggplot2::element_text(size = 15),
    axis.title.y = ggplot2::element_text(size = 15),
        panel.background = ggplot2::element_rect(fill = "white"),
        panel.grid.major = ggplot2::element_line(size = 0.5, linetype = 'solid',
                                                 colour = "grey"), 
        panel.grid.minor = ggplot2::element_line(size = 0.1, linetype = 'solid',
                                                 colour = "grey"),
        axis.line = ggplot2::element_line(colour = "black", size = 0.5, 
                                          linetype = 1, lineend = "butt")
  )
print(cohort.plot)


years <- lubridate::round_date(metadata$Cohort, unit = "year")
metadata$Cohort2 <- as.integer(sub("-.*$", "", years))



cohort2.plot <- ggplot2::ggplot(metadata, ggplot2::aes(x = STRATA1, y = Cohort2)) +
  ggplot2::geom_boxplot() +
  ggplot2::geom_jitter(ggplot2::aes(colour = Sex ), alpha = 0.8) +
  ggplot2::scale_colour_manual(values = c("#E69F00", "#56B4E9","#999999"), 
                             na.value = "#999999") +  
  ggplot2::labs(y = "Year of birth") +
  ggplot2::facet_wrap(~Year) +
  ggplot2::theme(
    axis.title.x = ggplot2::element_blank(),
        axis.text.x = ggplot2::element_text(color = "black", 
                                            size = 15, angle = 90, 
                                            hjust = 1, vjust = 0),
    axis.text.y = ggplot2::element_text(size = 15),
    axis.title.y = ggplot2::element_text(size = 15),
        panel.background = ggplot2::element_rect(fill = "white"),
        panel.grid.major = ggplot2::element_line(size = 0.5, linetype = 'solid',
                                                 colour = "grey"), 
        panel.grid.minor = ggplot2::element_line(size = 0.1, linetype = 'solid',
                                                 colour = "grey"),
        axis.line = ggplot2::element_line(colour = "black", size = 0.5, 
                                          linetype = 1, lineend = "butt")
  )
print(cohort2.plot)
ggplot2::ggsave(plot = cohort2.plot, filename = paste0(figure.path,"0.summary_cohort_NEW.png"),
                width = 30, height = 15, units = "cm")
```

## Maturity

```{r}
cohort2.plot <- ggplot2::ggplot(metadata) +
  ggplot2::geom_bar(ggplot2::aes(x = STRATA1, fill = Maturity_full),
                    stat = "count") +
  ggplot2::scale_fill_manual(values = c("#adcf9f", "#78B7C5", "#08306B","#E1AF00","#F21A00"),
                             na.value = "#999999") +
  ggplot2::labs(y = "Frequency of Maturity") +
  ggplot2::facet_wrap(~Year) +
  ggplot2::theme(
    axis.title.x = ggplot2::element_blank(),
        axis.text.x = ggplot2::element_text(color = "black", 
                                            size = 15, angle = 90, 
                                            hjust = 1, vjust = 0),
    axis.text.y = ggplot2::element_text(size = 15),
    axis.title.y = ggplot2::element_text(size = 15),
        panel.background = ggplot2::element_rect(fill = "white"),
        panel.grid.major = ggplot2::element_line(size = 0.5, linetype = 'solid',
                                                 colour = "grey"), 
        panel.grid.minor = ggplot2::element_line(size = 0.1, linetype = 'solid',
                                                 colour = "grey"),
        axis.line = ggplot2::element_line(colour = "black", size = 0.5, 
                                          linetype = 1, lineend = "butt")
  )
print(cohort2.plot)
ggplot2::ggsave(plot = cohort2.plot, filename = paste0(figure.path,"0.summary_maturity_NEW.png"),
                width = 30, height = 15, units = "cm")
```

## Load metdata

```{r, eval=FALSE}
# save(metadata, file = paste0(rdata.path,"Toothfish_metadata_unique_fish.Rdata"))
# save(metadata, file = paste0(rdata.path,"Toothfish_metadata.Rdata"))

print(load(paste0(rdata.path,"Toothfish_metadata_unique_fish.Rdata"))) #UNIQUE INDIVIDUALS
print(load(paste0(rdata.path,"Toothfish_metadata.Rdata"))) #ALL fish
```




## Donut plot for map
```{r}
print(load(paste0(rdata.path,"Toothfish_metadata_unique_fish.Rdata"))) #UNIQUE INDIVIDUALS

colours.region <- c("#E1AF00","#52B334","#4d6b40", "seagreen","#F21A00", "#78B7C5", "#08306B")
names(colours.region) <- c("48.2-2017", "58.4.1-2018", "58.4.2-2018", "58.4.2-2024",
                        "58.5.2-2017","88.1-2018", "88.2-2024")
colors.sex <- c("#E69F00", "#56B4E9")
names(colors.sex) <- c("Female", "Male")
colors.Mat <- c("#adcf9f", "#78B7C5", "#08306B","#E1AF00","#F21A00")
names(colors.Mat) <- c("Immature","Resting/Developing", "Developed", 
                       "Gravid", "Spent")
colours.comb <- c(colours.region, colors.sex, colors.Mat)
metadata2 <- metadata[!(metadata$STRATA1 == "58.4.2" & metadata$Year == "2024" &
                          metadata$Longitude < 70),]
data <- data.frame("Region" = paste0(metadata2$STRATA1,"-", metadata2$Year),
                   "Sex" = metadata2$Sex,
                   "Maturity_full" = metadata2$Maturity_full)

for (reg in unique(data$Region)) {
  print(reg)
  samp.size = length(data$Region[data$Region == reg])
  df.reg <- data.frame(Level = "Region", Value = reg, prct = 100)
  df.sex <- data %>%
    dplyr::filter(Region == reg) %>%
    dplyr::group_by(Sex) %>%
    dplyr::summarise(Sex.count = n()) %>%
    dplyr::mutate(Level = "Sex") %>%
    dplyr::mutate(Value = Sex) %>%
    dplyr::mutate(prct = round(100*Sex.count/sum(Sex.count),2)) %>%
    dplyr::select(-c("Sex", "Sex.count"))
  df.sex$Value[df.sex$Value == "F"] <- "Female"
  df.sex$Value[df.sex$Value == "M"] <- "Male"
  df.mat <- data %>%
    dplyr::filter(Region == reg) %>%
    dplyr::group_by(Maturity_full) %>%
    dplyr::summarise(Mat.count = n()) %>%
    dplyr::mutate(Level = "Maturity") %>%
    dplyr::mutate(Value = Maturity_full) %>%
    dplyr::mutate(prct = round(100*Mat.count/sum(Mat.count),2)) %>%
    dplyr::select(-c("Maturity_full", "Mat.count"))
  df <- rbind(df.reg, df.sex, df.mat)
  df$Level <- factor(df$Level, levels = c("Region", "Sex", "Maturity"))
  df$Value <- factor(df$Value, levels = names(c(colours.region, colors.sex, colors.Mat)))
  donut.plot <- ggplot2::ggplot(df, 
                                ggplot2::aes(x = Level, y = prct, fill = Value)) +
    ggplot2::geom_bar(stat = "identity") +
    ggplot2::geom_text(ggplot2::aes(label = paste0("n= ",samp.size)), 
                       x = 0.5, y = 0, fontface = "bold") +
    ggplot2::coord_polar(theta = "y") +
    ggplot2::scale_fill_manual(values = colours.comb,
                               na.value = "#999999") +
    ggplot2::theme( panel.background = 
                      ggplot2::element_rect(fill = 'transparent')) +
    ggplot2::theme_void()
    
  print(donut.plot)
  ggplot2::ggsave(donut.plot, bg = NULL,
                  filename = paste0(figure.path,
                                    "0.metadata_donutplot_for_map_",reg,".png"
                  ), 
                  width = 10, height = 10, units = "cm")
}
```



------------------------------------------------------------------------

<br> \pagebreak

<P style="page-break-before: always">


# 2_Genetic Data - SNP

## a. Individual samples







### Read data
```{r, eval=FALSE}
# vcf.file <- "/datasets/work/ncmi-toa-rrbs/work/1.Toothfish/GENOME/Combined_SNP/Toothfish_Nuclear_SNP_TOA_GLnexus_gatk_biallelic.vcf.gz"
# vcf.file <- "/datasets/work/ncmi-toa-rrbs/work/1.Toothfish/GENOME/Combined_SNP/Toothfish_Nuclear_SNP_TOA_revelio_biallelic.vcf.gz"
vcf.file <- "/datasets/work/ncmi-toa-rrbs/work/1.Toothfish/Toothfish_RRBS_array2018_SE/gatk/Toothfish_Nuclear_SNP_2018_TOA_revelio_biallelic.vcf.gz"
metafile <- "/datasets/work/ncmi-toa-rrbs/work/1.Toothfish/0.R-analysis/1.Toothfish_ALL_RRBS_metadata.csv"

TOA.all.gl <- dartR.base::gl.read.vcf(vcffile = vcf.file,
                                      ind.metafile = metafile,
                                      mode = "genotype",
                                      verbose = 5)
all.vcfr <- vcfR::read.vcfR(vcf.file, verbose = TRUE )

# Toothfish_Nuclear_SNP_TOA_GLnexus_gatk_biallelic.vcf.gz
# 304 samples
# 314 CHROMs
# 259,196 variants
# Object size: 709.9 Mb
# 0 percent missing data
 
# Toothfish_Nuclear_SNP_2018_TOA_revelio_biallelic.vcf.gz
# 144 samples
# 294 CHROMs
# 190,454 variants
# Object size: 292.8 Mb
# 0 percent missing data

TOA.all.gl[TOA.all.gl == 0,]

TOA.all.gl@other$loc.metrics$loc.names <- TOA.all.gl$loc.names
TOA.all.gl@other$loc.metrics$DP <- as.integer(TOA.all.gl@other$loc.metrics$DP)

TOA.all.gl@other$loc.metrics$rdepth <- 
  as.integer(TOA.all.gl@other$loc.metrics$DP)/
  adegenet::nInd(TOA.all.gl)

### ISSUE - MISSING DATA WAS SET TO 0/0 GT ### Here we adjust those based on coverage = 0
dp <- t(vcfR::extract.gt(all.vcfr, element = 'DP', as.numeric = TRUE))
mat <- as.matrix(TOA.all.gl)
sum(colnames(dp) == colnames(mat)) == length(colnames(mat))
sum(rownames(dp) == rownames(mat)) == length(rownames(mat))
mat[dp < 1] <- NA



gl3 <- new("dartR", gen = mat, ploidy = 2, ind.names = TOA.all.gl$ind.names, 
        loc.names = TOA.all.gl$loc.names, loc.all = TOA.all.gl$loc.all, 
        position = TOA.all.gl$position, parallel = FALSE)
gl3$chromosome <- TOA.all.gl$chromosome
gl3$pop <- TOA.all.gl$pop
gl3@other$loc.metrics <- TOA.all.gl@other$loc.metrics
gl3@other$ind.metrics <- TOA.all.gl@other$ind.metrics
gl3@other$loc.metrics.flags <- TOA.all.gl@other$loc.metrics.flags
gl3@other$verbose <- 5
gl3@other$history <- TOA.all.gl@other$history
gl3 <- dartR.base::gl.recalc.metrics(gl3)
TOA.all.gl <- gl3

# 144 genotypes,  190,454 SNPs , size: 135.9 Mb
# missing data: 12797792 (=46.66 %) scored as NA

# save(TOA.all.gl,all.vcfr, file = paste0(rdata.path,"TOA_All_SNP_genlight_raw.Rdata"))
save(TOA.all.gl,all.vcfr, file = paste0(rdata.path,"TOA_2018_SNP_genlight_raw.Rdata"))

```


### Raw coverage

```{r}
gt <- vcfR::extract.gt(all.vcfr)
# genotypes to be delimited with a '|' when they are phased and a '/' when unphased
sum(gt == "0|1", na.rm = T) #11598 HOM #2018: 14102
sum(gt == "1|1", na.rm = T) #2862 HET #2018: 4146

sum(gt == "0/0", na.rm = T) #78326285 #2018: 27080426
sum(gt == "1/1", na.rm = T) #190486 HET #2018: 106285
sum(gt == "0/1", na.rm = T) #249217 #2018: 204531

sum(gt == "0/2", na.rm = T) #0 #2018: 0
sum(is.na(gt)) #690 #2018: 1011
sum(vcfR::is.indel(all.vcfr)) #0 #2018: 0


gt2 <- vcfR::extract.gt(all.vcfr, element = 'GT', as.numeric = TRUE)
sum(gt2 == 1, na.rm = T) #193348 #2018: 110431


dp <- vcfR::extract.gt(all.vcfr, element = 'DP', as.numeric = TRUE)
dpf <- data.frame(dp, locus = rownames(dp))
dpf <- tidyr::gather(dpf, key = individual, 
                           value = Depth, -locus,
                           factor_key = TRUE) %>%
  dplyr::group_by(individual) %>%
  dplyr::mutate(NA_cnt = sum(is.na(Depth))/length(unique(locus))) %>%
  dplyr::ungroup()

dpf.NA <- dpf[!duplicated(dpf$individual),]
dpf.NA$NA_cnt <- round(dpf.NA$NA_cnt, 3)
dpf <- dpf[ dpf$Depth > 0 & !is.na(dpf$Depth),]
max(dpf$Depth, na.rm = T) #268

myPlot <- ggplot2::ggplot(dpf, ggplot2::aes(x = individual, y = Depth)) + 
  ggplot2::geom_boxplot() +
  ggplot2::geom_text(data = dpf.NA,ggplot2::aes(x = individual, y = 310, label = NA_cnt), 
                     size = 2, angle = 90) +
  ggplot2::theme_bw() + 
  ggplot2::theme(axis.title.x = ggplot2::element_blank(), 
                  axis.text.x = ggplot2::element_text(size = 7, angle = 90, hjust = 1)) +
  ggplot2::scale_y_continuous(trans = scales::log2_trans(), 
                  breaks = c(1, 10, 50, 300),
                  minor_breaks = c(1:10, 2:5*10, 1:3*100))
print(myPlot)
ggplot2::ggsave(myPlot, bg = "white",
                filename = paste0(figure.path,
                                  # "1A.TOA_vcfR_individual_coverage_149770_SNPs.png"
                                  "1A.TOA_vcfR_individual_coverage_2018_190454_SNPs.png"
                                  ), 
                width = 30, height = 15, units = "cm")


NA.SNP <- rowSums(is.na(dp))/ncol(dp)
hist(NA.SNP, breaks = 100)
NA.ind <- colSums(is.na(dp))/nrow(dp)
hist(NA.ind, xlim = c(0,1))


dp.SNP <- rowMeans(dp, na.rm = TRUE)
sum(dp.SNP > 5 & !is.na(dp.SNP)) #71831 #2018: 58452


subset <- grepl(x = colnames(dp),pattern = "48-2")
subset <- grepl(x = colnames(dp),pattern = "58.5")
subset <- grepl(x = colnames(dp),pattern = "88")
dp.SNP <- rowMeans(dp[,subset], na.rm = TRUE)
# hist(dp.SNP, breaks = 150)
hist(dp.SNP[dp.SNP < 10], breaks = 10)
sum(dp.SNP > 5 & !is.na(dp.SNP)) #71831

dp.ind <- colMeans(dp, na.rm = TRUE)
# dp.ind <- colSums(dp, na.rm = TRUE)
hist(dp.ind)
colnames(dp)[dp.ind < 1]
 # "TOA_58-4-2_B24_AE100883_muscle_F" "TOA_88-2_P79_AD294619_fin_F" 
```

### Sex-linked markers

```{r, eval=FALSE}
print(load(paste0(rdata.path,"TOA_All_SNP_genlight_raw.Rdata")))
print(load(paste0(rdata.path,"TOA_2018_SNP_genlight_raw.Rdata")))

TOA.all.gl@other$ind.metrics$id <- as.character(TOA.all.gl@other$ind.metrics$Genotype_names)
TOA.all.gl@other$ind.metrics$sex <- as.character(TOA.all.gl@other$ind.metrics$Sex)
TOA.all.gl@other$ind.metrics$sex[TOA.all.gl@other$ind.metrics$sex == "U"] <- NA
TOA.all.gl@other$ind.metrics <- as.data.frame(TOA.all.gl@other$ind.metrics)

res <- dartR.sexlinked::gl.report.sexlinked(x = TOA.all.gl, system = "xy", 
                                            plot.display = TRUE, ncores = 10,
                                            verbose = 5)

# 168 females and 136 males.
# **FINISHED** Total of analysed loci: 259196.
# Found 0 sex-linked loci:
#    0 Y-linked loci
#    0 sex-biased loci
#    0 X-linked loci
#    0 gametologs


#81 females and 63 males.
#    0 Y-linked loci
#    0 sex-biased loci
#    0 X-linked loci
#    0 gametologs.
# And 190454 autosomal loci.

# sexID <- dartR.sexlinked::gl.infer.sex(gl_sexlinked = res, system = "xy", seed = 124)
# identical(sexID$id,TOA.all.gl@other$ind.metrics$id)
# sexID$visual.sex <- TOA.all.gl@other$ind.metrics$sex
# readr::write_csv(sexID, file = "0_genetic_sexID_SNP_ALL.csv")
# save(res, sexID, file = paste0(rdata.path, "SNP_sex-linked_markers_ALL.Rdata"))
```


```{r}
print(load(paste0(rdata.path, "SNP_sex-linked_markers_ALL.Rdata")))
res$results.table[res$results.table$y.linked == TRUE, ]
res$results.table[res$results.table$x.linked == TRUE, ]
res$results.table[res$results.table$xy.gametolog == TRUE, ]
res$results.table[res$results.table$sex.biased == TRUE, ]

print(sexID)
```



### Filter data

```{r}
# print(load(paste0(rdata.path,"TOA_All_SNP_genlight_raw.Rdata")))
print(load(paste0(rdata.path,"TOA_2018_SNP_genlight_raw.Rdata")))

TOA.all.gl <- dartR.base::gl.filter.allna(TOA.all.gl, verbose = 5)
dartR.base::gl.report.monomorphs(TOA.all.gl, verbose = 5)
TOA.all.gl <- dartR.base::gl.filter.monomorphs(TOA.all.gl, verbose = 5) #254506 binary SNPs #180881 


dartR.base::gl.report.rdepth(TOA.all.gl, verbose = 5)
dev.print(file = paste0(figure.path,
                        # "TOA_SNP_read_depth_filter.png"
                        "TOA_2018_SNP_read_depth_filter.png"
                        ),
          device = png, res = 300,width = 30,height = 30,units = "cm")
TOA.all.gl <- dartR.base::gl.filter.rdepth(TOA.all.gl, lower = 5, upper = 40, verbose = 5) #68626 #52048 

dartR.base::gl.report.callrate(TOA.all.gl, method = "loc")
TOA.all.gl <- dartR.base::gl.filter.callrate(TOA.all.gl, method = "loc", 
                                                  threshold = 0.8, verbose = 5) #58105 binary SNPs #49643
dev.print(file = paste0(figure.path,
                        # "TOA_SNP_CallRate_filter.png"
                        "TOA_2018_SNP_CallRate_filter.png"
                        ),
          device = png, res = 300,width = 30,height = 30,units = "cm")

dartR.base::gl.report.callrate(TOA.all.gl, method = "ind")
TOA.all.gl <- dartR.base::gl.filter.callrate(TOA.all.gl, method = "ind", 
                                                  threshold = 0.95, #0.85
                                                  verbose = 5) #286 fish #131 
dev.print(file = paste0(figure.path,
                        # "TOA_IND_CallRate_filter.png"
                        "TOA_2018_IND_CallRate_filter.png"
                        ),
          device = png, res = 300,width = 30,height = 30,units = "cm")

# rm <- c("TOA_48-2_005_UK046_REP", "TOA_48-2_007", "TOA_48-2_024", 
#         "TOA_58-4-1_289_AN177617_REP", "TOA_58-4-2_B18_AE100308_muscle_F",
#         "TOA_58-4-2_B24_AE100883_muscle_F", "TOA_58-4-2_B28_AE100885_muscle_M",
#         "TOA_58-4-2_B30_AE100886_muscle_M", "TOA_58-4-2_B34_AE102049_muscle_F",
#         "TOA_58-4-2_B36_AE102050_muscle_M", "TOA_58-4-2_B38_AE102052_muscle_M",
#         "TOA_58-4-2_B42_AE102105_muscle_F", "TOA_58-4-2_B48_AE102108_muscle_M",
#         "TOA_58-4-2_B58_AE104944_muscle_F", "TOA_88-2_P24_AD292328_muscle_F", 
#         "TOA_88-2_P67_AD294613_fin_F", "TOA_88-2_P69_AD294614_fin_F", 
#         "TOA_88-2_P79_AD294619_fin_F")

# rm <- c(TOA_48-2_007[48.2-2017], TOA_48-2_024[48.2-2017], TOA_48-2_028[48.2-2017], TOA_58-4-1_291[58.4.1-2018], TOA_58-4-2_088[58.4.2-2018], TOA_58-4-2_095[58.4.2-2018], TOA_58-5-2_162[58.5.2-2017], TOA_88-1_224[88.1-2018], TOA_88-1_234[88.1-2018], TOA_88-1_237[88.1-2018], TOA_88-1_244[88.1-2018], TOA_88-1_247[88.1-2018], TOA_88-1_251[88.1-2018])


dartR.base::gl.report.callrate(TOA.all.gl, method = "loc")

het.res <- dartR.base::gl.report.heterozygosity(TOA.all.gl, method = "ind")
dev.print(file = paste0(figure.path,
                        # "TOA_IND_heterozygosity_filter.png"
                        "TOA_2018_IND_heterozygosity_filter.png"
                        ),
          device = png, res = 300,width = 30,height = 30,units = "cm")
het.res[order(het.res$Ho, decreasing = TRUE),]
dartR.base::gl.report.heterozygosity(TOA.all.gl, method = "pop")
dev.print(file = paste0(figure.path,
                        # "TOA_IND_heterozygosity_POP.png"
                        "TOA_2018_IND_heterozygosity_POP.png"
                        ),
          device = png, res = 300,width = 30,height = 30,units = "cm")

TOA.all.gl <- dartR.base::gl.filter.heterozygosity(TOA.all.gl, t.lower = 0, t.upper = 0.018,  verbose = 5)
# rm <- c(TOA_48-2_005[48.2-2017], TOA_48-2_031[48.2-2017])


TOA.all.gl <- dartR.base::gl.filter.allna(TOA.all.gl, verbose = 5)
dartR.base::gl.report.monomorphs(TOA.all.gl, verbose = 5)
TOA.all.gl <- dartR.base::gl.filter.monomorphs(TOA.all.gl, verbose = 5) #56402 binary SNPs #45866

dartR.base::gl.report.maf(TOA.all.gl, verbose = 5)
# TOA.all.gl <- dartR.base::gl.filter.maf(TOA.all.gl, threshold = 0.01, 
#                                      plot.colors = c("#2171B5", "#6BAED6"),
#                                      verbose = 5)

res_dup <- dartR.base::gl.report.replicates(TOA.all.gl,
                                            loc_threshold = 1000, perc_geno = 0.85)
res_dup$table.rep[order(res_dup$table.rep$perc, decreasing = TRUE),]

# save(TOA.all.gl, file = paste0(rdata.path,"TOA_All_SNP_genlight_filtered.Rdata"))
# save(TOA.all.gl, file = paste0(rdata.path,"TOA_2018_SNP_genlight_filtered.Rdata"))
```


### Load data

```{r}
# print(load(paste0(rdata.path,"TOA_All_SNP_genlight_filtered.Rdata")))
print(load(paste0(rdata.path,"TOA_2018_SNP_genlight_filtered.Rdata")))

table(TOA.all.gl$pop)
adegenet::nLoc(TOA.all.gl)
```

### Genetic differentiation

```{r}
# print(load(paste0(rdata.path,"TOA_All_SNP_genlight_filtered.Rdata")))
# print(load(paste0(rdata.path,"TOA_2018_SNP_genlight_filtered.Rdata")))
print(load(paste0(rdata.path,"TOA_2018_SNP_genlight_raw.Rdata")))

df <- as.data.frame(TOA.all.gl)
df[df == "0"] <- "11"
df[df == "1"] <- "12"
df[df == "2"] <- "22"
gen.gi <- adegenet::df2genind(df, ploidy = 2, ncode = 2, pop = TOA.all.gl$pop)

fst.basic <- hierfstat::basic.stats(gen.gi)

# fst.CI <- hierfstat::boot.ppfst(gen.gi, nboot = 100, quant = c(0.025,
#                                                                0.975), 
#                                 diploid = TRUE)

# df <- data.frame(fst.hierfstat, fst.CI.low = fst.CI$fis.ci[,1], fst.CI.high = fst.CI$fis.ci[, 2])
# df <- as.data.frame(t(df))
# save(fst.basic, df,fst.CI, file = paste0(rdata.path, 
                                         # "SNP_genetic_differentiation.Rdata"
                                         # "SNP_genetic_differentiation_2018.Rdata"
                                         # ))
save(fst.basic, df, file = paste0(rdata.path, 
                                         "SNP_genetic_differentiation_2018_raw.Rdata"
                                         ))
```


```{r}
print(load(paste0(rdata.path, "SNP_genetic_differentiation_2018.Rdata")))
fst.CI$ll
fst.CI$ul
```


### PCA

```{r, eval=FALSE}
pca.all <- adegenet::glPca(TOA.all.gl, nf = 4,
                           parallel = TRUE,
                           n.cores = parallel::detectCores())

save(pca.all,
     file = paste0(rdata.path,
                   # "PCA_ALL_objects2.Rdata"
                   "PCA_2018_objects.Rdata"
                   ))
```


```{r, warning=FALSE}
# print(load(paste0(rdata.path,"PCA_ALL_objects2.Rdata")))
print(load(paste0(rdata.path,"PCA_2018_objects.Rdata")))

var_frac <- pca.all$eig/sum(pca.all$eig)
print(paste0("Total variance explained: ",signif(sum(var_frac[1:4]) * 100, 3), "%")) 
barplot(var_frac[1:20], main = "variance explained", 
        col = heat.colors(length(var_frac[1:20])))

pop <- TOA.all.gl$pop
# title.plot <- "Toothfish SNP (144,655 markers)"
title.plot <- "Toothfish SNP (45,866 markers)"


# colours5 <- adegenet::funky(5)

data <- data.frame(individuals = rownames(pca.all$scores),
                    pop = pop, pca.all$scores)
data <- merge(data, aggregate(cbind(mean.x = PC1, mean.y = PC2) ~
                                pop, data, mean), by = "pop")

pca.plot <- ggplot2::ggplot(data, ggplot2::aes(PC1, PC2, color = pop)) +
  ggplot2::geom_point(size = 4) + 
  ggplot2::geom_point(ggplot2::aes(x = mean.x,
                                   y = mean.y, color = pop),
                      size = 7,
                      shape = 17) +
  ggplot2::scale_colour_manual(values = colours5) +
  ggplot2::scale_fill_manual(values = colours5) + 
  ggplot2::geom_hline(yintercept = 0) +
  ggplot2::geom_vline(xintercept = 0) + 
  ggplot2::labs(
    subtitle = paste0("Total variance explained: ",
                      signif(sum(var_frac[1:4]) * 100, 3), "%"),
    y = paste0("PC2; ", signif(var_frac[2] * 100, digits = 2), " %"),
    x = paste0("PC1; ", signif(var_frac[1] * 100, digits = 2), " %"),
    title = title.plot,
    caption = "") + 
  ggplot2::theme_minimal() +
  ggplot2::theme(
    axis.text = ggplot2::element_text(size = 15),
    axis.title.x = ggplot2::element_text(size = 20),
    axis.title.y = ggplot2::element_text(size = 20),
    legend.text = ggplot2::element_text(size = 15)
  )

print(pca.plot)
ggplot2::ggsave(pca.plot, bg = "white",
                filename = paste0(figure.path,
                                  # "2A.TOA_PCA_ALL_144655_SNPs_PC12.png"
                                  "2A.TOA_PCA_2018_45866_SNPs_PC12.png"
                                  ), 
                width = 30, height = 15, units = "cm")


# data$individuals[data$PC1 < -10] #All samples
# "TOA_48-2_020" "TOA_88-1_196"
# data$individuals[data$PC1 > 10] #All samples
# "TOA_48-2_024" #different species - mtDNA

# data$individuals[data$PC1 > 10] # "TOA_48-2_005" # high het
# data$individuals[data$PC2 > 5] # "TOA_48-2_031" # high het


data <- data.frame(individuals = rownames(pca.all$scores),
                    pop = pop, pca.all$scores)
data <- merge(data, aggregate(cbind(mean.x = PC2, mean.y = PC3) ~
                                pop, data, mean), by = "pop")

pca.plot <- ggplot2::ggplot(data, ggplot2::aes(PC2, PC3, color = pop)) +
  ggplot2::geom_point(size = 4) + 
  ggplot2::geom_point(ggplot2::aes(x = mean.x,
                                   y = mean.y, color = pop),
                      size = 7,
                      shape = 17) +
  ggplot2::scale_colour_manual(values = colours5) +
  ggplot2::scale_fill_manual(values = colours5) + 
  ggplot2::geom_hline(yintercept = 0) +
  ggplot2::geom_vline(xintercept = 0) + 
  ggplot2::labs(
    subtitle = paste0("Total variance explained: ",
                      signif(sum(var_frac[1:4]) * 100, 3), "%"),
    y = paste0("PC3; ", signif(var_frac[3] * 100, digits = 2), " %"),
    x = paste0("PC2; ", signif(var_frac[2] * 100, digits = 2), " %"),
    title = title.plot,
    caption = "") + 
  ggplot2::theme_minimal() +
  ggplot2::theme(
    axis.text = ggplot2::element_text(size = 15),
    axis.title.x = ggplot2::element_text(size = 20),
    axis.title.y = ggplot2::element_text(size = 20),
    legend.text = ggplot2::element_text(size = 15)
  )

print(pca.plot)
ggplot2::ggsave(pca.plot, bg = "white",
                filename = paste0(figure.path,
                                  # "2B.TOA_PCA_ALL_144655_SNPs_PC23.png"
                                  "2B.TOA_PCA_2018_45866_SNPs_PC23.png"
                                  ), 
                width = 30, height = 15, units = "cm")

# data$individuals[data$PC3 < -5] #All samples
# c("TOA_48-2_020", "TOA_48-2_024" ,"TOA_88-1_196")


data <- data.frame(individuals = rownames(pca.all$scores),
                    pop = pop, pca.all$scores)
data <- merge(data, aggregate(cbind(mean.x = PC1, mean.y = PC3) ~
                                pop, data, mean), by = "pop")

pca.plot <- ggplot2::ggplot(data, ggplot2::aes(PC1, PC3, color = pop)) +
  ggplot2::geom_point(size = 4) + 
  ggplot2::geom_point(ggplot2::aes(x = mean.x,
                                   y = mean.y, color = pop),
                      size = 7,
                      shape = 17) +
  ggplot2::scale_colour_manual(values = colours5) +
  ggplot2::scale_fill_manual(values = colours5) + 
  ggplot2::geom_hline(yintercept = 0) +
  ggplot2::geom_vline(xintercept = 0) + 
  ggplot2::labs(
    subtitle = paste0("Total variance explained: ",
                      signif(sum(var_frac[1:4]) * 100, 3), "%"),
    y = paste0("PC3; ", signif(var_frac[3] * 100, digits = 2), " %"),
    x = paste0("PC1; ", signif(var_frac[1] * 100, digits = 2), " %"),
    title = title.plot,
    caption = "") + 
  ggplot2::theme_minimal() +
  ggplot2::theme(
    axis.text = ggplot2::element_text(size = 15),
    axis.title.x = ggplot2::element_text(size = 20),
    axis.title.y = ggplot2::element_text(size = 20),
    legend.text = ggplot2::element_text(size = 15)
  )

print(pca.plot)
ggplot2::ggsave(pca.plot, bg = "white",
                filename = paste0(figure.path,
                                  # "2C.TOA_PCA_ALL_144655_SNPs_PC13.png"
                                  "2C.TOA_PCA_2018_45866_SNPs_PC13.png"
                                  ), 
                width = 30, height = 15, units = "cm")
```

### RDA

```{r}
gen <- as.data.frame(TOA.all.gl)
gen[1:10,1:10]
dim(gen)

sum(is.na(gen)) #1221754
gen.imp <- apply(gen, 2, function(x) replace(x, is.na(x), as.numeric(names(which.max(table(x))))))
sum(is.na(gen.imp)) # No NAs
gen.imp[1:10,1:10]

pred.pca <- rda(pred, scale=T)
summary(pred.pca)$cont
screeplot(pred.pca, main = "Screeplot: Eigenvalues of TOA Predictor Variables")
round(scores(pred.pca, choices=1:8, display="species", scaling=0), digits=3)




print(load(paste0(rdata.path,"Toothfish_metadata.Rdata")))

predictor <- metadata # define the predictor input file here
predictor <- predictor[predictor$File_names %in% TOA.all.gl$ind.names,]
predictor$SAMPLE_ID = as.character(predictor$File_names)
predictor$region = as.character(predictor$pop)
predictor$Sex = as.character(predictor$Sex)
predictor$Tissue = as.character(predictor$Tissue_type)
predictor$mat = as.character(predictor$Maturity)
predictor$mat[is.na(predictor$mat)] <- "2"
predictor$TL = as.numeric(predictor$Total_Length)
predictor$wt = as.numeric(predictor$`Weight_(kg)`)
predictor$dpth = as.numeric(predictor$Depth)
predictor$Age = as.numeric(predictor$Age)
predictor$Age[is.na(predictor$Age)] <- as.numeric(max(predictor$Age, na.rm = TRUE))
predictor$Cohort2 = as.integer(predictor$Cohort2)
predictor$Cohort2[is.na(predictor$Cohort2)] <- as.integer(min(predictor$Cohort2, na.rm = TRUE))



## Run RDA
#Test the country using growing condition as a Condition () (covariate):
SNP_rda = vegan::rda(gen.imp ~ region + Condition(Sex), 
                      data = predictor, scale = F) #I remove scale since I performed Hellinger transformation
SNP_rda = vegan::rda(gen.imp ~ region + Condition(Sex) + 
                        Condition(mat)+ Condition(TL) + Condition(wt) +
                        Condition(dpth) + Condition(Age) + Condition(Cohort2), 
                      na.action = "na.omit",
                      data = predictor, scale = TURE)


SNP_rda$pCCA
SNP_rda$CCA$wa

summary(SNP_rda) #Constrained Proportion: variance of Y explained by X, Unconstrained Proportion: unexplained variance in Y
# sink(paste0(rdata.path,"anova_rda.txt")) #with this you can save the output of the anova in a txt file, define the output directory
RsquareAdj(SNP_rda)
summary(eigenvals(SNP_rda, model = "constrained"))



SNP.aov <- anova(SNP_rda, test = "Chi", permutations = 999, parallel = 20) #run Anova on the result to see the significance level
print(SNP.aov)

# meth.aov.cca <- anova.cca(meth_rda, test = "Chi", permutations = 99)
SNP.aov.cca <- anova.cca(SNP_rda, permutations = 999, parallel = 20) # default is permutation=999
print(SNP.aov.cca)
# Model: rda(formula = gen.imp ~ region + Condition(Sex), data = predictor, scale = F)
#           Df Variance      F Pr(>F)    
# Model      4    78.52 1.0886  0.001 ***
# Residual 129  2326.04  


vegan::vif.cca(SNP_rda) # gives the variance inflation factors for each constraint

# screeplot(SNP_rda)
screeplot(SNP_rda, npcs = 20, type = "lines", bstick = T)

plot(SNP_rda, choices = c(1, 2), scaling = 3) #scaling=3:  “symmetrical scaling”
text(SNP_rda, choices = c(1, 2), scaling = 3, display="bp", col="#0868ac", cex=1)

plot(SNP_rda, choices = c(1, 3), scaling = 3)  # axes 1 and 3
text(SNP_rda, choices = c(1, 3), scaling = 3, display="bp", col="#0868ac", cex=1)

```

#### Highest contributing SNPs

```{r}
load.rda <- scores(SNP_rda, choices=c(1:3), display="species")  # Species scores for the first three constrained axes
# save(load.rda, file =  paste0(rdata.path, "SNP_2018_RDA.Rdata"))

hist(load.rda[,1], main="Loadings on RDA1")
hist(load.rda[,2], main="Loadings on RDA2")
hist(load.rda[,3], main="Loadings on RDA3") 

outliers <- function(x,z){
  lims <- mean(x) + c(-1, 1) * z * sd(x)     # find loadings +/-z sd from mean loading     
  x[x < lims[1] | x > lims[2]]               # locus names in these tails
}

cand1 <- outliers(load.rda[,1],3) # 38
cand2 <- outliers(load.rda[,2],3) # 69
cand3 <- outliers(load.rda[,3],3) # 34

ncand <- length(cand1) + length(cand2) + length(cand3)
ncand #1506


cand1 <- cbind.data.frame(rep(1,times=length(cand1)), names(cand1), unname(cand1))
cand2 <- cbind.data.frame(rep(2,times=length(cand2)), names(cand2), unname(cand2))
cand3 <- cbind.data.frame(rep(3,times=length(cand3)), names(cand3), unname(cand3))

colnames(cand1) <- colnames(cand2) <- colnames(cand3) <- c("axis","snp","loading")

cand <- rbind(cand1, cand2, cand3)
cand$snp <- as.character(cand$snp)


foo <- matrix(nrow=(ncand), ncol=8)  # 8 columns for 8 predictors
colnames(foo) <- c("AMT","MDR","sdT","AP","cvP","NDVI","Elev","Tree")

for (i in 1:length(cand$snp)) {
  nam <- cand[i,2]
  snp.gen <- gen.imp[,nam]
  foo[i,] <- apply(pred,2,function(x) cor(x,snp.gen))
}

cand <- cbind.data.frame(cand,foo)  
head(cand)
```

## b. Pooled samples

### Read data
```{r}
vcf.file <- "/datasets/work/ncmi-toa-rrbs/work/1.Toothfish/GENOME/Combined_SNP/Toothfish_Nuclear_SNP_POOL_revelio_biallelic.vcf.gz"

metafile <- "/datasets/work/ncmi-toa-rrbs/work/1.Toothfish/0.R-analysis/1.Toothfish_ALL_RRBS_metadata.csv"

TOA.pool.gl <- dartR.base::gl.read.vcf(vcffile = vcf.file,
                                      ind.metafile = metafile,
                                      mode = "genotype",
                                      verbose = 5)

pool.vcfr <- vcfR::read.vcfR(vcf.file, verbose = TRUE )
# Toothfish_Nuclear_SNP_POOL_GLnexus_gatk_biallelic.vcf.gz
# 8 samples
# 96 CHROMs
# 12,821 variants
# Object size: 5.9 Mb
# 0 percent missing data


TOA.pool.gl@other$loc.metrics$loc.names <- TOA.pool.gl$loc.names

TOA.pool.gl@other$loc.metrics$rdepth <- as.integer(TOA.pool.gl@other$loc.metrics$DP)/adegenet::nInd(TOA.pool.gl)
  

save(TOA.pool.gl, pool.vcfr, file = paste0(rdata.path,"TOA_POOL_SNP_genlight_raw.Rdata"))
```

### Raw coverage

```{r}
print(load(paste0(rdata.path,"TOA_POOL_SNP_genlight_raw.Rdata")))


gt <- vcfR::extract.gt(pool.vcfr)
gt[1:1000,1:8]

# genotypes to be delimited with a '|' when they are phased and a '/' when unphased
sum(gt == "0|1", na.rm = T) #909 HOM
sum(gt == "1|1", na.rm = T) #137 HET
sum(gt == "0/0", na.rm = T) #85981
sum(gt == "1/1", na.rm = T) #3535 HET
sum(gt == "0/1", na.rm = T) #11863
sum(gt == "0/2", na.rm = T) #0
sum(is.na(gt)) #5
sum(vcfR::is.indel(pool.vcfr)) #0


gt2 <- vcfR::extract.gt(pool.vcfr, element = 'GT', as.numeric = TRUE)
gt2[1:10,1:8]
sum(gt2==1, na.rm = T) #3672

dp <- vcfR::extract.gt(pool.vcfr, element = 'DP', as.numeric = TRUE)
dp[1:10,1:8]

dpf <- data.frame(dp, locus = rownames(dp))
dpf <- tidyr::gather(dpf, key = individual, 
                           value = Depth, -locus,
                           factor_key = TRUE) %>%
  dplyr::group_by(individual) %>%
  dplyr::mutate(NA_cnt = sum(is.na(Depth))/length(unique(locus))) %>%
  dplyr::ungroup()

dpf.NA <- dpf[!duplicated(dpf$individual),]
dpf.NA$NA_cnt <- round(dpf.NA$NA_cnt, 3)
dpf <- dpf[ dpf$Depth > 0 & !is.na(dpf$Depth),]
max(dpf$Depth, na.rm = T) #119

myPlot <- ggplot2::ggplot(dpf, ggplot2::aes(x = individual, y = Depth)) + 
  # geom_violin(fill="#8dd3c7", adjust=1.0, scale = "count", trim=TRUE) +
  ggplot2::geom_boxplot() +
  ggplot2::geom_text(data = dpf.NA,ggplot2::aes(x = individual, y = 310, label = NA_cnt), 
                     size = 2, angle = 90) +
  ggplot2::theme_bw() + 
  ggplot2::theme(axis.title.x = ggplot2::element_blank(), 
                  axis.text.x = ggplot2::element_text(size = 7, angle = 90, hjust = 1)) +
  ggplot2::scale_y_continuous(trans = scales::log2_trans(), 
                  breaks = c(1, 10, 50, 200),
                  minor_breaks = c(1:10, 2:5*10, 1:2*100))
print(myPlot)
ggplot2::ggsave(myPlot, bg = "white",
                filename = paste0(figure.path,
                                  "1A.TOA_POOL_vcfR_individual_coverage_12821_SNPs.png"
                                  ), 
                width = 30, height = 15, units = "cm")

dp.SNP <- rowMeans(dp, na.rm = TRUE)
sum(dp.SNP > 5 & !is.na(dp.SNP)) #6904

```
### Filter data

```{r}
# dna_file <- "/datasets/work/ncmi-toa-rrbs/work/1.Toothfish/Bismark_genome_index/D.mawsoni.fasta"
# gff_file <- "/datasets/work/ncmi-toa-rrbs/work/1.Toothfish/Bismark_genome_index/D.mawsoni_2.gtf"
# dna <- ape::read.dna(dna_file, format = "fasta")
# gff <- read.table(gff_file, sep = "\t", quote = "")
# gff2 <- gff[grep("JAAKFY010000027", gff[,1]),]




rm <- c("TOA_48-2_020","TOA_88-1_196","TOA_48-2_024")
metadata.new <- metadata.new[!metadata.new$SAMPLE_ID %in% rm,]
TOA.all.gl <- TOA.all.gl[TOA.all.gl$ind.names %in% metadata.new$SAMPLE_ID,]
TOA.all.gl$pop <- factor(metadata.new$region)




TOA.pool.gl <- dartR.base::gl.filter.allna(TOA.pool.gl, verbose = 5)
dartR.base::gl.report.monomorphs(TOA.pool.gl, verbose = 5)
TOA.pool.gl <- dartR.base::gl.filter.monomorphs(TOA.pool.gl, verbose = 5) #196011  binary SNPs

dartR.base::gl.report.rdepth(TOA.pool.gl, verbose = 5)
dev.print(file = paste0("TOA_POOL_SNP_read_depth_filter.png"),
          device = png, res = 300,width = 30,height = 30,units = "cm")
TOA.pool.gl <- dartR.base::gl.filter.rdepth(TOA.pool.gl, lower = 7, upper = 25, verbose = 5) #82 

dartR.base::gl.report.callrate(TOA.pool.gl, method = "loc")
TOA.pool.gl <- dartR.base::gl.filter.callrate(TOA.pool.gl, method = "loc", 
                                                  threshold = 0.9, verbose = 5) #56898 binary SNPs
dev.print(file = paste0("TOA_POOL_SNP_CallRate_filter.png"),
          device = png, res = 300,width = 30,height = 30,units = "cm")

dartR.base::gl.report.callrate(TOA.pool.gl, method = "ind")
TOA.pool.gl <- dartR.base::gl.filter.callrate(TOA.pool.gl, method = "ind", 
                                                  threshold = 0.94,
                                                  verbose = 5) #56 fish
dev.print(file = paste0("TOA_POOL_IND_CallRate_filter.png"),
          device = png, res = 300,width = 30,height = 30,units = "cm")

# GT13848-R[DR], Brev_17_CygnetCreek_01[Cy]

dartR.base::gl.report.callrate(TOA.pool.gl, method = "loc")



het.res <- dartR.base::gl.report.heterozygosity(TOA.pool.gl, method = "ind")
dev.print(file = paste0("TOA_POOL_IND_heterozygosity_filter.png"),
          device = png, res = 300,width = 30,height = 30,units = "cm")

dartR.base::gl.report.heterozygosity(TOA.pool.gl, method = "pop")
dev.print(file = paste0("TOA_POOL_IND_heterozygosity_POP.png"),
          device = png, res = 300,width = 30,height = 30,units = "cm")

# TOA.pool.gl <- dartR.base::gl.filter.heterozygosity(TOA.pool.gl, t.lower = 0, t.upper = 0.01,  verbose = 5)

TOA.pool.gl <- dartR.base::gl.filter.allna(TOA.pool.gl, verbose = 5)
dartR.base::gl.report.monomorphs(TOA.pool.gl, verbose = 5)
TOA.pool.gl <- dartR.base::gl.filter.monomorphs(TOA.pool.gl, verbose = 5) #56164 binary SNPs

dartR.base::gl.report.maf(TOA.pool.gl, verbose = 5)
TOA.pool.gl <- dartR.base::gl.filter.maf(TOA.pool.gl, threshold = 0.01, 
                                     plot.colors = c("#2171B5", "#6BAED6"),
                                     verbose = 5) #44465 binary SNPs

gl.replicates <- dartR.base::gl.report.replicates(TOA.pool.gl,
                                            loc_threshold = 1000, perc_geno = 0.80)
gl.replicates$table.rep[order(gl.replicates$table.rep$perc, decreasing = TRUE),]

TOA.pool.gl <- TOA.pool.gl[!TOA.pool.gl$ind.names %in% gl.replicates$ind.list.drop,]
# save(TOA.pool.gl, file = paste0("TOA_POOL_SNP_genlight_filtered.Rdata"))

```

### Popoolation2

```{r}
library("qqman")
library("DataCombine")

fst = read.table("all_PE_w10000_step10000.fst", header = T) 
fst <- fst[c("CHR", "BP", "X1.11", "X1.12", "X2.11", "X2.12", "X3.11", "X3.12", "X4.11", "X4.12", "X5.11", "X5.12", "X6.11", "X6.12", "X7.11", "X7.12", "X8.11", "X8.12", "X9.11", "X9.12", "X10.11", "X10.12", "X11.12")] #rename column headers
fst$ID <- paste(fst$CHR,fst$BP,sep=".")

#tidying file
Replaces <- data.frame(from=c("Chromosome_"), to=c(""))
fst2 <- FindReplace(data = fst, Var = "CHR", replaceData = Replaces,
                       from = "from", to = "to", exact = FALSE)
fst2$CHR <- as.numeric(fst2$CHR)

manhattan(fst2, chr="CHR", bp="BP", snp="ID", p="X1.11", suggestiveline = F, genomewideline = F, logp = F, ylim =c(0,1)) #p here refers to pvalue or in this case the Fst value, substitute with which column comparison you are interested in graphing
```

#### Pool-seq Ne
```{r}
mySync <- read.sync(file="/Path/to/data.sync", gen=c(0, 10, 20, 0, 10, 20), repl=c(1, 1, 1, 2, 2, 2), rising = FALSE)

estimateNe(p0=simTraj[,"F0"], pt=simTraj[,"F20"], cov0=simCov[,"F0"], covt=simCov[,"F20"], t=20, Ncensus=1000, poolSize=c(300, 300))
estimateSH(simTraj, Ne=1000, t=seq(0, 60, by=10), h=0.5)
estimateSH(simTraj, Ne=1000, t=seq(0, 60, by=10), h=0.5, simulate.p.value=TRUE)
est <- estimateSH(simTraj, Ne=1000, t=seq(0, 60, by=10), h=0.5)
confint(est)

```

# 3_Mitogenome

## Haplotype network
```{r}
print(load(paste0(rdata.path,"Toothfish_metadata.Rdata")))
metadata$region <- paste0(metadata$STRATA1, "_", metadata$Year)
metadata <- metadata[!metadata$pop %in% c("58.4.2-2024_50L", "88.2-2024"),]
table(metadata$region)
# f.in <- "/datasets/work/ncmi-toa-rrbs/work/1.Toothfish/0.R-analysis/D.mawsoni_mtgenome_alignment.fasta"
# f.in <- "/datasets/work/ncmi-toa-rrbs/work/1.Toothfish/0.R-analysis/0_Toothfish_mtgenome_aligned_MAFFT_edited_FINAL.fasta"
f.in <- "/datasets/work/ncmi-toa-rrbs/work/1.Toothfish/0.R-analysis/0_Toothfish_mtgenome_aligned_MAFFT_geneious-Edited-FINAL2.fasta"


TOA.mt.ml <- apex::read.multiFASTA(f.in)
(setLocusNames(TOA.mt.ml) <- "Mitogenome")
f.in2 <- "/datasets/work/ncmi-toa-rrbs/work/1.Toothfish/0.R-analysis/0_Toothfish_mtgenome_aligned_MAFFT_geneious-Edited-FINAL2_SUBSET.fasta"
id <- metadata$File_names
TOA.mt.dna.new <-   TOA.mt.ml[attributes(TOA.mt.ml)$labels %in% id, , drop = TRUE]
ape::write.FASTA(
  TOA.mt.dna.new@dna$Mitogenome,
  file = f.in2,
  header = NULL,
  append = FALSE
)

TOA.mt.ml <- apex::read.multiFASTA(f.in2)
(setLocusNames(TOA.mt.ml) <- "Mitogenome")
TOA.mt.gi <- apex::multidna2genind(TOA.mt.ml, mlst = TRUE)
adegenet::indNames(TOA.mt.gi) %in% metadata$File_names
adegenet::indNames(TOA.mt.gi)[!adegenet::indNames(TOA.mt.gi) %in% metadata$File_names]
metadata$File_names[!metadata$File_names %in% adegenet::indNames(TOA.mt.gi)]

metadata <- metadata[metadata$File_names %in% adegenet::indNames(TOA.mt.gi),]
TOA.mt.gi$pop <- factor(metadata$pop[order(match(metadata$File_names,
                                                    adegenet::indNames(TOA.mt.gi)))])

TOA.mt.dna <- haplotypes::read.fas(f.in2)
# TOA.mt.hap <- haplotypes::haplotype(TOA.mt.dna,indels = "sic")
TOA.mt.hap <- haplotypes::haplotype(TOA.mt.dna,indels = "sic")

TOA.mt.hapgroup <- haplotypes::grouping(TOA.mt.hap,TOA.mt.gi$pop)

TOA.mt.bin <- ape::read.dna(f.in2, format = "fasta", as.matrix = TRUE)
# HTs <- paste0("HT", sprintf('%0.3d', 1:59))
HTs <- paste0("HT", sprintf('%0.2d', 1:43))
TOA.mt.h <- pegas::haplotype(TOA.mt.bin, labels = HTs)
# TOA.mt.h <- pegas::haplotype(TOA.mt.bin)

# pop <- stringr::str_split_fixed(string = TOA.mt.ml@labels, pattern = "_", n = 3)[,2]
pop <- TOA.mt.gi$pop

my_strata <- data.frame(pop = TOA.mt.gi$pop)
adegenet::strata(TOA.mt.gi) <- my_strata



net <- pegas::haploNet(TOA.mt.h)
ind.hap <- with(
    utils::stack(setNames(attr(TOA.mt.h, "index"), rownames(TOA.mt.h))),
    table(hap = ind, pop = pop[values]))

size.new <- attr(net, "freq")
colours.all.new <- colours.all[names(colours.all) %in% pop]
colours.all.new <- colours.all.new[order(names(colours.all.new))]

par(mar = c(1,1,1,1))
plot(net, size = size.new, scale.ratio = 0.75, pie = ind.hap, 
     legend = c(-10,20),
     # legend = c(-5,5),
     show.mutation = 2,threshold = 0,  srt = 1, cex = 0.75, bg = colours.all.new)
title(paste0("HT network of 304 individuals per strata"))
dev.print(file = paste0(figure.path,"TOA_Mitogenome_HT_network_SUB.png"),
          device = png, res = 300,width = 30,height = 30,units = "cm")




size.new <- sqrt(size.new)


par(mar = c(1,1,1,1))
plot(net, size = size.new, scale.ratio = 1.8, pie = ind.hap, 
     # legend = c(-5,5),
     legend = c(-15,20),
     show.mutation = 2,threshold = 0,  srt = 1, cex = 0.75, bg = colours.all.new)
title(paste0("HT network of 302 individuals per strata; 6000bp fragment"))
dev.print(file = paste0(figure.path,"TOA_Mitogenome_HT_network2_SUB.png"),
          device = png, res = 300,width = 30,height = 30,units = "cm")


# xy <- replot()
# save(xy, file = paste0(rdata.path,"TOA_Mitogenome_HT_network_replot.Rdata"))
# save(xy, file = paste0(rdata.path,"TOA_Mitogenome_HT_network_replot_SUB.Rdata"))

# load(paste0(rdata.path,"TOA_Mitogenome_HT_network_replot.Rdata"))
load(paste0(rdata.path,"TOA_Mitogenome_HT_network_replot_SUB.Rdata"))

par(mar = c(1,1,1,1))
plot(net, xy = xy, size = size.new, scale.ratio = 0.75, pie = ind.hap, 
     # legend = c(-5,10),
     legend = c(-16.5,-9),
     show.mutation = 2,threshold = 0,  srt = 1, cex = 0.75, bg = colours.all.new)
title(paste0("HT network of 302 individuals per strata; 5927bp fragment"))
dev.print(file = paste0(figure.path,"TOA_Mitogenome_HT_network2_replot_SUB.png"),
          device = png, res = 300,width = 30,height = 30,units = "cm")
```

## Genetic differentiation

```{r}
TOA.mt.dist <- apex::dist.multidna(TOA.mt.ml, pool = TRUE)
{amova.result <- pegas::amova(TOA.mt.dist ~ pop,
                              data = adegenet::strata(TOA.mt.gi), nperm = 10000)
print("AMOVA result from pegas using the pooled genetic distance:")
print(amova.result)
sig2 <- setNames(amova.result$varcomp$sigma2, rownames(amova.result$varcomp))
knitr::kable(pegas::getPhi(sig2), format = "latex", caption = "PHIst from pegas AMOVA using the pooled genetic distance") %>%
  kableExtra::kable_styling(full_width = FALSE)
}


popstr.mt.river <-
  strataG::popStructTest(
    ggar.gt,
    nrep = 10000,
    stats = "all",
    type = "both",
    keep.null = FALSE,
    quietly = TRUE,
    max.cores = parallel::detectCores() - 1,
    write.output = TRUE
  )

```



# 4_Methylation QC

## Mapping efficiency & Conversion success

```{r}
print(load(paste0(rdata.path,"Toothfish_metadata.Rdata"))) #ALL fish
metadata$pop <- paste0(metadata$STRATA1, "-", metadata$Year)
```

### Lambda phage

Each sample was spike with approx 10% of unmethylated Lambda Phage DNA.

```{r, eval=FALSE}
lambda.path <- "/datasets/work/ncmi-toa-rrbs/work/1.Toothfish/LAMBDA/Combined_methylation/"

Map.eff <- data.frame(map = NULL, sample = NULL, pop = NULL)
cpg <- data.frame(cpg = NULL, sample = NULL, pop = NULL)
cHG <- data.frame(cHG = NULL, sample = NULL, pop = NULL)
cHH <- data.frame(cHH = NULL, sample = NULL, pop = NULL)

for (i in 1:312) {
  SE <- readr::read_tsv(
      paste0(lambda.path,
             metadata$File_names[i],
             "_R1_trimmed_sort_bismark_bt2_SE_report.txt"), skip = 4)
  Map.eff.sample <- SE[4, ]
  Map.eff.sample <- gsub(".*:\t","",Map.eff.sample)
  Map.eff.sample <- data.frame(map = gsub(".*:\t","",Map.eff.sample), 
                               sample = metadata$File_names[i],
                               pop = metadata$pop[i])
  Map.eff <- rbind(Map.eff, Map.eff.sample)
  
  cpg.sample <- SE[25, ]
  cpg.sample <- data.frame(cpg = gsub(".*:\t","",cpg.sample), 
                           sample = metadata$File_names[i],
                           pop = metadata$pop[i])
  cpg <- rbind(cpg, cpg.sample)
  
  cHG.sample <- SE[26, ]
  cHG.sample <- data.frame(cHG = gsub(".*:\t","",cHG.sample), 
                           sample = metadata$File_names[i],
                           pop = metadata$pop[i])
  cHG <- rbind(cHG, cHG.sample)
  
  cHH.sample <- SE[27, ]
  cHH.sample <- data.frame(cHH = gsub(".*:\t","",cHH.sample), 
                           sample = metadata$File_names[i],
                           pop = metadata$pop[i])
  cHH <- rbind(cHH, cHH.sample)
}
Map.eff$map <- as.numeric(gsub("%","",Map.eff$map))
cpg$cpg <- as.numeric(gsub("%","",cpg$cpg))
cHG$cHG <- as.numeric(gsub("%","",cHG$cHG))
cHH$cHH <- as.numeric(gsub("%","",cHH$cHH))

save(Map.eff, cpg, cHG, cHH, file = paste0(rdata.path,"Lambda_conversion_rate.Rdata"))
```

```{r}
print(load(paste0(rdata.path,"Lambda_conversion_rate.Rdata")))

Map.plot <- ggplot2::ggplot(Map.eff, ggplot2::aes(x = sample, y = map, col = pop)) +
  ggplot2::geom_point(size = 2) +
  ggplot2::ggtitle("Lambda phage - Mapping efficiency") +
  ggplot2::ylab("Mapping efficiency (%) of Lambda phage sequences") +
  ggplot2::scale_color_manual(values = colours.all) +
  ggplot2::theme_bw() +
  ggplot2::theme(
    legend.position = "top",
    axis.text.x = ggplot2::element_blank(),
    axis.text = ggplot2::element_text(size = 15),
    axis.title.x = ggplot2::element_text(size = 15),
    axis.title.y = ggplot2::element_text(size = 15)
  )
print(Map.plot)

ggplot2::ggsave(Map.plot,
                filename = paste0(figure.path,"Lambda_Mapping_efficiency.png"), 
                width = 30, height = 15, units = 'cm')

high.lambda.mapping <- Map.eff$sample[Map.eff$map > 20]
print(high.lambda.mapping)


cpg.plot <- ggplot2::ggplot(cpg, ggplot2::aes(x = sample, y = cpg, col = pop)) +
  ggplot2::geom_point(size = 2) +
  ggplot2::ggtitle("Lambda CpG") +
  ggplot2::ylab("CpG methylation (%) of Lambda phage") +
  ggplot2::scale_color_manual(values = colours.all) +
  ggplot2::theme_bw() +
  ggplot2::theme(
    legend.position = "top",
    axis.text.x = ggplot2::element_blank(),
    axis.text = ggplot2::element_text(size = 15),
    axis.title.x = ggplot2::element_text(size = 15),
    axis.title.y = ggplot2::element_text(size = 15)
  )
print(cpg.plot)

ggplot2::ggsave(cpg.plot,filename = paste0(figure.path,"Lambda_CpG_methylation.png"), width = 30, height = 15, units = 'cm')

high.lambda.methylation <- cpg$sample[cpg$cpg > 0.6]
print(high.lambda.methylation)


cHG.plot <- ggplot2::ggplot(cHG, ggplot2::aes(x = sample, y = cHG, col = pop)) +
  ggplot2::geom_point(size = 2) +
  ggplot2::ggtitle("Lambda CHG") +
  ggplot2::ylab("CHG methylation (%) of Lambda phage") +
  ggplot2::scale_color_manual(values = colours.all) +
  ggplot2::theme_bw() +
  ggplot2::theme(
    legend.position = "top",
    axis.text.x = ggplot2::element_blank(),
    axis.text = ggplot2::element_text(size = 15),
    axis.title.x = ggplot2::element_text(size = 15),
    axis.title.y = ggplot2::element_text(size = 15)
  )
print(cHG.plot)

ggplot2::ggsave(cHG.plot,filename = paste0(figure.path,"Lambda_CHG_methylation.png"), width = 30, height = 15, units = 'cm')

cHG$sample[cHG$cHG > 0.35]


cHH.plot <- ggplot2::ggplot(cHH, ggplot2::aes(x = sample, y = cHH, col = pop)) +
  ggplot2::geom_point(size = 2) +
  ggplot2::ggtitle("Lambda cHH") +
  ggplot2::ylab("cHH methylation (%) of Lambda phage") +
  ggplot2::scale_color_manual(values = colours.all) +
  ggplot2::theme_bw() +
  ggplot2::theme(
    legend.position = "top",
    axis.text.x = ggplot2::element_blank(),
    axis.text = ggplot2::element_text(size = 15),
    axis.title.x = ggplot2::element_text(size = 15),
    axis.title.y = ggplot2::element_text(size = 15)
  )
print(cHH.plot)

ggplot2::ggsave(cHH.plot,filename = paste0(figure.path,"Lambda_cHH_methylation.png"), width = 30, height = 15, units = 'cm')

cHH$sample[cHH$cHH > 0.5] 

save(high.lambda.mapping,high.lambda.methylation, file = paste0(rdata.path, "Low_quality_fish.Rdata"))
```


### Mitogenome

```{r, eval=FALSE}
mtgenome.path <- "/datasets/work/ncmi-toa-rrbs/work/1.Toothfish/MITOGENOME/Combined_methylation/"

Map.eff <- data.frame(map = NULL, sample = NULL, pop = NULL)
cpg <- data.frame(cpg = NULL, sample = NULL, pop = NULL)
cHG <- data.frame(cHG = NULL, sample = NULL, pop = NULL)
cHH <- data.frame(cHH = NULL, sample = NULL, pop = NULL)

for (i in 1:312) {
  SE <- readr::read_tsv(
      paste0(mtgenome.path,
             metadata$File_names[i],
             "_R1_trimmed_sort_bismark_bt2_SE_report.txt"), skip = 4)
  Map.eff.sample <- SE[4, ]
  Map.eff.sample <- gsub(".*:\t","",Map.eff.sample)
  Map.eff.sample <- data.frame(map = gsub(".*:\t","",Map.eff.sample), 
                               sample = metadata$File_names[i],
                               pop = metadata$pop[i])
  Map.eff <- rbind(Map.eff, Map.eff.sample)
  
  cpg.sample <- SE[25, ]
  cpg.sample <- data.frame(cpg = gsub(".*:\t","",cpg.sample), 
                           sample = metadata$File_names[i],
                           pop = metadata$pop[i])
  cpg <- rbind(cpg, cpg.sample)
  
  cHG.sample <- SE[26, ]
  cHG.sample <- data.frame(cHG = gsub(".*:\t","",cHG.sample), 
                           sample = metadata$File_names[i],
                           pop = metadata$pop[i])
  cHG <- rbind(cHG, cHG.sample)
  
  cHH.sample <- SE[27, ]
  cHH.sample <- data.frame(cHH = gsub(".*:\t","",cHH.sample),
                           sample = metadata$File_names[i],
                           pop = metadata$pop[i])
  cHH <- rbind(cHH, cHH.sample)
}
Map.eff$map <- as.numeric(gsub("%","",Map.eff$map))
cpg$cpg <- as.numeric(gsub("%","",cpg$cpg))
cHG$cHG <- as.numeric(gsub("%","",cHG$cHG))
cHH$cHH <- as.numeric(gsub("%","",cHH$cHH))

save(Map.eff, cpg, cHG, cHH, file = paste0(rdata.path,"mtgenome_conversion_rate.Rdata"))
```

```{r}
print(load(paste0(rdata.path,"mtgenome_conversion_rate.Rdata")))

Map.plot <- ggplot2::ggplot(Map.eff, ggplot2::aes(x = sample, y = map, col = pop)) +
  ggplot2::geom_point(size = 2) +
  ggplot2::ggtitle("Mitogenome - Mapping efficiency") +
  ggplot2::ylab("Mapping efficiency (%) of Mitogenome sequences") +
  ggplot2::theme_bw() +
  ggplot2::scale_color_manual(values = colours.all) +
  ggplot2::theme(
    legend.position = "top",
    axis.text.x = ggplot2::element_blank(),
    axis.text = ggplot2::element_text(size = 15),
    axis.title.x = ggplot2::element_text(size = 15),
    axis.title.y = ggplot2::element_text(size = 15)
  )
print(Map.plot)

ggplot2::ggsave(Map.plot,filename = paste0(figure.path,"mtgenome_Mapping_efficiency.png"), width = 30, height = 15, units = 'cm')

Map.eff$sample[Map.eff$map > 2]


cpg.plot <- ggplot2::ggplot(cpg, ggplot2::aes(x = sample, y = cpg, col = pop)) +
  ggplot2::geom_point(size = 2) +
  ggplot2::ggtitle("Mitogenome CpG") +
  ggplot2::ylab("CpG methylation (%) of Mitogenome") +
  ggplot2::scale_color_manual(values = colours.all) +
  ggplot2::theme_bw() +
  ggplot2::theme(
    legend.position = "top",
    axis.text.x = ggplot2::element_blank(),
    axis.text = ggplot2::element_text(size = 15),
    axis.title.x = ggplot2::element_text(size = 15),
    axis.title.y = ggplot2::element_text(size = 15)
  )
print(cpg.plot)

ggplot2::ggsave(cpg.plot,filename = paste0(figure.path,"mtgenome_CpG_methylation.png"), width = 30, height = 15, units = 'cm')

cpg$sample[cpg$cpg > 0.75] #"TOA_48-2_008" "TOA_48-2_013"


cHG.plot <- ggplot2::ggplot(cHG, ggplot2::aes(x = sample, y = cHG, col = pop)) +
  ggplot2::geom_point(size = 2) +
  ggplot2::ggtitle("Mitogenome CHG") +
  ggplot2::ylab("CHG methylation (%) of Mitogenome") +
  ggplot2::scale_color_manual(values = colours.all) +
  ggplot2::theme_bw() +
  ggplot2::theme(
    legend.position = "top",
    axis.text.x = ggplot2::element_blank(),
    axis.text = ggplot2::element_text(size = 15),
    axis.title.x = ggplot2::element_text(size = 15),
    axis.title.y = ggplot2::element_text(size = 15)
  )
print(cHG.plot)

ggplot2::ggsave(cHG.plot,filename = paste0(figure.path,"mtgenome_CHG_methylation.png"), width = 30, height = 15, units = 'cm')

cHG$sample[cHG$cHG > 0.5] 
#"TOA_48-2_005_UK046_REP"



cHH.plot <- ggplot2::ggplot(cHH, ggplot2::aes(x = sample, y = cHH, col = pop)) +
  ggplot2::geom_point(size = 2) +
  ggplot2::ggtitle("Mitogenome cHH") +
  ggplot2::ylab("cHH methylation (%) of Mitogenome") +
  ggplot2::scale_color_manual(values = colours.all) +
  ggplot2::theme_bw() +
  ggplot2::theme(
    legend.position = "top",
    axis.text.x = ggplot2::element_blank(),
    axis.text = ggplot2::element_text(size = 15),
    axis.title.x = ggplot2::element_text(size = 15),
    axis.title.y = ggplot2::element_text(size = 15)
  )
print(cHH.plot)

ggplot2::ggsave(cHH.plot,filename = paste0(figure.path,"mtgenome_cHH_methylation.png"), width = 30, height = 15, units = 'cm')

cHH$sample[cHH$cHH > 0.5] # "TOA_48-2_005_UK046_REP"
```

### Toothfish Genome
```{r, eval=FALSE}
genome.path <- "/datasets/work/ncmi-toa-rrbs/work/1.Toothfish/GENOME/Combined_methylation/"

Map.eff <- data.frame(map = NULL, sample = NULL, pop = NULL)
cpg <- data.frame(cpg = NULL, sample = NULL, pop = NULL)
cHG <- data.frame(cHG = NULL, sample = NULL, pop = NULL)
cHH <- data.frame(cHH = NULL, sample = NULL, pop = NULL)

for (i in 1:312) {
  SE <- readr::read_tsv(
      paste0(genome.path, metadata$File_names[i],
             "_R1_trimmed_sort_bismark_bt2_SE_report.txt"), skip = 4)
  Map.eff.sample <- SE[4, ]
  Map.eff.sample <- gsub(".*:\t","",Map.eff.sample)
  Map.eff.sample <- data.frame(map = gsub(".*:\t","",Map.eff.sample), 
                               sample = metadata$File_names[i],
                               pop = metadata$pop[i])
  Map.eff <- rbind(Map.eff, Map.eff.sample)
  
  cpg.sample <- SE[25, ]
  cpg.sample <- data.frame(cpg = gsub(".*:\t","",cpg.sample),
                           sample = metadata$File_names[i],
                           pop = metadata$pop[i])
  cpg <- rbind(cpg, cpg.sample)
  
  cHG.sample <- SE[26, ]
  cHG.sample <- data.frame(cHG = gsub(".*:\t","",cHG.sample), 
                           sample = metadata$File_names[i],
                           pop = metadata$pop[i])
  cHG <- rbind(cHG, cHG.sample)
  
  cHH.sample <- SE[27, ]
  cHH.sample <- data.frame(cHH = gsub(".*:\t","",cHH.sample), 
                           sample = metadata$File_names[i],
                           pop = metadata$pop[i])
  cHH <- rbind(cHH, cHH.sample)
}

Map.eff$map <- as.numeric(gsub("%","",Map.eff$map))
cpg$cpg <- as.numeric(gsub("%","",cpg$cpg))
cHG$cHG <- as.numeric(gsub("%","",cHG$cHG))
cHH$cHH <- as.numeric(gsub("%","",cHH$cHH))

save(Map.eff, cpg, cHG, cHH, file = paste0(rdata.path,"Toothfish_genome_conversion_rate.Rdata"))
```

```{r}
print(load(paste0(rdata.path,"Toothfish_genome_conversion_rate.Rdata")))

Map.plot <- ggplot2::ggplot(Map.eff, ggplot2::aes(x = sample, y = map, col = pop)) +
  ggplot2::geom_point(size = 2) +
  ggplot2::ggtitle("Toothfish - Mapping efficiency") +
  ggplot2::ylab("Mapping efficiency (%) of Toothfish genomic sequences") +
  ggplot2::scale_color_manual(values = colours.all) +
  ggplot2::theme_bw() +
  ggplot2::theme(
    legend.position = "top",
    axis.text.x = ggplot2::element_blank(),
    axis.text = ggplot2::element_text(size = 15),
    axis.title.x = ggplot2::element_text(size = 15),
    axis.title.y = ggplot2::element_text(size = 15)
  )
print(Map.plot)

ggplot2::ggsave(Map.plot,
                filename = paste0(figure.path,"Toothfish_genome_Mapping_efficiency.png"),
                width = 30, height = 15, units = 'cm')

low.genome.mapping <- Map.eff$sample[Map.eff$map < 35]
print(low.genome.mapping)


cpg.plot <- ggplot2::ggplot(cpg, ggplot2::aes(x = sample, y = cpg, col = pop)) +
  ggplot2::geom_point(size = 2) +
  ggplot2::ggtitle("Toothfish genome CpG") +
  ggplot2::ylab("CpG methylation (%) of Toothfish genome") +
  ggplot2::scale_color_manual(values = colours.all) +
  ggplot2::theme_bw() +
  ggplot2::ylim(0,100) +
  ggplot2::theme(
    legend.position = "top",
    axis.text.x = ggplot2::element_blank(),
    axis.text = ggplot2::element_text(size = 15),
    axis.title.x = ggplot2::element_text(size = 15),
    axis.title.y = ggplot2::element_text(size = 15)
  )
print(cpg.plot)

ggplot2::ggsave(cpg.plot,
                filename = paste0(figure.path,"Toothfish_genome_CpG_methylation2.png"), 
                width = 30, height = 15, units = 'cm')

low.genome.methylation <- cpg$sample[cpg$cpg < 55]
print(low.genome.methylation)

cHG.plot <- ggplot2::ggplot(cHG, ggplot2::aes(x = sample, y = cHG, col = pop)) +
  ggplot2::geom_point(size = 2) +
  ggplot2::ggtitle("Toothfish genome CHG") +
  ggplot2::ylab("CHG methylation (%) of Toothfish genome") +
  ggplot2::scale_color_manual(values = colours.all) +
  ggplot2::theme_bw() +
  ggplot2::theme(
    legend.position = "top",
    axis.text.x = ggplot2::element_blank(),
    axis.text = ggplot2::element_text(size = 15),
    axis.title.x = ggplot2::element_text(size = 15),
    axis.title.y = ggplot2::element_text(size = 15)
  )
print(cHG.plot)

ggplot2::ggsave(cHG.plot,
                filename = paste0(figure.path,"Toothfish_genome_CHG_methylation.png"), 
                width = 30, height = 15, units = 'cm')

cHG$sample[cHG$cHG > 1] 
# "TOA_48-2_005_UK046_REP" "TOA_48-2_028""TOA_88-2_P24_AD292328_muscle_F" 
# "TOA_88-2_P67_AD294613_fin_F" "TOA_88-2_P69_AD294614_fin_F_REP" "TOA_88-2_P69_AD294614_fin_F" 


cHH.plot <- ggplot2::ggplot(cHH, ggplot2::aes(x = sample, y = cHH, col = pop)) +
  ggplot2::geom_point(size = 2) +
  ggplot2::ggtitle("Toothfish genome cHH") +
  ggplot2::ylab("cHH methylation (%) of Toothfish genome") +
  ggplot2::scale_color_manual(values = colours.all) +
  ggplot2::theme_bw() +
  ggplot2::theme(
    legend.position = "top",
    axis.text.x = ggplot2::element_blank(),
    axis.text = ggplot2::element_text(size = 15),
    axis.title.x = ggplot2::element_text(size = 15),
    axis.title.y = ggplot2::element_text(size = 15)
  )
print(cHH.plot)

ggplot2::ggsave(cHH.plot,
                filename = paste0(figure.path,"Toothfish_genome_cHH_methylation.png"),
                width = 30, height = 15, units = 'cm')

cHH$sample[cHH$cHH > 0.75] 

# "TOA_48-2_005_UK046_REP" "TOA_48-2_028" "TOA_88-2_P24_AD292328_muscle_F" "TOA_88-2_P67_AD294613_fin_F" "TOA_88-2_P69_AD294614_fin_F_REP" "TOA_88-2_P69_AD294614_fin_F" "TOA_88-2_P79_AD294619_fin_F"

save(high.lambda.mapping,high.lambda.methylation, low.genome.mapping,low.genome.methylation,
     file = paste0(rdata.path, "Low_quality_fish.Rdata"))
```


## M-bias

Different read lengths between 2018 and 2024.
NuGen python script cuts 3bp from 3' (YGG) 5bp from 5' end. The 2024 fastq files were trimmed from 150 to 92bp afterwards to equalise the lengths. 

```{r, eval=FALSE}
mbias.path <- "/datasets/work/ncmi-toa-rrbs/work/1.Toothfish/GENOME/Combined_methylation/"

cpg.M <- data.frame(position = NULL, `count methylated` = NULL, 
                    `count unmethylated` = NULL, `% methylation` = NULL, 
                    coverage = NULL)
cHG.M <- data.frame(position = NULL, `count methylated` = NULL, 
                    `count unmethylated` = NULL, `% methylation` = NULL, 
                    coverage = NULL)
cHH.M <- data.frame(position = NULL, `count methylated` = NULL, 
                    `count unmethylated` = NULL, `% methylation` = NULL, 
                    coverage = NULL)

for (i in c(1:8,10,45,66,79:170,210,249:312)) {
  Mbias <- readr::read_tsv(
      paste0(mbias.path,
             metadata$File_names[i],
             "_R1_trimmed_sort_bismark_bt2.M-bias.txt"), skip = 2)
  Mbias$sample <- metadata$File_names[i]
  cpg.M.sample <- Mbias[1:97, ]
  cpg.M <- rbind(cpg.M, cpg.M.sample)
  cHG.M.sample <- Mbias[100:197, ]
  cHG.M <- rbind(cHG.M, cHG.M.sample)
  cHH.M.sample <- Mbias[200:297, ]
  cHH.M <- rbind(cHH.M, cHH.M.sample)
}

# for (i in c(9,11:44,46:65,67:78,171:209,211:248)) {
for (i in 1:312) {
  Mbias <- readr::read_tsv(
      paste0(mbias.path,
             metadata$File_names[i],
             "_R1_trimmed_sort_bismark_bt2.M-bias.txt"), skip = 2)
  Mbias$sample <- metadata$File_names[i]
  cpg.M.sample <- Mbias[1:92, ]
  cpg.M <- rbind(cpg.M, cpg.M.sample)
  cHG.M.sample <- Mbias[95:187, ]
  cHG.M <- rbind(cHG.M, cHG.M.sample)
  cHH.M.sample <- Mbias[190:282, ]
  cHH.M <- rbind(cHH.M, cHH.M.sample)
}

cpg.M$`% methylation` <- as.numeric(cpg.M$`% methylation`)
cpg.M$position <- as.integer(cpg.M$position)
cpg.M$coverage <- as.integer(cpg.M$coverage)

cHG.M$`% methylation` <- as.numeric(cHG.M$`% methylation`)
cHG.M$position <- as.integer(cHG.M$position)
cHG.M$coverage <- as.integer(cHG.M$coverage)

cHH.M$`% methylation` <- as.numeric(cHH.M$`% methylation`)
cHH.M$position <- as.integer(cHH.M$position)
cHH.M$coverage <- as.integer(cHH.M$coverage)


save(cpg.M, cHG.M, cHH.M, file = paste0(rdata.path,"Toothfish_Mbias.Rdata"))
```


```{r}
load(paste0(rdata.path,"Toothfish_Mbias.Rdata"))

M.plot <- ggplot2::ggplot(cpg.M, ggplot2::aes(x = position, y = `% methylation`,
                                              colour = sample)) +
  ggplot2::geom_line() +
  ggplot2::ggtitle("M-bias plot: CpG") +
  ggplot2::ylim(c(0,100)) +
  ggplot2::theme_bw() +
  ggplot2::theme(
    legend.position = "none",
    axis.text.x = ggplot2::element_text(size = 15),
    axis.text = ggplot2::element_text(size = 15),
    axis.title.x = ggplot2::element_text(size = 15),
    axis.title.y = ggplot2::element_text(size = 15)
  )

print(M.plot)
ggplot2::ggsave(M.plot,
                filename = paste0(figure.path,"Toothfish_M-bias_CpG.png"), 
                width = 30, height = 15, units = 'cm')

Mbias.coverage <- cpg.M %>%
  dplyr::group_by(sample) %>%
  dplyr::summarise(MEAN = mean(`% methylation`, na.rm = TRUE))
low.mbias.methylation <- Mbias.coverage$sample[Mbias.coverage$MEAN < 50]
print(low.mbias.methylation)


M.plot2 <- ggplot2::ggplot(cpg.M, 
                           ggplot2::aes(x = position, y = coverage , 
                                        colour = sample)) +
  ggplot2::geom_line() +
  ggplot2::ggtitle("M-bias plot: CpG coverage") +
  ggplot2::theme_bw() +
  ggplot2::theme(
    legend.position = "none",
    axis.text.x = ggplot2::element_text(size = 15),
    axis.text = ggplot2::element_text(size = 15),
    axis.title.x = ggplot2::element_text(size = 15),
    axis.title.y = ggplot2::element_text(size = 15)
  )
print(M.plot2)
ggplot2::ggsave(M.plot2,
                filename = 
                  paste0(figure.path,"Toothfish_M-bias_CpG_coverage.png"), 
                width = 30, height = 15, units = 'cm')

M.plot3 <- ggplot2::ggplot(data = cpg.M[cpg.M$position < 89,], 
                           ggplot2::aes(x = position, y = coverage , 
                                        colour = sample)) +
  ggplot2::geom_line() +
  ggplot2::ggtitle("M-bias plot: CpG coverage") +
  ggplot2::theme_bw() +
  ggplot2::theme(
    legend.position = "none",
    axis.text.x = ggplot2::element_text(size = 15),
    axis.text = ggplot2::element_text(size = 15),
    axis.title.x = ggplot2::element_text(size = 15),
    axis.title.y = ggplot2::element_text(size = 15)
  )
print(M.plot3)
ggplot2::ggsave(M.plot2,
                filename = paste0(figure.path,"Toothfish_M-bias_CpG_coverage4-88.png"), 
                width = 30, height = 15, units = 'cm')


Mbias.coverage <- cpg.M %>%
  dplyr::filter(position < 89) %>%
  dplyr::group_by(sample) %>%
  dplyr::summarise(MEAN_COV  = mean(coverage , na.rm = TRUE))
Mbias.coverage$sample[Mbias.coverage$MEAN_COV < 250000]




M.plot <- ggplot2::ggplot(cHG.M, 
                          ggplot2::aes(x = position, y = `% methylation`, 
                                       colour = sample)) +
  ggplot2::geom_line() +
    ggplot2::ggtitle("M-bias plot: CHG") +
  ggplot2::theme_bw() +
  ggplot2::theme(
    legend.position = "none",
    axis.text.x = ggplot2::element_text(size = 15),
    axis.text = ggplot2::element_text(size = 15),
    axis.title.x = ggplot2::element_text(size = 15),
    axis.title.y = ggplot2::element_text(size = 15)
  )
print(M.plot)
ggplot2::ggsave(M.plot,
                filename = paste0(figure.path,"Toothfish_M-bias_CHG.png"), 
                width = 30, height = 15, units = 'cm')


M.plot2 <- ggplot2::ggplot(cHG.M, 
                           ggplot2::aes(x = position, y = coverage , 
                                        colour = sample)) +
  ggplot2::geom_line() +
  ggplot2::ggtitle("M-bias plot: CHG coverage") +
  ggplot2::theme_bw() +
  ggplot2::theme(
    legend.position = "none",
    axis.text.x = ggplot2::element_text(size = 15),
    axis.text = ggplot2::element_text(size = 15),
    axis.title.x = ggplot2::element_text(size = 15),
    axis.title.y = ggplot2::element_text(size = 15)
  )
print(M.plot2)
ggplot2::ggsave(M.plot2,
                filename = paste0(figure.path,"Toothfish_M-bias_CHG_coverage.png"), 
                width = 30, height = 15, units = 'cm')




M.plot <- ggplot2::ggplot(cHH.M, 
                          ggplot2::aes(x = position, y = `% methylation`, 
                                       colour = sample)) +
  ggplot2::geom_line() +
    ggplot2::ggtitle("M-bias plot: CHH") +
  ggplot2::theme_bw() +
  ggplot2::theme(
    legend.position = "none",
    axis.text.x = ggplot2::element_text(size = 15),
    axis.text = ggplot2::element_text(size = 15),
    axis.title.x = ggplot2::element_text(size = 15),
    axis.title.y = ggplot2::element_text(size = 15)
  )
print(M.plot)
ggplot2::ggsave(M.plot,
                filename = paste0(figure.path,"Toothfish_M-bias_CHH.png"), 
                width = 30, height = 15, units = 'cm')


save(high.lambda.mapping,high.lambda.methylation, 
     low.genome.mapping,low.genome.methylation,
     low.mbias.methylation,
     file = paste0(rdata.path, "Low_quality_fish.Rdata"))
```

## Allele specific methylation


```{r}
asm.path <- 
  "/datasets/work/ncmi-toa-rrbs/work/1.Toothfish/GENOME/Combined_SNP_CGMAP/"
asm <- data.frame(Chr = NULL, Pos = NULL, Ref = NULL, Allele1 = NULL, Allele2 = NULL,
                  Region_start = NULL, Region_end = NULL, Allele1_linked_C = NULL, 
                  Allele2_linked_C = NULL, Allele1_linked_C_met = NULL, 
                  Allele2_linked_C_met = NULL, pvalue = NULL, fdr = NULL, 
                  ASM = NULL, sample = NULL)

for (i in 1:312) {
  asm.sample <- readr::read_tsv(
      paste0(asm.path, metadata$File_names[i],".asm"), skip = 0, col_names = TRUE)
  asm.sample$sample <- metadata$File_names[i]
  asm <- rbind(asm, asm.sample)
}
asm$ChrPos <- paste0(asm$Chr, "_", asm$Pos)
length(unique(asm$ChrPos[asm$ASM == TRUE])) #12351
asmCHR <- unique(asm$ChrPos[asm$ASM == TRUE])
asmTRUE <- asm[asm$ChrPos %in% asmCHR, ]

save(asm, asmTRUE,file = paste0(rdata.path,"Allele_specific_mutations_raw.Rdata"))
save(asmTRUE,file = paste0(rdata.path,"Allele_specific_mutations.Rdata"))
```


## Genome features annotation

cpgreport: 

As there is no official definition of what is a cpg island is, and worst where they begin and end, we have to live with 2 definitions and thus two methods. These are:

a putative island if the score is higher than a threshold (17 at the moment).


Obs/Exp ratio > 0.6
   % C + % G > 50%
   Length > 200.
   
```{r, eval=FALSE}
chr.file <- "/datasets/work/ncmi-toa-rrbs/work/1.Toothfish/Bismark_genome_index/sizes.chr"

chrom.info <- read.table(file = chr.file, header = F)
chrom.info$V3 <- FALSE
colnames(chrom.info) <- c("chrom", "length", "is_circular")

gtf.file <- "/datasets/work/ncmi-toa-rrbs/work/1.Toothfish/Bismark_genome_index/D.mawsoni_2.gtf"
bed.file <- "/datasets/work/ncmi-toa-rrbs/work/1.Toothfish/Bismark_genome_index/D.mawsoni_2.bed"
cpg.report <- "/datasets/work/ncmi-toa-rrbs/work/1.Toothfish/Bismark_genome_index/cpgreport_out.txt"

cpg.tab <- read.table(cpg.report)
colnames(cpg.tab) <- c("Sequence", "Begin", "End", "Score", "CpG",  "%CG", "CG/GC")
cpg.tab <- cpg.tab[grepl(pattern = "JAAKFY01", x = cpg.tab$Sequence),]

cpg.tab$Score <- as.numeric(cpg.tab$Score)
cpg.tab$CpG <- as.numeric(cpg.tab$CpG)
cpg.tab$`CG/GC` <- as.numeric(cpg.tab$`CG/GC`)
cpg.tab$`%CG` <- as.numeric(cpg.tab$`%CG`)
cpg.tab$Begin <- as.numeric(cpg.tab$Begin)
cpg.tab$End <- as.numeric(cpg.tab$End)
cpg.tab$width <- cpg.tab$End - cpg.tab$Begin

cpg.tab2 <- cpg.tab[cpg.tab$Score > 17 & 
                    cpg.tab$width > 200 & 
                    cpg.tab$`%CG` > 50,]

chrom.info.new <- chrom.info[chrom.info$chrom %in% unique(cpg.tab2$Sequence),]
chrom.seqinfo.new <- GenomeInfoDb::Seqinfo(seqnames = chrom.info.new$chrom, 
                                       seqlengths = chrom.info.new$length, 
                                       isCircular = chrom.info.new$is_circular, 
                                       genome = "Antarctic Toothfish")

cpgi.obj <- GenomicRanges::GRanges(seqnames = cpg.tab2$Sequence, 
                                   ranges = IRanges::IRanges(
                                     start = cpg.tab2$Begin,end = cpg.tab2$End),
                                   seqinfo = chrom.seqinfo.new, 
                                   mcols = cpg.tab2[4:7])



gencode <- txdbmaker::makeTxDbFromGFF(gtf.file,
                format = "auto",
                dataSource = "Annotation From Chen et al. 2021",
                organism = "Dissostichus mawsoni",
                taxonomyId = NA,
                circ_seqs = NULL,
                chrominfo = chrom.info,
                miRBaseBuild = NA,
                metadata = NULL)

# saveDb(gencode, file = paste0(rdata.path, "gencode_D_mawsoni.sqlite"))
gencode <- loadDb(paste0(rdata.path, "gencode_D_mawsoni.sqlite"))

#### Create GRAnges objects for the each feature
genes.gencode <- sort(GenomicFeatures::genes(gencode))
intergenic.gencode <- GenomicRanges::gaps(genes.gencode)
transcript.gencode <- sort(GenomicFeatures::transcripts(gencode, use.names = TRUE))
cds.gencode <- sort(GenomicFeatures::cds(gencode))
exons.gencode <- sort(GenomicFeatures::exons(gencode))
introns.gencode <- sort(unlist(GenomicFeatures::intronsByTranscript(gencode)))
promoters.gencode <- sort(GenomicFeatures::promoters(gencode, upstream = 1000, downstream = 1000))
TSS.gencode <- transcript.gencode
TSS.gencode@ranges@width <- rep(1L, times = length(TSS.gencode@ranges@width))

annotation.all.features <-  GenomicRanges::GRangesList(
  promoters = promoters.gencode, 
  exons = exons.gencode,
  genes = genes.gencode,
  introns = introns.gencode,
  intergenics = intergenic.gencode,
  transcripts = transcript.gencode,
  cds = cds.gencode,
  TSSes = TSS.gencode)


# gff3 <- rtracklayer::import.gff(gtf.file)


save(annotation.all.features, gff3,cpg.tab,cpg.tab2, cpgi.obj,
     file = paste0(rdata.path,"Toothfish_genome_annotation_features.Rdata"))
```

### Gene annotation track

```{r}
options(ucscChromosomeNames = FALSE)
print(load(paste0(rdata.path,"Toothfish_genome_annotation_features.Rdata")))
gencode <- loadDb(paste0(rdata.path, "gencode_D_mawsoni.sqlite"))

gtrack <- Gviz::GenomeAxisTrack()
geneTrack <- Gviz::GeneRegionTrack(gencode,
                                    name = "Annotation", geneSymbol = TRUE,
                                   transcriptAnnotation = "transcript",)

seqnames1 <- gsub(pattern = "JAAKFY01", replacement = "", as.character(cpgi.obj@seqnames))
seqnames2 <- paste0("chr", stringr::str_pad(as.integer(seqnames1), 2, pad = "0"))
seqlevels(cpgi.obj, pruning.mode = "coarse") <- as.character(unique(paste0("JAAKFY01",seqnames1)))
seqlevels(cpgi.obj) <- as.character(unique(seqnames2))
cpgi.track <- Gviz::AnnotationTrack(cpgi.obj,
                                    name = "CpG islands")
geneTrack@dp@pars$fill <- "blue"
geneTrack@dp@pars$col <- NULL
geneTrack@dp@pars$transcriptAnnotation <- "transcript"
plotTracks(geneTrack, chromosome = "chr02", from = 19200000, to = 19268000, showId=FALSE,
           transcriptAnnotation = "transcript")
plotTracks(list(gtrack, geneTrack, cpgi.track), chromosome = "chr02", from = 19200000, to = 19400000, showId=FALSE,
           transcriptAnnotation = "transcript")

plotTracks(list(gtrack, geneTrack, cpgi.track), chromosome = "chr02", from = 19200000, to = 19400000, showId=FALSE,
           transcriptAnnotation = "transcript")

getOption("Gviz.scheme")
## [1] "default"
scheme <- getScheme()
scheme$GeneRegionTrack$fill <- "salmon"
scheme$GeneRegionTrack$col <- NULL
scheme$GeneRegionTrack$transcriptAnnotation <- "transcript"
addScheme(scheme, "myScheme")
options(Gviz.scheme = "myScheme")
grtrack <- GeneRegionTrack(geneModels, genome = gen, transcriptAnnotation = "transcript",
                           chromosome = chr, name = "Gene Model")


 # littleTicks = TRUE

keys(gencode)
symbol(geneTrack) <- mapIds(gencode, 
                            keys=sub("\\.\\d+$", "", gene(geneTrack)), 
                            keytype="ENSEMBL", column="SYMBOL")
symbol(geneTrack) <- ifelse(is.na(symbol(geneTrack)), gene(geneTrack), symbol(geneTrack))


# using the rtracklayer import function I can read the GFF3 from GenCode and create conversion table between gene id and gene symbol
geneTrack <- Gviz::GeneRegionTrack(gff3, chromosome = "chr02"
                 , from = 19200000, to = 22000000,
                                    name = "Annotation")

gene2symbol <- mcols(gff3)[,c("gene_id")] #"gene_name"
gene2symbol <- unique(gene2symbol)
rownames(gene2symbol) <- gene2symbol$gene_id

# the GFF3 from GenCode can be parsed into TxDb object and after creating the GeneRegionTrack I could set the correct gene symbols using the table from previous step
# txdb <- makeTxDbFromGFF("gencode.v24.annotation.gff3.gz")

ranges(geneTrack)$symbol <- gene2symbol[ranges(geneTrack)$gene, "gene_name"]
ranges(geneTrack)$symbol <- ranges(geneTrack)$feature
# plotting with gene symbols
plotTracks(geneTrack, chromosome = "chr02", from = 19230000, to = 19250000, showId=FALSE,
           transcriptAnnotation = "symbol")

z <- ranges(geneTrack)
z$symbol <- mapIds(gencode, z$symbol, "SYMBOL", "TXNAME")
# 'select()' returned 1:1 mapping between keys and columns
ranges(gene) <- z
plotTracks(gene)


seqnames1 <- gsub(pattern = "JAAKFY01", replacement = "", as.character(gencode@.xData$user_seqlevels))
seqnames2 <- paste0("chr", stringr::str_pad(as.integer(seqnames1), 2, pad = "0"))
# seqlevels(annotation.genes, pruning.mode = "coarse") <- as.character(unique(paste0("JAAKFY01",seqnames1)))
gencode@.xData$user_seqlevels <- as.character(unique(seqnames2))

annot.track <- Gviz::GeneRegionTrack(gencode,
                                    name = "Annotation")
annot.track <- Gviz::AnnotationTrack(gencode,
                                    name = "Annotation")
Gviz::plotTracks(annot.track, chromosome = "chr02"
                 , from = 19200000, to = 22000000
                 )

df <- as.data.frame(annotation.all.features)
df[df$group_name == "genes",]
df$row <- as.integer(IRanges(1L, width = lengths(annotation.all.features)))
wide <- reshape(df[c("row", "group_name", "score")], direction = "wide", 
                timevar = "group_name", idvar = "row")

annotation.genes <- annotation.all.features$genes
seqnames1 <- gsub(pattern = "JAAKFY01", replacement = "", as.character(annotation.genes@seqnames))
seqnames2 <- paste0("chr", stringr::str_pad(as.integer(seqnames1), 2, pad = "0"))
seqlevels(annotation.genes, pruning.mode = "coarse") <- as.character(unique(paste0("JAAKFY01",seqnames1)))
seqlevels(annotation.genes) <- as.character(unique(seqnames2))
gene.track <- Gviz::AnnotationTrack(annotation.genes,
                                    name = "genes")


cyto.bands <- data.frame(chrom = rep("chrI", 4),
                         chromStart = c(0, 148071, 151524, 154977),
                         chromEnd = c(148071, 151524, 154977, 230218),
                         name = c(NA, "CEN1", "CEN1", NA),
                         gieStain = c("gneg", "acen", "acen", "gneg"))

ideogram.track <- IdeogramTrack(chromosome = "chrI", genome = "sacCer3", bands = cyto.bands)
data(geneModels)

```


### Gene ontology 

```{r, eval=FALSE}
library(GO.db)
GO <- as.list(GOTERM)
data1 <- readr::read_csv(
  file = 
    "/datasets/work/ncmi-toa-rrbs/work/1.Toothfish/Annotation_GO/go_enrichment/06_fisher_tests/all_go_annotations.csv",
  col_names = FALSE)
GOids <- strsplit(x = data1$X2, split = ";")

go_df <- data.frame(gene = NULL, go_id = NULL, go_name = NULL, root_nod = NULL)
for (i in 1:nrow(data1)) {
  print(i)
  dat <- GOfuncR::get_names(GOids[[i]], term_df = NULL, godir = NULL)
  go_df <- rbind(go_df, 
                 data.frame(gene = data1$X1[i], dat))
}

print(load(paste0(rdata.path,"Toothfish_genome_annotation_features.Rdata")))
gff3.df <- as.data.frame(gff3)
transID <- stringr::str_split(string = gff3.df$transcript_id, pattern = "\\|", n = 3, simplify = TRUE)
gff3.df$gene <- transID[,3]
sum(go_df$gene %in% transID) #314069

gff3.df <- gff3.df %>% dplyr::left_join(go_df, by = "gene")
gff3 <- as(gff3.df, "GRanges")

save(go_df,gff3.df,gff3, file = paste0(rdata.path, "Gene_ontology_terms.Rdata"))
```


```{r}
print(load(paste0(rdata.path, "Gene_ontology_terms.Rdata")))
length(unique(go_df$gene)) #26677 unique genes
go_df2 <- dplyr::group_by(go_df, root_node) %>%
  dplyr::count()
ggplot2::ggplot(go_df2) +
  ggplot2::geom_point(ggplot2::aes(x = root_node, y = n)) +
  ggplot2::theme_bw() +
  ggplot2::theme(axis.text.x = ggplot2::element_text(angle = 90))


```




# 5_Methylation: 2018 - muscle tissue

```{r}
print(load(paste0(rdata.path,"Toothfish_metadata.Rdata"))) #ALL fish
```

## Reading the methylation call files

```{r, eval=FALSE}
data.list <- list.files(
  path = "/datasets/work/ncmi-toa-rrbs/work/1.Toothfish/GENOME/Combined_methylation/", 
  pattern = "*.cov.gz")

subnames <- metadata$File_names[!(metadata$Year %in% 2024 | metadata$Tissue_type %in% c("fin", "Skin") |
                                    grepl(pattern = "REP", x = metadata$File_names))]
data.list <- data.list[data.list %in% paste0(subnames,
                                             "_R1_trimmed_sort_bismark_bt2.bismark.cov.gz")]

data.loc <- as.list(paste0(
  "/datasets/work/ncmi-toa-rrbs/work/1.Toothfish/GENOME/Combined_methylation/",
                           data.list))



data.list2 <-  gsub(pattern = '_R1_trimmed_sort_bismark_bt2.bismark.cov.gz',
                    replacement = '',x = as.character(data.list))
# data.list2 <-  gsub(pattern = '_sort_bismark_bt2_PE.bismark.cov.gz',
#                     replacement = '',x = as.character(data.list2))

metadata.new <- metadata[metadata$File_names %in% data.list2,]
metadata.new <- metadata.new[order(match(metadata.new$File_names, data.list2)),]

metadata.new$File_names == data.list2

sample.ids <- metadata.new$Genotype_names
pop <- as.integer(factor(metadata.new$pop))


myobjDB <- methylKit::methRead(location = data.loc,
           sample.id = as.list(sample.ids),
           assembly = "TOA_genome",
           treatment = pop,
           context = "CpG",
           header = FALSE, 
           dbtype = "tabix",
           pipeline = "bismarkCoverage",
           mincov = 3
           )
save(myobjDB, file = paste0(rdata.path,"Toothfish_2018_muscle_methylkit_cpgDB.Rdata"))
```


## Data QC

#### Number of sites in genome

```{r, eval=FALSE}
cov.df <- data.frame(chr = NA, start = NA,end = NA, strand = NA, coverage = NA, 
                     numCs = NA, numTs = NA, sample = NA)
for (i in 1:length(sample.ids)) {
  df <- methylKit::getData(myobjDB[[i]])
  df$sample <- sample.ids[i]
  cov.df <- rbind(cov.df, df)
}
cov.df <- cov.df[-1,]
save(cov.df, file = paste0(rdata.path,"Toothfish_2018_muscle_methylkit_cpg_cov.Rdata"))
```

### Coverage stats

```{r}
load(paste0(rdata.path,"Toothfish_2018_muscle_methylkit_cpgDB.Rdata"))
print(load(paste0(rdata.path,"Toothfish_2018_muscle_methylkit_cpg_cov.Rdata")))

length(unique(paste0(cov.df$chr, "_", cov.df$start))) #8,147,303

# summary(as.factor(cov.df$sample))
summary(cov.df$coverage)
    # Min.  1st Qu.   Median     Mean  3rd Qu.     Max.     NA's 
    # 4.00     7.00    15.00    25.22    33.00 81760.00        1

summary(cov.df$coverage[cov.df$coverage > 5 & cov.df$coverage < 50])
   # Min. 1st Qu.  Median    Mean 3rd Qu.    Max.    NA's 
   # 6.00    9.00   16.00   19.68   28.00   49.00       1 

cov.plot <- ggplot2::ggplot(cov.df, ggplot2::aes(x = log(coverage), fill = sample)) +
  ggplot2::geom_histogram() +
  ggplot2::theme(legend.position = "none")
# print(cov.plot)
ggplot2::ggsave(cov.plot,
                filename =
                  paste0(figure.path,"Toothfish_2018_muscle_coverage_hist_Combined.png"), 
                width = 45, height = 15, units = 'cm')

cov.plot <- ggplot2::ggplot(cov.df, ggplot2::aes(x = coverage, fill = sample)) +
  ggplot2::geom_histogram() +
  ggplot2::theme(legend.position = "none")
# print(cov.plot)
ggplot2::ggsave(cov.plot,
                filename =
                  paste0(figure.path,"Toothfish_2018_muscle_coverage_hist_Combined2.png"), width = 45, height = 15, units = 'cm')

cov.plot <- ggplot2::ggplot(cov.df[cov.df$coverage > 5 & cov.df$coverage < 160,], 
                            ggplot2::aes(x = coverage)) +
  ggplot2::geom_histogram(binwidth = 1) +
  ggplot2::ggtitle("coverage < 160") +
  ggplot2::theme(legend.position = "none")
# print(cov.plot)
ggplot2::ggsave(cov.plot,
                filename =
                  paste0(figure.path,"Toothfish_2018_muscle_coverage_hist_Combined4.png"),
                width = 45, height = 15, units = 'cm')



max(cov.df$coverage, na.rm = T) #81760
ind.covplot <- ggplot2::ggplot(cov.df, ggplot2::aes(x = sample, y = coverage)) + 
  ggplot2::geom_boxplot() +
  ggplot2::theme_bw() + 
  ggplot2::theme(axis.title.x = ggplot2::element_blank(), 
                 axis.text.y = ggplot2::element_text(size = 15),
                 axis.title.y = ggplot2::element_text(size = 15),
                 axis.text.x = ggplot2::element_text(size = 7, angle = 90, hjust = 1)) +
  ggplot2::scale_y_continuous(trans = scales::log2_trans(),
                  breaks = c(1, 10, 50, 750, 100000),
                  minor_breaks = c(1:10, 2:5*10, 1:7.5*100, 10:80*1000))

print(ind.covplot)

ggplot2::ggsave(ind.covplot, bg = "white",
                filename = 
                  paste0(figure.path,
                         "Toothfish_2018_muscle_raw_individual_coverage_CpGs.png"),
                width = 45, height = 15, units = 'cm')
```


### Coverage filter

High count filter based on coverage stats.

```{r, eval=FALSE}
load(paste0(rdata.path,"Toothfish_2018_muscle_methylkit_cpgDB.Rdata"))

filtered.myobj <- methylKit::filterByCoverage(myobjDB, lo.count = 10, lo.perc = NULL,
                                      hi.count = NULL, hi.perc = 99)
save(myobjDB,filtered.myobj, 
     file = paste0(rdata.path,"Toothfish_2018_muscle_methylkit_cpgDB.Rdata"))
```

### Normalise coverage

Normalize read coverage distributions between samples. Requires stringent filter on max coverage first, otherwise it will over-sample PCR bias reads. These two functions will help reduce the bias in the statistical tests that might occur due to systematic over-sampling of reads in certain samples.

```{r}
filtered.myobj.normal <- methylKit::normalizeCoverage(obj = filtered.myobj, 
                                                      method = "median", chunk.size = 1e6)
save(myobjDB,filtered.myobj, filtered.myobj.normal,
     file = paste0(rdata.path,"Toothfish_2018_muscle_methylkit_cpgDB.Rdata"))
```

### Number of sites per sample

```{r, eval = FALSE}
sites.df <- data.frame(sites = NA,  sample = NA)
for (i in 1:length(filtered.myobj.normal)) {
  print(i)
  sites.df[i,] <- c(filtered.myobj.normal[[i]]@num.records, filtered.myobj.normal[[i]]@sample.id)
}
sites.df$sites <- as.integer(sites.df$sites)
# sites.df$pop <- filtered.myobj.normal@treatment
sites.df$pop <- metadata.new$pop

save(sites.df, file = paste0(rdata.path,"Toothfish_2018_muscle_number_cpg_sample.Rdata"))
```


```{r}
print(load(paste0(rdata.path,"Toothfish_2018_muscle_number_cpg_sample.Rdata")))
sites.plot <- ggplot2::ggplot(sites.df, ggplot2::aes(x = sample, y = sites, colour = pop)) +
  ggplot2::geom_point() +
  ggplot2::scale_color_manual(values = colours4) +
  ggplot2::theme_bw() +
  ggplot2::theme(axis.text.x = ggplot2::element_text(angle = 90))
print(sites.plot)

ggplot2::ggsave(sites.plot,
                filename =
                  paste0(figure.path,"Toothfish_2018_muscle_number_cpg_sample.png"),
                width = 45, height = 15, units = 'cm')

sites.df$sample[sites.df$sites < 800000] #"TOA_58-4-2_088_AN170741"
low.sites <- sites.df$sample[sites.df$sites < 600000]
```


## Select samples

```{r}
print(load(paste0(rdata.path,"Toothfish_2018_muscle_methylkit_cpgDB.Rdata")))
print(load(paste0(rdata.path,"Toothfish_metadata.Rdata")))

metadata2 <- metadata[!(metadata$Year %in% 2024 | metadata$Tissue_type %in% 
                                    c("fin", "Skin")|
                                    grepl(pattern = "REP", x = metadata$File_names)),]

print(load(paste0(rdata.path, "Low_quality_fish.Rdata")))

bad.list <- list(high.lambda.mapping,
          low.genome.mapping,low.genome.methylation,
          low.mbias.methylation)
bad.plot <- ggVennDiagram::ggVennDiagram(
  x = bad.list, label_alpha = 0, set_color = c("black","black","black","black"),
  set_size = 7, label_size = 5,
  category.names = c("high.lambda.mapping",
          "low.genome.mapping","low.genome.methylation",
          "low.mbias.methylation")) +
  ggplot2::scale_fill_gradient2(low = "blue", mid = "yellow",high = "red", midpoint = 150) + 
  ggplot2::scale_color_manual(values = c("black","black","black","black")) +
    ggplot2::theme(
    legend.text = ggplot2::element_text(size = 20),
    legend.title = ggplot2::element_text(size = 30)
  )
print(bad.plot)


rm1 <- unique(c(low.genome.mapping,low.genome.methylation, low.sites))
print(rm1)
rm2 <- c("TOA_88-1_351", "TOA_58-4-1_115_AN177878", "TOA_58-4-1_277_AN177259",
         "TOA_58-4-1_289_AN177617", "TOA_58-4-1_290_AN177621",
         "TOA_58-4-1_295_AN177901", "TOA_58-4-2_111_AN172604") # bad PCA
rm3 <- c("TOA_88-2_P79_AD294619_fin_F", "TOA_58-4-2_088")
print(rm3) # bad coverage
rm4 <- c("TOA_48-2_031_UK116","TOA_58-4-2_111_AN172604","TOA_88-1_230_AD082543",
         "TOA_48-2_024_UK082")
print(rm4) #bad tissue PCA

rm <- c(rm1,rm2,rm3,rm4) # bad coverage

metadata2 <- metadata2[!(metadata2$Genotype_names %in% rm | metadata2$File_names %in% rm),]

filt.sample.id.full <- sapply(filtered.myobj@.Data, function(x){x@sample.id})

# filt.sample.id <- filt.sample.id.full[filt.sample.id.full %in% metadata2$File_names]
# metadata2 <- metadata2[order(match(metadata.new$File_names, filt.sample.id)),]

filt.sample.id <- filt.sample.id.full[filt.sample.id.full %in% metadata2$Genotype_names]
metadata2 <- metadata2[order(match(metadata2$Genotype_names, filt.sample.id)),]
identical(filt.sample.id,metadata2$Genotype_names)


pop <- metadata2$pop
pop.levels <- c("58.4.1-2018","58.4.2-2018",
                "58.5.2-2017", "88.1-2018")

pop <- factor(pop, levels = pop.levels)
pop <- as.numeric(pop)
sample.ids <- metadata2$Genotype_names



filtered.myobj2.normal <- methylKit::reorganize(
  methylObj = filtered.myobj.normal,
  sample.ids = sample.ids,
  treatment = pop,
  chunk.size = 1e+06,
  save.db = TRUE,
  suffix = "TOA_2018_muscle_normal"
)
save(myobjDB,filtered.myobj, filtered.myobj.normal,filtered.myobj2.normal,
     file = paste0(rdata.path,"Toothfish_2018_muscle_methylkit_cpgDB.Rdata"))
```


## Merging samples

```{r}
print(load(paste0(rdata.path,"Toothfish_2018_muscle_methylkit_cpgDB.Rdata")))
pop <- metadata2$pop
pop.levels <- c("58.4.1-2018","58.4.2-2018",
                "58.5.2-2017", "88.1-2018")
pop <- as.numeric(factor(pop, levels = pop.levels))
table(pop)


meth.2018M.norm11L <- methylKit::unite(filtered.myobj2.normal, destrand = FALSE,
                                       save.db = FALSE,
                                       min.per.group = 11L, #missing data allowed
                                       mc.cores = 25)

#1251500  common sites for 11 samples per group

save(meth.2018M.norm11L,
     file = paste0(rdata.path,"Toothfish_2018_muscle_methylkit_object.Rdata"))
```


## Data QC
### Remove ASM

```{r}
load(paste0(rdata.path,"Allele_specific_mutations.Rdata"))

mut <- as(asmTRUE,"GRanges")
meth.2018M.norm11L <- as(meth.2018M.norm11L,"methylBase") 


sum(as(meth.2018M.norm11L,"GRanges") %over% mut)#38220
meth.asm <- methylKit::selectByOverlap(meth.2018M.norm11L, mut) 
BOOLEAN <- paste0(meth.2018M.norm11L$chr,"_",meth.2018M.norm11L$start) %in% paste0(asmTRUE$Chr,"_",asmTRUE$Pos) #1
sum(BOOLEAN) #727
meth.2018M.norm11L <- methylKit::select(meth.2018M.norm11L, !BOOLEAN)

save(meth.2018M.norm11L,
     file = paste0(rdata.path,"Toothfish_2018_muscle_methylkit_object.Rdata"))
```


### Coverage 

```{r}
print(load(paste0(rdata.path,"Toothfish_2018_muscle_methylkit_object.Rdata")))

meth.df <- methylKit::getData(meth.2018M.norm11L)
cov <- meth.df[,meth.2018M.norm11L@coverage.index]
colnames(cov) <- meth.2018M.norm11L@sample.ids
dpf <- data.frame(cov, locus = paste0(meth.df$chr,"_",meth.df$start))
dpf <- tidyr::gather(dpf, key = individual, 
                           value = Depth, -locus,
                           factor_key = TRUE) %>%
  dplyr::group_by(individual) %>%
  dplyr::mutate(NA_cnt = sum(is.na(Depth))/length(unique(locus))) %>%
  dplyr::ungroup()
dpf.NA <- dpf[!duplicated(dpf$individual),]
dpf.NA$NA_cnt <- round(dpf.NA$NA_cnt, 3)

dpf <- dpf[ dpf$Depth > 0 & !is.na(dpf$Depth),]

max(dpf$Depth, na.rm = T) #570
myPlot <- ggplot2::ggplot(dpf, ggplot2::aes(x = individual, y = Depth)) + 
  ggplot2::geom_boxplot() +
  ggplot2::geom_text(data = dpf.NA,ggplot2::aes(x = individual, y = 150, label = NA_cnt), 
                     size = 2, angle = 90) +
  ggplot2::theme_bw() + 
  ggplot2::theme(axis.title.x = ggplot2::element_blank(), 
                 axis.text.y = ggplot2::element_text(size = 15),
                 axis.title.y = ggplot2::element_text(size = 15),
                 axis.text.x = ggplot2::element_text(size = 7, angle = 90, hjust = 1)) +
  ggplot2::scale_y_continuous(trans = scales::log2_trans(),
                  breaks = c(1, 10, 50, 100, 600),
                  minor_breaks = c(1:10, 2:5*10, 1:5*100))

print(myPlot)
ggplot2::ggsave(myPlot, bg = "white",
                filename = 
                  paste0(figure.path,
                  "1B.Toothfish_2018_muscle_normalised15_methylkit_individual_coverage_CpGs.png"
                                  ), 
                width = 45, height = 15, units = "cm")

nrow(cov) # 974309 SMPs
plot(x = 1:ncol(cov), y = colMeans(cov, na.rm = TRUE), col = meth.2018M.norm11L@treatment)
plot(x = 1:ncol(cov), y = colMeans(cov > 20, na.rm = TRUE), col = meth.2018M.norm11L@treatment)
plot(x = 1:ncol(cov), y = colMeans(cov > 10, na.rm = TRUE), col = meth.2018M.norm11L@treatment)
meth.2018M.norm11L@sample.ids[colMeans(cov > 10, na.rm = TRUE) < 0.85]
# samples with low coverage: "TOA_58-4-2_088_AN170741"
```


## Load data

```{r}
print(load(paste0(rdata.path,"Toothfish_2018_muscle_methylkit_object.Rdata")))
print(load(paste0(rdata.path,"Toothfish_metadata.Rdata"))) 
metadata.new <- metadata[metadata$Genotype_names %in% meth.2018M.norm11L@sample.ids,]
metadata.new <- metadata.new[order(match(metadata.new$Genotype_names,
                                         meth.2018M.norm11L@sample.ids)),]

length(meth.2018M.norm11L@row.names) #number of sites
```


## Global clustering analysis

```{r}
methylKit::clusterSamples(meth.2018M.norm11L, dist = "correlation", sd.filter = TRUE, 
                          method = "ward.D2", plot = TRUE, )
dev.print(file = paste0(figure.path,"1_TOA_2018_muscle_cluster_plot.png"),
          device = png, res = 300,width = 45,height = 15,units = "cm")
```


### PCA

```{r}
PCobj <- methylKit::PCASamples(meth.2018M.norm11L, comp = c(1,2), 
                               adj.lim = c(0,0), obj.return = TRUE)

var_frac <- PCobj$sdev/sum(PCobj$sdev)
print(paste0("Total variance explained: ",signif(sum(var_frac[1:3]) * 100, 3), "%")) 
barplot(var_frac[1:20], main = "variance explained", 
        col = heat.colors(length(var_frac[1:20])))

identical(metadata.new$Genotype_names, meth.2018M.norm11L@sample.ids)
identical(metadata.new$Genotype_names, rownames(PCobj$x))

pop <- metadata.new$pop

plot.title <- "Toothfish methylation - 2018 muscle (1,250,773 CpG sites)"

data <- data.frame(ind = rownames(PCobj$x),PCobj$x,  pop = pop)
# data <- data.frame(ind = rownames(PCobj$rotation),PCobj$rotation, pop = pop)
data <- merge(data, aggregate(cbind(mean.x = PC1, mean.y = PC2) ~
                                pop, data, mean), by = "pop")

pca.plot <- ggplot2::ggplot(data, ggplot2::aes(PC1, PC2, color = pop)) +
  ggplot2::geom_point(size = 4) + 
  ggplot2::geom_point(ggplot2::aes(x = mean.x,
                                   y = mean.y, color = pop),
                      size = 7,
                      shape = 17) +
  ggplot2::scale_colour_manual(values = colours4) +
  ggplot2::geom_hline(yintercept = 0) +
  ggplot2::geom_vline(xintercept = 0) + 
  ggplot2::labs(
    subtitle = paste0("Total variance explained: ",
                      signif(sum(var_frac[1:4]) * 100, 3), "%"),
    y = paste0("PC2; ", signif(var_frac[2] * 100, digits = 2), " %"),
    x = paste0("PC1; ", signif(var_frac[1] * 100, digits = 2), " %"),
    title = plot.title,
    caption = "") + 
  ggplot2::theme_minimal() +
  ggplot2::theme(
    axis.text = ggplot2::element_text(size = 15),
    axis.title.x = ggplot2::element_text(size = 20),
    axis.title.y = ggplot2::element_text(size = 20),
    legend.text = ggplot2::element_text(size = 15)
  )

print(pca.plot)
ggplot2::ggsave(pca.plot, bg = "white",
                filename = paste0(figure.path,"Toothfish_2018_muscle_PCA_CpG_PC12.png"), 
                width = 30, height = 15, units = "cm")


data <- data.frame(ind = rownames(PCobj$x),PCobj$x,  pop = pop)
# data <- data.frame(ind = rownames(PCobj$rotation),PCobj$rotation, pop = pop)
data <- merge(data, aggregate(cbind(mean.x = PC2, mean.y = PC3) ~
                                pop, data, mean), by = "pop")
pca.plot <- ggplot2::ggplot(data, ggplot2::aes(PC2, PC3, color = pop)) +
  ggplot2::geom_point(size = 4) + 
  ggplot2::geom_point(ggplot2::aes(x = mean.x,
                                   y = mean.y, color = pop),
                      size = 7,
                      shape = 17) +
  ggplot2::scale_colour_manual(values = colours4) +
  ggplot2::geom_hline(yintercept = 0) +
  ggplot2::geom_vline(xintercept = 0) + 
  ggplot2::labs(
    subtitle = paste0("Total variance explained: ",
                      signif(sum(var_frac[1:4]) * 100, 3), "%"),
    y = paste0("PC3; ", signif(var_frac[3] * 100, digits = 2), " %"),
    x = paste0("PC2; ", signif(var_frac[2] * 100, digits = 2), " %"),
    title = plot.title,
    caption = "") + 
  ggplot2::theme_minimal() +
  ggplot2::theme(
    axis.text = ggplot2::element_text(size = 15),
    axis.title.x = ggplot2::element_text(size = 20),
    axis.title.y = ggplot2::element_text(size = 20),
    legend.text = ggplot2::element_text(size = 15)
  )

print(pca.plot)
ggplot2::ggsave(pca.plot, bg = "white",
                filename = paste0(figure.path,"Toothfish_2018_muscle_PCA_CpG_PC23.png"), 
                width = 30, height = 15, units = "cm")

# data$ind[data$PC2 > 200 | data$PC3 > 100]
# "TOA_58-4-1_115_AN177878" "TOA_58-4-1_277_AN177259" "TOA_58-4-1_289_AN177617" "TOA_58-4-1_290_AN177621"
# "TOA_58-4-1_295_AN177901" "TOA_58-4-2_111_AN172604"
```


### RDA

```{r}
# setwd("/datasets/work/ncmi-toa-rrbs/work/1.Toothfish/Toothfish_RRBS_array_22098078/Bismark/")
# print(load(paste0(rdata.path,"Toothfish_methylation_objects_raw.Rdata"))) #36,635 SMPs
print(load(paste0(rdata.path,"Toothfish_subset_methylation_objects_raw.Rdata"))) #1,316,799 SMPs

# meth.RDA <- methylKit::unite(filtered.myobj2, destrand = FALSE,
#                              save.db = FALSE,
#                              suffix = "subset",
#                              min.per.group = NULL, #missing data allowed
#                              mc.cores = 25)


# meth.df <- methylKit::getData(meth)
# cov <- meth.df[,grep(colnames(meth.df), pattern = "coverage*")]
# cov <- meth.df[,meth@coverage.index]
# colnames(cov) <- meth@sample.ids
# meth.df <- methylKit::percMethylation(meth)

# meth.df <- methylKit::percMethylation(meth.RDA)

# save(meth.df,meth.RDA, 
#      file = paste0(rdata.path,"TOA_data_for_RDA_meth_percent.Rdata"))
print(load(paste0(rdata.path,"TOA_data_for_RDA_meth_percent.Rdata"))) #347,614 SMPs

hist(colSums(is.na(meth.df)), breaks = 100)
hist(rowSums(is.na(meth.df)), breaks = 100)





print(load(paste0(rdata.path,"Toothfish_metadata.Rdata")))

predictor <- metadata2 # define the predictor input file here
predictor <- predictor[predictor$SAMPLE_ID %in% meth@sample.ids,]
predictor$SAMPLE_ID = as.character(predictor$SAMPLE_ID)
predictor$region = as.character(predictor$region)
predictor$Sex = as.character(predictor$Sex)
predictor$Tissue = as.character(predictor$Tissue)
predictor$mat = as.character(predictor$mat)
predictor$mat[is.na(predictor$mat)] <- "2"
predictor$TL = as.numeric(predictor$TL)
predictor$wt = as.numeric(predictor$wt)
predictor$dpth = as.numeric(predictor$dpth)
predictor$Age = as.numeric(predictor$Age)
predictor$Age[is.na(predictor$Age)] <- as.numeric(max(predictor$Age, na.rm = TRUE))
predictor$Cohort2 = as.integer(predictor$Cohort2)
predictor$Cohort2[is.na(predictor$Cohort2)] <- as.integer(min(predictor$Cohort2, na.rm = TRUE))


meth.df1 = t(meth.df)
# meth.df2 = na.omit(meth.df1) #rda can't handle NAs, remove them if you have them
# meth.df1[1:10,1:10]
## Predictor
# Make names characters (not factors)


#if you had numbers, they had to be factors (as.factor)

# Confirm that methylation data and predictors are in the same order
identical(rownames(meth.df1), predictor$SAMPLE_ID)

# Perform Hellinger transformation on the methylation data:
meth.df3 = vegan::decostand(meth.df1, method = "hellinger", MARGIN = 1) #MARGIN=1 means rows, on rows I have samples now because I transposed the matrix

## Run RDA
#Test the country using growing condition as a Condition () (covariate):
meth_rda = vegan::rda(meth.df3 ~ region + Condition(Sex), 
                      data = predictor, scale = F) #I remove scale since I performed Hellinger transformation
meth_rda = vegan::rda(meth.df3 ~ region + Condition(Sex) + 
                        Condition(mat)+ Condition(TL) + Condition(wt) +
                        Condition(dpth) + Condition(Age) + Condition(Cohort2), 
                      na.action = "na.omit",
                      data = predictor, scale = F)

meth_rda$pCCA
meth_rda$CCA$wa

summary(meth_rda) #Constrained Proportion: variance of Y explained by X, Unconstrained Proportion: unexplained variance in Y
# sink(paste0(rdata.path,"anova_rda.txt")) #with this you can save the output of the anova in a txt file, define the output directory
meth.aov <- anova(meth_rda, test = "Chi", permutations = 99) #run Anova on the result to see the significance level
# meth.aov.cca <- anova.cca(meth_rda, test = "Chi", permutations = 99)
print(meth.aov)
# sink(paste0(rdata.path,"anova_rda_chi.txt"))

# Model: rda(formula = meth.df3 ~ region + Condition(Sex), data = predictor, scale = F)
#           Df  Variance      F Pr(>F)   
# Model      3 0.0003924 1.4126  0.004 **
# Residual 110 0.0101861

# Model: rda(formula = meth.df3 ~ region + Condition(Sex) + Condition(mat) + Condition(TL) + Condition(wt) + Condition(dpth) + Condition(Age) + Condition(Cohort2), data = predictor, scale = F, na.action = "na.omit")
#           Df  Variance     F Pr(>F)   
# Model      3 0.0003412 1.257   0.01 **
# Residual 101 0.0091396 

Pop.names <- predictor$region
plot.a <- vegan::ordiplot(meth_rda, choices=c(1,2))
Pop.ellipses <- vegan::ordiellipse(plot.a, groups=Pop.names, display="sites", kind="sd")
dev.print(file = paste0(figure.path,"TOA_RDA_ordiplot_complex_model.png"),
          device = png, res = 300,width = 45,height = 15,units = "cm")

# Type 1 scaling
vegan::ordiplot(meth_rda, scaling = 1, type = "text")
# Type 2 scaling
vegan::ordiplot(meth_rda, scaling = 2, type = "text")

ggvegan::autoplot(
  meth_rda,
  axes = c(1, 2),
  geom = "text",
  layers = c("species", "sites", "biplot", "centroids"),
  arrows = TRUE)

#CpG contribution
hist(meth_rda$CCA$v[,1])
hist(meth_rda$CCA$v[,2])
hist(meth_rda$CCA$v[,3])



#Test the interaction of your predictors:
meth_rda3 = vegan::rda(meth ~ region * Sex + Condition(region + Sex), data = predictor, scale = F)
summary(meth_rda3)
sink(paste0(rdata.path,"anova_rda_interaction.txt"))
print(anova(meth_rda3, test="Chi"))
sink(paste0(rdata.path,"anova_rda_interaction_chi.txt"))
```

#### plot RDA

```{r}

BioR.theme <- theme(
        panel.background = element_blank(),
        panel.border = element_blank(),
        panel.grid = element_blank(),
        axis.line = element_line("gray25"),
        text = element_text(size = 12),
        axis.text = element_text(size = 10, colour = "gray25"),
        axis.title = element_text(size = 14, colour = "gray25"),
        legend.title = element_text(size = 14),
        legend.text = element_text(size = 14),
        legend.key = element_blank())

plot.a <- vegan::ordiplot(meth_rda, choices=c(1,2))
Pop.names <- predictor$region

# sites1 <- BiodiversityR::sites.long(plot.a, env.data=predictor)
sites1 <- data.frame(cbind(predictor, axis1=plot.a$sites[, 1], 
                           axis2=plot.a$sites[, 2], 
                           labels=rownames(plot.a$sites)),
                     group = predictor$region)

# axis1 <- BiodiversityR::axis.long(meth_rda, choices=c(1, 2))
eigs.all <- meth_rda$CA$eig
eigs <- round(100 * eigs.all / meth_rda$tot.chi, digits=1)[c(1, 2)] 
xlab.label <- paste0(names(eigs)[1], " (", eigs[1], "%)")
ylab.label <- paste0(names(eigs)[2], " (", eigs[2], "%)")
axis1 <- data.frame(axis=c(1:2), 
                    ggplot=c("xlab.label", "ylab.label"), 
                    label=c(xlab.label, ylab.label))


# centroids.Pop1 <- BiodiversityR::centroids.long(sites1, 
#                                                 grouping=Pop.names, 
#                                                 centroids.only=TRUE)
# centroids.Pop2 <- BiodiversityR::centroids.long(sites1, 
#                                                 grouping=Pop.names, 
#                                                 centroids.only=FALSE)
gr.name <- rlang::enexpr(Pop.names)
cent.means <-  stats::aggregate(cbind(axis1, axis2) ~ Pop.names,
                                data = sites1,
                                FUN = mean)
names(cent.means) <- c("group", "axis1c", "axis2c")
centroids.Pop1 <- cent.means
cent.means$Centroid <- cent.means[, 1]
by.name <- names(sites1)[names(sites1) == "group"]
centroids.Pop2 <- dplyr::full_join(sites1, cent.means, by = by.name)


# Pop.ellipses.long2 <- BiodiversityR::ordiellipse.long(Pop.ellipses, 
#                                                     grouping.name="Pop.names")


plotgg.a <- ggplot() + 
    geom_vline(xintercept = c(0), color = "grey70", linetype = 2) +
    geom_hline(yintercept = c(0), color = "grey70", linetype = 2) +  
    xlab(axis1[1, "label"]) +
    ylab(axis1[2, "label"]) +  
    scale_x_continuous(sec.axis = dup_axis(labels=NULL, name=NULL)) +
    scale_y_continuous(sec.axis = dup_axis(labels=NULL, name=NULL)) +
    # geom_polygon(data=Pop.ellipses.long2, 
    #                aes(x=axis1, y=axis2, colour=Pop.names, 
    #                    fill=after_scale(alpha(colour, 0.2))), 
    #           size=0.2, show.legend=TRUE) +
    geom_segment(data=centroids.Pop2, 
                 aes(x=axis1c, y=axis2c, xend=axis1, yend=axis2, colour=Pop.names), 
                 size=0.7, show.legend=FALSE) +
    geom_point(data=centroids.Pop1, 
               aes(x=axis1c, y=axis2c), 
               shape="square", colour="black", alpha=0.7, size=2) +
    geom_label_repel(data=centroids.Pop1, 
               aes(x=axis1c, y=axis2c, label=group, colour=group), 
               alpha=1.0, size=2, show.legend=FALSE) +
    # labs(colour = Pop.names) +
    BioR.theme +
    theme(legend.text = element_text(size = 10)) +
    scale_colour_npg() +
    coord_fixed(ratio=1)

plotgg.a

ggsave(plotgg.a, file = paste0(figure.path, 
                               # "TOA_RDA_simple_model_A.png"
                               "TOA_RDA_complex_model_A.png"
                               ),
       height = 30, width = 30, units = "cm",  device = "png")
```



## DMS - 2018 - muscle

```{r, eval=FALSE}
print(load(paste0(rdata.path,"Toothfish_2018_muscle_methylkit_object.Rdata")))

print(load(paste0(rdata.path,"Toothfish_metadata.Rdata")))
metadata.new <- metadata[metadata$Genotype_names %in% attributes(meth.2018M.norm11L)$sample.ids,]
attributes(meth.2018M.norm11L)$sample.ids == metadata.new$Genotype_names

meth.2018M.norm11L@treatment <- as.integer(factor(metadata.new$pop))

# data.frame(attributes(meth.2018M.norm11L)$sample.ids, metadata.new$Genotype_names, metadata.new$strata, meth.2018M.norm11L@treatment)

covariates <- dplyr::select(metadata.new, Sex, Age, Cohort2, Maturity_full)
myDiff <- methylKit::calculateDiffMeth(meth.2018M.norm11L,
                                       covariates = NULL,
                                       adjust = "fdr",
                                       chunk.size = 1e+06,
                                       mc.cores = 25)
meth.diff.chr <- methylKit::diffMethPerChr(myDiff,plot = TRUE, 
                                           qvalue.cutoff = 0.01, meth.cutoff = 40)
myDiff40p <- methylKit::getMethylDiff(myDiff, difference = 40, qvalue = 0.01)
#35
myDiff25p <- methylKit::getMethylDiff(myDiff, difference = 25, qvalue = 0.01)
#1202  
myDiff15p <- methylKit::getMethylDiff(myDiff, difference = 15, qvalue = 0.01)
#11188 
top100.2018M <- slice_max(as.data.frame(as(myDiff,"GRanges")), order_by = meth.diff, n = 100)
top100.2018M <- as(top100.2018M,"GRanges")

save(myDiff,myDiff15p, myDiff25p, myDiff40p,top100.2018M,
     file = paste0(rdata.path,"Toothfish_2018_muscle_methylation_DMS.Rdata"))
```


```{r}
print(load(paste0(rdata.path,"Toothfish_2018_muscle_methylation_DMS.Rdata")))
volcano.plot <- ggplot2::ggplot() +
  ggplot2::geom_point(data = myDiff, ggplot2::aes(x = meth.diff, 
                                                  y = -log10(qvalue))) +
  ggplot2::geom_point(data = myDiff25p, ggplot2::aes(x = meth.diff, 
                                                  y = -log10(qvalue)),
                      colour = "red") +
  ggplot2::theme_minimal() +
  ggplot2::theme(
    axis.text = ggplot2::element_text(size = 15),
    axis.title.x = ggplot2::element_text(size = 20),
    axis.title.y = ggplot2::element_text(size = 20),
    legend.text = ggplot2::element_text(size = 15)
  )
print(volcano.plot)

ggplot2::ggsave(volcano.plot, bg = "white",
                filename =
                  paste0(figure.path,
                         "Toothfish_2018_muscle_DMS25_VOLCANO_methylkit_regions.png"), 
                width = 30, height = 15, units = "cm")


myDiff25p.gr <- as(myDiff25p, "GRanges")
grand.plot <- ggbio::plotGrandLinear(myDiff25p.gr,
                                     ggplot2::aes(y = meth.diff)) +
  ggplot2::ylab("Methylation difference (%)") +
  ggplot2::theme_bw() +
  ggplot2::theme(axis.text.x = ggplot2::element_text(angle = 90, hjust = 0), 
                 legend.position = "none")
print(grand.plot)
ggplot2::ggsave(grand.plot@ggplot, bg = "white",
                filename = paste0(figure.path,
                                  "Toothfish_2018_muscle_DMS25_Grandplot_methylkit_regions.png"
                ), 
                width = 30, height = 15, units = "cm")
```

### Old DMS
```{r}
print(load(paste0(rdata.path,"Toothfish_DMS_CpGsites.Rdata")))
meth.cpg <- paste0(myDiff25p$chr, "_", myDiff25p$start) #1202 
DMS.cpg <- unique(c(DMS$DMS)) #1243
sum(meth.cpg %in% DMS.cpg) #514
```

### Heatmap

```{r}
print(load(paste0(rdata.path,"Toothfish_2018_muscle_methylation_DMS.Rdata")))

myDiff25p.gr <- as(myDiff25p, "GRanges")
meth.diff <- methylKit::selectByOverlap(meth.2018M.norm11L, myDiff25p.gr)

meth.diff.df <- as.data.frame(t(methylKit::percMethylation(meth.diff)))
loc.names <- paste0(meth.diff$chr,"_",meth.diff$start)
colnames(meth.diff.df) <- loc.names

meth.diff.df$individual <- rownames(meth.diff.df)
pop <- metadata.new$pop
meth.diff.df$pop <- as.character(pop)


heat.plot.data <- meth.diff.df %>%
  tibble::as_tibble() %>%
  tidyr::gather(key = "CpG", value = "Methylation_percentage", -individual, -pop,
                factor_key = FALSE) %>%
  dplyr::group_by(pop, CpG) %>%
  dplyr::mutate(CpG_mean = mean(Methylation_percentage, na.rm = TRUE)) %>%
  dplyr::arrange(pop, individual, CpG) %>%
  dplyr::mutate(individual = factor(individual, levels = unique(individual))) %>%
  dplyr::mutate(CpG = factor(CpG, levels = unique(CpG)))
cpgs <- heat.plot.data$CpG[heat.plot.data$individual == unique(heat.plot.data$individual)[1]]
cpgs_mean <- heat.plot.data$CpG_mean[heat.plot.data$individual == unique(heat.plot.data$individual)[1]]

cpgs <- cpgs[order(cpgs_mean, decreasing = TRUE)]
heat.plot.data$CpG <- factor(heat.plot.data$CpG, levels = c("pop", unique(as.character(cpgs))))
heat.plot.data <- heat.plot.data[order(heat.plot.data$CpG),]

heat.plot <- ggplot2::ggplot(data = heat.plot.data, ggplot2::aes(x = individual)) + 
    ggplot2::geom_tile(data = heat.plot.data, 
                       ggplot2::aes(x = individual, y = CpG, fill = Methylation_percentage)) +
    ggplot2::scale_fill_gradient2(low = "blue", mid = "yellow", high = "red", 
                                  midpoint = 50, guide = "colourbar") +
    ggnewscale::new_scale_fill() +
    ggplot2::geom_tile(data = heat.plot.data, ggplot2::aes(x = individual, y = "pop", fill = pop,
                                    height = length(unique(heat.plot.data$CpG))*0.02)) +
    ggplot2::scale_fill_manual(values = colours4) +
    ggplot2::scale_y_discrete(drop = FALSE) +
    ggplot2::theme(legend.position = "top",
                   axis.title.y = ggplot2::element_blank(),
                   axis.title.x = ggplot2::element_blank(),
                   axis.text.x = ggplot2::element_text(face = "bold", color = "black", 
                                        size = 10, angle = 90, hjust = 1, vjust = 0.5))
print(heat.plot)

ggplot2::ggsave(heat.plot, bg = "white",
                filename = paste0(figure.path,
                                  "Toothfish_2018_muscle_methDiff_25_HEATMAP.png"), 
                width = 30, height = 30, units = "cm")
```

### Plot CpG difference

```{r}
load(file = paste0(rdata.path,"Toothfish_2018_muscle_methylation_DMS.Rdata"))

myDiff25p.gr <- as(myDiff25p, "GRanges")
meth.diff <- methylKit::selectByOverlap(meth.2018M.norm11L, myDiff25p.gr) #113 

length(unique(meth.diff$chr)) #41
sort(table(meth.diff$chr), decreasing = TRUE)[1:20]

meth.diff.df <- as.data.frame(t(methylKit::percMethylation(meth.diff)))
loc.names <- paste0(meth.diff$chr,"_",meth.diff$start)
colnames(meth.diff.df) <- loc.names

meth.diff.df$individual <- rownames(meth.diff.df) 
identical(metadata.new$Genotype_names, meth.diff.df$individual)
meth.diff.df$pop <- metadata.new$pop



meth.diff.df.plot <- tidyr::gather(meth.diff.df, key = locus, 
                           value = perc_methylation, -individual, -pop,
                           factor_key = TRUE) %>%
  dplyr::mutate(chr = stringr::str_split_fixed(locus, 
                                               pattern = "_", 
                                               n = 2)[,1])
meth.diff.df.plot.locus <- meth.diff.df.plot %>% dplyr::group_by(chr,locus,pop) %>% 
  dplyr::summarise(mean_pop = mean(perc_methylation, na.rm = TRUE),
                   sum_NA = sum(is.na(perc_methylation)),
            .groups = 'drop')


diff.plot <- ggplot2::ggplot() +
  ggplot2::geom_line(data = meth.diff.df.plot.locus,
                     ggplot2::aes(x = pop, y = mean_pop, colour = locus, group = locus)) +
  ggplot2::geom_point(data = meth.diff.df.plot.locus,
                     ggplot2::aes(x = pop, y = mean_pop, colour = locus, group = locus),
                     size = 5) +
  ggplot2::facet_wrap(~chr) +
  ggplot2::ylab("Mean methylation (%)") +
  ggplot2::theme(legend.position = "none",
                   axis.text.x = ggplot2::element_text(face = "bold", color = "black", 
                                        size = 10, angle = 90, hjust = 1, vjust = 0.5))

print(diff.plot)

ggplot2::ggsave(plot = diff.plot, 
                filename = paste0(figure.path,"1_Toothfish_2018_muscle_methDiff.png"),
                width = 30, height = 15, units = "cm", device = "png")
```


### DMS Clustering
```{r}
load(file = paste0(rdata.path,"Toothfish_2018_muscle_methylation_DMS.Rdata"))

myDiff25p.gr <- as(myDiff25p, "GRanges")
meth.diff <- methylKit::selectByOverlap(meth.2018M.norm11L, myDiff25p.gr) #112  
methdif.perc <- methylKit::percMethylation(meth.diff)

methdif.perc2 <- zoo::na.aggregate(methdif.perc, by = meth.2018M.norm11L@treatment)
PCobj <- prcomp(methdif.perc2, retx = TRUE, scale. = TRUE)

var_frac <- PCobj$sdev/sum(PCobj$sdev)
print(paste0("Total variance explained: ",signif(sum(var_frac[1:3]) * 100, 3), "%")) 
barplot(var_frac[1:3], main = "variance explained", 
        col = heat.colors(length(var_frac[1:3])))


identical(metadata.new$Genotype_names, meth.2018M.norm11L@sample.ids)
# identical(metadata.new$individual, rownames(PCobj$x))
identical(metadata.new$Genotype_names, rownames(PCobj$rotation))
pop <- metadata.new$pop

plot.title <- "Toothfish methylation - 2018 Muscle (1202 DMS sites - 25%)"

# data <- data.frame(ind = rownames(PCobj$x),PCobj$x,  pop = pop)
data <- data.frame(ind = rownames(PCobj$rotation),PCobj$rotation,  pop = pop)
data <- merge(data, aggregate(cbind(mean.x = PC1, mean.y = PC2) ~
                                pop, data, mean), by = "pop")

pca.plot <- ggplot2::ggplot(data, ggplot2::aes(PC1, PC2, color = pop)) +
  ggplot2::geom_point(size = 4) + 
  ggplot2::geom_point(ggplot2::aes(x = mean.x,
                                   y = mean.y, color = pop),
                      size = 7,
                      shape = 17) +
  ggplot2::scale_colour_manual(values = colours4) +
  ggplot2::geom_hline(yintercept = 0) +
  ggplot2::geom_vline(xintercept = 0) + 
  ggplot2::labs(
    subtitle = paste0("Total variance explained: ",
                      signif(sum(var_frac[1:3]) * 100, 3), "%"),
    y = paste0("PC2; ", signif(var_frac[2] * 100, digits = 2), " %"),
    x = paste0("PC1; ", signif(var_frac[1] * 100, digits = 2), " %"),
    title = plot.title,
    caption = "") + 
  ggplot2::theme_minimal() +
  ggplot2::theme(
    axis.text = ggplot2::element_text(size = 15),
    axis.title.x = ggplot2::element_text(size = 20),
    axis.title.y = ggplot2::element_text(size = 20),
    legend.text = ggplot2::element_text(size = 15)
  )

print(pca.plot)
ggplot2::ggsave(pca.plot, bg = "white",
                filename = paste0(figure.path,"Toothfish_2018_muscle_PCA_CpG_DMS_PC12.png"), 
                width = 30, height = 15, units = "cm")



# data <- data.frame(ind = rownames(PCobj$x),PCobj$x,  pop = pop)
data <- data.frame(ind = rownames(PCobj$rotation),PCobj$rotation,  pop = pop)
data <- merge(data, aggregate(cbind(mean.x = PC1, mean.y = PC3) ~
                                pop, data, mean), by = "pop")

pca.plot <- ggplot2::ggplot(data, ggplot2::aes(PC1, PC3, color = pop)) +
  ggplot2::geom_point(size = 4) + 
  ggplot2::geom_point(ggplot2::aes(x = mean.x,
                                   y = mean.y, color = pop),
                      size = 7,
                      shape = 17) +
  ggplot2::scale_colour_manual(values = colours4) +
  ggplot2::geom_hline(yintercept = 0) +
  ggplot2::geom_vline(xintercept = 0) + 
  ggplot2::labs(
    subtitle = paste0("Total variance explained: ",
                      signif(sum(var_frac[1:3]) * 100, 3), "%"),
    y = paste0("PC3; ", signif(var_frac[3] * 100, digits = 2), " %"),
    x = paste0("PC1; ", signif(var_frac[1] * 100, digits = 2), " %"),
    title = plot.title,
    caption = "") + 
  ggplot2::theme_minimal() +
  ggplot2::theme(
    axis.text = ggplot2::element_text(size = 15),
    axis.title.x = ggplot2::element_text(size = 20),
    axis.title.y = ggplot2::element_text(size = 20),
    legend.text = ggplot2::element_text(size = 15)
  )

print(pca.plot)
ggplot2::ggsave(pca.plot, bg = "white",
                filename = paste0(figure.path,
                                  "Toothfish_2018_muscle_PCA_CpG_DMS_PC13.png"), 
                width = 30, height = 15, units = "cm")




# data <- data.frame(ind = rownames(PCobj$x),PCobj$x,  pop = pop)
data <- data.frame(ind = rownames(PCobj$rotation),PCobj$rotation,  pop = pop)
data <- merge(data, aggregate(cbind(mean.x = PC2, mean.y = PC3) ~
                                pop, data, mean), by = "pop")

pca.plot <- ggplot2::ggplot(data, ggplot2::aes(PC2, PC3, color = pop)) +
  ggplot2::geom_point(size = 4) + 
  ggplot2::geom_point(ggplot2::aes(x = mean.x,
                                   y = mean.y, color = pop),
                      size = 7,
                      shape = 17) +
  ggplot2::scale_colour_manual(values = colours4) +
  ggplot2::geom_hline(yintercept = 0) +
  ggplot2::geom_vline(xintercept = 0) + 
  ggplot2::labs(
    subtitle = paste0("Total variance explained: ",
                      signif(sum(var_frac[1:3]) * 100, 3), "%"),
    y = paste0("PC3; ", signif(var_frac[3] * 100, digits = 2), " %"),
    x = paste0("PC2; ", signif(var_frac[2] * 100, digits = 2), " %"),
    title = plot.title,
    caption = "") + 
  ggplot2::theme_minimal() +
  ggplot2::theme(
    axis.text = ggplot2::element_text(size = 15),
    axis.title.x = ggplot2::element_text(size = 20),
    axis.title.y = ggplot2::element_text(size = 20),
    legend.text = ggplot2::element_text(size = 15)
  )

print(pca.plot)
ggplot2::ggsave(pca.plot, bg = "white",
                filename = paste0(figure.path,"Toothfish_2018_muscle_PCA_CpG_DMS_PC23.png"), 
                width = 30, height = 15, units = "cm")
```

### Karyogram

```{r}
chr.file <- "/datasets/work/ncmi-toa-rrbs/work/1.Toothfish/Bismark_genome_index/sizes.chr"
chrom.info <- read.table(file = chr.file, header = F)
chrom.info$V3 <- FALSE
colnames(chrom.info) <- c("chrom", "length", "is_circular")
# chrom.info <- chrom.info[order(chrom.info$chrom),]
# chrom.info2 <- chrom.info[chrom.info$chrom %in% c("JAAKFY010000002", "JAAKFY010000014"),]

chrom.info2 <- chrom.info[chrom.info$chrom %in% paste0("JAAKFY0100000", stringr::str_pad(1:27, 2, pad = "0")),]

chrom.info3 <- chrom.info2
seqnames <- gsub(pattern = "JAAKFY01", replacement = "", as.character(chrom.info3$chrom))
seqnames <- paste0("chr", stringr::str_pad(as.integer(seqnames), 2, pad = "0"))
chrom.info3$chrom <- seqnames

chrom.seqinfo <- GenomeInfoDb::Seqinfo(seqnames = chrom.info3$chrom, 
                                       seqlengths = chrom.info3$length, 
                                       isCircular = chrom.info3$is_circular, 
                                       genome = "Antarctic Toothfish")

gr <- GRanges(seqnames = chrom.info2$chrom,ranges = IRanges(start = c(0,0), end = chrom.info2$length))

meth.sub <- methylKit::selectByOverlap(meth.2018M.norm11L, gr) #1202  
meth.sub.gr <- as(meth.sub, "GRanges")
seqnames <- gsub(pattern = "JAAKFY01", replacement = "", as.character(meth.sub.gr@seqnames))
seqnames <- paste0("chr", stringr::str_pad(as.integer(seqnames), 2, pad = "0"))
seqlevels(meth.sub.gr) <- as.character(unique(seqnames))

myDiff25p.sub <- methylKit::selectByOverlap(myDiff25p, gr) #1202  
myDiff25p.sub.gr <- as(myDiff25p.sub, "GRanges")
seqnames <- gsub(pattern = "JAAKFY01", replacement = "", as.character(myDiff25p.sub.gr@seqnames))
seqnames <- paste0("chr", stringr::str_pad(as.integer(seqnames), 2, pad = "0"))
seqlevels(myDiff25p.sub.gr) <- as.character(unique(seqnames))


karyo <- GRanges(seqnames = meth.sub.gr@seqnames,
                 ranges = IRanges::ranges(meth.sub.gr))
karyo.diff <- GRanges(seqnames = myDiff25p.sub.gr@seqnames,
                 ranges = IRanges::ranges(myDiff25p.sub.gr))

p <- ggbio::autoplot(chrom.seqinfo, layout = "karyogram", cytobands = FALSE) + 
  ggbio::layout_karyogram(karyo, geom = "rect", color = "blue") +
  ggbio::layout_karyogram(karyo.diff, geom = "rect", color = "red")
# print(p)

ggplot2::ggsave(p@ggplot, bg = "white",
                filename = paste0(figure.path,
                                  "Toothfish_2018_muscle_ALL_karyogram2.png"
                ), 
                width = 45, height = 30, units = "cm")

```

### Gene region track

```{r, eval=FALSE}
options(ucscChromosomeNames = FALSE)
print(load(paste0(rdata.path,"Toothfish_genome_annotation_features.Rdata")))
gencode <- loadDb(paste0(rdata.path, "gencode_D_mawsoni.sqlite"))

chr.file <- "/datasets/work/ncmi-toa-rrbs/work/1.Toothfish/Bismark_genome_index/sizes.chr"
chrom.info <- read.table(file = chr.file, header = F)
chrom.info$V3 <- FALSE
colnames(chrom.info) <- c("chrom", "length", "is_circular")

chrom.info3 <- chrom.info
seqnames <- gsub(pattern = "JAAKFY01", replacement = "", as.character(chrom.info3$chrom))
seqnames <- paste0("chr", stringr::str_pad(as.integer(seqnames), 2, pad = "0"))
chrom.info3$chrom <- seqnames

chrom.seqinfo <- GenomeInfoDb::Seqinfo(seqnames = chrom.info3$chrom, 
                                       seqlengths = chrom.info3$length, 
                                       isCircular = chrom.info3$is_circular, 
                                       genome = "Antarctic Toothfish")


gtrack <- Gviz::GenomeAxisTrack()


seqnames1 <- gsub(pattern = "JAAKFY01", replacement = "", as.character(gencode@.xData$user_seqlevels))
seqnames2 <- paste0("chr", stringr::str_pad(as.integer(seqnames1), 2, pad = "0"))
# seqlevels(annotation.genes, pruning.mode = "coarse") <- as.character(unique(paste0("JAAKFY01",seqnames1)))
gencode@.xData$user_seqlevels <- as.character(unique(seqnames2))
geneTrack <- Gviz::GeneRegionTrack(gencode,
                                    name = "Gene", shape = "arrow",
                                   transcriptAnnotation = "transcript"
                                   )

geneTrack@dp@pars$fontcolor <- "black"
geneTrack@dp@pars$fontcolor.title <- "black"
geneTrack@dp@pars$background.title = "white"
geneTrack@dp@pars$col.border.title <- "black"


seqnames1 <- gsub(pattern = "JAAKFY01", replacement = "", as.character(cpgi.obj@seqnames))
seqnames2 <- paste0("chr", stringr::str_pad(as.integer(seqnames1), 2, pad = "0"))
seqlevels(cpgi.obj, pruning.mode = "coarse") <- as.character(unique(paste0("JAAKFY01",seqnames1)))
seqlevels(cpgi.obj) <- as.character(unique(seqnames2))
cpgi.track <- Gviz::AnnotationTrack(cpgi.obj,
                                    name = "CpGi")


cpgi.track@dp@pars$fontcolor.title <- "black"
cpgi.track@dp@pars$background.title = "white"
cpgi.track@dp@pars$col.border.title <- "black"

print(load(paste0(rdata.path,"Toothfish_2018_muscle_methylkit_object.Rdata")))
print(load(paste0(rdata.path,"Toothfish_metadata.Rdata"))) 
print(load(paste0(rdata.path,"Toothfish_2018_muscle_methylation_DMS.Rdata")))


meth.gr <- as(meth.2018M.norm11L, "GRanges")
seqnames <- gsub(pattern = "JAAKFY01", replacement = "", as.character(meth.gr@seqnames))
seqnames <- paste0("chr", stringr::str_pad(as.integer(seqnames), 2, pad = "0"))
seqlevels(meth.gr) <- as.character(unique(seqnames))
meth.perc <- methylKit::percMethylation(meth.2018M.norm11L)
meth.perc2 <- as.data.frame(t(meth.perc))

seqnames.start <- paste0(meth.gr@seqnames, "_", meth.gr@ranges)
colnames(meth.perc2) <- seqnames.start
treatment <- meth.2018M.norm11L@treatment
treatment <- factor(treatment)
levels(treatment) <- c("58.4.1-2018","58.4.2-2018", "58.5.2-2017", "88.1-2018")
meth.perc2$treatment <- treatment
meth.perc2$individual <- row.names(meth.perc2)
data <- meth.perc2 %>% 
  tidyr::pivot_longer(!c(treatment, individual), names_to = "CpG", values_to = "methylation") %>%
  dplyr::group_by(treatment, CpG) %>%
  dplyr::summarise(MEAN = mean(methylation, na.rm = TRUE))
data <- data[order(data$treatment, match(data$CpG,seqnames.start)),]
data2 <- data %>% tidyr::pivot_wider(names_from = treatment, values_from = MEAN)
meth2.gr <- GRanges(seqnames = meth.gr@seqnames,ranges = meth.gr@ranges, seqinfo = chrom.seqinfo)
mcols(meth2.gr) <- data2[,2:5]
# meth.track <- Gviz::DataTrack(meth2.gr,type = "b",color = "blue",
#                               groups = c("58.4.1", "58.4.2", "58.5.2", "88.1"),
#                              name = "Mean Methylation")
meth.track2 <- Gviz::DataTrack(meth2.gr,type = "heatmap",color = "blue",
                              name = "Mean meth")

meth3.gr <- GRanges(seqnames = meth.gr@seqnames,ranges = meth.gr@ranges, seqinfo = chrom.seqinfo)
meth.perc.na <- zoo::na.aggregate(meth.perc,by = treatment, FUN = mean, na.rm = TRUE)
# mcols(meth3.gr) <- meth.perc
mcols(meth3.gr) <- meth.perc.na
meth.track <- Gviz::DataTrack(meth3.gr,type = c("a", "confint"),
                              groups = treatment,
                              na.rm = TRUE,
                             name = "Methylation %")
meth.track@dp@pars$ylim <- c(0, 100)
meth.track@dp@pars$col.confint <- colours4
meth.track@dp@pars$col <- colours4
meth.track@dp@pars$legend <- FALSE
meth.track@dp@pars$cex.legend <- 0

meth.track@dp@pars$fontcolor.title <- "black"
meth.track@dp@pars$background.title = "white"
meth.track@dp@pars$col.border.title <- "black"
meth.track@dp@pars$col.axis <- "black"

meth.track2@dp@pars$fontcolor.title <- "black"
meth.track2@dp@pars$background.title = "white"
meth.track2@dp@pars$col.border.title <- "black"
meth.track2@dp@pars$col.axis <- "black"


myDiff25p.gr <- as(myDiff, "GRanges")
# myDiff25p.gr <- as(myDiff25p, "GRanges")
seqnames <- gsub(pattern = "JAAKFY01", replacement = "", as.character(myDiff25p.gr@seqnames))
seqnames <- paste0("chr", stringr::str_pad(as.integer(seqnames), 2, pad = "0"))
seqlevels(myDiff25p.gr) <- as.character(unique(seqnames))
diff.track <- Gviz::DataTrack(myDiff25p.gr,type = "b", data = myDiff25p.gr$meth.diff,
                              name = "Meth. diff %")
diff.track@dp@pars$fontcolor.title <- "black"
diff.track@dp@pars$background.title = "white"
diff.track@dp@pars$col.border.title <- "black"
diff.track@dp@pars$col.axis <- "black"
diff.track@dp@pars$ylim <- c(0, max(myDiff25p.gr$meth.diff))

meth.cov <- methylKit::getData(meth.2018M.norm11L)[,meth.2018M.norm11L@coverage.index]
colnames(meth.cov) <- meth.2018M.norm11L@sample.ids
meth.cov2 <- as.data.frame(t(meth.cov))

seqnames.start <- paste0(meth.gr@seqnames, "_", meth.gr@ranges)
colnames(meth.cov2) <- seqnames.start
treatment <- meth.2018M.norm11L@treatment
treatment <- factor(treatment)
levels(treatment) <- c("58.4.1-2018","58.4.2-2018", "58.5.2-2017", "88.1-2018")
meth.cov2$treatment <- treatment
meth.cov2$individual <- row.names(meth.cov2)
data.cov <- meth.cov2 %>% 
  tidyr::pivot_longer(!c(treatment, individual), names_to = "CpG", values_to = "coverage") %>%
  dplyr::group_by(treatment, CpG) %>%
  dplyr::summarise(MEAN = mean(coverage, na.rm = TRUE))
data.cov <- data.cov[order(data$treatment, match(data$CpG,seqnames.start)),]
data.cov2 <- data.cov %>% tidyr::pivot_wider(names_from = treatment, values_from = MEAN)
meth4.gr <- GRanges(seqnames = meth.gr@seqnames,ranges = meth.gr@ranges, seqinfo = chrom.seqinfo)
mcols(meth4.gr) <- data.cov2[,2:5]
# cov.track <- Gviz::DataTrack(meth4.gr,type =  c("a", "p"),color = "blue",
#                               groups = c("58.4.1", "58.4.2", "58.5.2", "88.1"),
#                              name = "Mean coverage")

meth4.gr <- GRanges(seqnames = meth.gr@seqnames,ranges = meth.gr@ranges, seqinfo = chrom.seqinfo)
meth.cov.na <- zoo::na.aggregate(meth.cov,by = treatment, FUN = mean, na.rm = TRUE)
mcols(meth4.gr) <- meth.cov.na
cov.track <- Gviz::DataTrack(meth4.gr,type =  c("a", "confint"),
                              groups = treatment,
                             na.rm = TRUE,
                             name = "Coverage")

cov.track@dp@pars$fontcolor.title <- "black"
cov.track@dp@pars$background.title = "white"
cov.track@dp@pars$col.border.title <- "black"
cov.track@dp@pars$col.axis <- "black"
cov.track@dp@pars$cex.legend <- 0

# cov.track@dp@pars$size <- 5
# cov.track@dp@pars$cex <- 5
# cov.track@dp@pars$cex.axis <- 5
cov.track@dp@pars$ylim <- c(0, max(data.cov$MEAN))
cov.track@dp@pars$col.confint <- colours4
cov.track@dp@pars$col <- colours4


print(load(paste0(rdata.path, "SNP_genetic_differentiation_2018.Rdata")))
print(load(paste0(rdata.path, "SNP_2018_RDA.Rdata")))
SNP.names <- rownames(fst.basic$perloc)
SNP.names <- stringr::str_split(string = SNP.names, pattern = "_", simplify = TRUE)

SNP.gr <- GRanges(seqnames = SNP.names[,1],
                   ranges = IRanges(start = as.integer(SNP.names[,2]),
                                    end = as.integer(SNP.names[,2])+1))
mcols(SNP.gr) <- fst.basic$perloc
seqnames <- gsub(pattern = "JAAKFY01", replacement = "", as.character(SNP.gr@seqnames))
seqnames <- paste0("chr", stringr::str_pad(as.integer(seqnames), 2, pad = "0"))
seqlevels(SNP.gr) <- as.character(unique(seqnames))

SNP2.gr <- GRanges(seqnames = SNP.names[,1],
                   ranges = IRanges(start = as.integer(SNP.names[,2]),
                                    end = as.integer(SNP.names[,2])+1))
mcols(SNP2.gr) <- fst.basic$Ho
seqnames <- gsub(pattern = "JAAKFY01", replacement = "", as.character(SNP2.gr@seqnames))
seqnames <- paste0("chr", stringr::str_pad(as.integer(seqnames), 2, pad = "0"))
seqlevels(SNP2.gr) <- as.character(unique(seqnames))

FST.track <- Gviz::DataTrack(SNP.gr,type ="b",
                              data = SNP.gr$Fst,
                             name = "Fst")

FST.track@dp@pars$fontcolor.title <- "black"
FST.track@dp@pars$background.title = "white"
FST.track@dp@pars$col.border.title <- "black"
FST.track@dp@pars$col.axis <- "black"
# FST.track@dp@pars$ylim <- c(0, max(SNP.gr$Fst))


SNP2.gr <- GRanges(seqnames = SNP.names[,1],
                   ranges = IRanges(start = as.integer(SNP.names[,2]),
                                    end = as.integer(SNP.names[,2])+1))
mcols(SNP2.gr) <- fst.basic$Ho
seqnames <- gsub(pattern = "JAAKFY01", replacement = "", as.character(SNP2.gr@seqnames))
seqnames <- paste0("chr", stringr::str_pad(as.integer(seqnames), 2, pad = "0"))
seqlevels(SNP2.gr) <- as.character(unique(seqnames))
Ho.track <- Gviz::DataTrack(SNP2.gr,type =  c("a", "p"),color = "blue",
                            # groups = c("48.2-2017", "58.4.1-2018", "58.4.2-2018", "58.4.2-2024",
                            #              "58.5.2-2017", "88.1-2018", "88.2-2024"),
                            groups = c("48.2-2017", "58.4.1-2018", "58.4.2-2018",
                                         "58.5.2-2017", "88.1-2018"),
                            legend = FALSE,
                             name = "Obs Het")
Ho.track@dp@pars$col.confint <- colours.all
Ho.track@dp@pars$col <- colours.all
Ho.track@dp@pars$fontcolor.title <- "black"
Ho.track@dp@pars$background.title = "white"
Ho.track@dp@pars$col.border.title <- "black"
Ho.track@dp@pars$col.axis <- "black"
# Ho.track@dp@pars$ylim <- c(0, 1)
Ho.track@dp@pars$legend <- FALSE
Ho.track@dp@pars$cex.legend <- 0

track.list <- c(geneTrack, cpgi.track, cov.track,diff.track,meth.track, meth.track2, 
                FST.track, Ho.track, gtrack)

save(track.list, file = paste0(rdata.path, "Toothfish_2018_Muscle_tracklist.Rdata"))
```


```{r}
print(load(paste0(rdata.path, "Toothfish_2018_Muscle_tracklist.Rdata")))

#### CHR02
Gviz::plotTracks(track.list, legend = TRUE, chromosome = "chr02", from = 19400000, to = 21540000,  
                 showId = FALSE)
dev.print(file = paste0(figure.path,"Toothfish_2018_muscle_TRACKPLOT_chr02_19400000_21540000.png"),
          device = png, res = 300,width = 30,height = 15,units = "cm")

Gviz::plotTracks(track.list, legend = TRUE, chromosome = "chr02", from = 19400000, to = 19600000,  
                 showId = FALSE)
dev.print(file = paste0(figure.path,"Toothfish_2018_muscle_TRACKPLOT_chr02_19400000_19600000.png"),
          device = png, res = 300,width = 30,height = 15,units = "cm")
Gviz::plotTracks(track.list, legend = TRUE, chromosome = "chr02", from = 19513500, to = 19516000)
dev.print(file = paste0(figure.path,"Toothfish_2018_muscle_TRACKPLOT_chr02_19513500_19516000.png"),
          device = png, res = 300,width = 15,height = 15,units = "cm")
Gviz::plotTracks(track.list, legend = TRUE, chromosome = "chr02", from = 21510000, to = 21540000,  
                 showId = FALSE)
dev.print(file = paste0(figure.path,"Toothfish_2018_muscle_TRACKPLOT_chr02_21500000_21540000.png"),
          device = png, res = 300,width = 30,height = 15,units = "cm")
Gviz::plotTracks(track.list, legend = TRUE, chromosome = "chr02", from = 21526500, to = 21527600)
dev.print(file = paste0(figure.path,"Toothfish_2018_muscle_TRACKPLOT_chr02_21526500_21527600.png"),
          device = png, res = 300,width = 15,height = 15,units = "cm")

#### CHR14
Gviz::plotTracks(track.list, legend = TRUE, chromosome = "chr14", from = 4650000, to = 4900000,
                 showId = FALSE)
dev.print(file = paste0(figure.path,"Toothfish_2018_muscle_TRACKPLOT_chr14_4650000_490000.png"),
          device = png, res = 300,width = 30,height = 15,units = "cm")
Gviz::plotTracks(track.list, legend = TRUE, chromosome = "chr14", from = 4722000, to = 4770000)
dev.print(file = paste0(figure.path,"Toothfish_2018_muscle_TRACKPLOT_chr14_4722000_4770000.png"),
          device = png, res = 300,width = 15,height = 15,units = "cm")
Gviz::plotTracks(track.list, legend = TRUE, chromosome = "chr14", from = 4722000, to = 4724000)
dev.print(file = paste0(figure.path,"Toothfish_2018_muscle_TRACKPLOT_chr14_4722000_4724000.png"),
          device = png, res = 300,width = 15,height = 15,units = "cm")
```

### Gene function

```{r}
chrom.info2 <- chrom.info[chrom.info$chrom %in% c("JAAKFY010000002", "JAAKFY010000014"),]
gr <- GRanges(seqnames = chrom.info2$chrom,ranges = IRanges(start = c(0,0), end = chrom.info2$length))

chr2.gr <- GRanges(seqnames = "JAAKFY010000002",
                   ranges = IRanges(start = 19450000,
                                    end = 19550000))
olap <- GenomicRanges::findOverlaps(gff3, chr2.gr)
gff3.sub <- gff3[unique(olap@from),]

chr2.gr <- GRanges(seqnames = "JAAKFY010000002",
                   ranges = IRanges(start = 21523000,
                                    end = 21525000))
olap <- GenomicRanges::findOverlaps(gff3, chr2.gr)
gff3.sub <- gff3[unique(olap@from),]

chr14.gr <- GRanges(seqnames = "JAAKFY010000014",
                   ranges = IRanges(start = 4722000,
                                    end = 4724000))
olap <- GenomicRanges::findOverlaps(gff3, chr14.gr)
gff3.sub <- gff3[unique(olap@from),]

View(as.data.frame(gff3.sub))


```

## DMS RANDOM - 2018 - muscle

```{r, eval=FALSE}
print(load(paste0(rdata.path,"Toothfish_2018_muscle_methylkit_object.Rdata")))

print(load(paste0(rdata.path,"Toothfish_metadata.Rdata")))
metadata.new <- metadata[metadata$Genotype_names %in% attributes(meth.2018M.norm11L)$sample.ids,]
attributes(meth.2018M.norm11L)$sample.ids == metadata.new$Genotype_names

strata.random <- sample(metadata.new$pop, size = length(metadata.new$pop), replace = FALSE)
meth.2018M.norm11L@treatment <- as.integer(factor(strata.random))

# data.frame(attributes(meth.2018M.norm11L)$sample.ids, metadata.new$Genotype_names, metadata.new$strata, meth.2018M.norm11L@treatment)

# covariates <- dplyr::select(metadata.new, Sex, TL_mm)
myDiff <- methylKit::calculateDiffMeth(meth.2018M.norm11L,
                                       covariates = NULL,
                                       adjust = "fdr",
                                       chunk.size = 1e+06,
                                       mc.cores = 25)
meth.diff.chr <- methylKit::diffMethPerChr(myDiff,plot = TRUE, 
                                           qvalue.cutoff = 0.01, meth.cutoff = 40)
myDiff40p <- methylKit::getMethylDiff(myDiff, difference = 40, qvalue = 0.01)
#13
myDiff25p <- methylKit::getMethylDiff(myDiff, difference = 25, qvalue = 0.01)
#537 
myDiff15p <- methylKit::getMethylDiff(myDiff, difference = 15, qvalue = 0.01)
#7310   

save(strata.random, myDiff,myDiff15p, myDiff25p, myDiff40p,
     file = paste0(rdata.path,"Toothfish_2018_muscle_methylation_RANDOM_DMS.Rdata"))
```

```{r}
print(load(paste0(rdata.path,"Toothfish_2018_muscle_methylation_RANDOM_DMS.Rdata")))
volcano.plot <- ggplot2::ggplot() +
  ggplot2::geom_point(data = myDiff, ggplot2::aes(x = meth.diff, 
                                                  y = -log10(qvalue))) +
  ggplot2::geom_point(data = myDiff25p, ggplot2::aes(x = meth.diff, 
                                                  y = -log10(qvalue)),
                      colour = "red") +
  
  ggplot2::theme_minimal() +
  ggplot2::theme(
    axis.text = ggplot2::element_text(size = 15),
    axis.title.x = ggplot2::element_text(size = 20),
    axis.title.y = ggplot2::element_text(size = 20),
    legend.text = ggplot2::element_text(size = 15)
  )
print(volcano.plot)

ggplot2::ggsave(volcano.plot, bg = "white",
                filename =
                  paste0(figure.path,
                         "Toothfish_2018_muscle_RANDOM_DMS25_VOLCANO_methylkit.png"), 
                width = 30, height = 15, units = "cm")



myDiff25p.gr <- as(myDiff25p, "GRanges")
grand.plot <- ggbio::plotGrandLinear(myDiff25p.gr,
                                     ggplot2::aes(y = meth.diff)) +
  ggplot2::ylab("Methylation difference (%)") +
  ggplot2::theme_bw() +
  ggplot2::theme(axis.text.x = ggplot2::element_text(angle = 90, hjust = 0), 
                 legend.position = "none")
print(grand.plot)
ggplot2::ggsave(grand.plot@ggplot, bg = "white",
                filename = paste0(figure.path,
                                  "Toothfish_2018_muscle_RANDOM_DMS25_Grandplot.png"
                ), 
                width = 30, height = 15, units = "cm")
```


### Heatmap

```{r}
print(load(paste0(rdata.path,"Toothfish_2018_muscle_methylation_RANDOM_DMS.Rdata")))

myDiff25p.gr <- as(myDiff25p, "GRanges")
meth.diff <- methylKit::selectByOverlap(meth.2018M.norm11L, myDiff25p.gr)

meth.diff.df <- as.data.frame(t(methylKit::percMethylation(meth.diff)))
loc.names <- paste0(meth.diff$chr,"_",meth.diff$start)
colnames(meth.diff.df) <- loc.names

meth.diff.df$individual <- rownames(meth.diff.df)
meth.diff.df$pop <- strata.random
meth.diff.df$pop2 <- as.character(metadata.new$pop)

heat.plot.data <- meth.diff.df %>%
  tibble::as_tibble() %>%
  tidyr::gather(key = "CpG", value = "Methylation_percentage", -individual, -pop, -pop2,
                factor_key = FALSE) %>%
  dplyr::group_by(pop, CpG) %>%
  dplyr::mutate(CpG_mean = mean(Methylation_percentage, na.rm = TRUE)) %>%
  dplyr::arrange(pop, individual, CpG) %>%
  dplyr::mutate(individual = factor(individual, levels = unique(individual))) %>%
  dplyr::mutate(CpG = factor(CpG, levels = unique(CpG)))
cpgs <- heat.plot.data$CpG[heat.plot.data$individual == unique(heat.plot.data$individual)[1]]
cpgs_mean <- heat.plot.data$CpG_mean[heat.plot.data$individual == unique(heat.plot.data$individual)[1]]

cpgs <- cpgs[order(cpgs_mean, decreasing = TRUE)]
heat.plot.data$CpG <- factor(heat.plot.data$CpG, levels = c("pop", unique(as.character(cpgs))))
heat.plot.data <- heat.plot.data[order(heat.plot.data$CpG),]

heat.plot <- ggplot2::ggplot(data = heat.plot.data, ggplot2::aes(x = individual)) + 
    ggplot2::geom_tile(data = heat.plot.data, 
                       ggplot2::aes(x = individual, y = CpG, fill = Methylation_percentage)) +
    ggplot2::scale_fill_gradient2(low = "blue", mid = "yellow", high = "red", 
                                  midpoint = 50, guide = "colourbar") +
    ggnewscale::new_scale_fill() +
    ggplot2::geom_tile(data = heat.plot.data, ggplot2::aes(x = individual, 
                                                           y = "pop-random", fill = pop,
                                    height = length(unique(heat.plot.data$CpG))*0.02)) +
    ggplot2::scale_fill_manual(values = colours4) +
    ggnewscale::new_scale_fill() +
    ggplot2::geom_tile(data = heat.plot.data, ggplot2::aes(x = individual, y = "pop", fill = pop2,
                                    height = length(unique(heat.plot.data$CpG))*0.02)) +
    ggplot2::scale_fill_manual(values = colours4) +
    ggplot2::scale_y_discrete(drop = FALSE) +
    ggplot2::theme(legend.position = "top",
                   axis.title.y = ggplot2::element_blank(),
                   axis.title.x = ggplot2::element_blank(),
                   axis.text.x = ggplot2::element_text(face = "bold", color = "black", 
                                        size = 10, angle = 90, hjust = 1, vjust = 0.5))
print(heat.plot)

ggplot2::ggsave(heat.plot, bg = "white",
                filename = paste0(figure.path,
                                  "Toothfish_2018_muscle_RANDOM_methDiff_25_HEATMAP.png"), 
                width = 30, height = 30, units = "cm")
```

### Plot CpG difference

```{r}
load(file = paste0(rdata.path,"Toothfish_2018_muscle_methylation_RANDOM_DMS.Rdata"))

myDiff25p.gr <- as(myDiff25p, "GRanges")
meth.diff <- methylKit::selectByOverlap(meth.2018M.norm11L, myDiff25p.gr) #113 

length(unique(meth.diff$chr)) #34
sort(table(meth.diff$chr), decreasing = TRUE)[1:20]

meth.diff.df <- as.data.frame(t(methylKit::percMethylation(meth.diff)))
loc.names <- paste0(meth.diff$chr,"_",meth.diff$start)
colnames(meth.diff.df) <- loc.names

meth.diff.df$individual <- rownames(meth.diff.df) 
identical(metadata.new$Genotype_names, meth.diff.df$individual)
meth.diff.df$pop <- metadata.new$pop


meth.diff.df.plot <- tidyr::gather(meth.diff.df, key = locus, 
                           value = perc_methylation, -individual, -pop,
                           factor_key = TRUE) %>%
  dplyr::mutate(chr = stringr::str_split_fixed(locus, 
                                               pattern = "_", 
                                               n = 2)[,1])
meth.diff.df.plot.locus <- meth.diff.df.plot %>% dplyr::group_by(chr,locus,pop) %>% 
  dplyr::summarise(mean_pop = mean(perc_methylation, na.rm = TRUE),
                   sum_NA = sum(is.na(perc_methylation)),
            .groups = 'drop')


diff.plot <- ggplot2::ggplot() +
  ggplot2::geom_line(data = meth.diff.df.plot.locus,
                     ggplot2::aes(x = pop, y = mean_pop, colour = locus, group = locus)) +
  ggplot2::geom_point(data = meth.diff.df.plot.locus,
                     ggplot2::aes(x = pop, y = mean_pop, colour = locus, group = locus),
                     size = 5) +
  ggplot2::facet_wrap(~chr) +
  ggplot2::ylab("Mean methylation (%)") +
  ggplot2::theme(legend.position = "none",
                   axis.text.x = ggplot2::element_text(face = "bold", color = "black", 
                                        size = 10, angle = 90, hjust = 1, vjust = 0.5))
    # ggplot2::guides(colour = ggplot2::guide_legend(ncol = 2))

print(diff.plot)

ggplot2::ggsave(plot = diff.plot, 
                filename = paste0(figure.path,"1_Toothfish_2018_muscle_RANDOM_methDiff.png"),
                width = 30, height = 15, units = "cm", device = "png")
```


### DMS Clustering
```{r}
load(file = paste0(rdata.path,"Toothfish_2018_muscle_methylation_RANDOM_DMS.Rdata"))

myDiff25p.gr <- as(myDiff25p, "GRanges")
meth.diff <- methylKit::selectByOverlap(meth.2018M.norm11L, myDiff25p.gr) #270  
methdif.perc <- methylKit::percMethylation(meth.diff)

methdif.perc2 <- zoo::na.aggregate(methdif.perc, by = meth.2018M.norm11L@treatment)
PCobj <- prcomp(methdif.perc2, retx = TRUE, scale. = TRUE)

var_frac <- PCobj$sdev/sum(PCobj$sdev)
print(paste0("Total variance explained: ",signif(sum(var_frac[1:3]) * 100, 3), "%")) 
barplot(var_frac[1:3], main = "variance explained", 
        col = heat.colors(length(var_frac[1:3])))


identical(metadata.new$Genotype_names, meth.2018M.norm11L@sample.ids)

identical(as.numeric(factor(metadata.new$pop)), meth.2018M.norm11L@treatment)

# identical(metadata.new$individual, rownames(PCobj$x))
identical(metadata.new$Genotype_names, rownames(PCobj$rotation))
pop <- metadata.new$pop

plot.title <- "Toothfish methylation - RANDOMISED 2018 Muscle (537 DMS sites - 25%)"

# data <- data.frame(ind = rownames(PCobj$x),PCobj$x,  pop = pop)
data <- data.frame(ind = rownames(PCobj$rotation),PCobj$rotation,  pop = pop)
data <- merge(data, aggregate(cbind(mean.x = PC1, mean.y = PC2) ~
                                pop, data, mean), by = "pop")

pca.plot <- ggplot2::ggplot(data, ggplot2::aes(PC1, PC2, color = pop)) +
  ggplot2::geom_point(size = 4) + 
  ggplot2::geom_point(ggplot2::aes(x = mean.x,
                                   y = mean.y, color = pop),
                      size = 7,
                      shape = 17) +
  ggplot2::scale_colour_manual(values = colours4) +
  ggplot2::geom_hline(yintercept = 0) +
  ggplot2::geom_vline(xintercept = 0) + 
  ggplot2::labs(
    subtitle = paste0("Total variance explained: ",
                      signif(sum(var_frac[1:3]) * 100, 3), "%"),
    y = paste0("PC2; ", signif(var_frac[2] * 100, digits = 2), " %"),
    x = paste0("PC1; ", signif(var_frac[1] * 100, digits = 2), " %"),
    title = plot.title,
    caption = "") + 
  ggplot2::theme_minimal() +
  ggplot2::theme(
    axis.text = ggplot2::element_text(size = 15),
    axis.title.x = ggplot2::element_text(size = 20),
    axis.title.y = ggplot2::element_text(size = 20),
    legend.text = ggplot2::element_text(size = 15)
  )

print(pca.plot)
ggplot2::ggsave(pca.plot, bg = "white",
                filename = 
                  paste0(figure.path,"Toothfish_2018_muscle_RANDOM_PCA_CpG_DMS_PC12.png"), 
                width = 30, height = 15, units = "cm")



# data <- data.frame(ind = rownames(PCobj$x),PCobj$x,  pop = pop)
data <- data.frame(ind = rownames(PCobj$rotation),PCobj$rotation,  pop = pop)
data <- merge(data, aggregate(cbind(mean.x = PC1, mean.y = PC3) ~
                                pop, data, mean), by = "pop")

pca.plot <- ggplot2::ggplot(data, ggplot2::aes(PC1, PC3, color = pop)) +
  ggplot2::geom_point(size = 4) + 
  ggplot2::geom_point(ggplot2::aes(x = mean.x,
                                   y = mean.y, color = pop),
                      size = 7,
                      shape = 17) +
  ggplot2::scale_colour_manual(values = colours4) +
  ggplot2::geom_hline(yintercept = 0) +
  ggplot2::geom_vline(xintercept = 0) + 
  ggplot2::labs(
    subtitle = paste0("Total variance explained: ",
                      signif(sum(var_frac[1:3]) * 100, 3), "%"),
    y = paste0("PC3; ", signif(var_frac[3] * 100, digits = 2), " %"),
    x = paste0("PC1; ", signif(var_frac[1] * 100, digits = 2), " %"),
    title = plot.title,
    caption = "") + 
  ggplot2::theme_minimal() +
  ggplot2::theme(
    axis.text = ggplot2::element_text(size = 15),
    axis.title.x = ggplot2::element_text(size = 20),
    axis.title.y = ggplot2::element_text(size = 20),
    legend.text = ggplot2::element_text(size = 15)
  )

print(pca.plot)
ggplot2::ggsave(pca.plot, bg = "white",
                filename = paste0(figure.path,"Toothfish_2018_muscle_RANDOM_PCA_CpG_DMS_PC13.png"), 
                width = 30, height = 15, units = "cm")




# data <- data.frame(ind = rownames(PCobj$x),PCobj$x,  pop = pop)
data <- data.frame(ind = rownames(PCobj$rotation),PCobj$rotation,  pop = pop)
data <- merge(data, aggregate(cbind(mean.x = PC2, mean.y = PC3) ~
                                pop, data, mean), by = "pop")

pca.plot <- ggplot2::ggplot(data, ggplot2::aes(PC2, PC3, color = pop)) +
  ggplot2::geom_point(size = 4) + 
  ggplot2::geom_point(ggplot2::aes(x = mean.x,
                                   y = mean.y, color = pop),
                      size = 7,
                      shape = 17) +
  ggplot2::scale_colour_manual(values = colours4) +
  ggplot2::geom_hline(yintercept = 0) +
  ggplot2::geom_vline(xintercept = 0) + 
  ggplot2::labs(
    subtitle = paste0("Total variance explained: ",
                      signif(sum(var_frac[1:3]) * 100, 3), "%"),
    y = paste0("PC3; ", signif(var_frac[3] * 100, digits = 2), " %"),
    x = paste0("PC2; ", signif(var_frac[2] * 100, digits = 2), " %"),
    title = plot.title,
    caption = "") + 
  ggplot2::theme_minimal() +
  ggplot2::theme(
    axis.text = ggplot2::element_text(size = 15),
    axis.title.x = ggplot2::element_text(size = 20),
    axis.title.y = ggplot2::element_text(size = 20),
    legend.text = ggplot2::element_text(size = 15)
  )

print(pca.plot)
ggplot2::ggsave(pca.plot, bg = "white",
                filename = 
                  paste0(figure.path,"Toothfish_2018_muscle_RANDOM_PCA_CpG_DMS_PC23.png"), 
                width = 30, height = 15, units = "cm")
```


## TOP100 DMS - 2018 - muscle

```{r}
print(load(paste0(rdata.path,"Toothfish_2018_muscle_methylkit_object.Rdata")))
# print(load(paste0(rdata.path,"Toothfish_subset_methylation_objects_raw_NEW.Rdata")))
# meth.2018M.norm11L <- meth.sub
print(load(paste0(rdata.path,"Toothfish_metadata.Rdata"))) 
metadata.new <- metadata[metadata$Genotype_names %in% meth.2018M.norm11L@sample.ids,]
metadata.new <- metadata.new[order(match(metadata.new$Genotype_names,
                                         meth.2018M.norm11L@sample.ids)),]
# metadata.new <- metadata[metadata$File_names %in% meth.2018M.norm11L@sample.ids,]
# metadata.new <- metadata.new[order(match(metadata.new$File_names,
#                                          meth.2018M.norm11L@sample.ids)),]
meth.cpg <- paste0(meth.2018M.norm11L$chr, "_", meth.2018M.norm11L$start) #1250773 
length(meth.cpg)
print(load(paste0(rdata.path,"Toothfish_2018_muscle_methylation_DMS.Rdata")))
DMS.cpg <- paste0(as.character(top100.2018M@seqnames), "_", as.data.frame(top100.2018M@ranges)$start) #283819 
sum(meth.cpg %in% DMS.cpg) #100
meth.DMS <- meth.2018M.norm11L[meth.cpg %in% DMS.cpg,]
```


### Heatmap

```{r}
meth.diff.df <- as.data.frame(t(methylKit::percMethylation(meth.DMS)))
loc.names <- paste0(meth.DMS$chr,"_",meth.DMS$start)
colnames(meth.diff.df) <- loc.names

meth.diff.df$individual <- rownames(meth.diff.df)
pop <- metadata.new$pop
meth.diff.df$pop <- as.character(pop)


heat.plot.data <- meth.diff.df %>%
  tibble::as_tibble() %>%
  tidyr::gather(key = "CpG", value = "Methylation_percentage", -individual, -pop,
                factor_key = FALSE) %>%
  dplyr::group_by(pop, CpG) %>%
  dplyr::mutate(CpG_mean = mean(Methylation_percentage, na.rm = TRUE)) %>%
  dplyr::arrange(pop, individual, CpG) %>%
  dplyr::mutate(individual = factor(individual, levels = unique(individual))) %>%
  dplyr::mutate(CpG = factor(CpG, levels = unique(CpG)))
cpgs <- heat.plot.data$CpG[heat.plot.data$individual == unique(heat.plot.data$individual)[1]]
cpgs_mean <- heat.plot.data$CpG_mean[heat.plot.data$individual == unique(heat.plot.data$individual)[1]]

cpgs <- cpgs[order(cpgs_mean, decreasing = TRUE)]
heat.plot.data$CpG <- factor(heat.plot.data$CpG, levels = c("pop", unique(as.character(cpgs))))
heat.plot.data <- heat.plot.data[order(heat.plot.data$CpG),]

heat.plot <- ggplot2::ggplot(data = heat.plot.data, ggplot2::aes(x = individual)) + 
    ggplot2::geom_tile(data = heat.plot.data, 
                       ggplot2::aes(x = individual, y = CpG, fill = Methylation_percentage)) +
    ggplot2::scale_fill_gradient2(low = "blue", mid = "yellow", high = "red", 
                                  midpoint = 50, guide = "colourbar") +
    ggnewscale::new_scale_fill() +
    ggplot2::geom_tile(data = heat.plot.data, ggplot2::aes(x = individual, y = "pop", fill = pop,
                                    height = length(unique(heat.plot.data$CpG))*0.02)) +
    ggplot2::scale_fill_manual(values = colours4) +
    ggplot2::scale_y_discrete(drop = FALSE) +
    ggplot2::theme(legend.position = "top",
                   axis.title.y = ggplot2::element_blank(),
                   axis.title.x = ggplot2::element_blank(),
                   axis.text.x = ggplot2::element_text(face = "bold", color = "black", 
                                        size = 10, angle = 90, hjust = 1, vjust = 0.5))
print(heat.plot)

ggplot2::ggsave(heat.plot, bg = "white",
                filename = paste0(figure.path,
                                  "Toothfish_2018_muscle_BASED_ON_2018_TOP100_DMS_HEATMAP.png"), 
                width = 30, height = 30, units = "cm")
```

### Plot CpG difference

```{r}
meth.diff <- meth.DMS

length(unique(meth.diff$chr)) #41
sort(table(meth.diff$chr), decreasing = TRUE)[1:20]

meth.diff.df <- as.data.frame(t(methylKit::percMethylation(meth.diff)))
loc.names <- paste0(meth.diff$chr,"_",meth.diff$start)
colnames(meth.diff.df) <- loc.names

meth.diff.df$individual <- rownames(meth.diff.df) 
identical(metadata.new$Genotype_names, meth.diff.df$individual)
meth.diff.df$pop <- metadata.new$pop



meth.diff.df.plot <- tidyr::gather(meth.diff.df, key = locus, 
                           value = perc_methylation, -individual, -pop,
                           factor_key = TRUE) %>%
  dplyr::mutate(chr = stringr::str_split_fixed(locus, 
                                               pattern = "_", 
                                               n = 2)[,1])
meth.diff.df.plot.locus <- meth.diff.df.plot %>% dplyr::group_by(chr,locus,pop) %>% 
  dplyr::summarise(mean_pop = mean(perc_methylation, na.rm = TRUE),
                   sum_NA = sum(is.na(perc_methylation)),
            .groups = 'drop')


diff.plot <- ggplot2::ggplot() +
  ggplot2::geom_line(data = meth.diff.df.plot.locus,
                     ggplot2::aes(x = pop, y = mean_pop, colour = locus, group = locus)) +
  ggplot2::geom_point(data = meth.diff.df.plot.locus,
                     ggplot2::aes(x = pop, y = mean_pop, colour = locus, group = locus),
                     size = 5) +
  ggplot2::facet_wrap(~chr) +
  ggplot2::ylab("Mean methylation (%)") +
  ggplot2::theme(legend.position = "none",
                   axis.text.x = ggplot2::element_text(face = "bold", color = "black", 
                                        size = 10, angle = 90, hjust = 1, vjust = 0.5))

print(diff.plot)

ggplot2::ggsave(plot = diff.plot, 
                filename = paste0(figure.path,"1_Toothfish_2018_muscle_BASED_ON_2018_TOP100_DMS_methDiff.png"),
                width = 30, height = 15, units = "cm", device = "png")
```



### DMS Clustering
```{r}
methdif.perc <- methylKit::percMethylation(meth.DMS)
methdif.perc2 <- zoo::na.aggregate(methdif.perc, by = meth.DMS@treatment)
PCobj <- prcomp(methdif.perc2, retx = TRUE, scale. = TRUE)

var_frac <- PCobj$sdev/sum(PCobj$sdev)
print(paste0("Total variance explained: ",signif(sum(var_frac[1:3]) * 100, 3), "%")) 
barplot(var_frac[1:3], main = "variance explained", 
        col = heat.colors(length(var_frac[1:3])))

# 
# identical(metadata.new$Genotype_names, meth.DMS@sample.ids)
# identical(metadata.new$Genotype_names, rownames(PCobj$rotation))
identical(metadata.new$File_names, meth.DMS@sample.ids)
identical(metadata.new$File_names, rownames(PCobj$rotation))
pop <- metadata.new$pop

plot.title <- "Toothfish methylation - 2018 Muscle (TOP100 DMS from 2018)"

data <- data.frame(ind = rownames(PCobj$rotation),PCobj$rotation,  pop = pop)
data <- merge(data, aggregate(cbind(mean.x = PC1, mean.y = PC2) ~
                                pop, data, mean), by = "pop")

pca.plot <- ggplot2::ggplot(data, ggplot2::aes(PC1, PC2, color = pop)) +
  ggplot2::geom_point(size = 4) + 
  ggplot2::geom_point(ggplot2::aes(x = mean.x,
                                   y = mean.y, color = pop),
                      size = 7,
                      shape = 17) +
  ggplot2::scale_colour_manual(values = colours4) +
  ggplot2::geom_hline(yintercept = 0) +
  ggplot2::geom_vline(xintercept = 0) + 
  ggplot2::labs(
    subtitle = paste0("Total variance explained: ",
                      signif(sum(var_frac[1:3]) * 100, 3), "%"),
    y = paste0("PC2; ", signif(var_frac[2] * 100, digits = 2), " %"),
    x = paste0("PC1; ", signif(var_frac[1] * 100, digits = 2), " %"),
    title = plot.title,
    caption = "") + 
  ggplot2::theme_minimal() +
  ggplot2::theme(
    axis.text = ggplot2::element_text(size = 15),
    axis.title.x = ggplot2::element_text(size = 20),
    axis.title.y = ggplot2::element_text(size = 20),
    legend.text = ggplot2::element_text(size = 15)
  )

print(pca.plot)
ggplot2::ggsave(pca.plot, bg = "white",
                filename = paste0(figure.path,"Toothfish_2018_muscle_BASED_ON_2018_TOP100_DMS_PC12.png"), 
                width = 30, height = 15, units = "cm")



data <- data.frame(ind = rownames(PCobj$rotation),PCobj$rotation,  pop = pop)
data <- merge(data, aggregate(cbind(mean.x = PC1, mean.y = PC3) ~
                                pop, data, mean), by = "pop")

pca.plot <- ggplot2::ggplot(data, ggplot2::aes(PC1, PC3, color = pop)) +
  ggplot2::geom_point(size = 4) + 
  ggplot2::geom_point(ggplot2::aes(x = mean.x,
                                   y = mean.y, color = pop),
                      size = 7,
                      shape = 17) +
  ggplot2::scale_colour_manual(values = colours4) +
  ggplot2::geom_hline(yintercept = 0) +
  ggplot2::geom_vline(xintercept = 0) + 
  ggplot2::labs(
    subtitle = paste0("Total variance explained: ",
                      signif(sum(var_frac[1:3]) * 100, 3), "%"),
    y = paste0("PC3; ", signif(var_frac[3] * 100, digits = 2), " %"),
    x = paste0("PC1; ", signif(var_frac[1] * 100, digits = 2), " %"),
    title = plot.title,
    caption = "") + 
  ggplot2::theme_minimal() +
  ggplot2::theme(
    axis.text = ggplot2::element_text(size = 15),
    axis.title.x = ggplot2::element_text(size = 20),
    axis.title.y = ggplot2::element_text(size = 20),
    legend.text = ggplot2::element_text(size = 15)
  )

print(pca.plot)
ggplot2::ggsave(pca.plot, bg = "white",
                filename = paste0(figure.path,
                                  "Toothfish_2018_muscle_BASED_ON_2018_TOP100_DMS_PC13.png"), 
                width = 30, height = 15, units = "cm")




data <- data.frame(ind = rownames(PCobj$rotation),PCobj$rotation,  pop = pop)
data <- merge(data, aggregate(cbind(mean.x = PC2, mean.y = PC3) ~
                                pop, data, mean), by = "pop")

pca.plot <- ggplot2::ggplot(data, ggplot2::aes(PC2, PC3, color = pop)) +
  ggplot2::geom_point(size = 4) + 
  ggplot2::geom_point(ggplot2::aes(x = mean.x,
                                   y = mean.y, color = pop),
                      size = 7,
                      shape = 17) +
  ggplot2::scale_colour_manual(values = colours4) +
  ggplot2::geom_hline(yintercept = 0) +
  ggplot2::geom_vline(xintercept = 0) + 
  ggplot2::labs(
    subtitle = paste0("Total variance explained: ",
                      signif(sum(var_frac[1:3]) * 100, 3), "%"),
    y = paste0("PC3; ", signif(var_frac[3] * 100, digits = 2), " %"),
    x = paste0("PC2; ", signif(var_frac[2] * 100, digits = 2), " %"),
    title = plot.title,
    caption = "") + 
  ggplot2::theme_minimal() +
  ggplot2::theme(
    axis.text = ggplot2::element_text(size = 15),
    axis.title.x = ggplot2::element_text(size = 20),
    axis.title.y = ggplot2::element_text(size = 20),
    legend.text = ggplot2::element_text(size = 15)
  )

print(pca.plot)
ggplot2::ggsave(pca.plot, bg = "white",
                filename = paste0(figure.path,"Toothfish_2018_muscle_BASED_ON_2018_TOP100_DMS_PC23.png"), 
                width = 30, height = 15, units = "cm")
```



## GLMnet Re-assignment success of 2018 - training/test

```{r}
print(load(paste0(rdata.path,"Toothfish_2018_muscle_methylkit_object.Rdata")))
print(load(paste0(rdata.path,"Toothfish_metadata.Rdata"))) 
metadata.new <- metadata[metadata$Genotype_names %in% meth.2018M.norm11L@sample.ids,]
metadata.new <- metadata.new[order(match(metadata.new$File_names,
                                         meth.2018M.norm11L@sample.ids)),]
meth.cpg <- paste0(meth.2018M.norm11L$chr, "_", meth.2018M.norm11L$start) #1250773 
print(load(paste0(rdata.path,"Toothfish_2018_muscle_methylation_DMS.Rdata")))
DMS.cpg <- paste0(myDiff25p$chr, "_", myDiff25p$start) #1202

sum(meth.cpg %in% DMS.cpg) #1202
DMS.cpg2 <- meth.cpg[meth.cpg %in% DMS.cpg]
meth.DMS1 <- meth.2018M.norm11L[meth.cpg %in% DMS.cpg,]


meth.pct.df <- methylKit::percMethylation(meth.DMS1)
meth.pct.df <- zoo::na.aggregate(meth.pct.df, by = meth.2018M.norm11L@treatment)
meth.pct.df <-  as.data.frame(t(meth.pct.df))
colnames(meth.pct.df) <- DMS.cpg2
meth.pct.df$pop <- metadata.new$STRATA2

splits <- rsample::initial_split(meth.pct.df, strata = pop, prop = 3/4)
# <Training/Testing/Total>
#   <82/29/111>
testTransformed <- rsample::testing(splits)
trainTransformed3 <- rsample::training(splits)
```

### glmnet model
```{r}
fitControl <- caret::trainControl(method = "repeatedcv", number = 10, repeats = 10)

elastic_model <-
  caret::train(x = trainTransformed3[,-ncol(trainTransformed3)],
               y = trainTransformed3$pop,
    method = "glmnet",
    trControl = fitControl,
    tuneLength = 10
  )


predicted.class <- caret::predict.train(elastic_model)
predicted.class <- as.character(predicted.class)
# caret::postResample(pred = predicted.class, trainTransformed3$pop)
chisq.test(predicted.class, trainTransformed3$pop) #X-squared = 164, df = 4, p-value < 2.2e-16
predicted.class == trainTransformed3$pop # all TRUE
sum(predicted.class == trainTransformed3$pop) #82
predict.enet.test <- predict(elastic_model, testTransformed)
predict.enet.test <- as.character(predict.enet.test)
# caret::postResample(pred = predict.enet.test, testTransformed$pop)
# plot(testTransformed$pop, predict.enet.test)
chisq.test(predict.enet.test, testTransformed$pop) #X-squared = 38.816, df = 4, p-value = 7.603e-08
predict.enet.test == testTransformed$pop 
sum(predict.enet.test == testTransformed$pop)/length(testTransformed$pop) #86.2%
```

### Variable contributions

```{r}
varImp.ridge <- caret::varImp(elastic_model, scale = FALSE)
plot(varImp.ridge, top = 100)
dev.print(file = paste0(figure.path,"TOA_2018_muscle_glmnet_varimp.png"),
          device = png, res = 300,width = 45,height = 15,units = "cm")

varImp.ridge.df <- varImp.ridge$importance
varImp.ridge.df$Overall <- rowSums(varImp.ridge.df)
varImp.ridge.df$locus <- rownames(varImp.ridge.df)
varImp.ridge.names_Overall <- varImp.ridge.df$locus[varImp.ridge.df$Overall > 0.001]
length(varImp.ridge.names_Overall) #136

# save(meth.pct.df,testTransformed, trainTransformed3,varImp.ridge,varImp.ridge.df,
#      varImp.ridge.names_Overall,
#      file = paste0(rdata.path, "TOA_2018_muscle_glmnet.Rdata"))

print(load(paste0(rdata.path, "TOA_2018_muscle_glmnet.Rdata")))


meth.diff.df <- meth.pct.df
meth.diff.df$individual <- rownames(meth.diff.df)
print(load(paste0(rdata.path,"Toothfish_metadata.Rdata")))
identical(metadata.new$Genotype_names, meth.diff.df$individual)

meth.diff.df.plot <- tidyr::gather(meth.diff.df, key = locus, 
                                   value = perc_methylation, -individual, -pop,
                                   factor_key = TRUE) %>%
  dplyr::mutate(chr = stringr::str_split_fixed(locus, 
                                               pattern = "_", 
                                               n = 2)[,1])
meth.diff.df.plot <- meth.diff.df.plot[meth.diff.df.plot$locus %in% varImp.ridge.names_Overall, ]
meth.diff.df.plot.locus <- meth.diff.df.plot %>% dplyr::group_by(chr,locus,pop) %>% 
  dplyr::summarise(mean_pop = mean(perc_methylation, na.rm = TRUE),
            .groups = 'drop')


region.meth.plot2 <- ggplot2::ggplot() +
  ggplot2::geom_point(data = meth.diff.df.plot,
                      ggplot2::aes(x = pop, y = perc_methylation, colour = locus)) +
  ggplot2::geom_line(data = meth.diff.df.plot.locus,
                     ggplot2::aes(x = pop, y = mean_pop, colour = locus, group = locus)) +
  ggplot2::geom_point(data = meth.diff.df.plot.locus,
                     ggplot2::aes(x = pop, y = mean_pop, colour = locus, group = locus),
                     size = 5) +
  ggplot2::facet_wrap(~chr) + 
  ggplot2::guides(colour = ggplot2::guide_legend(ncol = 2)) + 
  ggplot2::theme(axis.text.x = ggplot2::element_text(angle = 90))
print(region.meth.plot2)


ggplot2::ggsave(plot = region.meth.plot2,
                filename = paste0(figure.path,
                                  "TOA_2018_muscles_Regional_methylation_percentage_ridge.png"
                                  ),
                                  width = 45, height = 30,
                                  units = "cm", device = "png")
```







## Maturity DMS - 2018 - muscle

```{r, eval=FALSE}
print(load(paste0(rdata.path,"Toothfish_2018_muscle_methylkit_object.Rdata")))

print(load(paste0(rdata.path,"Toothfish_metadata.Rdata")))
metadata.new <- metadata[metadata$Genotype_names %in% attributes(meth.2018M.norm11L)$sample.ids,]
attributes(meth.2018M.norm11L)$sample.ids == metadata.new$Genotype_names
mat <- as.character(metadata.new$Maturity_full)
mat[is.na(mat)] <- "Resting/Developing"
mat[mat == "Gravid"] <- "Spent"

table(factor(mat))
meth.2018M.norm11L@treatment <- as.integer(factor(mat))


# data.frame(attributes(meth.2018M.norm11L)$sample.ids, metadata.new$Genotype_names, metadata.new$strata, meth.2018M.norm11L@treatment)

covariates <- dplyr::select(metadata.new, Sex, Age, Cohort2, Maturity_full)
myDiff <- methylKit::calculateDiffMeth(meth.2018M.norm11L,
                                       covariates = NULL,
                                       adjust = "fdr",
                                       chunk.size = 1e+06,
                                       mc.cores = 25)
meth.diff.chr <- methylKit::diffMethPerChr(myDiff,plot = TRUE, 
                                           qvalue.cutoff = 0.01, meth.cutoff = 40)
myDiff40p <- methylKit::getMethylDiff(myDiff, difference = 40, qvalue = 0.01)
#3
myDiff25p <- methylKit::getMethylDiff(myDiff, difference = 25, qvalue = 0.01)
#195
myDiff15p <- methylKit::getMethylDiff(myDiff, difference = 15, qvalue = 0.01)
#4851   
top100.2018Mmat <- slice_max(as.data.frame(as(myDiff,"GRanges")), order_by = meth.diff, n = 100)
top100.2018Mmat <- as(top100.2018Mmat,"GRanges")

save(myDiff,myDiff15p, myDiff25p, myDiff40p,top100.2018Mmat,
     file = paste0(rdata.path,"Toothfish_2018_muscle_MATURITY_methylation_DMS.Rdata"))
```

```{r}
print(load(paste0(rdata.path,"Toothfish_2018_muscle_MATURITY_methylation_DMS.Rdata")))
volcano.plot <- ggplot2::ggplot() +
  ggplot2::geom_point(data = myDiff, ggplot2::aes(x = meth.diff, 
                                                  y = -log10(qvalue))) +
  ggplot2::geom_point(data = myDiff25p, ggplot2::aes(x = meth.diff, 
                                                  y = -log10(qvalue)),
                      colour = "red") +
  ggplot2::theme_minimal() +
  ggplot2::theme(
    axis.text = ggplot2::element_text(size = 15),
    axis.title.x = ggplot2::element_text(size = 20),
    axis.title.y = ggplot2::element_text(size = 20),
    legend.text = ggplot2::element_text(size = 15)
  )
print(volcano.plot)

ggplot2::ggsave(volcano.plot, bg = "white",
                filename =
                  paste0(figure.path,
                         "Toothfish_2018_muscle_MATURITY_DMS25_VOLCANO_methylkit_regions.png"), 
                width = 30, height = 15, units = "cm")



myDiff25p.gr <- as(myDiff25p, "GRanges")
grand.plot <- ggbio::plotGrandLinear(myDiff25p.gr,
                                     ggplot2::aes(y = meth.diff)) +
  ggplot2::ylab("Methylation difference (%)") +
  ggplot2::theme_bw() +
  ggplot2::theme(axis.text.x = ggplot2::element_text(angle = 90, hjust = 0), 
                 legend.position = "none")
print(grand.plot)
ggplot2::ggsave(grand.plot@ggplot, bg = "white",
                filename = paste0(figure.path,
                                  "Toothfish_2018_muscle_MATURITY_DMS25_Grandplot_methylkit.png"
                ), 
                width = 30, height = 15, units = "cm")
```


### Heatmap

```{r}
print(load(paste0(rdata.path,"Toothfish_2018_muscle_MATURITY_methylation_DMS.Rdata")))

myDiff25p.gr <- as(myDiff25p, "GRanges")
meth.diff <- methylKit::selectByOverlap(meth.2018M.norm11L, myDiff25p.gr)

meth.diff.df <- as.data.frame(t(methylKit::percMethylation(meth.diff)))
loc.names <- paste0(meth.diff$chr,"_",meth.diff$start)
colnames(meth.diff.df) <- loc.names

meth.diff.df$individual <- rownames(meth.diff.df)
pop <- metadata.new$Maturity_full
meth.diff.df$pop <- as.character(pop)


heat.plot.data <- meth.diff.df %>%
  tibble::as_tibble() %>%
  tidyr::gather(key = "CpG", value = "Methylation_percentage", -individual, -pop,
                factor_key = FALSE) %>%
  dplyr::group_by(pop, CpG) %>%
  dplyr::mutate(CpG_mean = mean(Methylation_percentage, na.rm = TRUE)) %>%
  dplyr::arrange(pop, individual, CpG) %>%
  dplyr::mutate(individual = factor(individual, levels = unique(individual))) %>%
  dplyr::mutate(CpG = factor(CpG, levels = unique(CpG)))
cpgs <- heat.plot.data$CpG[heat.plot.data$individual == unique(heat.plot.data$individual)[1]]
cpgs_mean <- heat.plot.data$CpG_mean[heat.plot.data$individual == unique(heat.plot.data$individual)[1]]

cpgs <- cpgs[order(cpgs_mean, decreasing = TRUE)]
heat.plot.data$CpG <- factor(heat.plot.data$CpG, levels = c("pop", unique(as.character(cpgs))))
heat.plot.data <- heat.plot.data[order(heat.plot.data$CpG),]

heat.plot <- ggplot2::ggplot(data = heat.plot.data, ggplot2::aes(x = individual)) + 
    ggplot2::geom_tile(data = heat.plot.data, 
                       ggplot2::aes(x = individual, y = CpG, fill = Methylation_percentage)) +
    ggplot2::scale_fill_gradient2(low = "blue", mid = "yellow", high = "red", 
                                  midpoint = 50, guide = "colourbar") +
    ggnewscale::new_scale_fill() +
    ggplot2::geom_tile(data = heat.plot.data, ggplot2::aes(x = individual, y = "pop", fill = pop,
                                    height = length(unique(heat.plot.data$CpG))*0.02)) +
    ggplot2::scale_fill_manual(values = c("#78B7C5", "#08306B","#E1AF00","#F21A00"),
                             na.value = "#999999") +
    ggplot2::scale_y_discrete(drop = FALSE) +
    ggplot2::theme(legend.position = "top",
                   axis.title.y = ggplot2::element_blank(),
                   axis.title.x = ggplot2::element_blank(),
                   axis.text.x = ggplot2::element_text(face = "bold", color = "black", 
                                        size = 10, angle = 90, hjust = 1, vjust = 0.5))
print(heat.plot)

ggplot2::ggsave(heat.plot, bg = "white",
                filename = paste0(figure.path,
                                  "Toothfish_2018_muscle_MATURITY_methDiff_25_HEATMAP.png"), 
                width = 30, height = 30, units = "cm")
```

### Plot CpG difference

```{r}
load(file = paste0(rdata.path,"Toothfish_2018_muscle_methylation_DMS.Rdata"))

myDiff25p.gr <- as(myDiff25p, "GRanges")
meth.diff <- methylKit::selectByOverlap(meth.2018M.norm11L, myDiff25p.gr) #113 

length(unique(meth.diff$chr)) #41
sort(table(meth.diff$chr), decreasing = TRUE)[1:20]

meth.diff.df <- as.data.frame(t(methylKit::percMethylation(meth.diff)))
loc.names <- paste0(meth.diff$chr,"_",meth.diff$start)
colnames(meth.diff.df) <- loc.names

meth.diff.df$individual <- rownames(meth.diff.df) 
identical(metadata.new$Genotype_names, meth.diff.df$individual)
meth.diff.df$pop <- metadata.new$Maturity_full



meth.diff.df.plot <- tidyr::gather(meth.diff.df, key = locus, 
                           value = perc_methylation, -individual, -pop,
                           factor_key = TRUE) %>%
  dplyr::mutate(chr = stringr::str_split_fixed(locus, 
                                               pattern = "_", 
                                               n = 2)[,1])
meth.diff.df.plot.locus <- meth.diff.df.plot %>% dplyr::group_by(chr,locus,pop) %>% 
  dplyr::summarise(mean_pop = mean(perc_methylation, na.rm = TRUE),
                   sum_NA = sum(is.na(perc_methylation)),
            .groups = 'drop')


diff.plot <- ggplot2::ggplot() +
  ggplot2::geom_line(data = meth.diff.df.plot.locus,
                     ggplot2::aes(x = pop, y = mean_pop, colour = locus, group = locus)) +
  ggplot2::geom_point(data = meth.diff.df.plot.locus,
                     ggplot2::aes(x = pop, y = mean_pop, colour = locus, group = locus),
                     size = 5) +
  ggplot2::facet_wrap(~chr) +
  ggplot2::ylab("Mean methylation (%)") +
  ggplot2::theme(legend.position = "none",
                   axis.text.x = ggplot2::element_text(face = "bold", color = "black", 
                                        size = 10, angle = 90, hjust = 1, vjust = 0.5))

print(diff.plot)

ggplot2::ggsave(plot = diff.plot, 
                filename = paste0(figure.path,"1_Toothfish_2018_muscle_MATURITY_methDiff.png"),
                width = 30, height = 15, units = "cm", device = "png")
```


### DMS Clustering
```{r}
load(file = paste0(rdata.path,"Toothfish_2018_muscle_methylation_DMS.Rdata"))

myDiff25p.gr <- as(myDiff25p, "GRanges")
meth.diff <- methylKit::selectByOverlap(meth.2018M.norm11L, myDiff25p.gr) #112  
methdif.perc <- methylKit::percMethylation(meth.diff)

methdif.perc2 <- zoo::na.aggregate(methdif.perc, by = meth.2018M.norm11L@treatment)
PCobj <- prcomp(methdif.perc2, retx = TRUE, scale. = TRUE)

var_frac <- PCobj$sdev/sum(PCobj$sdev)
print(paste0("Total variance explained: ",signif(sum(var_frac[1:3]) * 100, 3), "%")) 
barplot(var_frac[1:3], main = "variance explained", 
        col = heat.colors(length(var_frac[1:3])))


identical(metadata.new$Genotype_names, meth.2018M.norm11L@sample.ids)
# identical(metadata.new$individual, rownames(PCobj$x))
identical(metadata.new$Genotype_names, rownames(PCobj$rotation))
pop <- metadata.new$Maturity_full

plot.title <- "Toothfish methylation - 2018 Muscle MATURITY (195 DMS sites - 25%)"

# data <- data.frame(ind = rownames(PCobj$x),PCobj$x,  pop = pop)
data <- data.frame(ind = rownames(PCobj$rotation),PCobj$rotation,  pop = pop)
data <- merge(data, aggregate(cbind(mean.x = PC1, mean.y = PC2) ~
                                pop, data, mean), by = "pop")

pca.plot <- ggplot2::ggplot(data, ggplot2::aes(PC1, PC2, color = pop)) +
  ggplot2::geom_point(size = 4) + 
  ggplot2::geom_point(ggplot2::aes(x = mean.x,
                                   y = mean.y, color = pop),
                      size = 7,
                      shape = 17) +
  ggplot2::scale_colour_manual(values = c("#78B7C5", "#08306B","#E1AF00","#F21A00"),
                             na.value = "#999999") +
  ggplot2::geom_hline(yintercept = 0) +
  ggplot2::geom_vline(xintercept = 0) + 
  ggplot2::labs(
    subtitle = paste0("Total variance explained: ",
                      signif(sum(var_frac[1:3]) * 100, 3), "%"),
    y = paste0("PC2; ", signif(var_frac[2] * 100, digits = 2), " %"),
    x = paste0("PC1; ", signif(var_frac[1] * 100, digits = 2), " %"),
    title = plot.title,
    caption = "") + 
  ggplot2::theme_minimal() +
  ggplot2::theme(
    axis.text = ggplot2::element_text(size = 15),
    axis.title.x = ggplot2::element_text(size = 20),
    axis.title.y = ggplot2::element_text(size = 20),
    legend.text = ggplot2::element_text(size = 15)
  )

print(pca.plot)
ggplot2::ggsave(pca.plot, bg = "white",
                filename = paste0(figure.path,"Toothfish_2018_muscle_MATURITY_PCA_CpG_DMS_PC12.png"), 
                width = 30, height = 15, units = "cm")



# data <- data.frame(ind = rownames(PCobj$x),PCobj$x,  pop = pop)
data <- data.frame(ind = rownames(PCobj$rotation),PCobj$rotation,  pop = pop)
data <- merge(data, aggregate(cbind(mean.x = PC1, mean.y = PC3) ~
                                pop, data, mean), by = "pop")

pca.plot <- ggplot2::ggplot(data, ggplot2::aes(PC1, PC3, color = pop)) +
  ggplot2::geom_point(size = 4) + 
  ggplot2::geom_point(ggplot2::aes(x = mean.x,
                                   y = mean.y, color = pop),
                      size = 7,
                      shape = 17) +
  ggplot2::scale_colour_manual(values = c("#78B7C5", "#08306B","#E1AF00","#F21A00"),
                             na.value = "#999999") +
  ggplot2::geom_hline(yintercept = 0) +
  ggplot2::geom_vline(xintercept = 0) + 
  ggplot2::labs(
    subtitle = paste0("Total variance explained: ",
                      signif(sum(var_frac[1:3]) * 100, 3), "%"),
    y = paste0("PC3; ", signif(var_frac[3] * 100, digits = 2), " %"),
    x = paste0("PC1; ", signif(var_frac[1] * 100, digits = 2), " %"),
    title = plot.title,
    caption = "") + 
  ggplot2::theme_minimal() +
  ggplot2::theme(
    axis.text = ggplot2::element_text(size = 15),
    axis.title.x = ggplot2::element_text(size = 20),
    axis.title.y = ggplot2::element_text(size = 20),
    legend.text = ggplot2::element_text(size = 15)
  )

print(pca.plot)
ggplot2::ggsave(pca.plot, bg = "white",
                filename = paste0(figure.path,
                                  "Toothfish_2018_muscle_MATURITY_PCA_CpG_DMS_PC13.png"), 
                width = 30, height = 15, units = "cm")




# data <- data.frame(ind = rownames(PCobj$x),PCobj$x,  pop = pop)
data <- data.frame(ind = rownames(PCobj$rotation),PCobj$rotation,  pop = pop)
data <- merge(data, aggregate(cbind(mean.x = PC2, mean.y = PC3) ~
                                pop, data, mean), by = "pop")

pca.plot <- ggplot2::ggplot(data, ggplot2::aes(PC2, PC3, color = pop)) +
  ggplot2::geom_point(size = 4) + 
  ggplot2::geom_point(ggplot2::aes(x = mean.x,
                                   y = mean.y, color = pop),
                      size = 7,
                      shape = 17) +
  ggplot2::scale_colour_manual(values = c("#78B7C5", "#08306B","#E1AF00","#F21A00"),
                             na.value = "#999999") +
  ggplot2::geom_hline(yintercept = 0) +
  ggplot2::geom_vline(xintercept = 0) + 
  ggplot2::labs(
    subtitle = paste0("Total variance explained: ",
                      signif(sum(var_frac[1:3]) * 100, 3), "%"),
    y = paste0("PC3; ", signif(var_frac[3] * 100, digits = 2), " %"),
    x = paste0("PC2; ", signif(var_frac[2] * 100, digits = 2), " %"),
    title = plot.title,
    caption = "") + 
  ggplot2::theme_minimal() +
  ggplot2::theme(
    axis.text = ggplot2::element_text(size = 15),
    axis.title.x = ggplot2::element_text(size = 20),
    axis.title.y = ggplot2::element_text(size = 20),
    legend.text = ggplot2::element_text(size = 15)
  )

print(pca.plot)
ggplot2::ggsave(pca.plot, bg = "white",
                filename = paste0(figure.path,"Toothfish_2018_muscle_MATURITY_PCA_CpG_DMS_PC23.png"), 
                width = 30, height = 15, units = "cm")
```



## Maturity DMS RANDOM - 2018 - muscle

```{r, eval=FALSE}
print(load(paste0(rdata.path,"Toothfish_2018_muscle_methylkit_object.Rdata")))

print(load(paste0(rdata.path,"Toothfish_metadata.Rdata")))
metadata.new <- metadata[metadata$Genotype_names %in% attributes(meth.2018M.norm11L)$sample.ids,]
attributes(meth.2018M.norm11L)$sample.ids == metadata.new$Genotype_names



strata.random <- sample(metadata.new$Maturity_full, size = length(metadata.new$Maturity_full), replace = FALSE)

strata.random <- as.character(strata.random)
strata.random[is.na(strata.random)] <- "Resting/Developing"
strata.random[strata.random == "Gravid"] <- "Spent"
meth.2018M.norm11L@treatment <- as.integer(factor(strata.random))

# data.frame(attributes(meth.2018M.norm11L)$sample.ids, metadata.new$Genotype_names, metadata.new$strata, meth.2018M.norm11L@treatment)

# covariates <- dplyr::select(metadata.new, Sex, TL_mm)
myDiff <- methylKit::calculateDiffMeth(meth.2018M.norm11L,
                                       covariates = NULL,
                                       adjust = "fdr",
                                       chunk.size = 1e+06,
                                       mc.cores = 25)
meth.diff.chr <- methylKit::diffMethPerChr(myDiff,plot = TRUE, 
                                           qvalue.cutoff = 0.01, meth.cutoff = 40)
myDiff40p <- methylKit::getMethylDiff(myDiff, difference = 40, qvalue = 0.01)
#3
myDiff25p <- methylKit::getMethylDiff(myDiff, difference = 25, qvalue = 0.01)
#131
myDiff15p <- methylKit::getMethylDiff(myDiff, difference = 15, qvalue = 0.01)
#2698   

save(strata.random, myDiff,myDiff15p, myDiff25p, myDiff40p,
     file = paste0(rdata.path,"Toothfish_2018_muscle_MATURITY_methylation_RANDOM_DMS.Rdata"))
```

```{r}
print(load(paste0(rdata.path,"Toothfish_2018_muscle_methylation_RANDOM_DMS.Rdata")))
volcano.plot <- ggplot2::ggplot() +
  ggplot2::geom_point(data = myDiff, ggplot2::aes(x = meth.diff, 
                                                  y = -log10(qvalue))) +
  ggplot2::geom_point(data = myDiff25p, ggplot2::aes(x = meth.diff, 
                                                  y = -log10(qvalue)),
                      colour = "red") +
  
  ggplot2::theme_minimal() +
  ggplot2::theme(
    axis.text = ggplot2::element_text(size = 15),
    axis.title.x = ggplot2::element_text(size = 20),
    axis.title.y = ggplot2::element_text(size = 20),
    legend.text = ggplot2::element_text(size = 15)
  )
print(volcano.plot)

ggplot2::ggsave(volcano.plot, bg = "white",
                filename =
                  paste0(figure.path,
                         "Toothfish_2018_muscle_MATURITY_RANDOM_DMS25_VOLCANO_methylkit.png"), 
                width = 30, height = 15, units = "cm")


myDiff25p.gr <- as(myDiff25p, "GRanges")
grand.plot <- ggbio::plotGrandLinear(myDiff25p.gr,
                                     ggplot2::aes(y = meth.diff)) +
  ggplot2::ylab("Methylation difference (%)") +
  ggplot2::theme_bw() +
  ggplot2::theme(axis.text.x = ggplot2::element_text(angle = 90, hjust = 0), 
                 legend.position = "none")
print(grand.plot)
ggplot2::ggsave(grand.plot@ggplot, bg = "white",
                filename = paste0(figure.path,
                                  "Toothfish_2018_muscle_MATURITY_RANDOM_DMS25_Grandplot_methylkit.png"
                ), 
                width = 30, height = 15, units = "cm")
```


### Heatmap

```{r}
print(load(paste0(rdata.path,"Toothfish_2018_muscle_methylation_RANDOM_DMS.Rdata")))

myDiff25p.gr <- as(myDiff25p, "GRanges")
meth.diff <- methylKit::selectByOverlap(meth.2018M.norm11L, myDiff25p.gr)

meth.diff.df <- as.data.frame(t(methylKit::percMethylation(meth.diff)))
loc.names <- paste0(meth.diff$chr,"_",meth.diff$start)
colnames(meth.diff.df) <- loc.names

meth.diff.df$individual <- rownames(meth.diff.df)
meth.diff.df$pop <- strata.random
meth.diff.df$pop2 <- as.character(metadata.new$pop)

heat.plot.data <- meth.diff.df %>%
  tibble::as_tibble() %>%
  tidyr::gather(key = "CpG", value = "Methylation_percentage", -individual, -pop, -pop2,
                factor_key = FALSE) %>%
  dplyr::group_by(pop, CpG) %>%
  dplyr::mutate(CpG_mean = mean(Methylation_percentage, na.rm = TRUE)) %>%
  dplyr::arrange(pop, individual, CpG) %>%
  dplyr::mutate(individual = factor(individual, levels = unique(individual))) %>%
  dplyr::mutate(CpG = factor(CpG, levels = unique(CpG)))
cpgs <- heat.plot.data$CpG[heat.plot.data$individual == unique(heat.plot.data$individual)[1]]
cpgs_mean <- heat.plot.data$CpG_mean[heat.plot.data$individual == unique(heat.plot.data$individual)[1]]

cpgs <- cpgs[order(cpgs_mean, decreasing = TRUE)]
heat.plot.data$CpG <- factor(heat.plot.data$CpG, levels = c("pop", unique(as.character(cpgs))))
heat.plot.data <- heat.plot.data[order(heat.plot.data$CpG),]

heat.plot <- ggplot2::ggplot(data = heat.plot.data, ggplot2::aes(x = individual)) + 
    ggplot2::geom_tile(data = heat.plot.data, 
                       ggplot2::aes(x = individual, y = CpG, fill = Methylation_percentage)) +
    ggplot2::scale_fill_gradient2(low = "blue", mid = "yellow", high = "red", 
                                  midpoint = 50, guide = "colourbar") +
    ggnewscale::new_scale_fill() +
    ggplot2::geom_tile(data = heat.plot.data, ggplot2::aes(x = individual, 
                                                           y = "pop-random", fill = pop,
                                    height = length(unique(heat.plot.data$CpG))*0.02)) +
    ggplot2::scale_fill_manual(values = colours4) +
    ggnewscale::new_scale_fill() +
    ggplot2::geom_tile(data = heat.plot.data, ggplot2::aes(x = individual, y = "pop", fill = pop2,
                                    height = length(unique(heat.plot.data$CpG))*0.02)) +
    ggplot2::scale_fill_manual(values = colours4) +
    ggplot2::scale_y_discrete(drop = FALSE) +
    ggplot2::theme(legend.position = "top",
                   axis.title.y = ggplot2::element_blank(),
                   axis.title.x = ggplot2::element_blank(),
                   axis.text.x = ggplot2::element_text(face = "bold", color = "black", 
                                        size = 10, angle = 90, hjust = 1, vjust = 0.5))
print(heat.plot)

ggplot2::ggsave(heat.plot, bg = "white",
                filename = paste0(figure.path,
                                  "Toothfish_2018_muscle_MATURITY_RANDOM_methDiff_25_HEATMAP.png"), 
                width = 30, height = 30, units = "cm")
```

### Plot CpG difference

```{r}
load(file = paste0(rdata.path,"Toothfish_2018_muscle_MATURITY_methylation_RANDOM_DMS.Rdata"))

myDiff25p.gr <- as(myDiff25p, "GRanges")
meth.diff <- methylKit::selectByOverlap(meth.2018M.norm11L, myDiff25p.gr) #113 

length(unique(meth.diff$chr)) #34
sort(table(meth.diff$chr), decreasing = TRUE)[1:20]

meth.diff.df <- as.data.frame(t(methylKit::percMethylation(meth.diff)))
loc.names <- paste0(meth.diff$chr,"_",meth.diff$start)
colnames(meth.diff.df) <- loc.names

meth.diff.df$individual <- rownames(meth.diff.df) 
identical(metadata.new$Genotype_names, meth.diff.df$individual)
meth.diff.df$pop <- metadata.new$pop


meth.diff.df.plot <- tidyr::gather(meth.diff.df, key = locus, 
                           value = perc_methylation, -individual, -pop,
                           factor_key = TRUE) %>%
  dplyr::mutate(chr = stringr::str_split_fixed(locus, 
                                               pattern = "_", 
                                               n = 2)[,1])
meth.diff.df.plot.locus <- meth.diff.df.plot %>% dplyr::group_by(chr,locus,pop) %>% 
  dplyr::summarise(mean_pop = mean(perc_methylation, na.rm = TRUE),
                   sum_NA = sum(is.na(perc_methylation)),
            .groups = 'drop')


diff.plot <- ggplot2::ggplot() +
  ggplot2::geom_line(data = meth.diff.df.plot.locus,
                     ggplot2::aes(x = pop, y = mean_pop, colour = locus, group = locus)) +
  ggplot2::geom_point(data = meth.diff.df.plot.locus,
                     ggplot2::aes(x = pop, y = mean_pop, colour = locus, group = locus),
                     size = 5) +
  ggplot2::facet_wrap(~chr) +
  ggplot2::ylab("Mean methylation (%)") +
  ggplot2::theme(legend.position = "none",
                   axis.text.x = ggplot2::element_text(face = "bold", color = "black", 
                                        size = 10, angle = 90, hjust = 1, vjust = 0.5))
    # ggplot2::guides(colour = ggplot2::guide_legend(ncol = 2))

print(diff.plot)

ggplot2::ggsave(plot = diff.plot, 
                filename = paste0(figure.path,"1_Toothfish_2018_muscle_MATURITY_RANDOM_methDiff.png"),
                width = 30, height = 15, units = "cm", device = "png")
```


### DMS Clustering
```{r}
load(file = paste0(rdata.path,"Toothfish_2018_muscle_methylation_RANDOM_DMS.Rdata"))

myDiff25p.gr <- as(myDiff25p, "GRanges")
meth.diff <- methylKit::selectByOverlap(meth.2018M.norm11L, myDiff25p.gr) #270  
methdif.perc <- methylKit::percMethylation(meth.diff)

methdif.perc2 <- zoo::na.aggregate(methdif.perc, by = meth.2018M.norm11L@treatment)
PCobj <- prcomp(methdif.perc2, retx = TRUE, scale. = TRUE)

var_frac <- PCobj$sdev/sum(PCobj$sdev)
print(paste0("Total variance explained: ",signif(sum(var_frac[1:3]) * 100, 3), "%")) 
barplot(var_frac[1:3], main = "variance explained", 
        col = heat.colors(length(var_frac[1:3])))


identical(metadata.new$Genotype_names, meth.2018M.norm11L@sample.ids)

identical(as.numeric(factor(metadata.new$pop)), meth.2018M.norm11L@treatment)

# identical(metadata.new$individual, rownames(PCobj$x))
identical(metadata.new$Genotype_names, rownames(PCobj$rotation))
pop <- metadata.new$pop

plot.title <- "Toothfish methylation - RANDOMISED 2018 Muscle (193 DMS sites - 25%)"

# data <- data.frame(ind = rownames(PCobj$x),PCobj$x,  pop = pop)
data <- data.frame(ind = rownames(PCobj$rotation),PCobj$rotation,  pop = pop)
data <- merge(data, aggregate(cbind(mean.x = PC1, mean.y = PC2) ~
                                pop, data, mean), by = "pop")

pca.plot <- ggplot2::ggplot(data, ggplot2::aes(PC1, PC2, color = pop)) +
  ggplot2::geom_point(size = 4) + 
  ggplot2::geom_point(ggplot2::aes(x = mean.x,
                                   y = mean.y, color = pop),
                      size = 7,
                      shape = 17) +
  ggplot2::scale_colour_manual(values = colours4) +
  ggplot2::geom_hline(yintercept = 0) +
  ggplot2::geom_vline(xintercept = 0) + 
  ggplot2::labs(
    subtitle = paste0("Total variance explained: ",
                      signif(sum(var_frac[1:3]) * 100, 3), "%"),
    y = paste0("PC2; ", signif(var_frac[2] * 100, digits = 2), " %"),
    x = paste0("PC1; ", signif(var_frac[1] * 100, digits = 2), " %"),
    title = plot.title,
    caption = "") + 
  ggplot2::theme_minimal() +
  ggplot2::theme(
    axis.text = ggplot2::element_text(size = 15),
    axis.title.x = ggplot2::element_text(size = 20),
    axis.title.y = ggplot2::element_text(size = 20),
    legend.text = ggplot2::element_text(size = 15)
  )

print(pca.plot)
ggplot2::ggsave(pca.plot, bg = "white",
                filename = 
                  paste0(figure.path,"Toothfish_2018_muscle_MATURITY_RANDOM_PCA_CpG_DMS_PC12.png"), 
                width = 30, height = 15, units = "cm")



# data <- data.frame(ind = rownames(PCobj$x),PCobj$x,  pop = pop)
data <- data.frame(ind = rownames(PCobj$rotation),PCobj$rotation,  pop = pop)
data <- merge(data, aggregate(cbind(mean.x = PC1, mean.y = PC3) ~
                                pop, data, mean), by = "pop")

pca.plot <- ggplot2::ggplot(data, ggplot2::aes(PC1, PC3, color = pop)) +
  ggplot2::geom_point(size = 4) + 
  ggplot2::geom_point(ggplot2::aes(x = mean.x,
                                   y = mean.y, color = pop),
                      size = 7,
                      shape = 17) +
  ggplot2::scale_colour_manual(values = colours4) +
  ggplot2::geom_hline(yintercept = 0) +
  ggplot2::geom_vline(xintercept = 0) + 
  ggplot2::labs(
    subtitle = paste0("Total variance explained: ",
                      signif(sum(var_frac[1:3]) * 100, 3), "%"),
    y = paste0("PC3; ", signif(var_frac[3] * 100, digits = 2), " %"),
    x = paste0("PC1; ", signif(var_frac[1] * 100, digits = 2), " %"),
    title = plot.title,
    caption = "") + 
  ggplot2::theme_minimal() +
  ggplot2::theme(
    axis.text = ggplot2::element_text(size = 15),
    axis.title.x = ggplot2::element_text(size = 20),
    axis.title.y = ggplot2::element_text(size = 20),
    legend.text = ggplot2::element_text(size = 15)
  )

print(pca.plot)
ggplot2::ggsave(pca.plot, bg = "white",
                filename = paste0(figure.path,"Toothfish_2018_muscle_MATURITY_RANDOM_PCA_CpG_DMS_PC13.png"), 
                width = 30, height = 15, units = "cm")




# data <- data.frame(ind = rownames(PCobj$x),PCobj$x,  pop = pop)
data <- data.frame(ind = rownames(PCobj$rotation),PCobj$rotation,  pop = pop)
data <- merge(data, aggregate(cbind(mean.x = PC2, mean.y = PC3) ~
                                pop, data, mean), by = "pop")

pca.plot <- ggplot2::ggplot(data, ggplot2::aes(PC2, PC3, color = pop)) +
  ggplot2::geom_point(size = 4) + 
  ggplot2::geom_point(ggplot2::aes(x = mean.x,
                                   y = mean.y, color = pop),
                      size = 7,
                      shape = 17) +
  ggplot2::scale_colour_manual(values = colours4) +
  ggplot2::geom_hline(yintercept = 0) +
  ggplot2::geom_vline(xintercept = 0) + 
  ggplot2::labs(
    subtitle = paste0("Total variance explained: ",
                      signif(sum(var_frac[1:3]) * 100, 3), "%"),
    y = paste0("PC3; ", signif(var_frac[3] * 100, digits = 2), " %"),
    x = paste0("PC2; ", signif(var_frac[2] * 100, digits = 2), " %"),
    title = plot.title,
    caption = "") + 
  ggplot2::theme_minimal() +
  ggplot2::theme(
    axis.text = ggplot2::element_text(size = 15),
    axis.title.x = ggplot2::element_text(size = 20),
    axis.title.y = ggplot2::element_text(size = 20),
    legend.text = ggplot2::element_text(size = 15)
  )

print(pca.plot)
ggplot2::ggsave(pca.plot, bg = "white",
                filename = 
                  paste0(figure.path,"Toothfish_2018_muscle_MATURITY_RANDOM_PCA_CpG_DMS_PC23.png"), 
                width = 30, height = 15, units = "cm")
```



# 6_Methylation: 2024 CHECK REPLICATES

## muscle tissues

```{r}
print(load(paste0(rdata.path,"Toothfish_metadata.Rdata"))) #ALL fish
```

### Reading the methylation call files

```{r, eval=FALSE}
data.list <- list.files(
  path = 
  "/datasets/work/ncmi-toa-rrbs/work/1.Toothfish/GENOME/Combined_methylation/",
  pattern = "*.cov.gz")

# subnames <- metadata$File_names[grepl(pattern = "REP", x = metadata$STRATA3)]
subnames <- metadata$File_names[grepl(pattern = "REP", x = metadata$STRATA3) & 
                                  metadata$Tissue_type == "muscle"]
data.list <- data.list[data.list %in% paste0(subnames,
                                             "_R1_trimmed_sort_bismark_bt2.bismark.cov.gz")]

data.loc <- as.list(paste0(
  "/datasets/work/ncmi-toa-rrbs/work/1.Toothfish/GENOME/Combined_methylation/",
                           data.list))


data.list2 <-  gsub(pattern = '_R1_trimmed_sort_bismark_bt2.bismark.cov.gz',
                    replacement = '',x = as.character(data.list))
# data.list2 <-  gsub(pattern = '_sort_bismark_bt2_PE.bismark.cov.gz',
#                     replacement = '',x = as.character(data.list2))

metadata.new <- metadata[metadata$File_names %in% data.list2,]
metadata.new <- metadata.new[order(match(metadata.new$File_names, data.list2)),]

metadata.new$File_names == data.list2

sample.ids <- metadata.new$Genotype_names
pop <- as.integer(factor(metadata.new$pop))


myobjDB <- methylKit::methRead(location = data.loc,
           sample.id = as.list(sample.ids),
           assembly = "TOA_genome",
           treatment = pop,
           context = "CpG",
           header = FALSE, 
           dbtype = "tabix",
           pipeline = "bismarkCoverage",
           mincov = 3
           )
save(myobjDB, file = paste0(rdata.path,"Toothfish_REP_muscle_methylkit_cpgDB.Rdata"))
```


### Data QC

#### Number of sites in genome

```{r, eval=FALSE}
cov.df <- data.frame(chr = NA, start = NA,end = NA, strand = NA, coverage = NA, 
                     numCs = NA, numTs = NA, sample = NA)
sample.ids <- sapply(myobjDB@.Data, function(x){x@sample.id})

for (i in 1:length(sample.ids)) {
  df <- methylKit::getData(myobjDB[[i]])
  df$sample <- sample.ids[i]
  cov.df <- rbind(cov.df, df)
}
cov.df <- cov.df[-1,]
save(cov.df, file = paste0(rdata.path,"Toothfish_REP_muscle_methylkit_cpg_cov.Rdata"))
```

#### Coverage stats

```{r}
load(paste0(rdata.path,"Toothfish_REP_muscle_methylkit_cpgDB.Rdata"))
print(load(paste0(rdata.path,"Toothfish_REP_muscle_methylkit_cpg_cov.Rdata")))

length(unique(paste0(cov.df$chr, "_", cov.df$start))) #2720146

# summary(as.factor(cov.df$sample))
summary(cov.df$coverage)
    # Min.  1st Qu.   Median     Mean  3rd Qu.     Max. 
    # 4.00     9.00    20.00    29.09    40.00 44225.00  

summary(cov.df$coverage[cov.df$coverage > 5 & cov.df$coverage < 50])
   # Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
   # 6.00   10.00   18.00   20.96   30.00   49.00 

cov.plot <- ggplot2::ggplot(cov.df, ggplot2::aes(x = log(coverage), fill = sample)) +
  ggplot2::geom_histogram() +
  ggplot2::theme(legend.position = "none")
# print(cov.plot)
ggplot2::ggsave(cov.plot,
                filename =
                  paste0(figure.path,"Toothfish_REP_muscle_coverage_hist_Combined.png"), 
                width = 45, height = 15, units = 'cm')

cov.plot <- ggplot2::ggplot(cov.df, ggplot2::aes(x = coverage, fill = sample)) +
  ggplot2::geom_histogram() +
  ggplot2::theme(legend.position = "none")
# print(cov.plot)
ggplot2::ggsave(cov.plot,
                filename =
                  paste0(figure.path,"Toothfish_REP_muscle_coverage_hist_Combined2.png"), 
                width = 45, height = 15, units = 'cm')

cov.plot <- ggplot2::ggplot(cov.df[cov.df$coverage > 5 & cov.df$coverage < 160,], 
                            ggplot2::aes(x = coverage)) +
  ggplot2::geom_histogram(binwidth = 1) +
  ggplot2::ggtitle("coverage < 160") +
  ggplot2::theme(legend.position = "none")
# print(cov.plot)
ggplot2::ggsave(cov.plot,
                filename =
                  paste0(figure.path,"Toothfish_REP_muscle_coverage_hist_Combined4.png"),
                width = 45, height = 15, units = 'cm')



max(cov.df$coverage, na.rm = T) #44225
ind.covplot <- ggplot2::ggplot(cov.df, ggplot2::aes(x = sample, y = coverage)) + 
  ggplot2::geom_boxplot() +
  ggplot2::theme_bw() + 
  ggplot2::theme(axis.title.x = ggplot2::element_blank(), 
                 axis.text.y = ggplot2::element_text(size = 15),
                 axis.title.y = ggplot2::element_text(size = 15),
                 axis.text.x = ggplot2::element_text(size = 7, angle = 90, hjust = 1)) +
  # ggplot2::scale_y_continuous(trans = scales::log2_trans(),
  #                 breaks = c(1, 10, 50, 750, 20000, 50000),
  #                 minor_breaks = c(1:10, 2:5*10, 1:7.5*100, 0.5:4.5*1000))
  ggplot2::scale_y_continuous(trans = scales::log2_trans(),
                  breaks = c(1, 10, 50, 150),
                  minor_breaks = c(1:10, 2:10*10))

print(ind.covplot)

ggplot2::ggsave(ind.covplot, bg = "white",
                filename = 
                  paste0(figure.path,
                         "Toothfish_REP_muscle_raw_individual_coverage_CpGs.png"),
                width = 45, height = 15, units = 'cm')
```


#### Coverage filter

```{r, eval=FALSE}
load(paste0(rdata.path,"Toothfish_REP_muscle_methylkit_cpgDB.Rdata"))

filtered.myobj <- methylKit::filterByCoverage(myobjDB, lo.count = 10, lo.perc = NULL,
                                      hi.count = NULL, hi.perc = 99)
save(myobjDB,filtered.myobj, 
     file = paste0(rdata.path,"Toothfish_REP_muscle_methylkit_cpgDB.Rdata"))
```


#### Normalise coverage

Normalize read coverage distributions between samples. Requires stringent filter on max coverage first, otherwise it will over-sample PCR bias reads. These two functions will help reduce the bias in the statistical tests that might occur due to systematic over-sampling of reads in certain samples.

```{r}
filtered.myobj.normal <- methylKit::normalizeCoverage(obj = filtered.myobj, 
                                                      method = "median", chunk.size = 1e6)
save(myobjDB,filtered.myobj, filtered.myobj.normal,
     file = paste0(rdata.path,"Toothfish_REP_muscle_methylkit_cpgDB.Rdata"))
```

#### Number of sites per sample

```{r, eval = FALSE}
sites.df <- data.frame(sites = NA,  sample = NA)
for (i in 1:length(filtered.myobj.normal)) {
  print(i)
  sites.df[i,] <- c(filtered.myobj.normal[[i]]@num.records, filtered.myobj.normal[[i]]@sample.id)
}
sites.df$sites <- as.integer(sites.df$sites)
# sites.df$pop <- filtered.myobj.normal@treatment
sites.df$pop <- metadata.new$pop
save(sites.df, file = paste0(rdata.path,"Toothfish_REP_muscle_number_cpg_sample.Rdata"))
```


```{r}
print(load(paste0(rdata.path,"Toothfish_REP_muscle_number_cpg_sample.Rdata")))
sites.plot <- ggplot2::ggplot(sites.df, ggplot2::aes(x = sample, y = sites, colour = pop)) +
  ggplot2::geom_point() +
  ggplot2::scale_color_manual(values = colours6m) +
  ggplot2::theme_bw() +
  ggplot2::theme(axis.text.x = ggplot2::element_text(angle = 90))
print(sites.plot)

ggplot2::ggsave(sites.plot,
                filename =
                  paste0(figure.path,"Toothfish_REP_muscle_number_cpg_sample.png"),
                width = 45, height = 15, units = 'cm')

sites.df$sample[sites.df$sites < 500000]
low.sites <- sites.df$sample[sites.df$sites < 500000]
# "TOA_58-4-2_B18_AE100308_muscle_F" "TOA_58-4-2_B24_AE100883_muscle_F" "TOA_58-4-2_B28_AE100885_muscle_M"
# "TOA_58-4-2_B30_AE100886_muscle_M" "TOA_58-4-2_B36_AE102050_muscle_M" "TOA_58-4-2_B38_AE102052_muscle_M"
# "TOA_58-4-2_B42_AE102105_muscle_F" "TOA_58-4-2_B48_AE102108_muscle_M" "TOA_58-4-2_B58_AE104944_muscle_F"
# "TOA_88-2_P24_AD292328_muscle_F"

## Could indicate an issue with bisulfite conversion?
```


#### Select samples

```{r}
print(load(paste0(rdata.path,"Toothfish_REP_muscle_methylkit_cpgDB.Rdata")))
print(load(paste0(rdata.path,"Toothfish_metadata.Rdata")))

metadata2 <- metadata[(grepl(pattern = "REP", x = metadata$STRATA3) & 
                                  metadata$Tissue_type == "muscle"
                                  ),]

filt.sample.id.full <- sapply(filtered.myobj@.Data, function(x){x@sample.id})
filt.sample.id <- filt.sample.id.full[filt.sample.id.full %in% metadata2$Genotype_names]
metadata2 <- metadata2[order(match(metadata2$Genotype_names, filt.sample.id)),]
identical(filt.sample.id,metadata2$Genotype_names)


pop <- metadata2$pop
pop.levels <- c("58.4.1-2018",
                "58.4.2-2018",
                "58.4.2-2024_50L",
                "58.4.2-2024_70L",
                "88.1-2018",
                "88.2-2024")

pop <- as.numeric(factor(pop, levels = pop.levels))
sample.ids <- metadata2$Genotype_names



filtered.myobj2.normal <- methylKit::reorganize(
  methylObj = filtered.myobj.normal,
  sample.ids = sample.ids,
  treatment = pop,
  chunk.size = 1e+06,
  save.db = TRUE,
  suffix = "TOA_REP_muscle_normal"
)
save(myobjDB,filtered.myobj,filtered.myobj2.normal,
     file = paste0(rdata.path,"Toothfish_REP_muscle_methylkit_cpgDB.Rdata"))
```


### Merging samples

```{r}
print(load(paste0(rdata.path,"Toothfish_REP_muscle_methylkit_cpgDB.Rdata")))
pop <- metadata2$pop
pop.levels <- c("58.4.1-2018",
                "58.4.2-2018",
                "58.4.2-2024_50L",
                "58.4.2-2024_70L",
                "88.1-2018",
                "88.2-2024")
pop <- factor(pop, levels = pop.levels)
pop <- as.numeric(pop)

table(pop)


meth.RepM.norm1L <- methylKit::unite(filtered.myobj2.normal, destrand = FALSE,
                                       save.db = FALSE,
                                       min.per.group = 1L, #missing data allowed
                                       mc.cores = 25)
#925,900 common sites for 1 sample per group

save(meth.RepM.norm1L,
     file = paste0(rdata.path,"Toothfish_REP_muscle_methylkit_object.Rdata"))
```


### Data QC

#### Remove ASM

```{r}
load(paste0(rdata.path,"Allele_specific_mutations.Rdata"))

mut <- as(asmTRUE,"GRanges")
meth.RepM.norm1L <- as(meth.RepM.norm1L,"methylBase") 

sum(as(meth.RepM.norm1L,"GRanges") %over% mut)#32849
meth.asm <- methylKit::selectByOverlap(meth.RepM.norm1L, mut) 
BOOLEAN <- paste0(meth.RepM.norm1L$chr,"_",meth.RepM.norm1L$start) %in%
  paste0(asmTRUE$Chr,"_",asmTRUE$Pos)
sum(BOOLEAN) #718
meth.RepM.norm1L <- methylKit::select(meth.RepM.norm1L, !BOOLEAN) 

save(meth.RepM.norm1L,
     file = paste0(rdata.path,"Toothfish_REP_muscle_methylkit_object.Rdata"))
```



#### Coverage 

```{r}
meth.df <- methylKit::getData(meth.RepM.norm1L)
cov <- meth.df[,meth.RepM.norm1L@coverage.index]
colnames(cov) <- meth.RepM.norm1L@sample.ids
dpf <- data.frame(cov, locus = paste0(meth.df$chr,"_",meth.df$start))
dpf <- tidyr::gather(dpf, key = individual, 
                           value = Depth, -locus,
                           factor_key = TRUE) %>%
  dplyr::group_by(individual) %>%
  dplyr::mutate(NA_cnt = sum(is.na(Depth))/length(unique(locus))) %>%
  dplyr::ungroup()
dpf.NA <- dpf[!duplicated(dpf$individual),]
dpf.NA$NA_cnt <- round(dpf.NA$NA_cnt, 3)

dpf <- dpf[ dpf$Depth > 0 & !is.na(dpf$Depth),]

max(dpf$Depth, na.rm = T) #99
myPlot <- ggplot2::ggplot(dpf, ggplot2::aes(x = individual, y = Depth)) + 
  ggplot2::geom_boxplot() +
  ggplot2::geom_text(data = dpf.NA,ggplot2::aes(x = individual, y = 150, label = NA_cnt), 
                     size = 2, angle = 90) +
  ggplot2::theme_bw() + 
  ggplot2::theme(axis.title.x = ggplot2::element_blank(), 
                 axis.text.y = ggplot2::element_text(size = 15),
                 axis.title.y = ggplot2::element_text(size = 15),
                 axis.text.x = ggplot2::element_text(size = 7, angle = 90, hjust = 1)) +
  ggplot2::scale_y_continuous(trans = scales::log2_trans(),
                  breaks = c(1, 10, 50, 100),
                  minor_breaks = c(1:10, 2:5*10))

print(myPlot)
ggplot2::ggsave(myPlot, bg = "white",
                filename = 
                  paste0(figure.path,
                  "1B.Toothfish_REP_muscle_normalised1L_methylkit_individual_coverage_CpGs.png"
                                  ), 
                width = 45, height = 15, units = "cm")

nrow(cov) # 930219 SMPs
plot(x = 1:ncol(cov), y = colMeans(cov, na.rm = TRUE), col = meth.RepM.norm1L@treatment)
plot(x = 1:ncol(cov), y = colMeans(cov > 20, na.rm = TRUE), col = meth.RepM.norm1L@treatment)
plot(x = 1:ncol(cov), y = colMeans(cov > 10, na.rm = TRUE), col = meth.RepM.norm1L@treatment)
meth.RepM.norm1L@sample.ids[colMeans(cov > 10, na.rm = TRUE) < 0.95]
# samples with low coverage:"TOA_58-4-1_289_AN177617_REP"          "TOA_58-4-2_091_AN170750_REP"          "TOA_58-4-2_B74_AE109497_muscle_M_REP"
```



### Load data

```{r}
print(load(paste0(rdata.path,"Toothfish_REP_muscle_methylkit_object.Rdata")))
print(load(paste0(rdata.path,"Toothfish_metadata.Rdata"))) 

metadata.new <- metadata[metadata$Genotype_names %in% meth.RepM.norm1L@sample.ids,]
metadata.new <- metadata.new[order(match(metadata.new$Genotype_names,
                                         meth.RepM.norm1L@sample.ids)),]

length(meth.RepM.norm1L@row.names) # number of sites
```



### Epigenetic distance

```{r}
meth.prc.df <- as.data.frame(t(methylKit::percMethylation(meth.RepM.norm1L)))
loc.names <- paste0(meth.RepM.norm1L$chr,"_",meth.RepM.norm1L$start)
colnames(meth.prc.df) <- loc.names
meth.dist <- dist(meth.prc.df, method = "euclidean")
meth.dist.df <- reshape2::melt(as.matrix(meth.dist), varnames = c("ind1", "ind2")) %>%
  dplyr::filter(value != 0)

# metadata.new$POP2 <- gsub(pattern = "_PUP|_MOM", replacement = "", x = metadata.new$STRATA3)
# metadata.new$POP2[is.na(metadata.new$POP2)] <- "Unrelated"
# metadata.new$POP[is.na(metadata.new$POP)] <- "Unrelated"
# POP.levels <- c("CA_LIT_5_MOM", "CA_LIT_5_PUP", "CA_LIT_6_MOM", "CA_LIT_6_PUP",
#                 "CA_LIT_7_MOM","CA_LIT_7_PUP", "CA_LIT_8_MOM", "CA_LIT_8_PUP", 
#                 "CA_LIT_9_MOM", "CA_LIT_9_PUP", "CA_LIT_10_MOM", "CA_LIT_10_PUP", 
#                 "Unrelated")
# metadata.new$POP <- factor(metadata.new$POP, levels = POP.levels)

metadata.new$POP2 <- metadata.new$STRATA3

pop.df1 <- metadata.new %>%
  dplyr::rename(ind1 = Genotype_names) %>%
  dplyr::rename(POP1 = POP2) %>%
  dplyr::select(ind1, POP1)
pop.df2 <- metadata.new %>%
  dplyr::rename(ind2 = Genotype_names) %>%
  dplyr::select(ind2, POP2) 

meth.dist.df <- dplyr::left_join(meth.dist.df, pop.df1, by = "ind1") %>%
  dplyr::left_join(pop.df2, by = "ind2")
POP1 <- pmin(meth.dist.df$POP1, meth.dist.df$POP2)
POP2 <- pmax(meth.dist.df$POP1, meth.dist.df$POP2)
meth.dist.df$POP1 <- POP1
meth.dist.df$POP2 <- POP2
meth.dist.df$group <- paste0(meth.dist.df$POP1,"_",meth.dist.df$POP2)
# group.levels <- c("CA_LIT_5_CA_LIT_5", "CA_LIT_5_CA_LIT_6", "CA_LIT_5_CA_LIT_7", 
#                   "CA_LIT_5_CA_LIT_8", "CA_LIT_5_CA_LIT_9", "CA_LIT_10_CA_LIT_5","CA_LIT_5_Unrelated",
#                   "CA_LIT_6_CA_LIT_6", "CA_LIT_6_CA_LIT_7", "CA_LIT_6_CA_LIT_8", "CA_LIT_6_CA_LIT_9",
#                   "CA_LIT_10_CA_LIT_6","CA_LIT_6_Unrelated", "CA_LIT_7_CA_LIT_7","CA_LIT_7_CA_LIT_8",
#                   "CA_LIT_7_CA_LIT_9","CA_LIT_10_CA_LIT_7", "CA_LIT_7_Unrelated","CA_LIT_8_CA_LIT_8",
#                   "CA_LIT_8_CA_LIT_9", "CA_LIT_10_CA_LIT_8", "CA_LIT_8_Unrelated","CA_LIT_9_CA_LIT_9",
#                   "CA_LIT_10_CA_LIT_9","CA_LIT_9_Unrelated","CA_LIT_10_CA_LIT_10","CA_LIT_10_Unrelated",
#                   "Unrelated_Unrelated")
# meth.dist.df$group <- factor(meth.dist.df$group, group.levels)
# meth.dist.df <- meth.dist.df[order(meth.dist.df$group),]
dist.plot <- ggplot2::ggplot(meth.dist.df, ggplot2::aes(x = group, y = value, 
                                                fill = group)) + 
  ggplot2::geom_boxplot() +
  ggplot2::geom_point(alpha = 0.5) +
  # ggplot2::facet_wrap(condition~.) +
  ggplot2::labs(x = "Replicates", y = "Epigentic distance (euclidean)") +
  ggplot2::theme_bw() +
  ggplot2::theme(legend.position = "none",
                   # axis.title.y = ggplot2::element_blank(),
                   # axis.title.x = ggplot2::element_blank(),
                 # panel.background = element_rect(fill = "white", colour = NA),
                   axis.text.x = ggplot2::element_text(face = "bold", color = "black", 
                                        size = 10, angle = 90, hjust = 1, vjust = 0.5))
print(dist.plot)
ggplot2::ggsave(dist.plot, bg = "white",
                filename = paste0(figure.path,
                                  "Toothfish_REP_muscle_normalised1L_meth_Distance.png"
                                  ), 
                width = 45, height = 30, units = "cm")



# ind.levels <- metadata.new$individual[order(metadata.new$POP)]
# meth.dist.df$ind1 <- factor(meth.dist.df$ind1, levels = ind.levels)
# meth.dist.df$ind2 <- factor(meth.dist.df$ind2, levels = ind.levels)


heat.plot <- ggplot2::ggplot(data = meth.dist.df, ggplot2::aes(x = ind1, y = ind2, fill = value)) + 
    ggplot2::geom_tile() +
    ggplot2::scale_fill_gradient2(low = "blue", mid = "yellow", high = "red", 
                                  midpoint = (max(meth.dist.df$value, na.rm = TRUE) + 
                                                min(meth.dist.df$value, na.rm = TRUE))/2, 
                                  guide = "colourbar") +
    # ggnewscale::new_scale_fill() +
    # ggplot2::geom_tile(data = heat.plot.data, ggplot2::aes(x = individual, y = "pop", fill = pop,
    #                                 height = 2)) +
    # ggplot2::scale_fill_manual(values = colours.time) +
    # ggplot2::scale_y_discrete(drop = FALSE) +
    ggplot2::theme(legend.position = "right",
                   axis.title.y = ggplot2::element_blank(),
                   axis.title.x = ggplot2::element_blank(),
                   axis.text.x = ggplot2::element_text(face = "bold", color = "black", 
                                        size = 10, angle = 90, hjust = 1, vjust = 0.5))
print(heat.plot)
ggplot2::ggsave(heat.plot, bg = "white",
                filename = paste0(figure.path,
                                  "Toothfish_REP_muscle_normalised1L_meth_Distance_HEAT.png"
                                  ), 
                width = 45, height = 30, units = "cm")
```


### Global PCA

```{r}
PCobj <- methylKit::PCASamples(meth.RepM.norm1L, comp = c(1,2), 
                               adj.lim = c(0,0), obj.return = TRUE)

var_frac <- PCobj$sdev/sum(PCobj$sdev)
print(paste0("Total variance explained: ",signif(sum(var_frac[1:3]) * 100, 3), "%")) 
barplot(var_frac[1:20], main = "variance explained", 
        col = heat.colors(length(var_frac[1:20])))

identical(metadata.new$Genotype_names, meth.RepM.norm1L@sample.ids)
identical(metadata.new$Genotype_names, rownames(PCobj$x))

pop <- metadata.new$pop

plot.title <- "Toothfish methylation - REP muscle (930,219 CpG sites)"

data <- data.frame(ind = rownames(PCobj$x),PCobj$x,  pop = pop)
data <- merge(data, aggregate(cbind(mean.x = PC1, mean.y = PC2) ~
                                pop, data, mean), by = "pop")

pca.plot <- ggplot2::ggplot(data, ggplot2::aes(PC1, PC2, color = pop)) +
  ggplot2::geom_point(size = 4) + 
  ggplot2::geom_point(ggplot2::aes(x = mean.x,
                                   y = mean.y, color = pop),
                      size = 7,
                      shape = 17) +
  ggplot2::scale_colour_manual(values = colours6m) +
  ggplot2::geom_hline(yintercept = 0) +
  ggplot2::geom_vline(xintercept = 0) + 
  ggplot2::labs(
    subtitle = paste0("Total variance explained: ",
                      signif(sum(var_frac[1:4]) * 100, 3), "%"),
    y = paste0("PC2; ", signif(var_frac[2] * 100, digits = 2), " %"),
    x = paste0("PC1; ", signif(var_frac[1] * 100, digits = 2), " %"),
    title = plot.title,
    caption = "") + 
  ggplot2::theme_minimal() +
  ggplot2::theme(
    axis.text = ggplot2::element_text(size = 15),
    axis.title.x = ggplot2::element_text(size = 20),
    axis.title.y = ggplot2::element_text(size = 20),
    legend.text = ggplot2::element_text(size = 15)
  )

print(pca.plot)
ggplot2::ggsave(pca.plot, bg = "white",
                filename = paste0(figure.path,"Toothfish_REP_muscle_normalised1L_PCA_CpG_PC12.png"), 
                width = 30, height = 15, units = "cm")


data <- data.frame(ind = rownames(PCobj$x),PCobj$x,  pop = pop)
data <- merge(data, aggregate(cbind(mean.x = PC2, mean.y = PC3) ~
                                pop, data, mean), by = "pop")
pca.plot <- ggplot2::ggplot(data, ggplot2::aes(PC2, PC3, color = pop)) +
  ggplot2::geom_point(size = 4) + 
  ggplot2::geom_point(ggplot2::aes(x = mean.x,
                                   y = mean.y, color = pop),
                      size = 7,
                      shape = 17) +
  ggplot2::scale_colour_manual(values = colours6m) +
  ggplot2::geom_hline(yintercept = 0) +
  ggplot2::geom_vline(xintercept = 0) + 
  ggplot2::labs(
    subtitle = paste0("Total variance explained: ",
                      signif(sum(var_frac[1:4]) * 100, 3), "%"),
    y = paste0("PC3; ", signif(var_frac[3] * 100, digits = 2), " %"),
    x = paste0("PC2; ", signif(var_frac[2] * 100, digits = 2), " %"),
    title = plot.title,
    caption = "") + 
  ggplot2::theme_minimal() +
  ggplot2::theme(
    axis.text = ggplot2::element_text(size = 15),
    axis.title.x = ggplot2::element_text(size = 20),
    axis.title.y = ggplot2::element_text(size = 20),
    legend.text = ggplot2::element_text(size = 15)
  )

print(pca.plot)
ggplot2::ggsave(pca.plot, bg = "white",
                filename = paste0(figure.path,"Toothfish_REP_muscle_normalised1L_PCA_CpG_PC23.png"), 
                width = 30, height = 15, units = "cm")
```




## Fin tissues

```{r}
print(load(paste0(rdata.path,"Toothfish_metadata.Rdata"))) #ALL fish
```

### Reading the methylation call files

```{r, eval=FALSE}
data.list <- list.files(
  path = 
  "/datasets/work/ncmi-toa-rrbs/work/1.Toothfish/GENOME/Combined_methylation/",
  pattern = "*.cov.gz")

# subnames <- metadata$File_names[grepl(pattern = "REP", x = metadata$STRATA3)]
subnames <- metadata$File_names[grepl(pattern = "REP", x = metadata$STRATA3) & 
                                  metadata$Tissue_type == "fin"]
data.list <- data.list[data.list %in% paste0(subnames,
                                             "_R1_trimmed_sort_bismark_bt2.bismark.cov.gz")]

data.loc <- as.list(paste0(
  "/datasets/work/ncmi-toa-rrbs/work/1.Toothfish/GENOME/Combined_methylation/",
                           data.list))


data.list2 <-  gsub(pattern = '_R1_trimmed_sort_bismark_bt2.bismark.cov.gz',
                    replacement = '',x = as.character(data.list))
# data.list2 <-  gsub(pattern = '_sort_bismark_bt2_PE.bismark.cov.gz',
#                     replacement = '',x = as.character(data.list2))

metadata.new <- metadata[metadata$File_names %in% data.list2,]
metadata.new <- metadata.new[order(match(metadata.new$File_names, data.list2)),]

metadata.new$File_names == data.list2

sample.ids <- metadata.new$Genotype_names
pop <- as.integer(factor(metadata.new$pop))


myobjDB <- methylKit::methRead(location = data.loc,
           sample.id = as.list(sample.ids),
           assembly = "TOA_genome",
           treatment = pop,
           context = "CpG",
           header = FALSE, 
           dbtype = "tabix",
           pipeline = "bismarkCoverage",
           mincov = 3
           )
save(myobjDB, file = paste0(rdata.path,"Toothfish_REP_fin_methylkit_cpgDB.Rdata"))
```


### Data QC

#### Number of sites in genome

```{r, eval=FALSE}
cov.df <- data.frame(chr = NA, start = NA,end = NA, strand = NA, coverage = NA, 
                     numCs = NA, numTs = NA, sample = NA)
sample.ids <- sapply(myobjDB@.Data, function(x){x@sample.id})

for (i in 1:length(sample.ids)) {
  df <- methylKit::getData(myobjDB[[i]])
  df$sample <- sample.ids[i]
  cov.df <- rbind(cov.df, df)
}
cov.df <- cov.df[-1,]
save(cov.df, file = paste0(rdata.path,"Toothfish_REP_fin_methylkit_cpg_cov.Rdata"))
```

#### Coverage stats

```{r}
load(paste0(rdata.path,"Toothfish_REP_fin_methylkit_cpgDB.Rdata"))
print(load(paste0(rdata.path,"Toothfish_REP_fin_methylkit_cpg_cov.Rdata")))

length(unique(paste0(cov.df$chr, "_", cov.df$start))) #2,486,314

# summary(as.factor(cov.df$sample))
summary(cov.df$coverage)
    # Min.  1st Qu.   Median     Mean  3rd Qu.     Max. 
    # 4.00     8.00    17.00    24.76    32.00 44225.00 

summary(cov.df$coverage[cov.df$coverage > 5 & cov.df$coverage < 50])
   # Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
   # 6.00   10.00   17.00   20.06   28.00   49.00 

cov.plot <- ggplot2::ggplot(cov.df, ggplot2::aes(x = log(coverage), fill = sample)) +
  ggplot2::geom_histogram() +
  ggplot2::theme(legend.position = "none")
# print(cov.plot)
ggplot2::ggsave(cov.plot,
                filename =
                  paste0(figure.path,"Toothfish_REP_fin_coverage_hist_Combined.png"), 
                width = 45, height = 15, units = 'cm')

cov.plot <- ggplot2::ggplot(cov.df, ggplot2::aes(x = coverage, fill = sample)) +
  ggplot2::geom_histogram() +
  ggplot2::theme(legend.position = "none")
# print(cov.plot)
ggplot2::ggsave(cov.plot,
                filename =
                  paste0(figure.path,"Toothfish_REP_fin_coverage_hist_Combined2.png"), 
                width = 45, height = 15, units = 'cm')

cov.plot <- ggplot2::ggplot(cov.df[cov.df$coverage > 5 & cov.df$coverage < 160,], 
                            ggplot2::aes(x = coverage)) +
  ggplot2::geom_histogram(binwidth = 1) +
  ggplot2::ggtitle("coverage < 160") +
  ggplot2::theme(legend.position = "none")
# print(cov.plot)
ggplot2::ggsave(cov.plot,
                filename =
                  paste0(figure.path,"Toothfish_REP_fin_coverage_hist_Combined4.png"),
                width = 45, height = 15, units = 'cm')



max(cov.df$coverage, na.rm = T) #44225
ind.covplot <- ggplot2::ggplot(cov.df, ggplot2::aes(x = sample, y = coverage)) + 
  ggplot2::geom_boxplot() +
  ggplot2::theme_bw() + 
  ggplot2::theme(axis.title.x = ggplot2::element_blank(), 
                 axis.text.y = ggplot2::element_text(size = 15),
                 axis.title.y = ggplot2::element_text(size = 15),
                 axis.text.x = ggplot2::element_text(size = 7, angle = 90, hjust = 1)) +
  # ggplot2::scale_y_continuous(trans = scales::log2_trans(),
  #                 breaks = c(1, 10, 50, 750, 20000, 50000),
  #                 minor_breaks = c(1:10, 2:5*10, 1:7.5*100, 0.5:4.5*1000))
  ggplot2::scale_y_continuous(trans = scales::log2_trans(),
                  breaks = c(1, 10, 50, 150),
                  minor_breaks = c(1:10, 2:10*10))

print(ind.covplot)

ggplot2::ggsave(ind.covplot, bg = "white",
                filename = 
                  paste0(figure.path,
                         "Toothfish_REP_fin_raw_individual_coverage_CpGs.png"),
                width = 45, height = 15, units = 'cm')
```


#### Coverage filter

High count filter based on coverage stats.

```{r, eval=FALSE}
load(paste0(rdata.path,"Toothfish_REP_fin_methylkit_cpgDB.Rdata"))

filtered.myobj <- methylKit::filterByCoverage(myobjDB, lo.count = 10, lo.perc = NULL,
                                      hi.count = NULL, hi.perc = 99)
save(myobjDB,filtered.myobj, 
     file = paste0(rdata.path,"Toothfish_REP_fin_methylkit_cpgDB.Rdata"))
```


#### Normalise coverage

Normalize read coverage distributions between samples. Requires stringent filter on max coverage first, otherwise it will over-sample PCR bias reads. These two functions will help reduce the bias in the statistical tests that might occur due to systematic over-sampling of reads in certain samples.

```{r}
filtered.myobj.normal <- methylKit::normalizeCoverage(obj = filtered.myobj, 
                                                      method = "median", chunk.size = 1e6)
save(myobjDB,filtered.myobj, filtered.myobj.normal,
     file = paste0(rdata.path,"Toothfish_REP_fin_methylkit_cpgDB.Rdata"))
```

#### Number of sites per sample

```{r, eval = FALSE}
sites.df <- data.frame(sites = NA,  sample = NA)
for (i in 1:length(filtered.myobj.normal)) {
  print(i)
  sites.df[i,] <- c(filtered.myobj.normal[[i]]@num.records, filtered.myobj.normal[[i]]@sample.id)
}
sites.df$sites <- as.integer(sites.df$sites)
# sites.df$pop <- filtered.myobj.normal@treatment
sites.df$pop <- metadata.new$pop
save(sites.df, file = paste0(rdata.path,"Toothfish_REP_fin_number_cpg_sample.Rdata"))
```


```{r}
print(load(paste0(rdata.path,"Toothfish_REP_fin_number_cpg_sample.Rdata")))
sites.plot <- ggplot2::ggplot(sites.df, ggplot2::aes(x = sample, y = sites, colour = pop)) +
  ggplot2::geom_point() +
  ggplot2::scale_color_manual(values = colours3f) +
  ggplot2::theme_bw() +
  ggplot2::theme(axis.text.x = ggplot2::element_text(angle = 90))
print(sites.plot)

ggplot2::ggsave(sites.plot,
                filename =
                  paste0(figure.path,"Toothfish_REP_fin_number_cpg_sample.png"),
                width = 45, height = 15, units = 'cm')

sites.df$sample[sites.df$sites < 500000]
low.sites <- sites.df$sample[sites.df$sites < 500000]
# "TOA_58-4-2_B18_AE100308_muscle_F" "TOA_58-4-2_B24_AE100883_muscle_F" "TOA_58-4-2_B28_AE100885_muscle_M"
# "TOA_58-4-2_B30_AE100886_muscle_M" "TOA_58-4-2_B36_AE102050_muscle_M" "TOA_58-4-2_B38_AE102052_muscle_M"
# "TOA_58-4-2_B42_AE102105_muscle_F" "TOA_58-4-2_B48_AE102108_muscle_M" "TOA_58-4-2_B58_AE104944_muscle_F"
# "TOA_88-2_P24_AD292328_muscle_F"

## Could indicate an issue with bisulfite conversion?
```


#### Select samples

```{r}
print(load(paste0(rdata.path,"Toothfish_REP_fin_methylkit_cpgDB.Rdata")))
print(load(paste0(rdata.path,"Toothfish_metadata.Rdata")))

metadata2 <- metadata[(grepl(pattern = "REP", x = metadata$STRATA3) & 
                                  metadata$Tissue_type == "fin"
                                  ),]

filt.sample.id.full <- sapply(filtered.myobj@.Data, function(x){x@sample.id})
filt.sample.id <- filt.sample.id.full[filt.sample.id.full %in% metadata2$Genotype_names]
metadata2 <- metadata2[order(match(metadata2$Genotype_names, filt.sample.id)),]
identical(filt.sample.id,metadata2$Genotype_names)


pop <- metadata2$pop
pop.levels <- c("48.2-2017",
                "58.4.2-2024_50L",
                "58.4.2-2024_70L",
                "88.2-2024")

pop <- as.numeric(factor(pop, levels = pop.levels))
sample.ids <- metadata2$Genotype_names



filtered.myobj2.normal <- methylKit::reorganize(
  methylObj = filtered.myobj.normal,
  sample.ids = sample.ids,
  treatment = pop,
  chunk.size = 1e+06,
  save.db = TRUE,
  suffix = "TOA_REP_fin_normal"
)
save(myobjDB,filtered.myobj,filtered.myobj2.normal,
     file = paste0(rdata.path,"Toothfish_REP_fin_methylkit_cpgDB.Rdata"))
```


### Merging samples

```{r}
print(load(paste0(rdata.path,"Toothfish_REP_fin_methylkit_cpgDB.Rdata")))
pop <- metadata2$pop
pop.levels <- c("48.2-2017",
                "58.4.2-2024_50L",
                "58.4.2-2024_70L",
                "88.2-2024")
pop <- factor(pop, levels = pop.levels)
pop <- as.numeric(pop)

table(pop)


meth.RepF.norm1L <- methylKit::unite(filtered.myobj2.normal, destrand = FALSE,
                                       save.db = FALSE,
                                       min.per.group = 1L, #missing data allowed
                                       mc.cores = 25)
#925,900 common sites for 1 sample per group

save(meth.RepF.norm1L,
     file = paste0(rdata.path,"Toothfish_REP_fin_methylkit_object.Rdata"))
```


### Data QC

#### Remove ASM

```{r}
load(paste0(rdata.path,"Allele_specific_mutations.Rdata"))

mut <- as(asmTRUE,"GRanges")
meth.RepF.norm1L <- as(meth.RepF.norm1L,"methylBase") 

sum(as(meth.RepF.norm1L,"GRanges") %over% mut)#33359
meth.asm <- methylKit::selectByOverlap(meth.RepF.norm1L, mut) 
BOOLEAN <- paste0(meth.RepF.norm1L$chr,"_",meth.RepF.norm1L$start) %in%
  paste0(asmTRUE$Chr,"_",asmTRUE$Pos)
sum(BOOLEAN) #742
meth.RepF.norm1L <- methylKit::select(meth.RepF.norm1L, !BOOLEAN) 

save(meth.RepF.norm1L,
     file = paste0(rdata.path,"Toothfish_REP_fin_methylkit_object.Rdata"))
```



#### Coverage 

```{r}
meth.df <- methylKit::getData(meth.RepF.norm1L)
cov <- meth.df[,meth.RepF.norm1L@coverage.index]
colnames(cov) <- meth.RepF.norm1L@sample.ids
dpf <- data.frame(cov, locus = paste0(meth.df$chr,"_",meth.df$start))
dpf <- tidyr::gather(dpf, key = individual, 
                           value = Depth, -locus,
                           factor_key = TRUE) %>%
  dplyr::group_by(individual) %>%
  dplyr::mutate(NA_cnt = sum(is.na(Depth))/length(unique(locus))) %>%
  dplyr::ungroup()
dpf.NA <- dpf[!duplicated(dpf$individual),]
dpf.NA$NA_cnt <- round(dpf.NA$NA_cnt, 3)

dpf <- dpf[ dpf$Depth > 0 & !is.na(dpf$Depth),]

max(dpf$Depth, na.rm = T) #99
myPlot <- ggplot2::ggplot(dpf, ggplot2::aes(x = individual, y = Depth)) + 
  ggplot2::geom_boxplot() +
  ggplot2::geom_text(data = dpf.NA,ggplot2::aes(x = individual, y = 150, label = NA_cnt), 
                     size = 2, angle = 90) +
  ggplot2::theme_bw() + 
  ggplot2::theme(axis.title.x = ggplot2::element_blank(), 
                 axis.text.y = ggplot2::element_text(size = 15),
                 axis.title.y = ggplot2::element_text(size = 15),
                 axis.text.x = ggplot2::element_text(size = 7, angle = 90, hjust = 1)) +
  ggplot2::scale_y_continuous(trans = scales::log2_trans(),
                  breaks = c(1, 10, 50, 100),
                  minor_breaks = c(1:10, 2:5*10))

print(myPlot)
ggplot2::ggsave(myPlot, bg = "white",
                filename = 
                  paste0(figure.path,
                  "1B.Toothfish_REP_fin_normalised1L_methylkit_individual_coverage_CpGs.png"
                                  ), 
                width = 45, height = 15, units = "cm")

nrow(cov) # 925158 SMPs
plot(x = 1:ncol(cov), y = colMeans(cov, na.rm = TRUE), col = meth.RepF.norm1L@treatment)
plot(x = 1:ncol(cov), y = colMeans(cov > 20, na.rm = TRUE), col = meth.RepF.norm1L@treatment)
plot(x = 1:ncol(cov), y = colMeans(cov > 10, na.rm = TRUE), col = meth.RepF.norm1L@treatment)
meth.RepF.norm1L@sample.ids[colMeans(cov > 10, na.rm = TRUE) < 0.95]
# samples with low coverage: "TOA_58-4-2_B24_AE100883_muscle_F", "TOA_88-2_P24_AD292328_muscle_F"
```



### Load data

```{r}
print(load(paste0(rdata.path,"Toothfish_REP_fin_methylkit_object.Rdata")))
print(load(paste0(rdata.path,"Toothfish_metadata.Rdata"))) 

metadata.new <- metadata[metadata$Genotype_names %in% meth.RepF.norm1L@sample.ids,]
metadata.new <- metadata.new[order(match(metadata.new$Genotype_names,
                                         meth.RepF.norm1L@sample.ids)),]

length(meth.RepF.norm1L@row.names) # number of sites
```



### Epigenetic distance

```{r}
meth.prc.df <- as.data.frame(t(methylKit::percMethylation(meth.RepF.norm1L)))
loc.names <- paste0(meth.RepF.norm1L$chr,"_",meth.RepF.norm1L$start)
colnames(meth.prc.df) <- loc.names
meth.dist <- dist(meth.prc.df, method = "euclidean")
meth.dist.df <- reshape2::melt(as.matrix(meth.dist), varnames = c("ind1", "ind2")) %>%
  dplyr::filter(value != 0)

# metadata.new$POP2 <- gsub(pattern = "_PUP|_MOM", replacement = "", x = metadata.new$STRATA3)
# metadata.new$POP2[is.na(metadata.new$POP2)] <- "Unrelated"
# metadata.new$POP[is.na(metadata.new$POP)] <- "Unrelated"
# POP.levels <- c("CA_LIT_5_MOM", "CA_LIT_5_PUP", "CA_LIT_6_MOM", "CA_LIT_6_PUP",
#                 "CA_LIT_7_MOM","CA_LIT_7_PUP", "CA_LIT_8_MOM", "CA_LIT_8_PUP", 
#                 "CA_LIT_9_MOM", "CA_LIT_9_PUP", "CA_LIT_10_MOM", "CA_LIT_10_PUP", 
#                 "Unrelated")
# metadata.new$POP <- factor(metadata.new$POP, levels = POP.levels)

metadata.new$POP2 <- metadata.new$STRATA3

pop.df1 <- metadata.new %>%
  dplyr::rename(ind1 = Genotype_names) %>%
  dplyr::rename(POP1 = POP2) %>%
  dplyr::select(ind1, POP1)
pop.df2 <- metadata.new %>%
  dplyr::rename(ind2 = Genotype_names) %>%
  dplyr::select(ind2, POP2) 

meth.dist.df <- dplyr::left_join(meth.dist.df, pop.df1, by = "ind1") %>%
  dplyr::left_join(pop.df2, by = "ind2")
POP1 <- pmin(meth.dist.df$POP1, meth.dist.df$POP2)
POP2 <- pmax(meth.dist.df$POP1, meth.dist.df$POP2)
meth.dist.df$POP1 <- POP1
meth.dist.df$POP2 <- POP2
meth.dist.df$group <- paste0(meth.dist.df$POP1,"_",meth.dist.df$POP2)
# group.levels <- c("CA_LIT_5_CA_LIT_5", "CA_LIT_5_CA_LIT_6", "CA_LIT_5_CA_LIT_7", 
#                   "CA_LIT_5_CA_LIT_8", "CA_LIT_5_CA_LIT_9", "CA_LIT_10_CA_LIT_5","CA_LIT_5_Unrelated",
#                   "CA_LIT_6_CA_LIT_6", "CA_LIT_6_CA_LIT_7", "CA_LIT_6_CA_LIT_8", "CA_LIT_6_CA_LIT_9",
#                   "CA_LIT_10_CA_LIT_6","CA_LIT_6_Unrelated", "CA_LIT_7_CA_LIT_7","CA_LIT_7_CA_LIT_8",
#                   "CA_LIT_7_CA_LIT_9","CA_LIT_10_CA_LIT_7", "CA_LIT_7_Unrelated","CA_LIT_8_CA_LIT_8",
#                   "CA_LIT_8_CA_LIT_9", "CA_LIT_10_CA_LIT_8", "CA_LIT_8_Unrelated","CA_LIT_9_CA_LIT_9",
#                   "CA_LIT_10_CA_LIT_9","CA_LIT_9_Unrelated","CA_LIT_10_CA_LIT_10","CA_LIT_10_Unrelated",
#                   "Unrelated_Unrelated")
# meth.dist.df$group <- factor(meth.dist.df$group, group.levels)
# meth.dist.df <- meth.dist.df[order(meth.dist.df$group),]
dist.plot <- ggplot2::ggplot(meth.dist.df, ggplot2::aes(x = group, y = value, 
                                                fill = group)) + 
  ggplot2::geom_boxplot() +
  ggplot2::geom_point(alpha = 0.5) +
  # ggplot2::facet_wrap(condition~.) +
  ggplot2::labs(x = "Replicates", y = "Epigentic distance (euclidean)") +
  ggplot2::theme_bw() +
  ggplot2::theme(legend.position = "none",
                   # axis.title.y = ggplot2::element_blank(),
                   # axis.title.x = ggplot2::element_blank(),
                 # panel.background = element_rect(fill = "white", colour = NA),
                   axis.text.x = ggplot2::element_text(face = "bold", color = "black", 
                                        size = 10, angle = 90, hjust = 1, vjust = 0.5))
print(dist.plot)
ggplot2::ggsave(dist.plot, bg = "white",
                filename = paste0(figure.path,
                                  "Toothfish_REP_fin_normalised1L_meth_Distance.png"
                                  ), 
                width = 45, height = 30, units = "cm")



# ind.levels <- metadata.new$individual[order(metadata.new$POP)]
# meth.dist.df$ind1 <- factor(meth.dist.df$ind1, levels = ind.levels)
# meth.dist.df$ind2 <- factor(meth.dist.df$ind2, levels = ind.levels)


heat.plot <- ggplot2::ggplot(data = meth.dist.df, ggplot2::aes(x = ind1, y = ind2, fill = value)) + 
    ggplot2::geom_tile() +
    ggplot2::scale_fill_gradient2(low = "blue", mid = "yellow", high = "red", 
                                  midpoint = (max(meth.dist.df$value, na.rm = TRUE) + 
                                                min(meth.dist.df$value, na.rm = TRUE))/2, 
                                  guide = "colourbar") +
    # ggnewscale::new_scale_fill() +
    # ggplot2::geom_tile(data = heat.plot.data, ggplot2::aes(x = individual, y = "pop", fill = pop,
    #                                 height = 2)) +
    # ggplot2::scale_fill_manual(values = colours.time) +
    # ggplot2::scale_y_discrete(drop = FALSE) +
    ggplot2::theme(legend.position = "right",
                   axis.title.y = ggplot2::element_blank(),
                   axis.title.x = ggplot2::element_blank(),
                   axis.text.x = ggplot2::element_text(face = "bold", color = "black", 
                                        size = 10, angle = 90, hjust = 1, vjust = 0.5))
print(heat.plot)
ggplot2::ggsave(heat.plot, bg = "white",
                filename = paste0(figure.path,
                                  "Toothfish_REP_fin_normalised1L_meth_Distance_HEAT.png"
                                  ), 
                width = 45, height = 30, units = "cm")
```


### Global PCA

```{r}
PCobj <- methylKit::PCASamples(meth.RepF.norm1L, comp = c(1,2), 
                               adj.lim = c(0,0), obj.return = TRUE)

var_frac <- PCobj$sdev/sum(PCobj$sdev)
print(paste0("Total variance explained: ",signif(sum(var_frac[1:3]) * 100, 3), "%")) 
barplot(var_frac[1:20], main = "variance explained", 
        col = heat.colors(length(var_frac[1:20])))

identical(metadata.new$Genotype_names, meth.RepF.norm1L@sample.ids)
identical(metadata.new$Genotype_names, rownames(PCobj$x))

pop <- metadata.new$pop

plot.title <- "Toothfish methylation - REP fin (925,158 CpG sites)"

data <- data.frame(ind = rownames(PCobj$x),PCobj$x,  pop = pop)
# data <- data.frame(ind = rownames(PCobj$rotation),PCobj$rotation, pop = pop)
data <- merge(data, aggregate(cbind(mean.x = PC1, mean.y = PC2) ~
                                pop, data, mean), by = "pop")

pca.plot <- ggplot2::ggplot(data, ggplot2::aes(PC1, PC2, color = pop)) +
  ggplot2::geom_point(size = 4) + 
  ggplot2::geom_point(ggplot2::aes(x = mean.x,
                                   y = mean.y, color = pop),
                      size = 7,
                      shape = 17) +
  ggplot2::scale_colour_manual(values = colours3f) +
  ggplot2::geom_hline(yintercept = 0) +
  ggplot2::geom_vline(xintercept = 0) + 
  ggplot2::labs(
    subtitle = paste0("Total variance explained: ",
                      signif(sum(var_frac[1:4]) * 100, 3), "%"),
    y = paste0("PC2; ", signif(var_frac[2] * 100, digits = 2), " %"),
    x = paste0("PC1; ", signif(var_frac[1] * 100, digits = 2), " %"),
    title = plot.title,
    caption = "") + 
  ggplot2::theme_minimal() +
  ggplot2::theme(
    axis.text = ggplot2::element_text(size = 15),
    axis.title.x = ggplot2::element_text(size = 20),
    axis.title.y = ggplot2::element_text(size = 20),
    legend.text = ggplot2::element_text(size = 15)
  )

print(pca.plot)
ggplot2::ggsave(pca.plot, bg = "white",
                filename = paste0(figure.path,"Toothfish_REP_fin_normalised1L_PCA_CpG_PC12.png"), 
                width = 30, height = 15, units = "cm")


data <- data.frame(ind = rownames(PCobj$x),PCobj$x,  pop = pop)
# data <- data.frame(ind = rownames(PCobj$rotation),PCobj$rotation, pop = pop)
data <- merge(data, aggregate(cbind(mean.x = PC2, mean.y = PC3) ~
                                pop, data, mean), by = "pop")
pca.plot <- ggplot2::ggplot(data, ggplot2::aes(PC2, PC3, color = pop)) +
  ggplot2::geom_point(size = 4) + 
  ggplot2::geom_point(ggplot2::aes(x = mean.x,
                                   y = mean.y, color = pop),
                      size = 7,
                      shape = 17) +
  ggplot2::scale_colour_manual(values = colours3f) +
  ggplot2::geom_hline(yintercept = 0) +
  ggplot2::geom_vline(xintercept = 0) + 
  ggplot2::labs(
    subtitle = paste0("Total variance explained: ",
                      signif(sum(var_frac[1:4]) * 100, 3), "%"),
    y = paste0("PC3; ", signif(var_frac[3] * 100, digits = 2), " %"),
    x = paste0("PC2; ", signif(var_frac[2] * 100, digits = 2), " %"),
    title = plot.title,
    caption = "") + 
  ggplot2::theme_minimal() +
  ggplot2::theme(
    axis.text = ggplot2::element_text(size = 15),
    axis.title.x = ggplot2::element_text(size = 20),
    axis.title.y = ggplot2::element_text(size = 20),
    legend.text = ggplot2::element_text(size = 15)
  )

print(pca.plot)
ggplot2::ggsave(pca.plot, bg = "white",
                filename = paste0(figure.path,"Toothfish_REP_fin_normalised1L_PCA_CpG_PC23.png"), 
                width = 30, height = 15, units = "cm")

data$ind[data$PC2 < -100]
# "TOA_88-2_P69_AD294614_fin_F"     "TOA_88-2_P69_AD294614_fin_F_REP"
# DUE TO LOW COVEVRAGE
```



# 7a_Methylation: 2024 - muscle tissue

```{r}
print(load(paste0(rdata.path,"Toothfish_metadata.Rdata"))) #ALL fish
```

## Reading the methylation call files

```{r, eval=FALSE}
data.list <- list.files(
  path = 
  "/datasets/work/ncmi-toa-rrbs/work/1.Toothfish/GENOME/Combined_methylation/",
  pattern = "*.cov.gz")

subnames <- metadata$File_names[!(metadata$Year %in% c(2017,2018) | metadata$Tissue_type == "fin" |
                                    grepl(pattern = "REP", x = metadata$File_names) |
                                    !grepl(pattern = "TOA", x = metadata$File_names) |
                                    metadata$pop == "58.4.2-2024_50L"
                                  )]
data.list <- data.list[data.list %in% paste0(subnames,
                                             "_R1_trimmed_sort_bismark_bt2.bismark.cov.gz")]

data.loc <- as.list(paste0(
  "/datasets/work/ncmi-toa-rrbs/work/1.Toothfish/GENOME/Combined_methylation/",
                           data.list))



data.list2 <-  gsub(pattern = '_R1_trimmed_sort_bismark_bt2.bismark.cov.gz',
                    replacement = '',x = as.character(data.list))
# data.list2 <-  gsub(pattern = '_sort_bismark_bt2_PE.bismark.cov.gz',
#                     replacement = '',x = as.character(data.list2))

metadata.new <- metadata[metadata$File_names %in% data.list2,]
metadata.new <- metadata.new[order(match(metadata.new$File_names, data.list2)),]

metadata.new$File_names == data.list2

sample.ids <- metadata.new$Genotype_names
pop <- as.integer(factor(metadata.new$pop))


myobjDB <- methylKit::methRead(location = data.loc,
           sample.id = as.list(sample.ids),
           assembly = "TOA_genome",
           treatment = pop,
           context = "CpG",
           header = FALSE, 
           dbtype = "tabix",
           pipeline = "bismarkCoverage",
           mincov = 3
           )
save(myobjDB, file = paste0(rdata.path,"Toothfish_2024_muscle_methylkit_cpgDB.Rdata"))
```


## Data QC

#### Number of sites in genome

```{r, eval=FALSE}

cov.df <- data.frame(chr = NA, start = NA,end = NA, strand = NA, coverage = NA, 
                     numCs = NA, numTs = NA, sample = NA)
sample.ids <- sapply(myobjDB@.Data, function(x){x@sample.id})

for (i in 1:length(sample.ids)) {
  df <- methylKit::getData(myobjDB[[i]])
  df$sample <- sample.ids[i]
  cov.df <- rbind(cov.df, df)
}
cov.df <- cov.df[-1,]
save(cov.df, file = paste0(rdata.path,"Toothfish_2024_muscle_methylkit_cpg_cov.Rdata"))
```

### Coverage stats

```{r}
load(paste0(rdata.path,"Toothfish_2024_muscle_methylkit_cpgDB.Rdata"))
print(load(paste0(rdata.path,"Toothfish_2024_muscle_methylkit_cpg_cov.Rdata")))

length(unique(paste0(cov.df$chr, "_", cov.df$start))) #2,908,558

# summary(as.factor(cov.df$sample))
summary(cov.df$coverage)
    # Min.  1st Qu.   Median     Mean  3rd Qu.     Max. 
    # 4.00     8.00    17.00    24.76    32.00 44225.00 

summary(cov.df$coverage[cov.df$coverage > 5 & cov.df$coverage < 50])
   # Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
   # 6.00   10.00   17.00   20.06   28.00   49.00 

cov.plot <- ggplot2::ggplot(cov.df, ggplot2::aes(x = log(coverage), fill = sample)) +
  ggplot2::geom_histogram() +
  ggplot2::theme(legend.position = "none")
# print(cov.plot)
ggplot2::ggsave(cov.plot,
                filename =
                  paste0(figure.path,"Toothfish_2024_muscle_coverage_hist_Combined.png"), 
                width = 45, height = 15, units = 'cm')

cov.plot <- ggplot2::ggplot(cov.df, ggplot2::aes(x = coverage, fill = sample)) +
  ggplot2::geom_histogram() +
  ggplot2::theme(legend.position = "none")
# print(cov.plot)
ggplot2::ggsave(cov.plot,
                filename =
                  paste0(figure.path,"Toothfish_2024_muscle_coverage_hist_Combined2.png"), 
                width = 45, height = 15, units = 'cm')

cov.plot <- ggplot2::ggplot(cov.df[cov.df$coverage > 5 & cov.df$coverage < 160,], 
                            ggplot2::aes(x = coverage)) +
  ggplot2::geom_histogram(binwidth = 1) +
  ggplot2::ggtitle("coverage < 160") +
  ggplot2::theme(legend.position = "none")
# print(cov.plot)
ggplot2::ggsave(cov.plot,
                filename =
                  paste0(figure.path,"Toothfish_2024_muscle_coverage_hist_Combined4.png"),
                width = 45, height = 15, units = 'cm')



max(cov.df$coverage, na.rm = T) #44225
ind.covplot <- ggplot2::ggplot(cov.df, ggplot2::aes(x = sample, y = coverage)) + 
  ggplot2::geom_boxplot() +
  ggplot2::theme_bw() + 
  ggplot2::theme(axis.title.x = ggplot2::element_blank(), 
                 axis.text.y = ggplot2::element_text(size = 15),
                 axis.title.y = ggplot2::element_text(size = 15),
                 axis.text.x = ggplot2::element_text(size = 7, angle = 90, hjust = 1)) +
  # ggplot2::scale_y_continuous(trans = scales::log2_trans(),
  #                 breaks = c(1, 10, 50, 750, 20000, 50000),
  #                 minor_breaks = c(1:10, 2:5*10, 1:7.5*100, 0.5:4.5*1000))
  ggplot2::scale_y_continuous(trans = scales::log2_trans(),
                  breaks = c(1, 10, 50, 150),
                  minor_breaks = c(1:10, 2:10*10))

print(ind.covplot)

ggplot2::ggsave(ind.covplot, bg = "white",
                filename = 
                  paste0(figure.path,
                         "Toothfish_2024_muscle_raw_individual_coverage_CpGs.png"),
                width = 45, height = 15, units = 'cm')
```


### Coverage filter

High count filter based on coverage stats.

```{r, eval=FALSE}
load(paste0(rdata.path,"Toothfish_2024_muscle_methylkit_cpgDB.Rdata"))

filtered.myobj <- methylKit::filterByCoverage(myobjDB, lo.count = 10, lo.perc = NULL,
                                      hi.count = NULL, hi.perc = 99)
save(myobjDB,filtered.myobj, 
     file = paste0(rdata.path,"Toothfish_2024_muscle_methylkit_cpgDB.Rdata"))
```


### Normalise coverage

Normalize read coverage distributions between samples. Requires stringent filter on max coverage first, otherwise it will over-sample PCR bias reads. These two functions will help reduce the bias in the statistical tests that might occur due to systematic over-sampling of reads in certain samples.

```{r}
filtered.myobj.normal <- methylKit::normalizeCoverage(obj = filtered.myobj, 
                                                      method = "median", chunk.size = 1e6)
save(myobjDB,filtered.myobj, filtered.myobj.normal,
     file = paste0(rdata.path,"Toothfish_2024_muscle_methylkit_cpgDB.Rdata"))
```

### Number of sites per sample

```{r, eval = FALSE}
sites.df <- data.frame(sites = NA,  sample = NA)
for (i in 1:length(filtered.myobj.normal)) {
  print(i)
  sites.df[i,] <- c(filtered.myobj.normal[[i]]@num.records, filtered.myobj.normal[[i]]@sample.id)
}
sites.df$sites <- as.integer(sites.df$sites)
# sites.df$pop <- filtered.myobj.normal@treatment
sites.df$pop <- metadata.new$pop
save(sites.df, file = paste0(rdata.path,"Toothfish_2024_muscle_number_cpg_sample.Rdata"))
```


```{r}
print(load(paste0(rdata.path,"Toothfish_2024_muscle_number_cpg_sample.Rdata")))
sites.plot <- ggplot2::ggplot(sites.df, ggplot2::aes(x = sample, y = sites, colour = pop)) +
  ggplot2::geom_point() +
  ggplot2::scale_color_manual(values = colours2) +
  ggplot2::theme_bw() +
  ggplot2::theme(axis.text.x = ggplot2::element_text(angle = 90))
print(sites.plot)

ggplot2::ggsave(sites.plot,
                filename =
                  paste0(figure.path,"Toothfish_2024_muscle_number_cpg_sample.png"),
                width = 45, height = 15, units = 'cm')

sites.df$sample[sites.df$sites < 500000]
low.sites <- sites.df$sample[sites.df$sites < 500000]
# "TOA_58-4-2_B18_AE100308_muscle_F" "TOA_58-4-2_B24_AE100883_muscle_F" "TOA_58-4-2_B28_AE100885_muscle_M"
# "TOA_58-4-2_B30_AE100886_muscle_M" "TOA_58-4-2_B36_AE102050_muscle_M" "TOA_58-4-2_B38_AE102052_muscle_M"
# "TOA_58-4-2_B42_AE102105_muscle_F" "TOA_58-4-2_B48_AE102108_muscle_M" "TOA_58-4-2_B58_AE104944_muscle_F"
# "TOA_88-2_P24_AD292328_muscle_F"

## Could indicate an issue with bisulfite conversion?
```


## Select samples

```{r}
print(load(paste0(rdata.path,"Toothfish_2024_muscle_methylkit_cpgDB.Rdata")))
print(load(paste0(rdata.path,"Toothfish_metadata.Rdata")))

metadata2 <- metadata[!(metadata$Year %in% c(2017,2018) | metadata$Tissue_type == "fin" |
                                    grepl(pattern = "REP", x = metadata$File_names) |
                                    !grepl(pattern = "TOA", x = metadata$File_names)
                                  ),]

print(load(paste0(rdata.path, "Low_quality_fish.Rdata")))

# rm1 <- Reduce(intersect, list(low.genome.mapping,low.genome.methylation))
rm1 <- unique(c(low.genome.mapping,low.genome.methylation, low.sites))
print(rm1)
rm2 <- c("TOA_58-4-2_B40_AE102056_muscle_M") # bad PCA
rm3 <- c("TOA_58-4-2_B24_AE100883_muscle_F", "TOA_88-2_P24_AD292328_muscle_F",
         "TOA_58-4-2_B64_AE107905_muscle_F") # bad coverage
rm4 <- c("TOA_48-2_031_UK116","TOA_58-4-2_111_AN172604","TOA_88-1_230_AD082543",
         "TOA_48-2_024_UK082") #bad tissue PCA

rm <- c(rm1,rm2,rm3,rm4) # bad coverage

metadata2 <- metadata2[!(metadata2$File_names %in% rm | metadata2$Genotype_names %in% rm),]

filt.sample.id.full <- sapply(filtered.myobj@.Data, function(x){x@sample.id})

# filt.sample.id <- filt.sample.id.full[filt.sample.id.full %in% metadata2$File_names]
# metadata2 <- metadata2[order(match(metadata.new$File_names, filt.sample.id)),]

filt.sample.id <- filt.sample.id.full[filt.sample.id.full %in% metadata2$Genotype_names]
metadata2 <- metadata2[order(match(metadata2$Genotype_names, filt.sample.id)),]
identical(filt.sample.id,metadata2$Genotype_names)


pop <- metadata2$pop
pop.levels <- c("58.4.2-2024_50L","58.4.2-2024_70L",
                "88.2-2024")

pop <- as.numeric(factor(pop, levels = pop.levels))
sample.ids <- metadata2$Genotype_names



filtered.myobj2.normal <- methylKit::reorganize(
  methylObj = filtered.myobj.normal,
  sample.ids = sample.ids,
  treatment = pop,
  chunk.size = 1e+06,
  save.db = TRUE,
  suffix = "TOA_2024_muscle_normal"
)
save(myobjDB,filtered.myobj,filtered.myobj2.normal,
     file = paste0(rdata.path,"Toothfish_2024_muscle_methylkit_cpgDB.Rdata"))
```


## Merging samples

```{r}
print(load(paste0(rdata.path,"Toothfish_2024_muscle_methylkit_cpgDB.Rdata")))
pop <- metadata2$pop
pop.levels <- c("58.4.2-2024_50L","58.4.2-2024_70L",
                "88.2-2024")
pop <- factor(pop, levels = pop.levels)
pop <- as.numeric(pop)

table(pop)


meth.2024M.norm15L <- methylKit::unite(filtered.myobj2.normal, destrand = FALSE,
                                       save.db = FALSE,
                                       min.per.group = 15L, #missing data allowed
                                       mc.cores = 25)
#775,653   common sites for 15 samples per group

save(meth.2024M.norm15L,
     file = paste0(rdata.path,"Toothfish_2024_muscle_methylkit_object.Rdata"))
```


## Data QC

### Remove ASM

```{r}
load(paste0(rdata.path,"Allele_specific_mutations.Rdata"))

mut <- as(asmTRUE,"GRanges")
meth.2024M.norm15L <- as(meth.2024M.norm15L,"methylBase") 

sum(as(meth.2024M.norm15L,"GRanges") %over% mut)#28018
meth.asm <- methylKit::selectByOverlap(meth.2024M.norm15L, mut) 
BOOLEAN <- paste0(meth.2024M.norm15L$chr,"_",meth.2024M.norm15L$start) %in%
  paste0(asmTRUE$Chr,"_",asmTRUE$Pos)
sum(BOOLEAN) #659
meth.2024M.norm15L <- methylKit::select(meth.2024M.norm15L, !BOOLEAN) 

save(meth.2024M.norm15L,
     file = paste0(rdata.path,"Toothfish_2024_muscle_methylkit_object.Rdata"))
```



### Coverage 

```{r}
meth.df <- methylKit::getData(meth.2024M.norm15L)
cov <- meth.df[,meth.2024M.norm15L@coverage.index]
colnames(cov) <- meth.2024M.norm15L@sample.ids
dpf <- data.frame(cov, locus = paste0(meth.df$chr,"_",meth.df$start))
dpf <- tidyr::gather(dpf, key = individual, 
                           value = Depth, -locus,
                           factor_key = TRUE) %>%
  dplyr::group_by(individual) %>%
  dplyr::mutate(NA_cnt = sum(is.na(Depth))/length(unique(locus))) %>%
  dplyr::ungroup()
dpf.NA <- dpf[!duplicated(dpf$individual),]
dpf.NA$NA_cnt <- round(dpf.NA$NA_cnt, 3)

dpf <- dpf[ dpf$Depth > 0 & !is.na(dpf$Depth),]

max(dpf$Depth, na.rm = T) #99
myPlot <- ggplot2::ggplot(dpf, ggplot2::aes(x = individual, y = Depth)) + 
  ggplot2::geom_boxplot() +
  ggplot2::geom_text(data = dpf.NA,ggplot2::aes(x = individual, y = 150, label = NA_cnt), 
                     size = 2, angle = 90) +
  ggplot2::theme_bw() + 
  ggplot2::theme(axis.title.x = ggplot2::element_blank(), 
                 axis.text.y = ggplot2::element_text(size = 15),
                 axis.title.y = ggplot2::element_text(size = 15),
                 axis.text.x = ggplot2::element_text(size = 7, angle = 90, hjust = 1)) +
  ggplot2::scale_y_continuous(trans = scales::log2_trans(),
                  breaks = c(1, 10, 50, 100),
                  minor_breaks = c(1:10, 2:5*10))

print(myPlot)
ggplot2::ggsave(myPlot, bg = "white",
                filename = 
                  paste0(figure.path,
                  "1B.Toothfish_2024_muscle_normalised15_methylkit_individual_coverage_CpGs.png"
                                  ), 
                width = 45, height = 15, units = "cm")

nrow(cov) # 776851 SMPs
plot(x = 1:ncol(cov), y = colMeans(cov, na.rm = TRUE), col = meth.2024M.norm15L@treatment)
plot(x = 1:ncol(cov), y = colMeans(cov > 20, na.rm = TRUE), col = meth.2024M.norm15L@treatment)
plot(x = 1:ncol(cov), y = colMeans(cov > 10, na.rm = TRUE), col = meth.2024M.norm15L@treatment)
meth.2024M.norm15L@sample.ids[colMeans(cov > 10, na.rm = TRUE) < 0.95]
# samples with low coverage: "TOA_58-4-2_B24_AE100883_muscle_F", "TOA_88-2_P24_AD292328_muscle_F"
```



## Load data

```{r}
print(load(paste0(rdata.path,"Toothfish_2024_muscle_methylkit_object.Rdata")))
print(load(paste0(rdata.path,"Toothfish_metadata.Rdata"))) 

metadata.new <- metadata[metadata$Genotype_names %in% meth.2024M.norm15L@sample.ids,]
metadata.new <- metadata.new[order(match(metadata.new$Genotype_names,
                                         meth.2024M.norm15L@sample.ids)),]

length(meth.2024M.norm15L@row.names) # number of sites
```


## Global clustering analysis

```{r}
methylKit::clusterSamples(meth.2024M.norm15L, dist = "correlation", sd.filter = TRUE, 
                          method = "ward.D2", plot = TRUE, )
dev.print(file = paste0(figure.path,"1_TOA_2024_muscle_cluster_plot.png"),
          device = png, res = 300,width = 45,height = 15,units = "cm")
```


### PCA

```{r}
PCobj <- methylKit::PCASamples(meth.2024M.norm15L, comp = c(1,2), 
                               adj.lim = c(0,0), obj.return = TRUE)

var_frac <- PCobj$sdev/sum(PCobj$sdev)
print(paste0("Total variance explained: ",signif(sum(var_frac[1:3]) * 100, 3), "%")) 
barplot(var_frac[1:20], main = "variance explained", 
        col = heat.colors(length(var_frac[1:20])))

identical(metadata.new$Genotype_names, meth.2024M.norm15L@sample.ids)
identical(metadata.new$Genotype_names, rownames(PCobj$x))

pop <- metadata.new$pop

plot.title <- "Toothfish methylation - 2024 muscle (774,994 CpG sites)"

data <- data.frame(ind = rownames(PCobj$x),PCobj$x,  pop = pop)
# data <- data.frame(ind = rownames(PCobj$rotation),PCobj$rotation, pop = pop)
data <- merge(data, aggregate(cbind(mean.x = PC1, mean.y = PC2) ~
                                pop, data, mean), by = "pop")

pca.plot <- ggplot2::ggplot(data, ggplot2::aes(PC1, PC2, color = pop)) +
  ggplot2::geom_point(size = 4) + 
  ggplot2::geom_point(ggplot2::aes(x = mean.x,
                                   y = mean.y, color = pop),
                      size = 7,
                      shape = 17) +
  ggplot2::scale_colour_manual(values = colours2) +
  ggplot2::geom_hline(yintercept = 0) +
  ggplot2::geom_vline(xintercept = 0) + 
  ggplot2::labs(
    subtitle = paste0("Total variance explained: ",
                      signif(sum(var_frac[1:4]) * 100, 3), "%"),
    y = paste0("PC2; ", signif(var_frac[2] * 100, digits = 2), " %"),
    x = paste0("PC1; ", signif(var_frac[1] * 100, digits = 2), " %"),
    title = plot.title,
    caption = "") + 
  ggplot2::theme_minimal() +
  ggplot2::theme(
    axis.text = ggplot2::element_text(size = 15),
    axis.title.x = ggplot2::element_text(size = 20),
    axis.title.y = ggplot2::element_text(size = 20),
    legend.text = ggplot2::element_text(size = 15)
  )

print(pca.plot)
ggplot2::ggsave(pca.plot, bg = "white",
                filename = paste0(figure.path,"Toothfish_2024_muscle_PCA_CpG_PC12.png"), 
                width = 30, height = 15, units = "cm")



data <- data.frame(ind = rownames(PCobj$x),PCobj$x,  pop = pop)
# data <- data.frame(ind = rownames(PCobj$rotation),PCobj$rotation, pop = pop)
data <- merge(data, aggregate(cbind(mean.x = PC2, mean.y = PC3) ~
                                pop, data, mean), by = "pop")
pca.plot <- ggplot2::ggplot(data, ggplot2::aes(PC2, PC3, color = pop)) +
  ggplot2::geom_point(size = 4) + 
  ggplot2::geom_point(ggplot2::aes(x = mean.x,
                                   y = mean.y, color = pop),
                      size = 7,
                      shape = 17) +
  ggplot2::scale_colour_manual(values = colours2) +
  ggplot2::geom_hline(yintercept = 0) +
  ggplot2::geom_vline(xintercept = 0) + 
  ggplot2::labs(
    subtitle = paste0("Total variance explained: ",
                      signif(sum(var_frac[1:4]) * 100, 3), "%"),
    y = paste0("PC3; ", signif(var_frac[3] * 100, digits = 2), " %"),
    x = paste0("PC2; ", signif(var_frac[2] * 100, digits = 2), " %"),
    title = plot.title,
    caption = "") + 
  ggplot2::theme_minimal() +
  ggplot2::theme(
    axis.text = ggplot2::element_text(size = 15),
    axis.title.x = ggplot2::element_text(size = 20),
    axis.title.y = ggplot2::element_text(size = 20),
    legend.text = ggplot2::element_text(size = 15)
  )

print(pca.plot)
ggplot2::ggsave(pca.plot, bg = "white",
                filename = paste0(figure.path,"Toothfish_2024_muscle_PCA_CpG_PC23.png"), 
                width = 30, height = 15, units = "cm")

# data$ind[data$PC3 > 100] #"TOA_58-4-2_B40_AE102056_muscle_M"
# data$ind[data$PC3 < -100] #"TOA_58-4-2_B64_AE107905_muscle_F"
```



### RDA

```{r}
# setwd("/datasets/work/ncmi-toa-rrbs/work/1.Toothfish/Toothfish_RRBS_array_22098078/Bismark/")
# print(load(paste0(rdata.path,"Toothfish_methylation_objects_raw.Rdata"))) #36,635 SMPs
print(load(paste0(rdata.path,"Toothfish_subset_methylation_objects_raw.Rdata"))) #1,316,799 SMPs

# meth.RDA <- methylKit::unite(filtered.myobj2.norm, destrand = FALSE,
#                              save.db = FALSE,
#                              suffix = "subset",
#                              min.per.group = NULL, #missing data allowed
#                              mc.cores = 25)


# meth.df <- methylKit::getData(meth)
# cov <- meth.df[,grep(colnames(meth.df), pattern = "coverage*")]
# cov <- meth.df[,meth@coverage.index]
# colnames(cov) <- meth@sample.ids
# meth.df <- methylKit::percMethylation(meth)

# meth.df <- methylKit::percMethylation(meth.RDA)

# save(meth.df,meth.RDA, 
#      file = paste0(rdata.path,"TOA_data_for_RDA_meth_percent.Rdata"))
print(load(paste0(rdata.path,"TOA_data_for_RDA_meth_percent.Rdata"))) #347,614 SMPs

hist(colSums(is.na(meth.df)), breaks = 100)
hist(rowSums(is.na(meth.df)), breaks = 100)





print(load(paste0(rdata.path,"Toothfish_metadata.Rdata")))

predictor <- metadata2 # define the predictor input file here
predictor <- predictor[predictor$SAMPLE_ID %in% meth@sample.ids,]
predictor$SAMPLE_ID = as.character(predictor$SAMPLE_ID)
predictor$region = as.character(predictor$region)
predictor$Sex = as.character(predictor$Sex)
predictor$Tissue = as.character(predictor$Tissue)
predictor$mat = as.character(predictor$mat)
predictor$mat[is.na(predictor$mat)] <- "2"
predictor$TL = as.numeric(predictor$TL)
predictor$wt = as.numeric(predictor$wt)
predictor$dpth = as.numeric(predictor$dpth)
predictor$Age = as.numeric(predictor$Age)
predictor$Age[is.na(predictor$Age)] <- as.numeric(max(predictor$Age, na.rm = TRUE))
predictor$Cohort2 = as.integer(predictor$Cohort2)
predictor$Cohort2[is.na(predictor$Cohort2)] <- as.integer(min(predictor$Cohort2, na.rm = TRUE))


meth.df1 = t(meth.df)
# meth.df2 = na.omit(meth.df1) #rda can't handle NAs, remove them if you have them
# meth.df1[1:10,1:10]
## Predictor
# Make names characters (not factors)


#if you had numbers, they had to be factors (as.factor)

# Confirm that methylation data and predictors are in the same order
identical(rownames(meth.df1), predictor$SAMPLE_ID)

# Perform Hellinger transformation on the methylation data:
meth.df3 = vegan::decostand(meth.df1, method = "hellinger", MARGIN = 1) #MARGIN=1 means rows, on rows I have samples now because I transposed the matrix

## Run RDA
#Test the country using growing condition as a Condition () (covariate):
meth_rda = vegan::rda(meth.df3 ~ region + Condition(Sex), 
                      data = predictor, scale = F) #I remove scale since I performed Hellinger transformation
meth_rda = vegan::rda(meth.df3 ~ region + Condition(Sex) + 
                        Condition(mat)+ Condition(TL) + Condition(wt) +
                        Condition(dpth) + Condition(Age) + Condition(Cohort2), 
                      na.action = "na.omit",
                      data = predictor, scale = F)

meth_rda$pCCA
meth_rda$CCA$wa

summary(meth_rda) #Constrained Proportion: variance of Y explained by X, Unconstrained Proportion: unexplained variance in Y
# sink(paste0(rdata.path,"anova_rda.txt")) #with this you can save the output of the anova in a txt file, define the output directory
meth.aov <- anova(meth_rda, test = "Chi", permutations = 99) #run Anova on the result to see the significance level
# meth.aov.cca <- anova.cca(meth_rda, test = "Chi", permutations = 99)
print(meth.aov)
# sink(paste0(rdata.path,"anova_rda_chi.txt"))

# Model: rda(formula = meth.df3 ~ region + Condition(Sex), data = predictor, scale = F)
#           Df  Variance      F Pr(>F)   
# Model      3 0.0003924 1.4126  0.004 **
# Residual 110 0.0101861

# Model: rda(formula = meth.df3 ~ region + Condition(Sex) + Condition(mat) + Condition(TL) + Condition(wt) + Condition(dpth) + Condition(Age) + Condition(Cohort2), data = predictor, scale = F, na.action = "na.omit")
#           Df  Variance     F Pr(>F)   
# Model      3 0.0003412 1.257   0.01 **
# Residual 101 0.0091396 

Pop.names <- predictor$region
plot.a <- vegan::ordiplot(meth_rda, choices=c(1,2))
Pop.ellipses <- vegan::ordiellipse(plot.a, groups=Pop.names, display="sites", kind="sd")
dev.print(file = paste0(figure.path,"TOA_RDA_ordiplot_complex_model.png"),
          device = png, res = 300,width = 45,height = 15,units = "cm")

# Type 1 scaling
vegan::ordiplot(meth_rda, scaling = 1, type = "text")
# Type 2 scaling
vegan::ordiplot(meth_rda, scaling = 2, type = "text")

ggvegan::autoplot(
  meth_rda,
  axes = c(1, 2),
  geom = "text",
  layers = c("species", "sites", "biplot", "centroids"),
  arrows = TRUE)

#CpG contribution
hist(meth_rda$CCA$v[,1])
hist(meth_rda$CCA$v[,2])
hist(meth_rda$CCA$v[,3])



#Test the interaction of your predictors:
meth_rda3 = vegan::rda(meth ~ region * Sex + Condition(region + Sex), data = predictor, scale = F)
summary(meth_rda3)
sink(paste0(rdata.path,"anova_rda_interaction.txt"))
print(anova(meth_rda3, test="Chi"))
sink(paste0(rdata.path,"anova_rda_interaction_chi.txt"))
```

#### plot RDA

```{r}

BioR.theme <- theme(
        panel.background = element_blank(),
        panel.border = element_blank(),
        panel.grid = element_blank(),
        axis.line = element_line("gray25"),
        text = element_text(size = 12),
        axis.text = element_text(size = 10, colour = "gray25"),
        axis.title = element_text(size = 14, colour = "gray25"),
        legend.title = element_text(size = 14),
        legend.text = element_text(size = 14),
        legend.key = element_blank())

plot.a <- vegan::ordiplot(meth_rda, choices=c(1,2))
Pop.names <- predictor$region

# sites1 <- BiodiversityR::sites.long(plot.a, env.data=predictor)
sites1 <- data.frame(cbind(predictor, axis1=plot.a$sites[, 1], 
                           axis2=plot.a$sites[, 2], 
                           labels=rownames(plot.a$sites)),
                     group = predictor$region)

# axis1 <- BiodiversityR::axis.long(meth_rda, choices=c(1, 2))
eigs.all <- meth_rda$CA$eig
eigs <- round(100 * eigs.all / meth_rda$tot.chi, digits=1)[c(1, 2)] 
xlab.label <- paste0(names(eigs)[1], " (", eigs[1], "%)")
ylab.label <- paste0(names(eigs)[2], " (", eigs[2], "%)")
axis1 <- data.frame(axis=c(1:2), 
                    ggplot=c("xlab.label", "ylab.label"), 
                    label=c(xlab.label, ylab.label))


# centroids.Pop1 <- BiodiversityR::centroids.long(sites1, 
#                                                 grouping=Pop.names, 
#                                                 centroids.only=TRUE)
# centroids.Pop2 <- BiodiversityR::centroids.long(sites1, 
#                                                 grouping=Pop.names, 
#                                                 centroids.only=FALSE)
gr.name <- rlang::enexpr(Pop.names)
cent.means <-  stats::aggregate(cbind(axis1, axis2) ~ Pop.names,
                                data = sites1,
                                FUN = mean)
names(cent.means) <- c("group", "axis1c", "axis2c")
centroids.Pop1 <- cent.means
cent.means$Centroid <- cent.means[, 1]
by.name <- names(sites1)[names(sites1) == "group"]
centroids.Pop2 <- dplyr::full_join(sites1, cent.means, by = by.name)


# Pop.ellipses.long2 <- BiodiversityR::ordiellipse.long(Pop.ellipses, 
#                                                     grouping.name="Pop.names")


plotgg.a <- ggplot() + 
    geom_vline(xintercept = c(0), color = "grey70", linetype = 2) +
    geom_hline(yintercept = c(0), color = "grey70", linetype = 2) +  
    xlab(axis1[1, "label"]) +
    ylab(axis1[2, "label"]) +  
    scale_x_continuous(sec.axis = dup_axis(labels=NULL, name=NULL)) +
    scale_y_continuous(sec.axis = dup_axis(labels=NULL, name=NULL)) +
    # geom_polygon(data=Pop.ellipses.long2, 
    #                aes(x=axis1, y=axis2, colour=Pop.names, 
    #                    fill=after_scale(alpha(colour, 0.2))), 
    #           size=0.2, show.legend=TRUE) +
    geom_segment(data=centroids.Pop2, 
                 aes(x=axis1c, y=axis2c, xend=axis1, yend=axis2, colour=Pop.names), 
                 size=0.7, show.legend=FALSE) +
    geom_point(data=centroids.Pop1, 
               aes(x=axis1c, y=axis2c), 
               shape="square", colour="black", alpha=0.7, size=2) +
    geom_label_repel(data=centroids.Pop1, 
               aes(x=axis1c, y=axis2c, label=group, colour=group), 
               alpha=1.0, size=2, show.legend=FALSE) +
    # labs(colour = Pop.names) +
    BioR.theme +
    theme(legend.text = element_text(size = 10)) +
    scale_colour_npg() +
    coord_fixed(ratio=1)

plotgg.a

ggsave(plotgg.a, file = paste0(figure.path, 
                               # "TOA_RDA_simple_model_A.png"
                               "TOA_RDA_complex_model_A.png"
                               ),
       height = 30, width = 30, units = "cm",  device = "png")
```



## DMS - 2024 - muscle

```{r, eval = FALSE}
print(load(paste0(rdata.path,"Toothfish_2024_muscle_methylkit_object.Rdata")))

print(load(paste0(rdata.path,"Toothfish_metadata.Rdata")))
metadata.new <- metadata[metadata$Genotype_names %in% attributes(meth.2024M.norm15L)$sample.ids,]
attributes(meth.2024M.norm15L)$sample.ids == metadata.new$Genotype_names

# meth.2024M.norm15L@treatment <- as.integer(factor(metadata.new$pop))
meth.2024M.norm15L@treatment <- as.integer(factor(metadata.new$STRATA2))

# data.frame(attributes(meth.2024M.norm15L)$sample.ids, metadata.new$Genotype_names, metadata.new$strata, meth.2024M.norm15L@treatment)

covariates <- dplyr::select(metadata.new, Sex, Age, Cohort2, Maturity_full)
myDiff <- methylKit::calculateDiffMeth(meth.2024M.norm15L,
                                       covariates = NULL,
                                       adjust = "fdr",
                                       chunk.size = 1e+06,
                                       mc.cores = 25)
meth.diff.chr <- methylKit::diffMethPerChr(myDiff,plot = TRUE, 
                                           qvalue.cutoff = 0.01, meth.cutoff = 40)
myDiff40p <- methylKit::getMethylDiff(myDiff, difference = 40, qvalue = 0.01)
#4
myDiff25p <- methylKit::getMethylDiff(myDiff, difference = 25, qvalue = 0.01)
#1211  
myDiff15p <- methylKit::getMethylDiff(myDiff, difference = 15, qvalue = 0.01)
#15709  
top100.2024M <- slice_max(as.data.frame(as(myDiff,"GRanges")), order_by = meth.diff, n = 100)
top100.2024M <- as(top100.2024M,"GRanges")

save(myDiff,myDiff15p, myDiff25p, myDiff40p,top100.2024M,
     file = paste0(rdata.path,"Toothfish_2024_muscle_methylation_DMS.Rdata"))
```


```{r}
print(load(paste0(rdata.path,"Toothfish_2024_muscle_methylation_DMS.Rdata")))

volcano.plot <- ggplot2::ggplot() +
  ggplot2::geom_point(data = myDiff, ggplot2::aes(x = meth.diff, 
                                                  y = -log10(qvalue))) +
  ggplot2::geom_point(data = myDiff25p, ggplot2::aes(x = meth.diff, 
                                                  y = -log10(qvalue)),
                      colour = "red") +
  ggplot2::theme_minimal() +
  ggplot2::theme(
    axis.text = ggplot2::element_text(size = 15),
    axis.title.x = ggplot2::element_text(size = 20),
    axis.title.y = ggplot2::element_text(size = 20),
    legend.text = ggplot2::element_text(size = 15)
  )
print(volcano.plot)

ggplot2::ggsave(volcano.plot, bg = "white",
                filename =
                  paste0(figure.path,
                         "Toothfish_2024_muscle_DMS25_VOLCANO_methylkit_regions.png"), 
                width = 30, height = 15, units = "cm")


myDiff25p.gr <- as(myDiff25p, "GRanges")
grand.plot <- ggbio::plotGrandLinear(myDiff25p.gr,
                                     ggplot2::aes(y = meth.diff)) +
  ggplot2::ylab("Methylation difference (%)") +
  ggplot2::theme_bw() +
  ggplot2::theme(axis.text.x = ggplot2::element_text(angle = 90, hjust = 0), 
                 legend.position = "none")
print(grand.plot)
ggplot2::ggsave(grand.plot@ggplot, bg = "white",
                filename = paste0(figure.path,
                                  "Toothfish_2024_muscle_DMS25_Grandplot_methylkit_regions.png"
                ), 
                width = 30, height = 15, units = "cm")
```

### DMS in common with 2018 DMS
```{r}
print(load(paste0(rdata.path,"Toothfish_DMS_CpGsites.Rdata")))
meth.cpg <- paste0(myDiff25p$chr, "_", myDiff25p$start) #1211 
DMS.cpg <- unique(c(DMS$DMS)) #1243
sum(meth.cpg %in% DMS.cpg) #32 = 4%

DMS.cpg <- unique(c(DMS.ER$DMS, DMS.EANT$DMS)) #74
sum(meth.cpg %in% DMS.cpg) #13 = 17%
```

### Heatmap

```{r}
print(load(paste0(rdata.path,"Toothfish_2024_muscle_methylation_DMS.Rdata")))

myDiff25p.gr <- as(myDiff25p, "GRanges")
meth.diff <- methylKit::selectByOverlap(meth.2024M.norm15L, myDiff25p.gr)

meth.diff.df <- as.data.frame(t(methylKit::percMethylation(meth.diff)))
loc.names <- paste0(meth.diff$chr,"_",meth.diff$start)
colnames(meth.diff.df) <- loc.names

meth.diff.df$individual <- rownames(meth.diff.df)
pop <- metadata.new$pop
meth.diff.df$pop <- as.character(pop)


heat.plot.data <- meth.diff.df %>%
  tibble::as_tibble() %>%
  tidyr::gather(key = "CpG", value = "Methylation_percentage", -individual, -pop,
                factor_key = FALSE) %>%
  dplyr::group_by(pop, CpG) %>%
  dplyr::mutate(CpG_mean = mean(Methylation_percentage, na.rm = TRUE)) %>%
  dplyr::arrange(pop, individual, CpG) %>%
  dplyr::mutate(individual = factor(individual, levels = unique(individual))) %>%
  dplyr::mutate(CpG = factor(CpG, levels = unique(CpG)))
cpgs <- heat.plot.data$CpG[heat.plot.data$individual == unique(heat.plot.data$individual)[1]]
cpgs_mean <- heat.plot.data$CpG_mean[heat.plot.data$individual == unique(heat.plot.data$individual)[1]]

cpgs <- cpgs[order(cpgs_mean, decreasing = TRUE)]
heat.plot.data$CpG <- factor(heat.plot.data$CpG, levels = c("pop", unique(as.character(cpgs))))
heat.plot.data <- heat.plot.data[order(heat.plot.data$CpG),]

heat.plot <- ggplot2::ggplot(data = heat.plot.data, ggplot2::aes(x = individual)) + 
    ggplot2::geom_tile(data = heat.plot.data, 
                       ggplot2::aes(x = individual, y = CpG, fill = Methylation_percentage)) +
    ggplot2::scale_fill_gradient2(low = "blue", mid = "yellow", high = "red", 
                                  midpoint = 50, guide = "colourbar") +
    ggnewscale::new_scale_fill() +
    ggplot2::geom_tile(data = heat.plot.data, ggplot2::aes(x = individual, y = "pop", fill = pop,
                                    height = length(unique(heat.plot.data$CpG))*0.02)) +
    ggplot2::scale_fill_manual(values = colours2) +
    ggplot2::scale_y_discrete(drop = FALSE) +
    ggplot2::theme(legend.position = "top",
                   axis.title.y = ggplot2::element_blank(),
                   axis.title.x = ggplot2::element_blank(),
                   axis.text.x = ggplot2::element_text(face = "bold", color = "black", 
                                        size = 10, angle = 90, hjust = 1, vjust = 0.5))
print(heat.plot)

ggplot2::ggsave(heat.plot, bg = "white",
                filename = paste0(figure.path,
                                  "Toothfish_2024_muscle_methDiff_25_HEATMAP.png"), 
                width = 30, height = 30, units = "cm")
```

### Plot CpG difference

```{r}
load(file = paste0(rdata.path,"Toothfish_2024_muscle_methylation_DMS.Rdata"))

myDiff25p.gr <- as(myDiff25p, "GRanges")
meth.diff <- methylKit::selectByOverlap(meth.2024M.norm15L, myDiff25p.gr) 

length(unique(meth.diff$chr))
sort(table(meth.diff$chr), decreasing = TRUE)[1:20]

meth.diff.df <- as.data.frame(t(methylKit::percMethylation(meth.diff)))
loc.names <- paste0(meth.diff$chr,"_",meth.diff$start)
colnames(meth.diff.df) <- loc.names

meth.diff.df$individual <- rownames(meth.diff.df) 
identical(metadata.new$Genotype_names, meth.diff.df$individual)
meth.diff.df$pop <- metadata.new$pop



meth.diff.df.plot <- tidyr::gather(meth.diff.df, key = locus, 
                           value = perc_methylation, -individual, -pop,
                           factor_key = TRUE) %>%
  dplyr::mutate(chr = stringr::str_split_fixed(locus, 
                                               pattern = "_", 
                                               n = 2)[,1])
meth.diff.df.plot.locus <- meth.diff.df.plot %>% dplyr::group_by(chr,locus,pop) %>% 
  dplyr::summarise(mean_pop = mean(perc_methylation, na.rm = TRUE),
                   sum_NA = sum(is.na(perc_methylation)),
            .groups = 'drop')


diff.plot <- ggplot2::ggplot() +
  ggplot2::geom_line(data = meth.diff.df.plot.locus,
                     ggplot2::aes(x = pop, y = mean_pop, colour = locus, group = locus)) +
  ggplot2::geom_point(data = meth.diff.df.plot.locus,
                     ggplot2::aes(x = pop, y = mean_pop, colour = locus, group = locus),
                     size = 5) +
  ggplot2::facet_wrap(~chr) +
  ggplot2::ylab("Mean methylation (%)") +
  ggplot2::theme(legend.position = "none",
                   axis.text.x = ggplot2::element_text(face = "bold", color = "black", 
                                        size = 10, angle = 90, hjust = 1, vjust = 0.5))

print(diff.plot)

ggplot2::ggsave(plot = diff.plot, 
                filename = paste0(figure.path,"1_Toothfish_2024_muscle_methDiff.png"),
                width = 30, height = 15, units = "cm", device = "png")
```


### DMS Clustering
```{r}
load(file = paste0(rdata.path,"Toothfish_2024_muscle_methylation_DMS.Rdata"))

myDiff25p.gr <- as(myDiff25p, "GRanges")
meth.diff <- methylKit::selectByOverlap(meth.2024M.norm15L, myDiff25p.gr) #112  
methdif.perc <- methylKit::percMethylation(meth.diff)

# methdif.perc2 <- zoo::na.aggregate(methdif.perc, by = meth.2024M.norm15L@treatment)
# PCobj <- prcomp(methdif.perc2, retx = TRUE, scale. = TRUE, center = TRUE)

PCobj <- methylKit::PCASamples(meth.diff, comp = c(1,2), 
                               adj.lim = c(0,0), obj.return = TRUE)

var_frac <- PCobj$sdev/sum(PCobj$sdev)
print(paste0("Total variance explained: ",signif(sum(var_frac[1:3]) * 100, 3), "%")) 
barplot(var_frac[1:3], main = "variance explained", 
        col = heat.colors(length(var_frac[1:3])))


identical(metadata.new$Genotype_names, meth.2024M.norm15L@sample.ids)
identical(metadata.new$Genotype_names, rownames(PCobj$x))
# identical(metadata.new$Genotype_names, rownames(PCobj$rotation))

pop <- metadata.new$pop


plot.title <- "Toothfish methylation - 2024 Muscle (1211 DMS sites - 25%)"

data <- data.frame(ind = rownames(PCobj$x),PCobj$x,  pop = pop)
# data <- data.frame(ind = rownames(PCobj$rotation),PCobj$rotation,  pop = pop)
data <- merge(data, aggregate(cbind(mean.x = PC1, mean.y = PC2) ~
                                pop, data, mean), by = "pop")

pca.plot <- ggplot2::ggplot(data, ggplot2::aes(PC1, PC2, color = pop)) +
  ggplot2::geom_point(size = 4) + 
  ggplot2::geom_point(ggplot2::aes(x = mean.x,
                                   y = mean.y, color = pop),
                      size = 7,
                      shape = 17) +
  ggplot2::scale_colour_manual(values = colours2.new) +
  ggplot2::geom_hline(yintercept = 0) +
  ggplot2::geom_vline(xintercept = 0) + 
  ggplot2::labs(
    subtitle = paste0("Total variance explained: ",
                      signif(sum(var_frac[1:3]) * 100, 3), "%"),
    y = paste0("PC2; ", signif(var_frac[2] * 100, digits = 2), " %"),
    x = paste0("PC1; ", signif(var_frac[1] * 100, digits = 2), " %"),
    title = plot.title,
    caption = "") + 
  ggplot2::theme_minimal() +
  ggplot2::theme(
    axis.text = ggplot2::element_text(size = 15),
    axis.title.x = ggplot2::element_text(size = 20),
    axis.title.y = ggplot2::element_text(size = 20),
    legend.text = ggplot2::element_text(size = 15)
  )

print(pca.plot)
ggplot2::ggsave(pca.plot, bg = "white",
                filename = paste0(figure.path,"Toothfish_2024_muscle_PCA_CpG_DMS_PC12.png"), 
                width = 30, height = 15, units = "cm")



data <- data.frame(ind = rownames(PCobj$x),PCobj$x,  pop = pop)
# data <- data.frame(ind = rownames(PCobj$rotation),PCobj$rotation,  pop = pop)
data <- merge(data, aggregate(cbind(mean.x = PC1, mean.y = PC3) ~
                                pop, data, mean), by = "pop")

pca.plot <- ggplot2::ggplot(data, ggplot2::aes(PC1, PC3, color = pop)) +
  ggplot2::geom_point(size = 4) + 
  ggplot2::geom_point(ggplot2::aes(x = mean.x,
                                   y = mean.y, color = pop),
                      size = 7,
                      shape = 17) +
  ggplot2::scale_colour_manual(values = colours2.new) +
  ggplot2::geom_hline(yintercept = 0) +
  ggplot2::geom_vline(xintercept = 0) + 
  ggplot2::labs(
    subtitle = paste0("Total variance explained: ",
                      signif(sum(var_frac[1:3]) * 100, 3), "%"),
    y = paste0("PC3; ", signif(var_frac[3] * 100, digits = 2), " %"),
    x = paste0("PC1; ", signif(var_frac[1] * 100, digits = 2), " %"),
    title = plot.title,
    caption = "") + 
  ggplot2::theme_minimal() +
  ggplot2::theme(
    axis.text = ggplot2::element_text(size = 15),
    axis.title.x = ggplot2::element_text(size = 20),
    axis.title.y = ggplot2::element_text(size = 20),
    legend.text = ggplot2::element_text(size = 15)
  )

print(pca.plot)
ggplot2::ggsave(pca.plot, bg = "white",
                filename = paste0(figure.path,
                                  "Toothfish_2024_muscle_PCA_CpG_DMS_PC13.png"), 
                width = 30, height = 15, units = "cm")




data <- data.frame(ind = rownames(PCobj$x),PCobj$x,  pop = pop)
# data <- data.frame(ind = rownames(PCobj$rotation),PCobj$rotation,  pop = pop)
data <- merge(data, aggregate(cbind(mean.x = PC2, mean.y = PC3) ~
                                pop, data, mean), by = "pop")

pca.plot <- ggplot2::ggplot(data, ggplot2::aes(PC2, PC3, color = pop)) +
  ggplot2::geom_point(size = 4) + 
  ggplot2::geom_point(ggplot2::aes(x = mean.x,
                                   y = mean.y, color = pop),
                      size = 7,
                      shape = 17) +
  ggplot2::scale_colour_manual(values = colours2.new) +
  ggplot2::geom_hline(yintercept = 0) +
  ggplot2::geom_vline(xintercept = 0) + 
  ggplot2::labs(
    subtitle = paste0("Total variance explained: ",
                      signif(sum(var_frac[1:3]) * 100, 3), "%"),
    y = paste0("PC3; ", signif(var_frac[3] * 100, digits = 2), " %"),
    x = paste0("PC2; ", signif(var_frac[2] * 100, digits = 2), " %"),
    title = plot.title,
    caption = "") + 
  ggplot2::theme_minimal() +
  ggplot2::theme(
    axis.text = ggplot2::element_text(size = 15),
    axis.title.x = ggplot2::element_text(size = 20),
    axis.title.y = ggplot2::element_text(size = 20),
    legend.text = ggplot2::element_text(size = 15)
  )

print(pca.plot)
ggplot2::ggsave(pca.plot, bg = "white",
                filename = paste0(figure.path,"Toothfish_2024_muscle_PCA_CpG_DMS_PC23.png"), 
                width = 30, height = 15, units = "cm")
```



## DMS RANDOM - 2024 - muscle

```{r,eval=FALSE}
print(load(paste0(rdata.path,"Toothfish_2024_muscle_methylkit_object.Rdata")))

print(load(paste0(rdata.path,"Toothfish_metadata.Rdata")))
metadata.new <- metadata[metadata$Genotype_names %in% attributes(meth.2024M.norm15L)$sample.ids,]
attributes(meth.2024M.norm15L)$sample.ids == metadata.new$Genotype_names

strata.random <- sample(metadata.new$pop, size = length(metadata.new$pop), replace = FALSE)
meth.2024M.norm15L@treatment <- as.integer(factor(strata.random))

# data.frame(attributes(meth.2024M.norm15L)$sample.ids, metadata.new$Genotype_names, metadata.new$strata, meth.2024M.norm15L@treatment)

# covariates <- dplyr::select(metadata.new, Sex, TL_mm)
myDiff <- methylKit::calculateDiffMeth(meth.2024M.norm15L,
                                       covariates = NULL,
                                       adjust = "fdr",
                                       chunk.size = 1e+06,
                                       mc.cores = 25)
meth.diff.chr <- methylKit::diffMethPerChr(myDiff,plot = TRUE, 
                                           qvalue.cutoff = 0.01, meth.cutoff = 40)
myDiff40p <- methylKit::getMethylDiff(myDiff, difference = 40, qvalue = 0.01)
#1
myDiff25p <- methylKit::getMethylDiff(myDiff, difference = 25, qvalue = 0.01)
#129
myDiff15p <- methylKit::getMethylDiff(myDiff, difference = 15, qvalue = 0.01)
#3755    

save(strata.random, myDiff,myDiff15p, myDiff25p, myDiff40p,
     file = paste0(rdata.path,"Toothfish_2024_muscle_methylation_RANDOM_DMS.Rdata"))
```



```{r}
print(load(paste0(rdata.path,"Toothfish_2024_muscle_methylation_RANDOM_DMS.Rdata")))

volcano.plot <- ggplot2::ggplot() +
  ggplot2::geom_point(data = myDiff, ggplot2::aes(x = meth.diff, 
                                                  y = -log10(qvalue))) +
  ggplot2::geom_point(data = myDiff25p, ggplot2::aes(x = meth.diff, 
                                                  y = -log10(qvalue)),
                      colour = "red") +
  
  ggplot2::theme_minimal() +
  ggplot2::theme(
    axis.text = ggplot2::element_text(size = 15),
    axis.title.x = ggplot2::element_text(size = 20),
    axis.title.y = ggplot2::element_text(size = 20),
    legend.text = ggplot2::element_text(size = 15)
  )
print(volcano.plot)

ggplot2::ggsave(volcano.plot, bg = "white",
                filename =
                  paste0(figure.path,
                         "Toothfish_2024_muscle_RANDOM_DMS25_VOLCANO_methylkit.png"), 
                width = 30, height = 15, units = "cm")



myDiff25p.gr <- as(myDiff25p, "GRanges")
grand.plot <- ggbio::plotGrandLinear(myDiff25p.gr,
                                     ggplot2::aes(y = meth.diff)) +
  ggplot2::ylab("Methylation difference (%)") +
  ggplot2::theme_bw() +
  ggplot2::theme(axis.text.x = ggplot2::element_text(angle = 90, hjust = 0), 
                 legend.position = "none")
print(grand.plot)
ggplot2::ggsave(grand.plot@ggplot, bg = "white",
                filename = paste0(figure.path,
                                  "Toothfish_2024_muscle_RANDOM_DMS25_Grandplot_methylkit_regions.png"
                ), 
                width = 30, height = 15, units = "cm")
```


### Heatmap

```{r}
print(load(paste0(rdata.path,"Toothfish_2024_muscle_methylation_RANDOM_DMS.Rdata")))

myDiff25p.gr <- as(myDiff25p, "GRanges")
meth.diff <- methylKit::selectByOverlap(meth.2024M.norm15L, myDiff25p.gr)

meth.diff.df <- as.data.frame(t(methylKit::percMethylation(meth.diff)))
loc.names <- paste0(meth.diff$chr,"_",meth.diff$start)
colnames(meth.diff.df) <- loc.names

meth.diff.df$individual <- rownames(meth.diff.df)
meth.diff.df$pop <- strata.random
meth.diff.df$pop2 <- as.character(metadata.new$pop)


heat.plot.data <- meth.diff.df %>%
  tibble::as_tibble() %>%
  tidyr::gather(key = "CpG", value = "Methylation_percentage", -individual, -pop, -pop2,
                factor_key = FALSE) %>%
  dplyr::group_by(pop, CpG) %>%
  dplyr::mutate(CpG_mean = mean(Methylation_percentage, na.rm = TRUE)) %>%
  dplyr::arrange(pop, individual, CpG) %>%
  dplyr::mutate(individual = factor(individual, levels = unique(individual))) %>%
  dplyr::mutate(CpG = factor(CpG, levels = unique(CpG)))
cpgs <- heat.plot.data$CpG[heat.plot.data$individual == unique(heat.plot.data$individual)[1]]
cpgs_mean <- heat.plot.data$CpG_mean[heat.plot.data$individual == unique(heat.plot.data$individual)[1]]

cpgs <- cpgs[order(cpgs_mean, decreasing = TRUE)]
heat.plot.data$CpG <- factor(heat.plot.data$CpG, levels = c("pop", unique(as.character(cpgs))))
heat.plot.data <- heat.plot.data[order(heat.plot.data$CpG),]

heat.plot <- ggplot2::ggplot(data = heat.plot.data, ggplot2::aes(x = individual)) + 
    ggplot2::geom_tile(data = heat.plot.data, 
                       ggplot2::aes(x = individual, y = CpG, fill = Methylation_percentage)) +
    ggplot2::scale_fill_gradient2(low = "blue", mid = "yellow", high = "red", 
                                  midpoint = 50, guide = "colourbar") +
    ggnewscale::new_scale_fill() +
    ggplot2::geom_tile(data = heat.plot.data, ggplot2::aes(x = individual, 
                                                           y = "pop - random", fill = pop,
                                    height = length(unique(heat.plot.data$CpG))*0.02)) +
    ggplot2::scale_fill_manual(values = colours2) +
    ggnewscale::new_scale_fill() +
    ggplot2::geom_tile(data = heat.plot.data, ggplot2::aes(x = individual, y = "pop", fill = pop2,
                                    height = length(unique(heat.plot.data$CpG))*0.02)) +
    ggplot2::scale_fill_manual(values = colours2) +
    ggplot2::scale_y_discrete(drop = FALSE) +
    ggplot2::theme(legend.position = "top",
                   axis.title.y = ggplot2::element_blank(),
                   axis.title.x = ggplot2::element_blank(),
                   axis.text.x = ggplot2::element_text(face = "bold", color = "black", 
                                        size = 10, angle = 90, hjust = 1, vjust = 0.5))
print(heat.plot)

ggplot2::ggsave(heat.plot, bg = "white",
                filename = paste0(figure.path,
                                  "Toothfish_2024_muscle_RANDOM_methDiff_25_HEATMAP.png"), 
                width = 30, height = 30, units = "cm")
```

### Plot CpG difference

```{r}
load(file = paste0(rdata.path,"Toothfish_2024_muscle_methylation_RANDOM_DMS.Rdata"))

myDiff25p.gr <- as(myDiff25p, "GRanges")
meth.diff <- methylKit::selectByOverlap(meth.2024M.norm15L, myDiff25p.gr) #113 

length(unique(meth.diff$chr)) #25
sort(table(meth.diff$chr), decreasing = TRUE)[1:20]

meth.diff.df <- as.data.frame(t(methylKit::percMethylation(meth.diff)))
loc.names <- paste0(meth.diff$chr,"_",meth.diff$start)
colnames(meth.diff.df) <- loc.names

meth.diff.df$individual <- rownames(meth.diff.df) 
identical(metadata.new$Genotype_names, meth.diff.df$individual)
meth.diff.df$pop <- metadata.new$pop


meth.diff.df.plot <- tidyr::gather(meth.diff.df, key = locus, 
                           value = perc_methylation, -individual, -pop,
                           factor_key = TRUE) %>%
  dplyr::mutate(chr = stringr::str_split_fixed(locus, 
                                               pattern = "_", 
                                               n = 2)[,1])
meth.diff.df.plot.locus <- meth.diff.df.plot %>% dplyr::group_by(chr,locus,pop) %>% 
  dplyr::summarise(mean_pop = mean(perc_methylation, na.rm = TRUE),
                   sum_NA = sum(is.na(perc_methylation)),
            .groups = 'drop')


diff.plot <- ggplot2::ggplot() +
  ggplot2::geom_line(data = meth.diff.df.plot.locus,
                     ggplot2::aes(x = pop, y = mean_pop, colour = locus, group = locus)) +
  ggplot2::geom_point(data = meth.diff.df.plot.locus,
                     ggplot2::aes(x = pop, y = mean_pop, colour = locus, group = locus),
                     size = 5) +
  ggplot2::facet_wrap(~chr) +
  ggplot2::ylab("Mean methylation (%)") +
  ggplot2::theme(legend.position = "none",
                   axis.text.x = ggplot2::element_text(face = "bold", color = "black", 
                                        size = 10, angle = 90, hjust = 1, vjust = 0.5))
    # ggplot2::guides(colour = ggplot2::guide_legend(ncol = 2))

print(diff.plot)

ggplot2::ggsave(plot = diff.plot, 
                filename = paste0(figure.path,"1_Toothfish_2024_muscle_RANDOM_methDiff.png"),
                width = 30, height = 15, units = "cm", device = "png")
```


### DMS Clustering
```{r}
load(file = paste0(rdata.path,"Toothfish_2024_muscle_methylation_RANDOM_DMS.Rdata"))

myDiff25p.gr <- as(myDiff25p, "GRanges")
meth.diff <- methylKit::selectByOverlap(meth.2024M.norm15L, myDiff25p.gr) #270  
methdif.perc <- methylKit::percMethylation(meth.diff)

methdif.perc2 <- zoo::na.aggregate(methdif.perc, by = meth.2024M.norm15L@treatment)
PCobj <- prcomp(methdif.perc2, retx = TRUE, scale. = TRUE) # Too many missing values

var_frac <- PCobj$sdev/sum(PCobj$sdev)
print(paste0("Total variance explained: ",signif(sum(var_frac[1:3]) * 100, 3), "%")) 
barplot(var_frac[1:3], main = "variance explained", 
        col = heat.colors(length(var_frac[1:3])))


identical(metadata.new$Genotype_names, meth.2024M.norm15L@sample.ids)

identical(as.numeric(factor(metadata.new$pop)), meth.2024M.norm15L@treatment)

# identical(metadata.new$individual, rownames(PCobj$x))
identical(metadata.new$Genotype_names, rownames(PCobj$rotation))
pop <- metadata.new$pop

plot.title <- "Toothfish methylation - RANDOMISED 2024 Muscle (129 DMS sites - 25%)"

# data <- data.frame(ind = rownames(PCobj$x),PCobj$x,  pop = pop)
data <- data.frame(ind = rownames(PCobj$rotation),PCobj$rotation,  pop = pop)
data <- merge(data, aggregate(cbind(mean.x = PC1, mean.y = PC2) ~
                                pop, data, mean), by = "pop")

pca.plot <- ggplot2::ggplot(data, ggplot2::aes(PC1, PC2, color = pop)) +
  ggplot2::geom_point(size = 4) + 
  ggplot2::geom_point(ggplot2::aes(x = mean.x,
                                   y = mean.y, color = pop),
                      size = 7,
                      shape = 17) +
  ggplot2::scale_colour_manual(values = colours2) +
  ggplot2::geom_hline(yintercept = 0) +
  ggplot2::geom_vline(xintercept = 0) + 
  ggplot2::labs(
    subtitle = paste0("Total variance explained: ",
                      signif(sum(var_frac[1:3]) * 100, 3), "%"),
    y = paste0("PC2; ", signif(var_frac[2] * 100, digits = 2), " %"),
    x = paste0("PC1; ", signif(var_frac[1] * 100, digits = 2), " %"),
    title = plot.title,
    caption = "") + 
  ggplot2::theme_minimal() +
  ggplot2::theme(
    axis.text = ggplot2::element_text(size = 15),
    axis.title.x = ggplot2::element_text(size = 20),
    axis.title.y = ggplot2::element_text(size = 20),
    legend.text = ggplot2::element_text(size = 15)
  )

print(pca.plot)
ggplot2::ggsave(pca.plot, bg = "white",
                filename = 
                  paste0(figure.path,"Toothfish_2024_muscle_RANDOM_PCA_CpG_DMS_PC12.png"), 
                width = 30, height = 15, units = "cm")



# data <- data.frame(ind = rownames(PCobj$x),PCobj$x,  pop = pop)
data <- data.frame(ind = rownames(PCobj$rotation),PCobj$rotation,  pop = pop)
data <- merge(data, aggregate(cbind(mean.x = PC1, mean.y = PC3) ~
                                pop, data, mean), by = "pop")

pca.plot <- ggplot2::ggplot(data, ggplot2::aes(PC1, PC3, color = pop)) +
  ggplot2::geom_point(size = 4) + 
  ggplot2::geom_point(ggplot2::aes(x = mean.x,
                                   y = mean.y, color = pop),
                      size = 7,
                      shape = 17) +
  ggplot2::scale_colour_manual(values = colours2) +
  ggplot2::geom_hline(yintercept = 0) +
  ggplot2::geom_vline(xintercept = 0) + 
  ggplot2::labs(
    subtitle = paste0("Total variance explained: ",
                      signif(sum(var_frac[1:3]) * 100, 3), "%"),
    y = paste0("PC3; ", signif(var_frac[3] * 100, digits = 2), " %"),
    x = paste0("PC1; ", signif(var_frac[1] * 100, digits = 2), " %"),
    title = plot.title,
    caption = "") + 
  ggplot2::theme_minimal() +
  ggplot2::theme(
    axis.text = ggplot2::element_text(size = 15),
    axis.title.x = ggplot2::element_text(size = 20),
    axis.title.y = ggplot2::element_text(size = 20),
    legend.text = ggplot2::element_text(size = 15)
  )

print(pca.plot)
ggplot2::ggsave(pca.plot, bg = "white",
                filename = paste0(figure.path,"Toothfish_2024_muscle_RANDOM_PCA_CpG_DMS_PC13.png"), 
                width = 30, height = 15, units = "cm")




# data <- data.frame(ind = rownames(PCobj$x),PCobj$x,  pop = pop)
data <- data.frame(ind = rownames(PCobj$rotation),PCobj$rotation,  pop = pop)
data <- merge(data, aggregate(cbind(mean.x = PC2, mean.y = PC3) ~
                                pop, data, mean), by = "pop")

pca.plot <- ggplot2::ggplot(data, ggplot2::aes(PC2, PC3, color = pop)) +
  ggplot2::geom_point(size = 4) + 
  ggplot2::geom_point(ggplot2::aes(x = mean.x,
                                   y = mean.y, color = pop),
                      size = 7,
                      shape = 17) +
  ggplot2::scale_colour_manual(values = colours2) +
  ggplot2::geom_hline(yintercept = 0) +
  ggplot2::geom_vline(xintercept = 0) + 
  ggplot2::labs(
    subtitle = paste0("Total variance explained: ",
                      signif(sum(var_frac[1:3]) * 100, 3), "%"),
    y = paste0("PC3; ", signif(var_frac[3] * 100, digits = 2), " %"),
    x = paste0("PC2; ", signif(var_frac[2] * 100, digits = 2), " %"),
    title = plot.title,
    caption = "") + 
  ggplot2::theme_minimal() +
  ggplot2::theme(
    axis.text = ggplot2::element_text(size = 15),
    axis.title.x = ggplot2::element_text(size = 20),
    axis.title.y = ggplot2::element_text(size = 20),
    legend.text = ggplot2::element_text(size = 15)
  )

print(pca.plot)
ggplot2::ggsave(pca.plot, bg = "white",
                filename = 
                  paste0(figure.path,"Toothfish_2024_muscle_RANDOM_PCA_CpG_DMS_PC23.png"), 
                width = 30, height = 15, units = "cm")
```


## TOP100 DMS - 2024 - muscle

```{r}
print(load(paste0(rdata.path,"Toothfish_2024_muscle_methylkit_object.Rdata")))
print(load(paste0(rdata.path,"Toothfish_metadata.Rdata"))) 
metadata.new <- metadata[metadata$Genotype_names %in% meth.2024M.norm15L@sample.ids,]
metadata.new <- metadata.new[order(match(metadata.new$Genotype_names,
                                         meth.2024M.norm15L@sample.ids)),]
meth.cpg <- paste0(meth.2024M.norm15L$chr, "_", meth.2024M.norm15L$start) #601560 
length(meth.cpg)
print(load(paste0(rdata.path,"Toothfish_2024_muscle_methylation_DMS.Rdata")))
DMS.cpg <- paste0(as.character(top100.2024M@seqnames), "_", as.data.frame(top100.2024M@ranges)$start) #283819 
sum(meth.cpg %in% DMS.cpg) #100
meth.DMS <- meth.2024M.norm15L[meth.cpg %in% DMS.cpg,]
```


### Heatmap

```{r}
meth.diff.df <- as.data.frame(t(methylKit::percMethylation(meth.DMS)))
loc.names <- paste0(meth.DMS$chr,"_",meth.DMS$start)
colnames(meth.diff.df) <- loc.names

meth.diff.df$individual <- rownames(meth.diff.df)
pop <- metadata.new$pop
meth.diff.df$pop <- as.character(pop)


heat.plot.data <- meth.diff.df %>%
  tibble::as_tibble() %>%
  tidyr::gather(key = "CpG", value = "Methylation_percentage", -individual, -pop,
                factor_key = FALSE) %>%
  dplyr::group_by(pop, CpG) %>%
  dplyr::mutate(CpG_mean = mean(Methylation_percentage, na.rm = TRUE)) %>%
  dplyr::arrange(pop, individual, CpG) %>%
  dplyr::mutate(individual = factor(individual, levels = unique(individual))) %>%
  dplyr::mutate(CpG = factor(CpG, levels = unique(CpG)))
cpgs <- heat.plot.data$CpG[heat.plot.data$individual == unique(heat.plot.data$individual)[1]]
cpgs_mean <- heat.plot.data$CpG_mean[heat.plot.data$individual == unique(heat.plot.data$individual)[1]]

cpgs <- cpgs[order(cpgs_mean, decreasing = TRUE)]
heat.plot.data$CpG <- factor(heat.plot.data$CpG, levels = c("pop", unique(as.character(cpgs))))
heat.plot.data <- heat.plot.data[order(heat.plot.data$CpG),]

heat.plot <- ggplot2::ggplot(data = heat.plot.data, ggplot2::aes(x = individual)) + 
    ggplot2::geom_tile(data = heat.plot.data, 
                       ggplot2::aes(x = individual, y = CpG, fill = Methylation_percentage)) +
    ggplot2::scale_fill_gradient2(low = "blue", mid = "yellow", high = "red", 
                                  midpoint = 50, guide = "colourbar") +
    ggnewscale::new_scale_fill() +
    ggplot2::geom_tile(data = heat.plot.data, ggplot2::aes(x = individual, y = "pop", fill = pop,
                                    height = length(unique(heat.plot.data$CpG))*0.02)) +
    ggplot2::scale_fill_manual(values = colours2) +
    ggplot2::scale_y_discrete(drop = FALSE) +
    ggplot2::theme(legend.position = "top",
                   axis.title.y = ggplot2::element_blank(),
                   axis.title.x = ggplot2::element_blank(),
                   axis.text.x = ggplot2::element_text(face = "bold", color = "black", 
                                        size = 10, angle = 90, hjust = 1, vjust = 0.5))
print(heat.plot)

ggplot2::ggsave(heat.plot, bg = "white",
                filename = paste0(figure.path,
                                  "Toothfish_2024_muscle_BASED_ON_2024_TOP100_DMS_HEATMAP.png"), 
                width = 30, height = 30, units = "cm")
```

### Plot CpG difference

```{r}
meth.diff <- meth.DMS

length(unique(meth.diff$chr)) #28
sort(table(meth.diff$chr), decreasing = TRUE)[1:20]

meth.diff.df <- as.data.frame(t(methylKit::percMethylation(meth.diff)))
loc.names <- paste0(meth.diff$chr,"_",meth.diff$start)
colnames(meth.diff.df) <- loc.names

meth.diff.df$individual <- rownames(meth.diff.df) 
identical(metadata.new$Genotype_names, meth.diff.df$individual)
meth.diff.df$pop <- metadata.new$pop



meth.diff.df.plot <- tidyr::gather(meth.diff.df, key = locus, 
                           value = perc_methylation, -individual, -pop,
                           factor_key = TRUE) %>%
  dplyr::mutate(chr = stringr::str_split_fixed(locus, 
                                               pattern = "_", 
                                               n = 2)[,1])
meth.diff.df.plot.locus <- meth.diff.df.plot %>% dplyr::group_by(chr,locus,pop) %>% 
  dplyr::summarise(mean_pop = mean(perc_methylation, na.rm = TRUE),
                   sum_NA = sum(is.na(perc_methylation)),
            .groups = 'drop')


diff.plot <- ggplot2::ggplot() +
  ggplot2::geom_line(data = meth.diff.df.plot.locus,
                     ggplot2::aes(x = pop, y = mean_pop, colour = locus, group = locus)) +
  ggplot2::geom_point(data = meth.diff.df.plot.locus,
                     ggplot2::aes(x = pop, y = mean_pop, colour = locus, group = locus),
                     size = 5) +
  ggplot2::facet_wrap(~chr) +
  ggplot2::ylab("Mean methylation (%)") +
  ggplot2::theme(legend.position = "none",
                   axis.text.x = ggplot2::element_text(face = "bold", color = "black", 
                                        size = 10, angle = 90, hjust = 1, vjust = 0.5))

print(diff.plot)

ggplot2::ggsave(plot = diff.plot, 
                filename = paste0(figure.path,"1_Toothfish_2024_muscle_BASED_ON_2024_TOP100_DMS_methDiff.png"),
                width = 30, height = 15, units = "cm", device = "png")
```



### DMS Clustering
```{r}
methdif.perc <- methylKit::percMethylation(meth.DMS)
methdif.perc2 <- zoo::na.aggregate(methdif.perc, by = meth.DMS@treatment)
PCobj <- prcomp(methdif.perc2, retx = TRUE, scale. = TRUE, center = TRUE)

var_frac <- PCobj$sdev/sum(PCobj$sdev)
print(paste0("Total variance explained: ",signif(sum(var_frac[1:3]) * 100, 3), "%")) 
barplot(var_frac[1:3], main = "variance explained", 
        col = heat.colors(length(var_frac[1:3])))

# 
# identical(metadata.new$Genotype_names, meth.DMS@sample.ids)
# identical(metadata.new$Genotype_names, rownames(PCobj$rotation))
identical(metadata.new$File_names, meth.DMS@sample.ids)
identical(metadata.new$File_names, rownames(PCobj$rotation))
pop <- metadata.new$pop

plot.title <- "Toothfish methylation - 2024 Muscle (TOP100 DMS from 2024)"

data <- data.frame(ind = rownames(PCobj$rotation),PCobj$rotation,  pop = pop)
data <- merge(data, aggregate(cbind(mean.x = PC1, mean.y = PC2) ~
                                pop, data, mean), by = "pop")

pca.plot <- ggplot2::ggplot(data, ggplot2::aes(PC1, PC2, color = pop)) +
  ggplot2::geom_point(size = 4) + 
  ggplot2::geom_point(ggplot2::aes(x = mean.x,
                                   y = mean.y, color = pop),
                      size = 7,
                      shape = 17) +
  ggplot2::scale_colour_manual(values = colours2) +
  ggplot2::geom_hline(yintercept = 0) +
  ggplot2::geom_vline(xintercept = 0) + 
  ggplot2::labs(
    subtitle = paste0("Total variance explained: ",
                      signif(sum(var_frac[1:3]) * 100, 3), "%"),
    y = paste0("PC2; ", signif(var_frac[2] * 100, digits = 2), " %"),
    x = paste0("PC1; ", signif(var_frac[1] * 100, digits = 2), " %"),
    title = plot.title,
    caption = "") + 
  ggplot2::theme_minimal() +
  ggplot2::theme(
    axis.text = ggplot2::element_text(size = 15),
    axis.title.x = ggplot2::element_text(size = 20),
    axis.title.y = ggplot2::element_text(size = 20),
    legend.text = ggplot2::element_text(size = 15)
  )

print(pca.plot)
ggplot2::ggsave(pca.plot, bg = "white",
                filename = paste0(figure.path,"Toothfish_2024_muscle_BASED_ON_2024_TOP100_DMS_PC12.png"), 
                width = 30, height = 15, units = "cm")



data <- data.frame(ind = rownames(PCobj$rotation),PCobj$rotation,  pop = pop)
data <- merge(data, aggregate(cbind(mean.x = PC1, mean.y = PC3) ~
                                pop, data, mean), by = "pop")

pca.plot <- ggplot2::ggplot(data, ggplot2::aes(PC1, PC3, color = pop)) +
  ggplot2::geom_point(size = 4) + 
  ggplot2::geom_point(ggplot2::aes(x = mean.x,
                                   y = mean.y, color = pop),
                      size = 7,
                      shape = 17) +
  ggplot2::scale_colour_manual(values = colours2) +
  ggplot2::geom_hline(yintercept = 0) +
  ggplot2::geom_vline(xintercept = 0) + 
  ggplot2::labs(
    subtitle = paste0("Total variance explained: ",
                      signif(sum(var_frac[1:3]) * 100, 3), "%"),
    y = paste0("PC3; ", signif(var_frac[3] * 100, digits = 2), " %"),
    x = paste0("PC1; ", signif(var_frac[1] * 100, digits = 2), " %"),
    title = plot.title,
    caption = "") + 
  ggplot2::theme_minimal() +
  ggplot2::theme(
    axis.text = ggplot2::element_text(size = 15),
    axis.title.x = ggplot2::element_text(size = 20),
    axis.title.y = ggplot2::element_text(size = 20),
    legend.text = ggplot2::element_text(size = 15)
  )

print(pca.plot)
ggplot2::ggsave(pca.plot, bg = "white",
                filename = paste0(figure.path,
                                  "Toothfish_2024_muscle_BASED_ON_2024_TOP100_DMS_PC13.png"), 
                width = 30, height = 15, units = "cm")




data <- data.frame(ind = rownames(PCobj$rotation),PCobj$rotation,  pop = pop)
data <- merge(data, aggregate(cbind(mean.x = PC2, mean.y = PC3) ~
                                pop, data, mean), by = "pop")

pca.plot <- ggplot2::ggplot(data, ggplot2::aes(PC2, PC3, color = pop)) +
  ggplot2::geom_point(size = 4) + 
  ggplot2::geom_point(ggplot2::aes(x = mean.x,
                                   y = mean.y, color = pop),
                      size = 7,
                      shape = 17) +
  ggplot2::scale_colour_manual(values = colours2) +
  ggplot2::geom_hline(yintercept = 0) +
  ggplot2::geom_vline(xintercept = 0) + 
  ggplot2::labs(
    subtitle = paste0("Total variance explained: ",
                      signif(sum(var_frac[1:3]) * 100, 3), "%"),
    y = paste0("PC3; ", signif(var_frac[3] * 100, digits = 2), " %"),
    x = paste0("PC2; ", signif(var_frac[2] * 100, digits = 2), " %"),
    title = plot.title,
    caption = "") + 
  ggplot2::theme_minimal() +
  ggplot2::theme(
    axis.text = ggplot2::element_text(size = 15),
    axis.title.x = ggplot2::element_text(size = 20),
    axis.title.y = ggplot2::element_text(size = 20),
    legend.text = ggplot2::element_text(size = 15)
  )

print(pca.plot)
ggplot2::ggsave(pca.plot, bg = "white",
                filename = paste0(figure.path,"Toothfish_2024_muscle_BASED_ON_2024_TOP100_DMS_PC23.png"), 
                width = 30, height = 15, units = "cm")
```





## 2024 fish muscle with DMS from 2017/2018
```{r}
print(load(paste0(rdata.path,"Toothfish_DMS_CpGsites.Rdata")))
print(load(paste0(rdata.path,"Toothfish_2018_muscle_methylation_DMS.Rdata")))

meth.cpg <- paste0(meth.2024M.norm15L$chr, "_", meth.2024M.norm15L$start) #601560 
DMS.cpg <- unique(c(DMS$DMS)) #1243
sum(meth.cpg %in% DMS.cpg) #516 = 41%

DMS.cpg <- unique(c(DMS.ER$DMS, DMS.EANT$DMS)) #74
sum(meth.cpg %in% DMS.cpg) #47 = 63%

meth.DMS <- meth.2024M.norm15L[meth.cpg %in% DMS.cpg,] 
```

### Heatmap

```{r}
meth.diff.df <- as.data.frame(t(methylKit::percMethylation(meth.DMS)))
loc.names <- paste0(meth.DMS$chr,"_",meth.DMS$start)
colnames(meth.diff.df) <- loc.names

meth.diff.df$individual <- rownames(meth.diff.df)
pop <- metadata.new$pop
meth.diff.df$pop <- as.character(pop)


heat.plot.data <- meth.diff.df %>%
  tibble::as_tibble() %>%
  tidyr::gather(key = "CpG", value = "Methylation_percentage", -individual, -pop,
                factor_key = FALSE) %>%
  dplyr::group_by(pop, CpG) %>%
  dplyr::mutate(CpG_mean = mean(Methylation_percentage, na.rm = TRUE)) %>%
  dplyr::arrange(pop, individual, CpG) %>%
  dplyr::mutate(individual = factor(individual, levels = unique(individual))) %>%
  dplyr::mutate(CpG = factor(CpG, levels = unique(CpG)))
cpgs <- heat.plot.data$CpG[heat.plot.data$individual == unique(heat.plot.data$individual)[1]]
cpgs_mean <- heat.plot.data$CpG_mean[heat.plot.data$individual == unique(heat.plot.data$individual)[1]]

cpgs <- cpgs[order(cpgs_mean, decreasing = TRUE)]
heat.plot.data$CpG <- factor(heat.plot.data$CpG, levels = c("pop", unique(as.character(cpgs))))
heat.plot.data <- heat.plot.data[order(heat.plot.data$CpG),]

heat.plot <- ggplot2::ggplot(data = heat.plot.data, ggplot2::aes(x = individual)) + 
    ggplot2::geom_tile(data = heat.plot.data, 
                       ggplot2::aes(x = individual, y = CpG, fill = Methylation_percentage)) +
    ggplot2::scale_fill_gradient2(low = "blue", mid = "yellow", high = "red", 
                                  midpoint = 50, guide = "colourbar") +
    ggnewscale::new_scale_fill() +
    ggplot2::geom_tile(data = heat.plot.data, ggplot2::aes(x = individual, y = "pop", fill = pop,
                                    height = length(unique(heat.plot.data$CpG))*0.02)) +
    ggplot2::scale_fill_manual(values = colours2) +
    ggplot2::scale_y_discrete(drop = FALSE) +
    ggplot2::theme(legend.position = "top",
                   axis.title.y = ggplot2::element_blank(),
                   axis.title.x = ggplot2::element_blank(),
                   axis.text.x = ggplot2::element_text(face = "bold", color = "black", 
                                        size = 10, angle = 90, hjust = 1, vjust = 0.5))
print(heat.plot)

ggplot2::ggsave(heat.plot, bg = "white",
                filename = paste0(figure.path,
                                  "Toothfish_2024_muscle_BASED_ON_2018_DMS_HEATMAP.png"), 
                width = 30, height = 30, units = "cm")
```


### DMS Clustering
```{r}
methdif.perc <- methylKit::percMethylation(meth.DMS)
methdif.perc2 <- zoo::na.aggregate(methdif.perc, by = meth.DMS@treatment)
PCobj <- prcomp(methdif.perc2, retx = TRUE, scale. = TRUE)

var_frac <- PCobj$sdev/sum(PCobj$sdev)
print(paste0("Total variance explained: ",signif(sum(var_frac[1:3]) * 100, 3), "%")) 
barplot(var_frac[1:3], main = "variance explained", 
        col = heat.colors(length(var_frac[1:3])))


identical(metadata.new$Genotype_names, meth.DMS@sample.ids)
# identical(metadata.new$individual, rownames(PCobj$x))
identical(metadata.new$Genotype_names, rownames(PCobj$rotation))
pop <- metadata.new$pop

plot.title <- "Toothfish methylation - 2024 Muscle (516 DMS from 2018 - 25%)"

# data <- data.frame(ind = rownames(PCobj$x),PCobj$x,  pop = pop)
data <- data.frame(ind = rownames(PCobj$rotation),PCobj$rotation,  pop = pop)
data <- merge(data, aggregate(cbind(mean.x = PC1, mean.y = PC2) ~
                                pop, data, mean), by = "pop")

pca.plot <- ggplot2::ggplot(data, ggplot2::aes(PC1, PC2, color = pop)) +
  ggplot2::geom_point(size = 4) + 
  ggplot2::geom_point(ggplot2::aes(x = mean.x,
                                   y = mean.y, color = pop),
                      size = 7,
                      shape = 17) +
  ggplot2::scale_colour_manual(values = colours2) +
  ggplot2::geom_hline(yintercept = 0) +
  ggplot2::geom_vline(xintercept = 0) + 
  ggplot2::labs(
    subtitle = paste0("Total variance explained: ",
                      signif(sum(var_frac[1:3]) * 100, 3), "%"),
    y = paste0("PC2; ", signif(var_frac[2] * 100, digits = 2), " %"),
    x = paste0("PC1; ", signif(var_frac[1] * 100, digits = 2), " %"),
    title = plot.title,
    caption = "") + 
  ggplot2::theme_minimal() +
  ggplot2::theme(
    axis.text = ggplot2::element_text(size = 15),
    axis.title.x = ggplot2::element_text(size = 20),
    axis.title.y = ggplot2::element_text(size = 20),
    legend.text = ggplot2::element_text(size = 15)
  )

print(pca.plot)
ggplot2::ggsave(pca.plot, bg = "white",
                filename = paste0(figure.path,"Toothfish_2024_muscle_PCA_CpG_DMS2018_PC12.png"), 
                width = 30, height = 15, units = "cm")



# data <- data.frame(ind = rownames(PCobj$x),PCobj$x,  pop = pop)
data <- data.frame(ind = rownames(PCobj$rotation),PCobj$rotation,  pop = pop)
data <- merge(data, aggregate(cbind(mean.x = PC1, mean.y = PC3) ~
                                pop, data, mean), by = "pop")

pca.plot <- ggplot2::ggplot(data, ggplot2::aes(PC1, PC3, color = pop)) +
  ggplot2::geom_point(size = 4) + 
  ggplot2::geom_point(ggplot2::aes(x = mean.x,
                                   y = mean.y, color = pop),
                      size = 7,
                      shape = 17) +
  ggplot2::scale_colour_manual(values = colours2) +
  ggplot2::geom_hline(yintercept = 0) +
  ggplot2::geom_vline(xintercept = 0) + 
  ggplot2::labs(
    subtitle = paste0("Total variance explained: ",
                      signif(sum(var_frac[1:3]) * 100, 3), "%"),
    y = paste0("PC3; ", signif(var_frac[3] * 100, digits = 2), " %"),
    x = paste0("PC1; ", signif(var_frac[1] * 100, digits = 2), " %"),
    title = plot.title,
    caption = "") + 
  ggplot2::theme_minimal() +
  ggplot2::theme(
    axis.text = ggplot2::element_text(size = 15),
    axis.title.x = ggplot2::element_text(size = 20),
    axis.title.y = ggplot2::element_text(size = 20),
    legend.text = ggplot2::element_text(size = 15)
  )

print(pca.plot)
ggplot2::ggsave(pca.plot, bg = "white",
                filename = paste0(figure.path,
                                  "Toothfish_2024_muscle_PCA_CpG_DMS2018_PC13.png"), 
                width = 30, height = 15, units = "cm")




# data <- data.frame(ind = rownames(PCobj$x),PCobj$x,  pop = pop)
data <- data.frame(ind = rownames(PCobj$rotation),PCobj$rotation,  pop = pop)
data <- merge(data, aggregate(cbind(mean.x = PC2, mean.y = PC3) ~
                                pop, data, mean), by = "pop")

pca.plot <- ggplot2::ggplot(data, ggplot2::aes(PC2, PC3, color = pop)) +
  ggplot2::geom_point(size = 4) + 
  ggplot2::geom_point(ggplot2::aes(x = mean.x,
                                   y = mean.y, color = pop),
                      size = 7,
                      shape = 17) +
  ggplot2::scale_colour_manual(values = colours2) +
  ggplot2::geom_hline(yintercept = 0) +
  ggplot2::geom_vline(xintercept = 0) + 
  ggplot2::labs(
    subtitle = paste0("Total variance explained: ",
                      signif(sum(var_frac[1:3]) * 100, 3), "%"),
    y = paste0("PC3; ", signif(var_frac[3] * 100, digits = 2), " %"),
    x = paste0("PC2; ", signif(var_frac[2] * 100, digits = 2), " %"),
    title = plot.title,
    caption = "") + 
  ggplot2::theme_minimal() +
  ggplot2::theme(
    axis.text = ggplot2::element_text(size = 15),
    axis.title.x = ggplot2::element_text(size = 20),
    axis.title.y = ggplot2::element_text(size = 20),
    legend.text = ggplot2::element_text(size = 15)
  )

print(pca.plot)
ggplot2::ggsave(pca.plot, bg = "white",
                filename = paste0(figure.path,"Toothfish_2024_muscle_PCA_CpG_DMS2018_PC23.png"), 
                width = 30, height = 15, units = "cm")
```


## 2018 fish muscle with DMS from 2024

```{r}
print(load(paste0(rdata.path,"Toothfish_2018_muscle_methylkit_object.Rdata")))
# print(load(paste0(rdata.path,"Toothfish_subset_methylation_objects_raw_NEW.Rdata")))
# meth.2018M.norm11L <- meth.sub

print(load(paste0(rdata.path,"Toothfish_metadata.Rdata"))) 
metadata.new <- metadata[metadata$Genotype_names %in% meth.2018M.norm11L@sample.ids,]
metadata.new <- metadata.new[order(match(metadata.new$Genotype_names,
                                         meth.2018M.norm11L@sample.ids)),]
# metadata.new <- metadata.new[order(match(metadata.new$Genotype_names,
#                                          meth.2018M.norm11L@sample.ids)),]
print(load(paste0(rdata.path,"Toothfish_2024_muscle_methylation_DMS.Rdata")))
meth.cpg <- paste0(meth.2018M.norm11L$chr, "_", meth.2018M.norm11L$start) #1308608 
DMS.cpg <- paste0(myDiff25p$chr, "_", myDiff25p$start) #1211
sum(meth.cpg %in% DMS.cpg) #917 = 75%
meth.DMS <- meth.2018M.norm11L[meth.cpg %in% DMS.cpg,]
```



### Heatmap

```{r}
meth.diff.df <- as.data.frame(t(methylKit::percMethylation(meth.DMS)))
loc.names <- paste0(meth.DMS$chr,"_",meth.DMS$start)
colnames(meth.diff.df) <- loc.names

meth.diff.df$individual <- rownames(meth.diff.df)
pop <- metadata.new$pop
meth.diff.df$pop <- as.character(pop)


heat.plot.data <- meth.diff.df %>%
  tibble::as_tibble() %>%
  tidyr::gather(key = "CpG", value = "Methylation_percentage", -individual, -pop,
                factor_key = FALSE) %>%
  dplyr::group_by(pop, CpG) %>%
  dplyr::mutate(CpG_mean = mean(Methylation_percentage, na.rm = TRUE)) %>%
  dplyr::arrange(pop, individual, CpG) %>%
  dplyr::mutate(individual = factor(individual, levels = unique(individual))) %>%
  dplyr::mutate(CpG = factor(CpG, levels = unique(CpG)))
cpgs <- heat.plot.data$CpG[heat.plot.data$individual == unique(heat.plot.data$individual)[1]]
cpgs_mean <- heat.plot.data$CpG_mean[heat.plot.data$individual == unique(heat.plot.data$individual)[1]]

cpgs <- cpgs[order(cpgs_mean, decreasing = TRUE)]
heat.plot.data$CpG <- factor(heat.plot.data$CpG, levels = c("pop", unique(as.character(cpgs))))
heat.plot.data <- heat.plot.data[order(heat.plot.data$CpG),]

heat.plot <- ggplot2::ggplot(data = heat.plot.data, ggplot2::aes(x = individual)) + 
    ggplot2::geom_tile(data = heat.plot.data, 
                       ggplot2::aes(x = individual, y = CpG, fill = Methylation_percentage)) +
    ggplot2::scale_fill_gradient2(low = "blue", mid = "yellow", high = "red", 
                                  midpoint = 50, guide = "colourbar") +
    ggnewscale::new_scale_fill() +
    ggplot2::geom_tile(data = heat.plot.data, ggplot2::aes(x = individual, y = "pop", fill = pop,
                                    height = length(unique(heat.plot.data$CpG))*0.02)) +
    ggplot2::scale_fill_manual(values = colours4) +
    ggplot2::scale_y_discrete(drop = FALSE) +
    ggplot2::theme(legend.position = "top",
                   axis.title.y = ggplot2::element_blank(),
                   axis.title.x = ggplot2::element_blank(),
                   axis.text.x = ggplot2::element_text(face = "bold", color = "black", 
                                        size = 10, angle = 90, hjust = 1, vjust = 0.5))
print(heat.plot)

ggplot2::ggsave(heat.plot, bg = "white",
                filename = paste0(figure.path,
                                  "Toothfish_2018_muscle_BASED_ON_2024_DMS_HEATMAP.png"), 
                width = 30, height = 30, units = "cm")
```


### DMS Clustering
```{r}
methdif.perc <- methylKit::percMethylation(meth.DMS)
methdif.perc2 <- zoo::na.aggregate(methdif.perc, by = meth.DMS@treatment)
PCobj <- prcomp(methdif.perc2, retx = TRUE, scale. = TRUE)

var_frac <- PCobj$sdev/sum(PCobj$sdev)
print(paste0("Total variance explained: ",signif(sum(var_frac[1:3]) * 100, 3), "%")) 
barplot(var_frac[1:3], main = "variance explained", 
        col = heat.colors(length(var_frac[1:3])))

# 
# identical(metadata.new$Genotype_names, meth.DMS@sample.ids)
# identical(metadata.new$Genotype_names, rownames(PCobj$rotation))
identical(metadata.new$File_names, meth.DMS@sample.ids)
identical(metadata.new$File_names, rownames(PCobj$rotation))
pop <- metadata.new$pop

plot.title <- "Toothfish methylation - 2018 Muscle (917 DMS from 2024 - 25%)"

data <- data.frame(ind = rownames(PCobj$rotation),PCobj$rotation,  pop = pop)
data <- merge(data, aggregate(cbind(mean.x = PC1, mean.y = PC2) ~
                                pop, data, mean), by = "pop")

pca.plot <- ggplot2::ggplot(data, ggplot2::aes(PC1, PC2, color = pop)) +
  ggplot2::geom_point(size = 4) + 
  ggplot2::geom_point(ggplot2::aes(x = mean.x,
                                   y = mean.y, color = pop),
                      size = 7,
                      shape = 17) +
  ggplot2::scale_colour_manual(values = colours4) +
  ggplot2::geom_hline(yintercept = 0) +
  ggplot2::geom_vline(xintercept = 0) + 
  ggplot2::labs(
    subtitle = paste0("Total variance explained: ",
                      signif(sum(var_frac[1:3]) * 100, 3), "%"),
    y = paste0("PC2; ", signif(var_frac[2] * 100, digits = 2), " %"),
    x = paste0("PC1; ", signif(var_frac[1] * 100, digits = 2), " %"),
    title = plot.title,
    caption = "") + 
  ggplot2::theme_minimal() +
  ggplot2::theme(
    axis.text = ggplot2::element_text(size = 15),
    axis.title.x = ggplot2::element_text(size = 20),
    axis.title.y = ggplot2::element_text(size = 20),
    legend.text = ggplot2::element_text(size = 15)
  )

print(pca.plot)
ggplot2::ggsave(pca.plot, bg = "white",
                filename = paste0(figure.path,"Toothfish_2018_muscle_PCA_CpG_DMS2024_PC12.png"), 
                width = 30, height = 15, units = "cm")



data <- data.frame(ind = rownames(PCobj$rotation),PCobj$rotation,  pop = pop)
data <- merge(data, aggregate(cbind(mean.x = PC1, mean.y = PC3) ~
                                pop, data, mean), by = "pop")

pca.plot <- ggplot2::ggplot(data, ggplot2::aes(PC1, PC3, color = pop)) +
  ggplot2::geom_point(size = 4) + 
  ggplot2::geom_point(ggplot2::aes(x = mean.x,
                                   y = mean.y, color = pop),
                      size = 7,
                      shape = 17) +
  ggplot2::scale_colour_manual(values = colours4) +
  ggplot2::geom_hline(yintercept = 0) +
  ggplot2::geom_vline(xintercept = 0) + 
  ggplot2::labs(
    subtitle = paste0("Total variance explained: ",
                      signif(sum(var_frac[1:3]) * 100, 3), "%"),
    y = paste0("PC3; ", signif(var_frac[3] * 100, digits = 2), " %"),
    x = paste0("PC1; ", signif(var_frac[1] * 100, digits = 2), " %"),
    title = plot.title,
    caption = "") + 
  ggplot2::theme_minimal() +
  ggplot2::theme(
    axis.text = ggplot2::element_text(size = 15),
    axis.title.x = ggplot2::element_text(size = 20),
    axis.title.y = ggplot2::element_text(size = 20),
    legend.text = ggplot2::element_text(size = 15)
  )

print(pca.plot)
ggplot2::ggsave(pca.plot, bg = "white",
                filename = paste0(figure.path,
                                  "Toothfish_2018_muscle_PCA_CpG_DMS2024_PC13.png"), 
                width = 30, height = 15, units = "cm")




data <- data.frame(ind = rownames(PCobj$rotation),PCobj$rotation,  pop = pop)
data <- merge(data, aggregate(cbind(mean.x = PC2, mean.y = PC3) ~
                                pop, data, mean), by = "pop")

pca.plot <- ggplot2::ggplot(data, ggplot2::aes(PC2, PC3, color = pop)) +
  ggplot2::geom_point(size = 4) + 
  ggplot2::geom_point(ggplot2::aes(x = mean.x,
                                   y = mean.y, color = pop),
                      size = 7,
                      shape = 17) +
  ggplot2::scale_colour_manual(values = colours4) +
  ggplot2::geom_hline(yintercept = 0) +
  ggplot2::geom_vline(xintercept = 0) + 
  ggplot2::labs(
    subtitle = paste0("Total variance explained: ",
                      signif(sum(var_frac[1:3]) * 100, 3), "%"),
    y = paste0("PC3; ", signif(var_frac[3] * 100, digits = 2), " %"),
    x = paste0("PC2; ", signif(var_frac[2] * 100, digits = 2), " %"),
    title = plot.title,
    caption = "") + 
  ggplot2::theme_minimal() +
  ggplot2::theme(
    axis.text = ggplot2::element_text(size = 15),
    axis.title.x = ggplot2::element_text(size = 20),
    axis.title.y = ggplot2::element_text(size = 20),
    legend.text = ggplot2::element_text(size = 15)
  )

print(pca.plot)
ggplot2::ggsave(pca.plot, bg = "white",
                filename = paste0(figure.path,"Toothfish_2018_muscle_PCA_CpG_DMS2024_PC23.png"), 
                width = 30, height = 15, units = "cm")
```




## Re-assignment success of 2018 based on 2024 DMS

```{r}
print(load(paste0(rdata.path,"Toothfish_2018_muscle_methylkit_object.Rdata")))
print(load(paste0(rdata.path,"Toothfish_subset_methylation_objects_raw_NEW.Rdata")))
meth.2018M.norm11L <- meth.sub

print(load(paste0(rdata.path,"Toothfish_metadata.Rdata"))) 
metadata.new <- metadata[metadata$File_names %in% meth.2018M.norm11L@sample.ids,]
metadata.new <- metadata.new[order(match(metadata.new$File_names,
                                         meth.2018M.norm11L@sample.ids)),]
print(load(paste0(rdata.path,"Toothfish_2024_muscle_methylation_DMS.Rdata")))
meth.cpg <- paste0(meth.2018M.norm11L$chr, "_", meth.2018M.norm11L$start) #1037176 
# DMS.cpg <- paste0(as.character(top100.2024M@seqnames), "_", as.data.frame(top100.2024M@ranges)$start) #283819 
DMS.cpg <- paste0(myDiff25p$chr, "_", myDiff25p$start) #1448

sum(meth.cpg %in% DMS.cpg) #65 or 1448
DMS.cpg2 <- meth.cpg[meth.cpg %in% DMS.cpg]
meth.DMS1 <- meth.2018M.norm11L[meth.cpg %in% DMS.cpg,]

testTransformed <- methylKit::percMethylation(meth.DMS1)
testTransformed <- zoo::na.aggregate(testTransformed, by = meth.DMS1@treatment)
testTransformed <-  as.data.frame(t(testTransformed))
colnames(testTransformed) <- DMS.cpg2
# testTransformed$pop <- metadata.new$pop
testTransformed$pop <- metadata.new$STRATA2
testTransformed <- testTransformed[testTransformed$pop %in% c("East_Antarctica", "Ross_sea"),]

print(load(paste0(rdata.path,"Toothfish_2024_muscle_methylkit_object.Rdata")))
metadata.new <- metadata[metadata$File_names %in% meth.2024M.norm15L@sample.ids,]
metadata.new <- metadata.new[order(match(metadata.new$Genotype_names,
                                         meth.2024M.norm15L@sample.ids)),]
meth.cpg <- paste0(meth.2024M.norm15L$chr, "_", meth.2024M.norm15L$start) 
meth.DMS2 <- meth.2024M.norm15L[meth.cpg %in% DMS.cpg2,]

trainTransformed3 <- methylKit::percMethylation(meth.DMS2)
trainTransformed3 <- zoo::na.aggregate(trainTransformed3, by = meth.DMS2@treatment)
trainTransformed3 <-  as.data.frame(t(trainTransformed3))
colnames(trainTransformed3) <- DMS.cpg2
# testTransformed$pop <- metadata.new$pop
trainTransformed3$pop <- metadata.new$STRATA2
```

### glmnet model
```{r}
fitControl <- caret::trainControl(method = "repeatedcv", number = 10, repeats = 10)

elastic_model <-
  caret::train(x = trainTransformed3[,-ncol(trainTransformed3)],
               y = trainTransformed3$pop,
    method = "glmnet",
    trControl = fitControl,
    tuneLength = 10
  )


predicted.class <- caret::predict.train(elastic_model)
predicted.class <- as.character(predicted.class)
# caret::postResample(pred = predicted.class, trainTransformed3$pop)
chisq.test(predicted.class, trainTransformed3$pop) #X-squared = 48.054, df = 1, p-value = 4.146e-12
predicted.class == trainTransformed3$pop # all TRUE
sum(predicted.class == trainTransformed3$pop) #52
# heatmap(as.character(predicted.class), trainTransformed3$pop)
predict.enet.test <- predict(elastic_model, testTransformed)
predict.enet.test <- as.character(predict.enet.test)
# caret::postResample(pred = predict.enet.test, testTransformed$pop)
# plot(testTransformed$pop, predict.enet.test)
chisq.test(predict.enet.test, testTransformed$pop) #X-squared = 0.027829, df = 1, p-value = 0.8675
predict.enet.test == testTransformed$pop 
sum(predict.enet.test == testTransformed$pop)/length(testTransformed$pop)
# TOP100_2024: 41 TRUE  = 47%
# DMS_2024M: 42 TRUE  = 48%
```

### Variable contributions

```{r}
varImp.ridge <- caret::varImp(elastic_model, scale = FALSE)
plot(varImp.ridge, top = 100)
varImp.ridge.df <- varImp.ridge$importance
varImp.ridge.df$locus <- rownames(varImp.ridge.df)
varImp.ridge.names_Overall <- varImp.ridge.df$locus[varImp.ridge.df$Overall > 0.001]

all.model.names <- c(varImp.ridge.names.58.5.2, varImp.ridge.names.88.1,
                     varImp.ridge.names.58.4.1, varImp.ridge.names.58.4.2)
length(unique(all.model.names))#76
sum(duplicated(all.model.names))# 2


meth.RDA.gr <- as(meth.RDA, "GRanges")
x <- as.integer(gsub(pattern = "V", replacement = "", x = all.model.names))
x <- unique(x )
print(data.frame(loc.number = x, 
                 loc.name = paste0(meth.RDA$chr[x],"_",meth.RDA$start[x])))
meth.regional.gr <- meth.RDA.gr[x,]
meth.regional.raw <- meth.RDA[x,]
# meth.aging.prc <- methylKit::percMethylation(meth.regional.raw)
meth.diff.df <- as.data.frame(t(methylKit::percMethylation(meth.regional.raw)))
loc.names <- paste0(meth.regional.raw$chr,"_",meth.regional.raw$start)
colnames(meth.diff.df) <- loc.names

meth.diff.df$individual <- rownames(meth.diff.df)
print(load(paste0(rdata.path,"Toothfish_metadata.Rdata")))
identical(metadata2$SAMPLE_ID, meth.diff.df$individual)
meth.diff.df$pop <- metadata2$region

meth.diff.df.plot <- tidyr::gather(meth.diff.df, key = locus, 
                                   value = perc_methylation, -individual, -pop,
                                   factor_key = TRUE) %>%
  dplyr::mutate(chr = stringr::str_split_fixed(locus, 
                                               pattern = "_", 
                                               n = 2)[,1])
meth.diff.df.plot.locus <- meth.diff.df.plot %>% dplyr::group_by(chr,locus,pop) %>% 
  dplyr::summarise(mean_pop = mean(perc_methylation, na.rm = TRUE),
            .groups = 'drop')

# meth.diff.df.plot.centroid <- meth.diff.df.plot %>% dplyr::group_by(chr,pop) %>% 
#   dplyr::summarise(mean_pop = mean(perc_methylation, na.rm = TRUE),
#             .groups = 'drop')
# 
# meth.diff.df.plot.label <- meth.diff.df.plot %>% dplyr::group_by(chr) %>% 
#   dplyr::summarise(nmarkers = length(unique(locus)),
#             .groups = 'drop')



region.meth.plot <- ggplot2::ggplot() +
  ggplot2::geom_point(data = meth.diff.df.plot[meth.diff.df.plot$locus 
                                               %in%
                                                 unique(meth.diff.df.plot$locus)
                                               [1:5],],
                      ggplot2::aes(x = pop, y = perc_methylation, colour = locus)) +
  ggplot2::geom_line(data = meth.diff.df.plot.locus[meth.diff.df.plot.locus$locus 
                                             %in% 
                                               unique(meth.diff.df.plot$locus)
                                             [1:5],],
                     ggplot2::aes(x = pop, y = mean_pop, colour = locus, group = locus)) +
  ggplot2::geom_point(data = meth.diff.df.plot.locus[meth.diff.df.plot.locus$locus 
                                             %in% 
                                               unique(meth.diff.df.plot$locus)[1:5],],
                     ggplot2::aes(x = pop, y = mean_pop, colour = locus, group = locus),
                     size = 5)
print(region.meth.plot)


region.meth.plot2 <- ggplot2::ggplot() +
  ggplot2::geom_point(data = meth.diff.df.plot,
                      ggplot2::aes(x = pop, y = perc_methylation, colour = locus)) +
  ggplot2::geom_line(data = meth.diff.df.plot.locus,
                     ggplot2::aes(x = pop, y = mean_pop, colour = locus, group = locus)) +
  ggplot2::geom_point(data = meth.diff.df.plot.locus,
                     ggplot2::aes(x = pop, y = mean_pop, colour = locus, group = locus),
                     size = 5) +
  ggplot2::facet_wrap(~chr) + 
  ggplot2::guides(colour = ggplot2::guide_legend(ncol = 2))
print(region.meth.plot2)


ggplot2::ggsave(plot = region.meth.plot2,
                filename = paste0(figure.path,
                                  # "TOA_aging_methylation_percentage_enet05.png"
                                  "TOA_Regional_methylation_percentage_ridge.png"
                                  # "TOA_aging_methylation_percentage_all.models.png"
                                  ),
                                  width = 45, height = 30,
                                  units = "cm", device = "png")
```




## Re-assignment success of 2024 based on 2018 DMS

```{r}
print(load(paste0(rdata.path,"Toothfish_2024_muscle_methylkit_object.Rdata")))
meth.cpg <- paste0(meth.2024M.norm15L$chr, "_", meth.2024M.norm15L$start) #1037176 

print(load(paste0(rdata.path,"Toothfish_metadata.Rdata"))) 
metadata.new <- metadata[metadata$Genotype_names %in% meth.2024M.norm15L@sample.ids,]
metadata.new <- metadata.new[order(match(metadata.new$Genotype_names,
                                         meth.2024M.norm15L@sample.ids)),]
print(load(paste0(rdata.path,"Toothfish_DMS_CpGsites.Rdata")))
DMS.cpg <- unique(DMS$DMS[DMS$region_pw %in% c("EANT4.1_EANT4.2", "EANT4.1_ROSS" , "EANT4.2_ROSS", "EANT_ROSS")]) #188
# print(load(paste0(rdata.path,"Toothfish_2018_muscle_methylation_DMS.Rdata")))
# DMS.cpg <- paste0(as.character(top100.2024M@seqnames), "_", as.data.frame(top100.2024M@ranges)$start) #283819 
# DMS.cpg <- paste0(myDiff25p$chr, "_", myDiff25p$start) #1448

sum(meth.cpg %in% DMS.cpg) #38
DMS.cpg2 <- meth.cpg[meth.cpg %in% DMS.cpg]
# meth.DMS1 <- meth.2024M.norm15L[meth.cpg %in% DMS.cpg2,]

# testTransformed <- methylKit::percMethylation(meth.DMS1)
# testTransformed <- zoo::na.aggregate(testTransformed, by = meth.DMS1@treatment)
# testTransformed <-  as.data.frame(t(testTransformed))
# colnames(testTransformed) <- DMS.cpg2
# # testTransformed$pop <- metadata.new$pop
# testTransformed$pop <- metadata.new$STRATA2
# testTransformed <- testTransformed[testTransformed$pop %in% c("East_Antarctica", "Ross_sea"),]

print(load(paste0(rdata.path,"Toothfish_2018_muscle_methylkit_object.Rdata")))
# print(load(paste0(rdata.path,"Toothfish_subset_methylation_objects_raw_NEW.Rdata")))
# meth.2018M.norm11L <- meth.sub
metadata.new <- metadata[metadata$Genotype_names %in% meth.2018M.norm11L@sample.ids,]
metadata.new <- metadata.new[order(match(metadata.new$Genotype_names,
                                         meth.2018M.norm11L@sample.ids)),]
# metadata.new <- metadata.new[order(match(metadata.new$File_names,
#                                          meth.2018M.norm11L@sample.ids)),]
meth.cpg <- paste0(meth.2018M.norm11L$chr, "_", meth.2018M.norm11L$start) 
sum(meth.cpg %in% DMS.cpg2) #35
DMS.cpg3 <- meth.cpg[meth.cpg %in% DMS.cpg2]
meth.DMS2 <- meth.2018M.norm11L[meth.cpg %in% DMS.cpg3,]

trainTransformed3 <- methylKit::percMethylation(meth.DMS2)
trainTransformed3 <- zoo::na.aggregate(trainTransformed3, by = meth.DMS2@treatment)
trainTransformed3 <-  as.data.frame(t(trainTransformed3))
colnames(trainTransformed3) <- DMS.cpg3
# testTransformed$pop <- metadata.new$pop
trainTransformed3$pop <- metadata.new$STRATA2
trainTransformed3 <- trainTransformed3[trainTransformed3$pop %in% c("East_Antarctica", "Ross_sea"),]


metadata.new <- metadata[metadata$Genotype_names %in% meth.2024M.norm15L@sample.ids,]
metadata.new <- metadata.new[order(match(metadata.new$Genotype_names,
                                         meth.2024M.norm15L@sample.ids)),]
meth.DMS1 <- meth.2024M.norm15L[meth.cpg %in% DMS.cpg3,]
testTransformed <- methylKit::percMethylation(meth.DMS1)
testTransformed <- zoo::na.aggregate(testTransformed, by = meth.DMS1@treatment)
testTransformed <-  as.data.frame(t(testTransformed))
colnames(testTransformed) <- DMS.cpg3
# testTransformed$pop <- metadata.new$pop
testTransformed$pop <- metadata.new$STRATA2
testTransformed <- testTransformed[testTransformed$pop %in% c("East_Antarctica", "Ross_sea"),]

boolean <- is.na(colSums(testTransformed[-ncol(testTransformed)]))
testTransformed <- testTransformed[,!boolean]
trainTransformed3 <- trainTransformed3[,!boolean]
```

### glmnet model
```{r}
fitControl <- caret::trainControl(method = "repeatedcv", number = 10, repeats = 10)

elastic_model <-
  caret::train(x = trainTransformed3[,-ncol(trainTransformed3)],
               y = trainTransformed3$pop,
    method = "glmnet",
    trControl = fitControl,
    tuneLength = 10
  )


predicted.class <- caret::predict.train(elastic_model)
predicted.class <- as.character(predicted.class)
# caret::postResample(pred = predicted.class, trainTransformed3$pop)
chisq.test(predicted.class, trainTransformed3$pop) #X-squared = 48.054, df = 1, p-value = 4.146e-12
predicted.class == trainTransformed3$pop # all TRUE
sum(predicted.class == trainTransformed3$pop) #76
# heatmap(as.character(predicted.class), trainTransformed3$pop)
predict.enet.test <- predict(elastic_model, testTransformed)
predict.enet.test <- as.character(predict.enet.test)
# caret::postResample(pred = predict.enet.test, testTransformed$pop)
# plot(testTransformed$pop, predict.enet.test)
chisq.test(predict.enet.test, testTransformed$pop) #X-squared = 0.027829, df = 1, p-value = 0.8675
predict.enet.test == testTransformed$pop 
sum(predict.enet.test == testTransformed$pop)/length(testTransformed$pop)
# TOP100_2018: x TRUE  = x%
# DMS_2018M: 23 TRUE  = 44%
```

### Variable contributions

```{r}
varImp.ridge <- caret::varImp(elastic_model, scale = FALSE)
plot(varImp.ridge, top = 28)
varImp.ridge.df <- varImp.ridge$importance
varImp.ridge.df$locus <- rownames(varImp.ridge.df)
varImp.ridge.names_Overall <- varImp.ridge.df$locus[varImp.ridge.df$Overall > 0.001]

all.model.names <- c(varImp.ridge.names.58.5.2, varImp.ridge.names.88.1,
                     varImp.ridge.names.58.4.1, varImp.ridge.names.58.4.2)
length(unique(all.model.names))#76
sum(duplicated(all.model.names))# 2


meth.RDA.gr <- as(meth.RDA, "GRanges")
x <- as.integer(gsub(pattern = "V", replacement = "", x = all.model.names))
x <- unique(x )
print(data.frame(loc.number = x, 
                 loc.name = paste0(meth.RDA$chr[x],"_",meth.RDA$start[x])))
meth.regional.gr <- meth.RDA.gr[x,]
meth.regional.raw <- meth.RDA[x,]
# meth.aging.prc <- methylKit::percMethylation(meth.regional.raw)
meth.diff.df <- as.data.frame(t(methylKit::percMethylation(meth.regional.raw)))
loc.names <- paste0(meth.regional.raw$chr,"_",meth.regional.raw$start)
colnames(meth.diff.df) <- loc.names

meth.diff.df$individual <- rownames(meth.diff.df)
print(load(paste0(rdata.path,"Toothfish_metadata.Rdata")))
identical(metadata2$SAMPLE_ID, meth.diff.df$individual)
meth.diff.df$pop <- metadata2$region

meth.diff.df.plot <- tidyr::gather(meth.diff.df, key = locus, 
                                   value = perc_methylation, -individual, -pop,
                                   factor_key = TRUE) %>%
  dplyr::mutate(chr = stringr::str_split_fixed(locus, 
                                               pattern = "_", 
                                               n = 2)[,1])
meth.diff.df.plot.locus <- meth.diff.df.plot %>% dplyr::group_by(chr,locus,pop) %>% 
  dplyr::summarise(mean_pop = mean(perc_methylation, na.rm = TRUE),
            .groups = 'drop')

# meth.diff.df.plot.centroid <- meth.diff.df.plot %>% dplyr::group_by(chr,pop) %>% 
#   dplyr::summarise(mean_pop = mean(perc_methylation, na.rm = TRUE),
#             .groups = 'drop')
# 
# meth.diff.df.plot.label <- meth.diff.df.plot %>% dplyr::group_by(chr) %>% 
#   dplyr::summarise(nmarkers = length(unique(locus)),
#             .groups = 'drop')



region.meth.plot <- ggplot2::ggplot() +
  ggplot2::geom_point(data = meth.diff.df.plot[meth.diff.df.plot$locus 
                                               %in%
                                                 unique(meth.diff.df.plot$locus)
                                               [1:5],],
                      ggplot2::aes(x = pop, y = perc_methylation, colour = locus)) +
  ggplot2::geom_line(data = meth.diff.df.plot.locus[meth.diff.df.plot.locus$locus 
                                             %in% 
                                               unique(meth.diff.df.plot$locus)
                                             [1:5],],
                     ggplot2::aes(x = pop, y = mean_pop, colour = locus, group = locus)) +
  ggplot2::geom_point(data = meth.diff.df.plot.locus[meth.diff.df.plot.locus$locus 
                                             %in% 
                                               unique(meth.diff.df.plot$locus)[1:5],],
                     ggplot2::aes(x = pop, y = mean_pop, colour = locus, group = locus),
                     size = 5)
print(region.meth.plot)


region.meth.plot2 <- ggplot2::ggplot() +
  ggplot2::geom_point(data = meth.diff.df.plot,
                      ggplot2::aes(x = pop, y = perc_methylation, colour = locus)) +
  ggplot2::geom_line(data = meth.diff.df.plot.locus,
                     ggplot2::aes(x = pop, y = mean_pop, colour = locus, group = locus)) +
  ggplot2::geom_point(data = meth.diff.df.plot.locus,
                     ggplot2::aes(x = pop, y = mean_pop, colour = locus, group = locus),
                     size = 5) +
  ggplot2::facet_wrap(~chr) + 
  ggplot2::guides(colour = ggplot2::guide_legend(ncol = 2))
print(region.meth.plot2)


ggplot2::ggsave(plot = region.meth.plot2,
                filename = paste0(figure.path,
                                  # "TOA_aging_methylation_percentage_enet05.png"
                                  "TOA_Regional_methylation_percentage_ridge.png"
                                  # "TOA_aging_methylation_percentage_all.models.png"
                                  ),
                                  width = 45, height = 30,
                                  units = "cm", device = "png")
```




# 7b_Methylation: 2024 E-ant - muscle tissue


## Select samples

```{r}
print(load(paste0(rdata.path,"Toothfish_2024_muscle_methylkit_cpgDB.Rdata")))
print(load(paste0(rdata.path,"Toothfish_metadata.Rdata")))

metadata2 <- metadata[metadata$pop == "58.4.2-2024_70L" & metadata$Tissue_type == "muscle" 
                      & !grepl(pattern = "REP", x = metadata$File_names) &
                      grepl(pattern = "TOA", x = metadata$File_names),]

print(load(paste0(rdata.path, "Low_quality_fish.Rdata")))

# rm1 <- Reduce(intersect, list(low.genome.mapping,low.genome.methylation))
rm1 <- unique(c(low.genome.mapping,low.genome.methylation))
print(rm1)
rm2 <- c("TOA_58-4-2_B40_AE102056_muscle_M") # bad PCA
rm3 <- c("TOA_58-4-2_B24_AE100883_muscle_F", "TOA_88-2_P24_AD292328_muscle_F",
         "TOA_58-4-2_B64_AE107905_muscle_F") # bad coverage
rm4 <- c("TOA_48-2_031_UK116","TOA_58-4-2_111_AN172604","TOA_88-1_230_AD082543",
         "TOA_48-2_024_UK082") #bad tissue PCA

rm <- c(rm1,rm2,rm3,rm4) # bad coverage

metadata2 <- metadata2[!(metadata2$File_names %in% rm | metadata2$Genotype_names %in% rm),]

filt.sample.id.full <- sapply(filtered.myobj@.Data, function(x){x@sample.id})

# filt.sample.id <- filt.sample.id.full[filt.sample.id.full %in% metadata2$File_names]
# metadata2 <- metadata2[order(match(metadata.new$File_names, filt.sample.id)),]

filt.sample.id <- filt.sample.id.full[filt.sample.id.full %in% metadata2$Genotype_names]
metadata2 <- metadata2[order(match(metadata2$Genotype_names, filt.sample.id)),]
identical(filt.sample.id,metadata2$Genotype_names)


pop <- metadata2$pop
pop.levels <- c("58.4.2-2024_70L")

pop <- as.numeric(factor(pop, levels = pop.levels))
sample.ids <- metadata2$Genotype_names



filtered.myobj2.normal <- methylKit::reorganize(
  methylObj = filtered.myobj2.normal,
  sample.ids = sample.ids,
  treatment = pop,
  chunk.size = 1e+06,
  save.db = TRUE,
  suffix = "TOA_2024_EANT_muscle_normal"
)
save(myobjDB,filtered.myobj,filtered.myobj2.normal,
     file = paste0(rdata.path,"Toothfish_2024_EANT_muscle_methylkit_cpgDB.Rdata"))
```


## Merging samples

```{r}
print(load(paste0(rdata.path,"Toothfish_2024_EANT_muscle_methylkit_cpgDB.Rdata")))
pop <- metadata2$pop
pop.levels <- c("58.4.2-2024_70L")
pop <- factor(pop, levels = pop.levels)
pop <- as.numeric(pop)

table(pop)


meth.2024EANTM.norm11L <- methylKit::unite(filtered.myobj2.normal, destrand = FALSE,
                                       save.db = FALSE,
                                       min.per.group = 11L, #missing data allowed
                                       mc.cores = 25)
#749,867 common sites for 11 samples per group

save(meth.2024EANTM.norm11L,
     file = paste0(rdata.path,"Toothfish_2024_EANT_muscle_methylkit_object.Rdata"))
```


## Data QC

### Remove ASM

```{r}
load(paste0(rdata.path,"Allele_specific_mutations.Rdata"))

mut <- as(asmTRUE,"GRanges")
meth.2024EANTM.norm11L <- as(meth.2024EANTM.norm11L,"methylBase") 

sum(as(meth.2024EANTM.norm11L,"GRanges") %over% mut)#28018
meth.asm <- methylKit::selectByOverlap(meth.2024EANTM.norm11L, mut) 
BOOLEAN <- paste0(meth.2024EANTM.norm11L$chr,"_",meth.2024EANTM.norm11L$start) %in%
  paste0(asmTRUE$Chr,"_",asmTRUE$Pos)
sum(BOOLEAN) #659
meth.2024EANTM.norm11L <- methylKit::select(meth.2024EANTM.norm11L, !BOOLEAN) 

save(meth.2024EANTM.norm11L,
     file = paste0(rdata.path,"Toothfish_2024_EANT_muscle_methylkit_object.Rdata"))
```



### Coverage 

```{r}
meth.df <- methylKit::getData(meth.2024EANTM.norm11L)
cov <- meth.df[,meth.2024EANTM.norm11L@coverage.index]
colnames(cov) <- meth.2024EANTM.norm11L@sample.ids
dpf <- data.frame(cov, locus = paste0(meth.df$chr,"_",meth.df$start))
dpf <- tidyr::gather(dpf, key = individual, 
                           value = Depth, -locus,
                           factor_key = TRUE) %>%
  dplyr::group_by(individual) %>%
  dplyr::mutate(NA_cnt = sum(is.na(Depth))/length(unique(locus))) %>%
  dplyr::ungroup()
dpf.NA <- dpf[!duplicated(dpf$individual),]
dpf.NA$NA_cnt <- round(dpf.NA$NA_cnt, 3)

dpf <- dpf[ dpf$Depth > 0 & !is.na(dpf$Depth),]

max(dpf$Depth, na.rm = T) #99
myPlot <- ggplot2::ggplot(dpf, ggplot2::aes(x = individual, y = Depth)) + 
  ggplot2::geom_boxplot() +
  ggplot2::geom_text(data = dpf.NA,ggplot2::aes(x = individual, y = 150, label = NA_cnt), 
                     size = 2, angle = 90) +
  ggplot2::theme_bw() + 
  ggplot2::theme(axis.title.x = ggplot2::element_blank(), 
                 axis.text.y = ggplot2::element_text(size = 15),
                 axis.title.y = ggplot2::element_text(size = 15),
                 axis.text.x = ggplot2::element_text(size = 7, angle = 90, hjust = 1)) +
  ggplot2::scale_y_continuous(trans = scales::log2_trans(),
                  breaks = c(1, 10, 50, 100),
                  minor_breaks = c(1:10, 2:5*10))

print(myPlot)
ggplot2::ggsave(myPlot, bg = "white",
                filename = 
                  paste0(figure.path,
                  "1B.Toothfish_2024_EANT_muscle_normalised15_methylkit_individual_coverage_CpGs.png"
                                  ), 
                width = 45, height = 15, units = "cm")

nrow(cov) # 749264 SMPs
plot(x = 1:ncol(cov), y = colMeans(cov, na.rm = TRUE), col = meth.2024EANTM.norm11L@treatment)
plot(x = 1:ncol(cov), y = colMeans(cov > 20, na.rm = TRUE), col = meth.2024EANTM.norm11L@treatment)
plot(x = 1:ncol(cov), y = colMeans(cov > 10, na.rm = TRUE), col = meth.2024EANTM.norm11L@treatment)
meth.2024EANTM.norm11L@sample.ids[colMeans(cov > 10, na.rm = TRUE) < 0.95]
# samples with low coverage: "TOA_58-4-2_B24_AE100883_muscle_F", "TOA_88-2_P24_AD292328_muscle_F"
```



## Load data

```{r}
print(load(paste0(rdata.path,"Toothfish_2024_EANT_muscle_methylkit_object.Rdata")))
print(load(paste0(rdata.path,"Toothfish_metadata.Rdata"))) 

metadata.new <- metadata[metadata$Genotype_names %in% meth.2024EANTM.norm11L@sample.ids,]
metadata.new <- metadata.new[order(match(metadata.new$Genotype_names,
                                         meth.2024EANTM.norm11L@sample.ids)),]

length(meth.2024EANTM.norm11L@row.names) # number of sites
```


## Global clustering analysis

```{r}
methylKit::clusterSamples(meth.2024EANTM.norm11L, dist = "correlation", sd.filter = TRUE, 
                          method = "ward.D2", plot = TRUE, )
dev.print(file = paste0(figure.path,"1_TOA_2024_EANT_muscle_cluster_plot.png"),
          device = png, res = 300,width = 45,height = 15,units = "cm")
```


### PCA

```{r}
PCobj <- methylKit::PCASamples(meth.2024EANTM.norm11L, comp = c(1,2), 
                               adj.lim = c(0,0), obj.return = TRUE)

var_frac <- PCobj$sdev/sum(PCobj$sdev)
print(paste0("Total variance explained: ",signif(sum(var_frac[1:3]) * 100, 3), "%")) 
barplot(var_frac[1:20], main = "variance explained", 
        col = heat.colors(length(var_frac[1:20])))

identical(metadata.new$Genotype_names, meth.2024EANTM.norm11L@sample.ids)
identical(metadata.new$Genotype_names, rownames(PCobj$x))

pop <- metadata.new$pop

plot.title <- "Toothfish methylation - 2024 EANT muscle (749,264 CpG sites)"

data <- data.frame(ind = rownames(PCobj$x),PCobj$x,  pop = pop)
# data <- data.frame(ind = rownames(PCobj$rotation),PCobj$rotation, pop = pop)
data <- merge(data, aggregate(cbind(mean.x = PC1, mean.y = PC2) ~
                                pop, data, mean), by = "pop")

pca.plot <- ggplot2::ggplot(data, ggplot2::aes(PC1, PC2, color = pop)) +
  ggplot2::geom_point(size = 4) + 
  ggplot2::geom_point(ggplot2::aes(x = mean.x,
                                   y = mean.y, color = pop),
                      size = 7,
                      shape = 17) +
  ggplot2::scale_colour_manual(values = colours2) +
  ggplot2::geom_hline(yintercept = 0) +
  ggplot2::geom_vline(xintercept = 0) + 
  ggplot2::labs(
    subtitle = paste0("Total variance explained: ",
                      signif(sum(var_frac[1:4]) * 100, 3), "%"),
    y = paste0("PC2; ", signif(var_frac[2] * 100, digits = 2), " %"),
    x = paste0("PC1; ", signif(var_frac[1] * 100, digits = 2), " %"),
    title = plot.title,
    caption = "") + 
  ggplot2::theme_minimal() +
  ggplot2::theme(
    axis.text = ggplot2::element_text(size = 15),
    axis.title.x = ggplot2::element_text(size = 20),
    axis.title.y = ggplot2::element_text(size = 20),
    legend.text = ggplot2::element_text(size = 15)
  )

print(pca.plot)
ggplot2::ggsave(pca.plot, bg = "white",
                filename = paste0(figure.path,"Toothfish_2024_EANT_muscle_PCA_CpG_PC12.png"), 
                width = 30, height = 15, units = "cm")



data <- data.frame(ind = rownames(PCobj$x),PCobj$x,  pop = pop)
# data <- data.frame(ind = rownames(PCobj$rotation),PCobj$rotation, pop = pop)
data <- merge(data, aggregate(cbind(mean.x = PC2, mean.y = PC3) ~
                                pop, data, mean), by = "pop")
pca.plot <- ggplot2::ggplot(data, ggplot2::aes(PC2, PC3, color = pop)) +
  ggplot2::geom_point(size = 4) + 
  ggplot2::geom_point(ggplot2::aes(x = mean.x,
                                   y = mean.y, color = pop),
                      size = 7,
                      shape = 17) +
  ggplot2::scale_colour_manual(values = colours2) +
  ggplot2::geom_hline(yintercept = 0) +
  ggplot2::geom_vline(xintercept = 0) + 
  ggplot2::labs(
    subtitle = paste0("Total variance explained: ",
                      signif(sum(var_frac[1:4]) * 100, 3), "%"),
    y = paste0("PC3; ", signif(var_frac[3] * 100, digits = 2), " %"),
    x = paste0("PC2; ", signif(var_frac[2] * 100, digits = 2), " %"),
    title = plot.title,
    caption = "") + 
  ggplot2::theme_minimal() +
  ggplot2::theme(
    axis.text = ggplot2::element_text(size = 15),
    axis.title.x = ggplot2::element_text(size = 20),
    axis.title.y = ggplot2::element_text(size = 20),
    legend.text = ggplot2::element_text(size = 15)
  )

print(pca.plot)
ggplot2::ggsave(pca.plot, bg = "white",
                filename = paste0(figure.path,"Toothfish_2024_EANT_muscle_PCA_CpG_PC23.png"), 
                width = 30, height = 15, units = "cm")

# data$ind[data$PC3 > 100] #"TOA_58-4-2_B40_AE102056_muscle_M"
# data$ind[data$PC3 < -100] #"TOA_58-4-2_B64_AE107905_muscle_F"
```





## 2024 fish muscle with DMS from 2017/2018
```{r}
# print(load(paste0(rdata.path,"Toothfish_DMS_CpGsites.Rdata")))
print(load(paste0(rdata.path,"Toothfish_2018_muscle_methylation_DMS.Rdata")))

meth.cpg.2024 <- paste0(meth.2024EANTM.norm11L$chr, "_", meth.2024EANTM.norm11L$start) 

DMS.cpg.2018 <- paste0(myDiff25p$chr, "_", myDiff25p$start) #1202
# DMS.cpg <- unique(c(DMS$DMS)) #1243
sum(meth.cpg.2024 %in% DMS.cpg.2018) #362 = 30%

meth.DMS.2024 <- meth.2024EANTM.norm11L[meth.cpg.2024 %in% DMS.cpg.2018,] 
```

### Heatmap

```{r}
meth.diff.df <- as.data.frame(t(methylKit::percMethylation(meth.DMS)))
loc.names <- paste0(meth.DMS$chr,"_",meth.DMS$start)
colnames(meth.diff.df) <- loc.names

meth.diff.df$individual <- rownames(meth.diff.df)
pop <- metadata.new$pop
meth.diff.df$pop <- as.character(pop)


heat.plot.data <- meth.diff.df %>%
  tibble::as_tibble() %>%
  tidyr::gather(key = "CpG", value = "Methylation_percentage", -individual, -pop,
                factor_key = FALSE) %>%
  dplyr::group_by(pop, CpG) %>%
  dplyr::mutate(CpG_mean = mean(Methylation_percentage, na.rm = TRUE)) %>%
  dplyr::arrange(pop, individual, CpG) %>%
  dplyr::mutate(individual = factor(individual, levels = unique(individual))) %>%
  dplyr::mutate(CpG = factor(CpG, levels = unique(CpG)))
cpgs <- heat.plot.data$CpG[heat.plot.data$individual == unique(heat.plot.data$individual)[1]]
cpgs_mean <- heat.plot.data$CpG_mean[heat.plot.data$individual == unique(heat.plot.data$individual)[1]]

cpgs <- cpgs[order(cpgs_mean, decreasing = TRUE)]
heat.plot.data$CpG <- factor(heat.plot.data$CpG, levels = c("pop", unique(as.character(cpgs))))
heat.plot.data <- heat.plot.data[order(heat.plot.data$CpG),]

heat.plot <- ggplot2::ggplot(data = heat.plot.data, ggplot2::aes(x = individual)) + 
    ggplot2::geom_tile(data = heat.plot.data, 
                       ggplot2::aes(x = individual, y = CpG, fill = Methylation_percentage)) +
    ggplot2::scale_fill_gradient2(low = "blue", mid = "yellow", high = "red", 
                                  midpoint = 50, guide = "colourbar") +
    ggnewscale::new_scale_fill() +
    ggplot2::geom_tile(data = heat.plot.data, ggplot2::aes(x = individual, y = "pop", fill = pop,
                                    height = length(unique(heat.plot.data$CpG))*0.02)) +
    ggplot2::scale_fill_manual(values = colours2) +
    ggplot2::scale_y_discrete(drop = FALSE) +
    ggplot2::theme(legend.position = "top",
                   axis.title.y = ggplot2::element_blank(),
                   axis.title.x = ggplot2::element_blank(),
                   axis.text.x = ggplot2::element_text(face = "bold", color = "black", 
                                        size = 10, angle = 90, hjust = 1, vjust = 0.5))
print(heat.plot)

ggplot2::ggsave(heat.plot, bg = "white",
                filename = paste0(figure.path,
                                  "Toothfish_2024_EANT_muscle_BASED_ON_2018_DMS_HEATMAP.png"), 
                width = 30, height = 30, units = "cm")
```


### DMS Clustering
```{r}
methdif.perc <- methylKit::percMethylation(meth.DMS)
methdif.perc2 <- zoo::na.aggregate(methdif.perc, by = meth.DMS@treatment)
PCobj <- prcomp(methdif.perc2, retx = TRUE, scale. = TRUE)

var_frac <- PCobj$sdev/sum(PCobj$sdev)
print(paste0("Total variance explained: ",signif(sum(var_frac[1:3]) * 100, 3), "%")) 
barplot(var_frac[1:3], main = "variance explained", 
        col = heat.colors(length(var_frac[1:3])))


identical(metadata.new$Genotype_names, meth.DMS@sample.ids)
# identical(metadata.new$individual, rownames(PCobj$x))
identical(metadata.new$Genotype_names, rownames(PCobj$rotation))
pop <- metadata.new$pop

plot.title <- "Toothfish methylation - 2024 EANT Muscle (362 DMS from 2018 - 25%)"

# data <- data.frame(ind = rownames(PCobj$x),PCobj$x,  pop = pop)
data <- data.frame(ind = rownames(PCobj$rotation),PCobj$rotation,  pop = pop)
data <- merge(data, aggregate(cbind(mean.x = PC1, mean.y = PC2) ~
                                pop, data, mean), by = "pop")

pca.plot <- ggplot2::ggplot(data, ggplot2::aes(PC1, PC2, color = pop)) +
  ggplot2::geom_point(size = 4) + 
  ggplot2::geom_point(ggplot2::aes(x = mean.x,
                                   y = mean.y, color = pop),
                      size = 7,
                      shape = 17) +
  ggplot2::scale_colour_manual(values = colours2) +
  ggplot2::geom_hline(yintercept = 0) +
  ggplot2::geom_vline(xintercept = 0) + 
  ggplot2::labs(
    subtitle = paste0("Total variance explained: ",
                      signif(sum(var_frac[1:3]) * 100, 3), "%"),
    y = paste0("PC2; ", signif(var_frac[2] * 100, digits = 2), " %"),
    x = paste0("PC1; ", signif(var_frac[1] * 100, digits = 2), " %"),
    title = plot.title,
    caption = "") + 
  ggplot2::theme_minimal() +
  ggplot2::theme(
    axis.text = ggplot2::element_text(size = 15),
    axis.title.x = ggplot2::element_text(size = 20),
    axis.title.y = ggplot2::element_text(size = 20),
    legend.text = ggplot2::element_text(size = 15)
  )

print(pca.plot)
ggplot2::ggsave(pca.plot, bg = "white",
                filename = paste0(figure.path,"Toothfish_2024_EANT_muscle_PCA_CpG_DMS2018_PC12.png"), 
                width = 30, height = 15, units = "cm")



# data <- data.frame(ind = rownames(PCobj$x),PCobj$x,  pop = pop)
data <- data.frame(ind = rownames(PCobj$rotation),PCobj$rotation,  pop = pop)
data <- merge(data, aggregate(cbind(mean.x = PC1, mean.y = PC3) ~
                                pop, data, mean), by = "pop")

pca.plot <- ggplot2::ggplot(data, ggplot2::aes(PC1, PC3, color = pop)) +
  ggplot2::geom_point(size = 4) + 
  ggplot2::geom_point(ggplot2::aes(x = mean.x,
                                   y = mean.y, color = pop),
                      size = 7,
                      shape = 17) +
  ggplot2::scale_colour_manual(values = colours2) +
  ggplot2::geom_hline(yintercept = 0) +
  ggplot2::geom_vline(xintercept = 0) + 
  ggplot2::labs(
    subtitle = paste0("Total variance explained: ",
                      signif(sum(var_frac[1:3]) * 100, 3), "%"),
    y = paste0("PC3; ", signif(var_frac[3] * 100, digits = 2), " %"),
    x = paste0("PC1; ", signif(var_frac[1] * 100, digits = 2), " %"),
    title = plot.title,
    caption = "") + 
  ggplot2::theme_minimal() +
  ggplot2::theme(
    axis.text = ggplot2::element_text(size = 15),
    axis.title.x = ggplot2::element_text(size = 20),
    axis.title.y = ggplot2::element_text(size = 20),
    legend.text = ggplot2::element_text(size = 15)
  )

print(pca.plot)
ggplot2::ggsave(pca.plot, bg = "white",
                filename = paste0(figure.path,
                                  "Toothfish_2024_EANT_muscle_PCA_CpG_DMS2018_PC13.png"), 
                width = 30, height = 15, units = "cm")




# data <- data.frame(ind = rownames(PCobj$x),PCobj$x,  pop = pop)
data <- data.frame(ind = rownames(PCobj$rotation),PCobj$rotation,  pop = pop)
data <- merge(data, aggregate(cbind(mean.x = PC2, mean.y = PC3) ~
                                pop, data, mean), by = "pop")

pca.plot <- ggplot2::ggplot(data, ggplot2::aes(PC2, PC3, color = pop)) +
  ggplot2::geom_point(size = 4) + 
  ggplot2::geom_point(ggplot2::aes(x = mean.x,
                                   y = mean.y, color = pop),
                      size = 7,
                      shape = 17) +
  ggplot2::scale_colour_manual(values = colours2) +
  ggplot2::geom_hline(yintercept = 0) +
  ggplot2::geom_vline(xintercept = 0) + 
  ggplot2::labs(
    subtitle = paste0("Total variance explained: ",
                      signif(sum(var_frac[1:3]) * 100, 3), "%"),
    y = paste0("PC3; ", signif(var_frac[3] * 100, digits = 2), " %"),
    x = paste0("PC2; ", signif(var_frac[2] * 100, digits = 2), " %"),
    title = plot.title,
    caption = "") + 
  ggplot2::theme_minimal() +
  ggplot2::theme(
    axis.text = ggplot2::element_text(size = 15),
    axis.title.x = ggplot2::element_text(size = 20),
    axis.title.y = ggplot2::element_text(size = 20),
    legend.text = ggplot2::element_text(size = 15)
  )

print(pca.plot)
ggplot2::ggsave(pca.plot, bg = "white",
                filename = paste0(figure.path,"Toothfish_2024_EANT_muscle_PCA_CpG_DMS2018_PC23.png"), 
                width = 30, height = 15, units = "cm")
```



## Re-assignment success of 2018 based on 2024 DMS

```{r}
print(load(paste0(rdata.path,"Toothfish_2018_muscle_methylkit_object.Rdata")))
print(load(paste0(rdata.path,"Toothfish_subset_methylation_objects_raw_NEW.Rdata")))
# meth.2018M.norm11L <- meth.sub

print(load(paste0(rdata.path,"Toothfish_metadata.Rdata"))) 
metadata.new <- metadata[metadata$Genotype_names %in% meth.2018M.norm11L@sample.ids,]
metadata.new <- metadata.new[order(match(metadata.new$File_names,
                                         meth.2018M.norm11L@sample.ids)),]

meth.cpg.2018 <- paste0(meth.2018M.norm11L$chr, "_", meth.2018M.norm11L$start) #1037176 
DMS.cpg.2018 <- paste0(myDiff25p$chr, "_", myDiff25p$start) #1202

sum(meth.cpg.2024 %in% DMS.cpg.2018) #362
sum(meth.cpg.2018 %in% DMS.cpg.2018) #1202

DMS.cpg.2024 <- paste0(meth.DMS.2024$chr, "_", meth.DMS.2024$start) 
meth.DMS.2018 <- meth.2018M.norm11L[meth.cpg.2018 %in% DMS.cpg.2024,]

trainTransformed3 <- methylKit::percMethylation(meth.DMS.2018)
trainTransformed3 <- zoo::na.aggregate(trainTransformed3, by = meth.DMS.2018@treatment)
trainTransformed3 <-  as.data.frame(t(trainTransformed3))
colnames(trainTransformed3) <- DMS.cpg.2024
trainTransformed3$pop <- metadata.new$STRATA2
# trainTransformed3 <- trainTransformed3[trainTransformed3$pop %in% c("East_Antarctica", "Ross_sea"),]

print(load(paste0(rdata.path,"Toothfish_2024_EANT_muscle_methylkit_object.Rdata")))
metadata.new <- metadata[metadata$Genotype_names %in% meth.2024EANTM.norm11L@sample.ids,]
metadata.new <- metadata.new[order(match(metadata.new$Genotype_names,
                                         meth.2024EANTM.norm11L@sample.ids)),]
meth.cpg.2024 <- paste0(meth.2024EANTM.norm11L$chr, "_", meth.2024EANTM.norm11L$start) 
meth.DMS.2024 <- meth.2024EANTM.norm11L[meth.cpg.2024 %in% DMS.cpg.2024,]

testTransformed <- methylKit::percMethylation(meth.DMS.2024)
testTransformed <- zoo::na.aggregate(testTransformed, by = meth.DMS.2024@treatment)
testTransformed <-  as.data.frame(t(testTransformed))
colnames(testTransformed) <- DMS.cpg.2024
testTransformed$pop <- metadata.new$STRATA2
```

### glmnet model
```{r}
fitControl <- caret::trainControl(method = "repeatedcv", number = 10, repeats = 10)

elastic_model <-
  caret::train(x = trainTransformed3[,-ncol(trainTransformed3)],
               y = trainTransformed3$pop,
    method = "glmnet",
    trControl = fitControl,
    tuneLength = 10
  )



predicted.class <- caret::predict.train(elastic_model)
predicted.class <- as.character(predicted.class)
# caret::postResample(pred = predicted.class, trainTransformed3$pop)
chisq.test(predicted.class, trainTransformed3$pop) #X-squared = 222, df = 4, p-value < 2.2e-16
predicted.class == trainTransformed3$pop # all TRUE
sum(predicted.class == trainTransformed3$pop) #111
# heatmap(as.character(predicted.class), trainTransformed3$pop)
predict.enet.test <- predict(elastic_model, testTransformed)
predict.enet.test <- as.character(predict.enet.test)
# caret::postResample(pred = predict.enet.test, testTransformed$pop)
# plot(testTransformed$pop, predict.enet.test)
chisq.test(predict.enet.test, testTransformed$pop) #X-squared = 0.027829, df = 1, p-value = 0.8675
predict.enet.test == testTransformed$pop 
sum(predict.enet.test == testTransformed$pop)/length(testTransformed$pop) #7%
```

### Variable contributions

```{r}
varImp.ridge <- caret::varImp(elastic_model, scale = FALSE)
plot(varImp.ridge, top = 100)
varImp.ridge.df <- varImp.ridge$importance
varImp.ridge.df$locus <- rownames(varImp.ridge.df)
varImp.ridge.names_Overall <- varImp.ridge.df$locus[varImp.ridge.df$Overall > 0.001]

all.model.names <- c(varImp.ridge.names.58.5.2, varImp.ridge.names.88.1,
                     varImp.ridge.names.58.4.1, varImp.ridge.names.58.4.2)
length(unique(all.model.names))#76
sum(duplicated(all.model.names))# 2


meth.RDA.gr <- as(meth.RDA, "GRanges")
x <- as.integer(gsub(pattern = "V", replacement = "", x = all.model.names))
x <- unique(x )
print(data.frame(loc.number = x, 
                 loc.name = paste0(meth.RDA$chr[x],"_",meth.RDA$start[x])))
meth.regional.gr <- meth.RDA.gr[x,]
meth.regional.raw <- meth.RDA[x,]
# meth.aging.prc <- methylKit::percMethylation(meth.regional.raw)
meth.diff.df <- as.data.frame(t(methylKit::percMethylation(meth.regional.raw)))
loc.names <- paste0(meth.regional.raw$chr,"_",meth.regional.raw$start)
colnames(meth.diff.df) <- loc.names

meth.diff.df$individual <- rownames(meth.diff.df)
print(load(paste0(rdata.path,"Toothfish_metadata.Rdata")))
identical(metadata2$SAMPLE_ID, meth.diff.df$individual)
meth.diff.df$pop <- metadata2$region

meth.diff.df.plot <- tidyr::gather(meth.diff.df, key = locus, 
                                   value = perc_methylation, -individual, -pop,
                                   factor_key = TRUE) %>%
  dplyr::mutate(chr = stringr::str_split_fixed(locus, 
                                               pattern = "_", 
                                               n = 2)[,1])
meth.diff.df.plot.locus <- meth.diff.df.plot %>% dplyr::group_by(chr,locus,pop) %>% 
  dplyr::summarise(mean_pop = mean(perc_methylation, na.rm = TRUE),
            .groups = 'drop')

# meth.diff.df.plot.centroid <- meth.diff.df.plot %>% dplyr::group_by(chr,pop) %>% 
#   dplyr::summarise(mean_pop = mean(perc_methylation, na.rm = TRUE),
#             .groups = 'drop')
# 
# meth.diff.df.plot.label <- meth.diff.df.plot %>% dplyr::group_by(chr) %>% 
#   dplyr::summarise(nmarkers = length(unique(locus)),
#             .groups = 'drop')



region.meth.plot <- ggplot2::ggplot() +
  ggplot2::geom_point(data = meth.diff.df.plot[meth.diff.df.plot$locus 
                                               %in%
                                                 unique(meth.diff.df.plot$locus)
                                               [1:5],],
                      ggplot2::aes(x = pop, y = perc_methylation, colour = locus)) +
  ggplot2::geom_line(data = meth.diff.df.plot.locus[meth.diff.df.plot.locus$locus 
                                             %in% 
                                               unique(meth.diff.df.plot$locus)
                                             [1:5],],
                     ggplot2::aes(x = pop, y = mean_pop, colour = locus, group = locus)) +
  ggplot2::geom_point(data = meth.diff.df.plot.locus[meth.diff.df.plot.locus$locus 
                                             %in% 
                                               unique(meth.diff.df.plot$locus)[1:5],],
                     ggplot2::aes(x = pop, y = mean_pop, colour = locus, group = locus),
                     size = 5)
print(region.meth.plot)


region.meth.plot2 <- ggplot2::ggplot() +
  ggplot2::geom_point(data = meth.diff.df.plot,
                      ggplot2::aes(x = pop, y = perc_methylation, colour = locus)) +
  ggplot2::geom_line(data = meth.diff.df.plot.locus,
                     ggplot2::aes(x = pop, y = mean_pop, colour = locus, group = locus)) +
  ggplot2::geom_point(data = meth.diff.df.plot.locus,
                     ggplot2::aes(x = pop, y = mean_pop, colour = locus, group = locus),
                     size = 5) +
  ggplot2::facet_wrap(~chr) + 
  ggplot2::guides(colour = ggplot2::guide_legend(ncol = 2))
print(region.meth.plot2)


ggplot2::ggsave(plot = region.meth.plot2,
                filename = paste0(figure.path,
                                  # "TOA_aging_methylation_percentage_enet05.png"
                                  "TOA_Regional_methylation_percentage_ridge.png"
                                  # "TOA_aging_methylation_percentage_all.models.png"
                                  ),
                                  width = 45, height = 30,
                                  units = "cm", device = "png")
```




## Re-assignment success of 2024 based on 2018 DMS

```{r}
print(load(paste0(rdata.path,"Toothfish_2024_EANT_muscle_methylkit_object.Rdata")))
meth.cpg <- paste0(meth.2024EANTM.norm11L$chr, "_", meth.2024EANTM.norm11L$start) #1037176 

print(load(paste0(rdata.path,"Toothfish_metadata.Rdata"))) 
metadata.new <- metadata[metadata$Genotype_names %in% meth.2024EANTM.norm11L@sample.ids,]
metadata.new <- metadata.new[order(match(metadata.new$Genotype_names,
                                         meth.2024EANTM.norm11L@sample.ids)),]
print(load(paste0(rdata.path,"Toothfish_DMS_CpGsites.Rdata")))
DMS.cpg <- unique(DMS$DMS[DMS$region_pw %in% c("EANT4.1_EANT4.2", "EANT4.1_ROSS" , "EANT4.2_ROSS", "EANT_ROSS")]) #188
# print(load(paste0(rdata.path,"Toothfish_2018_muscle_methylation_DMS.Rdata")))
# DMS.cpg <- paste0(as.character(top100.2024M@seqnames), "_", as.data.frame(top100.2024M@ranges)$start) #283819 
# DMS.cpg <- paste0(myDiff25p$chr, "_", myDiff25p$start) #1448

sum(meth.cpg %in% DMS.cpg) #38
DMS.cpg2 <- meth.cpg[meth.cpg %in% DMS.cpg]
# meth.DMS1 <- meth.2024EANTM.norm11L[meth.cpg %in% DMS.cpg2,]

# testTransformed <- methylKit::percMethylation(meth.DMS1)
# testTransformed <- zoo::na.aggregate(testTransformed, by = meth.DMS1@treatment)
# testTransformed <-  as.data.frame(t(testTransformed))
# colnames(testTransformed) <- DMS.cpg2
# # testTransformed$pop <- metadata.new$pop
# testTransformed$pop <- metadata.new$STRATA2
# testTransformed <- testTransformed[testTransformed$pop %in% c("East_Antarctica", "Ross_sea"),]

print(load(paste0(rdata.path,"Toothfish_2018_muscle_methylkit_object.Rdata")))
# print(load(paste0(rdata.path,"Toothfish_subset_methylation_objects_raw_NEW.Rdata")))
# meth.2018M.norm11L <- meth.sub
metadata.new <- metadata[metadata$Genotype_names %in% meth.2018M.norm11L@sample.ids,]
metadata.new <- metadata.new[order(match(metadata.new$Genotype_names,
                                         meth.2018M.norm11L@sample.ids)),]
# metadata.new <- metadata.new[order(match(metadata.new$File_names,
#                                          meth.2018M.norm11L@sample.ids)),]
meth.cpg <- paste0(meth.2018M.norm11L$chr, "_", meth.2018M.norm11L$start) 
sum(meth.cpg %in% DMS.cpg2) #35
DMS.cpg3 <- meth.cpg[meth.cpg %in% DMS.cpg2]
meth.DMS2 <- meth.2018M.norm11L[meth.cpg %in% DMS.cpg3,]

trainTransformed3 <- methylKit::percMethylation(meth.DMS2)
trainTransformed3 <- zoo::na.aggregate(trainTransformed3, by = meth.DMS2@treatment)
trainTransformed3 <-  as.data.frame(t(trainTransformed3))
colnames(trainTransformed3) <- DMS.cpg3
# testTransformed$pop <- metadata.new$pop
trainTransformed3$pop <- metadata.new$STRATA2
trainTransformed3 <- trainTransformed3[trainTransformed3$pop %in% c("East_Antarctica", "Ross_sea"),]


metadata.new <- metadata[metadata$Genotype_names %in% meth.2024EANTM.norm11L@sample.ids,]
metadata.new <- metadata.new[order(match(metadata.new$Genotype_names,
                                         meth.2024EANTM.norm11L@sample.ids)),]
meth.DMS1 <- meth.2024EANTM.norm11L[meth.cpg %in% DMS.cpg3,]
testTransformed <- methylKit::percMethylation(meth.DMS1)
testTransformed <- zoo::na.aggregate(testTransformed, by = meth.DMS1@treatment)
testTransformed <-  as.data.frame(t(testTransformed))
colnames(testTransformed) <- DMS.cpg3
# testTransformed$pop <- metadata.new$pop
testTransformed$pop <- metadata.new$STRATA2
testTransformed <- testTransformed[testTransformed$pop %in% c("East_Antarctica", "Ross_sea"),]

boolean <- is.na(colSums(testTransformed[-ncol(testTransformed)]))
testTransformed <- testTransformed[,!boolean]
trainTransformed3 <- trainTransformed3[,!boolean]
```

### glmnet model
```{r}
fitControl <- caret::trainControl(method = "repeatedcv", number = 10, repeats = 10)

elastic_model <-
  caret::train(x = trainTransformed3[,-ncol(trainTransformed3)],
               y = trainTransformed3$pop,
    method = "glmnet",
    trControl = fitControl,
    tuneLength = 10
  )


predicted.class <- caret::predict.train(elastic_model)
predicted.class <- as.character(predicted.class)
# caret::postResample(pred = predicted.class, trainTransformed3$pop)
chisq.test(predicted.class, trainTransformed3$pop) #X-squared = 48.054, df = 1, p-value = 4.146e-12
predicted.class == trainTransformed3$pop # all TRUE
sum(predicted.class == trainTransformed3$pop) #76
# heatmap(as.character(predicted.class), trainTransformed3$pop)
predict.enet.test <- predict(elastic_model, testTransformed)
predict.enet.test <- as.character(predict.enet.test)
# caret::postResample(pred = predict.enet.test, testTransformed$pop)
# plot(testTransformed$pop, predict.enet.test)
chisq.test(predict.enet.test, testTransformed$pop) #X-squared = 0.027829, df = 1, p-value = 0.8675
predict.enet.test == testTransformed$pop 
sum(predict.enet.test == testTransformed$pop)/length(testTransformed$pop)
# TOP100_2018: x TRUE  = x%
# DMS_2018M: 23 TRUE  = 44%
```

### Variable contributions

```{r}
varImp.ridge <- caret::varImp(elastic_model, scale = FALSE)
plot(varImp.ridge, top = 28)
varImp.ridge.df <- varImp.ridge$importance
varImp.ridge.df$locus <- rownames(varImp.ridge.df)
varImp.ridge.names_Overall <- varImp.ridge.df$locus[varImp.ridge.df$Overall > 0.001]

all.model.names <- c(varImp.ridge.names.58.5.2, varImp.ridge.names.88.1,
                     varImp.ridge.names.58.4.1, varImp.ridge.names.58.4.2)
length(unique(all.model.names))#76
sum(duplicated(all.model.names))# 2


meth.RDA.gr <- as(meth.RDA, "GRanges")
x <- as.integer(gsub(pattern = "V", replacement = "", x = all.model.names))
x <- unique(x )
print(data.frame(loc.number = x, 
                 loc.name = paste0(meth.RDA$chr[x],"_",meth.RDA$start[x])))
meth.regional.gr <- meth.RDA.gr[x,]
meth.regional.raw <- meth.RDA[x,]
# meth.aging.prc <- methylKit::percMethylation(meth.regional.raw)
meth.diff.df <- as.data.frame(t(methylKit::percMethylation(meth.regional.raw)))
loc.names <- paste0(meth.regional.raw$chr,"_",meth.regional.raw$start)
colnames(meth.diff.df) <- loc.names

meth.diff.df$individual <- rownames(meth.diff.df)
print(load(paste0(rdata.path,"Toothfish_metadata.Rdata")))
identical(metadata2$SAMPLE_ID, meth.diff.df$individual)
meth.diff.df$pop <- metadata2$region

meth.diff.df.plot <- tidyr::gather(meth.diff.df, key = locus, 
                                   value = perc_methylation, -individual, -pop,
                                   factor_key = TRUE) %>%
  dplyr::mutate(chr = stringr::str_split_fixed(locus, 
                                               pattern = "_", 
                                               n = 2)[,1])
meth.diff.df.plot.locus <- meth.diff.df.plot %>% dplyr::group_by(chr,locus,pop) %>% 
  dplyr::summarise(mean_pop = mean(perc_methylation, na.rm = TRUE),
            .groups = 'drop')

# meth.diff.df.plot.centroid <- meth.diff.df.plot %>% dplyr::group_by(chr,pop) %>% 
#   dplyr::summarise(mean_pop = mean(perc_methylation, na.rm = TRUE),
#             .groups = 'drop')
# 
# meth.diff.df.plot.label <- meth.diff.df.plot %>% dplyr::group_by(chr) %>% 
#   dplyr::summarise(nmarkers = length(unique(locus)),
#             .groups = 'drop')



region.meth.plot <- ggplot2::ggplot() +
  ggplot2::geom_point(data = meth.diff.df.plot[meth.diff.df.plot$locus 
                                               %in%
                                                 unique(meth.diff.df.plot$locus)
                                               [1:5],],
                      ggplot2::aes(x = pop, y = perc_methylation, colour = locus)) +
  ggplot2::geom_line(data = meth.diff.df.plot.locus[meth.diff.df.plot.locus$locus 
                                             %in% 
                                               unique(meth.diff.df.plot$locus)
                                             [1:5],],
                     ggplot2::aes(x = pop, y = mean_pop, colour = locus, group = locus)) +
  ggplot2::geom_point(data = meth.diff.df.plot.locus[meth.diff.df.plot.locus$locus 
                                             %in% 
                                               unique(meth.diff.df.plot$locus)[1:5],],
                     ggplot2::aes(x = pop, y = mean_pop, colour = locus, group = locus),
                     size = 5)
print(region.meth.plot)


region.meth.plot2 <- ggplot2::ggplot() +
  ggplot2::geom_point(data = meth.diff.df.plot,
                      ggplot2::aes(x = pop, y = perc_methylation, colour = locus)) +
  ggplot2::geom_line(data = meth.diff.df.plot.locus,
                     ggplot2::aes(x = pop, y = mean_pop, colour = locus, group = locus)) +
  ggplot2::geom_point(data = meth.diff.df.plot.locus,
                     ggplot2::aes(x = pop, y = mean_pop, colour = locus, group = locus),
                     size = 5) +
  ggplot2::facet_wrap(~chr) + 
  ggplot2::guides(colour = ggplot2::guide_legend(ncol = 2))
print(region.meth.plot2)


ggplot2::ggsave(plot = region.meth.plot2,
                filename = paste0(figure.path,
                                  # "TOA_aging_methylation_percentage_enet05.png"
                                  "TOA_Regional_methylation_percentage_ridge.png"
                                  # "TOA_aging_methylation_percentage_all.models.png"
                                  ),
                                  width = 45, height = 30,
                                  units = "cm", device = "png")
```




# 8_Methylation: ALL muscle tissue

```{r}
print(load(paste0(rdata.path,"Toothfish_metadata.Rdata"))) #ALL fish
```

## Reading the methylation call files

```{r, eval=FALSE}
data.list <- list.files(
  path = 
  "/datasets/work/ncmi-toa-rrbs/work/1.Toothfish/GENOME/Combined_methylation/",
  pattern = "*.cov.gz")

subnames <- metadata$File_names[!(metadata$Tissue_type %in% 
                                    c("fin", "Skin") | # "muscle from under SD"
                                    grepl(pattern = "REP", x = metadata$File_names) 
                                  )]

subnames <- subnames[!subnames %in% "TOA_88-2_P79_AD294619_muscle_F"] # low coverage

data.list <- data.list[data.list %in% paste0(subnames,
                                             "_R1_trimmed_sort_bismark_bt2.bismark.cov.gz")]

data.loc <- as.list(paste0(
  "/datasets/work/ncmi-toa-rrbs/work/1.Toothfish/GENOME/Combined_methylation/",
                           data.list))



data.list2 <-  gsub(pattern = '_R1_trimmed_sort_bismark_bt2.bismark.cov.gz',
                    replacement = '',x = as.character(data.list))

metadata.new <- metadata[metadata$File_names %in% data.list2,]
metadata.new <- metadata.new[order(match(metadata.new$File_names, data.list2)),]

metadata.new$File_names == data.list2

sample.ids <- metadata.new$Genotype_names
pop <- as.integer(factor(metadata.new$pop))

myobjDB <- methylKit::methRead(location = data.loc,
           sample.id = as.list(sample.ids),
           assembly = "TOA_genome",
           treatment = pop,
           context = "CpG",
           header = FALSE, 
           dbtype = "tabix",
           pipeline = "bismarkCoverage",
           mincov = 3
           )
save(myobjDB, file = paste0(rdata.path,"Toothfish_ALL_muscle_methylkit_cpgDB.Rdata"))
```


## Data QC

#### Number of sites in genome

```{r, eval=FALSE}
cov.df <- data.frame(chr = NA, start = NA,end = NA, strand = NA, coverage = NA, 
                     numCs = NA, numTs = NA, sample = NA)
for (i in 1:length(sample.ids)) {
  df <- methylKit::getData(myobjDB[[i]])
  df$sample <- sample.ids[i]
  cov.df <- rbind(cov.df, df)
}
cov.df <- cov.df[-1,]
save(cov.df, file = paste0(rdata.path,"Toothfish_ALL_muscle_methylkit_cpg_cov.Rdata"))
```

### Coverage stats

```{r}
load(paste0(rdata.path,"Toothfish_ALL_muscle_methylkit_cpgDB.Rdata"))
print(load(paste0(rdata.path,"Toothfish_ALL_muscle_methylkit_cpg_cov.Rdata")))

length(unique(paste0(cov.df$chr, "_", cov.df$start))) #4,154,270

summary(cov.df$coverage)
    # Min.  1st Qu.   Median     Mean  3rd Qu.     Max. 
    # 4.00     8.00    17.00    25.65    34.00 51120.00 

summary(cov.df$coverage[cov.df$coverage > 5 & cov.df$coverage < 50])
   # Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
   #  6.0    10.0    17.0    20.1    28.0    49.0 

cov.plot <- ggplot2::ggplot(cov.df, ggplot2::aes(x = log(coverage), fill = sample)) +
  ggplot2::geom_histogram() +
  ggplot2::theme(legend.position = "none")
# print(cov.plot)
ggplot2::ggsave(cov.plot,
                filename =
                  paste0(figure.path,"Toothfish_ALL_muscle_coverage_hist_Combined.png"), 
                width = 45, height = 15, units = 'cm')


cov.plot <- ggplot2::ggplot(cov.df, ggplot2::aes(x = coverage, fill = sample)) +
  ggplot2::geom_histogram() +
  ggplot2::theme(legend.position = "none")
# print(cov.plot)
ggplot2::ggsave(cov.plot,
                filename =
                  paste0(figure.path,"Toothfish_ALL_muscle_coverage_hist_Combined2.png"), 
                width = 45, height = 15, units = 'cm')

cov.plot <- ggplot2::ggplot(cov.df[cov.df$coverage > 5 & cov.df$coverage < 160,], 
                            ggplot2::aes(x = coverage)) +
  ggplot2::geom_histogram(binwidth = 1) +
  ggplot2::ggtitle("coverage < 160") +
  ggplot2::theme(legend.position = "none")
# print(cov.plot)
ggplot2::ggsave(cov.plot,
                filename =
                  paste0(figure.path,"Toothfish_ALL_muscle_coverage_hist_Combined4.png"),
                width = 45, height = 15, units = 'cm')



max(cov.df$coverage, na.rm = T) #51120
ind.covplot <- ggplot2::ggplot(cov.df, ggplot2::aes(x = sample, y = coverage)) + 
  ggplot2::geom_boxplot() +
  ggplot2::theme_bw() + 
  ggplot2::theme(axis.title.x = ggplot2::element_blank(), 
                 axis.text.y = ggplot2::element_text(size = 15),
                 axis.title.y = ggplot2::element_text(size = 15),
                 axis.text.x = ggplot2::element_text(size = 7, angle = 90, hjust = 1)) +
  ggplot2::scale_y_continuous(trans = scales::log2_trans(),
                  breaks = c(1, 10, 50, 750, 30000, 60000),
                  minor_breaks = c(1:10, 2:5*10, 1:7.5*100, 0.5:5*1000))

print(ind.covplot)

ggplot2::ggsave(ind.covplot, bg = "white",
                filename = 
                  paste0(figure.path,
                         "Toothfish_ALL_muscle_raw_individual_coverage_CpGs.png"),
                width = 45, height = 15, units = 'cm')
```


### Coverage filter

High count filter based on coverage stats.

```{r, eval=FALSE}
load(paste0(rdata.path,"Toothfish_ALL_muscle_methylkit_cpgDB.Rdata"))

filtered.myobj <- methylKit::filterByCoverage(myobjDB, lo.count = 10, lo.perc = NULL,
                                              save.db = FALSE,
                                      hi.count = NULL, hi.perc = 99)

save(myobjDB,filtered.myobj,
     file = paste0(rdata.path,"Toothfish_ALL_muscle_methylkit_cpgDB.Rdata"))
```

### Normalise coverage

Normalize read coverage distributions between samples. Requires stringent filter on max coverage first, otherwise it will over-sample PCR bias reads. These two functions will help reduce the bias in the statistical tests that might occur due to systematic over-sampling of reads in certain samples.

```{r}
filtered.myobj.normal <- methylKit::normalizeCoverage(obj = filtered.myobj, 
                                                      save.db = TRUE,
                                                      method = "median", chunk.size = 1e6)
save(myobjDB,filtered.myobj, filtered.myobj.normal,
     file = paste0(rdata.path,"Toothfish_ALL_muscle_methylkit_cpgDB.Rdata"))
```

### Number of sites per sample

```{r, eval = FALSE}
sites.df <- data.frame(sites = NA,  sample = NA)
for (i in 1:length(filtered.myobj.normal)) {
  print(i)
  sites.df[i,] <- c(filtered.myobj.normal[[i]]@num.records, filtered.myobj.normal[[i]]@sample.id)
}
sites.df$sites <- as.integer(sites.df$sites)
sites.df$pop <- metadata.new$pop
save(sites.df, file = paste0(rdata.path,"Toothfish_ALL_muscle_number_cpg_sample.Rdata"))
```


```{r}
print(load(paste0(rdata.path,"Toothfish_ALL_muscle_number_cpg_sample.Rdata")))
sites.plot <- ggplot2::ggplot(sites.df, ggplot2::aes(x = sample, y = sites, colour = pop)) +
  ggplot2::geom_point() +
  ggplot2::scale_color_manual(values = colours6m) +
  ggplot2::theme_bw() +
  ggplot2::theme(axis.text.x = ggplot2::element_text(angle = 90))
print(sites.plot)

ggplot2::ggsave(sites.plot,
                filename =
                  paste0(figure.path,"Toothfish_ALL_m_number_cpg_sample.png"),
                width = 45, height = 15, units = 'cm')

sites.df$sample[sites.df$sites < 500000]
# "TOA_58-4-2_088_AN170741" "TOA_58-4-2_B18_AE100308_muscle_F" "TOA_58-4-2_B24_AE100883_muscle_F" "TOA_58-4-2_B28_AE100885_muscle_M" "TOA_58-4-2_B30_AE100886_muscle_M" "TOA_58-4-2_B36_AE102050_muscle_M" "TOA_58-4-2_B38_AE102052_muscle_M" "TOA_58-4-2_B42_AE102105_muscle_F" "TOA_58-4-2_B48_AE102108_muscle_M" "TOA_58-4-2_B58_AE104944_muscle_F" "TOA_88-2_P24_AD292328_muscle_F"
low.sites <- sites.df$sample[sites.df$sites < 500000]
```


## Select samples

```{r}
print(load(paste0(rdata.path,"Toothfish_metadata.Rdata")))

metadata2 <- metadata[!(metadata$Tissue_type %in% 
                                    c("fin", "Skin") | # "muscle from under SD"
                                    grepl(pattern = "REP", x = metadata$File_names) 
                                  ),]
print(load(paste0(rdata.path, "Low_quality_fish.Rdata")))


rm1 <- unique(c(low.genome.mapping,low.genome.methylation, low.sites))
print(rm1)
rm2 <- c("TOA_58-4-2_B40_AE102056_muscle_M") # bad PCA
# rm2 <- c("TOA_58-4-2_B18_AE100308_muscle_F", "TOA_58-4-2_B36_AE102050_muscle_M", "TOA_58-4-2_B38_AE102052_muscle_M", "TOA_58-4-2_B42_AE102105_muscle_F", "TOA_58-4-2_B48_AE102108_muscle_M", "TOA_58-4-2_B58_AE104944_muscle_F") # bad PCA

rm3 <- c("TOA_58-4-2_B24_AE100883_muscle_F", "TOA_88-2_P24_AD292328_muscle_F",
         "TOA_58-4-2_B64_AE107905_muscle_F") # bad coverage
rm4 <- c("TOA_48-2_031_UK116","TOA_58-4-2_111_AN172604","TOA_88-1_230_AD082543",
         "TOA_48-2_024_UK082") #bad tissue PCA

rm <- c(rm1,rm2,rm3,rm4) # bad coverage
metadata2 <- metadata2[!(metadata2$File_names %in% rm | metadata2$Genotype_names %in% rm),]

filt.sample.id.full <- sapply(filtered.myobj.normal@.Data, function(x){x@sample.id})
filt.sample.id <- filt.sample.id.full[filt.sample.id.full %in% metadata2$Genotype_names]
metadata2 <- metadata2[order(match(metadata2$Genotype_names, filt.sample.id)),]
identical(filt.sample.id,metadata2$Genotype_names)


pop <- metadata2$pop
pop[pop == "58.4.2-2024_50L"] <- "58.4.2-2024"
pop[pop == "58.4.2-2024_70L"] <- "58.4.2-2024"

pop.levels <- c("58.4.1-2018", "58.4.2-2018", "58.4.2-2024", "58.5.2-2017",
                     "88.1-2018", "88.2-2024")
pop <- as.numeric(factor(pop, levels = pop.levels))
sample.ids <- metadata2$Genotype_names

filtered.myobj2.normal@treatment <- pop


filtered.myobj2.normal <- methylKit::reorganize(
  methylObj = filtered.myobj.normal,
  sample.ids = sample.ids,
  treatment = pop,
  chunk.size = 1e+06,
  save.db = TRUE,
  suffix = "TOA_ALL_muscle_normal"
)
save(myobjDB,filtered.myobj,
     filtered.myobj.normal,filtered.myobj2.normal,
     file = paste0(rdata.path,"Toothfish_ALL_muscle_methylkit_cpgDB.Rdata"))
```


## Merging samples

```{r}
print(load(paste0(rdata.path,"Toothfish_ALL_muscle_methylkit_cpgDB.Rdata")))
pop <- metadata2$pop
pop[pop == "58.4.2-2024_50L"] <- "58.4.2-2024"
pop[pop == "58.4.2-2024_70L"] <- "58.4.2-2024"

pop.levels <- c("58.4.1-2018", "58.4.2-2018", "58.4.2-2024", "58.5.2-2017",
                     "88.1-2018", "88.2-2024")
pop <- as.numeric(factor(pop, levels = pop.levels))
table(pop)


meth.ALLM.norm15L <- methylKit::unite(filtered.myobj2.normal, destrand = FALSE,
                                      save.db = FALSE, 
                                      min.per.group = 15L, #missing data allowed
                                      mc.cores = 25)
#283,250  for 15L missing per group

save(meth.ALLM.norm15L,
     file = paste0(rdata.path,"Toothfish_ALL_muscle_methylkit_object.Rdata"))
```


## Data QC



### Remove ASM

```{r}
load(paste0(rdata.path,"Allele_specific_mutations.Rdata"))

mut <- as(asmTRUE,"GRanges")
sum(as(meth.ALLM.norm15L,"GRanges") %over% mut)#10619
meth.asm <- methylKit::selectByOverlap(meth.ALLM.norm15L, mut)   
BOOLEAN <- paste0(meth.ALLM.norm15L$chr,"_",meth.ALLM.norm15L$start) %in%
  paste0(asmTRUE$Chr,"_",asmTRUE$Pos) #1
sum(BOOLEAN) #221
meth.ALLM.norm15L <- methylKit::select(meth.ALLM.norm15L, !BOOLEAN) 

save(meth.ALLM.norm15L,
     file = paste0(rdata.path,"Toothfish_ALL_muscle_methylkit_object.Rdata"))
```



### Coverage 

```{r}
meth.df <- methylKit::getData(meth.ALLM.norm15L)
cov <- meth.df[,meth.ALLM.norm15L@coverage.index]
colnames(cov) <- meth.ALLM.norm15L@sample.ids
dpf <- data.frame(cov, locus = paste0(meth.df$chr,"_",meth.df$start))
dpf <- tidyr::gather(dpf, key = individual, 
                           value = Depth, -locus,
                           factor_key = TRUE) %>%
  dplyr::group_by(individual) %>%
  dplyr::mutate(NA_cnt = sum(is.na(Depth))/length(unique(locus))) %>%
  dplyr::ungroup()
dpf.NA <- dpf[!duplicated(dpf$individual),]
dpf.NA$NA_cnt <- round(dpf.NA$NA_cnt, 3)

dpf <- dpf[ dpf$Depth > 0 & !is.na(dpf$Depth),]

max(dpf$Depth, na.rm = T) #256
myPlot <- ggplot2::ggplot(dpf, ggplot2::aes(x = individual, y = Depth)) + 
  ggplot2::geom_boxplot() +
  ggplot2::geom_text(data = dpf.NA,ggplot2::aes(x = individual, y = 300, label = NA_cnt), 
                     size = 2, angle = 90) +
  ggplot2::theme_bw() + 
  ggplot2::theme(axis.title.x = ggplot2::element_blank(), 
                 axis.text.y = ggplot2::element_text(size = 15),
                 axis.title.y = ggplot2::element_text(size = 15),
                 axis.text.x = ggplot2::element_text(size = 7, angle = 90, hjust = 1)) +
  ggplot2::scale_y_continuous(trans = scales::log2_trans(),
                  breaks = c(1, 10, 50, 150, 300),
                  minor_breaks = c(1:10, 2:5*10, 1:2.5*100))

print(myPlot)
ggplot2::ggsave(myPlot, bg = "white",
                filename = 
                  paste0(figure.path,
                  "1B.Toothfish_ALL_muscle_normalised15_methylkit_individual_coverage_CpGs.png"
                                  ), 
                width = 45, height = 15, units = "cm")

nrow(cov) # 305129 SMPs
plot(x = 1:ncol(cov), y = colMeans(cov, na.rm = TRUE), col = meth.ALLM.norm15L@treatment)
plot(x = 1:ncol(cov), y = colMeans(cov > 20, na.rm = TRUE), col = meth.ALLM.norm15L@treatment)
plot(x = 1:ncol(cov), y = colMeans(cov > 10, na.rm = TRUE), col = meth.ALLM.norm15L@treatment)
meth.ALLM.norm15L@sample.ids[colMeans(cov > 10, na.rm = TRUE) < 0.99]
# samples with low coverage: 
```


## Load data

```{r}
print(load(paste0(rdata.path,"Toothfish_ALL_muscle_methylkit_object.Rdata")))
print(load(paste0(rdata.path,"Toothfish_metadata.Rdata"))) 

metadata.new <- metadata[metadata$Genotype_names %in% meth.ALLM.norm15L@sample.ids,]
metadata.new <- metadata.new[order(match(metadata.new$Genotype_names,
                                         meth.ALLM.norm15L@sample.ids)),]

length(meth.ALLM.norm15L@row.names) # number of sites
```


## Global clustering analysis

```{r}
methylKit::clusterSamples(meth.ALLM.norm15L, dist = "correlation", sd.filter = TRUE, 
                          method = "ward.D2", plot = TRUE, )
dev.print(file = paste0(figure.path,"1_TOA_ALL_muscle_cluster_plot.png"),
          device = png, res = 300,width = 45,height = 15,units = "cm")
```


### PCA

```{r}
PCobj <- methylKit::PCASamples(meth.ALLM.norm15L, comp = c(1,2), 
                               adj.lim = c(0,0), obj.return = TRUE)

var_frac <- PCobj$sdev/sum(PCobj$sdev)
print(paste0("Total variance explained: ",signif(sum(var_frac[1:3]) * 100, 3), "%")) 
barplot(var_frac[1:20], main = "variance explained", 
        col = heat.colors(length(var_frac[1:20])))

identical(metadata.new$Genotype_names, meth.ALLM.norm15L@sample.ids)
identical(metadata.new$Genotype_names, rownames(PCobj$x))

pop <- metadata.new$pop
pool <- metadata.new$STRATA4

plot.title <- "Toothfish methylation - ALL muscle (259,441 CpG sites)"

data <- data.frame(ind = rownames(PCobj$x),PCobj$x,  pop = pop, pool = pool)
data <- merge(data, aggregate(cbind(mean.x = PC1, mean.y = PC2) ~
                                pop, data, mean), by = "pop")

pca.plot <- ggplot2::ggplot(data) +
  ggplot2::geom_point(ggplot2::aes(PC1, PC2, color = pop, shape = pool),
                      size = 4) + 
  ggplot2::stat_ellipse(ggplot2::aes(
    x = PC1, y = PC2,color = pop, group = pop),
    type = "norm") +
  ggplot2::geom_point(ggplot2::aes(x = mean.x,
                                   y = mean.y, color = pop),
                      size = 7,
                      shape = 17) +
  ggplot2::scale_shape_manual(values = c(19,18)) +
  ggplot2::scale_colour_manual(values = colours6m) +
  ggplot2::geom_hline(yintercept = 0) +
  ggplot2::geom_vline(xintercept = 0) + 
  ggplot2::labs(
    subtitle = paste0("Total variance explained: ",
                      signif(sum(var_frac[1:4]) * 100, 3), "%"),
    y = paste0("PC2; ", signif(var_frac[2] * 100, digits = 2), " %"),
    x = paste0("PC1; ", signif(var_frac[1] * 100, digits = 2), " %"),
    title = plot.title,
    caption = "") + 
  ggplot2::theme_minimal() +
  ggplot2::theme(
    axis.text = ggplot2::element_text(size = 15),
    axis.title.x = ggplot2::element_text(size = 20),
    axis.title.y = ggplot2::element_text(size = 20),
    legend.text = ggplot2::element_text(size = 15)
  )

print(pca.plot)
ggplot2::ggsave(pca.plot, bg = "white",
                filename = paste0(figure.path,"Toothfish_ALL_muscle_PCA_CpG_PC12.png"), 
                width = 30, height = 15, units = "cm")




data <- data.frame(ind = rownames(PCobj$x),PCobj$x,  pop = pop, pool = pool)
data <- merge(data, aggregate(cbind(mean.x = PC2, mean.y = PC3) ~
                                pop, data, mean), by = "pop")
pca.plot <- ggplot2::ggplot(data) +
  ggplot2::geom_point(ggplot2::aes(PC2, PC3, color = pop, shape = pool),
                      size = 4) + 
  ggplot2::geom_point(ggplot2::aes(x = mean.x,
                                   y = mean.y, color = pop),
                      size = 7,
                      shape = 17) +
  ggplot2::scale_shape_manual(values = c(19,18)) +
  ggplot2::scale_colour_manual(values = colours6m) +
  ggplot2::geom_hline(yintercept = 0) +
  ggplot2::geom_vline(xintercept = 0) + 
  ggplot2::labs(
    subtitle = paste0("Total variance explained: ",
                      signif(sum(var_frac[1:4]) * 100, 3), "%"),
    y = paste0("PC3; ", signif(var_frac[3] * 100, digits = 2), " %"),
    x = paste0("PC2; ", signif(var_frac[2] * 100, digits = 2), " %"),
    title = plot.title,
    caption = "") + 
  ggplot2::theme_minimal() +
  ggplot2::theme(
    axis.text = ggplot2::element_text(size = 15),
    axis.title.x = ggplot2::element_text(size = 20),
    axis.title.y = ggplot2::element_text(size = 20),
    legend.text = ggplot2::element_text(size = 15)
  )

print(pca.plot)
ggplot2::ggsave(pca.plot, bg = "white",
                filename = paste0(figure.path,"Toothfish_ALL_muscle_PCA_CpG_PC23.png"), 
                width = 30, height = 15, units = "cm")
```


### RDA

```{r}
# setwd("/datasets/work/ncmi-toa-rrbs/work/1.Toothfish/Toothfish_RRBS_array_22098078/Bismark/")
# print(load(paste0(rdata.path,"Toothfish_methylation_objects_raw.Rdata"))) #36,635 SMPs
print(load(paste0(rdata.path,"Toothfish_subset_methylation_objects_raw.Rdata"))) #1,316,799 SMPs

# meth.RDA <- methylKit::unite(filtered.myobj2.norm, destrand = FALSE,
#                              save.db = FALSE,
#                              suffix = "subset",
#                              min.per.group = NULL, #missing data allowed
#                              mc.cores = 25)


# meth.df <- methylKit::getData(meth)
# cov <- meth.df[,grep(colnames(meth.df), pattern = "coverage*")]
# cov <- meth.df[,meth@coverage.index]
# colnames(cov) <- meth@sample.ids
# meth.df <- methylKit::percMethylation(meth)

# meth.df <- methylKit::percMethylation(meth.RDA)

# save(meth.df,meth.RDA, 
#      file = paste0(rdata.path,"TOA_data_for_RDA_meth_percent.Rdata"))
print(load(paste0(rdata.path,"TOA_data_for_RDA_meth_percent.Rdata"))) #347,614 SMPs

hist(colSums(is.na(meth.df)), breaks = 100)
hist(rowSums(is.na(meth.df)), breaks = 100)





print(load(paste0(rdata.path,"Toothfish_metadata.Rdata")))

predictor <- metadata2 # define the predictor input file here
predictor <- predictor[predictor$SAMPLE_ID %in% meth@sample.ids,]
predictor$SAMPLE_ID = as.character(predictor$SAMPLE_ID)
predictor$region = as.character(predictor$region)
predictor$Sex = as.character(predictor$Sex)
predictor$Tissue = as.character(predictor$Tissue)
predictor$mat = as.character(predictor$mat)
predictor$mat[is.na(predictor$mat)] <- "2"
predictor$TL = as.numeric(predictor$TL)
predictor$wt = as.numeric(predictor$wt)
predictor$dpth = as.numeric(predictor$dpth)
predictor$Age = as.numeric(predictor$Age)
predictor$Age[is.na(predictor$Age)] <- as.numeric(max(predictor$Age, na.rm = TRUE))
predictor$Cohort2 = as.integer(predictor$Cohort2)
predictor$Cohort2[is.na(predictor$Cohort2)] <- as.integer(min(predictor$Cohort2, na.rm = TRUE))


meth.df1 = t(meth.df)
# meth.df2 = na.omit(meth.df1) #rda can't handle NAs, remove them if you have them
# meth.df1[1:10,1:10]
## Predictor
# Make names characters (not factors)


#if you had numbers, they had to be factors (as.factor)

# Confirm that methylation data and predictors are in the same order
identical(rownames(meth.df1), predictor$SAMPLE_ID)

# Perform Hellinger transformation on the methylation data:
meth.df3 = vegan::decostand(meth.df1, method = "hellinger", MARGIN = 1) #MARGIN=1 means rows, on rows I have samples now because I transposed the matrix

## Run RDA
#Test the country using growing condition as a Condition () (covariate):
meth_rda = vegan::rda(meth.df3 ~ region + Condition(Sex), 
                      data = predictor, scale = F) #I remove scale since I performed Hellinger transformation
meth_rda = vegan::rda(meth.df3 ~ region + Condition(Sex) + 
                        Condition(mat)+ Condition(TL) + Condition(wt) +
                        Condition(dpth) + Condition(Age) + Condition(Cohort2), 
                      na.action = "na.omit",
                      data = predictor, scale = F)

meth_rda$pCCA
meth_rda$CCA$wa

summary(meth_rda) #Constrained Proportion: variance of Y explained by X, Unconstrained Proportion: unexplained variance in Y
# sink(paste0(rdata.path,"anova_rda.txt")) #with this you can save the output of the anova in a txt file, define the output directory
meth.aov <- anova(meth_rda, test = "Chi", permutations = 99) #run Anova on the result to see the significance level
# meth.aov.cca <- anova.cca(meth_rda, test = "Chi", permutations = 99)
print(meth.aov)
# sink(paste0(rdata.path,"anova_rda_chi.txt"))

# Model: rda(formula = meth.df3 ~ region + Condition(Sex), data = predictor, scale = F)
#           Df  Variance      F Pr(>F)   
# Model      3 0.0003924 1.4126  0.004 **
# Residual 110 0.0101861

# Model: rda(formula = meth.df3 ~ region + Condition(Sex) + Condition(mat) + Condition(TL) + Condition(wt) + Condition(dpth) + Condition(Age) + Condition(Cohort2), data = predictor, scale = F, na.action = "na.omit")
#           Df  Variance     F Pr(>F)   
# Model      3 0.0003412 1.257   0.01 **
# Residual 101 0.0091396 

Pop.names <- predictor$region
plot.a <- vegan::ordiplot(meth_rda, choices=c(1,2))
Pop.ellipses <- vegan::ordiellipse(plot.a, groups=Pop.names, display="sites", kind="sd")
dev.print(file = paste0(figure.path,"TOA_RDA_ordiplot_complex_model.png"),
          device = png, res = 300,width = 45,height = 15,units = "cm")

# Type 1 scaling
vegan::ordiplot(meth_rda, scaling = 1, type = "text")
# Type 2 scaling
vegan::ordiplot(meth_rda, scaling = 2, type = "text")

ggvegan::autoplot(
  meth_rda,
  axes = c(1, 2),
  geom = "text",
  layers = c("species", "sites", "biplot", "centroids"),
  arrows = TRUE)

#CpG contribution
hist(meth_rda$CCA$v[,1])
hist(meth_rda$CCA$v[,2])
hist(meth_rda$CCA$v[,3])



#Test the interaction of your predictors:
meth_rda3 = vegan::rda(meth ~ region * Sex + Condition(region + Sex), data = predictor, scale = F)
summary(meth_rda3)
sink(paste0(rdata.path,"anova_rda_interaction.txt"))
print(anova(meth_rda3, test="Chi"))
sink(paste0(rdata.path,"anova_rda_interaction_chi.txt"))
```

#### plot RDA

```{r}

BioR.theme <- theme(
        panel.background = element_blank(),
        panel.border = element_blank(),
        panel.grid = element_blank(),
        axis.line = element_line("gray25"),
        text = element_text(size = 12),
        axis.text = element_text(size = 10, colour = "gray25"),
        axis.title = element_text(size = 14, colour = "gray25"),
        legend.title = element_text(size = 14),
        legend.text = element_text(size = 14),
        legend.key = element_blank())

plot.a <- vegan::ordiplot(meth_rda, choices=c(1,2))
Pop.names <- predictor$region

# sites1 <- BiodiversityR::sites.long(plot.a, env.data=predictor)
sites1 <- data.frame(cbind(predictor, axis1=plot.a$sites[, 1], 
                           axis2=plot.a$sites[, 2], 
                           labels=rownames(plot.a$sites)),
                     group = predictor$region)

# axis1 <- BiodiversityR::axis.long(meth_rda, choices=c(1, 2))
eigs.all <- meth_rda$CA$eig
eigs <- round(100 * eigs.all / meth_rda$tot.chi, digits=1)[c(1, 2)] 
xlab.label <- paste0(names(eigs)[1], " (", eigs[1], "%)")
ylab.label <- paste0(names(eigs)[2], " (", eigs[2], "%)")
axis1 <- data.frame(axis=c(1:2), 
                    ggplot=c("xlab.label", "ylab.label"), 
                    label=c(xlab.label, ylab.label))


# centroids.Pop1 <- BiodiversityR::centroids.long(sites1, 
#                                                 grouping=Pop.names, 
#                                                 centroids.only=TRUE)
# centroids.Pop2 <- BiodiversityR::centroids.long(sites1, 
#                                                 grouping=Pop.names, 
#                                                 centroids.only=FALSE)
gr.name <- rlang::enexpr(Pop.names)
cent.means <-  stats::aggregate(cbind(axis1, axis2) ~ Pop.names,
                                data = sites1,
                                FUN = mean)
names(cent.means) <- c("group", "axis1c", "axis2c")
centroids.Pop1 <- cent.means
cent.means$Centroid <- cent.means[, 1]
by.name <- names(sites1)[names(sites1) == "group"]
centroids.Pop2 <- dplyr::full_join(sites1, cent.means, by = by.name)


# Pop.ellipses.long2 <- BiodiversityR::ordiellipse.long(Pop.ellipses, 
#                                                     grouping.name="Pop.names")


plotgg.a <- ggplot() + 
    geom_vline(xintercept = c(0), color = "grey70", linetype = 2) +
    geom_hline(yintercept = c(0), color = "grey70", linetype = 2) +  
    xlab(axis1[1, "label"]) +
    ylab(axis1[2, "label"]) +  
    scale_x_continuous(sec.axis = dup_axis(labels=NULL, name=NULL)) +
    scale_y_continuous(sec.axis = dup_axis(labels=NULL, name=NULL)) +
    # geom_polygon(data=Pop.ellipses.long2, 
    #                aes(x=axis1, y=axis2, colour=Pop.names, 
    #                    fill=after_scale(alpha(colour, 0.2))), 
    #           size=0.2, show.legend=TRUE) +
    geom_segment(data=centroids.Pop2, 
                 aes(x=axis1c, y=axis2c, xend=axis1, yend=axis2, colour=Pop.names), 
                 size=0.7, show.legend=FALSE) +
    geom_point(data=centroids.Pop1, 
               aes(x=axis1c, y=axis2c), 
               shape="square", colour="black", alpha=0.7, size=2) +
    geom_label_repel(data=centroids.Pop1, 
               aes(x=axis1c, y=axis2c, label=group, colour=group), 
               alpha=1.0, size=2, show.legend=FALSE) +
    # labs(colour = Pop.names) +
    BioR.theme +
    theme(legend.text = element_text(size = 10)) +
    scale_colour_npg() +
    coord_fixed(ratio=1)

plotgg.a

ggsave(plotgg.a, file = paste0(figure.path, 
                               # "TOA_RDA_simple_model_A.png"
                               "TOA_RDA_complex_model_A.png"
                               ),
       height = 30, width = 30, units = "cm",  device = "png")
```



## DMS - ALL - muscle

```{r, eval = FALSE}
print(load(paste0(rdata.path,"Toothfish_ALL_muscle_methylkit_object.Rdata")))

print(load(paste0(rdata.path,"Toothfish_metadata.Rdata")))
metadata.new <- metadata[metadata$Genotype_names %in% attributes(meth.ALLM.norm15L)$sample.ids,]
attributes(meth.ALLM.norm15L)$sample.ids == metadata.new$Genotype_names

# meth.ALLM.norm15L@treatment <- as.integer(factor(metadata.new$pop))
meth.ALLM.norm15L@treatment <- as.integer(factor(metadata.new$STRATA2))

# data.frame(attributes(meth.ALLM.norm15L)$sample.ids, metadata.new$individual, metadata.new$strata, meth.ALLM.norm15L@treatment)

# covariates <- dplyr::select(metadata.new, Sex, Age, Cohort2, Maturity_full)
metadata.new$batch_effect <- metadata.new$Year
metadata.new$batch_effect[metadata.new$batch_effect == 2017] <- 2018

covariates <- dplyr::select(metadata.new, batch_effect)

myDiff <- methylKit::calculateDiffMeth(meth.ALLM.norm15L,
                                       covariates = covariates,
                                       adjust = "fdr",
                                       chunk.size = 1e+06,
                                       mc.cores = 25)
meth.diff.chr <- methylKit::diffMethPerChr(myDiff,plot = TRUE, 
                                           qvalue.cutoff = 0.01, meth.cutoff = 40)
myDiff40p <- methylKit::getMethylDiff(myDiff, difference = 40, qvalue = 0.001)
#379 for pop
#2 for strata2
myDiff25p <- methylKit::getMethylDiff(myDiff, difference = 25, qvalue = 0.001)
#2467 for pop
#49 for strata2
myDiff15p <- methylKit::getMethylDiff(myDiff, difference = 15, qvalue = 0.001)
#11746 for pop
#868 for strata2
top100.ALLM <- slice_max(as.data.frame(as(myDiff,"GRanges")), order_by = meth.diff, n = 100)
top100.ALLM <- as(top100.ALLM,"GRanges")

save(myDiff,myDiff15p, myDiff25p, myDiff40p,top100.ALLM,
     file = paste0(rdata.path,"Toothfish_ALL_muscle_methylation_DMS.Rdata"))
```


```{r}
print(load(paste0(rdata.path,"Toothfish_ALL_muscle_methylation_DMS.Rdata")))

volcano.plot <- ggplot2::ggplot() +
  ggplot2::geom_point(data = myDiff, ggplot2::aes(x = meth.diff, 
                                                  y = -log10(qvalue))) +
  ggplot2::geom_point(data = myDiff25p, ggplot2::aes(x = meth.diff, 
                                                  y = -log10(qvalue)),
                      colour = "red") +
  ggplot2::theme_minimal() +
  ggplot2::theme(
    axis.text = ggplot2::element_text(size = 15),
    axis.title.x = ggplot2::element_text(size = 20),
    axis.title.y = ggplot2::element_text(size = 20),
    legend.text = ggplot2::element_text(size = 15)
  )
print(volcano.plot)

ggplot2::ggsave(volcano.plot, bg = "white",
                filename =
                  paste0(figure.path,
                         "Toothfish_ALL_muscle_DMS25_VOLCANO_methylkit_regions.png"), 
                width = 30, height = 15, units = "cm")




myDiff25p.gr <- as(myDiff25p, "GRanges")
grand.plot <- ggbio::plotGrandLinear(myDiff25p.gr,
                                     ggplot2::aes(y = meth.diff)) +
  ggplot2::ylab("Methylation difference (%)") +
  ggplot2::theme_bw() +
  ggplot2::theme(axis.text.x = ggplot2::element_text(angle = 90, hjust = 0), 
                 legend.position = "none")
print(grand.plot)
ggplot2::ggsave(grand.plot@ggplot, bg = "white",
                filename = paste0(figure.path,
                                  "Toothfish_ALL_muscle_DMS25_Grandplot_methylkit_regions.png"
                ), 
                width = 30, height = 15, units = "cm")
```

### Heatmap

```{r}
print(load(paste0(rdata.path,"Toothfish_ALL_muscle_methylation_DMS.Rdata")))

myDiff25p.gr <- as(myDiff25p, "GRanges")
meth.diff <- methylKit::selectByOverlap(meth.ALLM.norm15L, myDiff25p.gr)

meth.diff.df <- as.data.frame(t(methylKit::percMethylation(meth.diff)))
loc.names <- paste0(meth.diff$chr,"_",meth.diff$start)
colnames(meth.diff.df) <- loc.names

meth.diff.df$individual <- rownames(meth.diff.df)

pop <- metadata.new$pop
meth.diff.df$pop <- as.character(pop)


heat.plot.data <- meth.diff.df %>%
  tibble::as_tibble() %>%
  tidyr::gather(key = "CpG", value = "Methylation_percentage", -individual, -pop,
                factor_key = FALSE) %>%
  dplyr::group_by(pop, CpG) %>%
  dplyr::mutate(CpG_mean = mean(Methylation_percentage, na.rm = TRUE)) %>%
  dplyr::arrange(pop, individual, CpG) %>%
  dplyr::mutate(individual = factor(individual, levels = unique(individual))) %>%
  dplyr::mutate(CpG = factor(CpG, levels = unique(CpG)))
cpgs <- heat.plot.data$CpG[heat.plot.data$individual == unique(heat.plot.data$individual)[1]]
cpgs_mean <- heat.plot.data$CpG_mean[heat.plot.data$individual == unique(heat.plot.data$individual)[1]]

cpgs <- cpgs[order(cpgs_mean, decreasing = TRUE)]
heat.plot.data$CpG <- factor(heat.plot.data$CpG, levels = c("pop", unique(as.character(cpgs))))
heat.plot.data <- heat.plot.data[order(heat.plot.data$CpG),]

heat.plot <- ggplot2::ggplot(data = heat.plot.data, ggplot2::aes(x = individual)) + 
    ggplot2::geom_tile(data = heat.plot.data, 
                       ggplot2::aes(x = individual, y = CpG, fill = Methylation_percentage)) +
    ggplot2::scale_fill_gradient2(low = "blue", mid = "yellow", high = "red", 
                                  midpoint = 50, guide = "colourbar") +
    ggnewscale::new_scale_fill() +
    ggplot2::geom_tile(data = heat.plot.data, ggplot2::aes(x = individual, y = "pop", fill = pop,
                                    height = length(unique(heat.plot.data$CpG))*0.02)) +
    ggplot2::scale_fill_manual(values = colours6m) +
    ggplot2::scale_y_discrete(drop = FALSE) +
    ggplot2::theme(legend.position = "top",
                   axis.title.y = ggplot2::element_blank(),
                   axis.title.x = ggplot2::element_blank(),
                   axis.text.x = ggplot2::element_text(face = "bold", color = "black", 
                                        size = 10, angle = 90, hjust = 1, vjust = 0.5))
print(heat.plot)

ggplot2::ggsave(heat.plot, bg = "white",
                filename = paste0(figure.path,
                                  "Toothfish_ALL_muscle_methDiff_25_HEATMAP.png"), 
                width = 30, height = 30, units = "cm")
```

### Plot CpG difference

```{r}
load(file = paste0(rdata.path,"Toothfish_ALL_muscle_methylation_DMS.Rdata"))

myDiff25p.gr <- as(myDiff25p, "GRanges")
meth.diff <- methylKit::selectByOverlap(meth.ALLM.norm15L, myDiff25p.gr) #113 

length(unique(meth.diff$chr)) #34
sort(table(meth.diff$chr), decreasing = TRUE)[1:20]

meth.diff.df <- as.data.frame(t(methylKit::percMethylation(meth.diff)))
loc.names <- paste0(meth.diff$chr,"_",meth.diff$start)
colnames(meth.diff.df) <- loc.names

meth.diff.df$individual <- rownames(meth.diff.df) 
identical(metadata.new$Genotype_names, meth.diff.df$individual)
meth.diff.df$pop <- metadata.new$pop



meth.diff.df.plot <- tidyr::gather(meth.diff.df, key = locus, 
                           value = perc_methylation, -individual, -pop,
                           factor_key = TRUE) %>%
  dplyr::mutate(chr = stringr::str_split_fixed(locus, 
                                               pattern = "_", 
                                               n = 2)[,1])
meth.diff.df.plot.locus <- meth.diff.df.plot %>% dplyr::group_by(chr,locus,pop) %>% 
  dplyr::summarise(mean_pop = mean(perc_methylation, na.rm = TRUE),
                   sum_NA = sum(is.na(perc_methylation)),
            .groups = 'drop')


diff.plot <- ggplot2::ggplot() +
  ggplot2::geom_line(data = meth.diff.df.plot.locus,
                     ggplot2::aes(x = pop, y = mean_pop, colour = locus, group = locus)) +
  ggplot2::geom_point(data = meth.diff.df.plot.locus,
                     ggplot2::aes(x = pop, y = mean_pop, colour = locus, group = locus),
                     size = 5) +
  ggplot2::facet_wrap(~chr) +
  ggplot2::ylab("Mean methylation (%)") +
  ggplot2::theme(legend.position = "none",
                   axis.text.x = ggplot2::element_text(face = "bold", color = "black", 
                                        size = 10, angle = 90, hjust = 1, vjust = 0.5))

print(diff.plot)

ggplot2::ggsave(plot = diff.plot, 
                filename = paste0(figure.path,"1_Toothfish_ALL_muscle_methDiff.png"),
                width = 30, height = 15, units = "cm", device = "png")
```


### DMS Clustering
```{r}
load(file = paste0(rdata.path,"Toothfish_ALL_muscle_methylation_DMS.Rdata"))

myDiff25p.gr <- as(myDiff25p, "GRanges")
meth.diff <- methylKit::selectByOverlap(meth.ALLM.norm15L, myDiff25p.gr) #112  
methdif.perc <- methylKit::percMethylation(meth.diff)

methdif.perc2 <- zoo::na.aggregate(methdif.perc, by = meth.ALLM.norm15L@treatment)
methdif.perc2[is.na(methdif.perc2)] <- 50
PCobj <- prcomp(methdif.perc2, retx = TRUE, scale. = TRUE)

var_frac <- PCobj$sdev/sum(PCobj$sdev)
print(paste0("Total variance explained: ",signif(sum(var_frac[1:3]) * 100, 3), "%")) 
barplot(var_frac[1:3], main = "variance explained", 
        col = heat.colors(length(var_frac[1:3])))


identical(metadata.new$Genotype_names, meth.ALLM.norm15L@sample.ids)
identical(metadata.new$Genotype_names, rownames(PCobj$rotation))

pop <- metadata.new$pop
pool <- metadata.new$STRATA4

plot.title <- "Toothfish methylation - ALL muscle (49 DMS sites - 25%)"

data <- data.frame(ind = rownames(PCobj$rotation),PCobj$rotation,  
                   pop = pop, pool = pool)
data <- merge(data, aggregate(cbind(mean.x = PC1, mean.y = PC2) ~
                                pop, data, mean), by = "pop")

pca.plot <- ggplot2::ggplot(data) +
  ggplot2::geom_point(ggplot2::aes(PC1, PC2, color = pop, shape = pool),
                                   size = 4) + 
  ggplot2::geom_point(ggplot2::aes(x = mean.x,
                                   y = mean.y, color = pop),
                      size = 7,
                      shape = 17) +
  ggplot2::scale_shape_manual(values = c(19,18)) +
  ggplot2::scale_colour_manual(values = colours6m) +
  ggplot2::geom_hline(yintercept = 0) +
  ggplot2::geom_vline(xintercept = 0) + 
  ggplot2::labs(
    subtitle = paste0("Total variance explained: ",
                      signif(sum(var_frac[1:3]) * 100, 3), "%"),
    y = paste0("PC2; ", signif(var_frac[2] * 100, digits = 2), " %"),
    x = paste0("PC1; ", signif(var_frac[1] * 100, digits = 2), " %"),
    title = plot.title,
    caption = "") + 
  ggplot2::theme_minimal() +
  ggplot2::theme(
    axis.text = ggplot2::element_text(size = 15),
    axis.title.x = ggplot2::element_text(size = 20),
    axis.title.y = ggplot2::element_text(size = 20),
    legend.text = ggplot2::element_text(size = 15)
  )

print(pca.plot)
ggplot2::ggsave(pca.plot, bg = "white",
                filename = paste0(figure.path,"Toothfish_ALL_muscle_PCA_CpG_DMS_PC12.png"), 
                width = 30, height = 15, units = "cm")



data <- data.frame(ind = rownames(PCobj$rotation),PCobj$rotation,  
                   pop = pop, pool = pool)
data <- merge(data, aggregate(cbind(mean.x = PC1, mean.y = PC3) ~
                                pop, data, mean), by = "pop")

pca.plot <- ggplot2::ggplot(data) +
  ggplot2::geom_point(ggplot2::aes(PC1, PC3, color = pop, shape = pool),
                      size = 4) + 
  ggplot2::geom_point(ggplot2::aes(x = mean.x,
                                   y = mean.y, color = pop),
                      size = 7,
                      shape = 17) +
  ggplot2::scale_shape_manual(values = c(19,18)) +
  ggplot2::scale_colour_manual(values = colours6m) +
  ggplot2::geom_hline(yintercept = 0) +
  ggplot2::geom_vline(xintercept = 0) + 
  ggplot2::labs(
    subtitle = paste0("Total variance explained: ",
                      signif(sum(var_frac[1:3]) * 100, 3), "%"),
    y = paste0("PC3; ", signif(var_frac[3] * 100, digits = 2), " %"),
    x = paste0("PC1; ", signif(var_frac[1] * 100, digits = 2), " %"),
    title = plot.title,
    caption = "") + 
  ggplot2::theme_minimal() +
  ggplot2::theme(
    axis.text = ggplot2::element_text(size = 15),
    axis.title.x = ggplot2::element_text(size = 20),
    axis.title.y = ggplot2::element_text(size = 20),
    legend.text = ggplot2::element_text(size = 15)
  )

print(pca.plot)
ggplot2::ggsave(pca.plot, bg = "white",
                filename = paste0(figure.path,
                                  "Toothfish_ALL_muscle_PCA_CpG_DMS_PC13.png"), 
                width = 30, height = 15, units = "cm")




data <- data.frame(ind = rownames(PCobj$rotation),PCobj$rotation,  pop = pop, pool = pool)
data <- merge(data, aggregate(cbind(mean.x = PC2, mean.y = PC3) ~
                                pop, data, mean), by = "pop")

pca.plot <- ggplot2::ggplot(data) +
  ggplot2::geom_point(ggplot2::aes(PC2, PC3, color = pop, shape = pool),
                      size = 4) + 
  ggplot2::geom_point(ggplot2::aes(x = mean.x,
                                   y = mean.y, color = pop),
                      size = 7,
                      shape = 17) +
  ggplot2::scale_shape_manual(values = c(19,18)) +
  ggplot2::scale_colour_manual(values = colours6m) +
  ggplot2::geom_hline(yintercept = 0) +
  ggplot2::geom_vline(xintercept = 0) + 
  ggplot2::labs(
    subtitle = paste0("Total variance explained: ",
                      signif(sum(var_frac[1:3]) * 100, 3), "%"),
    y = paste0("PC3; ", signif(var_frac[3] * 100, digits = 2), " %"),
    x = paste0("PC2; ", signif(var_frac[2] * 100, digits = 2), " %"),
    title = plot.title,
    caption = "") + 
  ggplot2::theme_minimal() +
  ggplot2::theme(
    axis.text = ggplot2::element_text(size = 15),
    axis.title.x = ggplot2::element_text(size = 20),
    axis.title.y = ggplot2::element_text(size = 20),
    legend.text = ggplot2::element_text(size = 15)
  )

print(pca.plot)
ggplot2::ggsave(pca.plot, bg = "white",
                filename = paste0(figure.path,"Toothfish_ALL_muscle_PCA_CpG_DMS_PC23.png"), 
                width = 30, height = 15, units = "cm")



data <- data.frame(ind = rownames(PCobj$rotation),PCobj$rotation,  pop = pop, pool = pool)
data <- merge(data, aggregate(cbind(mean.x = PC4, mean.y = PC3) ~
                                pop, data, mean), by = "pop")

pca.plot <- ggplot2::ggplot(data) +
  ggplot2::geom_point(ggplot2::aes(PC4, PC3, color = pop, shape = pool),
                      size = 4) + 
  ggplot2::stat_ellipse(ggplot2::aes(
    x = PC4, y = PC3,color = pop, group = pop),
    type = "norm", level = 0.95) +
  ggplot2::geom_point(ggplot2::aes(x = mean.x,
                                   y = mean.y, color = pop),
                      size = 7,
                      shape = 17) +
  ggplot2::scale_shape_manual(values = c(19,18)) +
  ggplot2::scale_colour_manual(values = colours6m) +
  ggplot2::geom_hline(yintercept = 0) +
  ggplot2::geom_vline(xintercept = 0) + 
  ggplot2::labs(
    subtitle = paste0("Total variance explained: ",
                      signif(sum(var_frac[1:4]) * 100, 3), "%"),
    y = paste0("PC3; ", signif(var_frac[3] * 100, digits = 2), " %"),
    x = paste0("PC4; ", signif(var_frac[4] * 100, digits = 2), " %"),
    title = plot.title,
    caption = "") + 
  ggplot2::theme_minimal() +
  ggplot2::theme(
    axis.text = ggplot2::element_text(size = 15),
    axis.title.x = ggplot2::element_text(size = 20),
    axis.title.y = ggplot2::element_text(size = 20),
    legend.text = ggplot2::element_text(size = 15)
  )

print(pca.plot)
ggplot2::ggsave(pca.plot, bg = "white",
                filename = paste0(figure.path,"Toothfish_ALL_muscle_PCA_CpG_DMS_PC34.png"), 
                width = 30, height = 15, units = "cm")
```


### Karyogram

```{r}
chr.file <- "/datasets/work/ncmi-toa-rrbs/work/1.Toothfish/Bismark_genome_index/sizes.chr"
chrom.info <- read.table(file = chr.file, header = F)
chrom.info$V3 <- FALSE
colnames(chrom.info) <- c("chrom", "length", "is_circular")
chrom.info2 <- chrom.info[chrom.info$chrom %in% paste0("JAAKFY0100000", stringr::str_pad(1:27, 2, pad = "0")),]

chrom.info3 <- chrom.info2
seqnames <- gsub(pattern = "JAAKFY01", replacement = "", as.character(chrom.info3$chrom))
seqnames <- paste0("chr", stringr::str_pad(as.integer(seqnames), 2, pad = "0"))
chrom.info3$chrom <- seqnames

chrom.seqinfo <- GenomeInfoDb::Seqinfo(seqnames = chrom.info3$chrom, 
                                       seqlengths = chrom.info3$length, 
                                       isCircular = chrom.info3$is_circular, 
                                       genome = "Antarctic Toothfish")

gr <- GRanges(seqnames = chrom.info2$chrom,ranges = IRanges(start = c(0,0), end = chrom.info2$length))

meth.sub <- methylKit::selectByOverlap(meth.ALLM.norm15L, gr) #1202  
meth.sub.gr <- as(meth.sub, "GRanges")
seqnames <- gsub(pattern = "JAAKFY01", replacement = "", as.character(meth.sub.gr@seqnames))
seqnames <- paste0("chr", stringr::str_pad(as.integer(seqnames), 2, pad = "0"))
seqlevels(meth.sub.gr) <- as.character(unique(seqnames))

myDiff25p.sub <- methylKit::selectByOverlap(myDiff25p, gr) #1202  
myDiff25p.sub.gr <- as(myDiff25p.sub, "GRanges")
seqnames <- gsub(pattern = "JAAKFY01", replacement = "", as.character(myDiff25p.sub.gr@seqnames))
seqnames <- paste0("chr", stringr::str_pad(as.integer(seqnames), 2, pad = "0"))
seqlevels(myDiff25p.sub.gr) <- as.character(unique(seqnames))


karyo <- GRanges(seqnames = meth.sub.gr@seqnames,
                 ranges = IRanges::ranges(meth.sub.gr))
karyo.diff <- GRanges(seqnames = myDiff25p.sub.gr@seqnames,
                 ranges = IRanges::ranges(myDiff25p.sub.gr))

p <- ggbio::autoplot(chrom.seqinfo, layout = "karyogram", cytobands = FALSE) + 
  ggbio::layout_karyogram(karyo, geom = "rect", color = "blue") +
  ggbio::layout_karyogram(karyo.diff, geom = "rect", color = "red")
print(p)

ggplot2::ggsave(p@ggplot, bg = "white",
                filename = paste0(figure.path,
                                  "Toothfish_ALL_muscle_ALL_karyogram.png"
                ), 
                width = 45, height = 30, units = "cm")

```



## DMS RANDOM - ALL - muscle

```{r,eval=FALSE}
print(load(paste0(rdata.path,"Toothfish_ALL_muscle_methylkit_object.Rdata")))

print(load(paste0(rdata.path,"Toothfish_metadata.Rdata")))
metadata.new <- metadata[metadata$Genotype_names %in% attributes(meth.ALLM.norm15L)$sample.ids,]
attributes(meth.ALLM.norm15L)$sample.ids == metadata.new$Genotype_names

strata.random <- sample(metadata.new$pop, size = length(metadata.new$pop), replace = FALSE)
meth.ALLM.norm15L@treatment <- as.integer(factor(strata.random))

# covariates <- dplyr::select(metadata.new, Sex, TL_mm)
myDiff <- methylKit::calculateDiffMeth(meth.ALLM.norm15L,
                                       covariates = NULL,
                                       adjust = "fdr",
                                       chunk.size = 1e+06,
                                       mc.cores = 25)
meth.diff.chr <- methylKit::diffMethPerChr(myDiff,plot = TRUE, 
                                           qvalue.cutoff = 0.01, meth.cutoff = 40)
myDiff40p <- methylKit::getMethylDiff(myDiff, difference = 40, qvalue = 0.01)
#1
myDiff25p <- methylKit::getMethylDiff(myDiff, difference = 25, qvalue = 0.01)
#101
myDiff15p <- methylKit::getMethylDiff(myDiff, difference = 15, qvalue = 0.01)
#1700     

save(strata.random, myDiff,myDiff15p, myDiff25p, myDiff40p,
     file = paste0(rdata.path,"Toothfish_ALL_muscle_methylation_RANDOM_DMS.Rdata"))
```


```{r}
print(load(paste0(rdata.path,"Toothfish_ALL_muscle_methylation_RANDOM_DMS.Rdata")))

volcano.plot <- ggplot2::ggplot() +
  ggplot2::geom_point(data = myDiff, ggplot2::aes(x = meth.diff, 
                                                  y = -log10(qvalue))) +
  ggplot2::geom_point(data = myDiff25p, ggplot2::aes(x = meth.diff, 
                                                  y = -log10(qvalue)),
                      colour = "red") +
  
  ggplot2::theme_minimal() +
  ggplot2::theme(
    axis.text = ggplot2::element_text(size = 15),
    axis.title.x = ggplot2::element_text(size = 20),
    axis.title.y = ggplot2::element_text(size = 20),
    legend.text = ggplot2::element_text(size = 15)
  )
print(volcano.plot)

ggplot2::ggsave(volcano.plot, bg = "white",
                filename =
                  paste0(figure.path,
                         "Toothfish_ALL_muscle_RANDOM_DMS25_VOLCANO_methylkit.png"), 
                width = 30, height = 15, units = "cm")
```


### Heatmap

```{r}
print(load(paste0(rdata.path,"Toothfish_ALL_muscle_methylation_RANDOM_DMS.Rdata")))

myDiff25p.gr <- as(myDiff25p, "GRanges")
meth.diff <- methylKit::selectByOverlap(meth.ALLM.norm15L, myDiff25p.gr)

meth.diff.df <- as.data.frame(t(methylKit::percMethylation(meth.diff)))
loc.names <- paste0(meth.diff$chr,"_",meth.diff$start)
colnames(meth.diff.df) <- loc.names

meth.diff.df$individual <- rownames(meth.diff.df)
meth.diff.df$pop <- strata.random
meth.diff.df$pop2 <- as.character(metadata.new$pop)


heat.plot.data <- meth.diff.df %>%
  tibble::as_tibble() %>%
  tidyr::gather(key = "CpG", value = "Methylation_percentage", -individual, -pop, -pop2,
                factor_key = FALSE) %>%
  dplyr::group_by(pop, CpG) %>%
  dplyr::mutate(CpG_mean = mean(Methylation_percentage, na.rm = TRUE)) %>%
  dplyr::arrange(pop, individual, CpG) %>%
  dplyr::mutate(individual = factor(individual, levels = unique(individual))) %>%
  dplyr::mutate(CpG = factor(CpG, levels = unique(CpG)))
cpgs <- heat.plot.data$CpG[heat.plot.data$individual == unique(heat.plot.data$individual)[1]]
cpgs_mean <- heat.plot.data$CpG_mean[heat.plot.data$individual == unique(heat.plot.data$individual)[1]]

cpgs <- cpgs[order(cpgs_mean, decreasing = TRUE)]
heat.plot.data$CpG <- factor(heat.plot.data$CpG, levels = c("pop", unique(as.character(cpgs))))
heat.plot.data <- heat.plot.data[order(heat.plot.data$CpG),]

heat.plot <- ggplot2::ggplot(data = heat.plot.data, ggplot2::aes(x = individual)) + 
    ggplot2::geom_tile(data = heat.plot.data, 
                       ggplot2::aes(x = individual, y = CpG, fill = Methylation_percentage)) +
    ggplot2::scale_fill_gradient2(low = "blue", mid = "yellow", high = "red", 
                                  midpoint = 50, guide = "colourbar") +
    ggnewscale::new_scale_fill() +
    ggplot2::geom_tile(data = heat.plot.data, ggplot2::aes(x = individual, 
                                                           y = "pop - random", fill = pop,
                                    height = length(unique(heat.plot.data$CpG))*0.02)) +
    ggplot2::scale_fill_manual(values = colours6m) +
    ggnewscale::new_scale_fill() +
    ggplot2::geom_tile(data = heat.plot.data, ggplot2::aes(x = individual, y = "pop", fill = pop2,
                                    height = length(unique(heat.plot.data$CpG))*0.02)) +
    ggplot2::scale_fill_manual(values = colours6m) +
    ggplot2::scale_y_discrete(drop = FALSE) +
    ggplot2::theme(legend.position = "top",
                   axis.title.y = ggplot2::element_blank(),
                   axis.title.x = ggplot2::element_blank(),
                   axis.text.x = ggplot2::element_text(face = "bold", color = "black", 
                                        size = 10, angle = 90, hjust = 1, vjust = 0.5))
print(heat.plot)

ggplot2::ggsave(heat.plot, bg = "white",
                filename = paste0(figure.path,
                                  "Toothfish_ALL_muscle_RANDOM_methDiff_25_HEATMAP.png"), 
                width = 30, height = 30, units = "cm")
```

### Plot CpG difference

```{r}
load(file = paste0(rdata.path,"Toothfish_ALL_muscle_methylation_RANDOM_DMS.Rdata"))

myDiff25p.gr <- as(myDiff25p, "GRanges")
meth.diff <- methylKit::selectByOverlap(meth.ALLM.norm15L, myDiff25p.gr) #113 

length(unique(meth.diff$chr)) #34
sort(table(meth.diff$chr), decreasing = TRUE)[1:20]

meth.diff.df <- as.data.frame(t(methylKit::percMethylation(meth.diff)))
loc.names <- paste0(meth.diff$chr,"_",meth.diff$start)
colnames(meth.diff.df) <- loc.names

meth.diff.df$individual <- rownames(meth.diff.df) 
identical(metadata.new$Genotype_names, meth.diff.df$individual)
meth.diff.df$pop <- metadata.new$pop


meth.diff.df.plot <- tidyr::gather(meth.diff.df, key = locus, 
                           value = perc_methylation, -individual, -pop,
                           factor_key = TRUE) %>%
  dplyr::mutate(chr = stringr::str_split_fixed(locus, 
                                               pattern = "_", 
                                               n = 2)[,1])
meth.diff.df.plot.locus <- meth.diff.df.plot %>% dplyr::group_by(chr,locus,pop) %>% 
  dplyr::summarise(mean_pop = mean(perc_methylation, na.rm = TRUE),
                   sum_NA = sum(is.na(perc_methylation)),
            .groups = 'drop')


diff.plot <- ggplot2::ggplot() +
  ggplot2::geom_line(data = meth.diff.df.plot.locus,
                     ggplot2::aes(x = pop, y = mean_pop, colour = locus, group = locus)) +
  ggplot2::geom_point(data = meth.diff.df.plot.locus,
                     ggplot2::aes(x = pop, y = mean_pop, colour = locus, group = locus),
                     size = 5) +
  ggplot2::facet_wrap(~chr) +
  ggplot2::ylab("Mean methylation (%)") +
  ggplot2::theme(legend.position = "none",
                   axis.text.x = ggplot2::element_text(face = "bold", color = "black", 
                                        size = 10, angle = 90, hjust = 1, vjust = 0.5))
    # ggplot2::guides(colour = ggplot2::guide_legend(ncol = 2))

print(diff.plot)

ggplot2::ggsave(plot = diff.plot, 
                filename = paste0(figure.path,"1_Toothfish_ALL_muscle_RANDOM_methDiff.png"),
                width = 30, height = 15, units = "cm", device = "png")
```


### DMS Clustering
```{r}
load(file = paste0(rdata.path,"Toothfish_ALL_muscle_methylation_RANDOM_DMS.Rdata"))

myDiff25p.gr <- as(myDiff25p, "GRanges")
meth.diff <- methylKit::selectByOverlap(meth.ALLM.norm15L, myDiff25p.gr) #270  
methdif.perc <- methylKit::percMethylation(meth.diff)

methdif.perc2 <- zoo::na.aggregate(methdif.perc, by = meth.ALLM.norm15L@treatment)
methdif.perc2[is.na(methdif.perc2)] <- 50
PCobj <- prcomp(methdif.perc2, retx = TRUE, scale. = TRUE) # Too many missing values

var_frac <- PCobj$sdev/sum(PCobj$sdev)
print(paste0("Total variance explained: ",signif(sum(var_frac[1:3]) * 100, 3), "%")) 
barplot(var_frac[1:3], main = "variance explained", 
        col = heat.colors(length(var_frac[1:3])))


identical(metadata.new$Genotype_names, meth.ALLM.norm15L@sample.ids)

identical(as.numeric(factor(metadata.new$pop)), meth.ALLM.norm15L@treatment)

# identical(metadata.new$individual, rownames(PCobj$x))
identical(metadata.new$Genotype_names, rownames(PCobj$rotation))
pop <- metadata.new$pop
pool <- metadata.new$STRATA4

plot.title <- "Toothfish methylation - RANDOMISED ALL muscle (276 DMS sites - 25%)"


data <- data.frame(ind = rownames(PCobj$rotation),PCobj$rotation,  
                   pop = pop, pool = pool)
data <- merge(data, aggregate(cbind(mean.x = PC1, mean.y = PC2) ~
                                pop, data, mean), by = "pop")

pca.plot <- ggplot2::ggplot(data) +
  ggplot2::geom_point(ggplot2::aes(PC1, PC2, color = pop, shape = pool),
                      size = 4) + 
  ggplot2::geom_point(ggplot2::aes(x = mean.x,
                                   y = mean.y, color = pop),
                      size = 7,
                      shape = 17) +
  ggplot2::scale_shape_manual(values = c(19,18)) +
  ggplot2::scale_colour_manual(values = colours6m) +
  ggplot2::geom_hline(yintercept = 0) +
  ggplot2::geom_vline(xintercept = 0) + 
  ggplot2::labs(
    subtitle = paste0("Total variance explained: ",
                      signif(sum(var_frac[1:3]) * 100, 3), "%"),
    y = paste0("PC2; ", signif(var_frac[2] * 100, digits = 2), " %"),
    x = paste0("PC1; ", signif(var_frac[1] * 100, digits = 2), " %"),
    title = plot.title,
    caption = "") + 
  ggplot2::theme_minimal() +
  ggplot2::theme(
    axis.text = ggplot2::element_text(size = 15),
    axis.title.x = ggplot2::element_text(size = 20),
    axis.title.y = ggplot2::element_text(size = 20),
    legend.text = ggplot2::element_text(size = 15)
  )

print(pca.plot)
ggplot2::ggsave(pca.plot, bg = "white",
                filename = 
                  paste0(figure.path,"Toothfish_ALL_muscle_RANDOM_PCA_CpG_DMS_PC12.png"), 
                width = 30, height = 15, units = "cm")



data <- data.frame(ind = rownames(PCobj$rotation),PCobj$rotation, 
                   pop = pop, pool = pool)
data <- merge(data, aggregate(cbind(mean.x = PC1, mean.y = PC3) ~
                                pop, data, mean), by = "pop")

pca.plot <- ggplot2::ggplot(data, ggplot2::aes(PC1, PC3, color = pop, shape = pool)) +
  ggplot2::geom_point(size = 4) + 
  ggplot2::geom_point(ggplot2::aes(x = mean.x,
                                   y = mean.y, color = pop),
                      size = 7,
                      shape = 17) +
  ggplot2::scale_shape_manual(values = c(19,18)) +
  ggplot2::scale_colour_manual(values = colours6m) +
  ggplot2::geom_hline(yintercept = 0) +
  ggplot2::geom_vline(xintercept = 0) + 
  ggplot2::labs(
    subtitle = paste0("Total variance explained: ",
                      signif(sum(var_frac[1:3]) * 100, 3), "%"),
    y = paste0("PC3; ", signif(var_frac[3] * 100, digits = 2), " %"),
    x = paste0("PC1; ", signif(var_frac[1] * 100, digits = 2), " %"),
    title = plot.title,
    caption = "") + 
  ggplot2::theme_minimal() +
  ggplot2::theme(
    axis.text = ggplot2::element_text(size = 15),
    axis.title.x = ggplot2::element_text(size = 20),
    axis.title.y = ggplot2::element_text(size = 20),
    legend.text = ggplot2::element_text(size = 15)
  )

print(pca.plot)
ggplot2::ggsave(pca.plot, bg = "white",
                filename = paste0(figure.path,"Toothfish_ALL_muscle_RANDOM_PCA_CpG_DMS_PC13.png"), 
                width = 30, height = 15, units = "cm")




data <- data.frame(ind = rownames(PCobj$rotation),PCobj$rotation, 
                   pop = pop, shape = pool)
data <- merge(data, aggregate(cbind(mean.x = PC2, mean.y = PC3) ~
                                pop, data, mean), by = "pop")

pca.plot <- ggplot2::ggplot(data) +
  ggplot2::geom_point(ggplot2::aes(PC2, PC3, color = pop, shape = pool), 
                      size = 4) + 
  ggplot2::geom_point(ggplot2::aes(x = mean.x,
                                   y = mean.y, color = pop),
                      size = 7,
                      shape = 17) +
  ggplot2::scale_shape_manual(values = c(19,18)) +
  ggplot2::scale_colour_manual(values = colours6m) +
  ggplot2::geom_hline(yintercept = 0) +
  ggplot2::geom_vline(xintercept = 0) + 
  ggplot2::labs(
    subtitle = paste0("Total variance explained: ",
                      signif(sum(var_frac[1:3]) * 100, 3), "%"),
    y = paste0("PC3; ", signif(var_frac[3] * 100, digits = 2), " %"),
    x = paste0("PC2; ", signif(var_frac[2] * 100, digits = 2), " %"),
    title = plot.title,
    caption = "") + 
  ggplot2::theme_minimal() +
  ggplot2::theme(
    axis.text = ggplot2::element_text(size = 15),
    axis.title.x = ggplot2::element_text(size = 20),
    axis.title.y = ggplot2::element_text(size = 20),
    legend.text = ggplot2::element_text(size = 15)
  )

print(pca.plot)
ggplot2::ggsave(pca.plot, bg = "white",
                filename = 
                  paste0(figure.path,"Toothfish_ALL_muscle_RANDOM_PCA_CpG_DMS_PC23.png"), 
                width = 30, height = 15, units = "cm")
```



## TOP100 DMS - ALL - muscle

```{r}
print(load(paste0(rdata.path,"Toothfish_ALL_muscle_methylkit_object.Rdata")))
print(load(paste0(rdata.path,"Toothfish_metadata.Rdata"))) 
metadata.new <- metadata[metadata$Genotype_names %in% meth.ALLM.norm15L@sample.ids,]
metadata.new <- metadata.new[order(match(metadata.new$Genotype_names,
                                         meth.ALLM.norm15L@sample.ids)),]
meth.cpg <- paste0(meth.ALLM.norm15L$chr, "_", meth.ALLM.norm15L$start) #601560 
length(meth.cpg)
print(load(paste0(rdata.path,"Toothfish_ALL_muscle_methylation_DMS.Rdata")))
DMS.cpg <- paste0(as.character(top100.ALLM@seqnames), "_", as.data.frame(top100.ALLM@ranges)$start) #283819 
sum(meth.cpg %in% DMS.cpg) #100
meth.DMS <- meth.ALLM.norm15L[meth.cpg %in% DMS.cpg,]
```


### Heatmap

```{r}
meth.diff.df <- as.data.frame(t(methylKit::percMethylation(meth.DMS)))
loc.names <- paste0(meth.DMS$chr,"_",meth.DMS$start)
colnames(meth.diff.df) <- loc.names

meth.diff.df$individual <- rownames(meth.diff.df)
pop <- metadata.new$pop
meth.diff.df$pop <- as.character(pop)


heat.plot.data <- meth.diff.df %>%
  tibble::as_tibble() %>%
  tidyr::gather(key = "CpG", value = "Methylation_percentage", -individual, -pop,
                factor_key = FALSE) %>%
  dplyr::group_by(pop, CpG) %>%
  dplyr::mutate(CpG_mean = mean(Methylation_percentage, na.rm = TRUE)) %>%
  dplyr::arrange(pop, individual, CpG) %>%
  dplyr::mutate(individual = factor(individual, levels = unique(individual))) %>%
  dplyr::mutate(CpG = factor(CpG, levels = unique(CpG)))
cpgs <- heat.plot.data$CpG[heat.plot.data$individual == unique(heat.plot.data$individual)[1]]
cpgs_mean <- heat.plot.data$CpG_mean[heat.plot.data$individual == unique(heat.plot.data$individual)[1]]

cpgs <- cpgs[order(cpgs_mean, decreasing = TRUE)]
heat.plot.data$CpG <- factor(heat.plot.data$CpG, levels = c("pop", unique(as.character(cpgs))))
heat.plot.data <- heat.plot.data[order(heat.plot.data$CpG),]

heat.plot <- ggplot2::ggplot(data = heat.plot.data, ggplot2::aes(x = individual)) + 
    ggplot2::geom_tile(data = heat.plot.data, 
                       ggplot2::aes(x = individual, y = CpG, fill = Methylation_percentage)) +
    ggplot2::scale_fill_gradient2(low = "blue", mid = "yellow", high = "red", 
                                  midpoint = 50, guide = "colourbar") +
    ggnewscale::new_scale_fill() +
    ggplot2::geom_tile(data = heat.plot.data, ggplot2::aes(x = individual, y = "pop", fill = pop,
                                    height = length(unique(heat.plot.data$CpG))*0.02)) +
    ggplot2::scale_fill_manual(values = colours6m) +
    ggplot2::scale_y_discrete(drop = FALSE) +
    ggplot2::theme(legend.position = "top",
                   axis.title.y = ggplot2::element_blank(),
                   axis.title.x = ggplot2::element_blank(),
                   axis.text.x = ggplot2::element_text(face = "bold", color = "black", 
                                        size = 10, angle = 90, hjust = 1, vjust = 0.5))
print(heat.plot)

ggplot2::ggsave(heat.plot, bg = "white",
                filename = paste0(figure.path,
                                  "Toothfish_ALL_muscle_BASED_ON_ALL_TOP100_DMS_HEATMAP.png"), 
                width = 30, height = 30, units = "cm")
```

### Plot CpG difference

```{r}
meth.diff <- meth.DMS

length(unique(meth.diff$chr)) #28
sort(table(meth.diff$chr), decreasing = TRUE)[1:20]

meth.diff.df <- as.data.frame(t(methylKit::percMethylation(meth.diff)))
loc.names <- paste0(meth.diff$chr,"_",meth.diff$start)
colnames(meth.diff.df) <- loc.names

meth.diff.df$individual <- rownames(meth.diff.df) 
identical(metadata.new$Genotype_names, meth.diff.df$individual)
meth.diff.df$pop <- metadata.new$pop



meth.diff.df.plot <- tidyr::gather(meth.diff.df, key = locus, 
                           value = perc_methylation, -individual, -pop,
                           factor_key = TRUE) %>%
  dplyr::mutate(chr = stringr::str_split_fixed(locus, 
                                               pattern = "_", 
                                               n = 2)[,1])
meth.diff.df.plot.locus <- meth.diff.df.plot %>% dplyr::group_by(chr,locus,pop) %>% 
  dplyr::summarise(mean_pop = mean(perc_methylation, na.rm = TRUE),
                   sum_NA = sum(is.na(perc_methylation)),
            .groups = 'drop')


diff.plot <- ggplot2::ggplot() +
  ggplot2::geom_line(data = meth.diff.df.plot.locus,
                     ggplot2::aes(x = pop, y = mean_pop, colour = locus, group = locus)) +
  ggplot2::geom_point(data = meth.diff.df.plot.locus,
                     ggplot2::aes(x = pop, y = mean_pop, colour = locus, group = locus),
                     size = 5) +
  ggplot2::facet_wrap(~chr) +
  ggplot2::ylab("Mean methylation (%)") +
  ggplot2::theme(legend.position = "none",
                   axis.text.x = ggplot2::element_text(face = "bold", color = "black", 
                                        size = 10, angle = 90, hjust = 1, vjust = 0.5))

print(diff.plot)

ggplot2::ggsave(plot = diff.plot, 
                filename = paste0(figure.path,"1_Toothfish_ALL_muscle_BASED_ON_ALL_TOP100_DMS_methDiff.png"),
                width = 30, height = 15, units = "cm", device = "png")
```



### DMS Clustering
```{r}
methdif.perc <- methylKit::percMethylation(meth.DMS)
methdif.perc2 <- zoo::na.aggregate(methdif.perc, by = meth.DMS@treatment)
methdif.perc2[is.na(methdif.perc2)] <- 50
PCobj <- prcomp(methdif.perc2, retx = TRUE, scale. = TRUE, center = TRUE)

var_frac <- PCobj$sdev/sum(PCobj$sdev)
print(paste0("Total variance explained: ",signif(sum(var_frac[1:3]) * 100, 3), "%")) 
barplot(var_frac[1:3], main = "variance explained", 
        col = heat.colors(length(var_frac[1:3])))

# 
identical(metadata.new$Genotype_names, meth.DMS@sample.ids)
identical(metadata.new$Genotype_names, rownames(PCobj$rotation))
# identical(metadata.new$File_names, meth.DMS@sample.ids)
# identical(metadata.new$File_names, rownames(PCobj$rotation))
pop <- metadata.new$pop

plot.title <- "Toothfish methylation - ALL Muscle (TOP100 DMS from ALL_M)"

data <- data.frame(ind = rownames(PCobj$rotation),PCobj$rotation,  pop = pop)
data <- merge(data, aggregate(cbind(mean.x = PC1, mean.y = PC2) ~
                                pop, data, mean), by = "pop")

pca.plot <- ggplot2::ggplot(data, ggplot2::aes(PC1, PC2, color = pop)) +
  ggplot2::geom_point(size = 4) + 
  ggplot2::geom_point(ggplot2::aes(x = mean.x,
                                   y = mean.y, color = pop),
                      size = 7,
                      shape = 17) +
  ggplot2::scale_colour_manual(values = colours6m) +
  ggplot2::geom_hline(yintercept = 0) +
  ggplot2::geom_vline(xintercept = 0) + 
  ggplot2::labs(
    subtitle = paste0("Total variance explained: ",
                      signif(sum(var_frac[1:3]) * 100, 3), "%"),
    y = paste0("PC2; ", signif(var_frac[2] * 100, digits = 2), " %"),
    x = paste0("PC1; ", signif(var_frac[1] * 100, digits = 2), " %"),
    title = plot.title,
    caption = "") + 
  ggplot2::theme_minimal() +
  ggplot2::theme(
    axis.text = ggplot2::element_text(size = 15),
    axis.title.x = ggplot2::element_text(size = 20),
    axis.title.y = ggplot2::element_text(size = 20),
    legend.text = ggplot2::element_text(size = 15)
  )

print(pca.plot)
ggplot2::ggsave(pca.plot, bg = "white",
                filename = paste0(figure.path,"Toothfish_ALL_muscle_BASED_ON_ALL_TOP100_DMS_PC12.png"), 
                width = 30, height = 15, units = "cm")



data <- data.frame(ind = rownames(PCobj$rotation),PCobj$rotation,  pop = pop)
data <- merge(data, aggregate(cbind(mean.x = PC1, mean.y = PC3) ~
                                pop, data, mean), by = "pop")

pca.plot <- ggplot2::ggplot(data, ggplot2::aes(PC1, PC3, color = pop)) +
  ggplot2::geom_point(size = 4) + 
  ggplot2::geom_point(ggplot2::aes(x = mean.x,
                                   y = mean.y, color = pop),
                      size = 7,
                      shape = 17) +
  ggplot2::scale_colour_manual(values = colours6m) +
  ggplot2::geom_hline(yintercept = 0) +
  ggplot2::geom_vline(xintercept = 0) + 
  ggplot2::labs(
    subtitle = paste0("Total variance explained: ",
                      signif(sum(var_frac[1:3]) * 100, 3), "%"),
    y = paste0("PC3; ", signif(var_frac[3] * 100, digits = 2), " %"),
    x = paste0("PC1; ", signif(var_frac[1] * 100, digits = 2), " %"),
    title = plot.title,
    caption = "") + 
  ggplot2::theme_minimal() +
  ggplot2::theme(
    axis.text = ggplot2::element_text(size = 15),
    axis.title.x = ggplot2::element_text(size = 20),
    axis.title.y = ggplot2::element_text(size = 20),
    legend.text = ggplot2::element_text(size = 15)
  )

print(pca.plot)
ggplot2::ggsave(pca.plot, bg = "white",
                filename = paste0(figure.path,
                                  "Toothfish_ALL_muscle_BASED_ON_ALL_TOP100_DMS_PC13.png"), 
                width = 30, height = 15, units = "cm")




data <- data.frame(ind = rownames(PCobj$rotation),PCobj$rotation,  pop = pop)
data <- merge(data, aggregate(cbind(mean.x = PC2, mean.y = PC3) ~
                                pop, data, mean), by = "pop")

pca.plot <- ggplot2::ggplot(data, ggplot2::aes(PC2, PC3, color = pop)) +
  ggplot2::geom_point(size = 4) + 
  ggplot2::geom_point(ggplot2::aes(x = mean.x,
                                   y = mean.y, color = pop),
                      size = 7,
                      shape = 17) +
  ggplot2::scale_colour_manual(values = colours6m) +
  ggplot2::geom_hline(yintercept = 0) +
  ggplot2::geom_vline(xintercept = 0) + 
  ggplot2::labs(
    subtitle = paste0("Total variance explained: ",
                      signif(sum(var_frac[1:3]) * 100, 3), "%"),
    y = paste0("PC3; ", signif(var_frac[3] * 100, digits = 2), " %"),
    x = paste0("PC2; ", signif(var_frac[2] * 100, digits = 2), " %"),
    title = plot.title,
    caption = "") + 
  ggplot2::theme_minimal() +
  ggplot2::theme(
    axis.text = ggplot2::element_text(size = 15),
    axis.title.x = ggplot2::element_text(size = 20),
    axis.title.y = ggplot2::element_text(size = 20),
    legend.text = ggplot2::element_text(size = 15)
  )

print(pca.plot)
ggplot2::ggsave(pca.plot, bg = "white",
                filename = paste0(figure.path,"Toothfish_ALL_muscle_BASED_ON_ALL_TOP100_DMS_PC23.png"), 
                width = 30, height = 15, units = "cm")
```




## ALL fish muscle with DMS from 2017/2018
```{r}
print(load(paste0(rdata.path,"Toothfish_DMS_CpGsites.Rdata")))
meth.cpg <- paste0(meth.ALLM.norm15L$chr, "_", meth.ALLM.norm15L$start) #601560 
DMS.cpg <- unique(c(DMS$DMS)) #1243
sum(meth.cpg %in% DMS.cpg) #230 = 18%
meth.DMS <- meth.ALLM.norm15L[meth.cpg %in% DMS.cpg,] 
```

### Heatmap

```{r}
meth.diff.df <- as.data.frame(t(methylKit::percMethylation(meth.DMS)))
loc.names <- paste0(meth.DMS$chr,"_",meth.DMS$start)
colnames(meth.diff.df) <- loc.names

meth.diff.df$individual <- rownames(meth.diff.df)
pop <- metadata.new$pop
meth.diff.df$pop <- as.character(pop)


heat.plot.data <- meth.diff.df %>%
  tibble::as_tibble() %>%
  tidyr::gather(key = "CpG", value = "Methylation_percentage", -individual, -pop,
                factor_key = FALSE) %>%
  dplyr::group_by(pop, CpG) %>%
  dplyr::mutate(CpG_mean = mean(Methylation_percentage, na.rm = TRUE)) %>%
  dplyr::arrange(pop, individual, CpG) %>%
  dplyr::mutate(individual = factor(individual, levels = unique(individual))) %>%
  dplyr::mutate(CpG = factor(CpG, levels = unique(CpG)))
cpgs <- heat.plot.data$CpG[heat.plot.data$individual == unique(heat.plot.data$individual)[1]]
cpgs_mean <- heat.plot.data$CpG_mean[heat.plot.data$individual == unique(heat.plot.data$individual)[1]]

cpgs <- cpgs[order(cpgs_mean, decreasing = TRUE)]
heat.plot.data$CpG <- factor(heat.plot.data$CpG, levels = c("pop", unique(as.character(cpgs))))
heat.plot.data <- heat.plot.data[order(heat.plot.data$CpG),]

heat.plot <- ggplot2::ggplot(data = heat.plot.data, ggplot2::aes(x = individual)) + 
    ggplot2::geom_tile(data = heat.plot.data, 
                       ggplot2::aes(x = individual, y = CpG, fill = Methylation_percentage)) +
    ggplot2::scale_fill_gradient2(low = "blue", mid = "yellow", high = "red", 
                                  midpoint = 50, guide = "colourbar") +
    ggnewscale::new_scale_fill() +
    ggplot2::geom_tile(data = heat.plot.data, ggplot2::aes(x = individual, y = "pop", fill = pop,
                                    height = length(unique(heat.plot.data$CpG))*0.02)) +
    ggplot2::scale_fill_manual(values = colours6m) +
    ggplot2::scale_y_discrete(drop = FALSE) +
    ggplot2::theme(legend.position = "top",
                   axis.title.y = ggplot2::element_blank(),
                   axis.title.x = ggplot2::element_blank(),
                   axis.text.x = ggplot2::element_text(face = "bold", color = "black", 
                                        size = 10, angle = 90, hjust = 1, vjust = 0.5))
print(heat.plot)

ggplot2::ggsave(heat.plot, bg = "white",
                filename = paste0(figure.path,
                                  "Toothfish_ALL_muscle_BASED_ON_2018_DMS_HEATMAP.png"), 
                width = 30, height = 30, units = "cm")
```


### DMS Clustering
```{r}
methdif.perc <- methylKit::percMethylation(meth.DMS)
methdif.perc2 <- zoo::na.aggregate(methdif.perc, by = meth.DMS@treatment)
PCobj <- prcomp(methdif.perc2, retx = TRUE, scale. = TRUE)

var_frac <- PCobj$sdev/sum(PCobj$sdev)
print(paste0("Total variance explained: ",signif(sum(var_frac[1:3]) * 100, 3), "%")) 
barplot(var_frac[1:3], main = "variance explained", 
        col = heat.colors(length(var_frac[1:3])))


identical(metadata.new$Genotype_names, meth.DMS@sample.ids)
# identical(metadata.new$individual, rownames(PCobj$x))
identical(metadata.new$Genotype_names, rownames(PCobj$rotation))
pop <- metadata.new$pop

plot.title <- "Toothfish methylation - ALL Muscle (230 DMS from 2018 - 25%)"

# data <- data.frame(ind = rownames(PCobj$x),PCobj$x,  pop = pop)
data <- data.frame(ind = rownames(PCobj$rotation),PCobj$rotation,  pop = pop)
data <- merge(data, aggregate(cbind(mean.x = PC1, mean.y = PC2) ~
                                pop, data, mean), by = "pop")

pca.plot <- ggplot2::ggplot(data, ggplot2::aes(PC1, PC2, color = pop)) +
  ggplot2::geom_point(size = 4) + 
  ggplot2::geom_point(ggplot2::aes(x = mean.x,
                                   y = mean.y, color = pop),
                      size = 7,
                      shape = 17) +
  ggplot2::scale_colour_manual(values = colours6m) +
  ggplot2::geom_hline(yintercept = 0) +
  ggplot2::geom_vline(xintercept = 0) + 
  ggplot2::labs(
    subtitle = paste0("Total variance explained: ",
                      signif(sum(var_frac[1:3]) * 100, 3), "%"),
    y = paste0("PC2; ", signif(var_frac[2] * 100, digits = 2), " %"),
    x = paste0("PC1; ", signif(var_frac[1] * 100, digits = 2), " %"),
    title = plot.title,
    caption = "") + 
  ggplot2::theme_minimal() +
  ggplot2::theme(
    axis.text = ggplot2::element_text(size = 15),
    axis.title.x = ggplot2::element_text(size = 20),
    axis.title.y = ggplot2::element_text(size = 20),
    legend.text = ggplot2::element_text(size = 15)
  )

print(pca.plot)
ggplot2::ggsave(pca.plot, bg = "white",
                filename = paste0(figure.path,"Toothfish_ALL_muscle_PCA_CpG_DMS2018_PC12.png"), 
                width = 30, height = 15, units = "cm")



# data <- data.frame(ind = rownames(PCobj$x),PCobj$x,  pop = pop)
data <- data.frame(ind = rownames(PCobj$rotation),PCobj$rotation,  pop = pop)
data <- merge(data, aggregate(cbind(mean.x = PC1, mean.y = PC3) ~
                                pop, data, mean), by = "pop")

pca.plot <- ggplot2::ggplot(data, ggplot2::aes(PC1, PC3, color = pop)) +
  ggplot2::geom_point(size = 4) + 
  ggplot2::geom_point(ggplot2::aes(x = mean.x,
                                   y = mean.y, color = pop),
                      size = 7,
                      shape = 17) +
  ggplot2::scale_colour_manual(values = colours6m) +
  ggplot2::geom_hline(yintercept = 0) +
  ggplot2::geom_vline(xintercept = 0) + 
  ggplot2::labs(
    subtitle = paste0("Total variance explained: ",
                      signif(sum(var_frac[1:3]) * 100, 3), "%"),
    y = paste0("PC3; ", signif(var_frac[3] * 100, digits = 2), " %"),
    x = paste0("PC1; ", signif(var_frac[1] * 100, digits = 2), " %"),
    title = plot.title,
    caption = "") + 
  ggplot2::theme_minimal() +
  ggplot2::theme(
    axis.text = ggplot2::element_text(size = 15),
    axis.title.x = ggplot2::element_text(size = 20),
    axis.title.y = ggplot2::element_text(size = 20),
    legend.text = ggplot2::element_text(size = 15)
  )

print(pca.plot)
ggplot2::ggsave(pca.plot, bg = "white",
                filename = paste0(figure.path,
                                  "Toothfish_ALL_muscle_PCA_CpG_DMS2018_PC13.png"), 
                width = 30, height = 15, units = "cm")




# data <- data.frame(ind = rownames(PCobj$x),PCobj$x,  pop = pop)
data <- data.frame(ind = rownames(PCobj$rotation),PCobj$rotation,  pop = pop)
data <- merge(data, aggregate(cbind(mean.x = PC2, mean.y = PC3) ~
                                pop, data, mean), by = "pop")

pca.plot <- ggplot2::ggplot(data, ggplot2::aes(PC2, PC3, color = pop)) +
  ggplot2::geom_point(size = 4) + 
  ggplot2::geom_point(ggplot2::aes(x = mean.x,
                                   y = mean.y, color = pop),
                      size = 7,
                      shape = 17) +
  ggplot2::scale_colour_manual(values = colours6m) +
  ggplot2::geom_hline(yintercept = 0) +
  ggplot2::geom_vline(xintercept = 0) + 
  ggplot2::labs(
    subtitle = paste0("Total variance explained: ",
                      signif(sum(var_frac[1:3]) * 100, 3), "%"),
    y = paste0("PC3; ", signif(var_frac[3] * 100, digits = 2), " %"),
    x = paste0("PC2; ", signif(var_frac[2] * 100, digits = 2), " %"),
    title = plot.title,
    caption = "") + 
  ggplot2::theme_minimal() +
  ggplot2::theme(
    axis.text = ggplot2::element_text(size = 15),
    axis.title.x = ggplot2::element_text(size = 20),
    axis.title.y = ggplot2::element_text(size = 20),
    legend.text = ggplot2::element_text(size = 15)
  )

print(pca.plot)
ggplot2::ggsave(pca.plot, bg = "white",
                filename = paste0(figure.path,"Toothfish_ALL_muscle_PCA_CpG_DMS2018_PC23.png"), 
                width = 30, height = 15, units = "cm")
```



# 9_Methylation: 2024 - fin tissue

```{r}
print(load(paste0(rdata.path,"Toothfish_metadata.Rdata"))) #ALL fish
```

## Reading the methylation call files

```{r, eval=FALSE}
data.list <- list.files(
  path = 
  "/datasets/work/ncmi-toa-rrbs/work/1.Toothfish/GENOME/Combined_methylation/",
  pattern = "*.cov.gz")

subnames <- metadata$File_names[!(metadata$Year %in% c(2017,2018) | 
                                    metadata$Tissue_type == "muscle" |
                                    grepl(pattern = "REP", x = metadata$File_names) |
                                    !grepl(pattern = "TOA", x = metadata$File_names)
                                  )]

subnames <- subnames[!subnames %in% "TOA_88-2_P79_AD294619_fin_F"] # low coverage

data.list <- data.list[data.list %in% paste0(subnames,
                                             "_R1_trimmed_sort_bismark_bt2.bismark.cov.gz")]

data.loc <- as.list(paste0(
  "/datasets/work/ncmi-toa-rrbs/work/1.Toothfish/GENOME/Combined_methylation/",
                           data.list))



data.list2 <-  gsub(pattern = '_R1_trimmed_sort_bismark_bt2.bismark.cov.gz',
                    replacement = '',x = as.character(data.list))
# data.list2 <-  gsub(pattern = '_sort_bismark_bt2_PE.bismark.cov.gz',
#                     replacement = '',x = as.character(data.list2))

metadata.new <- metadata[metadata$File_names %in% data.list2,]
metadata.new <- metadata.new[order(match(metadata.new$File_names, data.list2)),]

metadata.new$File_names == data.list2

sample.ids <- metadata.new$Genotype_names
pop <- as.integer(factor(metadata.new$pop))

myobjDB <- methylKit::methRead(location = data.loc,
           sample.id = as.list(sample.ids),
           assembly = "TOA_genome",
           treatment = pop,
           context = "CpG",
           header = FALSE, 
           dbtype = "tabix",
           pipeline = "bismarkCoverage",
           mincov = 3
           )
save(myobjDB, file = paste0(rdata.path,"Toothfish_2024_fin_methylkit_cpgDB.Rdata"))
```


## Data QC

#### Number of sites in genome

```{r, eval=FALSE}
cov.df <- data.frame(chr = NA, start = NA,end = NA, strand = NA, coverage = NA, 
                     numCs = NA, numTs = NA, sample = NA)
for (i in 1:length(sample.ids)) {
  df <- methylKit::getData(myobjDB[[i]])
  df$sample <- sample.ids[i]
  cov.df <- rbind(cov.df, df)
}
cov.df <- cov.df[-1,]
save(cov.df, file = paste0(rdata.path,"Toothfish_2024_fin_methylkit_cpg_cov.Rdata"))
```

### Coverage stats

```{r}
print(load(paste0(rdata.path,"Toothfish_2024_fin_methylkit_cpgDB.Rdata")))
print(load(paste0(rdata.path,"Toothfish_2024_fin_methylkit_cpg_cov.Rdata")))

length(unique(paste0(cov.df$chr, "_", cov.df$start))) #2,830,251

# summary(as.factor(cov.df$sample))
summary(cov.df$coverage)
    # Min.  1st Qu.   Median     Mean  3rd Qu.     Max. 
    # 4.00     8.00    17.00    23.82    31.00 43148.00

summary(cov.df$coverage[cov.df$coverage > 5 & cov.df$coverage < 50])
   # Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
   # 6.00   10.00   18.00   20.48   29.00   49.00 

cov.plot <- ggplot2::ggplot(cov.df, ggplot2::aes(x = log(coverage), fill = sample)) +
  ggplot2::geom_histogram() +
  ggplot2::theme(legend.position = "none")
# print(cov.plot)
ggplot2::ggsave(cov.plot,
                filename =
                  paste0(figure.path,"Toothfish_2024_fin_coverage_hist_Combined.png"), 
                width = 45, height = 15, units = 'cm')

cov.plot <- ggplot2::ggplot(cov.df, ggplot2::aes(x = coverage, fill = sample)) +
  ggplot2::geom_histogram() +
  ggplot2::theme(legend.position = "none")
# print(cov.plot)
ggplot2::ggsave(cov.plot,
                filename =
                  paste0(figure.path,"Toothfish_2024_fin_coverage_hist_Combined2.png"), 
                width = 45, height = 15, units = 'cm')

cov.plot <- ggplot2::ggplot(cov.df[cov.df$coverage > 5 & cov.df$coverage < 160,], 
                            ggplot2::aes(x = coverage)) +
  ggplot2::geom_histogram(binwidth = 1) +
  ggplot2::ggtitle("coverage < 160") +
  ggplot2::theme(legend.position = "none")
# print(cov.plot)
ggplot2::ggsave(cov.plot,
                filename =
                  paste0(figure.path,"Toothfish_2024_fin_coverage_hist_Combined4.png"),
                width = 45, height = 15, units = 'cm')



max(cov.df$coverage, na.rm = T) #43148
ind.covplot <- ggplot2::ggplot(cov.df, ggplot2::aes(x = sample, y = coverage)) + 
  ggplot2::geom_boxplot() +
  ggplot2::theme_bw() + 
  ggplot2::theme(axis.title.x = ggplot2::element_blank(), 
                 axis.text.y = ggplot2::element_text(size = 15),
                 axis.title.y = ggplot2::element_text(size = 15),
                 axis.text.x = ggplot2::element_text(size = 7, angle = 90, hjust = 1)) +
  ggplot2::scale_y_continuous(trans = scales::log2_trans(),
                  breaks = c(1, 10, 50, 750, 20000, 50000),
                  minor_breaks = c(1:10, 2:5*10, 1:7.5*100, 0.5:4.5*1000))

print(ind.covplot)

ggplot2::ggsave(ind.covplot, bg = "white",
                filename = 
                  paste0(figure.path,
                         "Toothfish_2024_fin_raw_individual_coverage_CpGs.png"),
                width = 45, height = 15, units = 'cm')
```


### Coverage filter

High count filter based on coverage stats.

```{r, eval=FALSE}
load(paste0(rdata.path,"Toothfish_2024_fin_methylkit_cpgDB.Rdata"))

filtered.myobj <- methylKit::filterByCoverage(myobjDB, lo.count = 10, lo.perc = NULL,
                                              save.db = FALSE,
                                      hi.count = NULL, hi.perc = 99)

save(myobjDB,filtered.myobj,
     file = paste0(rdata.path,"Toothfish_2024_fin_methylkit_cpgDB.Rdata"))
```

### Normalise coverage

Normalize read coverage distributions between samples. Requires stringent filter on max coverage first, otherwise it will over-sample PCR bias reads. These two functions will help reduce the bias in the statistical tests that might occur due to systematic over-sampling of reads in certain samples.

```{r}
filtered.myobj.normal <- methylKit::normalizeCoverage(obj = filtered.myobj, 
                                                      save.db = TRUE,
                                                      method = "median", chunk.size = 1e6)
save(myobjDB,filtered.myobj, filtered.myobj.normal,
     file = paste0(rdata.path,"Toothfish_2024_fin_methylkit_cpgDB.Rdata"))
```

### Number of sites per sample

```{r, eval = FALSE}
sites.df <- data.frame(sites = NA,  sample = NA)
for (i in 1:length(filtered.myobj.normal)) {
  print(i)
  sites.df[i,] <- c(filtered.myobj.normal[[i]]@num.records, filtered.myobj.normal[[i]]@sample.id)
}
sites.df$sites <- as.integer(sites.df$sites)
# sites.df$pop <- filtered.myobj.normal@treatment
sites.df$pop <- metadata.new$pop
save(sites.df, file = paste0(rdata.path,"Toothfish_2024_fin_number_cpg_sample.Rdata"))
```


```{r}
print(load(paste0(rdata.path,"Toothfish_2024_fin_number_cpg_sample.Rdata")))
sites.plot <- ggplot2::ggplot(sites.df, ggplot2::aes(x = sample, y = sites, colour = pop)) +
  ggplot2::geom_point() +
  ggplot2::scale_color_manual(values = colours2) +
  ggplot2::theme_bw() +
  ggplot2::theme(axis.text.x = ggplot2::element_text(angle = 90))
print(sites.plot)

ggplot2::ggsave(sites.plot,
                filename =
                  paste0(figure.path,"Toothfish_2024_fin_number_cpg_sample.png"),
                width = 45, height = 15, units = 'cm')

sites.df$sample[sites.df$sites < 500000]
# "TOA_88-2_P67_AD294613_fin_F"
low.sites <- sites.df$sample[sites.df$sites < 500000]
```


## Select samples

```{r}
print(load(paste0(rdata.path,"Toothfish_metadata.Rdata")))

metadata2 <- metadata[!(metadata$Year %in% c(2017,2018) | 
                                    metadata$Tissue_type == "muscle" |
                                    grepl(pattern = "REP", x = metadata$File_names) |
                                    !grepl(pattern = "TOA", x = metadata$File_names)
                                  ),]

print(load(paste0(rdata.path, "Low_quality_fish.Rdata")))

# rm1 <- Reduce(intersect, list(low.genome.mapping,low.genome.methylation))
rm1 <- unique(c(low.genome.mapping,low.genome.methylation, low.sites))
print(rm1)
rm3 <- c("TOA_58-4-2_B1_AE100278_fin_M",  "TOA_58-4-2_B11_AE100305_fin_M", 
         "TOA_58-4-2_B17_AE100308_fin_F", "TOA_58-4-2_B3_AE100279_fin_M",
         "TOA_58-4-2_B5_AE100280_fin_F" ) # bad coverage
rm4 <- c("TOA_48-2_031_UK116","TOA_58-4-2_111_AN172604","TOA_88-1_230_AD082543",
         "TOA_48-2_024_UK082")
print(rm4) #bad tissue PCA
rm <- c(rm1,rm3,rm4)
# rm <- c(rm1,rm2,rm3,rm4)

metadata2 <- metadata2[!(metadata2$File_names %in% rm | metadata2$Genotype_names %in% rm),]
filt.sample.id.full <- sapply(filtered.myobj.normal@.Data, function(x){x@sample.id})

# filt.sample.id <- filt.sample.id.full[filt.sample.id.full %in% metadata2$File_names]
# metadata2 <- metadata2[order(match(metadata.new$File_names, filt.sample.id)),]

filt.sample.id <- filt.sample.id.full[filt.sample.id.full %in% metadata2$Genotype_names]
metadata2 <- metadata2[order(match(metadata2$Genotype_names, filt.sample.id)),]
identical(filt.sample.id,metadata2$Genotype_names)


pop <- metadata2$pop
pop.levels <- c("58.4.2-2024_50L",
                "58.4.2-2024_70L",
                "88.2-2024")

pop <- factor(pop, levels = pop.levels)
pop <- as.numeric(pop)
filtered.myobj.normal@treatment <- pop
sample.ids <- metadata2$Genotype_names


filtered.myobj2.normal <- methylKit::reorganize(
  methylObj = filtered.myobj.normal,
  sample.ids = sample.ids,
  treatment = pop,
  chunk.size = 1e+06,
  save.db = TRUE,
  suffix = "TOA_2024_fin_normal"
)
save(myobjDB,filtered.myobj, 
     filtered.myobj.normal,filtered.myobj2.normal,
     file = paste0(rdata.path,"Toothfish_2024_fin_methylkit_cpgDB.Rdata"))
```


## Merging samples

```{r}
print(load(paste0(rdata.path,"Toothfish_2024_fin_methylkit_cpgDB.Rdata")))
table(metadata2$pop)
table(metadata2$STRATA2)

filtered.myobj2.normal@treatment <- as.numeric(factor(metadata2$STRATA2))

meth.2024F.norm15L <- methylKit::unite(filtered.myobj2.normal, destrand = FALSE,
                                       save.db = FALSE,
                             min.per.group = 15L, #missing data allowed
                             mc.cores = 25)
#564,013 common sites for 15 samples per group

save(meth.2024F.norm15L,
     file = paste0(rdata.path,"Toothfish_2024_fin_methylkit_object.Rdata"))
```


## Data QC

### Remove ASM

```{r}
load(paste0(rdata.path,"Allele_specific_mutations.Rdata"))

mut <- as(asmTRUE,"GRanges")
sum(as(meth.2024F.norm15L,"GRanges") %over% mut)#18887
meth.asm <- methylKit::selectByOverlap(meth.2024F.norm15L, mut) 
BOOLEAN <- paste0(meth.2024F.norm15L$chr,"_",meth.2024F.norm15L$start) %in%
  paste0(asmTRUE$Chr,"_",asmTRUE$Pos) #1
sum(BOOLEAN) #392
meth.2024F.norm15L <- methylKit::select(meth.2024F.norm15L, !BOOLEAN)

save(meth.2024F.norm15L,
     file = paste0(rdata.path,"Toothfish_2024_fin_methylkit_object.Rdata"))
```



### Coverage 

```{r}
meth.df <- methylKit::getData(meth.2024F.norm15L)
cov <- meth.df[,meth.2024F.norm15L@coverage.index]
colnames(cov) <- meth.2024F.norm15L@sample.ids
dpf <- data.frame(cov, locus = paste0(meth.df$chr,"_",meth.df$start))
dpf <- tidyr::gather(dpf, key = individual, 
                           value = Depth, -locus,
                           factor_key = TRUE) %>%
  dplyr::group_by(individual) %>%
  dplyr::mutate(NA_cnt = sum(is.na(Depth))/length(unique(locus))) %>%
  dplyr::ungroup()
dpf.NA <- dpf[!duplicated(dpf$individual),]
dpf.NA$NA_cnt <- round(dpf.NA$NA_cnt, 3)

dpf <- dpf[ dpf$Depth > 0 & !is.na(dpf$Depth),]

max(dpf$Depth, na.rm = T) #150
myPlot <- ggplot2::ggplot(dpf, ggplot2::aes(x = individual, y = Depth)) + 
  ggplot2::geom_boxplot() +
  ggplot2::geom_text(data = dpf.NA,ggplot2::aes(x = individual, y = 200, label = NA_cnt), 
                     size = 2, angle = 90) +
  ggplot2::theme_bw() + 
  ggplot2::theme(axis.title.x = ggplot2::element_blank(), 
                 axis.text.y = ggplot2::element_text(size = 15),
                 axis.title.y = ggplot2::element_text(size = 15),
                 axis.text.x = ggplot2::element_text(size = 7, angle = 90, hjust = 1)) +
  ggplot2::scale_y_continuous(trans = scales::log2_trans(),
                  breaks = c(1, 10, 50, 100, 200),
                  minor_breaks = c(1:10, 2:10*10))

print(myPlot)
ggplot2::ggsave(myPlot, bg = "white",
                filename = 
                  paste0(figure.path,
                  "1B.Toothfish_2024_fin_normalised15_methylkit_individual_coverage_CpGs.png"
                                  ), 
                width = 45, height = 15, units = "cm")

nrow(cov) # 541160 SMPs
plot(x = 1:ncol(cov), y = colMeans(cov, na.rm = TRUE), col = meth.2024F.norm15L@treatment)
plot(x = 1:ncol(cov), y = colMeans(cov > 20, na.rm = TRUE), col = meth.2024F.norm15L@treatment)
plot(x = 1:ncol(cov), y = colMeans(cov > 10, na.rm = TRUE), col = meth.2024F.norm15L@treatment)
meth.2024F.norm15L@sample.ids[colMeans(cov > 10, na.rm = TRUE) < 0.99]
# samples with low coverage: "TOA_58-4-2_B1_AE100278_fin_M"  "TOA_58-4-2_B11_AE100305_fin_M" "TOA_58-4-2_B17_AE100308_fin_F" "TOA_58-4-2_B3_AE100279_fin_M"  "TOA_58-4-2_B5_AE100280_fin_F" 
```



## Load data

```{r}
print(load(paste0(rdata.path,"Toothfish_2024_fin_methylkit_object.Rdata")))
print(load(paste0(rdata.path,"Toothfish_metadata.Rdata"))) 

metadata.new <- metadata[metadata$Genotype_names %in% meth.2024F.norm15L@sample.ids,]
metadata.new <- metadata.new[order(match(metadata.new$Genotype_names,
                                         meth.2024F.norm15L@sample.ids)),]

length(meth.2024F.norm15L@row.names) #number of sites
```


## Global clustering analysis

```{r}
methylKit::clusterSamples(meth.2024F.norm15L, dist = "correlation", sd.filter = TRUE, 
                          method = "ward.D2", plot = TRUE, )
dev.print(file = paste0(figure.path,"1_TOA_2024_fin_cluster_plot.png"),
          device = png, res = 300,width = 45,height = 15,units = "cm")
```


### PCA

```{r}
# methdif.perc <- methylKit::percMethylation(meth.2024F.norm15L)
# methdif.perc2 <- zoo::na.aggregate(methdif.perc, by = meth.2024F.norm15L@treatment)
# PCobj <- prcomp(methdif.perc2, retx = TRUE, scale. = TRUE)
PCobj <- methylKit::PCASamples(meth.2024F.norm15L, comp = c(1,2), 
                               adj.lim = c(0,0), obj.return = TRUE)

var_frac <- PCobj$sdev/sum(PCobj$sdev)
print(paste0("Total variance explained: ",signif(sum(var_frac[1:3]) * 100, 3), "%")) 
barplot(var_frac[1:20], main = "variance explained", 
        col = heat.colors(length(var_frac[1:20])))

identical(metadata.new$Genotype_names, meth.2024F.norm15L@sample.ids)
identical(metadata.new$Genotype_names, rownames(PCobj$x))

pop <- metadata.new$pop

plot.title <- "Toothfish methylation - 2024 fin (561,964 CpG sites)"

data <- data.frame(ind = rownames(PCobj$x),PCobj$x,  pop = pop)
# data <- data.frame(ind = rownames(PCobj$rotation),PCobj$rotation, pop = pop)
data <- merge(data, aggregate(cbind(mean.x = PC1, mean.y = PC2) ~
                                pop, data, mean), by = "pop")

pca.plot <- ggplot2::ggplot(data, ggplot2::aes(PC1, PC2, color = pop)) +
  ggplot2::geom_point(size = 4) + 
  ggplot2::geom_point(ggplot2::aes(x = mean.x,
                                   y = mean.y, color = pop),
                      size = 7,
                      shape = 17) +
  ggplot2::scale_colour_manual(values = colours2) +
  ggplot2::geom_hline(yintercept = 0) +
  ggplot2::geom_vline(xintercept = 0) + 
  ggplot2::labs(
    subtitle = paste0("Total variance explained: ",
                      signif(sum(var_frac[1:4]) * 100, 3), "%"),
    y = paste0("PC2; ", signif(var_frac[2] * 100, digits = 2), " %"),
    x = paste0("PC1; ", signif(var_frac[1] * 100, digits = 2), " %"),
    title = plot.title,
    caption = "") + 
  ggplot2::theme_minimal() +
  ggplot2::theme(
    axis.text = ggplot2::element_text(size = 15),
    axis.title.x = ggplot2::element_text(size = 20),
    axis.title.y = ggplot2::element_text(size = 20),
    legend.text = ggplot2::element_text(size = 15)
  )

print(pca.plot)
ggplot2::ggsave(pca.plot, bg = "white",
                filename = paste0(figure.path,"Toothfish_2024_fin_PCA_CpG_PC12.png"), 
                width = 30, height = 15, units = "cm")


data <- data.frame(ind = rownames(PCobj$x),PCobj$x,  pop = pop)
# data <- data.frame(ind = rownames(PCobj$rotation),PCobj$rotation, pop = pop)
data <- merge(data, aggregate(cbind(mean.x = PC2, mean.y = PC3) ~
                                pop, data, mean), by = "pop")
pca.plot <- ggplot2::ggplot(data, ggplot2::aes(PC2, PC3, color = pop)) +
  ggplot2::geom_point(size = 4) + 
  ggplot2::geom_point(ggplot2::aes(x = mean.x,
                                   y = mean.y, color = pop),
                      size = 7,
                      shape = 17) +
  ggplot2::scale_colour_manual(values = colours2) +
  ggplot2::geom_hline(yintercept = 0) +
  ggplot2::geom_vline(xintercept = 0) + 
  ggplot2::labs(
    subtitle = paste0("Total variance explained: ",
                      signif(sum(var_frac[1:4]) * 100, 3), "%"),
    y = paste0("PC3; ", signif(var_frac[3] * 100, digits = 2), " %"),
    x = paste0("PC2; ", signif(var_frac[2] * 100, digits = 2), " %"),
    title = plot.title,
    caption = "") + 
  ggplot2::theme_minimal() +
  ggplot2::theme(
    axis.text = ggplot2::element_text(size = 15),
    axis.title.x = ggplot2::element_text(size = 20),
    axis.title.y = ggplot2::element_text(size = 20),
    legend.text = ggplot2::element_text(size = 15)
  )

print(pca.plot)
ggplot2::ggsave(pca.plot, bg = "white",
                filename = paste0(figure.path,"Toothfish_2024_fin_PCA_CpG_PC23.png"), 
                width = 30, height = 15, units = "cm")
```


### RDA

```{r}
# setwd("/datasets/work/ncmi-toa-rrbs/work/1.Toothfish/Toothfish_RRBS_array_22098078/Bismark/")
# print(load(paste0(rdata.path,"Toothfish_methylation_objects_raw.Rdata"))) #36,635 SMPs
print(load(paste0(rdata.path,"Toothfish_subset_methylation_objects_raw.Rdata"))) #1,316,799 SMPs

# meth.RDA <- methylKit::unite(filtered.myobj2.norm, destrand = FALSE,
#                              save.db = FALSE,
#                              suffix = "subset",
#                              min.per.group = NULL, #missing data allowed
#                              mc.cores = 25)


# meth.df <- methylKit::getData(meth)
# cov <- meth.df[,grep(colnames(meth.df), pattern = "coverage*")]
# cov <- meth.df[,meth@coverage.index]
# colnames(cov) <- meth@sample.ids
# meth.df <- methylKit::percMethylation(meth)

# meth.df <- methylKit::percMethylation(meth.RDA)

# save(meth.df,meth.RDA, 
#      file = paste0(rdata.path,"TOA_data_for_RDA_meth_percent.Rdata"))
print(load(paste0(rdata.path,"TOA_data_for_RDA_meth_percent.Rdata"))) #347,614 SMPs

hist(colSums(is.na(meth.df)), breaks = 100)
hist(rowSums(is.na(meth.df)), breaks = 100)





print(load(paste0(rdata.path,"Toothfish_metadata.Rdata")))

predictor <- metadata2 # define the predictor input file here
predictor <- predictor[predictor$SAMPLE_ID %in% meth@sample.ids,]
predictor$SAMPLE_ID = as.character(predictor$SAMPLE_ID)
predictor$region = as.character(predictor$region)
predictor$Sex = as.character(predictor$Sex)
predictor$Tissue = as.character(predictor$Tissue)
predictor$mat = as.character(predictor$mat)
predictor$mat[is.na(predictor$mat)] <- "2"
predictor$TL = as.numeric(predictor$TL)
predictor$wt = as.numeric(predictor$wt)
predictor$dpth = as.numeric(predictor$dpth)
predictor$Age = as.numeric(predictor$Age)
predictor$Age[is.na(predictor$Age)] <- as.numeric(max(predictor$Age, na.rm = TRUE))
predictor$Cohort2 = as.integer(predictor$Cohort2)
predictor$Cohort2[is.na(predictor$Cohort2)] <- as.integer(min(predictor$Cohort2, na.rm = TRUE))


meth.df1 = t(meth.df)
# meth.df2 = na.omit(meth.df1) #rda can't handle NAs, remove them if you have them
# meth.df1[1:10,1:10]
## Predictor
# Make names characters (not factors)


#if you had numbers, they had to be factors (as.factor)

# Confirm that methylation data and predictors are in the same order
identical(rownames(meth.df1), predictor$SAMPLE_ID)

# Perform Hellinger transformation on the methylation data:
meth.df3 = vegan::decostand(meth.df1, method = "hellinger", MARGIN = 1) #MARGIN=1 means rows, on rows I have samples now because I transposed the matrix

## Run RDA
#Test the country using growing condition as a Condition () (covariate):
meth_rda = vegan::rda(meth.df3 ~ region + Condition(Sex), 
                      data = predictor, scale = F) #I remove scale since I performed Hellinger transformation
meth_rda = vegan::rda(meth.df3 ~ region + Condition(Sex) + 
                        Condition(mat)+ Condition(TL) + Condition(wt) +
                        Condition(dpth) + Condition(Age) + Condition(Cohort2), 
                      na.action = "na.omit",
                      data = predictor, scale = F)

meth_rda$pCCA
meth_rda$CCA$wa

summary(meth_rda) #Constrained Proportion: variance of Y explained by X, Unconstrained Proportion: unexplained variance in Y
# sink(paste0(rdata.path,"anova_rda.txt")) #with this you can save the output of the anova in a txt file, define the output directory
meth.aov <- anova(meth_rda, test = "Chi", permutations = 99) #run Anova on the result to see the significance level
# meth.aov.cca <- anova.cca(meth_rda, test = "Chi", permutations = 99)
print(meth.aov)
# sink(paste0(rdata.path,"anova_rda_chi.txt"))

# Model: rda(formula = meth.df3 ~ region + Condition(Sex), data = predictor, scale = F)
#           Df  Variance      F Pr(>F)   
# Model      3 0.0003924 1.4126  0.004 **
# Residual 110 0.0101861

# Model: rda(formula = meth.df3 ~ region + Condition(Sex) + Condition(mat) + Condition(TL) + Condition(wt) + Condition(dpth) + Condition(Age) + Condition(Cohort2), data = predictor, scale = F, na.action = "na.omit")
#           Df  Variance     F Pr(>F)   
# Model      3 0.0003412 1.257   0.01 **
# Residual 101 0.0091396 

Pop.names <- predictor$region
plot.a <- vegan::ordiplot(meth_rda, choices=c(1,2))
Pop.ellipses <- vegan::ordiellipse(plot.a, groups=Pop.names, display="sites", kind="sd")
dev.print(file = paste0(figure.path,"TOA_RDA_ordiplot_complex_model.png"),
          device = png, res = 300,width = 45,height = 15,units = "cm")

# Type 1 scaling
vegan::ordiplot(meth_rda, scaling = 1, type = "text")
# Type 2 scaling
vegan::ordiplot(meth_rda, scaling = 2, type = "text")

ggvegan::autoplot(
  meth_rda,
  axes = c(1, 2),
  geom = "text",
  layers = c("species", "sites", "biplot", "centroids"),
  arrows = TRUE)

#CpG contribution
hist(meth_rda$CCA$v[,1])
hist(meth_rda$CCA$v[,2])
hist(meth_rda$CCA$v[,3])



#Test the interaction of your predictors:
meth_rda3 = vegan::rda(meth ~ region * Sex + Condition(region + Sex), data = predictor, scale = F)
summary(meth_rda3)
sink(paste0(rdata.path,"anova_rda_interaction.txt"))
print(anova(meth_rda3, test="Chi"))
sink(paste0(rdata.path,"anova_rda_interaction_chi.txt"))
```

#### plot RDA

```{r}

BioR.theme <- theme(
        panel.background = element_blank(),
        panel.border = element_blank(),
        panel.grid = element_blank(),
        axis.line = element_line("gray25"),
        text = element_text(size = 12),
        axis.text = element_text(size = 10, colour = "gray25"),
        axis.title = element_text(size = 14, colour = "gray25"),
        legend.title = element_text(size = 14),
        legend.text = element_text(size = 14),
        legend.key = element_blank())

plot.a <- vegan::ordiplot(meth_rda, choices=c(1,2))
Pop.names <- predictor$region

# sites1 <- BiodiversityR::sites.long(plot.a, env.data=predictor)
sites1 <- data.frame(cbind(predictor, axis1=plot.a$sites[, 1], 
                           axis2=plot.a$sites[, 2], 
                           labels=rownames(plot.a$sites)),
                     group = predictor$region)

# axis1 <- BiodiversityR::axis.long(meth_rda, choices=c(1, 2))
eigs.all <- meth_rda$CA$eig
eigs <- round(100 * eigs.all / meth_rda$tot.chi, digits=1)[c(1, 2)] 
xlab.label <- paste0(names(eigs)[1], " (", eigs[1], "%)")
ylab.label <- paste0(names(eigs)[2], " (", eigs[2], "%)")
axis1 <- data.frame(axis=c(1:2), 
                    ggplot=c("xlab.label", "ylab.label"), 
                    label=c(xlab.label, ylab.label))


# centroids.Pop1 <- BiodiversityR::centroids.long(sites1, 
#                                                 grouping=Pop.names, 
#                                                 centroids.only=TRUE)
# centroids.Pop2 <- BiodiversityR::centroids.long(sites1, 
#                                                 grouping=Pop.names, 
#                                                 centroids.only=FALSE)
gr.name <- rlang::enexpr(Pop.names)
cent.means <-  stats::aggregate(cbind(axis1, axis2) ~ Pop.names,
                                data = sites1,
                                FUN = mean)
names(cent.means) <- c("group", "axis1c", "axis2c")
centroids.Pop1 <- cent.means
cent.means$Centroid <- cent.means[, 1]
by.name <- names(sites1)[names(sites1) == "group"]
centroids.Pop2 <- dplyr::full_join(sites1, cent.means, by = by.name)


# Pop.ellipses.long2 <- BiodiversityR::ordiellipse.long(Pop.ellipses, 
#                                                     grouping.name="Pop.names")


plotgg.a <- ggplot() + 
    geom_vline(xintercept = c(0), color = "grey70", linetype = 2) +
    geom_hline(yintercept = c(0), color = "grey70", linetype = 2) +  
    xlab(axis1[1, "label"]) +
    ylab(axis1[2, "label"]) +  
    scale_x_continuous(sec.axis = dup_axis(labels=NULL, name=NULL)) +
    scale_y_continuous(sec.axis = dup_axis(labels=NULL, name=NULL)) +
    # geom_polygon(data=Pop.ellipses.long2, 
    #                aes(x=axis1, y=axis2, colour=Pop.names, 
    #                    fill=after_scale(alpha(colour, 0.2))), 
    #           size=0.2, show.legend=TRUE) +
    geom_segment(data=centroids.Pop2, 
                 aes(x=axis1c, y=axis2c, xend=axis1, yend=axis2, colour=Pop.names), 
                 size=0.7, show.legend=FALSE) +
    geom_point(data=centroids.Pop1, 
               aes(x=axis1c, y=axis2c), 
               shape="square", colour="black", alpha=0.7, size=2) +
    geom_label_repel(data=centroids.Pop1, 
               aes(x=axis1c, y=axis2c, label=group, colour=group), 
               alpha=1.0, size=2, show.legend=FALSE) +
    # labs(colour = Pop.names) +
    BioR.theme +
    theme(legend.text = element_text(size = 10)) +
    scale_colour_npg() +
    coord_fixed(ratio=1)

plotgg.a

ggsave(plotgg.a, file = paste0(figure.path, 
                               # "TOA_RDA_simple_model_A.png"
                               "TOA_RDA_complex_model_A.png"
                               ),
       height = 30, width = 30, units = "cm",  device = "png")
```



## DMS - 2024 - fin

```{r, eval = FALSE}
print(load(paste0(rdata.path,"Toothfish_2024_fin_methylkit_object.Rdata")))

print(load(paste0(rdata.path,"Toothfish_metadata.Rdata")))
metadata.new <- metadata[metadata$Genotype_names %in% attributes(meth.2024F.norm15L)$sample.ids,]
attributes(meth.2024F.norm15L)$sample.ids == metadata.new$Genotype_names

# meth.2024F.norm15L@treatment <- as.integer(factor(metadata.new$pop))
meth.2024F.norm15L@treatment <- as.integer(factor(metadata.new$STRATA2))

# data.frame(attributes(meth.2024F.norm15L)$sample.ids, metadata.new$individual, metadata.new$strata, meth.2024F.norm15L@treatment)

covariates <- dplyr::select(metadata.new, Sex, Age, Cohort2, Maturity_full)
myDiff <- methylKit::calculateDiffMeth(meth.2024F.norm15L,
                                       covariates = NULL,
                                       adjust = "fdr",
                                       chunk.size = 1e+06,
                                       mc.cores = 25)
meth.diff.chr <- methylKit::diffMethPerChr(myDiff,plot = TRUE, 
                                           qvalue.cutoff = 0.01, meth.cutoff = 40)
myDiff40p <- methylKit::getMethylDiff(myDiff, difference = 40, qvalue = 0.01)
#2
myDiff25p <- methylKit::getMethylDiff(myDiff, difference = 25, qvalue = 0.01)
#115 
myDiff15p <- methylKit::getMethylDiff(myDiff, difference = 15, qvalue = 0.01)
#2872  
top100.2024F <- slice_max(as.data.frame(as(myDiff,"GRanges")), order_by = meth.diff, n = 100)
top100.2024F <- as(top100.2024F,"GRanges")

save(myDiff,myDiff15p, myDiff25p, myDiff40p,top100.2024F,
     file = paste0(rdata.path,"Toothfish_2024_fin_methylation_DMS.Rdata"))
```


```{r}
print(load(paste0(rdata.path,"Toothfish_2024_fin_methylation_DMS.Rdata")))

volcano.plot <- ggplot2::ggplot() +
  ggplot2::geom_point(data = myDiff, ggplot2::aes(x = meth.diff, 
                                                  y = -log10(qvalue))) +
  ggplot2::geom_point(data = myDiff25p, ggplot2::aes(x = meth.diff, 
                                                  y = -log10(qvalue)),
                      colour = "red") +
  ggplot2::theme_minimal() +
  ggplot2::theme(
    axis.text = ggplot2::element_text(size = 15),
    axis.title.x = ggplot2::element_text(size = 20),
    axis.title.y = ggplot2::element_text(size = 20),
    legend.text = ggplot2::element_text(size = 15)
  )
print(volcano.plot)

ggplot2::ggsave(volcano.plot, bg = "white",
                filename =
                  paste0(figure.path,
                         "Toothfish_2024_fin_DMS25_VOLCANO_methylkit_regions.png"), 
                width = 30, height = 15, units = "cm")
```

### Heatmap

```{r}
print(load(paste0(rdata.path,"Toothfish_2024_fin_methylation_DMS.Rdata")))

myDiff25p.gr <- as(myDiff25p, "GRanges")
meth.diff <- methylKit::selectByOverlap(meth.2024F.norm15L, myDiff25p.gr)

meth.diff.df <- as.data.frame(t(methylKit::percMethylation(meth.diff)))
loc.names <- paste0(meth.diff$chr,"_",meth.diff$start)
colnames(meth.diff.df) <- loc.names

meth.diff.df$individual <- rownames(meth.diff.df)
pop <- metadata.new$pop
meth.diff.df$pop <- as.character(pop)


heat.plot.data <- meth.diff.df %>%
  tibble::as_tibble() %>%
  tidyr::gather(key = "CpG", value = "Methylation_percentage", -individual, -pop,
                factor_key = FALSE) %>%
  dplyr::group_by(pop, CpG) %>%
  dplyr::mutate(CpG_mean = mean(Methylation_percentage, na.rm = TRUE)) %>%
  dplyr::arrange(pop, individual, CpG) %>%
  dplyr::mutate(individual = factor(individual, levels = unique(individual))) %>%
  dplyr::mutate(CpG = factor(CpG, levels = unique(CpG)))
cpgs <- heat.plot.data$CpG[heat.plot.data$individual == unique(heat.plot.data$individual)[1]]
cpgs_mean <- heat.plot.data$CpG_mean[heat.plot.data$individual == unique(heat.plot.data$individual)[1]]

cpgs <- cpgs[order(cpgs_mean, decreasing = TRUE)]
heat.plot.data$CpG <- factor(heat.plot.data$CpG, levels = c("pop", unique(as.character(cpgs))))
heat.plot.data <- heat.plot.data[order(heat.plot.data$CpG),]

heat.plot <- ggplot2::ggplot(data = heat.plot.data, ggplot2::aes(x = individual)) + 
    ggplot2::geom_tile(data = heat.plot.data, 
                       ggplot2::aes(x = individual, y = CpG, fill = Methylation_percentage)) +
    ggplot2::scale_fill_gradient2(low = "blue", mid = "yellow", high = "red", 
                                  midpoint = 50, guide = "colourbar") +
    ggnewscale::new_scale_fill() +
    ggplot2::geom_tile(data = heat.plot.data, ggplot2::aes(x = individual, y = "pop", fill = pop,
                                    height = length(unique(heat.plot.data$CpG))*0.02)) +
    ggplot2::scale_fill_manual(values = colours2) +
    ggplot2::scale_y_discrete(drop = FALSE) +
    ggplot2::theme(legend.position = "top",
                   axis.title.y = ggplot2::element_blank(),
                   axis.title.x = ggplot2::element_blank(),
                   axis.text.x = ggplot2::element_text(face = "bold", color = "black", 
                                        size = 10, angle = 90, hjust = 1, vjust = 0.5))
print(heat.plot)

ggplot2::ggsave(heat.plot, bg = "white",
                filename = paste0(figure.path,
                                  "Toothfish_2024_fin_methDiff_25_HEATMAP.png"), 
                width = 30, height = 30, units = "cm")
```

### Plot CpG difference

```{r}
load(file = paste0(rdata.path,"Toothfish_2024_fin_methylation_DMS.Rdata"))

myDiff25p.gr <- as(myDiff25p, "GRanges")
meth.diff <- methylKit::selectByOverlap(meth.2024F.norm15L, myDiff25p.gr) #113 

length(unique(meth.diff$chr)) #34
sort(table(meth.diff$chr), decreasing = TRUE)[1:20]

meth.diff.df <- as.data.frame(t(methylKit::percMethylation(meth.diff)))
loc.names <- paste0(meth.diff$chr,"_",meth.diff$start)
colnames(meth.diff.df) <- loc.names

meth.diff.df$individual <- rownames(meth.diff.df) 
identical(metadata.new$Genotype_names, meth.diff.df$individual)
meth.diff.df$pop <- metadata.new$pop



meth.diff.df.plot <- tidyr::gather(meth.diff.df, key = locus, 
                           value = perc_methylation, -individual, -pop,
                           factor_key = TRUE) %>%
  dplyr::mutate(chr = stringr::str_split_fixed(locus, 
                                               pattern = "_", 
                                               n = 2)[,1])
meth.diff.df.plot.locus <- meth.diff.df.plot %>% dplyr::group_by(chr,locus,pop) %>% 
  dplyr::summarise(mean_pop = mean(perc_methylation, na.rm = TRUE),
                   sum_NA = sum(is.na(perc_methylation)),
            .groups = 'drop')


diff.plot <- ggplot2::ggplot() +
  ggplot2::geom_line(data = meth.diff.df.plot.locus,
                     ggplot2::aes(x = pop, y = mean_pop, colour = locus, group = locus)) +
  ggplot2::geom_point(data = meth.diff.df.plot.locus,
                     ggplot2::aes(x = pop, y = mean_pop, colour = locus, group = locus),
                     size = 5) +
  ggplot2::facet_wrap(~chr) +
  ggplot2::ylab("Mean methylation (%)") +
  ggplot2::theme(legend.position = "none",
                   axis.text.x = ggplot2::element_text(face = "bold", color = "black", 
                                        size = 10, angle = 90, hjust = 1, vjust = 0.5))

print(diff.plot)

ggplot2::ggsave(plot = diff.plot, 
                filename = paste0(figure.path,"1_Toothfish_2024_fin_methDiff.png"),
                width = 30, height = 15, units = "cm", device = "png")
```


### DMS Clustering
```{r}
load(file = paste0(rdata.path,"Toothfish_2024_fin_methylation_DMS.Rdata"))

myDiff25p.gr <- as(myDiff25p, "GRanges")
meth.diff <- methylKit::selectByOverlap(meth.2024F.norm15L, myDiff25p.gr) #112  
methdif.perc <- methylKit::percMethylation(meth.diff)

methdif.perc2 <- zoo::na.aggregate(methdif.perc, by = meth.2024F.norm15L@treatment)
PCobj <- prcomp(methdif.perc2, retx = TRUE, scale. = TRUE)

var_frac <- PCobj$sdev/sum(PCobj$sdev)
print(paste0("Total variance explained: ",signif(sum(var_frac[1:3]) * 100, 3), "%")) 
barplot(var_frac[1:3], main = "variance explained", 
        col = heat.colors(length(var_frac[1:3])))


identical(metadata.new$Genotype_names, meth.2024F.norm15L@sample.ids)
# identical(metadata.new$individual, rownames(PCobj$x))
identical(metadata.new$Genotype_names, rownames(PCobj$rotation))
pop <- metadata.new$pop

plot.title <- "Toothfish methylation - 2024 fin (115 DMS sites - 25%)"

# data <- data.frame(ind = rownames(PCobj$x),PCobj$x,  pop = pop)
data <- data.frame(ind = rownames(PCobj$rotation),PCobj$rotation,  pop = pop)
data <- merge(data, aggregate(cbind(mean.x = PC1, mean.y = PC2) ~
                                pop, data, mean), by = "pop")

pca.plot <- ggplot2::ggplot(data, ggplot2::aes(PC1, PC2, color = pop)) +
  ggplot2::geom_point(size = 4) + 
  ggplot2::geom_point(ggplot2::aes(x = mean.x,
                                   y = mean.y, color = pop),
                      size = 7,
                      shape = 17) +
  ggplot2::scale_colour_manual(values = colours2) +
  ggplot2::geom_hline(yintercept = 0) +
  ggplot2::geom_vline(xintercept = 0) + 
  ggplot2::labs(
    subtitle = paste0("Total variance explained: ",
                      signif(sum(var_frac[1:3]) * 100, 3), "%"),
    y = paste0("PC2; ", signif(var_frac[2] * 100, digits = 2), " %"),
    x = paste0("PC1; ", signif(var_frac[1] * 100, digits = 2), " %"),
    title = plot.title,
    caption = "") + 
  ggplot2::theme_minimal() +
  ggplot2::theme(
    axis.text = ggplot2::element_text(size = 15),
    axis.title.x = ggplot2::element_text(size = 20),
    axis.title.y = ggplot2::element_text(size = 20),
    legend.text = ggplot2::element_text(size = 15)
  )

print(pca.plot)
ggplot2::ggsave(pca.plot, bg = "white",
                filename = paste0(figure.path,"Toothfish_2024_fin_PCA_CpG_DMS_PC12.png"), 
                width = 30, height = 15, units = "cm")



# data <- data.frame(ind = rownames(PCobj$x),PCobj$x,  pop = pop)
data <- data.frame(ind = rownames(PCobj$rotation),PCobj$rotation,  pop = pop)
data <- merge(data, aggregate(cbind(mean.x = PC1, mean.y = PC3) ~
                                pop, data, mean), by = "pop")

pca.plot <- ggplot2::ggplot(data, ggplot2::aes(PC1, PC3, color = pop)) +
  ggplot2::geom_point(size = 4) + 
  ggplot2::geom_point(ggplot2::aes(x = mean.x,
                                   y = mean.y, color = pop),
                      size = 7,
                      shape = 17) +
  ggplot2::scale_colour_manual(values = colours2) +
  ggplot2::geom_hline(yintercept = 0) +
  ggplot2::geom_vline(xintercept = 0) + 
  ggplot2::labs(
    subtitle = paste0("Total variance explained: ",
                      signif(sum(var_frac[1:3]) * 100, 3), "%"),
    y = paste0("PC3; ", signif(var_frac[3] * 100, digits = 2), " %"),
    x = paste0("PC1; ", signif(var_frac[1] * 100, digits = 2), " %"),
    title = plot.title,
    caption = "") + 
  ggplot2::theme_minimal() +
  ggplot2::theme(
    axis.text = ggplot2::element_text(size = 15),
    axis.title.x = ggplot2::element_text(size = 20),
    axis.title.y = ggplot2::element_text(size = 20),
    legend.text = ggplot2::element_text(size = 15)
  )

print(pca.plot)
ggplot2::ggsave(pca.plot, bg = "white",
                filename = paste0(figure.path,
                                  "Toothfish_2024_fin_PCA_CpG_DMS_PC13.png"), 
                width = 30, height = 15, units = "cm")




# data <- data.frame(ind = rownames(PCobj$x),PCobj$x,  pop = pop)
data <- data.frame(ind = rownames(PCobj$rotation),PCobj$rotation,  pop = pop)
data <- merge(data, aggregate(cbind(mean.x = PC2, mean.y = PC3) ~
                                pop, data, mean), by = "pop")

pca.plot <- ggplot2::ggplot(data, ggplot2::aes(PC2, PC3, color = pop)) +
  ggplot2::geom_point(size = 4) + 
  ggplot2::geom_point(ggplot2::aes(x = mean.x,
                                   y = mean.y, color = pop),
                      size = 7,
                      shape = 17) +
  ggplot2::scale_colour_manual(values = colours2) +
  ggplot2::geom_hline(yintercept = 0) +
  ggplot2::geom_vline(xintercept = 0) + 
  ggplot2::labs(
    subtitle = paste0("Total variance explained: ",
                      signif(sum(var_frac[1:3]) * 100, 3), "%"),
    y = paste0("PC3; ", signif(var_frac[3] * 100, digits = 2), " %"),
    x = paste0("PC2; ", signif(var_frac[2] * 100, digits = 2), " %"),
    title = plot.title,
    caption = "") + 
  ggplot2::theme_minimal() +
  ggplot2::theme(
    axis.text = ggplot2::element_text(size = 15),
    axis.title.x = ggplot2::element_text(size = 20),
    axis.title.y = ggplot2::element_text(size = 20),
    legend.text = ggplot2::element_text(size = 15)
  )

print(pca.plot)
ggplot2::ggsave(pca.plot, bg = "white",
                filename = paste0(figure.path,"Toothfish_2024_fin_PCA_CpG_DMS_PC23.png"), 
                width = 30, height = 15, units = "cm")
```



## DMS RANDOM - 2024 - fin

```{r,eval=FALSE}
print(load(paste0(rdata.path,"Toothfish_2024_fin_methylkit_object.Rdata")))

print(load(paste0(rdata.path,"Toothfish_metadata.Rdata")))
metadata.new <- metadata[metadata$Genotype_names %in% attributes(meth.2024F.norm15L)$sample.ids,]
attributes(meth.2024F.norm15L)$sample.ids == metadata.new$Genotype_names

# strata.random <- sample(metadata.new$pop, size = length(metadata.new$pop), replace = FALSE)
strata.random <- sample(metadata.new$STRATA2, size = length(metadata.new$STRATA2), replace = FALSE)

meth.2024F.norm15L@treatment <- as.integer(factor(strata.random))

# data.frame(attributes(meth.2024F.norm15L)$sample.ids, metadata.new$individual, metadata.new$strata, meth.2024F.norm15L@treatment)

# covariates <- dplyr::select(metadata.new, Sex, TL_mm)
myDiff <- methylKit::calculateDiffMeth(meth.2024F.norm15L,
                                       covariates = NULL,
                                       adjust = "fdr",
                                       chunk.size = 1e+06,
                                       mc.cores = 25)
meth.diff.chr <- methylKit::diffMethPerChr(myDiff,plot = TRUE, 
                                           qvalue.cutoff = 0.01, meth.cutoff = 40)
myDiff40p <- methylKit::getMethylDiff(myDiff, difference = 40, qvalue = 0.01)
#0
myDiff25p <- methylKit::getMethylDiff(myDiff, difference = 25, qvalue = 0.01)
#54
myDiff15p <- methylKit::getMethylDiff(myDiff, difference = 15, qvalue = 0.01)
#1019

save(strata.random, myDiff,myDiff15p, myDiff25p, myDiff40p,
     file = paste0(rdata.path,"Toothfish_2024_fin_methylation_RANDOM_DMS.Rdata"))
```


```{r}
print(load(paste0(rdata.path,"Toothfish_2024_fin_methylation_RANDOM_DMS.Rdata")))

volcano.plot <- ggplot2::ggplot() +
  ggplot2::geom_point(data = myDiff, ggplot2::aes(x = meth.diff, 
                                                  y = -log10(qvalue))) +
  ggplot2::geom_point(data = myDiff25p, ggplot2::aes(x = meth.diff, 
                                                  y = -log10(qvalue)),
                      colour = "red") +
  
  ggplot2::theme_minimal() +
  ggplot2::theme(
    axis.text = ggplot2::element_text(size = 15),
    axis.title.x = ggplot2::element_text(size = 20),
    axis.title.y = ggplot2::element_text(size = 20),
    legend.text = ggplot2::element_text(size = 15)
  )
print(volcano.plot)

ggplot2::ggsave(volcano.plot, bg = "white",
                filename =
                  paste0(figure.path,
                         "Toothfish_2024_fin_RANDOM_DMS25_VOLCANO_methylkit.png"), 
                width = 30, height = 15, units = "cm")
```


### Heatmap

```{r}
print(load(paste0(rdata.path,"Toothfish_2024_fin_methylation_RANDOM_DMS.Rdata")))

myDiff25p.gr <- as(myDiff25p, "GRanges")
meth.diff <- methylKit::selectByOverlap(meth.2024F.norm15L, myDiff25p.gr)

meth.diff.df <- as.data.frame(t(methylKit::percMethylation(meth.diff)))
loc.names <- paste0(meth.diff$chr,"_",meth.diff$start)
colnames(meth.diff.df) <- loc.names

meth.diff.df$individual <- rownames(meth.diff.df)
meth.diff.df$pop <- strata.random
meth.diff.df$pop2 <- as.character(metadata.new$pop)


heat.plot.data <- meth.diff.df %>%
  tibble::as_tibble() %>%
  tidyr::gather(key = "CpG", value = "Methylation_percentage", -individual, -pop, -pop2,
                factor_key = FALSE) %>%
  dplyr::group_by(pop, CpG) %>%
  dplyr::mutate(CpG_mean = mean(Methylation_percentage, na.rm = TRUE)) %>%
  dplyr::arrange(pop, individual, CpG) %>%
  dplyr::mutate(individual = factor(individual, levels = unique(individual))) %>%
  dplyr::mutate(CpG = factor(CpG, levels = unique(CpG)))
cpgs <- heat.plot.data$CpG[heat.plot.data$individual == unique(heat.plot.data$individual)[1]]
cpgs_mean <- heat.plot.data$CpG_mean[heat.plot.data$individual == unique(heat.plot.data$individual)[1]]

cpgs <- cpgs[order(cpgs_mean, decreasing = TRUE)]
heat.plot.data$CpG <- factor(heat.plot.data$CpG, levels = c("pop", unique(as.character(cpgs))))
heat.plot.data <- heat.plot.data[order(heat.plot.data$CpG),]

heat.plot <- ggplot2::ggplot(data = heat.plot.data, ggplot2::aes(x = individual)) + 
    ggplot2::geom_tile(data = heat.plot.data, 
                       ggplot2::aes(x = individual, y = CpG, fill = Methylation_percentage)) +
    ggplot2::scale_fill_gradient2(low = "blue", mid = "yellow", high = "red", 
                                  midpoint = 50, guide = "colourbar") +
    ggnewscale::new_scale_fill() +
    ggplot2::geom_tile(data = heat.plot.data, ggplot2::aes(x = individual, 
                                                           y = "pop - random", fill = pop,
                                    height = length(unique(heat.plot.data$CpG))*0.02)) +
    # ggplot2::scale_fill_manual(values = colours2) +
    ggplot2::scale_fill_manual(values = c("#adcf9f", "#08306B")) +
    ggnewscale::new_scale_fill() +
    ggplot2::geom_tile(data = heat.plot.data, ggplot2::aes(x = individual, y = "pop", fill = pop2,
                                    height = length(unique(heat.plot.data$CpG))*0.02)) +
    ggplot2::scale_fill_manual(values = colours2) +
    ggplot2::scale_y_discrete(drop = FALSE) +
    ggplot2::theme(legend.position = "top",
                   axis.title.y = ggplot2::element_blank(),
                   axis.title.x = ggplot2::element_blank(),
                   axis.text.x = ggplot2::element_text(face = "bold", color = "black", 
                                        size = 10, angle = 90, hjust = 1, vjust = 0.5))
print(heat.plot)

ggplot2::ggsave(heat.plot, bg = "white",
                filename = paste0(figure.path,
                                  "Toothfish_2024_fin_RANDOM_methDiff_25_HEATMAP.png"), 
                width = 30, height = 30, units = "cm")
```

### Plot CpG difference

```{r}
load(file = paste0(rdata.path,"Toothfish_2024_fin_methylation_RANDOM_DMS.Rdata"))

myDiff25p.gr <- as(myDiff25p, "GRanges")
meth.diff <- methylKit::selectByOverlap(meth.2024F.norm15L, myDiff25p.gr) #113 

length(unique(meth.diff$chr)) #34
sort(table(meth.diff$chr), decreasing = TRUE)[1:20]

meth.diff.df <- as.data.frame(t(methylKit::percMethylation(meth.diff)))
loc.names <- paste0(meth.diff$chr,"_",meth.diff$start)
colnames(meth.diff.df) <- loc.names

meth.diff.df$individual <- rownames(meth.diff.df) 
identical(metadata.new$Genotype_names, meth.diff.df$individual)
meth.diff.df$pop <- metadata.new$pop


meth.diff.df.plot <- tidyr::gather(meth.diff.df, key = locus, 
                           value = perc_methylation, -individual, -pop,
                           factor_key = TRUE) %>%
  dplyr::mutate(chr = stringr::str_split_fixed(locus, 
                                               pattern = "_", 
                                               n = 2)[,1])
meth.diff.df.plot.locus <- meth.diff.df.plot %>% dplyr::group_by(chr,locus,pop) %>% 
  dplyr::summarise(mean_pop = mean(perc_methylation, na.rm = TRUE),
                   sum_NA = sum(is.na(perc_methylation)),
            .groups = 'drop')


diff.plot <- ggplot2::ggplot() +
  ggplot2::geom_line(data = meth.diff.df.plot.locus,
                     ggplot2::aes(x = pop, y = mean_pop, colour = locus, group = locus)) +
  ggplot2::geom_point(data = meth.diff.df.plot.locus,
                     ggplot2::aes(x = pop, y = mean_pop, colour = locus, group = locus),
                     size = 5) +
  ggplot2::facet_wrap(~chr) +
  ggplot2::ylab("Mean methylation (%)") +
  ggplot2::theme(legend.position = "none",
                   axis.text.x = ggplot2::element_text(face = "bold", color = "black", 
                                        size = 10, angle = 90, hjust = 1, vjust = 0.5))
    # ggplot2::guides(colour = ggplot2::guide_legend(ncol = 2))

print(diff.plot)

ggplot2::ggsave(plot = diff.plot, 
                filename = paste0(figure.path,"1_Toothfish_2024_fin_RANDOM_methDiff.png"),
                width = 30, height = 15, units = "cm", device = "png")
```


### DMS Clustering
```{r}
load(file = paste0(rdata.path,"Toothfish_2024_fin_methylation_RANDOM_DMS.Rdata"))

myDiff25p.gr <- as(myDiff25p, "GRanges")
meth.diff <- methylKit::selectByOverlap(meth.2024F.norm15L, myDiff25p.gr) #270  
methdif.perc <- methylKit::percMethylation(meth.diff)

methdif.perc2 <- zoo::na.aggregate(methdif.perc, by = meth.2024F.norm15L@treatment)
PCobj <- prcomp(methdif.perc2, retx = TRUE, scale. = TRUE) # Too many missing values

var_frac <- PCobj$sdev/sum(PCobj$sdev)
print(paste0("Total variance explained: ",signif(sum(var_frac[1:3]) * 100, 3), "%")) 
barplot(var_frac[1:3], main = "variance explained", 
        col = heat.colors(length(var_frac[1:3])))


identical(metadata.new$Genotype_names, meth.2024F.norm15L@sample.ids)

identical(as.numeric(factor(metadata.new$pop)), meth.2024F.norm15L@treatment)

# identical(metadata.new$individual, rownames(PCobj$x))
identical(metadata.new$Genotype_names, rownames(PCobj$rotation))
pop <- metadata.new$pop

plot.title <- "Toothfish methylation - RANDOMISED 2024 fin (54 DMS sites - 25%)"

# data <- data.frame(ind = rownames(PCobj$x),PCobj$x,  pop = pop)
data <- data.frame(ind = rownames(PCobj$rotation),PCobj$rotation,  pop = pop)
data <- merge(data, aggregate(cbind(mean.x = PC1, mean.y = PC2) ~
                                pop, data, mean), by = "pop")

pca.plot <- ggplot2::ggplot(data, ggplot2::aes(PC1, PC2, color = pop)) +
  ggplot2::geom_point(size = 4) + 
  ggplot2::geom_point(ggplot2::aes(x = mean.x,
                                   y = mean.y, color = pop),
                      size = 7,
                      shape = 17) +
  ggplot2::scale_colour_manual(values = colours2) +
  ggplot2::geom_hline(yintercept = 0) +
  ggplot2::geom_vline(xintercept = 0) + 
  ggplot2::labs(
    subtitle = paste0("Total variance explained: ",
                      signif(sum(var_frac[1:3]) * 100, 3), "%"),
    y = paste0("PC2; ", signif(var_frac[2] * 100, digits = 2), " %"),
    x = paste0("PC1; ", signif(var_frac[1] * 100, digits = 2), " %"),
    title = plot.title,
    caption = "") + 
  ggplot2::theme_minimal() +
  ggplot2::theme(
    axis.text = ggplot2::element_text(size = 15),
    axis.title.x = ggplot2::element_text(size = 20),
    axis.title.y = ggplot2::element_text(size = 20),
    legend.text = ggplot2::element_text(size = 15)
  )

print(pca.plot)
ggplot2::ggsave(pca.plot, bg = "white",
                filename = 
                  paste0(figure.path,"Toothfish_2024_fin_RANDOM_PCA_CpG_DMS_PC12.png"), 
                width = 30, height = 15, units = "cm")



# data <- data.frame(ind = rownames(PCobj$x),PCobj$x,  pop = pop)
data <- data.frame(ind = rownames(PCobj$rotation),PCobj$rotation,  pop = pop)
data <- merge(data, aggregate(cbind(mean.x = PC1, mean.y = PC3) ~
                                pop, data, mean), by = "pop")

pca.plot <- ggplot2::ggplot(data, ggplot2::aes(PC1, PC3, color = pop)) +
  ggplot2::geom_point(size = 4) + 
  ggplot2::geom_point(ggplot2::aes(x = mean.x,
                                   y = mean.y, color = pop),
                      size = 7,
                      shape = 17) +
  ggplot2::scale_colour_manual(values = colours2) +
  ggplot2::geom_hline(yintercept = 0) +
  ggplot2::geom_vline(xintercept = 0) + 
  ggplot2::labs(
    subtitle = paste0("Total variance explained: ",
                      signif(sum(var_frac[1:3]) * 100, 3), "%"),
    y = paste0("PC3; ", signif(var_frac[3] * 100, digits = 2), " %"),
    x = paste0("PC1; ", signif(var_frac[1] * 100, digits = 2), " %"),
    title = plot.title,
    caption = "") + 
  ggplot2::theme_minimal() +
  ggplot2::theme(
    axis.text = ggplot2::element_text(size = 15),
    axis.title.x = ggplot2::element_text(size = 20),
    axis.title.y = ggplot2::element_text(size = 20),
    legend.text = ggplot2::element_text(size = 15)
  )

print(pca.plot)
ggplot2::ggsave(pca.plot, bg = "white",
                filename = paste0(figure.path,"Toothfish_2024_fin_RANDOM_PCA_CpG_DMS_PC13.png"), 
                width = 30, height = 15, units = "cm")




# data <- data.frame(ind = rownames(PCobj$x),PCobj$x,  pop = pop)
data <- data.frame(ind = rownames(PCobj$rotation),PCobj$rotation,  pop = pop)
data <- merge(data, aggregate(cbind(mean.x = PC2, mean.y = PC3) ~
                                pop, data, mean), by = "pop")

pca.plot <- ggplot2::ggplot(data, ggplot2::aes(PC2, PC3, color = pop)) +
  ggplot2::geom_point(size = 4) + 
  ggplot2::geom_point(ggplot2::aes(x = mean.x,
                                   y = mean.y, color = pop),
                      size = 7,
                      shape = 17) +
  ggplot2::scale_colour_manual(values = colours2) +
  ggplot2::geom_hline(yintercept = 0) +
  ggplot2::geom_vline(xintercept = 0) + 
  ggplot2::labs(
    subtitle = paste0("Total variance explained: ",
                      signif(sum(var_frac[1:3]) * 100, 3), "%"),
    y = paste0("PC3; ", signif(var_frac[3] * 100, digits = 2), " %"),
    x = paste0("PC2; ", signif(var_frac[2] * 100, digits = 2), " %"),
    title = plot.title,
    caption = "") + 
  ggplot2::theme_minimal() +
  ggplot2::theme(
    axis.text = ggplot2::element_text(size = 15),
    axis.title.x = ggplot2::element_text(size = 20),
    axis.title.y = ggplot2::element_text(size = 20),
    legend.text = ggplot2::element_text(size = 15)
  )

print(pca.plot)
ggplot2::ggsave(pca.plot, bg = "white",
                filename = 
                  paste0(figure.path,"Toothfish_2024_fin_RANDOM_PCA_CpG_DMS_PC23.png"), 
                width = 30, height = 15, units = "cm")
```


## DMS from 2017/2018
```{r}
print(load(paste0(rdata.path,"Toothfish_DMS_CpGsites.Rdata")))
# meth <- meth.2024F.norm15L
meth.cpg <- paste0(myDiff25p$chr, "_", myDiff25p$start) #80 
DMS.cpg <- unique(c(DMS$DMS)) #1243
sum(meth.cpg %in% DMS.cpg) #3 = 0.2%
# meth.DMS <- meth[meth.cpg %in% DMS.cpg,] #
```


## DMS from 2024 muscle
```{r}
print(load(paste0(rdata.path,"Toothfish_2024_muscle_methylation_DMS.Rdata")))
DMS.cpg <- paste0(myDiff25p$chr, "_", myDiff25p$start) #1211
print(load(paste0(rdata.path,"Toothfish_2024_fin_methylation_DMS.Rdata")))
meth.cpg <- paste0(myDiff25p$chr, "_", myDiff25p$start) #115 
sum(meth.cpg %in% DMS.cpg) #8 = 0.6%
```


## TOP100 DMS - 2024 - fin

```{r}
print(load(paste0(rdata.path,"Toothfish_2024_fin_methylkit_object.Rdata")))
print(load(paste0(rdata.path,"Toothfish_metadata.Rdata"))) 
metadata.new <- metadata[metadata$Genotype_names %in% meth.2024F.norm15L@sample.ids,]
metadata.new <- metadata.new[order(match(metadata.new$Genotype_names,
                                         meth.2024F.norm15L@sample.ids)),]
meth.cpg <- paste0(meth.2024F.norm15L$chr, "_", meth.2024F.norm15L$start) #561964 
length(meth.cpg)
print(load(paste0(rdata.path,"Toothfish_2024_fin_methylation_DMS.Rdata")))
DMS.cpg <- paste0(as.character(top100.2024F@seqnames), "_", as.data.frame(top100.2024F@ranges)$start) 
sum(meth.cpg %in% DMS.cpg) #100
meth.DMS <- meth.2024F.norm15L[meth.cpg %in% DMS.cpg,]
```


### Heatmap

```{r}
meth.diff.df <- as.data.frame(t(methylKit::percMethylation(meth.DMS)))
loc.names <- paste0(meth.DMS$chr,"_",meth.DMS$start)
colnames(meth.diff.df) <- loc.names

meth.diff.df$individual <- rownames(meth.diff.df)
pop <- metadata.new$pop
meth.diff.df$pop <- as.character(pop)


heat.plot.data <- meth.diff.df %>%
  tibble::as_tibble() %>%
  tidyr::gather(key = "CpG", value = "Methylation_percentage", -individual, -pop,
                factor_key = FALSE) %>%
  dplyr::group_by(pop, CpG) %>%
  dplyr::mutate(CpG_mean = mean(Methylation_percentage, na.rm = TRUE)) %>%
  dplyr::arrange(pop, individual, CpG) %>%
  dplyr::mutate(individual = factor(individual, levels = unique(individual))) %>%
  dplyr::mutate(CpG = factor(CpG, levels = unique(CpG)))
cpgs <- heat.plot.data$CpG[heat.plot.data$individual == unique(heat.plot.data$individual)[1]]
cpgs_mean <- heat.plot.data$CpG_mean[heat.plot.data$individual == unique(heat.plot.data$individual)[1]]

cpgs <- cpgs[order(cpgs_mean, decreasing = TRUE)]
heat.plot.data$CpG <- factor(heat.plot.data$CpG, levels = c("pop", unique(as.character(cpgs))))
heat.plot.data <- heat.plot.data[order(heat.plot.data$CpG),]

heat.plot <- ggplot2::ggplot(data = heat.plot.data, ggplot2::aes(x = individual)) + 
    ggplot2::geom_tile(data = heat.plot.data, 
                       ggplot2::aes(x = individual, y = CpG, fill = Methylation_percentage)) +
    ggplot2::scale_fill_gradient2(low = "blue", mid = "yellow", high = "red", 
                                  midpoint = 50, guide = "colourbar") +
    ggnewscale::new_scale_fill() +
    ggplot2::geom_tile(data = heat.plot.data, ggplot2::aes(x = individual, y = "pop", fill = pop,
                                    height = length(unique(heat.plot.data$CpG))*0.02)) +
    ggplot2::scale_fill_manual(values = colours6m) +
    ggplot2::scale_y_discrete(drop = FALSE) +
    ggplot2::theme(legend.position = "top",
                   axis.title.y = ggplot2::element_blank(),
                   axis.title.x = ggplot2::element_blank(),
                   axis.text.x = ggplot2::element_text(face = "bold", color = "black", 
                                        size = 10, angle = 90, hjust = 1, vjust = 0.5))
print(heat.plot)

ggplot2::ggsave(heat.plot, bg = "white",
                filename = paste0(figure.path,
                                  "Toothfish_2024_fin_BASED_ON_ALL_TOP100_DMS_HEATMAP.png"), 
                width = 30, height = 30, units = "cm")
```

### Plot CpG difference

```{r}
meth.diff <- meth.DMS

length(unique(meth.diff$chr)) #28
sort(table(meth.diff$chr), decreasing = TRUE)[1:20]

meth.diff.df <- as.data.frame(t(methylKit::percMethylation(meth.diff)))
loc.names <- paste0(meth.diff$chr,"_",meth.diff$start)
colnames(meth.diff.df) <- loc.names

meth.diff.df$individual <- rownames(meth.diff.df) 
identical(metadata.new$Genotype_names, meth.diff.df$individual)
meth.diff.df$pop <- metadata.new$pop



meth.diff.df.plot <- tidyr::gather(meth.diff.df, key = locus, 
                           value = perc_methylation, -individual, -pop,
                           factor_key = TRUE) %>%
  dplyr::mutate(chr = stringr::str_split_fixed(locus, 
                                               pattern = "_", 
                                               n = 2)[,1])
meth.diff.df.plot.locus <- meth.diff.df.plot %>% dplyr::group_by(chr,locus,pop) %>% 
  dplyr::summarise(mean_pop = mean(perc_methylation, na.rm = TRUE),
                   sum_NA = sum(is.na(perc_methylation)),
            .groups = 'drop')


diff.plot <- ggplot2::ggplot() +
  ggplot2::geom_line(data = meth.diff.df.plot.locus,
                     ggplot2::aes(x = pop, y = mean_pop, colour = locus, group = locus)) +
  ggplot2::geom_point(data = meth.diff.df.plot.locus,
                     ggplot2::aes(x = pop, y = mean_pop, colour = locus, group = locus),
                     size = 5) +
  ggplot2::facet_wrap(~chr) +
  ggplot2::ylab("Mean methylation (%)") +
  ggplot2::theme(legend.position = "none",
                   axis.text.x = ggplot2::element_text(face = "bold", color = "black", 
                                        size = 10, angle = 90, hjust = 1, vjust = 0.5))

print(diff.plot)

ggplot2::ggsave(plot = diff.plot, 
                filename = paste0(figure.path,"1_Toothfish_2024_fin_BASED_ON_ALL_TOP100_DMS_methDiff.png"),
                width = 30, height = 15, units = "cm", device = "png")
```



### DMS Clustering
```{r}
methdif.perc <- methylKit::percMethylation(meth.DMS)
methdif.perc2 <- zoo::na.aggregate(methdif.perc, by = meth.DMS@treatment)
PCobj <- prcomp(methdif.perc2, retx = TRUE, scale. = TRUE, center = TRUE)

var_frac <- PCobj$sdev/sum(PCobj$sdev)
print(paste0("Total variance explained: ",signif(sum(var_frac[1:3]) * 100, 3), "%")) 
barplot(var_frac[1:3], main = "variance explained", 
        col = heat.colors(length(var_frac[1:3])))

# 
identical(metadata.new$Genotype_names, meth.DMS@sample.ids)
identical(metadata.new$Genotype_names, rownames(PCobj$rotation))
# identical(metadata.new$File_names, meth.DMS@sample.ids)
# identical(metadata.new$File_names, rownames(PCobj$rotation))
pop <- metadata.new$pop

plot.title <- "Toothfish methylation - 2024 fin (TOP100 DMS from 2024 fin)"

data <- data.frame(ind = rownames(PCobj$rotation),PCobj$rotation,  pop = pop)
data <- merge(data, aggregate(cbind(mean.x = PC1, mean.y = PC2) ~
                                pop, data, mean), by = "pop")

pca.plot <- ggplot2::ggplot(data, ggplot2::aes(PC1, PC2, color = pop)) +
  ggplot2::geom_point(size = 4) + 
  ggplot2::geom_point(ggplot2::aes(x = mean.x,
                                   y = mean.y, color = pop),
                      size = 7,
                      shape = 17) +
  ggplot2::scale_colour_manual(values = colours2) +
  ggplot2::geom_hline(yintercept = 0) +
  ggplot2::geom_vline(xintercept = 0) + 
  ggplot2::labs(
    subtitle = paste0("Total variance explained: ",
                      signif(sum(var_frac[1:3]) * 100, 3), "%"),
    y = paste0("PC2; ", signif(var_frac[2] * 100, digits = 2), " %"),
    x = paste0("PC1; ", signif(var_frac[1] * 100, digits = 2), " %"),
    title = plot.title,
    caption = "") + 
  ggplot2::theme_minimal() +
  ggplot2::theme(
    axis.text = ggplot2::element_text(size = 15),
    axis.title.x = ggplot2::element_text(size = 20),
    axis.title.y = ggplot2::element_text(size = 20),
    legend.text = ggplot2::element_text(size = 15)
  )

print(pca.plot)
ggplot2::ggsave(pca.plot, bg = "white",
                filename = paste0(figure.path,"Toothfish_2024_fin_BASED_ON_ALL_TOP100_DMS_PC12.png"), 
                width = 30, height = 15, units = "cm")



data <- data.frame(ind = rownames(PCobj$rotation),PCobj$rotation,  pop = pop)
data <- merge(data, aggregate(cbind(mean.x = PC1, mean.y = PC3) ~
                                pop, data, mean), by = "pop")

pca.plot <- ggplot2::ggplot(data, ggplot2::aes(PC1, PC3, color = pop)) +
  ggplot2::geom_point(size = 4) + 
  ggplot2::geom_point(ggplot2::aes(x = mean.x,
                                   y = mean.y, color = pop),
                      size = 7,
                      shape = 17) +
  ggplot2::scale_colour_manual(values = colours6m) +
  ggplot2::geom_hline(yintercept = 0) +
  ggplot2::geom_vline(xintercept = 0) + 
  ggplot2::labs(
    subtitle = paste0("Total variance explained: ",
                      signif(sum(var_frac[1:3]) * 100, 3), "%"),
    y = paste0("PC3; ", signif(var_frac[3] * 100, digits = 2), " %"),
    x = paste0("PC1; ", signif(var_frac[1] * 100, digits = 2), " %"),
    title = plot.title,
    caption = "") + 
  ggplot2::theme_minimal() +
  ggplot2::theme(
    axis.text = ggplot2::element_text(size = 15),
    axis.title.x = ggplot2::element_text(size = 20),
    axis.title.y = ggplot2::element_text(size = 20),
    legend.text = ggplot2::element_text(size = 15)
  )

print(pca.plot)
ggplot2::ggsave(pca.plot, bg = "white",
                filename = paste0(figure.path,
                                  "Toothfish_2024_fin_BASED_ON_ALL_TOP100_DMS_PC13.png"), 
                width = 30, height = 15, units = "cm")




data <- data.frame(ind = rownames(PCobj$rotation),PCobj$rotation,  pop = pop)
data <- merge(data, aggregate(cbind(mean.x = PC2, mean.y = PC3) ~
                                pop, data, mean), by = "pop")

pca.plot <- ggplot2::ggplot(data, ggplot2::aes(PC2, PC3, color = pop)) +
  ggplot2::geom_point(size = 4) + 
  ggplot2::geom_point(ggplot2::aes(x = mean.x,
                                   y = mean.y, color = pop),
                      size = 7,
                      shape = 17) +
  ggplot2::scale_colour_manual(values = colours6m) +
  ggplot2::geom_hline(yintercept = 0) +
  ggplot2::geom_vline(xintercept = 0) + 
  ggplot2::labs(
    subtitle = paste0("Total variance explained: ",
                      signif(sum(var_frac[1:3]) * 100, 3), "%"),
    y = paste0("PC3; ", signif(var_frac[3] * 100, digits = 2), " %"),
    x = paste0("PC2; ", signif(var_frac[2] * 100, digits = 2), " %"),
    title = plot.title,
    caption = "") + 
  ggplot2::theme_minimal() +
  ggplot2::theme(
    axis.text = ggplot2::element_text(size = 15),
    axis.title.x = ggplot2::element_text(size = 20),
    axis.title.y = ggplot2::element_text(size = 20),
    legend.text = ggplot2::element_text(size = 15)
  )

print(pca.plot)
ggplot2::ggsave(pca.plot, bg = "white",
                filename = paste0(figure.path,"Toothfish_2024_fin_BASED_ON_ALL_TOP100_DMS_PC23.png"), 
                width = 30, height = 15, units = "cm")
```






# 10_Methylation: ALL fin tissue

```{r}
print(load(paste0(rdata.path,"Toothfish_metadata.Rdata"))) #ALL fish
```

## Reading the methylation call files

```{r, eval=FALSE}
data.list <- list.files(
  path = 
  "/datasets/work/ncmi-toa-rrbs/work/1.Toothfish/GENOME/Combined_methylation/",
  pattern = "*.cov.gz")

subnames <- metadata$File_names[!(metadata$Tissue_type %in% 
                                    c("muscle", "muscle from under SD", "Skin") |
                                    grepl(pattern = "REP", x = metadata$File_names) 
                                  )]

subnames <- subnames[!subnames %in% "TOA_88-2_P79_AD294619_fin_F"] # low coverage

data.list <- data.list[data.list %in% paste0(subnames,
                                             "_R1_trimmed_sort_bismark_bt2.bismark.cov.gz")]

data.loc <- as.list(paste0(
  "/datasets/work/ncmi-toa-rrbs/work/1.Toothfish/GENOME/Combined_methylation/",
                           data.list))



data.list2 <-  gsub(pattern = '_R1_trimmed_sort_bismark_bt2.bismark.cov.gz',
                    replacement = '',x = as.character(data.list))
# data.list2 <-  gsub(pattern = '_sort_bismark_bt2_PE.bismark.cov.gz',
#                     replacement = '',x = as.character(data.list2))

metadata.new <- metadata[metadata$File_names %in% data.list2,]
metadata.new <- metadata.new[order(match(metadata.new$File_names, data.list2)),]

metadata.new$File_names == data.list2

sample.ids <- metadata.new$Genotype_names
pop <- as.integer(factor(metadata.new$pop))

myobjDB <- methylKit::methRead(location = data.loc,
           sample.id = as.list(sample.ids),
           assembly = "TOA_genome",
           treatment = pop,
           context = "CpG",
           header = FALSE, 
           dbtype = "tabix",
           pipeline = "bismarkCoverage",
           mincov = 3
           )
save(myobjDB, file = paste0(rdata.path,"Toothfish_ALL_fin_methylkit_cpgDB.Rdata"))
```


## Data QC

#### Number of sites in genome

```{r, eval=FALSE}
cov.df <- data.frame(chr = NA, start = NA,end = NA, strand = NA, coverage = NA, 
                     numCs = NA, numTs = NA, sample = NA)
for (i in 1:length(sample.ids)) {
  df <- methylKit::getData(myobjDB[[i]])
  df$sample <- sample.ids[i]
  cov.df <- rbind(cov.df, df)
}
cov.df <- cov.df[-1,]
save(cov.df, file = paste0(rdata.path,"Toothfish_ALL_fin_methylkit_cpg_cov.Rdata"))
```

### Coverage stats

```{r}
load(paste0(rdata.path,"Toothfish_ALL_fin_methylkit_cpgDB.Rdata"))
print(load(paste0(rdata.path,"Toothfish_ALL_fin_methylkit_cpg_cov.Rdata")))

length(unique(paste0(cov.df$chr, "_", cov.df$start))) #4,387,684

# summary(as.factor(cov.df$sample))
summary(cov.df$coverage)
    # Min.  1st Qu.   Median     Mean  3rd Qu.     Max. 
    # 4.00     6.00    11.00    14.63    18.00 23671.00

summary(cov.df$coverage[cov.df$coverage > 5 & cov.df$coverage < 50])
   # Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
   # 6.00    8.00   13.00   15.07   19.00   49.00 

cov.plot <- ggplot2::ggplot(cov.df, ggplot2::aes(x = log(coverage), fill = sample)) +
  ggplot2::geom_histogram() +
  ggplot2::theme(legend.position = "none")
# print(cov.plot)
ggplot2::ggsave(cov.plot,
                filename =
                  paste0(figure.path,"Toothfish_ALL_fin_coverage_hist_Combined.png"), 
                width = 45, height = 15, units = 'cm')

cov.plot <- ggplot2::ggplot(cov.df, ggplot2::aes(x = coverage, fill = sample)) +
  ggplot2::geom_histogram() +
  ggplot2::theme(legend.position = "none")
# print(cov.plot)
ggplot2::ggsave(cov.plot,
                filename =
                  paste0(figure.path,"Toothfish_ALL_fin_coverage_hist_Combined2.png"), 
                width = 45, height = 15, units = 'cm')

cov.plot <- ggplot2::ggplot(cov.df[cov.df$coverage > 5 & cov.df$coverage < 160,], 
                            ggplot2::aes(x = coverage)) +
  ggplot2::geom_histogram(binwidth = 1) +
  ggplot2::ggtitle("coverage < 160") +
  ggplot2::theme(legend.position = "none")
# print(cov.plot)
ggplot2::ggsave(cov.plot,
                filename =
                  paste0(figure.path,"Toothfish_ALL_fin_coverage_hist_Combined4.png"),
                width = 45, height = 15, units = 'cm')



max(cov.df$coverage, na.rm = T) #11562
ind.covplot <- ggplot2::ggplot(cov.df, ggplot2::aes(x = sample, y = coverage)) + 
  ggplot2::geom_boxplot() +
  ggplot2::theme_bw() + 
  ggplot2::theme(axis.title.x = ggplot2::element_blank(), 
                 axis.text.y = ggplot2::element_text(size = 15),
                 axis.title.y = ggplot2::element_text(size = 15),
                 axis.text.x = ggplot2::element_text(size = 7, angle = 90, hjust = 1)) +
  ggplot2::scale_y_continuous(trans = scales::log2_trans(),
                  breaks = c(1, 10, 50, 750, 20000),
                  minor_breaks = c(1:10, 2:5*10, 1:7.5*100, 0.5:2*1000))

print(ind.covplot)

ggplot2::ggsave(ind.covplot, bg = "white",
                filename = 
                  paste0(figure.path,
                         "Toothfish_ALL_fin_raw_individual_coverage_CpGs.png"),
                width = 45, height = 15, units = 'cm')
```


### Coverage filter

High count filter based on coverage stats.

```{r, eval=FALSE}
load(paste0(rdata.path,"Toothfish_ALL_fin_methylkit_cpgDB.Rdata"))

filtered.myobj <- methylKit::filterByCoverage(myobjDB, lo.count = 10, lo.perc = NULL,
                                              save.db = FALSE,
                                      hi.count = NULL, hi.perc = 99)

save(myobjDB,filtered.myobj,
     file = paste0(rdata.path,"Toothfish_ALL_fin_methylkit_cpgDB.Rdata"))
```

### Normalise coverage

Normalize read coverage distributions between samples. Requires stringent filter on max coverage first, otherwise it will over-sample PCR bias reads. These two functions will help reduce the bias in the statistical tests that might occur due to systematic over-sampling of reads in certain samples.

```{r}
filtered.myobj.normal <- methylKit::normalizeCoverage(obj = filtered.myobj, 
                                                      save.db = FALSE,
                                                      method = "median", chunk.size = 1e6)
save(myobjDB,filtered.myobj, filtered.myobj.normal,
     file = paste0(rdata.path,"Toothfish_ALL_fin_methylkit_cpgDB.Rdata"))
```

### Number of sites per sample

```{r, eval = FALSE}
sites.df <- data.frame(sites = NA,  sample = NA)
for (i in 1:length(filtered.myobj.normal)) {
  print(i)
  sites.df[i,] <- c(filtered.myobj.normal[[i]]@num.records, filtered.myobj.normal[[i]]@sample.id)
}
sites.df$sites <- as.integer(sites.df$sites)
# sites.df$pop <- filtered.myobj.normal@treatment
sites.df$pop <- metadata.new$pop
save(sites.df, file = paste0(rdata.path,"Toothfish_ALL_fin_number_cpg_sample.Rdata"))
```


```{r}
print(load(paste0(rdata.path,"Toothfish_ALL_fin_number_cpg_sample.Rdata")))
sites.plot <- ggplot2::ggplot(sites.df, ggplot2::aes(x = sample, y = sites, colour = pop)) +
  ggplot2::geom_point() +
  ggplot2::scale_color_manual(values = colours3f) +
  ggplot2::theme_bw() +
  ggplot2::theme(axis.text.x = ggplot2::element_text(angle = 90))
print(sites.plot)

ggplot2::ggsave(sites.plot,
                filename =
                  paste0(figure.path,"Toothfish_ALL_fin_number_cpg_sample.png"),
                width = 45, height = 15, units = 'cm')

sites.df$sample[sites.df$sites < 500000]
# "TOA_48-2_007_UK048"          "TOA_48-2_028_UK103"          "TOA_88-2_P67_AD294613_fin_F"
low.sites <- sites.df$sample[sites.df$sites < 500000]
```


## Select samples

```{r}
print(load(paste0(rdata.path,"Toothfish_metadata.Rdata")))

metadata2 <- metadata[!(metadata$Tissue_type %in% 
                                    c("muscle", "muscle from under SD", "Skin") |
                                    grepl(pattern = "REP", x = metadata$File_names) 
                                  ),]
print(load(paste0(rdata.path, "Low_quality_fish.Rdata")))


# rm1 <- Reduce(intersect, list(low.genome.mapping,low.genome.methylation))
rm1 <- unique(c(low.genome.mapping,low.genome.methylation, low.sites))
print(rm1)
rm3 <- c("TOA_58-4-2_B1_AE100278_fin_M",  "TOA_58-4-2_B11_AE100305_fin_M", 
         "TOA_58-4-2_B17_AE100308_fin_F", "TOA_58-4-2_B3_AE100279_fin_M",
         "TOA_58-4-2_B5_AE100280_fin_F" ) # bad coverage
rm4 <- c("TOA_48-2_031_UK116","TOA_58-4-2_111_AN172604","TOA_88-1_230_AD082543",
         "TOA_48-2_024_UK082")
print(rm4) #bad tissue PCA
rm <- c(rm1,rm3,rm4)
# rm <- c(rm1,rm2,rm3,rm4)


metadata2 <- metadata2[!(metadata2$File_names %in% rm | metadata2$Genotype_names %in% rm),]

filt.sample.id.full <- sapply(filtered.myobj.normal@.Data, function(x){x@sample.id})
filt.sample.id <- filt.sample.id.full[filt.sample.id.full %in% metadata2$Genotype_names]
metadata2 <- metadata2[order(match(metadata2$Genotype_names, filt.sample.id)),]
identical(filt.sample.id,metadata2$Genotype_names)


pop <- metadata2$pop
pop.levels <- c("48.2-2017", 
                "58.4.2-2024",
                "58.4.2-2024_50L",
                "58.4.2-2024_70L",
                "88.2-2024")

pop <- factor(pop, levels = pop.levels)
pop <- as.numeric(pop)

sample.ids <- metadata2$Genotype_names

filtered.myobj.normal@treatment <- as.numeric((factor(metadata2$STRATA2)))


filtered.myobj2.normal <- methylKit::reorganize(
  methylObj = filtered.myobj.normal,
  sample.ids = sample.ids,
  treatment = pop,
  chunk.size = 1e+06,
  save.db = TRUE,
  suffix = "TOA_2024_fin_normal"
)
save(myobjDB,filtered.myobj,
     filtered.myobj.normal,filtered.myobj2.normal,
     file = paste0(rdata.path,"Toothfish_ALL_fin_methylkit_cpgDB.Rdata"))
```


## Merging samples

```{r}
print(load(paste0(rdata.path,"Toothfish_ALL_fin_methylkit_cpgDB.Rdata")))
pop <- metadata2$pop
pop.levels <- c("48.2-2017", 
                "58.4.2-2024",
                "58.4.2-2024_50L",
                "58.4.2-2024_70L",
                "88.2-2024")
pop <- as.numeric(factor(pop, levels = pop.levels))
table(pop)

filtered.myobj2.normal@treatment <- as.numeric((factor(metadata2$STRATA2)))
table(filtered.myobj2.normal@treatment )

meth.ALLF.norm10L <- methylKit::unite(filtered.myobj2.normal, destrand = FALSE,
                                      save.db = FALSE, 
                                      min.per.group = 10L, #missing data allowed
                                      mc.cores = 25)
#318784  for 10L missing per group

save(meth.ALLF.norm10L,
     file = paste0(rdata.path,"Toothfish_ALL_fin_methylkit_object.Rdata"))
```


## Data QC

### Remove ASM

```{r}
load(paste0(rdata.path,"Allele_specific_mutations.Rdata"))

mut <- as(asmTRUE,"GRanges")
sum(as(meth.ALLF.norm10L,"GRanges") %over% mut)#13022
meth.asm <- methylKit::selectByOverlap(meth.ALLF.norm10L, mut) 
BOOLEAN <- paste0(meth.ALLF.norm10L$chr,"_",meth.ALLF.norm10L$start) %in%
  paste0(asmTRUE$Chr,"_",asmTRUE$Pos)
sum(BOOLEAN) #288
meth.ALLF.norm10L <- methylKit::select(meth.ALLF.norm10L, !BOOLEAN)    

save(meth.ALLF.norm10L,
     file = paste0(rdata.path,"Toothfish_ALL_fin_methylkit_object.Rdata"))
```



### Coverage 

```{r}
meth.df <- methylKit::getData(meth.ALLF.norm10L)
cov <- meth.df[,meth.ALLF.norm10L@coverage.index]
colnames(cov) <- meth.ALLF.norm10L@sample.ids
dpf <- data.frame(cov, locus = paste0(meth.df$chr,"_",meth.df$start))
dpf <- tidyr::gather(dpf, key = individual, 
                           value = Depth, -locus,
                           factor_key = TRUE) %>%
  dplyr::group_by(individual) %>%
  dplyr::mutate(NA_cnt = sum(is.na(Depth))/length(unique(locus))) %>%
  dplyr::ungroup()
dpf.NA <- dpf[!duplicated(dpf$individual),]
dpf.NA$NA_cnt <- round(dpf.NA$NA_cnt, 3)

dpf <- dpf[ dpf$Depth > 0 & !is.na(dpf$Depth),]

max(dpf$Depth, na.rm = T) #137
myPlot <- ggplot2::ggplot(dpf, ggplot2::aes(x = individual, y = Depth)) + 
  ggplot2::geom_boxplot() +
  ggplot2::geom_text(data = dpf.NA,ggplot2::aes(x = individual, y = 150, label = NA_cnt), 
                     size = 2, angle = 90) +
  ggplot2::theme_bw() + 
  ggplot2::theme(axis.title.x = ggplot2::element_blank(), 
                 axis.text.y = ggplot2::element_text(size = 15),
                 axis.title.y = ggplot2::element_text(size = 15),
                 axis.text.x = ggplot2::element_text(size = 7, angle = 90, hjust = 1)) +
  ggplot2::scale_y_continuous(trans = scales::log2_trans(),
                  breaks = c(1, 10, 50, 150),
                  minor_breaks = c(1:10, 2:5*10))

print(myPlot)
ggplot2::ggsave(myPlot, bg = "white",
                filename = 
                  paste0(figure.path,
                  "1B.Toothfish_ALL_fin_normalised15_methylkit_individual_coverage_CpGs.png"
                                  ), 
                width = 45, height = 15, units = "cm")

nrow(cov) # 305129 SMPs
plot(x = 1:ncol(cov), y = colMeans(cov, na.rm = TRUE), col = meth.ALLF.norm10L@treatment)
plot(x = 1:ncol(cov), y = colMeans(cov > 20, na.rm = TRUE), col = meth.ALLF.norm10L@treatment)
plot(x = 1:ncol(cov), y = colMeans(cov > 10, na.rm = TRUE), col = meth.ALLF.norm10L@treatment)
meth.ALLF.norm10L@sample.ids[colMeans(cov > 10, na.rm = TRUE) < 0.99]
# samples with low coverage: 
# [1] "TOA_58-4-2_B1_AE100278_fin_M"  "TOA_58-4-2_B11_AE100305_fin_M" "TOA_58-4-2_B17_AE100308_fin_F"
# [4] "TOA_58-4-2_B3_AE100279_fin_M"  "TOA_58-4-2_B5_AE100280_fin_F" 
```


## Load data

```{r}
print(load(paste0(rdata.path,"Toothfish_ALL_fin_methylkit_object.Rdata")))
print(load(paste0(rdata.path,"Toothfish_metadata.Rdata"))) 

metadata.new <- metadata[metadata$Genotype_names %in% meth.ALLF.norm10L@sample.ids,]
metadata.new <- metadata.new[order(match(metadata.new$Genotype_names,
                                         meth.ALLF.norm10L@sample.ids)),]

length(meth.ALLF.norm10L@row.names) #number of sites
```


## Global clustering analysis

```{r}
methylKit::clusterSamples(meth.ALLF.norm10L, dist = "correlation", sd.filter = TRUE, 
                          method = "ward.D2", plot = TRUE, )
dev.print(file = paste0(figure.path,"1_TOA_ALL_fin_cluster_plot.png"),
          device = png, res = 300,width = 45,height = 15,units = "cm")
```


### PCA

```{r}
PCobj <- methylKit::PCASamples(meth.ALLF.norm10L, comp = c(1,2), 
                               adj.lim = c(0,0), obj.return = TRUE)

var_frac <- PCobj$sdev/sum(PCobj$sdev)
print(paste0("Total variance explained: ",signif(sum(var_frac[1:3]) * 100, 3), "%")) 
barplot(var_frac[1:20], main = "variance explained", 
        col = heat.colors(length(var_frac[1:20])))

identical(metadata.new$Genotype_names, meth.ALLF.norm10L@sample.ids)
identical(metadata.new$Genotype_names, rownames(PCobj$x))

pop <- metadata.new$pop
pool <- metadata.new$STRATA4

plot.title <- "Toothfish methylation - ALL fin (318,496 CpG sites)"

data <- data.frame(ind = rownames(PCobj$x),PCobj$x,  pop = pop, pool = pool)
data <- merge(data, aggregate(cbind(mean.x = PC1, mean.y = PC2) ~
                                pop, data, mean), by = "pop")

pca.plot <- ggplot2::ggplot(data) +
  ggplot2::geom_point(ggplot2::aes(PC1, PC2, color = pop, shape = pool),
                      size = 4) + 
  ggplot2::geom_point(ggplot2::aes(x = mean.x,
                                   y = mean.y, color = pop),
                      size = 7,
                      shape = 17) +
  ggplot2::scale_shape_manual(values = c(19,18)) +
  ggplot2::scale_colour_manual(values = colours3f) +
  ggplot2::geom_hline(yintercept = 0) +
  ggplot2::geom_vline(xintercept = 0) + 
  ggplot2::labs(
    subtitle = paste0("Total variance explained: ",
                      signif(sum(var_frac[1:4]) * 100, 3), "%"),
    y = paste0("PC2; ", signif(var_frac[2] * 100, digits = 2), " %"),
    x = paste0("PC1; ", signif(var_frac[1] * 100, digits = 2), " %"),
    title = plot.title,
    caption = "") + 
  ggplot2::theme_minimal() +
  ggplot2::theme(
    axis.text = ggplot2::element_text(size = 15),
    axis.title.x = ggplot2::element_text(size = 20),
    axis.title.y = ggplot2::element_text(size = 20),
    legend.text = ggplot2::element_text(size = 15)
  )

print(pca.plot)
ggplot2::ggsave(pca.plot, bg = "white",
                filename = paste0(figure.path,"Toothfish_ALL_fin_PCA_CpG_PC12.png"), 
                width = 30, height = 15, units = "cm")




data <- data.frame(ind = rownames(PCobj$x),PCobj$x,  pop = pop, pool = pool)
data <- merge(data, aggregate(cbind(mean.x = PC2, mean.y = PC3) ~
                                pop, data, mean), by = "pop")
pca.plot <- ggplot2::ggplot(data) +
  ggplot2::geom_point(ggplot2::aes(PC2, PC3, color = pop, shape = pool),
                      size = 4) + 
  ggplot2::geom_point(ggplot2::aes(x = mean.x,
                                   y = mean.y, color = pop),
                      size = 7,
                      shape = 17) +
  ggplot2::scale_shape_manual(values = c(19,18)) +
  ggplot2::scale_colour_manual(values = colours3f) +
  ggplot2::geom_hline(yintercept = 0) +
  ggplot2::geom_vline(xintercept = 0) + 
  ggplot2::labs(
    subtitle = paste0("Total variance explained: ",
                      signif(sum(var_frac[1:4]) * 100, 3), "%"),
    y = paste0("PC3; ", signif(var_frac[3] * 100, digits = 2), " %"),
    x = paste0("PC2; ", signif(var_frac[2] * 100, digits = 2), " %"),
    title = plot.title,
    caption = "") + 
  ggplot2::theme_minimal() +
  ggplot2::theme(
    axis.text = ggplot2::element_text(size = 15),
    axis.title.x = ggplot2::element_text(size = 20),
    axis.title.y = ggplot2::element_text(size = 20),
    legend.text = ggplot2::element_text(size = 15)
  )

print(pca.plot)
ggplot2::ggsave(pca.plot, bg = "white",
                filename = paste0(figure.path,"Toothfish_ALL_fin_PCA_CpG_PC23.png"), 
                width = 30, height = 15, units = "cm")
```


### RDA

```{r}
# setwd("/datasets/work/ncmi-toa-rrbs/work/1.Toothfish/Toothfish_RRBS_array_22098078/Bismark/")
# print(load(paste0(rdata.path,"Toothfish_methylation_objects_raw.Rdata"))) #36,635 SMPs
print(load(paste0(rdata.path,"Toothfish_subset_methylation_objects_raw.Rdata"))) #1,316,799 SMPs

# meth.RDA <- methylKit::unite(filtered.myobj2.norm, destrand = FALSE,
#                              save.db = FALSE,
#                              suffix = "subset",
#                              min.per.group = NULL, #missing data allowed
#                              mc.cores = 25)


# meth.df <- methylKit::getData(meth)
# cov <- meth.df[,grep(colnames(meth.df), pattern = "coverage*")]
# cov <- meth.df[,meth@coverage.index]
# colnames(cov) <- meth@sample.ids
# meth.df <- methylKit::percMethylation(meth)

# meth.df <- methylKit::percMethylation(meth.RDA)

# save(meth.df,meth.RDA, 
#      file = paste0(rdata.path,"TOA_data_for_RDA_meth_percent.Rdata"))
print(load(paste0(rdata.path,"TOA_data_for_RDA_meth_percent.Rdata"))) #347,614 SMPs

hist(colSums(is.na(meth.df)), breaks = 100)
hist(rowSums(is.na(meth.df)), breaks = 100)





print(load(paste0(rdata.path,"Toothfish_metadata.Rdata")))

predictor <- metadata2 # define the predictor input file here
predictor <- predictor[predictor$SAMPLE_ID %in% meth@sample.ids,]
predictor$SAMPLE_ID = as.character(predictor$SAMPLE_ID)
predictor$region = as.character(predictor$region)
predictor$Sex = as.character(predictor$Sex)
predictor$Tissue = as.character(predictor$Tissue)
predictor$mat = as.character(predictor$mat)
predictor$mat[is.na(predictor$mat)] <- "2"
predictor$TL = as.numeric(predictor$TL)
predictor$wt = as.numeric(predictor$wt)
predictor$dpth = as.numeric(predictor$dpth)
predictor$Age = as.numeric(predictor$Age)
predictor$Age[is.na(predictor$Age)] <- as.numeric(max(predictor$Age, na.rm = TRUE))
predictor$Cohort2 = as.integer(predictor$Cohort2)
predictor$Cohort2[is.na(predictor$Cohort2)] <- as.integer(min(predictor$Cohort2, na.rm = TRUE))


meth.df1 = t(meth.df)
# meth.df2 = na.omit(meth.df1) #rda can't handle NAs, remove them if you have them
# meth.df1[1:10,1:10]
## Predictor
# Make names characters (not factors)


#if you had numbers, they had to be factors (as.factor)

# Confirm that methylation data and predictors are in the same order
identical(rownames(meth.df1), predictor$SAMPLE_ID)

# Perform Hellinger transformation on the methylation data:
meth.df3 = vegan::decostand(meth.df1, method = "hellinger", MARGIN = 1) #MARGIN=1 means rows, on rows I have samples now because I transposed the matrix

## Run RDA
#Test the country using growing condition as a Condition () (covariate):
meth_rda = vegan::rda(meth.df3 ~ region + Condition(Sex), 
                      data = predictor, scale = F) #I remove scale since I performed Hellinger transformation
meth_rda = vegan::rda(meth.df3 ~ region + Condition(Sex) + 
                        Condition(mat)+ Condition(TL) + Condition(wt) +
                        Condition(dpth) + Condition(Age) + Condition(Cohort2), 
                      na.action = "na.omit",
                      data = predictor, scale = F)

meth_rda$pCCA
meth_rda$CCA$wa

summary(meth_rda) #Constrained Proportion: variance of Y explained by X, Unconstrained Proportion: unexplained variance in Y
# sink(paste0(rdata.path,"anova_rda.txt")) #with this you can save the output of the anova in a txt file, define the output directory
meth.aov <- anova(meth_rda, test = "Chi", permutations = 99) #run Anova on the result to see the significance level
# meth.aov.cca <- anova.cca(meth_rda, test = "Chi", permutations = 99)
print(meth.aov)
# sink(paste0(rdata.path,"anova_rda_chi.txt"))

# Model: rda(formula = meth.df3 ~ region + Condition(Sex), data = predictor, scale = F)
#           Df  Variance      F Pr(>F)   
# Model      3 0.0003924 1.4126  0.004 **
# Residual 110 0.0101861

# Model: rda(formula = meth.df3 ~ region + Condition(Sex) + Condition(mat) + Condition(TL) + Condition(wt) + Condition(dpth) + Condition(Age) + Condition(Cohort2), data = predictor, scale = F, na.action = "na.omit")
#           Df  Variance     F Pr(>F)   
# Model      3 0.0003412 1.257   0.01 **
# Residual 101 0.0091396 

Pop.names <- predictor$region
plot.a <- vegan::ordiplot(meth_rda, choices=c(1,2))
Pop.ellipses <- vegan::ordiellipse(plot.a, groups=Pop.names, display="sites", kind="sd")
dev.print(file = paste0(figure.path,"TOA_RDA_ordiplot_complex_model.png"),
          device = png, res = 300,width = 45,height = 15,units = "cm")

# Type 1 scaling
vegan::ordiplot(meth_rda, scaling = 1, type = "text")
# Type 2 scaling
vegan::ordiplot(meth_rda, scaling = 2, type = "text")

ggvegan::autoplot(
  meth_rda,
  axes = c(1, 2),
  geom = "text",
  layers = c("species", "sites", "biplot", "centroids"),
  arrows = TRUE)

#CpG contribution
hist(meth_rda$CCA$v[,1])
hist(meth_rda$CCA$v[,2])
hist(meth_rda$CCA$v[,3])



#Test the interaction of your predictors:
meth_rda3 = vegan::rda(meth ~ region * Sex + Condition(region + Sex), data = predictor, scale = F)
summary(meth_rda3)
sink(paste0(rdata.path,"anova_rda_interaction.txt"))
print(anova(meth_rda3, test="Chi"))
sink(paste0(rdata.path,"anova_rda_interaction_chi.txt"))
```

#### plot RDA

```{r}

BioR.theme <- theme(
        panel.background = element_blank(),
        panel.border = element_blank(),
        panel.grid = element_blank(),
        axis.line = element_line("gray25"),
        text = element_text(size = 12),
        axis.text = element_text(size = 10, colour = "gray25"),
        axis.title = element_text(size = 14, colour = "gray25"),
        legend.title = element_text(size = 14),
        legend.text = element_text(size = 14),
        legend.key = element_blank())

plot.a <- vegan::ordiplot(meth_rda, choices=c(1,2))
Pop.names <- predictor$region

# sites1 <- BiodiversityR::sites.long(plot.a, env.data=predictor)
sites1 <- data.frame(cbind(predictor, axis1=plot.a$sites[, 1], 
                           axis2=plot.a$sites[, 2], 
                           labels=rownames(plot.a$sites)),
                     group = predictor$region)

# axis1 <- BiodiversityR::axis.long(meth_rda, choices=c(1, 2))
eigs.all <- meth_rda$CA$eig
eigs <- round(100 * eigs.all / meth_rda$tot.chi, digits=1)[c(1, 2)] 
xlab.label <- paste0(names(eigs)[1], " (", eigs[1], "%)")
ylab.label <- paste0(names(eigs)[2], " (", eigs[2], "%)")
axis1 <- data.frame(axis=c(1:2), 
                    ggplot=c("xlab.label", "ylab.label"), 
                    label=c(xlab.label, ylab.label))


# centroids.Pop1 <- BiodiversityR::centroids.long(sites1, 
#                                                 grouping=Pop.names, 
#                                                 centroids.only=TRUE)
# centroids.Pop2 <- BiodiversityR::centroids.long(sites1, 
#                                                 grouping=Pop.names, 
#                                                 centroids.only=FALSE)
gr.name <- rlang::enexpr(Pop.names)
cent.means <-  stats::aggregate(cbind(axis1, axis2) ~ Pop.names,
                                data = sites1,
                                FUN = mean)
names(cent.means) <- c("group", "axis1c", "axis2c")
centroids.Pop1 <- cent.means
cent.means$Centroid <- cent.means[, 1]
by.name <- names(sites1)[names(sites1) == "group"]
centroids.Pop2 <- dplyr::full_join(sites1, cent.means, by = by.name)


# Pop.ellipses.long2 <- BiodiversityR::ordiellipse.long(Pop.ellipses, 
#                                                     grouping.name="Pop.names")


plotgg.a <- ggplot() + 
    geom_vline(xintercept = c(0), color = "grey70", linetype = 2) +
    geom_hline(yintercept = c(0), color = "grey70", linetype = 2) +  
    xlab(axis1[1, "label"]) +
    ylab(axis1[2, "label"]) +  
    scale_x_continuous(sec.axis = dup_axis(labels=NULL, name=NULL)) +
    scale_y_continuous(sec.axis = dup_axis(labels=NULL, name=NULL)) +
    # geom_polygon(data=Pop.ellipses.long2, 
    #                aes(x=axis1, y=axis2, colour=Pop.names, 
    #                    fill=after_scale(alpha(colour, 0.2))), 
    #           size=0.2, show.legend=TRUE) +
    geom_segment(data=centroids.Pop2, 
                 aes(x=axis1c, y=axis2c, xend=axis1, yend=axis2, colour=Pop.names), 
                 size=0.7, show.legend=FALSE) +
    geom_point(data=centroids.Pop1, 
               aes(x=axis1c, y=axis2c), 
               shape="square", colour="black", alpha=0.7, size=2) +
    geom_label_repel(data=centroids.Pop1, 
               aes(x=axis1c, y=axis2c, label=group, colour=group), 
               alpha=1.0, size=2, show.legend=FALSE) +
    # labs(colour = Pop.names) +
    BioR.theme +
    theme(legend.text = element_text(size = 10)) +
    scale_colour_npg() +
    coord_fixed(ratio=1)

plotgg.a

ggsave(plotgg.a, file = paste0(figure.path, 
                               # "TOA_RDA_simple_model_A.png"
                               "TOA_RDA_complex_model_A.png"
                               ),
       height = 30, width = 30, units = "cm",  device = "png")
```



## DMS - ALL - fin

```{r, eval = FALSE}
print(load(paste0(rdata.path,"Toothfish_ALL_fin_methylkit_object.Rdata")))

print(load(paste0(rdata.path,"Toothfish_metadata.Rdata")))
metadata.new <- metadata[metadata$Genotype_names %in% attributes(meth.ALLF.norm10L)$sample.ids,]
attributes(meth.ALLF.norm10L)$sample.ids == metadata.new$Genotype_names

# meth.ALLF.norm10L@treatment <- as.integer(factor(metadata.new$pop))
meth.ALLF.norm10L@treatment <- as.integer(factor(metadata.new$STRATA2))

# data.frame(attributes(meth.ALLF.norm10L)$sample.ids, metadata.new$individual, metadata.new$strata, meth.ALLF.norm10L@treatment)

covariates <- dplyr::select(metadata.new, Sex, Age, Cohort2, Maturity_full)
myDiff <- methylKit::calculateDiffMeth(meth.ALLF.norm10L,
                                       covariates = NULL,
                                       adjust = "fdr",
                                       chunk.size = 1e+06,
                                       mc.cores = 25)
meth.diff.chr <- methylKit::diffMethPerChr(myDiff,plot = TRUE, 
                                           qvalue.cutoff = 0.01, meth.cutoff = 40)
myDiff40p <- methylKit::getMethylDiff(myDiff, difference = 40, qvalue = 0.01)
#378 
myDiff25p <- methylKit::getMethylDiff(myDiff, difference = 25, qvalue = 0.01)
#2975 
myDiff15p <- methylKit::getMethylDiff(myDiff, difference = 15, qvalue = 0.01)
#15749 
top100.ALLF <- slice_max(as.data.frame(as(myDiff,"GRanges")), order_by = meth.diff, n = 100)
top100.ALLF <- as(top100.ALLF,"GRanges")

save(myDiff,myDiff15p, myDiff25p, myDiff40p,top100.ALLF,
     file = paste0(rdata.path,"Toothfish_ALL_fin_methylation_DMS.Rdata"))
```


```{r}
print(load(paste0(rdata.path,"Toothfish_ALL_fin_methylation_DMS.Rdata")))

volcano.plot <- ggplot2::ggplot() +
  ggplot2::geom_point(data = myDiff, ggplot2::aes(x = meth.diff, 
                                                  y = -log10(qvalue))) +
  ggplot2::geom_point(data = myDiff25p, ggplot2::aes(x = meth.diff, 
                                                  y = -log10(qvalue)),
                      colour = "red") +
  ggplot2::theme_minimal() +
  ggplot2::theme(
    axis.text = ggplot2::element_text(size = 15),
    axis.title.x = ggplot2::element_text(size = 20),
    axis.title.y = ggplot2::element_text(size = 20),
    legend.text = ggplot2::element_text(size = 15)
  )
print(volcano.plot)

ggplot2::ggsave(volcano.plot, bg = "white",
                filename =
                  paste0(figure.path,
                         "Toothfish_ALL_fin_DMS25_VOLCANO_methylkit_regions.png"), 
                width = 30, height = 15, units = "cm")
```

### Heatmap

```{r}
print(load(paste0(rdata.path,"Toothfish_ALL_fin_methylation_DMS.Rdata")))

myDiff25p.gr <- as(myDiff25p, "GRanges")
meth.diff <- methylKit::selectByOverlap(meth.ALLF.norm10L, myDiff25p.gr)

meth.diff.df <- as.data.frame(t(methylKit::percMethylation(meth.diff)))
loc.names <- paste0(meth.diff$chr,"_",meth.diff$start)
colnames(meth.diff.df) <- loc.names

meth.diff.df$individual <- rownames(meth.diff.df)
pop <- metadata.new$pop
meth.diff.df$pop <- as.character(pop)


heat.plot.data <- meth.diff.df %>%
  tibble::as_tibble() %>%
  tidyr::gather(key = "CpG", value = "Methylation_percentage", -individual, -pop,
                factor_key = FALSE) %>%
  dplyr::group_by(pop, CpG) %>%
  dplyr::mutate(CpG_mean = mean(Methylation_percentage, na.rm = TRUE)) %>%
  dplyr::arrange(pop, individual, CpG) %>%
  dplyr::mutate(individual = factor(individual, levels = unique(individual))) %>%
  dplyr::mutate(CpG = factor(CpG, levels = unique(CpG)))
cpgs <- heat.plot.data$CpG[heat.plot.data$individual == unique(heat.plot.data$individual)[1]]
cpgs_mean <- heat.plot.data$CpG_mean[heat.plot.data$individual == unique(heat.plot.data$individual)[1]]

cpgs <- cpgs[order(cpgs_mean, decreasing = TRUE)]
heat.plot.data$CpG <- factor(heat.plot.data$CpG, levels = c("pop", unique(as.character(cpgs))))
heat.plot.data <- heat.plot.data[order(heat.plot.data$CpG),]

heat.plot <- ggplot2::ggplot(data = heat.plot.data, ggplot2::aes(x = individual)) + 
    ggplot2::geom_tile(data = heat.plot.data, 
                       ggplot2::aes(x = individual, y = CpG, fill = Methylation_percentage)) +
    ggplot2::scale_fill_gradient2(low = "blue", mid = "yellow", high = "red", 
                                  midpoint = 50, guide = "colourbar") +
    ggnewscale::new_scale_fill() +
    ggplot2::geom_tile(data = heat.plot.data, ggplot2::aes(x = individual, y = "pop", fill = pop,
                                    height = length(unique(heat.plot.data$CpG))*0.02)) +
    ggplot2::scale_fill_manual(values = colours3f) +
    ggplot2::scale_y_discrete(drop = FALSE) +
    ggplot2::theme(legend.position = "top",
                   axis.title.y = ggplot2::element_blank(),
                   axis.title.x = ggplot2::element_blank(),
                   axis.text.x = ggplot2::element_text(face = "bold", color = "black", 
                                        size = 10, angle = 90, hjust = 1, vjust = 0.5))
print(heat.plot)

ggplot2::ggsave(heat.plot, bg = "white",
                filename = paste0(figure.path,
                                  "Toothfish_ALL_fin_methDiff_25_HEATMAP.png"), 
                width = 30, height = 30, units = "cm")
```

### Plot CpG difference

```{r}
load(file = paste0(rdata.path,"Toothfish_ALL_fin_methylation_DMS.Rdata"))

myDiff25p.gr <- as(myDiff25p, "GRanges")
meth.diff <- methylKit::selectByOverlap(meth.ALLF.norm10L, myDiff25p.gr) #113 

length(unique(meth.diff$chr)) #34
sort(table(meth.diff$chr), decreasing = TRUE)[1:20]

meth.diff.df <- as.data.frame(t(methylKit::percMethylation(meth.diff)))
loc.names <- paste0(meth.diff$chr,"_",meth.diff$start)
colnames(meth.diff.df) <- loc.names

meth.diff.df$individual <- rownames(meth.diff.df) 
identical(metadata.new$Genotype_names, meth.diff.df$individual)
meth.diff.df$pop <- metadata.new$pop



meth.diff.df.plot <- tidyr::gather(meth.diff.df, key = locus, 
                           value = perc_methylation, -individual, -pop,
                           factor_key = TRUE) %>%
  dplyr::mutate(chr = stringr::str_split_fixed(locus, 
                                               pattern = "_", 
                                               n = 2)[,1])
meth.diff.df.plot.locus <- meth.diff.df.plot %>% dplyr::group_by(chr,locus,pop) %>% 
  dplyr::summarise(mean_pop = mean(perc_methylation, na.rm = TRUE),
                   sum_NA = sum(is.na(perc_methylation)),
            .groups = 'drop')


diff.plot <- ggplot2::ggplot() +
  ggplot2::geom_line(data = meth.diff.df.plot.locus,
                     ggplot2::aes(x = pop, y = mean_pop, colour = locus, group = locus)) +
  ggplot2::geom_point(data = meth.diff.df.plot.locus,
                     ggplot2::aes(x = pop, y = mean_pop, colour = locus, group = locus),
                     size = 5) +
  ggplot2::facet_wrap(~chr) +
  ggplot2::ylab("Mean methylation (%)") +
  ggplot2::theme(legend.position = "none",
                   axis.text.x = ggplot2::element_text(face = "bold", color = "black", 
                                        size = 10, angle = 90, hjust = 1, vjust = 0.5))

print(diff.plot)

ggplot2::ggsave(plot = diff.plot, 
                filename = paste0(figure.path,"1_Toothfish_ALL_fin_methDiff.png"),
                width = 30, height = 15, units = "cm", device = "png")
```


### DMS Clustering
```{r}
load(file = paste0(rdata.path,"Toothfish_ALL_fin_methylation_DMS.Rdata"))

myDiff25p.gr <- as(myDiff25p, "GRanges")
meth.diff <- methylKit::selectByOverlap(meth.ALLF.norm10L, myDiff25p.gr) #112  
methdif.perc <- methylKit::percMethylation(meth.diff)

methdif.perc2 <- zoo::na.aggregate(methdif.perc, by = meth.ALLF.norm10L@treatment)
PCobj <- prcomp(methdif.perc2, retx = TRUE, scale. = TRUE)

var_frac <- PCobj$sdev/sum(PCobj$sdev)
print(paste0("Total variance explained: ",signif(sum(var_frac[1:3]) * 100, 3), "%")) 
barplot(var_frac[1:3], main = "variance explained", 
        col = heat.colors(length(var_frac[1:3])))


identical(metadata.new$Genotype_names, meth.ALLF.norm10L@sample.ids)
identical(metadata.new$Genotype_names, rownames(PCobj$rotation))

pop <- metadata.new$pop
pool <- metadata.new$STRATA4

plot.title <- "Toothfish methylation - ALL fin (2975 DMS sites - 25%)"

data <- data.frame(ind = rownames(PCobj$rotation),PCobj$rotation,  
                   pop = pop, pool = pool)
data <- merge(data, aggregate(cbind(mean.x = PC1, mean.y = PC2) ~
                                pop, data, mean), by = "pop")

pca.plot <- ggplot2::ggplot(data) +
  ggplot2::geom_point(ggplot2::aes(PC1, PC2, color = pop, shape = pool),
                                   size = 4) + 
  ggplot2::geom_point(ggplot2::aes(x = mean.x,
                                   y = mean.y, color = pop),
                      size = 7,
                      shape = 17) +
  ggplot2::scale_shape_manual(values = c(19,18)) +
  ggplot2::scale_colour_manual(values = colours3f) +
  ggplot2::geom_hline(yintercept = 0) +
  ggplot2::geom_vline(xintercept = 0) + 
  ggplot2::labs(
    subtitle = paste0("Total variance explained: ",
                      signif(sum(var_frac[1:3]) * 100, 3), "%"),
    y = paste0("PC2; ", signif(var_frac[2] * 100, digits = 2), " %"),
    x = paste0("PC1; ", signif(var_frac[1] * 100, digits = 2), " %"),
    title = plot.title,
    caption = "") + 
  ggplot2::theme_minimal() +
  ggplot2::theme(
    axis.text = ggplot2::element_text(size = 15),
    axis.title.x = ggplot2::element_text(size = 20),
    axis.title.y = ggplot2::element_text(size = 20),
    legend.text = ggplot2::element_text(size = 15)
  )

print(pca.plot)
ggplot2::ggsave(pca.plot, bg = "white",
                filename = paste0(figure.path,"Toothfish_ALL_fin_PCA_CpG_DMS_PC12.png"), 
                width = 30, height = 15, units = "cm")



data <- data.frame(ind = rownames(PCobj$rotation),PCobj$rotation,  
                   pop = pop, pool = pool)
data <- merge(data, aggregate(cbind(mean.x = PC1, mean.y = PC3) ~
                                pop, data, mean), by = "pop")

pca.plot <- ggplot2::ggplot(data) +
  ggplot2::geom_point(ggplot2::aes(PC1, PC3, color = pop, shape = pool),
                      size = 4) + 
  ggplot2::geom_point(ggplot2::aes(x = mean.x,
                                   y = mean.y, color = pop),
                      size = 7,
                      shape = 17) +
  ggplot2::scale_shape_manual(values = c(19,18)) +
  ggplot2::scale_colour_manual(values = colours3f) +
  ggplot2::geom_hline(yintercept = 0) +
  ggplot2::geom_vline(xintercept = 0) + 
  ggplot2::labs(
    subtitle = paste0("Total variance explained: ",
                      signif(sum(var_frac[1:3]) * 100, 3), "%"),
    y = paste0("PC3; ", signif(var_frac[3] * 100, digits = 2), " %"),
    x = paste0("PC1; ", signif(var_frac[1] * 100, digits = 2), " %"),
    title = plot.title,
    caption = "") + 
  ggplot2::theme_minimal() +
  ggplot2::theme(
    axis.text = ggplot2::element_text(size = 15),
    axis.title.x = ggplot2::element_text(size = 20),
    axis.title.y = ggplot2::element_text(size = 20),
    legend.text = ggplot2::element_text(size = 15)
  )

print(pca.plot)
ggplot2::ggsave(pca.plot, bg = "white",
                filename = paste0(figure.path,
                                  "Toothfish_ALL_fin_PCA_CpG_DMS_PC13.png"), 
                width = 30, height = 15, units = "cm")




data <- data.frame(ind = rownames(PCobj$rotation),PCobj$rotation,  pop = pop, pool = pool)
data <- merge(data, aggregate(cbind(mean.x = PC2, mean.y = PC3) ~
                                pop, data, mean), by = "pop")

pca.plot <- ggplot2::ggplot(data) +
  ggplot2::geom_point(ggplot2::aes(PC2, PC3, color = pop, shape = pool),
                      size = 4) + 
  ggplot2::geom_point(ggplot2::aes(x = mean.x,
                                   y = mean.y, color = pop),
                      size = 7,
                      shape = 17) +
  ggplot2::scale_shape_manual(values = c(19,18)) +
  ggplot2::scale_colour_manual(values = colours3f) +
  ggplot2::geom_hline(yintercept = 0) +
  ggplot2::geom_vline(xintercept = 0) + 
  ggplot2::labs(
    subtitle = paste0("Total variance explained: ",
                      signif(sum(var_frac[1:3]) * 100, 3), "%"),
    y = paste0("PC3; ", signif(var_frac[3] * 100, digits = 2), " %"),
    x = paste0("PC2; ", signif(var_frac[2] * 100, digits = 2), " %"),
    title = plot.title,
    caption = "") + 
  ggplot2::theme_minimal() +
  ggplot2::theme(
    axis.text = ggplot2::element_text(size = 15),
    axis.title.x = ggplot2::element_text(size = 20),
    axis.title.y = ggplot2::element_text(size = 20),
    legend.text = ggplot2::element_text(size = 15)
  )

print(pca.plot)
ggplot2::ggsave(pca.plot, bg = "white",
                filename = paste0(figure.path,"Toothfish_ALL_fin_PCA_CpG_DMS_PC23.png"), 
                width = 30, height = 15, units = "cm")
```



## DMS RANDOM - ALL - fin

```{r,eval=FALSE}
print(load(paste0(rdata.path,"Toothfish_ALL_fin_methylkit_object.Rdata")))

print(load(paste0(rdata.path,"Toothfish_metadata.Rdata")))
metadata.new <- metadata[metadata$Genotype_names %in% attributes(meth.ALLF.norm10L)$sample.ids,]
attributes(meth.ALLF.norm10L)$sample.ids == metadata.new$Genotype_names

# strata.random <- sample(metadata.new$pop, size = length(metadata.new$pop), replace = FALSE)
strata.random <- sample(metadata.new$STRATA2, size = length(metadata.new$STRATA2), replace = FALSE)

meth.ALLF.norm10L@treatment <- as.integer(factor(strata.random))

# covariates <- dplyr::select(metadata.new, Sex, TL_mm)
myDiff <- methylKit::calculateDiffMeth(meth.ALLF.norm10L,
                                       covariates = NULL,
                                       adjust = "fdr",
                                       chunk.size = 1e+06,
                                       mc.cores = 25)
meth.diff.chr <- methylKit::diffMethPerChr(myDiff,plot = TRUE, 
                                           qvalue.cutoff = 0.01, meth.cutoff = 40)
myDiff40p <- methylKit::getMethylDiff(myDiff, difference = 40, qvalue = 0.01)
#12
myDiff25p <- methylKit::getMethylDiff(myDiff, difference = 25, qvalue = 0.01)
#325
myDiff15p <- methylKit::getMethylDiff(myDiff, difference = 15, qvalue = 0.01)
#2739    

save(strata.random, myDiff,myDiff15p, myDiff25p, myDiff40p,
     file = paste0(rdata.path,"Toothfish_ALL_fin_methylation_RANDOM_DMS.Rdata"))
```


```{r}
print(load(paste0(rdata.path,"Toothfish_ALL_fin_methylation_RANDOM_DMS.Rdata")))

volcano.plot <- ggplot2::ggplot() +
  ggplot2::geom_point(data = myDiff, ggplot2::aes(x = meth.diff, 
                                                  y = -log10(qvalue))) +
  ggplot2::geom_point(data = myDiff25p, ggplot2::aes(x = meth.diff, 
                                                  y = -log10(qvalue)),
                      colour = "red") +
  
  ggplot2::theme_minimal() +
  ggplot2::theme(
    axis.text = ggplot2::element_text(size = 15),
    axis.title.x = ggplot2::element_text(size = 20),
    axis.title.y = ggplot2::element_text(size = 20),
    legend.text = ggplot2::element_text(size = 15)
  )
print(volcano.plot)

ggplot2::ggsave(volcano.plot, bg = "white",
                filename =
                  paste0(figure.path,
                         "Toothfish_ALL_fin_RANDOM_DMS25_VOLCANO_methylkit.png"), 
                width = 30, height = 15, units = "cm")
```


### Heatmap

```{r}
print(load(paste0(rdata.path,"Toothfish_ALL_fin_methylation_RANDOM_DMS.Rdata")))

myDiff25p.gr <- as(myDiff25p, "GRanges")
meth.diff <- methylKit::selectByOverlap(meth.ALLF.norm10L, myDiff25p.gr)

meth.diff.df <- as.data.frame(t(methylKit::percMethylation(meth.diff)))
loc.names <- paste0(meth.diff$chr,"_",meth.diff$start)
colnames(meth.diff.df) <- loc.names

meth.diff.df$individual <- rownames(meth.diff.df)
meth.diff.df$pop <- strata.random
meth.diff.df$pop2 <- as.character(metadata.new$pop)


heat.plot.data <- meth.diff.df %>%
  tibble::as_tibble() %>%
  tidyr::gather(key = "CpG", value = "Methylation_percentage", -individual, -pop, -pop2,
                factor_key = FALSE) %>%
  dplyr::group_by(pop, CpG) %>%
  dplyr::mutate(CpG_mean = mean(Methylation_percentage, na.rm = TRUE)) %>%
  dplyr::arrange(pop, individual, CpG) %>%
  dplyr::mutate(individual = factor(individual, levels = unique(individual))) %>%
  dplyr::mutate(CpG = factor(CpG, levels = unique(CpG)))
cpgs <- heat.plot.data$CpG[heat.plot.data$individual == unique(heat.plot.data$individual)[1]]
cpgs_mean <- heat.plot.data$CpG_mean[heat.plot.data$individual == unique(heat.plot.data$individual)[1]]

cpgs <- cpgs[order(cpgs_mean, decreasing = TRUE)]
heat.plot.data$CpG <- factor(heat.plot.data$CpG, levels = c("pop", unique(as.character(cpgs))))
heat.plot.data <- heat.plot.data[order(heat.plot.data$CpG),]

heat.plot <- ggplot2::ggplot(data = heat.plot.data, ggplot2::aes(x = individual)) + 
    ggplot2::geom_tile(data = heat.plot.data, 
                       ggplot2::aes(x = individual, y = CpG, fill = Methylation_percentage)) +
    ggplot2::scale_fill_gradient2(low = "blue", mid = "yellow", high = "red", 
                                  midpoint = 50, guide = "colourbar") +
    ggnewscale::new_scale_fill() +
    ggplot2::geom_tile(data = heat.plot.data, ggplot2::aes(x = individual, 
                                                           y = "pop - random", fill = pop,
                                    height = length(unique(heat.plot.data$CpG))*0.02)) +
    # ggplot2::scale_fill_manual(values = colours3f) +
    ggplot2::scale_fill_manual(values = c("#E1AF00","#adcf9f", "#08306B")) +
    ggnewscale::new_scale_fill() +
    ggplot2::geom_tile(data = heat.plot.data, ggplot2::aes(x = individual, y = "pop", fill = pop2,
                                    height = length(unique(heat.plot.data$CpG))*0.02)) +
    ggplot2::scale_fill_manual(values = colours3f) +
    ggplot2::scale_y_discrete(drop = FALSE) +
    ggplot2::theme(legend.position = "top",
                   axis.title.y = ggplot2::element_blank(),
                   axis.title.x = ggplot2::element_blank(),
                   axis.text.x = ggplot2::element_text(face = "bold", color = "black", 
                                        size = 10, angle = 90, hjust = 1, vjust = 0.5))
print(heat.plot)

ggplot2::ggsave(heat.plot, bg = "white",
                filename = paste0(figure.path,
                                  "Toothfish_ALL_fin_RANDOM_methDiff_25_HEATMAP.png"), 
                width = 30, height = 30, units = "cm")
```

### Plot CpG difference

```{r}
load(file = paste0(rdata.path,"Toothfish_ALL_fin_methylation_RANDOM_DMS.Rdata"))

myDiff25p.gr <- as(myDiff25p, "GRanges")
meth.diff <- methylKit::selectByOverlap(meth.ALLF.norm10L, myDiff25p.gr) #113 

length(unique(meth.diff$chr)) #34
sort(table(meth.diff$chr), decreasing = TRUE)[1:20]

meth.diff.df <- as.data.frame(t(methylKit::percMethylation(meth.diff)))
loc.names <- paste0(meth.diff$chr,"_",meth.diff$start)
colnames(meth.diff.df) <- loc.names

meth.diff.df$individual <- rownames(meth.diff.df) 
identical(metadata.new$Genotype_names, meth.diff.df$individual)
meth.diff.df$pop <- metadata.new$pop


meth.diff.df.plot <- tidyr::gather(meth.diff.df, key = locus, 
                           value = perc_methylation, -individual, -pop,
                           factor_key = TRUE) %>%
  dplyr::mutate(chr = stringr::str_split_fixed(locus, 
                                               pattern = "_", 
                                               n = 2)[,1])
meth.diff.df.plot.locus <- meth.diff.df.plot %>% dplyr::group_by(chr,locus,pop) %>% 
  dplyr::summarise(mean_pop = mean(perc_methylation, na.rm = TRUE),
                   sum_NA = sum(is.na(perc_methylation)),
            .groups = 'drop')


diff.plot <- ggplot2::ggplot() +
  ggplot2::geom_line(data = meth.diff.df.plot.locus,
                     ggplot2::aes(x = pop, y = mean_pop, colour = locus, group = locus)) +
  ggplot2::geom_point(data = meth.diff.df.plot.locus,
                     ggplot2::aes(x = pop, y = mean_pop, colour = locus, group = locus),
                     size = 5) +
  ggplot2::facet_wrap(~chr) +
  ggplot2::ylab("Mean methylation (%)") +
  ggplot2::theme(legend.position = "none",
                   axis.text.x = ggplot2::element_text(face = "bold", color = "black", 
                                        size = 10, angle = 90, hjust = 1, vjust = 0.5))
    # ggplot2::guides(colour = ggplot2::guide_legend(ncol = 2))

print(diff.plot)

ggplot2::ggsave(plot = diff.plot, 
                filename = paste0(figure.path,"1_Toothfish_ALL_fin_RANDOM_methDiff.png"),
                width = 30, height = 15, units = "cm", device = "png")
```


### DMS Clustering
```{r}
load(file = paste0(rdata.path,"Toothfish_ALL_fin_methylation_RANDOM_DMS.Rdata"))

myDiff25p.gr <- as(myDiff25p, "GRanges")
meth.diff <- methylKit::selectByOverlap(meth.ALLF.norm10L, myDiff25p.gr) #270  
methdif.perc <- methylKit::percMethylation(meth.diff)

methdif.perc2 <- zoo::na.aggregate(methdif.perc, by = meth.ALLF.norm10L@treatment)
PCobj <- prcomp(methdif.perc2, retx = TRUE, scale. = TRUE) # Too many missing values

var_frac <- PCobj$sdev/sum(PCobj$sdev)
print(paste0("Total variance explained: ",signif(sum(var_frac[1:3]) * 100, 3), "%")) 
barplot(var_frac[1:3], main = "variance explained", 
        col = heat.colors(length(var_frac[1:3])))


identical(metadata.new$Genotype_names, meth.ALLF.norm10L@sample.ids)

identical(as.numeric(factor(metadata.new$pop)), meth.ALLF.norm10L@treatment)

# identical(metadata.new$individual, rownames(PCobj$x))
identical(metadata.new$Genotype_names, rownames(PCobj$rotation))
pop <- metadata.new$pop
pool <- metadata.new$STRATA4

plot.title <- "Toothfish methylation - RANDOMISED ALL fin (325 DMS sites - 25%)"


data <- data.frame(ind = rownames(PCobj$rotation),PCobj$rotation,  
                   pop = pop, pool = pool)
data <- merge(data, aggregate(cbind(mean.x = PC1, mean.y = PC2) ~
                                pop, data, mean), by = "pop")

pca.plot <- ggplot2::ggplot(data) +
  ggplot2::geom_point(ggplot2::aes(PC1, PC2, color = pop, shape = pool),
                      size = 4) + 
  ggplot2::geom_point(ggplot2::aes(x = mean.x,
                                   y = mean.y, color = pop),
                      size = 7,
                      shape = 17) +
  ggplot2::scale_shape_manual(values = c(19,18)) +
  ggplot2::scale_colour_manual(values = colours3f) +
  ggplot2::geom_hline(yintercept = 0) +
  ggplot2::geom_vline(xintercept = 0) + 
  ggplot2::labs(
    subtitle = paste0("Total variance explained: ",
                      signif(sum(var_frac[1:3]) * 100, 3), "%"),
    y = paste0("PC2; ", signif(var_frac[2] * 100, digits = 2), " %"),
    x = paste0("PC1; ", signif(var_frac[1] * 100, digits = 2), " %"),
    title = plot.title,
    caption = "") + 
  ggplot2::theme_minimal() +
  ggplot2::theme(
    axis.text = ggplot2::element_text(size = 15),
    axis.title.x = ggplot2::element_text(size = 20),
    axis.title.y = ggplot2::element_text(size = 20),
    legend.text = ggplot2::element_text(size = 15)
  )

print(pca.plot)
ggplot2::ggsave(pca.plot, bg = "white",
                filename = 
                  paste0(figure.path,"Toothfish_ALL_fin_RANDOM_PCA_CpG_DMS_PC12.png"), 
                width = 30, height = 15, units = "cm")



data <- data.frame(ind = rownames(PCobj$rotation),PCobj$rotation, 
                   pop = pop, pool = pool)
data <- merge(data, aggregate(cbind(mean.x = PC1, mean.y = PC3) ~
                                pop, data, mean), by = "pop")

pca.plot <- ggplot2::ggplot(data, ggplot2::aes(PC1, PC3, color = pop, shape = pool)) +
  ggplot2::geom_point(size = 4) + 
  ggplot2::geom_point(ggplot2::aes(x = mean.x,
                                   y = mean.y, color = pop),
                      size = 7,
                      shape = 17) +
  ggplot2::scale_shape_manual(values = c(19,18)) +
  ggplot2::scale_colour_manual(values = colours3f) +
  ggplot2::geom_hline(yintercept = 0) +
  ggplot2::geom_vline(xintercept = 0) + 
  ggplot2::labs(
    subtitle = paste0("Total variance explained: ",
                      signif(sum(var_frac[1:3]) * 100, 3), "%"),
    y = paste0("PC3; ", signif(var_frac[3] * 100, digits = 2), " %"),
    x = paste0("PC1; ", signif(var_frac[1] * 100, digits = 2), " %"),
    title = plot.title,
    caption = "") + 
  ggplot2::theme_minimal() +
  ggplot2::theme(
    axis.text = ggplot2::element_text(size = 15),
    axis.title.x = ggplot2::element_text(size = 20),
    axis.title.y = ggplot2::element_text(size = 20),
    legend.text = ggplot2::element_text(size = 15)
  )

print(pca.plot)
ggplot2::ggsave(pca.plot, bg = "white",
                filename = paste0(figure.path,"Toothfish_ALL_fin_RANDOM_PCA_CpG_DMS_PC13.png"), 
                width = 30, height = 15, units = "cm")




data <- data.frame(ind = rownames(PCobj$rotation),PCobj$rotation, 
                   pop = pop, shape = pool)
data <- merge(data, aggregate(cbind(mean.x = PC2, mean.y = PC3) ~
                                pop, data, mean), by = "pop")

pca.plot <- ggplot2::ggplot(data) +
  ggplot2::geom_point(ggplot2::aes(PC2, PC3, color = pop, shape = pool), 
                      size = 4) + 
  ggplot2::geom_point(ggplot2::aes(x = mean.x,
                                   y = mean.y, color = pop),
                      size = 7,
                      shape = 17) +
  ggplot2::scale_shape_manual(values = c(19,18)) +
  ggplot2::scale_colour_manual(values = colours3f) +
  ggplot2::geom_hline(yintercept = 0) +
  ggplot2::geom_vline(xintercept = 0) + 
  ggplot2::labs(
    subtitle = paste0("Total variance explained: ",
                      signif(sum(var_frac[1:3]) * 100, 3), "%"),
    y = paste0("PC3; ", signif(var_frac[3] * 100, digits = 2), " %"),
    x = paste0("PC2; ", signif(var_frac[2] * 100, digits = 2), " %"),
    title = plot.title,
    caption = "") + 
  ggplot2::theme_minimal() +
  ggplot2::theme(
    axis.text = ggplot2::element_text(size = 15),
    axis.title.x = ggplot2::element_text(size = 20),
    axis.title.y = ggplot2::element_text(size = 20),
    legend.text = ggplot2::element_text(size = 15)
  )

print(pca.plot)
ggplot2::ggsave(pca.plot, bg = "white",
                filename = 
                  paste0(figure.path,"Toothfish_ALL_fin_RANDOM_PCA_CpG_DMS_PC23.png"), 
                width = 30, height = 15, units = "cm")
```









# 11_Methylation: 2024 - pool - muscle tissue

```{r}
print(load(paste0(rdata.path,"Toothfish_metadata.Rdata"))) #ALL fish
```

## Reading the methylation call files

```{r, eval=FALSE}
data.list <- list.files(
  path = 
  "/datasets/work/ncmi-toa-rrbs/work/1.Toothfish/GENOME/Combined_methylation/",
  pattern = "*.cov.gz")

subnames <- metadata$File_names[!(metadata$Tissue_type == "fin" |
                                    grepl(pattern = "TOA", x = metadata$File_names)
                                  )]
data.list <- data.list[data.list %in% paste0(subnames,
                                             "_R1_trimmed_sort_bismark_bt2.bismark.cov.gz")]
data.loc <- as.list(paste0(
  "/datasets/work/ncmi-toa-rrbs/work/1.Toothfish/GENOME/Combined_methylation/",
                           data.list))



data.list2 <-  gsub(pattern = '_R1_trimmed_sort_bismark_bt2.bismark.cov.gz',
                    replacement = '',x = as.character(data.list))


metadata.new <- metadata[metadata$File_names %in% data.list2,]
metadata.new <- metadata.new[order(match(metadata.new$File_names, data.list2)),]

metadata.new$File_names == data.list2

sample.ids <- metadata.new$Genotype_names
pop <- as.integer(factor(metadata.new$pop))

myobjDB <- methylKit::methRead(location = data.loc,
           sample.id = as.list(sample.ids),
           assembly = "TOA_genome",
           treatment = pop,
           context = "CpG",
           header = FALSE, 
           dbtype = "tabix",
           pipeline = "bismarkCoverage",
           mincov = 3
           )
save(myobjDB, file = paste0(rdata.path,"Toothfish_2024_pool_muscle_methylkit_cpgDB.Rdata"))
```


## Data QC

#### Number of sites in genome

```{r, eval=FALSE}
cov.df <- data.frame(chr = NA, start = NA,end = NA, strand = NA, coverage = NA, 
                     numCs = NA, numTs = NA, sample = NA)
for (i in 1:length(sample.ids)) {
  df <- methylKit::getData(myobjDB[[i]])
  df$sample <- sample.ids[i]
  cov.df <- rbind(cov.df, df)
}
cov.df <- cov.df[-1,]
save(cov.df, file = paste0(rdata.path,"Toothfish_2024_pool_muscle_methylkit_cpg_cov.Rdata"))
```

### Coverage stats

```{r}
load(paste0(rdata.path,"Toothfish_2024_pool_muscle_methylkit_cpgDB.Rdata"))
print(load(paste0(rdata.path,"Toothfish_2024_pool_muscle_methylkit_cpg_cov.Rdata")))

length(unique(paste0(cov.df$chr, "_", cov.df$start))) #1,894,182

# summary(as.factor(cov.df$sample))
summary(cov.df$coverage)
   # Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
   # 4.00    6.00   10.00   12.73   17.00 2601.00 

summary(cov.df$coverage[cov.df$coverage > 5 & cov.df$coverage < 50])
   # Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
   # 6.00    8.00   13.00   14.26   18.00   49.00

cov.plot <- ggplot2::ggplot(cov.df, ggplot2::aes(x = log(coverage), fill = sample)) +
  ggplot2::geom_histogram() +
  ggplot2::theme(legend.position = "none")
# print(cov.plot)
ggplot2::ggsave(cov.plot,
                filename =
                  paste0(figure.path,"Toothfish_2024_pool_muscle_coverage_hist_Combined.png"), 
                width = 45, height = 15, units = 'cm')

cov.plot <- ggplot2::ggplot(cov.df, ggplot2::aes(x = coverage, fill = sample)) +
  ggplot2::geom_histogram() +
  ggplot2::theme(legend.position = "none")
# print(cov.plot)
ggplot2::ggsave(cov.plot,
                filename =
                  paste0(figure.path,"Toothfish_2024_pool_muscle_coverage_hist_Combined2.png"), 
                width = 45, height = 15, units = 'cm')

cov.plot <- ggplot2::ggplot(cov.df[cov.df$coverage > 5 & cov.df$coverage < 160,], 
                            ggplot2::aes(x = coverage)) +
  ggplot2::geom_histogram(binwidth = 1) +
  ggplot2::ggtitle("coverage < 160") +
  ggplot2::theme(legend.position = "none")
# print(cov.plot)
ggplot2::ggsave(cov.plot,
                filename =
                  paste0(figure.path,"Toothfish_2024_pool_muscle_coverage_hist_Combined4.png"),
                width = 45, height = 15, units = 'cm')



max(cov.df$coverage, na.rm = T) #11562
ind.covplot <- ggplot2::ggplot(cov.df, ggplot2::aes(x = sample, y = coverage)) + 
  ggplot2::geom_boxplot() +
  ggplot2::theme_bw() + 
  ggplot2::theme(axis.title.x = ggplot2::element_blank(), 
                 axis.text.y = ggplot2::element_text(size = 15),
                 axis.title.y = ggplot2::element_text(size = 15),
                 axis.text.x = ggplot2::element_text(size = 7, angle = 90, hjust = 1)) +
  ggplot2::scale_y_continuous(trans = scales::log2_trans(),
                  breaks = c(1, 10, 50, 750, 20000),
                  minor_breaks = c(1:10, 2:5*10, 1:7.5*100, 0.5:2*1000))

print(ind.covplot)

ggplot2::ggsave(ind.covplot, bg = "white",
                filename = 
                  paste0(figure.path,
                         "Toothfish_2024_pool_muscle_raw_individual_coverage_CpGs.png"),
                width = 45, height = 15, units = 'cm')
```


### Coverage filter

High count filter based on coverage stats.

```{r, eval=FALSE}
load(paste0(rdata.path,"Toothfish_2024_pool_muscle_methylkit_cpgDB.Rdata"))

filtered.myobj <- methylKit::filterByCoverage(myobjDB, lo.count = 10, lo.perc = NULL,
                                              save.db = FALSE,
                                      hi.count = NULL, hi.perc = 99)
save(myobjDB,filtered.myobj,
     file = paste0(rdata.path,"Toothfish_2024_pool_muscle_methylkit_cpgDB.Rdata"))
```

### Normalise coverage

Normalize read coverage distributions between samples. Requires stringent filter on max coverage first, otherwise it will over-sample PCR bias reads. These two functions will help reduce the bias in the statistical tests that might occur due to systematic over-sampling of reads in certain samples.

```{r}
filtered.myobj.normal <- methylKit::normalizeCoverage(obj = filtered.myobj, 
                                                      save.db = TRUE,
                                                      method = "median", chunk.size = 1e6)
save(myobjDB,filtered.myobj, filtered.myobj.normal,
     file = paste0(rdata.path,"Toothfish_2024_pool_muscle_methylkit_cpgDB.Rdata"))
```

### Number of sites per sample

```{r, eval = FALSE}
sites.df <- data.frame(sites = NA,  sample = NA)
for (i in 1:length(filtered.myobj.normal)) {
  print(i)
  sites.df[i,] <- c(filtered.myobj.normal[[i]]@num.records, filtered.myobj.normal[[i]]@sample.id)
}
sites.df$sites <- as.integer(sites.df$sites)
# sites.df$pop <- filtered.myobj.normal@treatment
sites.df$pop <- metadata.new$pop
save(sites.df, file = paste0(rdata.path,"Toothfish_2024_pool_muscle_number_cpg_sample.Rdata"))
```


```{r}
print(load(paste0(rdata.path,"Toothfish_2024_pool_muscle_number_cpg_sample.Rdata")))
sites.plot <- ggplot2::ggplot(sites.df, ggplot2::aes(x = sample, y = sites, colour = pop)) +
  ggplot2::geom_point() +
  ggplot2::scale_color_manual(values = colours2) +
  ggplot2::theme_bw() +
  ggplot2::theme(axis.text.x = ggplot2::element_text(angle = 90))
print(sites.plot)

ggplot2::ggsave(sites.plot,
                filename =
                  paste0(figure.path,"Toothfish_2024_pool_muscle_number_cpg_sample.png"),
                width = 45, height = 15, units = 'cm')

sites.df$sample[sites.df$sites < 720000]

# "88-2_Muscle_Pool1"
```


## Select samples

```{r}
print(load(paste0(rdata.path,"Toothfish_metadata.Rdata")))

metadata2 <- metadata[!(metadata$Tissue_type == "fin" |
                                    grepl(pattern = "TOA", x = metadata$File_names)
                                  ),]

print(load(paste0(rdata.path, "Low_quality_fish.Rdata")))

bad.list <- list(high.lambda.mapping,
          low.genome.mapping,low.genome.methylation,
          low.mbias.methylation)

rm <- Reduce(intersect, list(high.lambda.mapping,
          low.genome.mapping,low.genome.methylation,
          low.mbias.methylation))
print(rm)

# rm <- c(rm,) # bad coverage

metadata2 <- metadata2[!metadata2$File_names %in% rm,]

filt.sample.id.full <- sapply(filtered.myobj@.Data, function(x){x@sample.id})

# filt.sample.id <- filt.sample.id.full[filt.sample.id.full %in% metadata2$File_names]
# metadata2 <- metadata2[order(match(metadata.new$File_names, filt.sample.id)),]

filt.sample.id <- filt.sample.id.full[filt.sample.id.full %in% metadata2$Genotype_names]
metadata2 <- metadata2[order(match(metadata2$Genotype_names, filt.sample.id)),]
identical(filt.sample.id,metadata2$Genotype_names)


pop <- metadata2$pop
pop.levels <- c("58.4.2-2024",
                "88.2-2024")

pop <- factor(pop, levels = pop.levels)
pop <- as.numeric(pop)
sample.ids <- metadata2$Genotype_names



# filtered.myobj2 <- methylKit::reorganize(
#   methylObj = filtered.myobj,
#   sample.ids = sample.ids,
#   treatment = pop,
#   chunk.size = 1e+06,
#   save.db = TRUE,
#   suffix = "TOA_2024_pool_muscle"
# )

filtered.myobj2.normal <- methylKit::reorganize(
  methylObj = filtered.myobj.normal,
  sample.ids = sample.ids,
  treatment = pop,
  chunk.size = 1e+06,
  save.db = TRUE,
  suffix = "TOA_2024_pool_muscle_normal"
)
save(myobjDB,filtered.myobj, 
     filtered.myobj.normal,filtered.myobj2.normal,
     file = paste0(rdata.path,"Toothfish_2024_pool_muscle_methylkit_cpgDB.Rdata"))
```


## Merging samples

```{r}
print(load(paste0(rdata.path,"Toothfish_2024_pool_muscle_methylkit_cpgDB.Rdata")))
pop <- metadata2$pop
pop.levels <- c("58.4.2-2024",
                "88.2-2024")
pop <- factor(pop, levels = pop.levels)
pop <- as.numeric(pop)

table(pop)


# meth.2024M15L <- methylKit::unite(filtered.myobj2, destrand = FALSE,
#                              min.per.group = 15L, #missing data allowed
#                              mc.cores = 25)


meth.2024poolM.norm <- methylKit::unite(filtered.myobj2.normal, destrand = FALSE,
                             min.per.group = NULL, #missing data allowed
                             mc.cores = 25)
#751879 common sites for 1 samples per group
#453761 common sites for no missing data per group

 
meth.2024poolM.norm <- as(meth.2024poolM.norm,"methylBase")

save(meth.2024poolM.norm,
     file = paste0(rdata.path,"Toothfish_2024_pool_muscle_methylkit_object.Rdata"))
```


## Data QC

### Coverage 

```{r}
meth.df <- methylKit::getData(meth.2024poolM.norm)
cov <- meth.df[,meth.2024poolM.norm@coverage.index]
colnames(cov) <- meth.2024poolM.norm@sample.ids
dpf <- data.frame(cov, locus = paste0(meth.df$chr,"_",meth.df$start))
dpf <- tidyr::gather(dpf, key = individual, 
                           value = Depth, -locus,
                           factor_key = TRUE) %>%
  dplyr::group_by(individual) %>%
  dplyr::mutate(NA_cnt = sum(is.na(Depth))/length(unique(locus))) %>%
  dplyr::ungroup()
dpf.NA <- dpf[!duplicated(dpf$individual),]
dpf.NA$NA_cnt <- round(dpf.NA$NA_cnt, 3)

dpf <- dpf[ dpf$Depth > 0 & !is.na(dpf$Depth),]

max(dpf$Depth, na.rm = T) #99
myPlot <- ggplot2::ggplot(dpf, ggplot2::aes(x = individual, y = Depth)) + 
  ggplot2::geom_boxplot() +
  ggplot2::geom_text(data = dpf.NA,ggplot2::aes(x = individual, y = 150, label = NA_cnt), 
                     size = 2, angle = 90) +
  ggplot2::theme_bw() + 
  ggplot2::theme(axis.title.x = ggplot2::element_blank(), 
                 axis.text.y = ggplot2::element_text(size = 15),
                 axis.title.y = ggplot2::element_text(size = 15),
                 axis.text.x = ggplot2::element_text(size = 7, angle = 90, hjust = 1)) +
  ggplot2::scale_y_continuous(trans = scales::log2_trans(),
                  breaks = c(1, 10, 50, 100),
                  minor_breaks = c(1:10, 2:5*10))

print(myPlot)
ggplot2::ggsave(myPlot, bg = "white",
                filename = 
                  paste0(figure.path,
                  "1B.Toothfish_2024_pool_muscle_normalised15_methylkit_individual_coverage_CpGs.png"
                                  ), 
                width = 45, height = 15, units = "cm")

nrow(cov) # 453761 SMPs
plot(x = 1:ncol(cov), y = colMeans(cov, na.rm = TRUE), col = meth.2024poolM.norm@treatment)
plot(x = 1:ncol(cov), y = colMeans(cov > 20, na.rm = TRUE), col = meth.2024poolM.norm@treatment)
plot(x = 1:ncol(cov), y = colMeans(cov > 10, na.rm = TRUE), col = meth.2024poolM.norm@treatment)
meth.2024poolM.norm@sample.ids[colMeans(cov > 10, na.rm = TRUE) < 0.98]
# samples with low coverage: "58-4-2_Muscle_Pool2"
```

### Remove ASM

```{r}
load(paste0(rdata.path,"Allele_specific_mutations.Rdata"))

mut <- as(asmTRUE,"GRanges")
meth.2024poolM.norm <- as(meth.2024poolM.norm,"methylBase") 

sum(as(meth.2024poolM.norm,"GRanges") %over% mut)#12389
meth.asm <- methylKit::selectByOverlap(meth.2024poolM.norm, mut) #411698 
BOOLEAN <- paste0(meth.2024poolM.norm$chr,"_",meth.2024poolM.norm$start) %in%
  paste0(asmTRUE$Chr,"_",asmTRUE$Pos) #1
sum(BOOLEAN) #98
meth.2024poolM.norm <- methylKit::select(meth.2024poolM.norm, !BOOLEAN) # 453663    

save(meth.2024poolM.norm,
     file = paste0(rdata.path,"Toothfish_2024_pool_muscle_methylkit_object.Rdata"))
```



## Load data

```{r}
print(load(paste0(rdata.path,"Toothfish_2024_pool_muscle_methylkit_object.Rdata")))
print(load(paste0(rdata.path,"Toothfish_metadata.Rdata"))) 

metadata.new <- metadata[metadata$Genotype_names %in% meth.2024poolM.norm@sample.ids,]
metadata.new <- metadata.new[order(match(metadata.new$Genotype_names,
                                         meth.2024poolM.norm@sample.ids)),]
```


## Global clustering analysis

```{r}
methylKit::clusterSamples(meth.2024poolM.norm, dist = "correlation", sd.filter = TRUE, 
                          method = "ward.D2", plot = TRUE, )
dev.print(file = paste0(figure.path,"1_TOA_2024_pool_muscle_cluster_plot.png"),
          device = png, res = 300,width = 45,height = 15,units = "cm")
```


### PCA

```{r}
# methdif.perc <- methylKit::percMethylation(meth.2024poolM.norm)
# methdif.perc2 <- zoo::na.aggregate(methdif.perc, by = meth.2024poolM.norm@treatment)
# PCobj <- prcomp(methdif.perc2, retx = TRUE, scale. = TRUE)
PCobj <- methylKit::PCASamples(meth.2024poolM.norm, comp = c(1,2), 
                               adj.lim = c(0,0), obj.return = TRUE)

var_frac <- PCobj$sdev/sum(PCobj$sdev)
print(paste0("Total variance explained: ",signif(sum(var_frac[1:3]) * 100, 3), "%")) 
barplot(var_frac[1:20], main = "variance explained", 
        col = heat.colors(length(var_frac[1:20])))

identical(metadata.new$Genotype_names, meth.2024poolM.norm@sample.ids)
identical(metadata.new$Genotype_names, rownames(PCobj$x))

pop <- metadata.new$pop

plot.title <- "Toothfish methylation - 2024 pool_muscle (453,663  CpG sites)"

data <- data.frame(ind = rownames(PCobj$x),PCobj$x,  pop = pop)
# data <- data.frame(ind = rownames(PCobj$rotation),PCobj$rotation, pop = pop)
data <- merge(data, aggregate(cbind(mean.x = PC1, mean.y = PC2) ~
                                pop, data, mean), by = "pop")

pca.plot <- ggplot2::ggplot(data, ggplot2::aes(PC1, PC2, color = pop)) +
  ggplot2::geom_point(size = 4) + 
  ggplot2::geom_point(ggplot2::aes(x = mean.x,
                                   y = mean.y, color = pop),
                      size = 7,
                      shape = 17) +
  ggplot2::scale_colour_manual(values = colours2) +
  ggplot2::geom_hline(yintercept = 0) +
  ggplot2::geom_vline(xintercept = 0) + 
  ggplot2::labs(
    subtitle = paste0("Total variance explained: ",
                      signif(sum(var_frac[1:4]) * 100, 3), "%"),
    y = paste0("PC2; ", signif(var_frac[2] * 100, digits = 2), " %"),
    x = paste0("PC1; ", signif(var_frac[1] * 100, digits = 2), " %"),
    title = plot.title,
    caption = "") + 
  ggplot2::theme_minimal() +
  ggplot2::theme(
    axis.text = ggplot2::element_text(size = 15),
    axis.title.x = ggplot2::element_text(size = 20),
    axis.title.y = ggplot2::element_text(size = 20),
    legend.text = ggplot2::element_text(size = 15)
  )

print(pca.plot)
ggplot2::ggsave(pca.plot, bg = "white",
                filename = paste0(figure.path,"Toothfish_2024_pool_muscle_PCA_CpG_PC12.png"), 
                width = 30, height = 15, units = "cm")


data <- data.frame(ind = rownames(PCobj$x),PCobj$x,  pop = pop)
# data <- data.frame(ind = rownames(PCobj$rotation),PCobj$rotation, pop = pop)
data <- merge(data, aggregate(cbind(mean.x = PC2, mean.y = PC3) ~
                                pop, data, mean), by = "pop")
pca.plot <- ggplot2::ggplot(data, ggplot2::aes(PC2, PC3, color = pop)) +
  ggplot2::geom_point(size = 4) + 
  ggplot2::geom_point(ggplot2::aes(x = mean.x,
                                   y = mean.y, color = pop),
                      size = 7,
                      shape = 17) +
  ggplot2::scale_colour_manual(values = colours2) +
  ggplot2::geom_hline(yintercept = 0) +
  ggplot2::geom_vline(xintercept = 0) + 
  ggplot2::labs(
    subtitle = paste0("Total variance explained: ",
                      signif(sum(var_frac[1:4]) * 100, 3), "%"),
    y = paste0("PC3; ", signif(var_frac[3] * 100, digits = 2), " %"),
    x = paste0("PC2; ", signif(var_frac[2] * 100, digits = 2), " %"),
    title = plot.title,
    caption = "") + 
  ggplot2::theme_minimal() +
  ggplot2::theme(
    axis.text = ggplot2::element_text(size = 15),
    axis.title.x = ggplot2::element_text(size = 20),
    axis.title.y = ggplot2::element_text(size = 20),
    legend.text = ggplot2::element_text(size = 15)
  )

print(pca.plot)
ggplot2::ggsave(pca.plot, bg = "white",
                filename = paste0(figure.path,"Toothfish_2024_pool_muscle_PCA_CpG_PC23.png"), 
                width = 30, height = 15, units = "cm")
```


### RDA

```{r}
# setwd("/datasets/work/ncmi-toa-rrbs/work/1.Toothfish/Toothfish_RRBS_array_22098078/Bismark/")
# print(load(paste0(rdata.path,"Toothfish_methylation_objects_raw.Rdata"))) #36,635 SMPs
print(load(paste0(rdata.path,"Toothfish_subset_methylation_objects_raw.Rdata"))) #1,316,799 SMPs

# meth.RDA <- methylKit::unite(filtered.myobj2.norm, destrand = FALSE,
#                              save.db = FALSE,
#                              suffix = "subset",
#                              min.per.group = NULL, #missing data allowed
#                              mc.cores = 25)


# meth.df <- methylKit::getData(meth)
# cov <- meth.df[,grep(colnames(meth.df), pattern = "coverage*")]
# cov <- meth.df[,meth@coverage.index]
# colnames(cov) <- meth@sample.ids
# meth.df <- methylKit::percMethylation(meth)

# meth.df <- methylKit::percMethylation(meth.RDA)

# save(meth.df,meth.RDA, 
#      file = paste0(rdata.path,"TOA_data_for_RDA_meth_percent.Rdata"))
print(load(paste0(rdata.path,"TOA_data_for_RDA_meth_percent.Rdata"))) #347,614 SMPs

hist(colSums(is.na(meth.df)), breaks = 100)
hist(rowSums(is.na(meth.df)), breaks = 100)





print(load(paste0(rdata.path,"Toothfish_metadata.Rdata")))

predictor <- metadata2 # depool_musclee the predictor input file here
predictor <- predictor[predictor$SAMPLE_ID %in% meth@sample.ids,]
predictor$SAMPLE_ID = as.character(predictor$SAMPLE_ID)
predictor$region = as.character(predictor$region)
predictor$Sex = as.character(predictor$Sex)
predictor$Tissue = as.character(predictor$Tissue)
predictor$mat = as.character(predictor$mat)
predictor$mat[is.na(predictor$mat)] <- "2"
predictor$TL = as.numeric(predictor$TL)
predictor$wt = as.numeric(predictor$wt)
predictor$dpth = as.numeric(predictor$dpth)
predictor$Age = as.numeric(predictor$Age)
predictor$Age[is.na(predictor$Age)] <- as.numeric(max(predictor$Age, na.rm = TRUE))
predictor$Cohort2 = as.integer(predictor$Cohort2)
predictor$Cohort2[is.na(predictor$Cohort2)] <- as.integer(min(predictor$Cohort2, na.rm = TRUE))


meth.df1 = t(meth.df)
# meth.df2 = na.omit(meth.df1) #rda can't handle NAs, remove them if you have them
# meth.df1[1:10,1:10]
## Predictor
# Make names characters (not factors)


#if you had numbers, they had to be factors (as.factor)

# Confirm that methylation data and predictors are in the same order
identical(rownames(meth.df1), predictor$SAMPLE_ID)

# Perform Hellinger transformation on the methylation data:
meth.df3 = vegan::decostand(meth.df1, method = "hellinger", MARGIN = 1) #MARGIN=1 means rows, on rows I have samples now because I transposed the matrix

## Run RDA
#Test the country using growing condition as a Condition () (covariate):
meth_rda = vegan::rda(meth.df3 ~ region + Condition(Sex), 
                      data = predictor, scale = F) #I remove scale since I performed Hellinger transformation
meth_rda = vegan::rda(meth.df3 ~ region + Condition(Sex) + 
                        Condition(mat)+ Condition(TL) + Condition(wt) +
                        Condition(dpth) + Condition(Age) + Condition(Cohort2), 
                      na.action = "na.omit",
                      data = predictor, scale = F)

meth_rda$pCCA
meth_rda$CCA$wa

summary(meth_rda) #Constrained Proportion: variance of Y explained by X, Unconstrained Proportion: unexplained variance in Y
# sink(paste0(rdata.path,"anova_rda.txt")) #with this you can save the output of the anova in a txt file, depool_musclee the output directory
meth.aov <- anova(meth_rda, test = "Chi", permutations = 99) #run Anova on the result to see the significance level
# meth.aov.cca <- anova.cca(meth_rda, test = "Chi", permutations = 99)
print(meth.aov)
# sink(paste0(rdata.path,"anova_rda_chi.txt"))

# Model: rda(formula = meth.df3 ~ region + Condition(Sex), data = predictor, scale = F)
#           Df  Variance      F Pr(>F)   
# Model      3 0.0003924 1.4126  0.004 **
# Residual 110 0.0101861

# Model: rda(formula = meth.df3 ~ region + Condition(Sex) + Condition(mat) + Condition(TL) + Condition(wt) + Condition(dpth) + Condition(Age) + Condition(Cohort2), data = predictor, scale = F, na.action = "na.omit")
#           Df  Variance     F Pr(>F)   
# Model      3 0.0003412 1.257   0.01 **
# Residual 101 0.0091396 

Pop.names <- predictor$region
plot.a <- vegan::ordiplot(meth_rda, choices=c(1,2))
Pop.ellipses <- vegan::ordiellipse(plot.a, groups=Pop.names, display="sites", kind="sd")
dev.print(file = paste0(figure.path,"TOA_RDA_ordiplot_complex_model.png"),
          device = png, res = 300,width = 45,height = 15,units = "cm")

# Type 1 scaling
vegan::ordiplot(meth_rda, scaling = 1, type = "text")
# Type 2 scaling
vegan::ordiplot(meth_rda, scaling = 2, type = "text")

ggvegan::autoplot(
  meth_rda,
  axes = c(1, 2),
  geom = "text",
  layers = c("species", "sites", "biplot", "centroids"),
  arrows = TRUE)

#CpG contribution
hist(meth_rda$CCA$v[,1])
hist(meth_rda$CCA$v[,2])
hist(meth_rda$CCA$v[,3])



#Test the interaction of your predictors:
meth_rda3 = vegan::rda(meth ~ region * Sex + Condition(region + Sex), data = predictor, scale = F)
summary(meth_rda3)
sink(paste0(rdata.path,"anova_rda_interaction.txt"))
print(anova(meth_rda3, test="Chi"))
sink(paste0(rdata.path,"anova_rda_interaction_chi.txt"))
```

#### plot RDA

```{r}

BioR.theme <- theme(
        panel.background = element_blank(),
        panel.border = element_blank(),
        panel.grid = element_blank(),
        axis.line = element_line("gray25"),
        text = element_text(size = 12),
        axis.text = element_text(size = 10, colour = "gray25"),
        axis.title = element_text(size = 14, colour = "gray25"),
        legend.title = element_text(size = 14),
        legend.text = element_text(size = 14),
        legend.key = element_blank())

plot.a <- vegan::ordiplot(meth_rda, choices=c(1,2))
Pop.names <- predictor$region

# sites1 <- BiodiversityR::sites.long(plot.a, env.data=predictor)
sites1 <- data.frame(cbind(predictor, axis1=plot.a$sites[, 1], 
                           axis2=plot.a$sites[, 2], 
                           labels=rownames(plot.a$sites)),
                     group = predictor$region)

# axis1 <- BiodiversityR::axis.long(meth_rda, choices=c(1, 2))
eigs.all <- meth_rda$CA$eig
eigs <- round(100 * eigs.all / meth_rda$tot.chi, digits=1)[c(1, 2)] 
xlab.label <- paste0(names(eigs)[1], " (", eigs[1], "%)")
ylab.label <- paste0(names(eigs)[2], " (", eigs[2], "%)")
axis1 <- data.frame(axis=c(1:2), 
                    ggplot=c("xlab.label", "ylab.label"), 
                    label=c(xlab.label, ylab.label))


# centroids.Pop1 <- BiodiversityR::centroids.long(sites1, 
#                                                 grouping=Pop.names, 
#                                                 centroids.only=TRUE)
# centroids.Pop2 <- BiodiversityR::centroids.long(sites1, 
#                                                 grouping=Pop.names, 
#                                                 centroids.only=FALSE)
gr.name <- rlang::enexpr(Pop.names)
cent.means <-  stats::aggregate(cbind(axis1, axis2) ~ Pop.names,
                                data = sites1,
                                FUN = mean)
names(cent.means) <- c("group", "axis1c", "axis2c")
centroids.Pop1 <- cent.means
cent.means$Centroid <- cent.means[, 1]
by.name <- names(sites1)[names(sites1) == "group"]
centroids.Pop2 <- dplyr::full_join(sites1, cent.means, by = by.name)


# Pop.ellipses.long2 <- BiodiversityR::ordiellipse.long(Pop.ellipses, 
#                                                     grouping.name="Pop.names")


plotgg.a <- ggplot() + 
    geom_vline(xintercept = c(0), color = "grey70", linetype = 2) +
    geom_hline(yintercept = c(0), color = "grey70", linetype = 2) +  
    xlab(axis1[1, "label"]) +
    ylab(axis1[2, "label"]) +  
    scale_x_continuous(sec.axis = dup_axis(labels=NULL, name=NULL)) +
    scale_y_continuous(sec.axis = dup_axis(labels=NULL, name=NULL)) +
    # geom_polygon(data=Pop.ellipses.long2, 
    #                aes(x=axis1, y=axis2, colour=Pop.names, 
    #                    fill=after_scale(alpha(colour, 0.2))), 
    #           size=0.2, show.legend=TRUE) +
    geom_segment(data=centroids.Pop2, 
                 aes(x=axis1c, y=axis2c, xend=axis1, yend=axis2, colour=Pop.names), 
                 size=0.7, show.legend=FALSE) +
    geom_point(data=centroids.Pop1, 
               aes(x=axis1c, y=axis2c), 
               shape="square", colour="black", alpha=0.7, size=2) +
    geom_label_repel(data=centroids.Pop1, 
               aes(x=axis1c, y=axis2c, label=group, colour=group), 
               alpha=1.0, size=2, show.legend=FALSE) +
    # labs(colour = Pop.names) +
    BioR.theme +
    theme(legend.text = element_text(size = 10)) +
    scale_colour_npg() +
    coord_fixed(ratio=1)

plotgg.a

ggsave(plotgg.a, file = paste0(figure.path, 
                               # "TOA_RDA_simple_model_A.png"
                               "TOA_RDA_complex_model_A.png"
                               ),
       height = 30, width = 30, units = "cm",  device = "png")
```



## DMS - 2024 - pool_muscle

```{r, eval = FALSE}
print(load(paste0(rdata.path,"Toothfish_2024_pool_muscle_methylkit_object.Rdata")))

print(load(paste0(rdata.path,"Toothfish_metadata.Rdata")))
metadata.new <- metadata[metadata$Genotype_names %in% attributes(meth.2024poolM.norm)$sample.ids,]
attributes(meth.2024poolM.norm)$sample.ids == metadata.new$Genotype_names

meth.2024poolM.norm@treatment <- as.integer(factor(metadata.new$pop))

# data.frame(attributes(meth.2024poolM.norm)$sample.ids, metadata.new$individual, metadata.new$strata, meth.2024poolM.norm@treatment)

covariates <- dplyr::select(metadata.new, Sex, Age, Cohort2, Maturity_full)
myDiff <- methylKit::calculateDiffMeth(meth.2024poolM.norm,
                                       covariates = NULL,
                                       adjust = "fdr",
                                       chunk.size = 1e+06,
                                       mc.cores = 24)
meth.diff.chr <- methylKit::diffMethPerChr(myDiff,plot = TRUE, 
                                           qvalue.cutoff = 0.01, meth.cutoff = 40)
myDiff40p <- methylKit::getMethylDiff(myDiff, difference = 40, qvalue = 0.01)
#589 
myDiff25p <- methylKit::getMethylDiff(myDiff, difference = 25, qvalue = 0.01)
#796  
myDiff15p <- methylKit::getMethylDiff(myDiff, difference = 15, qvalue = 0.01)
#810
top100.POOLM <- slice_max(as.data.frame(as(myDiff,"GRanges")), order_by = meth.diff, n = 100)
top100.POOLM <- as(top100.POOLM,"GRanges")

save(myDiff,myDiff15p, myDiff25p, myDiff40p,top100.POOLM,
     file = paste0(rdata.path,"Toothfish_2024_pool_muscle_methylation_DMS.Rdata"))
```


```{r}
print(load(paste0(rdata.path,"Toothfish_2024_pool_muscle_methylation_DMS.Rdata")))

volcano.plot <- ggplot2::ggplot() +
  ggplot2::geom_point(data = myDiff, ggplot2::aes(x = meth.diff, 
                                                  y = -log10(qvalue))) +
  ggplot2::geom_point(data = myDiff25p, ggplot2::aes(x = meth.diff, 
                                                  y = -log10(qvalue)),
                      colour = "red") +
  ggplot2::theme_minimal() +
  ggplot2::theme(
    axis.text = ggplot2::element_text(size = 15),
    axis.title.x = ggplot2::element_text(size = 20),
    axis.title.y = ggplot2::element_text(size = 20),
    legend.text = ggplot2::element_text(size = 15)
  )
print(volcano.plot)

ggplot2::ggsave(volcano.plot, bg = "white",
                filename =
                  paste0(figure.path,
                         "Toothfish_2024_pool_muscle_DMS25_VOLCANO_methylkit_regions.png"), 
                width = 30, height = 15, units = "cm")
```

### Old DMS
```{r}
print(load(paste0(rdata.path,"Toothfish_DMS_CpGsites.Rdata")))
meth.cpg <- paste0(myDiff25p$chr, "_", myDiff25p$start) #796 
DMS.cpg <- unique(c(DMS$DMS)) #1243
sum(meth.cpg %in% DMS.cpg) #10

print(load(paste0(rdata.path,"Toothfish_2024_muscle_methylation_DMS.Rdata")))
meth_2024_muscle.cpg <- paste0(myDiff25p$chr, "_", myDiff25p$start) #2359 
sum(meth.cpg %in% meth_2024_muscle.cpg) #105
```

### Heatmap

```{r}
print(load(paste0(rdata.path,"Toothfish_2024_pool_muscle_methylation_DMS.Rdata")))

myDiff25p.gr <- as(myDiff25p, "GRanges")
meth.diff <- methylKit::selectByOverlap(meth.2024poolM.norm, myDiff25p.gr)

meth.diff.df <- as.data.frame(t(methylKit::percMethylation(meth.diff)))
loc.names <- paste0(meth.diff$chr,"_",meth.diff$start)
colnames(meth.diff.df) <- loc.names

meth.diff.df$individual <- rownames(meth.diff.df)
pop <- metadata.new$pop
meth.diff.df$pop <- as.character(pop)


heat.plot.data <- meth.diff.df %>%
  tibble::as_tibble() %>%
  tidyr::gather(key = "CpG", value = "Methylation_percentage", -individual, -pop,
                factor_key = FALSE) %>%
  dplyr::group_by(pop, CpG) %>%
  dplyr::mutate(CpG_mean = mean(Methylation_percentage, na.rm = TRUE)) %>%
  dplyr::arrange(pop, individual, CpG) %>%
  dplyr::mutate(individual = factor(individual, levels = unique(individual))) %>%
  dplyr::mutate(CpG = factor(CpG, levels = unique(CpG)))
cpgs <- heat.plot.data$CpG[heat.plot.data$individual == unique(heat.plot.data$individual)[1]]
cpgs_mean <- heat.plot.data$CpG_mean[heat.plot.data$individual == unique(heat.plot.data$individual)[1]]

cpgs <- cpgs[order(cpgs_mean, decreasing = TRUE)]
heat.plot.data$CpG <- factor(heat.plot.data$CpG, levels = c("pop", unique(as.character(cpgs))))
heat.plot.data <- heat.plot.data[order(heat.plot.data$CpG),]

heat.plot <- ggplot2::ggplot(data = heat.plot.data, ggplot2::aes(x = individual)) + 
    ggplot2::geom_tile(data = heat.plot.data, 
                       ggplot2::aes(x = individual, y = CpG, fill = Methylation_percentage)) +
    ggplot2::scale_fill_gradient2(low = "blue", mid = "yellow", high = "red", 
                                  midpoint = 50, guide = "colourbar") +
    ggnewscale::new_scale_fill() +
    ggplot2::geom_tile(data = heat.plot.data, ggplot2::aes(x = individual, y = "pop", fill = pop,
                                    height = length(unique(heat.plot.data$CpG))*0.02)) +
    ggplot2::scale_fill_manual(values = colours2) +
    ggplot2::scale_y_discrete(drop = FALSE) +
    ggplot2::theme(legend.position = "top",
                   axis.title.y = ggplot2::element_blank(),
                   axis.title.x = ggplot2::element_blank(),
                   axis.text.x = ggplot2::element_text(face = "bold", color = "black", 
                                        size = 10, angle = 90, hjust = 1, vjust = 0.5))
print(heat.plot)

ggplot2::ggsave(heat.plot, bg = "white",
                filename = paste0(figure.path,
                                  "Toothfish_2024_pool_muscle_methDiff_25_HEATMAP.png"), 
                width = 30, height = 30, units = "cm")
```

### Plot CpG difference

```{r}
load(file = paste0(rdata.path,"Toothfish_2024_pool_muscle_methylation_DMS.Rdata"))

myDiff25p.gr <- as(myDiff25p, "GRanges")
meth.diff <- methylKit::selectByOverlap(meth.2024poolM.norm, myDiff25p.gr) #113 

length(unique(meth.diff$chr)) #34
sort(table(meth.diff$chr), decreasing = TRUE)[1:20]

meth.diff.df <- as.data.frame(t(methylKit::percMethylation(meth.diff)))
loc.names <- paste0(meth.diff$chr,"_",meth.diff$start)
colnames(meth.diff.df) <- loc.names

meth.diff.df$individual <- rownames(meth.diff.df) 
identical(metadata.new$Genotype_names, meth.diff.df$individual)
meth.diff.df$pop <- metadata.new$pop



meth.diff.df.plot <- tidyr::gather(meth.diff.df, key = locus, 
                           value = perc_methylation, -individual, -pop,
                           factor_key = TRUE) %>%
  dplyr::mutate(chr = stringr::str_split_fixed(locus, 
                                               pattern = "_", 
                                               n = 2)[,1])
meth.diff.df.plot.locus <- meth.diff.df.plot %>% dplyr::group_by(chr,locus,pop) %>% 
  dplyr::summarise(mean_pop = mean(perc_methylation, na.rm = TRUE),
                   sum_NA = sum(is.na(perc_methylation)),
            .groups = 'drop')


diff.plot <- ggplot2::ggplot() +
  ggplot2::geom_line(data = meth.diff.df.plot.locus,
                     ggplot2::aes(x = pop, y = mean_pop, colour = locus, group = locus)) +
  ggplot2::geom_point(data = meth.diff.df.plot.locus,
                     ggplot2::aes(x = pop, y = mean_pop, colour = locus, group = locus),
                     size = 5) +
  ggplot2::facet_wrap(~chr) +
  ggplot2::ylab("Mean methylation (%)") +
  ggplot2::theme(legend.position = "none",
                   axis.text.x = ggplot2::element_text(face = "bold", color = "black", 
                                        size = 10, angle = 90, hjust = 1, vjust = 0.5))

print(diff.plot)

ggplot2::ggsave(plot = diff.plot, 
                filename = paste0(figure.path,"1_Toothfish_2024_pool_muscle_methDiff.png"),
                width = 30, height = 15, units = "cm", device = "png")
```


### DMS Clustering
```{r}
load(file = paste0(rdata.path,"Toothfish_2024_pool_muscle_methylation_DMS.Rdata"))

myDiff25p.gr <- as(myDiff25p, "GRanges")
meth.diff <- methylKit::selectByOverlap(meth.2024poolM.norm, myDiff25p.gr) #112  
methdif.perc <- methylKit::percMethylation(meth.diff)

methdif.perc2 <- zoo::na.aggregate(methdif.perc, by = meth.2024poolM.norm@treatment)
PCobj <- prcomp(methdif.perc2, retx = TRUE, scale. = TRUE)

var_frac <- PCobj$sdev/sum(PCobj$sdev)
print(paste0("Total variance explained: ",signif(sum(var_frac[1:3]) * 100, 3), "%")) 
barplot(var_frac[1:3], main = "variance explained", 
        col = heat.colors(length(var_frac[1:3])))


identical(metadata.new$Genotype_names, meth.2024poolM.norm@sample.ids)
# identical(metadata.new$individual, rownames(PCobj$x))
identical(metadata.new$Genotype_names, rownames(PCobj$rotation))
pop <- metadata.new$pop

plot.title <- "Toothfish methylation - 2024 pool_muscle (796 DMS sites - 25%)"

# data <- data.frame(ind = rownames(PCobj$x),PCobj$x,  pop = pop)
data <- data.frame(ind = rownames(PCobj$rotation),PCobj$rotation,  pop = pop)
data <- merge(data, aggregate(cbind(mean.x = PC1, mean.y = PC2) ~
                                pop, data, mean), by = "pop")

pca.plot <- ggplot2::ggplot(data, ggplot2::aes(PC1, PC2, color = pop)) +
  ggplot2::geom_point(size = 4) + 
  ggplot2::geom_point(ggplot2::aes(x = mean.x,
                                   y = mean.y, color = pop),
                      size = 7,
                      shape = 17) +
  ggplot2::scale_colour_manual(values = colours2) +
  ggplot2::geom_hline(yintercept = 0) +
  ggplot2::geom_vline(xintercept = 0) + 
  ggplot2::labs(
    subtitle = paste0("Total variance explained: ",
                      signif(sum(var_frac[1:3]) * 100, 3), "%"),
    y = paste0("PC2; ", signif(var_frac[2] * 100, digits = 2), " %"),
    x = paste0("PC1; ", signif(var_frac[1] * 100, digits = 2), " %"),
    title = plot.title,
    caption = "") + 
  ggplot2::theme_minimal() +
  ggplot2::theme(
    axis.text = ggplot2::element_text(size = 15),
    axis.title.x = ggplot2::element_text(size = 20),
    axis.title.y = ggplot2::element_text(size = 20),
    legend.text = ggplot2::element_text(size = 15)
  )

print(pca.plot)
ggplot2::ggsave(pca.plot, bg = "white",
                filename = paste0(figure.path,"Toothfish_2024_pool_muscle_PCA_CpG_DMS_PC12.png"), 
                width = 30, height = 15, units = "cm")



# data <- data.frame(ind = rownames(PCobj$x),PCobj$x,  pop = pop)
data <- data.frame(ind = rownames(PCobj$rotation),PCobj$rotation,  pop = pop)
data <- merge(data, aggregate(cbind(mean.x = PC1, mean.y = PC3) ~
                                pop, data, mean), by = "pop")

pca.plot <- ggplot2::ggplot(data, ggplot2::aes(PC1, PC3, color = pop)) +
  ggplot2::geom_point(size = 4) + 
  ggplot2::geom_point(ggplot2::aes(x = mean.x,
                                   y = mean.y, color = pop),
                      size = 7,
                      shape = 17) +
  ggplot2::scale_colour_manual(values = colours2) +
  ggplot2::geom_hline(yintercept = 0) +
  ggplot2::geom_vline(xintercept = 0) + 
  ggplot2::labs(
    subtitle = paste0("Total variance explained: ",
                      signif(sum(var_frac[1:3]) * 100, 3), "%"),
    y = paste0("PC3; ", signif(var_frac[3] * 100, digits = 2), " %"),
    x = paste0("PC1; ", signif(var_frac[1] * 100, digits = 2), " %"),
    title = plot.title,
    caption = "") + 
  ggplot2::theme_minimal() +
  ggplot2::theme(
    axis.text = ggplot2::element_text(size = 15),
    axis.title.x = ggplot2::element_text(size = 20),
    axis.title.y = ggplot2::element_text(size = 20),
    legend.text = ggplot2::element_text(size = 15)
  )

print(pca.plot)
ggplot2::ggsave(pca.plot, bg = "white",
                filename = paste0(figure.path,
                                  "Toothfish_2024_pool_muscle_PCA_CpG_DMS_PC13.png"), 
                width = 30, height = 15, units = "cm")




# data <- data.frame(ind = rownames(PCobj$x),PCobj$x,  pop = pop)
data <- data.frame(ind = rownames(PCobj$rotation),PCobj$rotation,  pop = pop)
data <- merge(data, aggregate(cbind(mean.x = PC2, mean.y = PC3) ~
                                pop, data, mean), by = "pop")

pca.plot <- ggplot2::ggplot(data, ggplot2::aes(PC2, PC3, color = pop)) +
  ggplot2::geom_point(size = 4) + 
  ggplot2::geom_point(ggplot2::aes(x = mean.x,
                                   y = mean.y, color = pop),
                      size = 7,
                      shape = 17) +
  ggplot2::scale_colour_manual(values = colours2) +
  ggplot2::geom_hline(yintercept = 0) +
  ggplot2::geom_vline(xintercept = 0) + 
  ggplot2::labs(
    subtitle = paste0("Total variance explained: ",
                      signif(sum(var_frac[1:3]) * 100, 3), "%"),
    y = paste0("PC3; ", signif(var_frac[3] * 100, digits = 2), " %"),
    x = paste0("PC2; ", signif(var_frac[2] * 100, digits = 2), " %"),
    title = plot.title,
    caption = "") + 
  ggplot2::theme_minimal() +
  ggplot2::theme(
    axis.text = ggplot2::element_text(size = 15),
    axis.title.x = ggplot2::element_text(size = 20),
    axis.title.y = ggplot2::element_text(size = 20),
    legend.text = ggplot2::element_text(size = 15)
  )

print(pca.plot)
ggplot2::ggsave(pca.plot, bg = "white",
                filename = paste0(figure.path,"Toothfish_2024_pool_muscle_PCA_CpG_DMS_PC23.png"), 
                width = 30, height = 15, units = "cm")
```



## DMS RANDOM - 2024 - pool_muscle

```{r,eval=FALSE}
print(load(paste0(rdata.path,"Toothfish_2024_pool_muscle_methylkit_object.Rdata")))

print(load(paste0(rdata.path,"Toothfish_metadata.Rdata")))
metadata.new <- metadata[metadata$Genotype_names %in% attributes(meth.2024poolM.norm)$sample.ids,]
attributes(meth.2024poolM.norm)$sample.ids == metadata.new$Genotype_names

strata.random <- sample(metadata.new$pop, size = length(metadata.new$pop), replace = FALSE)
meth.2024poolM.norm@treatment <- as.integer(factor(strata.random))

# data.frame(attributes(meth.2024poolM.norm)$sample.ids, metadata.new$individual, metadata.new$strata, meth.2024poolM.norm@treatment)

# covariates <- dplyr::select(metadata.new, Sex, TL_mm)
myDiff <- methylKit::calculateDiffMeth(meth.2024poolM.norm,
                                       covariates = NULL,
                                       adjust = "fdr",
                                       chunk.size = 1e+06,
                                       mc.cores = 25)
meth.diff.chr <- methylKit::diffMethPerChr(myDiff,plot = TRUE, 
                                           qvalue.cutoff = 0.01, meth.cutoff = 40)
myDiff40p <- methylKit::getMethylDiff(myDiff, difference = 40, qvalue = 0.01)
#2
myDiff25p <- methylKit::getMethylDiff(myDiff, difference = 25, qvalue = 0.01)
#3
myDiff15p <- methylKit::getMethylDiff(myDiff, difference = 15, qvalue = 0.01)
#3  

save(strata.random, myDiff,myDiff15p, myDiff25p, myDiff40p,
     file = paste0(rdata.path,"Toothfish_2024_pool_muscle_methylation_RANDOM_DMS.Rdata"))
```


```{r}
print(load(paste0(rdata.path,"Toothfish_2024_pool_muscle_methylation_RANDOM_DMS.Rdata")))

volcano.plot <- ggplot2::ggplot() +
  ggplot2::geom_point(data = myDiff, ggplot2::aes(x = meth.diff, 
                                                  y = -log10(qvalue))) +
  ggplot2::geom_point(data = myDiff25p, ggplot2::aes(x = meth.diff, 
                                                  y = -log10(qvalue)),
                      colour = "red") +
  
  ggplot2::theme_minimal() +
  ggplot2::theme(
    axis.text = ggplot2::element_text(size = 15),
    axis.title.x = ggplot2::element_text(size = 20),
    axis.title.y = ggplot2::element_text(size = 20),
    legend.text = ggplot2::element_text(size = 15)
  )
print(volcano.plot)

ggplot2::ggsave(volcano.plot, bg = "white",
                filename =
                  paste0(figure.path,
                         "Toothfish_2024_pool_muscle_RANDOM_DMS25_VOLCANO_methylkit.png"), 
                width = 30, height = 15, units = "cm")
```


### Heatmap

```{r}
print(load(paste0(rdata.path,"Toothfish_2024_pool_muscle_methylation_RANDOM_DMS.Rdata")))

myDiff25p.gr <- as(myDiff25p, "GRanges")
meth.diff <- methylKit::selectByOverlap(meth.2024poolM.norm, myDiff25p.gr)

meth.diff.df <- as.data.frame(t(methylKit::percMethylation(meth.diff)))
loc.names <- paste0(meth.diff$chr,"_",meth.diff$start)
colnames(meth.diff.df) <- loc.names

meth.diff.df$individual <- rownames(meth.diff.df)
meth.diff.df$pop <- strata.random
meth.diff.df$pop2 <- as.character(metadata.new$pop)


heat.plot.data <- meth.diff.df %>%
  tibble::as_tibble() %>%
  tidyr::gather(key = "CpG", value = "Methylation_percentage", -individual, -pop, -pop2,
                factor_key = FALSE) %>%
  dplyr::group_by(pop, CpG) %>%
  dplyr::mutate(CpG_mean = mean(Methylation_percentage, na.rm = TRUE)) %>%
  dplyr::arrange(pop, individual, CpG) %>%
  dplyr::mutate(individual = factor(individual, levels = unique(individual))) %>%
  dplyr::mutate(CpG = factor(CpG, levels = unique(CpG)))
cpgs <- heat.plot.data$CpG[heat.plot.data$individual == unique(heat.plot.data$individual)[1]]
cpgs_mean <- heat.plot.data$CpG_mean[heat.plot.data$individual == unique(heat.plot.data$individual)[1]]

cpgs <- cpgs[order(cpgs_mean, decreasing = TRUE)]
heat.plot.data$CpG <- factor(heat.plot.data$CpG, levels = c("pop", unique(as.character(cpgs))))
heat.plot.data <- heat.plot.data[order(heat.plot.data$CpG),]

heat.plot <- ggplot2::ggplot(data = heat.plot.data, ggplot2::aes(x = individual)) + 
    ggplot2::geom_tile(data = heat.plot.data, 
                       ggplot2::aes(x = individual, y = CpG, fill = Methylation_percentage)) +
    ggplot2::scale_fill_gradient2(low = "blue", mid = "yellow", high = "red", 
                                  midpoint = 50, guide = "colourbar") +
    ggnewscale::new_scale_fill() +
    ggplot2::geom_tile(data = heat.plot.data, ggplot2::aes(x = individual, 
                                                           y = "pop - random", fill = pop,
                                    height = length(unique(heat.plot.data$CpG))*0.02)) +
    ggplot2::scale_fill_manual(values = colours2) +
    ggnewscale::new_scale_fill() +
    ggplot2::geom_tile(data = heat.plot.data, ggplot2::aes(x = individual, y = "pop", fill = pop2,
                                    height = length(unique(heat.plot.data$CpG))*0.02)) +
    ggplot2::scale_fill_manual(values = colours2) +
    ggplot2::scale_y_discrete(drop = FALSE) +
    ggplot2::theme(legend.position = "top",
                   axis.title.y = ggplot2::element_blank(),
                   axis.title.x = ggplot2::element_blank(),
                   axis.text.x = ggplot2::element_text(face = "bold", color = "black", 
                                        size = 10, angle = 90, hjust = 1, vjust = 0.5))
print(heat.plot)

ggplot2::ggsave(heat.plot, bg = "white",
                filename = paste0(figure.path,
                                  "Toothfish_2024_pool_muscle_RANDOM_methDiff_25_HEATMAP.png"), 
                width = 30, height = 30, units = "cm")
```

### Plot CpG difference

```{r}
load(file = paste0(rdata.path,"Toothfish_2024_pool_muscle_methylation_RANDOM_DMS.Rdata"))

myDiff25p.gr <- as(myDiff25p, "GRanges")
meth.diff <- methylKit::selectByOverlap(meth.2024poolM.norm, myDiff25p.gr) #113 

length(unique(meth.diff$chr)) #34
sort(table(meth.diff$chr), decreasing = TRUE)[1:20]

meth.diff.df <- as.data.frame(t(methylKit::percMethylation(meth.diff)))
loc.names <- paste0(meth.diff$chr,"_",meth.diff$start)
colnames(meth.diff.df) <- loc.names

meth.diff.df$individual <- rownames(meth.diff.df) 
identical(metadata.new$Genotype_names, meth.diff.df$individual)
meth.diff.df$pop <- metadata.new$pop


meth.diff.df.plot <- tidyr::gather(meth.diff.df, key = locus, 
                           value = perc_methylation, -individual, -pop,
                           factor_key = TRUE) %>%
  dplyr::mutate(chr = stringr::str_split_fixed(locus, 
                                               pattern = "_", 
                                               n = 2)[,1])
meth.diff.df.plot.locus <- meth.diff.df.plot %>% dplyr::group_by(chr,locus,pop) %>% 
  dplyr::summarise(mean_pop = mean(perc_methylation, na.rm = TRUE),
                   sum_NA = sum(is.na(perc_methylation)),
            .groups = 'drop')


diff.plot <- ggplot2::ggplot() +
  ggplot2::geom_line(data = meth.diff.df.plot.locus,
                     ggplot2::aes(x = pop, y = mean_pop, colour = locus, group = locus)) +
  ggplot2::geom_point(data = meth.diff.df.plot.locus,
                     ggplot2::aes(x = pop, y = mean_pop, colour = locus, group = locus),
                     size = 5) +
  ggplot2::facet_wrap(~chr) +
  ggplot2::ylab("Mean methylation (%)") +
  ggplot2::theme(legend.position = "none",
                   axis.text.x = ggplot2::element_text(face = "bold", color = "black", 
                                        size = 10, angle = 90, hjust = 1, vjust = 0.5))
    # ggplot2::guides(colour = ggplot2::guide_legend(ncol = 2))

print(diff.plot)

ggplot2::ggsave(plot = diff.plot, 
                filename = paste0(figure.path,"1_Toothfish_2024_pool_muscle_RANDOM_methDiff.png"),
                width = 30, height = 15, units = "cm", device = "png")
```


### DMS Clustering
```{r}
load(file = paste0(rdata.path,"Toothfish_2024_pool_muscle_methylation_RANDOM_DMS.Rdata"))

myDiff25p.gr <- as(myDiff25p, "GRanges")
meth.diff <- methylKit::selectByOverlap(meth.2024poolM.norm, myDiff25p.gr) #270  
methdif.perc <- methylKit::percMethylation(meth.diff)

methdif.perc2 <- zoo::na.aggregate(methdif.perc, by = meth.2024poolM.norm@treatment)
PCobj <- prcomp(methdif.perc2, retx = TRUE, scale. = TRUE) # Too many missing values

var_frac <- PCobj$sdev/sum(PCobj$sdev)
print(paste0("Total variance explained: ",signif(sum(var_frac[1:3]) * 100, 3), "%")) 
barplot(var_frac[1:3], main = "variance explained", 
        col = heat.colors(length(var_frac[1:3])))


identical(metadata.new$Genotype_names, meth.2024poolM.norm@sample.ids)

identical(as.numeric(factor(metadata.new$pop)), meth.2024poolM.norm@treatment)

# identical(metadata.new$individual, rownames(PCobj$x))
identical(metadata.new$Genotype_names, rownames(PCobj$rotation))
pop <- metadata.new$pop

plot.title <- "Toothfish methylation - RANDOMISED 2024 pool_muscle (3 DMS sites - 25%)"

# data <- data.frame(ind = rownames(PCobj$x),PCobj$x,  pop = pop)
data <- data.frame(ind = rownames(PCobj$rotation),PCobj$rotation,  pop = pop)
data <- merge(data, aggregate(cbind(mean.x = PC1, mean.y = PC2) ~
                                pop, data, mean), by = "pop")

pca.plot <- ggplot2::ggplot(data, ggplot2::aes(PC1, PC2, color = pop)) +
  ggplot2::geom_point(size = 4) + 
  ggplot2::geom_point(ggplot2::aes(x = mean.x,
                                   y = mean.y, color = pop),
                      size = 7,
                      shape = 17) +
  ggplot2::scale_colour_manual(values = colours2) +
  ggplot2::geom_hline(yintercept = 0) +
  ggplot2::geom_vline(xintercept = 0) + 
  ggplot2::labs(
    subtitle = paste0("Total variance explained: ",
                      signif(sum(var_frac[1:3]) * 100, 3), "%"),
    y = paste0("PC2; ", signif(var_frac[2] * 100, digits = 2), " %"),
    x = paste0("PC1; ", signif(var_frac[1] * 100, digits = 2), " %"),
    title = plot.title,
    caption = "") + 
  ggplot2::theme_minimal() +
  ggplot2::theme(
    axis.text = ggplot2::element_text(size = 15),
    axis.title.x = ggplot2::element_text(size = 20),
    axis.title.y = ggplot2::element_text(size = 20),
    legend.text = ggplot2::element_text(size = 15)
  )

print(pca.plot)
ggplot2::ggsave(pca.plot, bg = "white",
                filename = 
                  paste0(figure.path,"Toothfish_2024_pool_muscle_RANDOM_PCA_CpG_DMS_PC12.png"), 
                width = 30, height = 15, units = "cm")



# data <- data.frame(ind = rownames(PCobj$x),PCobj$x,  pop = pop)
data <- data.frame(ind = rownames(PCobj$rotation),PCobj$rotation,  pop = pop)
data <- merge(data, aggregate(cbind(mean.x = PC1, mean.y = PC3) ~
                                pop, data, mean), by = "pop")

pca.plot <- ggplot2::ggplot(data, ggplot2::aes(PC1, PC3, color = pop)) +
  ggplot2::geom_point(size = 4) + 
  ggplot2::geom_point(ggplot2::aes(x = mean.x,
                                   y = mean.y, color = pop),
                      size = 7,
                      shape = 17) +
  ggplot2::scale_colour_manual(values = colours2) +
  ggplot2::geom_hline(yintercept = 0) +
  ggplot2::geom_vline(xintercept = 0) + 
  ggplot2::labs(
    subtitle = paste0("Total variance explained: ",
                      signif(sum(var_frac[1:3]) * 100, 3), "%"),
    y = paste0("PC3; ", signif(var_frac[3] * 100, digits = 2), " %"),
    x = paste0("PC1; ", signif(var_frac[1] * 100, digits = 2), " %"),
    title = plot.title,
    caption = "") + 
  ggplot2::theme_minimal() +
  ggplot2::theme(
    axis.text = ggplot2::element_text(size = 15),
    axis.title.x = ggplot2::element_text(size = 20),
    axis.title.y = ggplot2::element_text(size = 20),
    legend.text = ggplot2::element_text(size = 15)
  )

print(pca.plot)
ggplot2::ggsave(pca.plot, bg = "white",
                filename = paste0(figure.path,"Toothfish_2024_pool_muscle_RANDOM_PCA_CpG_DMS_PC13.png"), 
                width = 30, height = 15, units = "cm")




# data <- data.frame(ind = rownames(PCobj$x),PCobj$x,  pop = pop)
data <- data.frame(ind = rownames(PCobj$rotation),PCobj$rotation,  pop = pop)
data <- merge(data, aggregate(cbind(mean.x = PC2, mean.y = PC3) ~
                                pop, data, mean), by = "pop")

pca.plot <- ggplot2::ggplot(data, ggplot2::aes(PC2, PC3, color = pop)) +
  ggplot2::geom_point(size = 4) + 
  ggplot2::geom_point(ggplot2::aes(x = mean.x,
                                   y = mean.y, color = pop),
                      size = 7,
                      shape = 17) +
  ggplot2::scale_colour_manual(values = colours2) +
  ggplot2::geom_hline(yintercept = 0) +
  ggplot2::geom_vline(xintercept = 0) + 
  ggplot2::labs(
    subtitle = paste0("Total variance explained: ",
                      signif(sum(var_frac[1:3]) * 100, 3), "%"),
    y = paste0("PC3; ", signif(var_frac[3] * 100, digits = 2), " %"),
    x = paste0("PC2; ", signif(var_frac[2] * 100, digits = 2), " %"),
    title = plot.title,
    caption = "") + 
  ggplot2::theme_minimal() +
  ggplot2::theme(
    axis.text = ggplot2::element_text(size = 15),
    axis.title.x = ggplot2::element_text(size = 20),
    axis.title.y = ggplot2::element_text(size = 20),
    legend.text = ggplot2::element_text(size = 15)
  )

print(pca.plot)
ggplot2::ggsave(pca.plot, bg = "white",
                filename = 
                  paste0(figure.path,"Toothfish_2024_pool_muscle_RANDOM_PCA_CpG_DMS_PC23.png"), 
                width = 30, height = 15, units = "cm")
```


## 2024 fish muscle with DMS from pooled samples
```{r}
print(load(paste0(rdata.path,"Toothfish_2024_EANT_muscle_methylkit_object.Rdata")))
print(load(paste0(rdata.path,"Toothfish_2024_pool_muscle_methylation_DMS.Rdata")))

meth.cpg <- paste0(meth.2024EANTM.norm11L$chr, "_", meth.2024EANTM.norm11L$start) #612803 
DMS.cpg <- paste0(myDiff25p$chr, "_", myDiff25p$start) #796
sum(meth.cpg %in% DMS.cpg) #752 = 94%

meth.DMS <- meth.2024EANTM.norm11L[meth.cpg %in% DMS.cpg,] #
# myDiff25p.gr <- as(myDiff25p, "GRanges")
# meth.diff <- methylKit::selectByOverlap(meth.2024EANTM.norm11L, myDiff25p.gr)


print(load(paste0(rdata.path,"Toothfish_metadata.Rdata")))
metadata.new <- metadata[metadata$Genotype_names %in% meth.DMS@sample.ids,]
metadata.new <- metadata.new[order(match(metadata.new$Genotype_names,
                                         meth.DMS@sample.ids)),]
```

### Heatmap

```{r}
meth.diff.df <- as.data.frame(t(methylKit::percMethylation(meth.DMS)))
loc.names <- paste0(meth.DMS$chr,"_",meth.DMS$start)
colnames(meth.diff.df) <- loc.names

meth.diff.df$individual <- rownames(meth.diff.df)
identical(metadata.new$individual,meth.diff.df$Genotype_names)
meth.diff.df$pop <- metadata.new$pop

heat.plot.data <- meth.diff.df %>%
  tibble::as_tibble() %>%
  tidyr::gather(key = "CpG", value = "Methylation_percentage", -individual, -pop,
                factor_key = FALSE) %>%
  dplyr::group_by(pop, CpG) %>%
  dplyr::mutate(CpG_mean = mean(Methylation_percentage, na.rm = TRUE)) %>%
  dplyr::arrange(pop, individual, CpG) %>%
  dplyr::mutate(individual = factor(individual, levels = unique(individual))) %>%
  dplyr::mutate(CpG = factor(CpG, levels = unique(CpG)))
cpgs <- heat.plot.data$CpG[heat.plot.data$individual == unique(heat.plot.data$individual)[1]]
cpgs_mean <- heat.plot.data$CpG_mean[heat.plot.data$individual == unique(heat.plot.data$individual)[1]]

cpgs <- cpgs[order(cpgs_mean, decreasing = TRUE)]
heat.plot.data$CpG <- factor(heat.plot.data$CpG, levels = c("pop", unique(as.character(cpgs))))
heat.plot.data <- heat.plot.data[order(heat.plot.data$CpG),]

heat.plot <- ggplot2::ggplot(data = heat.plot.data, ggplot2::aes(x = individual)) + 
    ggplot2::geom_tile(data = heat.plot.data, 
                       ggplot2::aes(x = individual, y = CpG, fill = Methylation_percentage)) +
    ggplot2::scale_fill_gradient2(low = "blue", mid = "yellow", high = "red", 
                                  midpoint = 50, guide = "colourbar") +
    ggnewscale::new_scale_fill() +
    ggplot2::geom_tile(data = heat.plot.data, ggplot2::aes(x = individual, y = "pop", fill = pop,
                                    height = length(unique(heat.plot.data$CpG))*0.02)) +
    ggplot2::scale_fill_manual(values = colours2) +
    ggplot2::scale_y_discrete(drop = FALSE) +
    ggplot2::theme(legend.position = "top",
                   axis.title.y = ggplot2::element_blank(),
                   axis.title.x = ggplot2::element_blank(),
                   axis.text.x = ggplot2::element_text(face = "bold", color = "black", 
                                        size = 10, angle = 90, hjust = 1, vjust = 0.5))
print(heat.plot)

ggplot2::ggsave(heat.plot, bg = "white",
                filename = 
                  paste0(figure.path,
                         "Toothfish_2024_muscle_BASED_ON_2024_pool_muscle_DMS_HEATMAP.png"), 
                width = 30, height = 30, units = "cm")
```


### DMS Clustering
```{r}
methdif.perc <- methylKit::percMethylation(meth.DMS)
methdif.perc2 <- zoo::na.aggregate(methdif.perc, by = meth.DMS@treatment)
PCobj <- prcomp(methdif.perc2, retx = TRUE, scale. = TRUE)

var_frac <- PCobj$sdev/sum(PCobj$sdev)
print(paste0("Total variance explained: ",signif(sum(var_frac[1:3]) * 100, 3), "%")) 
barplot(var_frac[1:3], main = "variance explained", 
        col = heat.colors(length(var_frac[1:3])))


identical(metadata.new$Genotype_names, meth.DMS@sample.ids)
# identical(metadata.new$individual, rownames(PCobj$x))
identical(metadata.new$Genotype_names, rownames(PCobj$rotation))
pop <- metadata.new$pop

plot.title <- "Toothfish methylation - 2024 Muscle (752 DMS sites from 2024_pool_muscle - 25%)"

# data <- data.frame(ind = rownames(PCobj$x),PCobj$x,  pop = pop)
data <- data.frame(ind = rownames(PCobj$rotation),PCobj$rotation,  pop = pop)
data <- merge(data, aggregate(cbind(mean.x = PC1, mean.y = PC2) ~
                                pop, data, mean), by = "pop")

pca.plot <- ggplot2::ggplot(data, ggplot2::aes(PC1, PC2, color = pop)) +
  ggplot2::geom_point(size = 4) + 
  ggplot2::geom_point(ggplot2::aes(x = mean.x,
                                   y = mean.y, color = pop),
                      size = 7,
                      shape = 17) +
  ggplot2::scale_colour_manual(values = colours2) +
  ggplot2::geom_hline(yintercept = 0) +
  ggplot2::geom_vline(xintercept = 0) + 
  ggplot2::labs(
    subtitle = paste0("Total variance explained: ",
                      signif(sum(var_frac[1:3]) * 100, 3), "%"),
    y = paste0("PC2; ", signif(var_frac[2] * 100, digits = 2), " %"),
    x = paste0("PC1; ", signif(var_frac[1] * 100, digits = 2), " %"),
    title = plot.title,
    caption = "") + 
  ggplot2::theme_minimal() +
  ggplot2::theme(
    axis.text = ggplot2::element_text(size = 15),
    axis.title.x = ggplot2::element_text(size = 20),
    axis.title.y = ggplot2::element_text(size = 20),
    legend.text = ggplot2::element_text(size = 15)
  )

print(pca.plot)
ggplot2::ggsave(pca.plot, bg = "white",
                filename = paste0(figure.path,"Toothfish_2024_muscle_BASED_ON_2024_pool_muscle_DMS_PC12.png"), 
                width = 30, height = 15, units = "cm")



# data <- data.frame(ind = rownames(PCobj$x),PCobj$x,  pop = pop)
data <- data.frame(ind = rownames(PCobj$rotation),PCobj$rotation,  pop = pop)
data <- merge(data, aggregate(cbind(mean.x = PC1, mean.y = PC3) ~
                                pop, data, mean), by = "pop")

pca.plot <- ggplot2::ggplot(data, ggplot2::aes(PC1, PC3, color = pop)) +
  ggplot2::geom_point(size = 4) + 
  ggplot2::geom_point(ggplot2::aes(x = mean.x,
                                   y = mean.y, color = pop),
                      size = 7,
                      shape = 17) +
  ggplot2::scale_colour_manual(values = colours2) +
  ggplot2::geom_hline(yintercept = 0) +
  ggplot2::geom_vline(xintercept = 0) + 
  ggplot2::labs(
    subtitle = paste0("Total variance explained: ",
                      signif(sum(var_frac[1:3]) * 100, 3), "%"),
    y = paste0("PC3; ", signif(var_frac[3] * 100, digits = 2), " %"),
    x = paste0("PC1; ", signif(var_frac[1] * 100, digits = 2), " %"),
    title = plot.title,
    caption = "") + 
  ggplot2::theme_minimal() +
  ggplot2::theme(
    axis.text = ggplot2::element_text(size = 15),
    axis.title.x = ggplot2::element_text(size = 20),
    axis.title.y = ggplot2::element_text(size = 20),
    legend.text = ggplot2::element_text(size = 15)
  )

print(pca.plot)
ggplot2::ggsave(pca.plot, bg = "white",
                filename = paste0(figure.path,
                                  "Toothfish_2024_muscle_BASED_ON_2024_pool_muscle_DMS_PC13.png"), 
                width = 30, height = 15, units = "cm")




# data <- data.frame(ind = rownames(PCobj$x),PCobj$x,  pop = pop)
data <- data.frame(ind = rownames(PCobj$rotation),PCobj$rotation,  pop = pop)
data <- merge(data, aggregate(cbind(mean.x = PC2, mean.y = PC3) ~
                                pop, data, mean), by = "pop")

pca.plot <- ggplot2::ggplot(data, ggplot2::aes(PC2, PC3, color = pop)) +
  ggplot2::geom_point(size = 4) + 
  ggplot2::geom_point(ggplot2::aes(x = mean.x,
                                   y = mean.y, color = pop),
                      size = 7,
                      shape = 17) +
  ggplot2::scale_colour_manual(values = colours2) +
  ggplot2::geom_hline(yintercept = 0) +
  ggplot2::geom_vline(xintercept = 0) + 
  ggplot2::labs(
    subtitle = paste0("Total variance explained: ",
                      signif(sum(var_frac[1:3]) * 100, 3), "%"),
    y = paste0("PC3; ", signif(var_frac[3] * 100, digits = 2), " %"),
    x = paste0("PC2; ", signif(var_frac[2] * 100, digits = 2), " %"),
    title = plot.title,
    caption = "") + 
  ggplot2::theme_minimal() +
  ggplot2::theme(
    axis.text = ggplot2::element_text(size = 15),
    axis.title.x = ggplot2::element_text(size = 20),
    axis.title.y = ggplot2::element_text(size = 20),
    legend.text = ggplot2::element_text(size = 15)
  )

print(pca.plot)
ggplot2::ggsave(pca.plot, bg = "white",
                filename = paste0(figure.path,"Toothfish_2024_muscle_BASED_ON_2024_pool_muscle_DMS_PC23.png"), 
                width = 30, height = 15, units = "cm")
```




## 2018 fish muscle with DMS from pooled samples

```{r}
print(load(paste0(rdata.path,"Toothfish_2018_muscle_methylkit_object.Rdata")))
print(load(paste0(rdata.path,"Toothfish_2024_pool_muscle_methylation_DMS.Rdata")))

meth.cpg <- paste0(meth.2018M.norm11L$chr, "_", meth.2018M.norm11L$start) #1232112 
DMS.cpg <- paste0(myDiff25p$chr, "_", myDiff25p$start) #796
sum(meth.cpg %in% DMS.cpg) #459 = 94%

meth.DMS <- meth.2018M.norm11L[meth.cpg %in% DMS.cpg,] #
# myDiff25p.gr <- as(myDiff25p, "GRanges")
# meth.diff <- methylKit::selectByOverlap(meth.2018M.norm11L, myDiff25p.gr)


print(load(paste0(rdata.path,"Toothfish_metadata.Rdata")))
metadata.new <- metadata[metadata$Genotype_names %in% meth.DMS@sample.ids,]
metadata.new <- metadata.new[order(match(metadata.new$Genotype_names,
                                         meth.DMS@sample.ids)),]
```

### Heatmap

```{r}
meth.diff.df <- as.data.frame(t(methylKit::percMethylation(meth.DMS)))
loc.names <- paste0(meth.DMS$chr,"_",meth.DMS$start)
colnames(meth.diff.df) <- loc.names

meth.diff.df$individual <- rownames(meth.diff.df)
identical(metadata.new$individual,meth.diff.df$Genotype_names)
meth.diff.df$pop <- metadata.new$pop


heat.plot.data <- meth.diff.df %>%
  tibble::as_tibble() %>%
  tidyr::gather(key = "CpG", value = "Methylation_percentage", -individual, -pop,
                factor_key = FALSE) %>%
  dplyr::group_by(pop, CpG) %>%
  dplyr::mutate(CpG_mean = mean(Methylation_percentage, na.rm = TRUE)) %>%
  dplyr::arrange(pop, individual, CpG) %>%
  dplyr::mutate(individual = factor(individual, levels = unique(individual))) %>%
  dplyr::mutate(CpG = factor(CpG, levels = unique(CpG)))
cpgs <- heat.plot.data$CpG[heat.plot.data$individual == unique(heat.plot.data$individual)[1]]
cpgs_mean <- heat.plot.data$CpG_mean[heat.plot.data$individual == unique(heat.plot.data$individual)[1]]

cpgs <- cpgs[order(cpgs_mean, decreasing = TRUE)]
heat.plot.data$CpG <- factor(heat.plot.data$CpG, levels = c("pop", unique(as.character(cpgs))))
heat.plot.data <- heat.plot.data[order(heat.plot.data$CpG),]

heat.plot <- ggplot2::ggplot(data = heat.plot.data, ggplot2::aes(x = individual)) + 
    ggplot2::geom_tile(data = heat.plot.data, 
                       ggplot2::aes(x = individual, y = CpG, fill = Methylation_percentage)) +
    ggplot2::scale_fill_gradient2(low = "blue", mid = "yellow", high = "red", 
                                  midpoint = 50, guide = "colourbar") +
    ggnewscale::new_scale_fill() +
    ggplot2::geom_tile(data = heat.plot.data, ggplot2::aes(x = individual, y = "pop", fill = pop,
                                    height = length(unique(heat.plot.data$CpG))*0.02)) +
    ggplot2::scale_fill_manual(values = colours4) +
    ggplot2::scale_y_discrete(drop = FALSE) +
    ggplot2::theme(legend.position = "top",
                   axis.title.y = ggplot2::element_blank(),
                   axis.title.x = ggplot2::element_blank(),
                   axis.text.x = ggplot2::element_text(face = "bold", color = "black", 
                                        size = 10, angle = 90, hjust = 1, vjust = 0.5))
print(heat.plot)

ggplot2::ggsave(heat.plot, bg = "white",
                filename = 
                  paste0(figure.path,
                         "Toothfish_2018_muscle_BASED_ON_2024_pool_muscle_DMS_HEATMAP.png"), 
                width = 30, height = 30, units = "cm")
```


### DMS Clustering
```{r}
methdif.perc <- methylKit::percMethylation(meth.DMS)
methdif.perc2 <- zoo::na.aggregate(methdif.perc, by = meth.DMS@treatment)
PCobj <- prcomp(methdif.perc2, retx = TRUE, scale. = TRUE)

var_frac <- PCobj$sdev/sum(PCobj$sdev)
print(paste0("Total variance explained: ",signif(sum(var_frac[1:3]) * 100, 3), "%")) 
barplot(var_frac[1:3], main = "variance explained", 
        col = heat.colors(length(var_frac[1:3])))


identical(metadata.new$Genotype_names, meth.DMS@sample.ids)
# identical(metadata.new$individual, rownames(PCobj$x))
identical(metadata.new$Genotype_names, rownames(PCobj$rotation))
pop <- metadata.new$pop

plot.title <- "Toothfish methylation - 2018 Muscle (459 DMS sites from 2024_pool_muscle - 25%)"

# data <- data.frame(ind = rownames(PCobj$x),PCobj$x,  pop = pop)
data <- data.frame(ind = rownames(PCobj$rotation),PCobj$rotation,  pop = pop)
data <- merge(data, aggregate(cbind(mean.x = PC1, mean.y = PC2) ~
                                pop, data, mean), by = "pop")

pca.plot <- ggplot2::ggplot(data, ggplot2::aes(PC1, PC2, color = pop)) +
  ggplot2::geom_point(size = 4) + 
  ggplot2::geom_point(ggplot2::aes(x = mean.x,
                                   y = mean.y, color = pop),
                      size = 7,
                      shape = 17) +
  ggplot2::scale_colour_manual(values = colours4) +
  ggplot2::geom_hline(yintercept = 0) +
  ggplot2::geom_vline(xintercept = 0) + 
  ggplot2::labs(
    subtitle = paste0("Total variance explained: ",
                      signif(sum(var_frac[1:3]) * 100, 3), "%"),
    y = paste0("PC2; ", signif(var_frac[2] * 100, digits = 2), " %"),
    x = paste0("PC1; ", signif(var_frac[1] * 100, digits = 2), " %"),
    title = plot.title,
    caption = "") + 
  ggplot2::theme_minimal() +
  ggplot2::theme(
    axis.text = ggplot2::element_text(size = 15),
    axis.title.x = ggplot2::element_text(size = 20),
    axis.title.y = ggplot2::element_text(size = 20),
    legend.text = ggplot2::element_text(size = 15)
  )

print(pca.plot)
ggplot2::ggsave(pca.plot, bg = "white",
                filename =
                  paste0(figure.path,
                         "Toothfish_2018_muscle_BASED_ON_2024_pool_muscle_DMS_PC12.png"), 
                width = 30, height = 15, units = "cm")



# data <- data.frame(ind = rownames(PCobj$x),PCobj$x,  pop = pop)
data <- data.frame(ind = rownames(PCobj$rotation),PCobj$rotation,  pop = pop)
data <- merge(data, aggregate(cbind(mean.x = PC1, mean.y = PC3) ~
                                pop, data, mean), by = "pop")

pca.plot <- ggplot2::ggplot(data, ggplot2::aes(PC1, PC3, color = pop)) +
  ggplot2::geom_point(size = 4) + 
  ggplot2::geom_point(ggplot2::aes(x = mean.x,
                                   y = mean.y, color = pop),
                      size = 7,
                      shape = 17) +
  ggplot2::scale_colour_manual(values = colours4) +
  ggplot2::geom_hline(yintercept = 0) +
  ggplot2::geom_vline(xintercept = 0) + 
  ggplot2::labs(
    subtitle = paste0("Total variance explained: ",
                      signif(sum(var_frac[1:3]) * 100, 3), "%"),
    y = paste0("PC3; ", signif(var_frac[3] * 100, digits = 2), " %"),
    x = paste0("PC1; ", signif(var_frac[1] * 100, digits = 2), " %"),
    title = plot.title,
    caption = "") + 
  ggplot2::theme_minimal() +
  ggplot2::theme(
    axis.text = ggplot2::element_text(size = 15),
    axis.title.x = ggplot2::element_text(size = 20),
    axis.title.y = ggplot2::element_text(size = 20),
    legend.text = ggplot2::element_text(size = 15)
  )

print(pca.plot)
ggplot2::ggsave(pca.plot, bg = "white",
                filename = paste0(figure.path,
                                  "Toothfish_2018_muscle_BASED_ON_2024_pool_muscle_DMS_PC13.png"), 
                width = 30, height = 15, units = "cm")




# data <- data.frame(ind = rownames(PCobj$x),PCobj$x,  pop = pop)
data <- data.frame(ind = rownames(PCobj$rotation),PCobj$rotation,  pop = pop)
data <- merge(data, aggregate(cbind(mean.x = PC2, mean.y = PC3) ~
                                pop, data, mean), by = "pop")

pca.plot <- ggplot2::ggplot(data, ggplot2::aes(PC2, PC3, color = pop)) +
  ggplot2::geom_point(size = 4) + 
  ggplot2::geom_point(ggplot2::aes(x = mean.x,
                                   y = mean.y, color = pop),
                      size = 7,
                      shape = 17) +
  ggplot2::scale_colour_manual(values = colours4) +
  ggplot2::geom_hline(yintercept = 0) +
  ggplot2::geom_vline(xintercept = 0) + 
  ggplot2::labs(
    subtitle = paste0("Total variance explained: ",
                      signif(sum(var_frac[1:3]) * 100, 3), "%"),
    y = paste0("PC3; ", signif(var_frac[3] * 100, digits = 2), " %"),
    x = paste0("PC2; ", signif(var_frac[2] * 100, digits = 2), " %"),
    title = plot.title,
    caption = "") + 
  ggplot2::theme_minimal() +
  ggplot2::theme(
    axis.text = ggplot2::element_text(size = 15),
    axis.title.x = ggplot2::element_text(size = 20),
    axis.title.y = ggplot2::element_text(size = 20),
    legend.text = ggplot2::element_text(size = 15)
  )

print(pca.plot)
ggplot2::ggsave(pca.plot, bg = "white",
                filename =
                  paste0(figure.path,
                         "Toothfish_2018_muscle_BASED_ON_2024_pool_muscle_DMS_PC23.png"), 
                width = 30, height = 15, units = "cm")
```







# 12_Methylation: 2024 - pool - fin tissue

```{r}
print(load(paste0(rdata.path,"Toothfish_metadata.Rdata"))) #ALL fish
```

## Reading the methylation call files

```{r, eval=FALSE}
data.list <- list.files(
  path = 
  "/datasets/work/ncmi-toa-rrbs/work/1.Toothfish/GENOME/Combined_methylation/",
  pattern = "*.cov.gz")

subnames <- metadata$File_names[!(metadata$Tissue_type == "muscle" |
                                    grepl(pattern = "TOA", x = metadata$File_names)
                                  )]
data.list <- data.list[data.list %in% paste0(subnames,
                                             "_R1_trimmed_sort_bismark_bt2.bismark.cov.gz")]
data.loc <- as.list(paste0(
  "/datasets/work/ncmi-toa-rrbs/work/1.Toothfish/GENOME/Combined_methylation/",
                           data.list))



data.list2 <-  gsub(pattern = '_R1_trimmed_sort_bismark_bt2.bismark.cov.gz',
                    replacement = '',x = as.character(data.list))


metadata.new <- metadata[metadata$File_names %in% data.list2,]
metadata.new <- metadata.new[order(match(metadata.new$File_names, data.list2)),]

metadata.new$File_names == data.list2

sample.ids <- metadata.new$Genotype_names
pop <- as.integer(factor(metadata.new$pop))

myobjDB <- methylKit::methRead(location = data.loc,
           sample.id = as.list(sample.ids),
           assembly = "TOA_genome",
           treatment = pop,
           context = "CpG",
           header = FALSE, 
           dbtype = "tabix",
           pipeline = "bismarkCoverage",
           mincov = 3
           )
save(myobjDB, file = paste0(rdata.path,"Toothfish_2024_pool_fin_methylkit_cpgDB.Rdata"))
```


## Data QC

#### Number of sites in genome

```{r, eval=FALSE}
cov.df <- data.frame(chr = NA, start = NA,end = NA, strand = NA, coverage = NA, 
                     numCs = NA, numTs = NA, sample = NA)
for (i in 1:length(sample.ids)) {
  df <- methylKit::getData(myobjDB[[i]])
  df$sample <- sample.ids[i]
  cov.df <- rbind(cov.df, df)
}
cov.df <- cov.df[-1,]
save(cov.df, file = paste0(rdata.path,"Toothfish_2024_pool_fin_methylkit_cpg_cov.Rdata"))
```

### Coverage stats

```{r}
load(paste0(rdata.path,"Toothfish_2024_pool_fin_methylkit_cpgDB.Rdata"))
print(load(paste0(rdata.path,"Toothfish_2024_pool_fin_methylkit_cpg_cov.Rdata")))

length(unique(paste0(cov.df$chr, "_", cov.df$start))) #1,894,182

# summary(as.factor(cov.df$sample))
summary(cov.df$coverage)
   # Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
   # 4.00    6.00   10.00   12.17   16.00 3000.00 

summary(cov.df$coverage[cov.df$coverage > 5 & cov.df$coverage < 50])
   # Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
   # 6.00    8.00   12.00   13.69   18.00   49.00 

cov.plot <- ggplot2::ggplot(cov.df, ggplot2::aes(x = log(coverage), fill = sample)) +
  ggplot2::geom_histogram() +
  ggplot2::theme(legend.position = "none")
# print(cov.plot)
ggplot2::ggsave(cov.plot,
                filename =
                  paste0(figure.path,"Toothfish_2024_pool_fin_coverage_hist_Combined.png"), 
                width = 45, height = 15, units = 'cm')

cov.plot <- ggplot2::ggplot(cov.df, ggplot2::aes(x = coverage, fill = sample)) +
  ggplot2::geom_histogram() +
  ggplot2::theme(legend.position = "none")
# print(cov.plot)
ggplot2::ggsave(cov.plot,
                filename =
                  paste0(figure.path,"Toothfish_2024_pool_fin_coverage_hist_Combined2.png"), 
                width = 45, height = 15, units = 'cm')

cov.plot <- ggplot2::ggplot(cov.df[cov.df$coverage > 5 & cov.df$coverage < 160,], 
                            ggplot2::aes(x = coverage)) +
  ggplot2::geom_histogram(binwidth = 1) +
  ggplot2::ggtitle("coverage < 160") +
  ggplot2::theme(legend.position = "none")
# print(cov.plot)
ggplot2::ggsave(cov.plot,
                filename =
                  paste0(figure.path,"Toothfish_2024_pool_fin_coverage_hist_Combined4.png"),
                width = 45, height = 15, units = 'cm')



max(cov.df$coverage, na.rm = T) #3000
ind.covplot <- ggplot2::ggplot(cov.df, ggplot2::aes(x = sample, y = coverage)) + 
  ggplot2::geom_boxplot() +
  ggplot2::theme_bw() + 
  ggplot2::theme(axis.title.x = ggplot2::element_blank(), 
                 axis.text.y = ggplot2::element_text(size = 15),
                 axis.title.y = ggplot2::element_text(size = 15),
                 axis.text.x = ggplot2::element_text(size = 7, angle = 90, hjust = 1)) +
  ggplot2::scale_y_continuous(trans = scales::log2_trans(),
                  breaks = c(1, 10, 50, 750, 3000),
                  minor_breaks = c(1:10, 2:5*10, 1:7.5*100, 0.5:2*1000))

print(ind.covplot)

ggplot2::ggsave(ind.covplot, bg = "white",
                filename = 
                  paste0(figure.path,
                         "Toothfish_2024_pool_fin_raw_individual_coverage_CpGs.png"),
                width = 45, height = 15, units = 'cm')
```


### Coverage filter

High count filter based on coverage stats.

```{r, eval=FALSE}
load(paste0(rdata.path,"Toothfish_2024_pool_fin_methylkit_cpgDB.Rdata"))

filtered.myobj <- methylKit::filterByCoverage(myobjDB, lo.count = 10, lo.perc = NULL,
                                              save.db = FALSE,
                                      hi.count = NULL, hi.perc = 99)
save(myobjDB,filtered.myobj,
     file = paste0(rdata.path,"Toothfish_2024_pool_fin_methylkit_cpgDB.Rdata"))
```

### Normalise coverage

Normalize read coverage distributions between samples. Requires stringent filter on max coverage first, otherwise it will over-sample PCR bias reads. These two functions will help reduce the bias in the statistical tests that might occur due to systematic over-sampling of reads in certain samples.

```{r}
filtered.myobj.normal <- methylKit::normalizeCoverage(obj = filtered.myobj, 
                                                      save.db = TRUE,
                                                      method = "median", chunk.size = 1e6)
save(myobjDB,filtered.myobj, filtered.myobj.normal,
     file = paste0(rdata.path,"Toothfish_2024_pool_fin_methylkit_cpgDB.Rdata"))
```

### Number of sites per sample

```{r, eval = FALSE}
sites.df <- data.frame(sites = NA,  sample = NA)
for (i in 1:length(filtered.myobj.normal)) {
  print(i)
  sites.df[i,] <- c(filtered.myobj.normal[[i]]@num.records, filtered.myobj.normal[[i]]@sample.id)
}
sites.df$sites <- as.integer(sites.df$sites)
# sites.df$pop <- filtered.myobj.normal@treatment
sites.df$pop <- metadata.new$pop
save(sites.df, file = paste0(rdata.path,"Toothfish_2024_pool_fin_number_cpg_sample.Rdata"))
```


```{r}
print(load(paste0(rdata.path,"Toothfish_2024_pool_fin_number_cpg_sample.Rdata")))
sites.plot <- ggplot2::ggplot(sites.df, ggplot2::aes(x = sample, y = sites, colour = pop)) +
  ggplot2::geom_point() +
  ggplot2::scale_color_manual(values = colours2) +
  ggplot2::theme_bw() +
  ggplot2::theme(axis.text.x = ggplot2::element_text(angle = 90))
print(sites.plot)

ggplot2::ggsave(sites.plot,
                filename =
                  paste0(figure.path,"Toothfish_2024_pool_fin_number_cpg_sample.png"),
                width = 45, height = 15, units = 'cm')

sites.df$sample[sites.df$sites < 610000]

# "88-2_Fin_Pool2"
```


## Select samples

```{r}
print(load(paste0(rdata.path,"Toothfish_metadata.Rdata")))

metadata2 <- metadata[!(metadata$Tissue_type == "muscle" |
                                    grepl(pattern = "TOA", x = metadata$File_names)
                                  ),]

print(load(paste0(rdata.path, "Low_quality_fish.Rdata")))

bad.list <- list(high.lambda.mapping,
          low.genome.mapping,low.genome.methylation,
          low.mbias.methylation)

rm <- Reduce(intersect, list(high.lambda.mapping,
          low.genome.mapping,low.genome.methylation,
          low.mbias.methylation))
print(rm)

# rm <- c(rm,) # bad coverage

metadata2 <- metadata2[!metadata2$File_names %in% rm,]

filt.sample.id.full <- sapply(filtered.myobj@.Data, function(x){x@sample.id})

# filt.sample.id <- filt.sample.id.full[filt.sample.id.full %in% metadata2$File_names]
# metadata2 <- metadata2[order(match(metadata.new$File_names, filt.sample.id)),]

filt.sample.id <- filt.sample.id.full[filt.sample.id.full %in% metadata2$Genotype_names]
metadata2 <- metadata2[order(match(metadata2$Genotype_names, filt.sample.id)),]
identical(filt.sample.id,metadata2$Genotype_names)


pop <- metadata2$pop
pop.levels <- c("58.4.2-2024",
                "88.2-2024")

pop <- factor(pop, levels = pop.levels)
pop <- as.numeric(pop)
sample.ids <- metadata2$Genotype_names



# filtered.myobj2 <- methylKit::reorganize(
#   methylObj = filtered.myobj,
#   sample.ids = sample.ids,
#   treatment = pop,
#   chunk.size = 1e+06,
#   save.db = TRUE,
#   suffix = "TOA_2024_pool_fin"
# )

filtered.myobj2.normal <- methylKit::reorganize(
  methylObj = filtered.myobj.normal,
  sample.ids = sample.ids,
  treatment = pop,
  chunk.size = 1e+06,
  save.db = TRUE,
  suffix = "TOA_2024_pool_fin_normal"
)
save(myobjDB,filtered.myobj, 
     filtered.myobj.normal,filtered.myobj2.normal,
     file = paste0(rdata.path,"Toothfish_2024_pool_fin_methylkit_cpgDB.Rdata"))
```


## Merging samples

```{r}
print(load(paste0(rdata.path,"Toothfish_2024_pool_fin_methylkit_cpgDB.Rdata")))
pop <- metadata2$pop
pop.levels <- c("58.4.2-2024",
                "88.2-2024")
pop <- factor(pop, levels = pop.levels)
pop <- as.numeric(pop)

table(pop)


# meth.2024M15L <- methylKit::unite(filtered.myobj2, destrand = FALSE,
#                              min.per.group = 15L, #missing data allowed
#                              mc.cores = 25)


meth.2024poolF.norm <- methylKit::unite(filtered.myobj2.normal, destrand = FALSE,
                             min.per.group = NULL, #missing data allowed
                             mc.cores = 25)
#400692 common sites for all samples per group

meth.2024poolF.norm <- as(meth.2024poolF.norm,"methylBase")

save(meth.2024poolF.norm,
     file = paste0(rdata.path,"Toothfish_2024_pool_fin_methylkit_object.Rdata"))
```


## Data QC


### Remove ASM

```{r}
load(paste0(rdata.path,"Allele_specific_mutations.Rdata"))

# example SNP
mut <- as(asmTRUE,"GRanges")
meth.2024poolF.norm <- as(meth.2024poolF.norm,"methylBase") 


sum(as(meth.2024poolF.norm,"GRanges") %over% mut)#21705
meth.asm <- methylKit::selectByOverlap(meth.2024poolF.norm, mut) #411698 
BOOLEAN <- paste0(meth.2024poolF.norm$chr,"_",meth.2024poolF.norm$start) %in%
  paste0(asmTRUE$Chr,"_",asmTRUE$Pos) #1
sum(BOOLEAN) #178
meth.2024poolF.norm <- methylKit::select(meth.2024poolF.norm, !BOOLEAN) # 751701   

save(meth.2024poolF.norm,
     file = paste0(rdata.path,"Toothfish_2024_pool_fin_methylkit_object.Rdata"))
```


### Coverage 

```{r}
meth.df <- methylKit::getData(meth.2024poolF.norm)
cov <- meth.df[,meth.2024poolF.norm@coverage.index]
colnames(cov) <- meth.2024poolF.norm@sample.ids
dpf <- data.frame(cov, locus = paste0(meth.df$chr,"_",meth.df$start))
dpf <- tidyr::gather(dpf, key = individual, 
                           value = Depth, -locus,
                           factor_key = TRUE) %>%
  dplyr::group_by(individual) %>%
  dplyr::mutate(NA_cnt = sum(is.na(Depth))/length(unique(locus))) %>%
  dplyr::ungroup()
dpf.NA <- dpf[!duplicated(dpf$individual),]
dpf.NA$NA_cnt <- round(dpf.NA$NA_cnt, 3)

dpf <- dpf[ dpf$Depth > 0 & !is.na(dpf$Depth),]

max(dpf$Depth, na.rm = T) #112
myPlot <- ggplot2::ggplot(dpf, ggplot2::aes(x = individual, y = Depth)) + 
  ggplot2::geom_boxplot() +
  ggplot2::geom_text(data = dpf.NA,ggplot2::aes(x = individual, y = 150, label = NA_cnt), 
                     size = 2, angle = 90) +
  ggplot2::theme_bw() + 
  ggplot2::theme(axis.title.x = ggplot2::element_blank(), 
                 axis.text.y = ggplot2::element_text(size = 15),
                 axis.title.y = ggplot2::element_text(size = 15),
                 axis.text.x = ggplot2::element_text(size = 7, angle = 90, hjust = 1)) +
  ggplot2::scale_y_continuous(trans = scales::log2_trans(),
                  breaks = c(1, 10, 50, 100),
                  minor_breaks = c(1:10, 2:5*10))

print(myPlot)
ggplot2::ggsave(myPlot, bg = "white",
                filename = 
                  paste0(figure.path,
                  "1B.Toothfish_2024_pool_fin_normalised15_methylkit_individual_coverage_CpGs.png"
                                  ), 
                width = 45, height = 15, units = "cm")

nrow(cov) # 400692 SMPs
plot(x = 1:ncol(cov), y = colMeans(cov, na.rm = TRUE), col = meth.2024poolF.norm@treatment)
plot(x = 1:ncol(cov), y = colMeans(cov > 20, na.rm = TRUE), col = meth.2024poolF.norm@treatment)
plot(x = 1:ncol(cov), y = colMeans(cov > 10, na.rm = TRUE), col = meth.2024poolF.norm@treatment)
meth.2024poolF.norm@sample.ids[colMeans(cov > 10, na.rm = TRUE) < 0.98]
# samples with low coverage: "58-4-2_Fin_Pool2"
```


## Load data

```{r}
print(load(paste0(rdata.path,"Toothfish_2024_pool_fin_methylkit_object.Rdata")))
print(load(paste0(rdata.path,"Toothfish_metadata.Rdata"))) 

metadata.new <- metadata[metadata$Genotype_names %in% meth.2024poolF.norm@sample.ids,]
metadata.new <- metadata.new[order(match(metadata.new$Genotype_names,
                                         meth.2024poolF.norm@sample.ids)),]
```


## Global clustering analysis

```{r}
methylKit::clusterSamples(meth.2024poolF.norm, dist = "correlation", sd.filter = TRUE, 
                          method = "ward.D2", plot = TRUE, )
dev.print(file = paste0(figure.path,"1_TOA_2024_pool_fin_cluster_plot.png"),
          device = png, res = 300,width = 45,height = 15,units = "cm")
```


### PCA

```{r}
# methdif.perc <- methylKit::percMethylation(meth.2024poolF.norm)
# methdif.perc2 <- zoo::na.aggregate(methdif.perc, by = meth.2024poolF.norm@treatment)
# PCobj <- prcomp(methdif.perc2, retx = TRUE, scale. = TRUE)
PCobj <- methylKit::PCASamples(meth.2024poolF.norm, comp = c(1,2), 
                               adj.lim = c(0,0), obj.return = TRUE)

var_frac <- PCobj$sdev/sum(PCobj$sdev)
print(paste0("Total variance explained: ",signif(sum(var_frac[1:3]) * 100, 3), "%")) 
barplot(var_frac[1:20], main = "variance explained", 
        col = heat.colors(length(var_frac[1:20])))

identical(metadata.new$Genotype_names, meth.2024poolF.norm@sample.ids)
identical(metadata.new$Genotype_names, rownames(PCobj$x))

pop <- metadata.new$pop

plot.title <- "Toothfish methylation - 2024 pool_fin (400692 CpG sites)"

data <- data.frame(ind = rownames(PCobj$x),PCobj$x,  pop = pop)
# data <- data.frame(ind = rownames(PCobj$rotation),PCobj$rotation, pop = pop)
data <- merge(data, aggregate(cbind(mean.x = PC1, mean.y = PC2) ~
                                pop, data, mean), by = "pop")

pca.plot <- ggplot2::ggplot(data, ggplot2::aes(PC1, PC2, color = pop)) +
  ggplot2::geom_point(size = 4) + 
  ggplot2::geom_point(ggplot2::aes(x = mean.x,
                                   y = mean.y, color = pop),
                      size = 7,
                      shape = 17) +
  ggplot2::scale_colour_manual(values = colours2) +
  ggplot2::geom_hline(yintercept = 0) +
  ggplot2::geom_vline(xintercept = 0) + 
  ggplot2::labs(
    subtitle = paste0("Total variance explained: ",
                      signif(sum(var_frac[1:4]) * 100, 3), "%"),
    y = paste0("PC2; ", signif(var_frac[2] * 100, digits = 2), " %"),
    x = paste0("PC1; ", signif(var_frac[1] * 100, digits = 2), " %"),
    title = plot.title,
    caption = "") + 
  ggplot2::theme_minimal() +
  ggplot2::theme(
    axis.text = ggplot2::element_text(size = 15),
    axis.title.x = ggplot2::element_text(size = 20),
    axis.title.y = ggplot2::element_text(size = 20),
    legend.text = ggplot2::element_text(size = 15)
  )

print(pca.plot)
ggplot2::ggsave(pca.plot, bg = "white",
                filename = paste0(figure.path,"Toothfish_2024_pool_fin_PCA_CpG_PC12.png"), 
                width = 30, height = 15, units = "cm")


data <- data.frame(ind = rownames(PCobj$x),PCobj$x,  pop = pop)
# data <- data.frame(ind = rownames(PCobj$rotation),PCobj$rotation, pop = pop)
data <- merge(data, aggregate(cbind(mean.x = PC2, mean.y = PC3) ~
                                pop, data, mean), by = "pop")
pca.plot <- ggplot2::ggplot(data, ggplot2::aes(PC2, PC3, color = pop)) +
  ggplot2::geom_point(size = 4) + 
  ggplot2::geom_point(ggplot2::aes(x = mean.x,
                                   y = mean.y, color = pop),
                      size = 7,
                      shape = 17) +
  ggplot2::scale_colour_manual(values = colours2) +
  ggplot2::geom_hline(yintercept = 0) +
  ggplot2::geom_vline(xintercept = 0) + 
  ggplot2::labs(
    subtitle = paste0("Total variance explained: ",
                      signif(sum(var_frac[1:4]) * 100, 3), "%"),
    y = paste0("PC3; ", signif(var_frac[3] * 100, digits = 2), " %"),
    x = paste0("PC2; ", signif(var_frac[2] * 100, digits = 2), " %"),
    title = plot.title,
    caption = "") + 
  ggplot2::theme_minimal() +
  ggplot2::theme(
    axis.text = ggplot2::element_text(size = 15),
    axis.title.x = ggplot2::element_text(size = 20),
    axis.title.y = ggplot2::element_text(size = 20),
    legend.text = ggplot2::element_text(size = 15)
  )

print(pca.plot)
ggplot2::ggsave(pca.plot, bg = "white",
                filename = paste0(figure.path,"Toothfish_2024_pool_fin_PCA_CpG_PC23.png"), 
                width = 30, height = 15, units = "cm")
```


### RDA

```{r}
# setwd("/datasets/work/ncmi-toa-rrbs/work/1.Toothfish/Toothfish_RRBS_array_22098078/Bismark/")
# print(load(paste0(rdata.path,"Toothfish_methylation_objects_raw.Rdata"))) #36,635 SMPs
print(load(paste0(rdata.path,"Toothfish_subset_methylation_objects_raw.Rdata"))) #1,316,799 SMPs

# meth.RDA <- methylKit::unite(filtered.myobj2.norm, destrand = FALSE,
#                              save.db = FALSE,
#                              suffix = "subset",
#                              min.per.group = NULL, #missing data allowed
#                              mc.cores = 25)


# meth.df <- methylKit::getData(meth)
# cov <- meth.df[,grep(colnames(meth.df), pattern = "coverage*")]
# cov <- meth.df[,meth@coverage.index]
# colnames(cov) <- meth@sample.ids
# meth.df <- methylKit::percMethylation(meth)

# meth.df <- methylKit::percMethylation(meth.RDA)

# save(meth.df,meth.RDA, 
#      file = paste0(rdata.path,"TOA_data_for_RDA_meth_percent.Rdata"))
print(load(paste0(rdata.path,"TOA_data_for_RDA_meth_percent.Rdata"))) #347,614 SMPs

hist(colSums(is.na(meth.df)), breaks = 100)
hist(rowSums(is.na(meth.df)), breaks = 100)





print(load(paste0(rdata.path,"Toothfish_metadata.Rdata")))

predictor <- metadata2 # depool_musclee the predictor input file here
predictor <- predictor[predictor$SAMPLE_ID %in% meth@sample.ids,]
predictor$SAMPLE_ID = as.character(predictor$SAMPLE_ID)
predictor$region = as.character(predictor$region)
predictor$Sex = as.character(predictor$Sex)
predictor$Tissue = as.character(predictor$Tissue)
predictor$mat = as.character(predictor$mat)
predictor$mat[is.na(predictor$mat)] <- "2"
predictor$TL = as.numeric(predictor$TL)
predictor$wt = as.numeric(predictor$wt)
predictor$dpth = as.numeric(predictor$dpth)
predictor$Age = as.numeric(predictor$Age)
predictor$Age[is.na(predictor$Age)] <- as.numeric(max(predictor$Age, na.rm = TRUE))
predictor$Cohort2 = as.integer(predictor$Cohort2)
predictor$Cohort2[is.na(predictor$Cohort2)] <- as.integer(min(predictor$Cohort2, na.rm = TRUE))


meth.df1 = t(meth.df)
# meth.df2 = na.omit(meth.df1) #rda can't handle NAs, remove them if you have them
# meth.df1[1:10,1:10]
## Predictor
# Make names characters (not factors)


#if you had numbers, they had to be factors (as.factor)

# Confirm that methylation data and predictors are in the same order
identical(rownames(meth.df1), predictor$SAMPLE_ID)

# Perform Hellinger transformation on the methylation data:
meth.df3 = vegan::decostand(meth.df1, method = "hellinger", MARGIN = 1) #MARGIN=1 means rows, on rows I have samples now because I transposed the matrix

## Run RDA
#Test the country using growing condition as a Condition () (covariate):
meth_rda = vegan::rda(meth.df3 ~ region + Condition(Sex), 
                      data = predictor, scale = F) #I remove scale since I performed Hellinger transformation
meth_rda = vegan::rda(meth.df3 ~ region + Condition(Sex) + 
                        Condition(mat)+ Condition(TL) + Condition(wt) +
                        Condition(dpth) + Condition(Age) + Condition(Cohort2), 
                      na.action = "na.omit",
                      data = predictor, scale = F)

meth_rda$pCCA
meth_rda$CCA$wa

summary(meth_rda) #Constrained Proportion: variance of Y explained by X, Unconstrained Proportion: unexplained variance in Y
# sink(paste0(rdata.path,"anova_rda.txt")) #with this you can save the output of the anova in a txt file, depool_musclee the output directory
meth.aov <- anova(meth_rda, test = "Chi", permutations = 99) #run Anova on the result to see the significance level
# meth.aov.cca <- anova.cca(meth_rda, test = "Chi", permutations = 99)
print(meth.aov)
# sink(paste0(rdata.path,"anova_rda_chi.txt"))

# Model: rda(formula = meth.df3 ~ region + Condition(Sex), data = predictor, scale = F)
#           Df  Variance      F Pr(>F)   
# Model      3 0.0003924 1.4126  0.004 **
# Residual 110 0.0101861

# Model: rda(formula = meth.df3 ~ region + Condition(Sex) + Condition(mat) + Condition(TL) + Condition(wt) + Condition(dpth) + Condition(Age) + Condition(Cohort2), data = predictor, scale = F, na.action = "na.omit")
#           Df  Variance     F Pr(>F)   
# Model      3 0.0003412 1.257   0.01 **
# Residual 101 0.0091396 

Pop.names <- predictor$region
plot.a <- vegan::ordiplot(meth_rda, choices=c(1,2))
Pop.ellipses <- vegan::ordiellipse(plot.a, groups=Pop.names, display="sites", kind="sd")
dev.print(file = paste0(figure.path,"TOA_RDA_ordiplot_complex_model.png"),
          device = png, res = 300,width = 45,height = 15,units = "cm")

# Type 1 scaling
vegan::ordiplot(meth_rda, scaling = 1, type = "text")
# Type 2 scaling
vegan::ordiplot(meth_rda, scaling = 2, type = "text")

ggvegan::autoplot(
  meth_rda,
  axes = c(1, 2),
  geom = "text",
  layers = c("species", "sites", "biplot", "centroids"),
  arrows = TRUE)

#CpG contribution
hist(meth_rda$CCA$v[,1])
hist(meth_rda$CCA$v[,2])
hist(meth_rda$CCA$v[,3])



#Test the interaction of your predictors:
meth_rda3 = vegan::rda(meth ~ region * Sex + Condition(region + Sex), data = predictor, scale = F)
summary(meth_rda3)
sink(paste0(rdata.path,"anova_rda_interaction.txt"))
print(anova(meth_rda3, test="Chi"))
sink(paste0(rdata.path,"anova_rda_interaction_chi.txt"))
```

#### plot RDA

```{r}

BioR.theme <- theme(
        panel.background = element_blank(),
        panel.border = element_blank(),
        panel.grid = element_blank(),
        axis.line = element_line("gray25"),
        text = element_text(size = 12),
        axis.text = element_text(size = 10, colour = "gray25"),
        axis.title = element_text(size = 14, colour = "gray25"),
        legend.title = element_text(size = 14),
        legend.text = element_text(size = 14),
        legend.key = element_blank())

plot.a <- vegan::ordiplot(meth_rda, choices=c(1,2))
Pop.names <- predictor$region

# sites1 <- BiodiversityR::sites.long(plot.a, env.data=predictor)
sites1 <- data.frame(cbind(predictor, axis1=plot.a$sites[, 1], 
                           axis2=plot.a$sites[, 2], 
                           labels=rownames(plot.a$sites)),
                     group = predictor$region)

# axis1 <- BiodiversityR::axis.long(meth_rda, choices=c(1, 2))
eigs.all <- meth_rda$CA$eig
eigs <- round(100 * eigs.all / meth_rda$tot.chi, digits=1)[c(1, 2)] 
xlab.label <- paste0(names(eigs)[1], " (", eigs[1], "%)")
ylab.label <- paste0(names(eigs)[2], " (", eigs[2], "%)")
axis1 <- data.frame(axis=c(1:2), 
                    ggplot=c("xlab.label", "ylab.label"), 
                    label=c(xlab.label, ylab.label))


# centroids.Pop1 <- BiodiversityR::centroids.long(sites1, 
#                                                 grouping=Pop.names, 
#                                                 centroids.only=TRUE)
# centroids.Pop2 <- BiodiversityR::centroids.long(sites1, 
#                                                 grouping=Pop.names, 
#                                                 centroids.only=FALSE)
gr.name <- rlang::enexpr(Pop.names)
cent.means <-  stats::aggregate(cbind(axis1, axis2) ~ Pop.names,
                                data = sites1,
                                FUN = mean)
names(cent.means) <- c("group", "axis1c", "axis2c")
centroids.Pop1 <- cent.means
cent.means$Centroid <- cent.means[, 1]
by.name <- names(sites1)[names(sites1) == "group"]
centroids.Pop2 <- dplyr::full_join(sites1, cent.means, by = by.name)


# Pop.ellipses.long2 <- BiodiversityR::ordiellipse.long(Pop.ellipses, 
#                                                     grouping.name="Pop.names")


plotgg.a <- ggplot() + 
    geom_vline(xintercept = c(0), color = "grey70", linetype = 2) +
    geom_hline(yintercept = c(0), color = "grey70", linetype = 2) +  
    xlab(axis1[1, "label"]) +
    ylab(axis1[2, "label"]) +  
    scale_x_continuous(sec.axis = dup_axis(labels=NULL, name=NULL)) +
    scale_y_continuous(sec.axis = dup_axis(labels=NULL, name=NULL)) +
    # geom_polygon(data=Pop.ellipses.long2, 
    #                aes(x=axis1, y=axis2, colour=Pop.names, 
    #                    fill=after_scale(alpha(colour, 0.2))), 
    #           size=0.2, show.legend=TRUE) +
    geom_segment(data=centroids.Pop2, 
                 aes(x=axis1c, y=axis2c, xend=axis1, yend=axis2, colour=Pop.names), 
                 size=0.7, show.legend=FALSE) +
    geom_point(data=centroids.Pop1, 
               aes(x=axis1c, y=axis2c), 
               shape="square", colour="black", alpha=0.7, size=2) +
    geom_label_repel(data=centroids.Pop1, 
               aes(x=axis1c, y=axis2c, label=group, colour=group), 
               alpha=1.0, size=2, show.legend=FALSE) +
    # labs(colour = Pop.names) +
    BioR.theme +
    theme(legend.text = element_text(size = 10)) +
    scale_colour_npg() +
    coord_fixed(ratio=1)

plotgg.a

ggsave(plotgg.a, file = paste0(figure.path, 
                               # "TOA_RDA_simple_model_A.png"
                               "TOA_RDA_complex_model_A.png"
                               ),
       height = 30, width = 30, units = "cm",  device = "png")
```



## DMS - 2024 - pool_fin

```{r, eval = FALSE}
print(load(paste0(rdata.path,"Toothfish_2024_pool_fin_methylkit_object.Rdata")))

print(load(paste0(rdata.path,"Toothfish_metadata.Rdata")))
metadata.new <- metadata[metadata$Genotype_names %in% attributes(meth.2024poolF.norm)$sample.ids,]
attributes(meth.2024poolF.norm)$sample.ids == metadata.new$Genotype_names

meth.2024poolF.norm@treatment <- as.integer(factor(metadata.new$pop))


covariates <- dplyr::select(metadata.new, Sex, Age, Cohort2, Maturity_full)
myDiff <- methylKit::calculateDiffMeth(meth.2024poolF.norm,
                                       covariates = NULL,
                                       adjust = "fdr",
                                       chunk.size = 1e+06,
                                       mc.cores = 24)
meth.diff.chr <- methylKit::diffMethPerChr(myDiff,plot = TRUE, 
                                           qvalue.cutoff = 0.05, meth.cutoff = 40)
myDiff40p <- methylKit::getMethylDiff(myDiff, difference = 40, qvalue = 0.05)
#114
myDiff25p <- methylKit::getMethylDiff(myDiff, difference = 25, qvalue = 0.05)
#184
myDiff15p <- methylKit::getMethylDiff(myDiff, difference = 15, qvalue = 0.05)
#193
top100.POOLF <- slice_max(as.data.frame(as(myDiff,"GRanges")), order_by = meth.diff, n = 100)
top100.POOLF <- as(top100.POOLF,"GRanges")

save(myDiff,myDiff15p, myDiff25p, myDiff40p,top100.POOLF,
     file = paste0(rdata.path,"Toothfish_2024_pool_fin_methylation_DMS.Rdata"))
```


```{r}
print(load(paste0(rdata.path,"Toothfish_2024_pool_fin_methylation_DMS.Rdata")))

volcano.plot <- ggplot2::ggplot() +
  ggplot2::geom_point(data = myDiff, ggplot2::aes(x = meth.diff, 
                                                  y = -log10(qvalue))) +
  ggplot2::geom_point(data = myDiff25p, ggplot2::aes(x = meth.diff, 
                                                  y = -log10(qvalue)),
                      colour = "red") +
  ggplot2::theme_minimal() +
  ggplot2::theme(
    axis.text = ggplot2::element_text(size = 15),
    axis.title.x = ggplot2::element_text(size = 20),
    axis.title.y = ggplot2::element_text(size = 20),
    legend.text = ggplot2::element_text(size = 15)
  )
print(volcano.plot)

ggplot2::ggsave(volcano.plot, bg = "white",
                filename =
                  paste0(figure.path,
                         "Toothfish_2024_pool_fin_DMS25_VOLCANO_methylkit_regions.png"), 
                width = 30, height = 15, units = "cm")
```

### Old DMS
```{r}
print(load(paste0(rdata.path,"Toothfish_DMS_CpGsites.Rdata")))
# meth <- meth.2024poolF.norm
meth.cpg <- paste0(myDiff25p$chr, "_", myDiff25p$start) #184 
DMS.cpg <- unique(c(DMS$DMS)) #1243
sum(meth.cpg %in% DMS.cpg) #0
# meth.DMS <- meth[meth.cpg %in% DMS.cpg,] #1144

print(load(paste0(rdata.path,"Toothfish_2024_fin_methylation_DMS.Rdata")))
meth_2024_fin.cpg <- paste0(myDiff25p$chr, "_", myDiff25p$start) #2359 
sum(meth.cpg %in% meth_2024_fin.cpg) #0
```

### Heatmap

```{r}
print(load(paste0(rdata.path,"Toothfish_2024_pool_fin_methylation_DMS.Rdata")))

myDiff25p.gr <- as(myDiff25p, "GRanges")
meth.diff <- methylKit::selectByOverlap(meth.2024poolF.norm, myDiff25p.gr)

meth.diff.df <- as.data.frame(t(methylKit::percMethylation(meth.diff)))
loc.names <- paste0(meth.diff$chr,"_",meth.diff$start)
colnames(meth.diff.df) <- loc.names

meth.diff.df$individual <- rownames(meth.diff.df)
pop <- metadata.new$pop
meth.diff.df$pop <- as.character(pop)


heat.plot.data <- meth.diff.df %>%
  tibble::as_tibble() %>%
  tidyr::gather(key = "CpG", value = "Methylation_percentage", -individual, -pop,
                factor_key = FALSE) %>%
  dplyr::group_by(pop, CpG) %>%
  dplyr::mutate(CpG_mean = mean(Methylation_percentage, na.rm = TRUE)) %>%
  dplyr::arrange(pop, individual, CpG) %>%
  dplyr::mutate(individual = factor(individual, levels = unique(individual))) %>%
  dplyr::mutate(CpG = factor(CpG, levels = unique(CpG)))
cpgs <- heat.plot.data$CpG[heat.plot.data$individual == unique(heat.plot.data$individual)[1]]
cpgs_mean <- heat.plot.data$CpG_mean[heat.plot.data$individual == unique(heat.plot.data$individual)[1]]

cpgs <- cpgs[order(cpgs_mean, decreasing = TRUE)]
heat.plot.data$CpG <- factor(heat.plot.data$CpG, levels = c("pop", unique(as.character(cpgs))))
heat.plot.data <- heat.plot.data[order(heat.plot.data$CpG),]

heat.plot <- ggplot2::ggplot(data = heat.plot.data, ggplot2::aes(x = individual)) + 
    ggplot2::geom_tile(data = heat.plot.data, 
                       ggplot2::aes(x = individual, y = CpG, fill = Methylation_percentage)) +
    ggplot2::scale_fill_gradient2(low = "blue", mid = "yellow", high = "red", 
                                  midpoint = 50, guide = "colourbar") +
    ggnewscale::new_scale_fill() +
    ggplot2::geom_tile(data = heat.plot.data, ggplot2::aes(x = individual, y = "pop", fill = pop,
                                    height = length(unique(heat.plot.data$CpG))*0.02)) +
    ggplot2::scale_fill_manual(values = colours2) +
    ggplot2::scale_y_discrete(drop = FALSE) +
    ggplot2::theme(legend.position = "top",
                   axis.title.y = ggplot2::element_blank(),
                   axis.title.x = ggplot2::element_blank(),
                   axis.text.x = ggplot2::element_text(face = "bold", color = "black", 
                                        size = 10, angle = 90, hjust = 1, vjust = 0.5))
print(heat.plot)

ggplot2::ggsave(heat.plot, bg = "white",
                filename = paste0(figure.path,
                                  "Toothfish_2024_pool_fin_methDiff_25_HEATMAP.png"), 
                width = 30, height = 30, units = "cm")
```

### Plot CpG difference

```{r}
load(file = paste0(rdata.path,"Toothfish_2024_pool_fin_methylation_DMS.Rdata"))

myDiff25p.gr <- as(myDiff25p, "GRanges")
meth.diff <- methylKit::selectByOverlap(meth.2024poolF.norm, myDiff25p.gr) #113 

length(unique(meth.diff$chr)) #34
sort(table(meth.diff$chr), decreasing = TRUE)[1:20]

meth.diff.df <- as.data.frame(t(methylKit::percMethylation(meth.diff)))
loc.names <- paste0(meth.diff$chr,"_",meth.diff$start)
colnames(meth.diff.df) <- loc.names

meth.diff.df$individual <- rownames(meth.diff.df) 
identical(metadata.new$Genotype_names, meth.diff.df$individual)
meth.diff.df$pop <- metadata.new$pop



meth.diff.df.plot <- tidyr::gather(meth.diff.df, key = locus, 
                           value = perc_methylation, -individual, -pop,
                           factor_key = TRUE) %>%
  dplyr::mutate(chr = stringr::str_split_fixed(locus, 
                                               pattern = "_", 
                                               n = 2)[,1])
meth.diff.df.plot.locus <- meth.diff.df.plot %>% dplyr::group_by(chr,locus,pop) %>% 
  dplyr::summarise(mean_pop = mean(perc_methylation, na.rm = TRUE),
                   sum_NA = sum(is.na(perc_methylation)),
            .groups = 'drop')


diff.plot <- ggplot2::ggplot() +
  ggplot2::geom_line(data = meth.diff.df.plot.locus,
                     ggplot2::aes(x = pop, y = mean_pop, colour = locus, group = locus)) +
  ggplot2::geom_point(data = meth.diff.df.plot.locus,
                     ggplot2::aes(x = pop, y = mean_pop, colour = locus, group = locus),
                     size = 5) +
  ggplot2::facet_wrap(~chr) +
  ggplot2::ylab("Mean methylation (%)") +
  ggplot2::theme(legend.position = "none",
                   axis.text.x = ggplot2::element_text(face = "bold", color = "black", 
                                        size = 10, angle = 90, hjust = 1, vjust = 0.5))

print(diff.plot)

ggplot2::ggsave(plot = diff.plot, 
                filename = paste0(figure.path,"1_Toothfish_2024_pool_fin_methDiff.png"),
                width = 30, height = 15, units = "cm", device = "png")
```


### DMS Clustering
```{r}
load(file = paste0(rdata.path,"Toothfish_2024_pool_fin_methylation_DMS.Rdata"))

myDiff25p.gr <- as(myDiff25p, "GRanges")
meth.diff <- methylKit::selectByOverlap(meth.2024poolF.norm, myDiff25p.gr) #112  
methdif.perc <- methylKit::percMethylation(meth.diff)

methdif.perc2 <- zoo::na.aggregate(methdif.perc, by = meth.2024poolF.norm@treatment)
PCobj <- prcomp(methdif.perc2, retx = TRUE, scale. = TRUE)

var_frac <- PCobj$sdev/sum(PCobj$sdev)
print(paste0("Total variance explained: ",signif(sum(var_frac[1:3]) * 100, 3), "%")) 
barplot(var_frac[1:3], main = "variance explained", 
        col = heat.colors(length(var_frac[1:3])))


identical(metadata.new$Genotype_names, meth.2024poolF.norm@sample.ids)
# identical(metadata.new$individual, rownames(PCobj$x))
identical(metadata.new$Genotype_names, rownames(PCobj$rotation))
pop <- metadata.new$pop

plot.title <- "Toothfish methylation - 2024 pool_fin (2359 DMS sites - 25%)"

# data <- data.frame(ind = rownames(PCobj$x),PCobj$x,  pop = pop)
data <- data.frame(ind = rownames(PCobj$rotation),PCobj$rotation,  pop = pop)
data <- merge(data, aggregate(cbind(mean.x = PC1, mean.y = PC2) ~
                                pop, data, mean), by = "pop")

pca.plot <- ggplot2::ggplot(data, ggplot2::aes(PC1, PC2, color = pop)) +
  ggplot2::geom_point(size = 4) + 
  ggplot2::geom_point(ggplot2::aes(x = mean.x,
                                   y = mean.y, color = pop),
                      size = 7,
                      shape = 17) +
  ggplot2::scale_colour_manual(values = colours2) +
  ggplot2::geom_hline(yintercept = 0) +
  ggplot2::geom_vline(xintercept = 0) + 
  ggplot2::labs(
    subtitle = paste0("Total variance explained: ",
                      signif(sum(var_frac[1:3]) * 100, 3), "%"),
    y = paste0("PC2; ", signif(var_frac[2] * 100, digits = 2), " %"),
    x = paste0("PC1; ", signif(var_frac[1] * 100, digits = 2), " %"),
    title = plot.title,
    caption = "") + 
  ggplot2::theme_minimal() +
  ggplot2::theme(
    axis.text = ggplot2::element_text(size = 15),
    axis.title.x = ggplot2::element_text(size = 20),
    axis.title.y = ggplot2::element_text(size = 20),
    legend.text = ggplot2::element_text(size = 15)
  )

print(pca.plot)
ggplot2::ggsave(pca.plot, bg = "white",
                filename = paste0(figure.path,"Toothfish_2024_pool_fin_PCA_CpG_DMS_PC12.png"), 
                width = 30, height = 15, units = "cm")



# data <- data.frame(ind = rownames(PCobj$x),PCobj$x,  pop = pop)
data <- data.frame(ind = rownames(PCobj$rotation),PCobj$rotation,  pop = pop)
data <- merge(data, aggregate(cbind(mean.x = PC1, mean.y = PC3) ~
                                pop, data, mean), by = "pop")

pca.plot <- ggplot2::ggplot(data, ggplot2::aes(PC1, PC3, color = pop)) +
  ggplot2::geom_point(size = 4) + 
  ggplot2::geom_point(ggplot2::aes(x = mean.x,
                                   y = mean.y, color = pop),
                      size = 7,
                      shape = 17) +
  ggplot2::scale_colour_manual(values = colours2) +
  ggplot2::geom_hline(yintercept = 0) +
  ggplot2::geom_vline(xintercept = 0) + 
  ggplot2::labs(
    subtitle = paste0("Total variance explained: ",
                      signif(sum(var_frac[1:3]) * 100, 3), "%"),
    y = paste0("PC3; ", signif(var_frac[3] * 100, digits = 2), " %"),
    x = paste0("PC1; ", signif(var_frac[1] * 100, digits = 2), " %"),
    title = plot.title,
    caption = "") + 
  ggplot2::theme_minimal() +
  ggplot2::theme(
    axis.text = ggplot2::element_text(size = 15),
    axis.title.x = ggplot2::element_text(size = 20),
    axis.title.y = ggplot2::element_text(size = 20),
    legend.text = ggplot2::element_text(size = 15)
  )

print(pca.plot)
ggplot2::ggsave(pca.plot, bg = "white",
                filename = paste0(figure.path,
                                  "Toothfish_2024_pool_fin_PCA_CpG_DMS_PC13.png"), 
                width = 30, height = 15, units = "cm")




# data <- data.frame(ind = rownames(PCobj$x),PCobj$x,  pop = pop)
data <- data.frame(ind = rownames(PCobj$rotation),PCobj$rotation,  pop = pop)
data <- merge(data, aggregate(cbind(mean.x = PC2, mean.y = PC3) ~
                                pop, data, mean), by = "pop")

pca.plot <- ggplot2::ggplot(data, ggplot2::aes(PC2, PC3, color = pop)) +
  ggplot2::geom_point(size = 4) + 
  ggplot2::geom_point(ggplot2::aes(x = mean.x,
                                   y = mean.y, color = pop),
                      size = 7,
                      shape = 17) +
  ggplot2::scale_colour_manual(values = colours2) +
  ggplot2::geom_hline(yintercept = 0) +
  ggplot2::geom_vline(xintercept = 0) + 
  ggplot2::labs(
    subtitle = paste0("Total variance explained: ",
                      signif(sum(var_frac[1:3]) * 100, 3), "%"),
    y = paste0("PC3; ", signif(var_frac[3] * 100, digits = 2), " %"),
    x = paste0("PC2; ", signif(var_frac[2] * 100, digits = 2), " %"),
    title = plot.title,
    caption = "") + 
  ggplot2::theme_minimal() +
  ggplot2::theme(
    axis.text = ggplot2::element_text(size = 15),
    axis.title.x = ggplot2::element_text(size = 20),
    axis.title.y = ggplot2::element_text(size = 20),
    legend.text = ggplot2::element_text(size = 15)
  )

print(pca.plot)
ggplot2::ggsave(pca.plot, bg = "white",
                filename = paste0(figure.path,"Toothfish_2024_pool_fin_PCA_CpG_DMS_PC23.png"), 
                width = 30, height = 15, units = "cm")
```



## DMS RANDOM - 2024 - pool_fin

```{r,eval=FALSE}
print(load(paste0(rdata.path,"Toothfish_2024_pool_fin_methylkit_object.Rdata")))

print(load(paste0(rdata.path,"Toothfish_metadata.Rdata")))
metadata.new <- metadata[metadata$Genotype_names %in% attributes(meth.2024poolF.norm)$sample.ids,]
attributes(meth.2024poolF.norm)$sample.ids == metadata.new$Genotype_names

strata.random <- sample(metadata.new$pop, size = length(metadata.new$pop), replace = FALSE)
meth.2024poolF.norm@treatment <- as.integer(factor(strata.random))

# data.frame(attributes(meth.2024poolF.norm)$sample.ids, metadata.new$individual, metadata.new$strata, meth.2024poolF.norm@treatment)

# covariates <- dplyr::select(metadata.new, Sex, TL_mm)
myDiff <- methylKit::calculateDiffMeth(meth.2024poolF.norm,
                                       covariates = NULL,
                                       adjust = "fdr",
                                       chunk.size = 1e+06,
                                       mc.cores = 25)
meth.diff.chr <- methylKit::diffMethPerChr(myDiff,plot = TRUE, 
                                           qvalue.cutoff = 0.05, meth.cutoff = 40)
myDiff40p <- methylKit::getMethylDiff(myDiff, difference = 40, qvalue = 0.05)
#16
myDiff25p <- methylKit::getMethylDiff(myDiff, difference = 25, qvalue = 0.05)
#24
myDiff15p <- methylKit::getMethylDiff(myDiff, difference = 15, qvalue = 0.05)
#26

save(strata.random, myDiff,myDiff15p, myDiff25p, myDiff40p,
     file = paste0(rdata.path,"Toothfish_2024_pool_fin_methylation_RANDOM_DMS.Rdata"))
```


```{r}
print(load(paste0(rdata.path,"Toothfish_2024_pool_fin_methylation_RANDOM_DMS.Rdata")))

volcano.plot <- ggplot2::ggplot() +
  ggplot2::geom_point(data = myDiff, ggplot2::aes(x = meth.diff, 
                                                  y = -log10(qvalue))) +
  ggplot2::geom_point(data = myDiff25p, ggplot2::aes(x = meth.diff, 
                                                  y = -log10(qvalue)),
                      colour = "red") +
  
  ggplot2::theme_minimal() +
  ggplot2::theme(
    axis.text = ggplot2::element_text(size = 15),
    axis.title.x = ggplot2::element_text(size = 20),
    axis.title.y = ggplot2::element_text(size = 20),
    legend.text = ggplot2::element_text(size = 15)
  )
print(volcano.plot)

ggplot2::ggsave(volcano.plot, bg = "white",
                filename =
                  paste0(figure.path,
                         "Toothfish_2024_pool_fin_RANDOM_DMS25_VOLCANO_methylkit.png"), 
                width = 30, height = 15, units = "cm")
```


### Heatmap

```{r}
print(load(paste0(rdata.path,"Toothfish_2024_pool_fin_methylation_RANDOM_DMS.Rdata")))

myDiff25p.gr <- as(myDiff25p, "GRanges")
meth.diff <- methylKit::selectByOverlap(meth.2024poolF.norm, myDiff25p.gr)

meth.diff.df <- as.data.frame(t(methylKit::percMethylation(meth.diff)))
loc.names <- paste0(meth.diff$chr,"_",meth.diff$start)
colnames(meth.diff.df) <- loc.names

meth.diff.df$individual <- rownames(meth.diff.df)
meth.diff.df$pop <- strata.random
meth.diff.df$pop2 <- as.character(metadata.new$pop)


heat.plot.data <- meth.diff.df %>%
  tibble::as_tibble() %>%
  tidyr::gather(key = "CpG", value = "Methylation_percentage", -individual, -pop, -pop2,
                factor_key = FALSE) %>%
  dplyr::group_by(pop, CpG) %>%
  dplyr::mutate(CpG_mean = mean(Methylation_percentage, na.rm = TRUE)) %>%
  dplyr::arrange(pop, individual, CpG) %>%
  dplyr::mutate(individual = factor(individual, levels = unique(individual))) %>%
  dplyr::mutate(CpG = factor(CpG, levels = unique(CpG)))
cpgs <- heat.plot.data$CpG[heat.plot.data$individual == unique(heat.plot.data$individual)[1]]
cpgs_mean <- heat.plot.data$CpG_mean[heat.plot.data$individual == unique(heat.plot.data$individual)[1]]

cpgs <- cpgs[order(cpgs_mean, decreasing = TRUE)]
heat.plot.data$CpG <- factor(heat.plot.data$CpG, levels = c("pop", unique(as.character(cpgs))))
heat.plot.data <- heat.plot.data[order(heat.plot.data$CpG),]

heat.plot <- ggplot2::ggplot(data = heat.plot.data, ggplot2::aes(x = individual)) + 
    ggplot2::geom_tile(data = heat.plot.data, 
                       ggplot2::aes(x = individual, y = CpG, fill = Methylation_percentage)) +
    ggplot2::scale_fill_gradient2(low = "blue", mid = "yellow", high = "red", 
                                  midpoint = 50, guide = "colourbar") +
    ggnewscale::new_scale_fill() +
    ggplot2::geom_tile(data = heat.plot.data, ggplot2::aes(x = individual, 
                                                           y = "pop - random", fill = pop,
                                    height = length(unique(heat.plot.data$CpG))*0.02)) +
    ggplot2::scale_fill_manual(values = colours2) +
    ggnewscale::new_scale_fill() +
    ggplot2::geom_tile(data = heat.plot.data, ggplot2::aes(x = individual, y = "pop", fill = pop2,
                                    height = length(unique(heat.plot.data$CpG))*0.02)) +
    ggplot2::scale_fill_manual(values = colours2) +
    ggplot2::scale_y_discrete(drop = FALSE) +
    ggplot2::theme(legend.position = "top",
                   axis.title.y = ggplot2::element_blank(),
                   axis.title.x = ggplot2::element_blank(),
                   axis.text.x = ggplot2::element_text(face = "bold", color = "black", 
                                        size = 10, angle = 90, hjust = 1, vjust = 0.5))
print(heat.plot)

ggplot2::ggsave(heat.plot, bg = "white",
                filename = paste0(figure.path,
                                  "Toothfish_2024_pool_fin_RANDOM_methDiff_25_HEATMAP.png"), 
                width = 30, height = 30, units = "cm")
```

### Plot CpG difference

```{r}
load(file = paste0(rdata.path,"Toothfish_2024_pool_fin_methylation_RANDOM_DMS.Rdata"))

myDiff25p.gr <- as(myDiff25p, "GRanges")
meth.diff <- methylKit::selectByOverlap(meth.2024poolF.norm, myDiff25p.gr) #113 

length(unique(meth.diff$chr)) #34
sort(table(meth.diff$chr), decreasing = TRUE)[1:20]

meth.diff.df <- as.data.frame(t(methylKit::percMethylation(meth.diff)))
loc.names <- paste0(meth.diff$chr,"_",meth.diff$start)
colnames(meth.diff.df) <- loc.names

meth.diff.df$individual <- rownames(meth.diff.df) 
identical(metadata.new$Genotype_names, meth.diff.df$individual)
meth.diff.df$pop <- metadata.new$pop


meth.diff.df.plot <- tidyr::gather(meth.diff.df, key = locus, 
                           value = perc_methylation, -individual, -pop,
                           factor_key = TRUE) %>%
  dplyr::mutate(chr = stringr::str_split_fixed(locus, 
                                               pattern = "_", 
                                               n = 2)[,1])
meth.diff.df.plot.locus <- meth.diff.df.plot %>% dplyr::group_by(chr,locus,pop) %>% 
  dplyr::summarise(mean_pop = mean(perc_methylation, na.rm = TRUE),
                   sum_NA = sum(is.na(perc_methylation)),
            .groups = 'drop')


diff.plot <- ggplot2::ggplot() +
  ggplot2::geom_line(data = meth.diff.df.plot.locus,
                     ggplot2::aes(x = pop, y = mean_pop, colour = locus, group = locus)) +
  ggplot2::geom_point(data = meth.diff.df.plot.locus,
                     ggplot2::aes(x = pop, y = mean_pop, colour = locus, group = locus),
                     size = 5) +
  ggplot2::facet_wrap(~chr) +
  ggplot2::ylab("Mean methylation (%)") +
  ggplot2::theme(legend.position = "none",
                   axis.text.x = ggplot2::element_text(face = "bold", color = "black", 
                                        size = 10, angle = 90, hjust = 1, vjust = 0.5))
    # ggplot2::guides(colour = ggplot2::guide_legend(ncol = 2))

print(diff.plot)

ggplot2::ggsave(plot = diff.plot, 
                filename = paste0(figure.path,"1_Toothfish_2024_pool_fin_RANDOM_methDiff.png"),
                width = 30, height = 15, units = "cm", device = "png")
```


### DMS Clustering
```{r}
load(file = paste0(rdata.path,"Toothfish_2024_pool_fin_methylation_RANDOM_DMS.Rdata"))

myDiff25p.gr <- as(myDiff25p, "GRanges")
meth.diff <- methylKit::selectByOverlap(meth.2024poolF.norm, myDiff25p.gr) #270  
methdif.perc <- methylKit::percMethylation(meth.diff)

methdif.perc2 <- zoo::na.aggregate(methdif.perc, by = meth.2024poolF.norm@treatment)
PCobj <- prcomp(methdif.perc2, retx = TRUE, scale. = TRUE) # Too many missing values

var_frac <- PCobj$sdev/sum(PCobj$sdev)
print(paste0("Total variance explained: ",signif(sum(var_frac[1:3]) * 100, 3), "%")) 
barplot(var_frac[1:3], main = "variance explained", 
        col = heat.colors(length(var_frac[1:3])))


identical(metadata.new$Genotype_names, meth.2024poolF.norm@sample.ids)

identical(as.numeric(factor(metadata.new$pop)), meth.2024poolF.norm@treatment)

# identical(metadata.new$individual, rownames(PCobj$x))
identical(metadata.new$Genotype_names, rownames(PCobj$rotation))
pop <- metadata.new$pop

plot.title <- "Toothfish methylation - RANDOMISED 2024 pool_muscle (24 DMS sites - 25%)"

# data <- data.frame(ind = rownames(PCobj$x),PCobj$x,  pop = pop)
data <- data.frame(ind = rownames(PCobj$rotation),PCobj$rotation,  pop = pop)
data <- merge(data, aggregate(cbind(mean.x = PC1, mean.y = PC2) ~
                                pop, data, mean), by = "pop")

pca.plot <- ggplot2::ggplot(data, ggplot2::aes(PC1, PC2, color = pop)) +
  ggplot2::geom_point(size = 4) + 
  ggplot2::geom_point(ggplot2::aes(x = mean.x,
                                   y = mean.y, color = pop),
                      size = 7,
                      shape = 17) +
  ggplot2::scale_colour_manual(values = colours2) +
  ggplot2::geom_hline(yintercept = 0) +
  ggplot2::geom_vline(xintercept = 0) + 
  ggplot2::labs(
    subtitle = paste0("Total variance explained: ",
                      signif(sum(var_frac[1:3]) * 100, 3), "%"),
    y = paste0("PC2; ", signif(var_frac[2] * 100, digits = 2), " %"),
    x = paste0("PC1; ", signif(var_frac[1] * 100, digits = 2), " %"),
    title = plot.title,
    caption = "") + 
  ggplot2::theme_minimal() +
  ggplot2::theme(
    axis.text = ggplot2::element_text(size = 15),
    axis.title.x = ggplot2::element_text(size = 20),
    axis.title.y = ggplot2::element_text(size = 20),
    legend.text = ggplot2::element_text(size = 15)
  )

print(pca.plot)
ggplot2::ggsave(pca.plot, bg = "white",
                filename = 
                  paste0(figure.path,"Toothfish_2024_pool_fin_RANDOM_PCA_CpG_DMS_PC12.png"), 
                width = 30, height = 15, units = "cm")



# data <- data.frame(ind = rownames(PCobj$x),PCobj$x,  pop = pop)
data <- data.frame(ind = rownames(PCobj$rotation),PCobj$rotation,  pop = pop)
data <- merge(data, aggregate(cbind(mean.x = PC1, mean.y = PC3) ~
                                pop, data, mean), by = "pop")

pca.plot <- ggplot2::ggplot(data, ggplot2::aes(PC1, PC3, color = pop)) +
  ggplot2::geom_point(size = 4) + 
  ggplot2::geom_point(ggplot2::aes(x = mean.x,
                                   y = mean.y, color = pop),
                      size = 7,
                      shape = 17) +
  ggplot2::scale_colour_manual(values = colours2) +
  ggplot2::geom_hline(yintercept = 0) +
  ggplot2::geom_vline(xintercept = 0) + 
  ggplot2::labs(
    subtitle = paste0("Total variance explained: ",
                      signif(sum(var_frac[1:3]) * 100, 3), "%"),
    y = paste0("PC3; ", signif(var_frac[3] * 100, digits = 2), " %"),
    x = paste0("PC1; ", signif(var_frac[1] * 100, digits = 2), " %"),
    title = plot.title,
    caption = "") + 
  ggplot2::theme_minimal() +
  ggplot2::theme(
    axis.text = ggplot2::element_text(size = 15),
    axis.title.x = ggplot2::element_text(size = 20),
    axis.title.y = ggplot2::element_text(size = 20),
    legend.text = ggplot2::element_text(size = 15)
  )

print(pca.plot)
ggplot2::ggsave(pca.plot, bg = "white",
                filename = paste0(figure.path,"Toothfish_2024_pool_fin_RANDOM_PCA_CpG_DMS_PC13.png"), 
                width = 30, height = 15, units = "cm")




# data <- data.frame(ind = rownames(PCobj$x),PCobj$x,  pop = pop)
data <- data.frame(ind = rownames(PCobj$rotation),PCobj$rotation,  pop = pop)
data <- merge(data, aggregate(cbind(mean.x = PC2, mean.y = PC3) ~
                                pop, data, mean), by = "pop")

pca.plot <- ggplot2::ggplot(data, ggplot2::aes(PC2, PC3, color = pop)) +
  ggplot2::geom_point(size = 4) + 
  ggplot2::geom_point(ggplot2::aes(x = mean.x,
                                   y = mean.y, color = pop),
                      size = 7,
                      shape = 17) +
  ggplot2::scale_colour_manual(values = colours2) +
  ggplot2::geom_hline(yintercept = 0) +
  ggplot2::geom_vline(xintercept = 0) + 
  ggplot2::labs(
    subtitle = paste0("Total variance explained: ",
                      signif(sum(var_frac[1:3]) * 100, 3), "%"),
    y = paste0("PC3; ", signif(var_frac[3] * 100, digits = 2), " %"),
    x = paste0("PC2; ", signif(var_frac[2] * 100, digits = 2), " %"),
    title = plot.title,
    caption = "") + 
  ggplot2::theme_minimal() +
  ggplot2::theme(
    axis.text = ggplot2::element_text(size = 15),
    axis.title.x = ggplot2::element_text(size = 20),
    axis.title.y = ggplot2::element_text(size = 20),
    legend.text = ggplot2::element_text(size = 15)
  )

print(pca.plot)
ggplot2::ggsave(pca.plot, bg = "white",
                filename = 
                  paste0(figure.path,"Toothfish_2024_pool_fin_RANDOM_PCA_CpG_DMS_PC23.png"), 
                width = 30, height = 15, units = "cm")
```





# 13_Methylation - Tissue comparison ALL samples

```{r}
print(load(paste0(rdata.path,"Toothfish_metadata.Rdata"))) #ALL fish
```

## Reading the methylation call files

```{r, eval=FALSE}
data.list <- list.files(
  path = 
  "/datasets/work/ncmi-toa-rrbs/work/1.Toothfish/GENOME/Combined_methylation/",
  pattern = "*.cov.gz")

subnames <- metadata$File_names[!(grepl(pattern = "REP", x = metadata$File_names) | 
                                  grepl(pattern = "muscle from under SD|Skin",
                                         x = metadata$Tissue_type) |
                                    !grepl(pattern = "TOA", x = metadata$File_names)
                                  )]
subnames <- subnames[! subnames %in% c("TOA_88-2_P79_AD294619_fin_F")]
data.list <- data.list[data.list %in% paste0(subnames,
                                             "_R1_trimmed_sort_bismark_bt2.bismark.cov.gz")]

data.loc <- as.list(paste0(
  "/datasets/work/ncmi-toa-rrbs/work/1.Toothfish/GENOME/Combined_methylation/",
                           data.list))



data.list2 <-  gsub(pattern = '_R1_trimmed_sort_bismark_bt2.bismark.cov.gz',
                    replacement = '',x = as.character(data.list))
# data.list2 <-  gsub(pattern = '_sort_bismark_bt2_PE.bismark.cov.gz',
#                     replacement = '',x = as.character(data.list2))

metadata.new <- metadata[metadata$File_names %in% data.list2,]
metadata.new <- metadata.new[order(match(metadata.new$File_names, data.list2)),]

metadata.new$File_names == data.list2

sample.ids <- metadata.new$Genotype_names
pop <- as.integer(factor(metadata.new$Tissue_type))


myobjDB <- methylKit::methRead(location = data.loc,
           sample.id = as.list(sample.ids),
           assembly = "TOA_genome",
           treatment = pop,
           context = "CpG",
           header = FALSE, 
           dbtype = "tabix",
           pipeline = "bismarkCoverage",
           mincov = 3
           )
save(myobjDB, file = paste0(rdata.path,"Toothfish_tissue_comparison_methylkit_cpgDB.Rdata"))
```


## Data QC

#### Number of sites in genome

```{r, eval=FALSE}

cov.df <- data.frame(chr = NA, start = NA,end = NA, strand = NA, coverage = NA, 
                     numCs = NA, numTs = NA, sample = NA)
for (i in 1:length(sample.ids)) {
  df <- methylKit::getData(myobjDB[[i]])
  df$sample <- sample.ids[i]
  cov.df <- rbind(cov.df, df)
}
cov.df <- cov.df[-1,]
save(cov.df, file = paste0(rdata.path,"Toothfish_tissue_comparison_methylkit_cpg_cov.Rdata"))
```

### Coverage stats

```{r}
load(paste0(rdata.path,"Toothfish_tissue_comparison_methylkit_cpgDB.Rdata"))
print(load(paste0(rdata.path,"Toothfish_tissue_comparison_methylkit_cpg_cov.Rdata")))

length(unique(paste0(cov.df$chr, "_", cov.df$start))) #5,121,570

# summary(as.factor(cov.df$sample))
summary(cov.df$coverage)
    # Min.  1st Qu.   Median     Mean  3rd Qu.     Max. 
    # 4.00     7.00    13.00    20.09    24.00 51120.00 

summary(cov.df$coverage[cov.df$coverage > 5 & cov.df$coverage < 50])
   # Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
   # 6.00    9.00   14.00   17.23   23.00   49.00 

cov.plot <- ggplot2::ggplot(cov.df, ggplot2::aes(x = log(coverage), fill = sample)) +
  ggplot2::geom_histogram() +
  ggplot2::theme(legend.position = "none")
# print(cov.plot)
ggplot2::ggsave(cov.plot,
                filename =
                  paste0(figure.path,"Toothfish_tissue_comparison_coverage_hist_Combined.png"), 
                width = 45, height = 15, units = 'cm')

cov.plot <- ggplot2::ggplot(cov.df, ggplot2::aes(x = coverage, fill = sample)) +
  ggplot2::geom_histogram() +
  ggplot2::theme(legend.position = "none")
# print(cov.plot)
ggplot2::ggsave(cov.plot,
                filename =
                  paste0(figure.path,"Toothfish_tissue_comparison_coverage_hist_Combined2.png"), 
                width = 45, height = 15, units = 'cm')

cov.plot <- ggplot2::ggplot(cov.df[cov.df$coverage > 5 & cov.df$coverage < 160,], 
                            ggplot2::aes(x = coverage)) +
  ggplot2::geom_histogram(binwidth = 1) +
  ggplot2::ggtitle("coverage < 160") +
  ggplot2::theme(legend.position = "none")
# print(cov.plot)
ggplot2::ggsave(cov.plot,
                filename =
                  paste0(figure.path,"Toothfish_tissue_comparison_coverage_hist_Combined4.png"),
                width = 45, height = 15, units = 'cm')



max(cov.df$coverage, na.rm = T) #14582
ind.covplot <- ggplot2::ggplot(cov.df, ggplot2::aes(x = sample, y = coverage)) + 
  ggplot2::geom_boxplot() +
  ggplot2::theme_bw() + 
  ggplot2::theme(axis.title.x = ggplot2::element_blank(), 
                 axis.text.y = ggplot2::element_text(size = 15),
                 axis.title.y = ggplot2::element_text(size = 15),
                 axis.text.x = ggplot2::element_text(size = 7, angle = 90, hjust = 1)) +
  ggplot2::scale_y_continuous(trans = scales::log2_trans(),
                  breaks = c(1, 10, 50, 750, 20000),
                  minor_breaks = c(1:10, 2:5*10, 1:7.5*100, 0.5:2*1000))

print(ind.covplot)

ggplot2::ggsave(ind.covplot, bg = "white",
                filename = 
                  paste0(figure.path,
                         "Toothfish_tissue_comparison_raw_individual_coverage_CpGs.png"),
                width = 45, height = 15, units = 'cm')
```


### Coverage filter

High count filter based on coverage stats.

```{r, eval=FALSE}
load(paste0(rdata.path,"Toothfish_tissue_comparison_methylkit_cpgDB.Rdata"))

filtered.myobj <- methylKit::filterByCoverage(myobjDB, lo.count = 10, lo.perc = NULL,
                                      hi.count = NULL, hi.perc = 99)
save(myobjDB,filtered.myobj, 
     file = paste0(rdata.path,"Toothfish_tissue_comparison_methylkit_cpgDB.Rdata"))
```

### Normalise coverage

Normalize read coverage distributions between samples. Requires stringent filter on max coverage first, otherwise it will over-sample PCR bias reads. These two functions will help reduce the bias in the statistical tests that might occur due to systematic over-sampling of reads in certain samples.

```{r}
filtered.myobj.normal <- methylKit::normalizeCoverage(obj = filtered.myobj, 
                                                      method = "median", chunk.size = 1e6)
save(myobjDB,filtered.myobj, filtered.myobj.normal,
     file = paste0(rdata.path,"Toothfish_tissue_comparison_methylkit_cpgDB.Rdata"))
```

### Number of sites per sample

```{r, eval = FALSE}
sites.df <- data.frame(sites = NA,  sample = NA)
for (i in 1:length(filtered.myobj.normal)) {
  print(i)
  sites.df[i,] <- c(filtered.myobj.normal[[i]]@num.records, filtered.myobj.normal[[i]]@sample.id)
}
sites.df$sites <- as.integer(sites.df$sites)
# sites.df$pop <- filtered.myobj.normal@treatment
sites.df$pop <- metadata.new$pop

save(sites.df, file = paste0(rdata.path,"Toothfish_tissue_comparison_number_cpg_sample.Rdata"))
```


```{r}
print(load(paste0(rdata.path,"Toothfish_tissue_comparison_number_cpg_sample.Rdata")))
sites.plot <- ggplot2::ggplot(sites.df, ggplot2::aes(x = sample, y = sites, colour = pop)) +
  ggplot2::geom_point() +
  ggplot2::scale_color_manual(values = colours.all) +
  ggplot2::theme_bw() +
  ggplot2::theme(axis.text.x = ggplot2::element_text(angle = 90))
print(sites.plot)

ggplot2::ggsave(sites.plot,
                filename =
                  paste0(figure.path,"Toothfish_tissue_comparison_number_cpg_sample.png"),
                width = 45, height = 15, units = 'cm')

sites.df$sample[sites.df$sites < 500000]
rm2 <- sites.df$sample[sites.df$sites < 500000]

# rm2 <- c("TOA_48-2_007_UK048","TOA_48-2_028_UK103","TOA_58-4-2_088_AN170741",
#          "TOA_58-4-2_B14_AE100306_muscle_F", "TOA_58-4-2_B18_AE100308_muscle_F",
#          "TOA_58-4-2_B24_AE100883_muscle_F", "TOA_58-4-2_B28_AE100885_muscle_M",
#          "TOA_58-4-2_B30_AE100886_muscle_M", "TOA_58-4-2_B31_AE102045_fin_F",
#          "TOA_58-4-2_B34_AE102049_muscle_F", "TOA_58-4-2_B36_AE102050_muscle_M", 
#          "TOA_58-4-2_B38_AE102052_muscle_M","TOA_58-4-2_B42_AE102105_muscle_F",
#          "TOA_58-4-2_B48_AE102108_muscle_M", "TOA_58-4-2_B55_AE104943_fin_M",
#          "TOA_58-4-2_B58_AE104944_muscle_F", "TOA_58-4-2_B63_AE107905_fin_F",
#          "TOA_58-4-2_B64_AE107905_muscle_F", "TOA_58-4-2_B78_AE109499_muscle_M",
#          "TOA_58-4-2_B79_AE109500_fin_M","TOA_88-2_P24_AD292328_muscle_F",
#          "TOA_88-2_P52_AD293511_muscle_M","TOA_88-2_P59_AD293515_fin_F",
#          "TOA_88-2_P67_AD294613_fin_F", "TOA_88-2_P69_AD294614_fin_F",
#          "TOA_88-2_P73_AD294616_fin_M")
```


## Select samples

```{r}
print(load(paste0(rdata.path,"Toothfish_metadata.Rdata")))

metadata2 <- metadata[!(grepl(pattern = "REP", x = metadata$File_names) | 
                                  grepl(pattern = "muscle from under SD|Skin",
                                         x = metadata$Tissue_type) |
                                    !grepl(pattern = "TOA", x = metadata$File_names)
                                  ),]

print(load(paste0(rdata.path, "Low_quality_fish.Rdata")))

bad.list <- list(high.lambda.mapping,
          low.genome.mapping,low.genome.methylation,
          low.mbias.methylation)
bad.plot <- ggVennDiagram::ggVennDiagram(
  x = bad.list, label_alpha = 0, set_color = c("black","black","black","black"),
  set_size = 7, label_size = 5,
  category.names = c("high.lambda.mapping",
          "low.genome.mapping","low.genome.methylation",
          "low.mbias.methylation")) +
  ggplot2::scale_fill_gradient2(low = "blue", mid = "yellow",high = "red", midpoint = 150) + 
  ggplot2::scale_color_manual(values = c("black","black","black","black")) +
    ggplot2::theme(
    legend.text = ggplot2::element_text(size = 20),
    legend.title = ggplot2::element_text(size = 30)
  )
print(bad.plot)

# rm1 <- Reduce(intersect, list(high.lambda.mapping,
#           low.genome.mapping,low.genome.methylation,
#           low.mbias.methylation))
rm1 <- Reduce(intersect, list(low.genome.mapping,low.genome.methylation))
print(rm1) #bad mapping & low methylation
print(rm2) # low number of cpg sites
rm3 <- "TOA_88-2_P79_AD294619_fin_F"
print(rm3) # bad coverage
rm4 <- c("TOA_48-2_031_UK116","TOA_58-4-2_111_AN172604","TOA_88-1_230_AD082543",
         "TOA_48-2_024_UK082")
print(rm4) #bad tissue PCA

rm <- c(rm, rm2,rm3, rm4) # bad coverage

metadata2 <- metadata2[!metadata2$File_names %in% rm,]

filt.sample.id.full <- sapply(filtered.myobj@.Data, function(x){x@sample.id})

# filt.sample.id <- filt.sample.id.full[filt.sample.id.full %in% metadata2$File_names]
# metadata2 <- metadata2[order(match(metadata.new$File_names, filt.sample.id)),]

filt.sample.id <- filt.sample.id.full[filt.sample.id.full %in% metadata2$Genotype_names]
metadata2 <- metadata2[order(match(metadata2$Genotype_names, filt.sample.id)),]
identical(filt.sample.id,metadata2$Genotype_names)


pop <- metadata2$Tissue_type
# pop.levels <- c("48.2-2017","58.4.1-2018", "58.4.2-2018", "58.4.2-2024", 
#                 "58.5.2-2017", "88.1-2018", "88.2-2024")

pop <- as.numeric(factor(pop))
sample.ids <- metadata2$Genotype_names



# filtered.myobj2 <- methylKit::reorganize(
#   methylObj = filtered.myobj,
#   sample.ids = sample.ids,
#   treatment = pop,
#   chunk.size = 1e+06,
#   save.db = TRUE,
#   suffix = "TOA_tissue_comparison"
# )

filtered.myobj2.normal <- methylKit::reorganize(
  methylObj = filtered.myobj.normal,
  sample.ids = sample.ids,
  treatment = pop,
  chunk.size = 1e+06,
  save.db = TRUE,
  suffix = "TOA_tissue_comparison_normal"
)
save(myobjDB,filtered.myobj,#filtered.myobj2, 
     filtered.myobj.normal,filtered.myobj2.normal,
     file = paste0(rdata.path,"Toothfish_tissue_comparison_methylkit_cpgDB.Rdata"))
```


## Merging samples

```{r}
print(load(paste0(rdata.path,"Toothfish_tissue_comparison_methylkit_cpgDB.Rdata")))
pop <- metadata2$Tissue_type
pop.levels <- c("fin","muscle")
pop <- as.numeric(factor(pop, levels = pop.levels))
filtered.myobj2.normal@treatment <- pop
table(pop)


# meth.2024M15L <- methylKit::unite(filtered.myobj2, destrand = FALSE,
#                              min.per.group = 15L, #missing data allowed
#                              mc.cores = 25)


meth.tissue.norm50L <- methylKit::unite(filtered.myobj2.normal, destrand = FALSE,
                             min.per.group = 50L, #missing data allowed
                             save.db = FALSE,
                             mc.cores = 25)
# meth.tissue.norm50L@num.records
#432227 common sites for 50 samples per group

# meth.2024M.norm <- as(meth.2024M.norm,"methylBase")
# meth.tissue.norm50L <- as(meth.tissue.norm50L,"methylBase")

save(meth.tissue.norm50L, #meth.2018M.norm,meth.2024M,meth.2024M15L,
     file = paste0(rdata.path,"Toothfish_tissue_comparison_methylkit_object.Rdata"))
```


## Data QC


### Remove ASM
```{r}
load(paste0(rdata.path,"Allele_specific_mutations.Rdata"))

# example SNP
mut <- as(asmTRUE,"GRanges")
# meth.tissue.norm50L <- as(meth.tissue.norm50L,"methylBase") 


sum(as(meth.tissue.norm50L,"GRanges") %over% mut)#13411
meth.asm <- methylKit::selectByOverlap(meth.tissue.norm50L, mut) #411698 
BOOLEAN <- paste0(meth.tissue.norm50L$chr,"_",meth.tissue.norm50L$start) %in%
  paste0(asmTRUE$Chr,"_",asmTRUE$Pos) #1
sum(BOOLEAN) #97
meth.tissue.norm50L <- methylKit::select(meth.tissue.norm50L, !BOOLEAN) # 432130   

save(meth.tissue.norm50L,
     file = paste0(rdata.path,"Toothfish_tissue_comparison_methylkit_object.Rdata"))
```


### Coverage 

```{r}
meth.df <- methylKit::getData(meth.tissue.norm50L)
cov <- meth.df[,meth.tissue.norm50L@coverage.index]
colnames(cov) <- meth.tissue.norm50L@sample.ids
dpf <- data.frame(cov, locus = paste0(meth.df$chr,"_",meth.df$start))
dpf <- tidyr::gather(dpf, key = individual, 
                           value = Depth, -locus,
                           factor_key = TRUE) %>%
  dplyr::group_by(individual) %>%
  dplyr::mutate(NA_cnt = sum(is.na(Depth))/length(unique(locus))) %>%
  dplyr::ungroup()
dpf.NA <- dpf[!duplicated(dpf$individual),]
dpf.NA$NA_cnt <- round(dpf.NA$NA_cnt, 3)

dpf <- dpf[ dpf$Depth > 0 & !is.na(dpf$Depth),]

max(dpf$Depth, na.rm = T) #99
myPlot <- ggplot2::ggplot(dpf, ggplot2::aes(x = individual, y = Depth)) + 
  ggplot2::geom_boxplot() +
  ggplot2::geom_text(data = dpf.NA,ggplot2::aes(x = individual, y = 150, label = NA_cnt), 
                     size = 2, angle = 90) +
  ggplot2::theme_bw() + 
  ggplot2::theme(axis.title.x = ggplot2::element_blank(), 
                 axis.text.y = ggplot2::element_text(size = 15),
                 axis.title.y = ggplot2::element_text(size = 15),
                 axis.text.x = ggplot2::element_text(size = 7, angle = 90, hjust = 1)) +
  ggplot2::scale_y_continuous(trans = scales::log2_trans(),
                  breaks = c(1, 10, 50, 100),
                  minor_breaks = c(1:10, 2:5*10))

print(myPlot)
ggplot2::ggsave(myPlot, bg = "white",
                filename = 
                  paste0(figure.path,
                  "1B.Toothfish_tissue_comparison_normalised50_methylkit_individual_coverage_CpGs.png"
                                  ), 
                width = 45, height = 15, units = "cm")

nrow(cov) # 612935 SMPs
plot(x = 1:ncol(cov), y = colMeans(cov, na.rm = TRUE), col = meth.tissue.norm50L@treatment)
plot(x = 1:ncol(cov), y = colMeans(cov > 20, na.rm = TRUE), col = meth.tissue.norm50L@treatment)
plot(x = 1:ncol(cov), y = colMeans(cov > 10, na.rm = TRUE), col = meth.tissue.norm50L@treatment)
meth.tissue.norm50L@sample.ids[colMeans(cov > 10, na.rm = TRUE) < 0.88]
# samples with low coverage: "TOA_48-2_028_UK103"      "TOA_58-4-2_088_AN170741"
```



## Load data

```{r}
print(load(paste0(rdata.path,"Toothfish_tissue_comparison_methylkit_object.Rdata")))
print(load(paste0(rdata.path,"Toothfish_metadata.Rdata"))) 

metadata.new <- metadata[metadata$Genotype_names %in% meth.tissue.norm50L@sample.ids,]
metadata.new <- metadata.new[order(match(metadata.new$Genotype_names,
                                         meth.tissue.norm50L@sample.ids)),]
```


## Global clustering analysis

```{r}
methylKit::clusterSamples(meth.tissue.norm50L, dist = "correlation", sd.filter = TRUE, 
                          method = "ward.D2", plot = TRUE, )
dev.print(file = paste0(figure.path,"1_TOA_tissue_comparison_cluster_plot.png"),
          device = png, res = 300,width = 45,height = 15,units = "cm")
```


### PCA

```{r}
# methdif.perc <- methylKit::percMethylation(meth.tissue.norm50L)
# methdif.perc2 <- zoo::na.aggregate(methdif.perc, by = meth.tissue.norm50L@treatment)
# PCobj <- prcomp(methdif.perc2, retx = TRUE, scale. = TRUE)
PCobj <- methylKit::PCASamples(meth.tissue.norm50L, comp = c(1,2), 
                               adj.lim = c(0,0), obj.return = TRUE)

var_frac <- PCobj$sdev/sum(PCobj$sdev)
print(paste0("Total variance explained: ",signif(sum(var_frac[1:3]) * 100, 3), "%")) 
barplot(var_frac[1:20], main = "variance explained", 
        col = heat.colors(length(var_frac[1:20])))

identical(metadata.new$Genotype_names, meth.tissue.norm50L@sample.ids)
identical(metadata.new$Genotype_names, rownames(PCobj$x))

pop <- metadata.new$Tissue_type

plot.title <- "Toothfish methylation - tissue comparison (432,130  CpG sites)"

data <- data.frame(ind = rownames(PCobj$x),PCobj$x,  pop = pop)
# data <- data.frame(ind = rownames(PCobj$rotation),PCobj$rotation, pop = pop)
data <- merge(data, aggregate(cbind(mean.x = PC1, mean.y = PC2) ~
                                pop, data, mean), by = "pop")

pca.plot <- ggplot2::ggplot(data, ggplot2::aes(PC1, PC2, color = pop)) +
  ggplot2::geom_point(size = 4) + 
  ggplot2::geom_point(ggplot2::aes(x = mean.x,
                                   y = mean.y, color = pop),
                      size = 7,
                      shape = 17) +
  ggplot2::scale_colour_manual(values = as.character(colours2)) +
  ggplot2::geom_hline(yintercept = 0) +
  ggplot2::geom_vline(xintercept = 0) + 
  ggplot2::labs(
    subtitle = paste0("Total variance explained: ",
                      signif(sum(var_frac[1:4]) * 100, 3), "%"),
    y = paste0("PC2; ", signif(var_frac[2] * 100, digits = 2), " %"),
    x = paste0("PC1; ", signif(var_frac[1] * 100, digits = 2), " %"),
    title = plot.title,
    caption = "") + 
  ggplot2::theme_minimal() +
  ggplot2::theme(
    axis.text = ggplot2::element_text(size = 15),
    axis.title.x = ggplot2::element_text(size = 20),
    axis.title.y = ggplot2::element_text(size = 20),
    legend.text = ggplot2::element_text(size = 15)
  )

print(pca.plot)
ggplot2::ggsave(pca.plot, bg = "white",
                filename = paste0(figure.path,"Toothfish_tissue_comparison_PCA_CpG_PC12.png"), 
                width = 30, height = 15, units = "cm")


data$ind[data$PC2 > 10]
# "TOA_48-2_031_UK116"      "TOA_58-4-2_111_AN172604" "TOA_88-1_230_AD082543" 

data <- data.frame(ind = rownames(PCobj$x),PCobj$x,  pop = pop)
# data <- data.frame(ind = rownames(PCobj$rotation),PCobj$rotation, pop = pop)
data <- merge(data, aggregate(cbind(mean.x = PC2, mean.y = PC3) ~
                                pop, data, mean), by = "pop")
pca.plot <- ggplot2::ggplot(data, ggplot2::aes(PC2, PC3, color = pop)) +
  ggplot2::geom_point(size = 4) + 
  ggplot2::geom_point(ggplot2::aes(x = mean.x,
                                   y = mean.y, color = pop),
                      size = 7,
                      shape = 17) +
  ggplot2::scale_colour_manual(values = as.character(colours2)) +
  ggplot2::geom_hline(yintercept = 0) +
  ggplot2::geom_vline(xintercept = 0) + 
  ggplot2::labs(
    subtitle = paste0("Total variance explained: ",
                      signif(sum(var_frac[1:4]) * 100, 3), "%"),
    y = paste0("PC3; ", signif(var_frac[3] * 100, digits = 2), " %"),
    x = paste0("PC2; ", signif(var_frac[2] * 100, digits = 2), " %"),
    title = plot.title,
    caption = "") + 
  ggplot2::theme_minimal() +
  ggplot2::theme(
    axis.text = ggplot2::element_text(size = 15),
    axis.title.x = ggplot2::element_text(size = 20),
    axis.title.y = ggplot2::element_text(size = 20),
    legend.text = ggplot2::element_text(size = 15)
  )

print(pca.plot)
ggplot2::ggsave(pca.plot, bg = "white",
                filename = paste0(figure.path,"Toothfish_tissue_comparison_PCA_CpG_PC23.png"), 
                width = 30, height = 15, units = "cm")


data$ind[data$PC3 > 20]
# "TOA_48-2_024_UK082"
```


### RDA

```{r}
# setwd("/datasets/work/ncmi-toa-rrbs/work/1.Toothfish/Toothfish_RRBS_array_22098078/Bismark/")
# print(load(paste0(rdata.path,"Toothfish_methylation_objects_raw.Rdata"))) #36,635 SMPs
print(load(paste0(rdata.path,"Toothfish_subset_methylation_objects_raw.Rdata"))) #1,316,799 SMPs

# meth.RDA <- methylKit::unite(filtered.myobj2.norm, destrand = FALSE,
#                              save.db = FALSE,
#                              suffix = "subset",
#                              min.per.group = NULL, #missing data allowed
#                              mc.cores = 25)


# meth.df <- methylKit::getData(meth)
# cov <- meth.df[,grep(colnames(meth.df), pattern = "coverage*")]
# cov <- meth.df[,meth@coverage.index]
# colnames(cov) <- meth@sample.ids
# meth.df <- methylKit::percMethylation(meth)

# meth.df <- methylKit::percMethylation(meth.RDA)

# save(meth.df,meth.RDA, 
#      file = paste0(rdata.path,"TOA_data_for_RDA_meth_percent.Rdata"))
print(load(paste0(rdata.path,"TOA_data_for_RDA_meth_percent.Rdata"))) #347,614 SMPs

hist(colSums(is.na(meth.df)), breaks = 100)
hist(rowSums(is.na(meth.df)), breaks = 100)





print(load(paste0(rdata.path,"Toothfish_metadata.Rdata")))

predictor <- metadata2 # define the predictor input file here
predictor <- predictor[predictor$SAMPLE_ID %in% meth@sample.ids,]
predictor$SAMPLE_ID = as.character(predictor$SAMPLE_ID)
predictor$region = as.character(predictor$region)
predictor$Sex = as.character(predictor$Sex)
predictor$Tissue = as.character(predictor$Tissue)
predictor$mat = as.character(predictor$mat)
predictor$mat[is.na(predictor$mat)] <- "2"
predictor$TL = as.numeric(predictor$TL)
predictor$wt = as.numeric(predictor$wt)
predictor$dpth = as.numeric(predictor$dpth)
predictor$Age = as.numeric(predictor$Age)
predictor$Age[is.na(predictor$Age)] <- as.numeric(max(predictor$Age, na.rm = TRUE))
predictor$Cohort2 = as.integer(predictor$Cohort2)
predictor$Cohort2[is.na(predictor$Cohort2)] <- as.integer(min(predictor$Cohort2, na.rm = TRUE))


meth.df1 = t(meth.df)
# meth.df2 = na.omit(meth.df1) #rda can't handle NAs, remove them if you have them
# meth.df1[1:10,1:10]
## Predictor
# Make names characters (not factors)


#if you had numbers, they had to be factors (as.factor)

# Confirm that methylation data and predictors are in the same order
identical(rownames(meth.df1), predictor$SAMPLE_ID)

# Perform Hellinger transformation on the methylation data:
meth.df3 = vegan::decostand(meth.df1, method = "hellinger", MARGIN = 1) #MARGIN=1 means rows, on rows I have samples now because I transposed the matrix

## Run RDA
#Test the country using growing condition as a Condition () (covariate):
meth_rda = vegan::rda(meth.df3 ~ region + Condition(Sex), 
                      data = predictor, scale = F) #I remove scale since I performed Hellinger transformation
meth_rda = vegan::rda(meth.df3 ~ region + Condition(Sex) + 
                        Condition(mat)+ Condition(TL) + Condition(wt) +
                        Condition(dpth) + Condition(Age) + Condition(Cohort2), 
                      na.action = "na.omit",
                      data = predictor, scale = F)

meth_rda$pCCA
meth_rda$CCA$wa

summary(meth_rda) #Constrained Proportion: variance of Y explained by X, Unconstrained Proportion: unexplained variance in Y
# sink(paste0(rdata.path,"anova_rda.txt")) #with this you can save the output of the anova in a txt file, define the output directory
meth.aov <- anova(meth_rda, test = "Chi", permutations = 99) #run Anova on the result to see the significance level
# meth.aov.cca <- anova.cca(meth_rda, test = "Chi", permutations = 99)
print(meth.aov)
# sink(paste0(rdata.path,"anova_rda_chi.txt"))

# Model: rda(formula = meth.df3 ~ region + Condition(Sex), data = predictor, scale = F)
#           Df  Variance      F Pr(>F)   
# Model      3 0.0003924 1.4126  0.004 **
# Residual 110 0.0101861

# Model: rda(formula = meth.df3 ~ region + Condition(Sex) + Condition(mat) + Condition(TL) + Condition(wt) + Condition(dpth) + Condition(Age) + Condition(Cohort2), data = predictor, scale = F, na.action = "na.omit")
#           Df  Variance     F Pr(>F)   
# Model      3 0.0003412 1.257   0.01 **
# Residual 101 0.0091396 

Pop.names <- predictor$region
plot.a <- vegan::ordiplot(meth_rda, choices=c(1,2))
Pop.ellipses <- vegan::ordiellipse(plot.a, groups=Pop.names, display="sites", kind="sd")
dev.print(file = paste0(figure.path,"TOA_RDA_ordiplot_complex_model.png"),
          device = png, res = 300,width = 45,height = 15,units = "cm")

# Type 1 scaling
vegan::ordiplot(meth_rda, scaling = 1, type = "text")
# Type 2 scaling
vegan::ordiplot(meth_rda, scaling = 2, type = "text")

ggvegan::autoplot(
  meth_rda,
  axes = c(1, 2),
  geom = "text",
  layers = c("species", "sites", "biplot", "centroids"),
  arrows = TRUE)

#CpG contribution
hist(meth_rda$CCA$v[,1])
hist(meth_rda$CCA$v[,2])
hist(meth_rda$CCA$v[,3])



#Test the interaction of your predictors:
meth_rda3 = vegan::rda(meth ~ region * Sex + Condition(region + Sex), data = predictor, scale = F)
summary(meth_rda3)
sink(paste0(rdata.path,"anova_rda_interaction.txt"))
print(anova(meth_rda3, test="Chi"))
sink(paste0(rdata.path,"anova_rda_interaction_chi.txt"))
```

#### plot RDA

```{r}

BioR.theme <- theme(
        panel.background = element_blank(),
        panel.border = element_blank(),
        panel.grid = element_blank(),
        axis.line = element_line("gray25"),
        text = element_text(size = 12),
        axis.text = element_text(size = 10, colour = "gray25"),
        axis.title = element_text(size = 14, colour = "gray25"),
        legend.title = element_text(size = 14),
        legend.text = element_text(size = 14),
        legend.key = element_blank())

plot.a <- vegan::ordiplot(meth_rda, choices=c(1,2))
Pop.names <- predictor$region

# sites1 <- BiodiversityR::sites.long(plot.a, env.data=predictor)
sites1 <- data.frame(cbind(predictor, axis1=plot.a$sites[, 1], 
                           axis2=plot.a$sites[, 2], 
                           labels=rownames(plot.a$sites)),
                     group = predictor$region)

# axis1 <- BiodiversityR::axis.long(meth_rda, choices=c(1, 2))
eigs.all <- meth_rda$CA$eig
eigs <- round(100 * eigs.all / meth_rda$tot.chi, digits=1)[c(1, 2)] 
xlab.label <- paste0(names(eigs)[1], " (", eigs[1], "%)")
ylab.label <- paste0(names(eigs)[2], " (", eigs[2], "%)")
axis1 <- data.frame(axis=c(1:2), 
                    ggplot=c("xlab.label", "ylab.label"), 
                    label=c(xlab.label, ylab.label))


# centroids.Pop1 <- BiodiversityR::centroids.long(sites1, 
#                                                 grouping=Pop.names, 
#                                                 centroids.only=TRUE)
# centroids.Pop2 <- BiodiversityR::centroids.long(sites1, 
#                                                 grouping=Pop.names, 
#                                                 centroids.only=FALSE)
gr.name <- rlang::enexpr(Pop.names)
cent.means <-  stats::aggregate(cbind(axis1, axis2) ~ Pop.names,
                                data = sites1,
                                FUN = mean)
names(cent.means) <- c("group", "axis1c", "axis2c")
centroids.Pop1 <- cent.means
cent.means$Centroid <- cent.means[, 1]
by.name <- names(sites1)[names(sites1) == "group"]
centroids.Pop2 <- dplyr::full_join(sites1, cent.means, by = by.name)


# Pop.ellipses.long2 <- BiodiversityR::ordiellipse.long(Pop.ellipses, 
#                                                     grouping.name="Pop.names")


plotgg.a <- ggplot() + 
    geom_vline(xintercept = c(0), color = "grey70", linetype = 2) +
    geom_hline(yintercept = c(0), color = "grey70", linetype = 2) +  
    xlab(axis1[1, "label"]) +
    ylab(axis1[2, "label"]) +  
    scale_x_continuous(sec.axis = dup_axis(labels=NULL, name=NULL)) +
    scale_y_continuous(sec.axis = dup_axis(labels=NULL, name=NULL)) +
    # geom_polygon(data=Pop.ellipses.long2, 
    #                aes(x=axis1, y=axis2, colour=Pop.names, 
    #                    fill=after_scale(alpha(colour, 0.2))), 
    #           size=0.2, show.legend=TRUE) +
    geom_segment(data=centroids.Pop2, 
                 aes(x=axis1c, y=axis2c, xend=axis1, yend=axis2, colour=Pop.names), 
                 size=0.7, show.legend=FALSE) +
    geom_point(data=centroids.Pop1, 
               aes(x=axis1c, y=axis2c), 
               shape="square", colour="black", alpha=0.7, size=2) +
    geom_label_repel(data=centroids.Pop1, 
               aes(x=axis1c, y=axis2c, label=group, colour=group), 
               alpha=1.0, size=2, show.legend=FALSE) +
    # labs(colour = Pop.names) +
    BioR.theme +
    theme(legend.text = element_text(size = 10)) +
    scale_colour_npg() +
    coord_fixed(ratio=1)

plotgg.a

ggsave(plotgg.a, file = paste0(figure.path, 
                               # "TOA_RDA_simple_model_A.png"
                               "TOA_RDA_complex_model_A.png"
                               ),
       height = 30, width = 30, units = "cm",  device = "png")
```



## DMS - Tissue comparison

```{r, eval = FALSE}
print(load(paste0(rdata.path,"Toothfish_tissue_comparison_methylkit_object.Rdata")))

print(load(paste0(rdata.path,"Toothfish_metadata.Rdata")))
metadata.new <- metadata[metadata$Genotype_names %in% attributes(meth.tissue.norm50L)$sample.ids,]
attributes(meth.tissue.norm50L)$sample.ids == metadata.new$Genotype_names

meth.tissue.norm50L@treatment <- as.integer(factor(metadata.new$Tissue_type))

# data.frame(attributes(meth.tissue.norm50L)$sample.ids, metadata.new$Genotype_names, metadata.new$strata, meth.tissue.norm50L@treatment)

covariates <- dplyr::select(metadata.new, Sex, Age, Cohort2, Maturity_full)
myDiff <- methylKit::calculateDiffMeth(meth.tissue.norm50L,
                                       covariates = NULL,
                                       adjust = "fdr",
                                       chunk.size = 1e+06,
                                       mc.cores = 25)
meth.diff.chr <- methylKit::diffMethPerChr(myDiff,plot = TRUE, 
                                           qvalue.cutoff = 0.01, meth.cutoff = 40)
myDiff40p <- methylKit::getMethylDiff(myDiff, difference = 40, qvalue = 0.01)
#14329 
myDiff25p <- methylKit::getMethylDiff(myDiff, difference = 25, qvalue = 0.01)
#33400  
myDiff15p <- methylKit::getMethylDiff(myDiff, difference = 15, qvalue = 0.01)
#58750
top100.tissue <- slice_max(as.data.frame(as(myDiff,"GRanges")), order_by = meth.diff, n = 100)
top100.tissue <- as(top100.tissue,"GRanges")

save(myDiff,myDiff15p, myDiff25p, myDiff40p,top100.tissue,
     file = paste0(rdata.path,"Toothfish_tissue_comparison_methylation_DMS.Rdata"))
```


```{r}
print(load(paste0(rdata.path,"Toothfish_tissue_comparison_methylation_DMS.Rdata")))

volcano.plot <- ggplot2::ggplot() +
  ggplot2::geom_point(data = myDiff, ggplot2::aes(x = meth.diff, 
                                                  y = -log10(qvalue))) +
  ggplot2::geom_point(data = myDiff25p, ggplot2::aes(x = meth.diff, 
                                                  y = -log10(qvalue)),
                      colour = "red") +
  ggplot2::theme_minimal() +
  ggplot2::theme(
    axis.text = ggplot2::element_text(size = 15),
    axis.title.x = ggplot2::element_text(size = 20),
    axis.title.y = ggplot2::element_text(size = 20),
    legend.text = ggplot2::element_text(size = 15)
  )
print(volcano.plot)

ggplot2::ggsave(volcano.plot, bg = "white",
                filename =
                  paste0(figure.path,
                         "Toothfish_tissue_comparison_DMS25_VOLCANO_methylkit_regions.png"), 
                width = 30, height = 15, units = "cm")
```

### Heatmap

```{r}
print(load(file = paste0(rdata.path,"Toothfish_tissue_comparison_methylation_DMS.Rdata")))

myDiff25p.gr <- as(myDiff25p, "GRanges")
meth.diff <- methylKit::selectByOverlap(meth.tissue.norm50L, myDiff25p.gr)

meth.diff.df <- as.data.frame(t(methylKit::percMethylation(meth.diff)))
loc.names <- paste0(meth.diff$chr,"_",meth.diff$start)
colnames(meth.diff.df) <- loc.names

meth.diff.df$individual <- rownames(meth.diff.df)

# metadata.new <- metadata.new[order(match(metadata.new$individual, meth.diff.df$individual)),]
# identical(metadata.new$individual,meth.diff.df$individual)
# meth.diff.df$pop <- metadata.new$strata

pop <- metadata.new$Tissue_type
meth.diff.df$pop <- as.character(pop)


heat.plot.data <- meth.diff.df %>%
  tibble::as_tibble() %>%
  tidyr::gather(key = "CpG", value = "Methylation_percentage", -individual, -pop,
                factor_key = FALSE) %>%
  dplyr::group_by(pop, CpG) %>%
  dplyr::mutate(CpG_mean = mean(Methylation_percentage, na.rm = TRUE)) %>%
  dplyr::arrange(pop, individual, CpG) %>%
  dplyr::mutate(individual = factor(individual, levels = unique(individual))) %>%
  dplyr::mutate(CpG = factor(CpG, levels = unique(CpG)))
cpgs <- heat.plot.data$CpG[heat.plot.data$individual == unique(heat.plot.data$individual)[1]]
cpgs_mean <- heat.plot.data$CpG_mean[heat.plot.data$individual == unique(heat.plot.data$individual)[1]]

cpgs <- cpgs[order(cpgs_mean, decreasing = TRUE)]
heat.plot.data$CpG <- factor(heat.plot.data$CpG, levels = c("pop", unique(as.character(cpgs))))
heat.plot.data <- heat.plot.data[order(heat.plot.data$CpG),]

heat.plot <- ggplot2::ggplot(data = heat.plot.data, ggplot2::aes(x = individual)) + 
    ggplot2::geom_tile(data = heat.plot.data, 
                       ggplot2::aes(x = individual, y = CpG, fill = Methylation_percentage)) +
    ggplot2::scale_fill_gradient2(low = "blue", mid = "yellow", high = "red", 
                                  midpoint = 50, guide = "colourbar") +
    ggnewscale::new_scale_fill() +
    ggplot2::geom_tile(data = heat.plot.data, ggplot2::aes(x = individual, y = "pop", fill = pop,
                                    height = length(unique(heat.plot.data$CpG))*0.02)) +
    ggplot2::scale_fill_manual(values = as.character(colours2)) +
    ggplot2::scale_y_discrete(drop = FALSE) +
    ggplot2::theme(legend.position = "top",
                   axis.title.y = ggplot2::element_blank(),
                   axis.title.x = ggplot2::element_blank(),
                   axis.text.x = ggplot2::element_text(face = "bold", color = "black", 
                                        size = 10, angle = 90, hjust = 1, vjust = 0.5))
print(heat.plot)

ggplot2::ggsave(heat.plot, bg = "white",
                filename = paste0(figure.path,
                                  "Toothfish_tissue_comparison_methDiff_25_HEATMAP_long.png"), 
                width = 60, height = 30, units = "cm")
```

### Plot CpG difference

```{r}
load(file = paste0(rdata.path,"Toothfish_tissue_comparison_methylation_DMS.Rdata"))

myDiff25p.gr <- as(myDiff25p, "GRanges")
meth.diff <- methylKit::selectByOverlap(meth.tissue.norm50L, myDiff25p.gr) #113 

length(unique(meth.diff$chr)) #34
sort(table(meth.diff$chr), decreasing = TRUE)[1:20]

meth.diff.df <- as.data.frame(t(methylKit::percMethylation(meth.diff)))
loc.names <- paste0(meth.diff$chr,"_",meth.diff$start)
colnames(meth.diff.df) <- loc.names

meth.diff.df$individual <- rownames(meth.diff.df) 
identical(metadata.new$Genotype_names, meth.diff.df$individual)
meth.diff.df$pop <- metadata.new$Tissue_type



meth.diff.df.plot <- tidyr::gather(meth.diff.df, key = locus, 
                           value = perc_methylation, -individual, -pop,
                           factor_key = TRUE) %>%
  dplyr::mutate(chr = stringr::str_split_fixed(locus, 
                                               pattern = "_", 
                                               n = 2)[,1])
meth.diff.df.plot.locus <- meth.diff.df.plot %>% dplyr::group_by(chr,locus,pop) %>% 
  dplyr::summarise(mean_pop = mean(perc_methylation, na.rm = TRUE),
                   sum_NA = sum(is.na(perc_methylation)),
            .groups = 'drop')


diff.plot <- ggplot2::ggplot() +
  ggplot2::geom_line(data = meth.diff.df.plot.locus,
                     ggplot2::aes(x = pop, y = mean_pop, colour = locus, group = locus)) +
  ggplot2::geom_point(data = meth.diff.df.plot.locus,
                     ggplot2::aes(x = pop, y = mean_pop, colour = locus, group = locus),
                     size = 5) +
  ggplot2::facet_wrap(~chr) +
  ggplot2::ylab("Mean methylation (%)") +
  ggplot2::theme(legend.position = "none",
                   axis.text.x = ggplot2::element_text(face = "bold", color = "black", 
                                        size = 10, angle = 90, hjust = 1, vjust = 0.5))

print(diff.plot)

ggplot2::ggsave(plot = diff.plot, 
                filename = paste0(figure.path,"1_Toothfish_tissue_comparison_methDiff.png"),
                width = 60, height = 60, units = "cm", device = "png")
```


### DMS Clustering
```{r}
load(file = paste0(rdata.path,"Toothfish_tissue_comparison_methylation_DMS.Rdata"))

myDiff25p.gr <- as(myDiff25p, "GRanges")
meth.diff <- methylKit::selectByOverlap(meth.tissue.norm50L, myDiff25p.gr) #112  
methdif.perc <- methylKit::percMethylation(meth.diff)

methdif.perc2 <- zoo::na.aggregate(methdif.perc, by = meth.tissue.norm50L@treatment)
PCobj <- prcomp(methdif.perc2, retx = TRUE, scale. = TRUE)

var_frac <- PCobj$sdev/sum(PCobj$sdev)
print(paste0("Total variance explained: ",signif(sum(var_frac[1:3]) * 100, 3), "%")) 
barplot(var_frac[1:3], main = "variance explained", 
        col = heat.colors(length(var_frac[1:3])))


identical(metadata.new$Genotype_names, meth.tissue.norm50L@sample.ids)
# identical(metadata.new$individual, rownames(PCobj$x))
identical(metadata.new$Genotype_names, rownames(PCobj$rotation))
pop <- metadata.new$Tissue_type

plot.title <- "Toothfish methylation - tissue comparison (33400 DMS sites - 25%)"

# data <- data.frame(ind = rownames(PCobj$x),PCobj$x,  pop = pop)
data <- data.frame(ind = rownames(PCobj$rotation),PCobj$rotation,  pop = pop)
data <- merge(data, aggregate(cbind(mean.x = PC1, mean.y = PC2) ~
                                pop, data, mean), by = "pop")

pca.plot <- ggplot2::ggplot(data, ggplot2::aes(PC1, PC2, color = pop)) +
  ggplot2::geom_point(size = 4) + 
  ggplot2::geom_point(ggplot2::aes(x = mean.x,
                                   y = mean.y, color = pop),
                      size = 7,
                      shape = 17) +
  ggplot2::scale_colour_manual(values = as.character(colours2)) +
  ggplot2::geom_hline(yintercept = 0) +
  ggplot2::geom_vline(xintercept = 0) + 
  ggplot2::labs(
    subtitle = paste0("Total variance explained: ",
                      signif(sum(var_frac[1:3]) * 100, 3), "%"),
    y = paste0("PC2; ", signif(var_frac[2] * 100, digits = 2), " %"),
    x = paste0("PC1; ", signif(var_frac[1] * 100, digits = 2), " %"),
    title = plot.title,
    caption = "") + 
  ggplot2::theme_minimal() +
  ggplot2::theme(
    axis.text = ggplot2::element_text(size = 15),
    axis.title.x = ggplot2::element_text(size = 20),
    axis.title.y = ggplot2::element_text(size = 20),
    legend.text = ggplot2::element_text(size = 15)
  )

print(pca.plot)
ggplot2::ggsave(pca.plot, bg = "white",
                filename = paste0(figure.path,"Toothfish_tissue_comparison_PCA_CpG_DMS_PC12.png"), 
                width = 30, height = 15, units = "cm")



# data <- data.frame(ind = rownames(PCobj$x),PCobj$x,  pop = pop)
data <- data.frame(ind = rownames(PCobj$rotation),PCobj$rotation,  pop = pop)
data <- merge(data, aggregate(cbind(mean.x = PC1, mean.y = PC3) ~
                                pop, data, mean), by = "pop")

pca.plot <- ggplot2::ggplot(data, ggplot2::aes(PC1, PC3, color = pop)) +
  ggplot2::geom_point(size = 4) + 
  ggplot2::geom_point(ggplot2::aes(x = mean.x,
                                   y = mean.y, color = pop),
                      size = 7,
                      shape = 17) +
  ggplot2::scale_colour_manual(values = as.character(colours2)) +
  ggplot2::geom_hline(yintercept = 0) +
  ggplot2::geom_vline(xintercept = 0) + 
  ggplot2::labs(
    subtitle = paste0("Total variance explained: ",
                      signif(sum(var_frac[1:3]) * 100, 3), "%"),
    y = paste0("PC3; ", signif(var_frac[3] * 100, digits = 2), " %"),
    x = paste0("PC1; ", signif(var_frac[1] * 100, digits = 2), " %"),
    title = plot.title,
    caption = "") + 
  ggplot2::theme_minimal() +
  ggplot2::theme(
    axis.text = ggplot2::element_text(size = 15),
    axis.title.x = ggplot2::element_text(size = 20),
    axis.title.y = ggplot2::element_text(size = 20),
    legend.text = ggplot2::element_text(size = 15)
  )

print(pca.plot)
ggplot2::ggsave(pca.plot, bg = "white",
                filename = paste0(figure.path,
                                  "Toothfish_tissue_comparison_PCA_CpG_DMS_PC13.png"), 
                width = 30, height = 15, units = "cm")




# data <- data.frame(ind = rownames(PCobj$x),PCobj$x,  pop = pop)
data <- data.frame(ind = rownames(PCobj$rotation),PCobj$rotation,  pop = pop)
data <- merge(data, aggregate(cbind(mean.x = PC2, mean.y = PC3) ~
                                pop, data, mean), by = "pop")

pca.plot <- ggplot2::ggplot(data, ggplot2::aes(PC2, PC3, color = pop)) +
  ggplot2::geom_point(size = 4) + 
  ggplot2::geom_point(ggplot2::aes(x = mean.x,
                                   y = mean.y, color = pop),
                      size = 7,
                      shape = 17) +
  ggplot2::scale_colour_manual(values = as.character(colours2)) +
  ggplot2::geom_hline(yintercept = 0) +
  ggplot2::geom_vline(xintercept = 0) + 
  ggplot2::labs(
    subtitle = paste0("Total variance explained: ",
                      signif(sum(var_frac[1:3]) * 100, 3), "%"),
    y = paste0("PC3; ", signif(var_frac[3] * 100, digits = 2), " %"),
    x = paste0("PC2; ", signif(var_frac[2] * 100, digits = 2), " %"),
    title = plot.title,
    caption = "") + 
  ggplot2::theme_minimal() +
  ggplot2::theme(
    axis.text = ggplot2::element_text(size = 15),
    axis.title.x = ggplot2::element_text(size = 20),
    axis.title.y = ggplot2::element_text(size = 20),
    legend.text = ggplot2::element_text(size = 15)
  )

print(pca.plot)
ggplot2::ggsave(pca.plot, bg = "white",
                filename = paste0(figure.path,"Toothfish_tissue_comparison_PCA_CpG_DMS_PC23.png"), 
                width = 30, height = 15, units = "cm")
```



## DMS RANDOM - Tissue comparison

```{r,eval=FALSE}
print(load(paste0(rdata.path,"Toothfish_tissue_comparison_methylkit_object.Rdata")))

print(load(paste0(rdata.path,"Toothfish_metadata.Rdata")))
metadata.new <- metadata[metadata$Genotype_names %in% attributes(meth.tissue.norm50L)$sample.ids,]
attributes(meth.tissue.norm50L)$sample.ids == metadata.new$Genotype_names

strata.random <- sample(metadata.new$Tissue_type, size = length(metadata.new$Tissue_type), replace = FALSE)
meth.tissue.norm50L@treatment <- as.integer(factor(strata.random))

# data.frame(attributes(meth.tissue.norm50L)$sample.ids, metadata.new$Genotype_names, metadata.new$strata, meth.tissue.norm50L@treatment)

# covariates <- dplyr::select(metadata.new, Sex, TL_mm)
myDiff <- methylKit::calculateDiffMeth(meth.tissue.norm50L,
                                       covariates = NULL,
                                       adjust = "fdr",
                                       chunk.size = 1e+06,
                                       mc.cores = 25)
meth.diff.chr <- methylKit::diffMethPerChr(myDiff,plot = TRUE, 
                                           qvalue.cutoff = 0.01, meth.cutoff = 40)
myDiff40p <- methylKit::getMethylDiff(myDiff, difference = 40, qvalue = 0.01)
#0
myDiff25p <- methylKit::getMethylDiff(myDiff, difference = 25, qvalue = 0.01)
#0
myDiff15p <- methylKit::getMethylDiff(myDiff, difference = 15, qvalue = 0.01)
#26   

save(strata.random, myDiff,myDiff15p, myDiff25p, myDiff40p,
     file = paste0(rdata.path,"Toothfish_tissue_comparison_methylation_RANDOM_DMS.Rdata"))
```


```{r}
print(load(paste0(rdata.path,"Toothfish_tissue_comparison_methylation_RANDOM_DMS.Rdata")))

volcano.plot <- ggplot2::ggplot() +
  ggplot2::geom_point(data = myDiff, ggplot2::aes(x = meth.diff, 
                                                  y = -log10(qvalue))) +
  ggplot2::geom_point(data = myDiff15p, ggplot2::aes(x = meth.diff, 
                                                  y = -log10(qvalue)),
                      colour = "red") +
  
  ggplot2::theme_minimal() +
  ggplot2::theme(
    axis.text = ggplot2::element_text(size = 15),
    axis.title.x = ggplot2::element_text(size = 20),
    axis.title.y = ggplot2::element_text(size = 20),
    legend.text = ggplot2::element_text(size = 15)
  )
print(volcano.plot)

ggplot2::ggsave(volcano.plot, bg = "white",
                filename =
                  paste0(figure.path,
                         "Toothfish_tissue_comparison_RANDOM_DMS15_VOLCANO_methylkit.png"), 
                width = 30, height = 15, units = "cm")
```


### Heatmap

```{r}
print(load(file = paste0(rdata.path,"Toothfish_tissue_comparison_methylation_RANDOM_DMS.Rdata")))

myDiff15p.gr <- as(myDiff15p, "GRanges")
meth.diff <- methylKit::selectByOverlap(meth.tissue.norm50L, myDiff15p.gr)

meth.diff.df <- as.data.frame(t(methylKit::percMethylation(meth.diff)))
loc.names <- paste0(meth.diff$chr,"_",meth.diff$start)
colnames(meth.diff.df) <- loc.names

meth.diff.df$individual <- rownames(meth.diff.df)

# metadata.new <- metadata.new[order(match(metadata.new$individual, meth.diff.df$individual)),]
# identical(metadata.new$Genotype_names,meth.diff.df$individual)

meth.diff.df$pop <- strata.random
meth.diff.df$pop2 <- as.character(metadata.new$Tissue_type)


heat.plot.data <- meth.diff.df %>%
  tibble::as_tibble() %>%
  tidyr::gather(key = "CpG", value = "Methylation_percentage", -individual, -pop, -pop2,
                factor_key = FALSE) %>%
  dplyr::group_by(pop, CpG) %>%
  dplyr::mutate(CpG_mean = mean(Methylation_percentage, na.rm = TRUE)) %>%
  dplyr::arrange(pop, individual, CpG) %>%
  dplyr::mutate(individual = factor(individual, levels = unique(individual))) %>%
  dplyr::mutate(CpG = factor(CpG, levels = unique(CpG)))
cpgs <- heat.plot.data$CpG[heat.plot.data$individual == unique(heat.plot.data$individual)[1]]
cpgs_mean <- heat.plot.data$CpG_mean[heat.plot.data$individual == unique(heat.plot.data$individual)[1]]

cpgs <- cpgs[order(cpgs_mean, decreasing = TRUE)]
heat.plot.data$CpG <- factor(heat.plot.data$CpG, levels = c("pop", unique(as.character(cpgs))))
heat.plot.data <- heat.plot.data[order(heat.plot.data$CpG),]

heat.plot <- ggplot2::ggplot(data = heat.plot.data, ggplot2::aes(x = individual)) + 
    ggplot2::geom_tile(data = heat.plot.data, 
                       ggplot2::aes(x = individual, y = CpG, fill = Methylation_percentage)) +
    ggplot2::scale_fill_gradient2(low = "blue", mid = "yellow", high = "red", 
                                  midpoint = 50, guide = "colourbar") +
    ggnewscale::new_scale_fill() +
    ggplot2::geom_tile(data = heat.plot.data, ggplot2::aes(x = individual, 
                                                           y = "pop - random", fill = pop,
                                    height = length(unique(heat.plot.data$CpG))*0.02)) +
    ggplot2::scale_fill_manual(values = as.character(colours2)) +
    ggnewscale::new_scale_fill() +
    ggplot2::geom_tile(data = heat.plot.data, ggplot2::aes(x = individual, y = "pop", fill = pop2,
                                    height = length(unique(heat.plot.data$CpG))*0.02)) +
    ggplot2::scale_fill_manual(values = as.character(colours2)) +
    ggplot2::scale_y_discrete(drop = FALSE) +
    ggplot2::theme(legend.position = "top",
                   axis.title.y = ggplot2::element_blank(),
                   axis.title.x = ggplot2::element_blank(),
                   axis.text.x = ggplot2::element_text(face = "bold", color = "black", 
                                        size = 10, angle = 90, hjust = 1, vjust = 0.5))
print(heat.plot)

ggplot2::ggsave(heat.plot, bg = "white",
                filename = paste0(figure.path,
                   "Toothfish_tissue_comparison_RANDOM_methDiff_15_HEATMAP_long.png"), 
                width = 60, height = 30, units = "cm")
```

### Plot CpG difference

```{r}
load(file = paste0(rdata.path,"Toothfish_tissue_comparison_methylation_RANDOM_DMS.Rdata"))

myDiff15p.gr <- as(myDiff15p, "GRanges")
meth.diff <- methylKit::selectByOverlap(meth.tissue.norm50L, myDiff15p.gr) #113 

length(unique(meth.diff$chr)) #34
sort(table(meth.diff$chr), decreasing = TRUE)[1:20]

meth.diff.df <- as.data.frame(t(methylKit::percMethylation(meth.diff)))
loc.names <- paste0(meth.diff$chr,"_",meth.diff$start)
colnames(meth.diff.df) <- loc.names

meth.diff.df$individual <- rownames(meth.diff.df) 
identical(metadata.new$Genotype_names, meth.diff.df$individual)
meth.diff.df$pop <- metadata.new$Tissue_type


meth.diff.df.plot <- tidyr::gather(meth.diff.df, key = locus, 
                           value = perc_methylation, -individual, -pop,
                           factor_key = TRUE) %>%
  dplyr::mutate(chr = stringr::str_split_fixed(locus, 
                                               pattern = "_", 
                                               n = 2)[,1])
meth.diff.df.plot.locus <- meth.diff.df.plot %>% dplyr::group_by(chr,locus,pop) %>% 
  dplyr::summarise(mean_pop = mean(perc_methylation, na.rm = TRUE),
                   sum_NA = sum(is.na(perc_methylation)),
            .groups = 'drop')


diff.plot <- ggplot2::ggplot() +
  ggplot2::geom_line(data = meth.diff.df.plot.locus,
                     ggplot2::aes(x = pop, y = mean_pop, colour = locus, group = locus)) +
  ggplot2::geom_point(data = meth.diff.df.plot.locus,
                     ggplot2::aes(x = pop, y = mean_pop, colour = locus, group = locus),
                     size = 5) +
  ggplot2::facet_wrap(~chr) +
  ggplot2::ylab("Mean methylation (%)") +
  ggplot2::theme(legend.position = "none",
                   axis.text.x = ggplot2::element_text(face = "bold", color = "black", 
                                        size = 10, angle = 90, hjust = 1, vjust = 0.5))
    # ggplot2::guides(colour = ggplot2::guide_legend(ncol = 2))

print(diff.plot)

ggplot2::ggsave(plot = diff.plot, 
                filename = paste0(figure.path,"1_Toothfish_tissue_comparison_RANDOM_methDiff.png"),
                width = 30, height = 15, units = "cm", device = "png")
```


### DMS Clustering
```{r}
load(file = paste0(rdata.path,"Toothfish_tissue_comparison_methylation_RANDOM_DMS.Rdata"))

myDiff15p.gr <- as(myDiff15p, "GRanges")
meth.diff <- methylKit::selectByOverlap(meth.tissue.norm50L, myDiff15p.gr) #270  
methdif.perc <- methylKit::percMethylation(meth.diff)

methdif.perc2 <- zoo::na.aggregate(methdif.perc, by = meth.tissue.norm50L@treatment)
PCobj <- prcomp(methdif.perc2, retx = TRUE, scale. = TRUE) # Too many missing values

var_frac <- PCobj$sdev/sum(PCobj$sdev)
print(paste0("Total variance explained: ",signif(sum(var_frac[1:3]) * 100, 3), "%")) 
barplot(var_frac[1:3], main = "variance explained", 
        col = heat.colors(length(var_frac[1:3])))


identical(metadata.new$Genotype_names, meth.tissue.norm50L@sample.ids)

identical(as.numeric(factor(metadata.new$pop)), meth.tissue.norm50L@treatment)

# identical(metadata.new$individual, rownames(PCobj$x))
identical(metadata.new$Genotype_names, rownames(PCobj$rotation))
pop <- metadata.new$pop

plot.title <- "Toothfish methylation - RANDOMISED tissue comparison (26 DMS sites - 15%)"

# data <- data.frame(ind = rownames(PCobj$x),PCobj$x,  pop = pop)
data <- data.frame(ind = rownames(PCobj$rotation),PCobj$rotation,  pop = pop)
data <- merge(data, aggregate(cbind(mean.x = PC1, mean.y = PC2) ~
                                pop, data, mean), by = "pop")

pca.plot <- ggplot2::ggplot(data, ggplot2::aes(PC1, PC2, color = pop)) +
  ggplot2::geom_point(size = 4) + 
  ggplot2::geom_point(ggplot2::aes(x = mean.x,
                                   y = mean.y, color = pop),
                      size = 7,
                      shape = 17) +
  ggplot2::scale_colour_manual(values = as.character(colours2)) +
  ggplot2::geom_hline(yintercept = 0) +
  ggplot2::geom_vline(xintercept = 0) + 
  ggplot2::labs(
    subtitle = paste0("Total variance explained: ",
                      signif(sum(var_frac[1:3]) * 100, 3), "%"),
    y = paste0("PC2; ", signif(var_frac[2] * 100, digits = 2), " %"),
    x = paste0("PC1; ", signif(var_frac[1] * 100, digits = 2), " %"),
    title = plot.title,
    caption = "") + 
  ggplot2::theme_minimal() +
  ggplot2::theme(
    axis.text = ggplot2::element_text(size = 15),
    axis.title.x = ggplot2::element_text(size = 20),
    axis.title.y = ggplot2::element_text(size = 20),
    legend.text = ggplot2::element_text(size = 15)
  )

print(pca.plot)
ggplot2::ggsave(pca.plot, bg = "white",
                filename = 
                  paste0(figure.path,"Toothfish_tissue_comparison_RANDOM_PCA_CpG_DMS_PC12.png"), 
                width = 30, height = 15, units = "cm")



# data <- data.frame(ind = rownames(PCobj$x),PCobj$x,  pop = pop)
data <- data.frame(ind = rownames(PCobj$rotation),PCobj$rotation,  pop = pop)
data <- merge(data, aggregate(cbind(mean.x = PC1, mean.y = PC3) ~
                                pop, data, mean), by = "pop")

pca.plot <- ggplot2::ggplot(data, ggplot2::aes(PC1, PC3, color = pop)) +
  ggplot2::geom_point(size = 4) + 
  ggplot2::geom_point(ggplot2::aes(x = mean.x,
                                   y = mean.y, color = pop),
                      size = 7,
                      shape = 17) +
  ggplot2::scale_colour_manual(values = as.character(colours2)) +
  ggplot2::geom_hline(yintercept = 0) +
  ggplot2::geom_vline(xintercept = 0) + 
  ggplot2::labs(
    subtitle = paste0("Total variance explained: ",
                      signif(sum(var_frac[1:3]) * 100, 3), "%"),
    y = paste0("PC3; ", signif(var_frac[3] * 100, digits = 2), " %"),
    x = paste0("PC1; ", signif(var_frac[1] * 100, digits = 2), " %"),
    title = plot.title,
    caption = "") + 
  ggplot2::theme_minimal() +
  ggplot2::theme(
    axis.text = ggplot2::element_text(size = 15),
    axis.title.x = ggplot2::element_text(size = 20),
    axis.title.y = ggplot2::element_text(size = 20),
    legend.text = ggplot2::element_text(size = 15)
  )

print(pca.plot)
ggplot2::ggsave(pca.plot, bg = "white",
                filename = paste0(figure.path,"Toothfish_tissue_comparison_RANDOM_PCA_CpG_DMS_PC13.png"), 
                width = 30, height = 15, units = "cm")




# data <- data.frame(ind = rownames(PCobj$x),PCobj$x,  pop = pop)
data <- data.frame(ind = rownames(PCobj$rotation),PCobj$rotation,  pop = pop)
data <- merge(data, aggregate(cbind(mean.x = PC2, mean.y = PC3) ~
                                pop, data, mean), by = "pop")

pca.plot <- ggplot2::ggplot(data, ggplot2::aes(PC2, PC3, color = pop)) +
  ggplot2::geom_point(size = 4) + 
  ggplot2::geom_point(ggplot2::aes(x = mean.x,
                                   y = mean.y, color = pop),
                      size = 7,
                      shape = 17) +
  ggplot2::scale_colour_manual(values = as.character(colours2)) +
  ggplot2::geom_hline(yintercept = 0) +
  ggplot2::geom_vline(xintercept = 0) + 
  ggplot2::labs(
    subtitle = paste0("Total variance explained: ",
                      signif(sum(var_frac[1:3]) * 100, 3), "%"),
    y = paste0("PC3; ", signif(var_frac[3] * 100, digits = 2), " %"),
    x = paste0("PC2; ", signif(var_frac[2] * 100, digits = 2), " %"),
    title = plot.title,
    caption = "") + 
  ggplot2::theme_minimal() +
  ggplot2::theme(
    axis.text = ggplot2::element_text(size = 15),
    axis.title.x = ggplot2::element_text(size = 20),
    axis.title.y = ggplot2::element_text(size = 20),
    legend.text = ggplot2::element_text(size = 15)
  )

print(pca.plot)
ggplot2::ggsave(pca.plot, bg = "white",
                filename = 
                  paste0(figure.path,"Toothfish_tissue_comparison_RANDOM_PCA_CpG_DMS_PC23.png"), 
                width = 30, height = 15, units = "cm")
```



## DMS methylation by Feature

```{r}
print(load(paste0(rdata.path,"Toothfish_tissue_comparison_methylkit_object.Rdata")))
print(load(paste0(rdata.path,"Toothfish_metadata.Rdata"))) 
print(load(file = paste0(rdata.path,"Toothfish_tissue_comparison_methylation_DMS.Rdata")))
print(load(file = paste0(rdata.path,"Toothfish_genome_annotation_features.Rdata")))

myDiff25p.gr <- as(myDiff25p, "GRanges")
meth.diff <- methylKit::selectByOverlap(meth.tissue.norm50L, myDiff25p.gr) #113 

TOA.gr <- as(meth.diff, "GRanges")

feat.df <- data.frame(annot = NULL, target = NULL, feature = NULL)
ann <- genomation::annotateWithFeature(TOA.gr, annotation.all.features$promoters, 
                                       intersect.chr = TRUE)
feat.df <- rbind(feat.df, data.frame(annot = ann@annotation[1], 
                                     target = ann@perc.of.OlapFeat, 
                                     feature = "Promoters"))
ann <- genomation::annotateWithFeature(TOA.gr, annotation.all.features$TSSes, 
                                       intersect.chr = TRUE)
feat.df <- rbind(feat.df, data.frame(annot = ann@annotation[1], 
                                     target = ann@perc.of.OlapFeat, 
                                     feature = "TSS"))
ann <- genomation::annotateWithFeature(TOA.gr, annotation.all.features$genes, 
                                       intersect.chr = TRUE)
feat.df <- rbind(feat.df, data.frame(annot = ann@annotation[1], 
                                     target = ann@perc.of.OlapFeat, 
                                     feature = "Genes"))
ann <- genomation::annotateWithFeature(TOA.gr, annotation.all.features$transcripts, 
                                       intersect.chr = TRUE)
feat.df <- rbind(feat.df, data.frame(annot = ann@annotation[1], 
                                     target = ann@perc.of.OlapFeat, 
                                     feature = "Transcipts"))
ann <- genomation::annotateWithFeature(TOA.gr, annotation.all.features$cds, 
                                       intersect.chr = TRUE)
feat.df <- rbind(feat.df, data.frame(annot = ann@annotation[1], 
                                     target = ann@perc.of.OlapFeat, 
                                     feature = "CDS"))
ann <- genomation::annotateWithFeature(TOA.gr, annotation.all.features$exons,
                                       intersect.chr = TRUE) 
feat.df <- rbind(feat.df, data.frame(annot = ann@annotation[1], 
                                     target = ann@perc.of.OlapFeat, 
                                     feature = "Exons"))
ann <- genomation::annotateWithFeature(TOA.gr, annotation.all.features$introns, 
                                       intersect.chr = TRUE) 
feat.df <- rbind(feat.df, data.frame(annot = ann@annotation[1], 
                                     target = ann@perc.of.OlapFeat, 
                                     feature = "Introns"))
# ann <- genomation::annotateWithFeature(TOA.gr, annotation.all.features$intergenics,
#                                        intersect.chr = TRUE) 
# feat.df <- rbind(feat.df, data.frame(annot = ann@annotation[1], 
#                                      target = ann@perc.of.OlapFeat, 
#                                      feature = "Intergenic"))

feat.df2 <- reshape2::melt(feat.df, id.vars = 'feature')
feat.df2$variable <- as.character(feat.df2$variable)
feat.df2$variable[feat.df2$variable == "annot"] <- "Percentage of CpG sites overlapping with features"
feat.df2$variable[feat.df2$variable == "target"] <- "Percentage of features covered by CpG sites"

plot.title <- "Overlap between annotations and 33,400 CpG sites"

annot.plot <- ggplot2::ggplot(data = feat.df2, 
                              ggplot2::aes(x =  feature, y = value, fill = variable)) +
  ggplot2::geom_bar(stat='identity', position = 'dodge') +
    ggplot2::labs(
    y = "Overlap %",
    x = "Annotation feature",
    title = plot.title,
    caption = "") +
  ggplot2::ylim(0,100) +
  ggplot2::theme_minimal() +
  ggplot2::theme(
    axis.text = ggplot2::element_text(size = 15),
    axis.text.x = ggplot2::element_text(size = 15, angle = 90),
    axis.title.x = ggplot2::element_text(size = 20),
    axis.title.y = ggplot2::element_text(size = 20),
    legend.text = ggplot2::element_text(size = 15)
  )
print(annot.plot)
ggplot2::ggsave(annot.plot, bg = "white",
                filename =
                  paste0(figure.path,
                         "Toothfish_tissue_comparison_annotation_overlap.png"), 
                width = 30, height = 15, units = "cm")

```


### Promotors

```{r}
idx_overlaps <- S4Vectors::queryHits(GenomicRanges::findOverlaps(
  TOA.gr,
  annotation.all.features$promoters, 
  type = 'any'))
promotor.gr <- TOA.gr[unique(idx_overlaps),]
promotor.raw <- meth.diff[unique(idx_overlaps),]

# promoters <- methylKit::regionCounts(meth.diff,annotation.all.features$promoters)
# methylKit::clusterSamples(promoters, dist = "correlation", method = "ward.D2", plot = TRUE)


promotor.prc <- methylKit::percMethylation(promotor.raw) # get percent methylation matrix
mds <- matrixStats::rowSds(promotor.prc, na.rm = TRUE) # calculate standard deviation of CpGs
hist(mds,col="cornflowerblue",xlab = "Std. dev. per CpG")


promotor.prc <- as.data.frame(promotor.prc)
promotor.prc$seqs <- paste0(promotor.gr@seqnames,"_", promotor.gr@ranges@start)

promotor.prc.long <- tidyr::gather(data = promotor.prc, sample, methylation, -seqs)
strata <- data.frame(sample  = metadata.new$Genotype_names, region = metadata.new$Tissue_type)
promotor.prc.long <- dplyr::left_join(promotor.prc.long, strata, by = "sample" )
promotor.grp <- dplyr::group_by(promotor.prc.long, region, seqs) %>%
    dplyr::summarise(mean_meth  = mean(methylation, na.rm = TRUE))
promotor.grp$feat <- "Promoter"

feat.plot <- ggplot2::ggplot(data = promotor.grp,ggplot2::aes(x =  feat, y = mean_meth)) +
  ggplot2::geom_violin(width = 1.5) +
  ggplot2::geom_jitter(size = 0.5, alpha = 0.4) +
    ggplot2::labs(
    y = "Mean methylation (%) per CpG site per region",
    x = "Annotation feature",
    title = "Each point is a CpG site",
    caption = "") +
  ggplot2::facet_wrap(~region) +
  ggplot2::theme_minimal() +
  ggplot2::theme(
    axis.text = ggplot2::element_text(size = 15),
    axis.text.x = ggplot2::element_text(size = 15, angle = 90),
    axis.title.x = ggplot2::element_text(size = 20),
    axis.title.y = ggplot2::element_text(size = 20),
    legend.text = ggplot2::element_text(size = 15)
  )
print(feat.plot)


ggplot2::ggsave(feat.plot, bg = "white",
                filename =
                  paste0(figure.path,"Toothfish_tissue_comparison_PROMOTOR_Methylation_per_CpG.png"), 
                width = 30, height = 15, units = "cm")

```


```{r}
sm <- genomation::ScoreMatrix(target = promotor.gr, windows = annotation.all.features$promoters,
                              strand.aware = TRUE,
                               weight.col = NULL, is.noCovNA = FALSE, type = "auto", 
                              rpm = FALSE, unique = FALSE, extend = 0, param = NULL, 
                              bam.paired.end = FALSE, library.size = NULL)
genomation::heatMatrix(sm, xcoords = c(-1000, 1000))


sml.sub <- intersectScoreMatrixList(sml, reorder = FALSE) # because scoreMatrices have different dimensions, we need them to cover the same promoters for the heatmap
multiHeatMatrix(sml.sub, xcoords=c(-2000, 2000), matrix.main=names(targets), xlab="region around TSS")


```


# 14_Overlapping DMS sites

## load data

```{r}
setwd("C:/Users/dev093/OneDrive - CSIRO/Postdoc/3.Toothfish/4.Analysis/R-analysis")
figure.path <- "C:/Users/dev093/OneDrive - CSIRO/Postdoc/3.Toothfish/4.Analysis/R-analysis/Figures/"
rdata.path <- "C:/Users/dev093/OneDrive - CSIRO/Postdoc/3.Toothfish/4.Analysis/R-analysis/Rdata/"

print(load(paste0(rdata.path,"Toothfish_2018_muscle_methylation_DMS.Rdata")))
Toothfish_2018_muscle_DMS <- paste0(myDiff25p$chr,"_", myDiff25p$start)
Toothfish_2018_muscle_TOP100_DMS <- paste0(as.character(top100.2018M@seqnames),"_", as.data.frame(top100.2018M@ranges)$start)

print(load(paste0(rdata.path,"Toothfish_2024_muscle_methylation_DMS.Rdata")))
Toothfish_2024_muscle_DMS <- paste0(myDiff25p$chr,"_", myDiff25p$start)
Toothfish_2024_muscle_TOP100_DMS <- paste0(as.character(top100.2024M@seqnames),"_", as.data.frame(top100.2024M@ranges)$start)

print(load(paste0(rdata.path,"Toothfish_ALL_muscle_methylation_DMS.Rdata")))
Toothfish_ALL_muscle_DMS <- paste0(myDiff25p$chr,"_", myDiff25p$start)
Toothfish_ALL_muscle_TOP100_DMS <- paste0(as.character(top100.ALLM@seqnames),"_", as.data.frame(top100.ALLM@ranges)$start)

print(load(paste0(rdata.path,"Toothfish_2024_fin_methylation_DMS.Rdata")))
Toothfish_2024_fin_DMS <- paste0(myDiff25p$chr,"_", myDiff25p$start)
Toothfish_2024_fin_TOP100_DMS <- paste0(as.character(top100.2024F@seqnames),"_", as.data.frame(top100.2024F@ranges)$start)

print(load(paste0(rdata.path,"Toothfish_ALL_fin_methylation_DMS.Rdata")))
Toothfish_ALL_fin_DMS <- paste0(myDiff25p$chr,"_", myDiff25p$start)
Toothfish_ALL_fin_TOP100_DMS <- paste0(as.character(top100.ALLF@seqnames),"_", as.data.frame(top100.ALLF@ranges)$start)

print(load(paste0(rdata.path,"Toothfish_2024_pool_muscle_methylation_DMS.Rdata")))
Pooled_2024_muscle_DMS <- paste0(myDiff25p$chr,"_", myDiff25p$start)
Pooled_2024_muscle_TOP100_DMS <- paste0(as.character(top100.POOLM@seqnames),"_", as.data.frame(top100.POOLM@ranges)$start)

print(load(paste0(rdata.path,"Toothfish_2024_pool_fin_methylation_DMS.Rdata")))
Pooled_2024_fin_DMS <- paste0(myDiff25p$chr,"_", myDiff25p$start)
Pooled_2024_fin_TOP100_DMS <- paste0(as.character(top100.POOLF@seqnames),"_", as.data.frame(top100.POOLF@ranges)$start)

print(load(paste0(rdata.path,"Toothfish_tissue_comparison_methylation_DMS.Rdata")))
Toothfish_ALL_tissue_DMS <- paste0(myDiff25p$chr,"_", myDiff25p$start)
Toothfish_ALL_tissue_TOP100_DMS <- paste0(as.character(top100.2018M@seqnames),"_", as.data.frame(top100.2018M@ranges)$start)

print(load(paste0(rdata.path,"Toothfish_2018_muscle_methylation_RANDOM_DMS.Rdata")))
Toothfish_2018_muscle_RANDOM_DMS <- paste0(myDiff25p$chr,"_", myDiff25p$start)
print(load(paste0(rdata.path,"Toothfish_2024_muscle_methylation_RANDOM_DMS.Rdata")))
Toothfish_2024_muscle_RANDOM_DMS <- paste0(myDiff25p$chr,"_", myDiff25p$start)
print(load(paste0(rdata.path,"Toothfish_ALL_muscle_methylation_RANDOM_DMS.Rdata")))
Toothfish_ALL_muscle_RANDOM_DMS <- paste0(myDiff25p$chr,"_", myDiff25p$start)
print(load(paste0(rdata.path,"Toothfish_2024_fin_methylation_RANDOM_DMS.Rdata")))
Toothfish_2024_fin_RANDOM_DMS <- paste0(myDiff25p$chr,"_", myDiff25p$start)
print(load(paste0(rdata.path,"Toothfish_ALL_fin_methylation_RANDOM_DMS.Rdata")))
Toothfish_ALL_fin_RANDOM_DMS <- paste0(myDiff25p$chr,"_", myDiff25p$start)
print(load(paste0(rdata.path,"Toothfish_2024_pool_muscle_methylation_RANDOM_DMS.Rdata")))
Pooled_2024_muscle_RANDOM_DMS <- paste0(myDiff25p$chr,"_", myDiff25p$start)
print(load(paste0(rdata.path,"Toothfish_2024_pool_fin_methylation_RANDOM_DMS.Rdata")))
Pooled_2024_fin_RANDOM_DMS <- paste0(myDiff25p$chr,"_", myDiff25p$start)
print(load(paste0(rdata.path,"Toothfish_tissue_comparison_methylation_RANDOM_DMS.Rdata")))
Toothfish_ALL_tissue_RANDOM_DMS <- paste0(myDiff25p$chr,"_", myDiff25p$start)





# print(load(paste0(rdata.path,"Toothfish_2018_muscle_methylation_DMS.Rdata")))
# Toothfish_2018_muscle_DMS <- paste0(myDiff15p$chr,"_", myDiff15p$start)
# print(load(paste0(rdata.path,"Toothfish_2024_muscle_methylation_DMS.Rdata")))
# Toothfish_2024_muscle_DMS <- paste0(myDiff15p$chr,"_", myDiff15p$start)
# print(load(paste0(rdata.path,"Toothfish_ALL_muscle_methylation_DMS.Rdata")))
# Toothfish_ALL_muscle_DMS <- paste0(myDiff15p$chr,"_", myDiff15p$start)
# print(load(paste0(rdata.path,"Toothfish_2024_fin_methylation_DMS.Rdata")))
# Toothfish_2024_fin_DMS <- paste0(myDiff15p$chr,"_", myDiff15p$start)
# print(load(paste0(rdata.path,"Toothfish_ALL_fin_methylation_DMS.Rdata")))
# Toothfish_ALL_fin_DMS <- paste0(myDiff15p$chr,"_", myDiff15p$start)
# print(load(paste0(rdata.path,"Toothfish_2024_pool_muscle_methylation_DMS.Rdata")))
# Pooled_2024_muscle_DMS <- paste0(myDiff15p$chr,"_", myDiff15p$start)
# print(load(paste0(rdata.path,"Toothfish_2024_pool_fin_methylation_DMS.Rdata")))
# Pooled_2024_fin_DMS <- paste0(myDiff15p$chr,"_", myDiff15p$start)
# print(load(paste0(rdata.path,"Toothfish_tissue_comparison_methylation_DMS.Rdata")))
# Toothfish_ALL_tissue_DMS <- paste0(myDiff15p$chr,"_", myDiff15p$start)
```

## Histogram full vs random

```{r}
sort(table(factor(stringr::str_extract(Toothfish_2018_muscle_DMS, "[^_]+"))), decreasing = TRUE)
sort(table(factor(stringr::str_extract(Toothfish_2024_muscle_DMS, "[^_]+"))), decreasing = TRUE)
sort(table(factor(stringr::str_extract(Pooled_2024_muscle_DMS, "[^_]+"))), decreasing = TRUE)
sort(table(factor(stringr::str_extract(Toothfish_ALL_muscle_DMS, "[^_]+"))), decreasing = TRUE)
sort(table(factor(stringr::str_extract(Toothfish_2024_fin_DMS, "[^_]+"))), decreasing = TRUE)
sort(table(factor(stringr::str_extract(Pooled_2024_fin_DMS, "[^_]+"))), decreasing = TRUE)
sort(table(factor(stringr::str_extract(Toothfish_ALL_fin_DMS, "[^_]+"))), decreasing = TRUE)
sort(table(factor(stringr::str_extract(Toothfish_ALL_tissue_DMS, "[^_]+"))), decreasing = TRUE)

DMS.df <- data.frame(
  DMS = c(length(Toothfish_2018_muscle_DMS),
          length(Toothfish_2024_muscle_DMS),
          length(Toothfish_ALL_muscle_DMS),
          length(Toothfish_2024_fin_DMS),
          length(Toothfish_ALL_fin_DMS),
          length(Pooled_2024_muscle_DMS),
          length(Pooled_2024_fin_DMS),
          length(Toothfish_ALL_tissue_DMS),
          length(Toothfish_2018_muscle_RANDOM_DMS),
          length(Toothfish_2024_muscle_RANDOM_DMS),
          length(Toothfish_ALL_muscle_RANDOM_DMS),
          length(Toothfish_2024_fin_RANDOM_DMS),
          length(Toothfish_ALL_fin_RANDOM_DMS),
          length(Pooled_2024_muscle_RANDOM_DMS),
          length(Pooled_2024_fin_RANDOM_DMS),
          length(Toothfish_ALL_tissue_RANDOM_DMS)),
  CHR = c(length(unique(stringr::str_extract(Toothfish_2018_muscle_DMS, "[^_]+"))),
          length(unique(stringr::str_extract(Toothfish_2024_muscle_DMS, "[^_]+"))),
          length(unique(stringr::str_extract(Toothfish_ALL_muscle_DMS, "[^_]+"))),
          length(unique(stringr::str_extract(Toothfish_2024_fin_DMS, "[^_]+"))),
          length(unique(stringr::str_extract(Toothfish_ALL_fin_DMS, "[^_]+"))),
          length(unique(stringr::str_extract(Pooled_2024_muscle_DMS, "[^_]+"))),
          length(unique(stringr::str_extract(Pooled_2024_fin_DMS, "[^_]+"))),
          length(unique(stringr::str_extract(Toothfish_ALL_tissue_DMS, "[^_]+"))),
          length(unique(stringr::str_extract(Toothfish_2018_muscle_RANDOM_DMS, "[^_]+"))),
          length(unique(stringr::str_extract(Toothfish_2024_muscle_RANDOM_DMS, "[^_]+"))),
          length(unique(stringr::str_extract(Toothfish_ALL_muscle_RANDOM_DMS, "[^_]+"))),
          length(unique(stringr::str_extract(Toothfish_2024_fin_RANDOM_DMS, "[^_]+"))),
          length(unique(stringr::str_extract(Toothfish_ALL_fin_RANDOM_DMS, "[^_]+"))),
          length(unique(stringr::str_extract(Pooled_2024_muscle_RANDOM_DMS, "[^_]+"))),
          length(unique(stringr::str_extract(Pooled_2024_fin_RANDOM_DMS, "[^_]+"))),
          length(unique(stringr::str_extract(Toothfish_ALL_tissue_RANDOM_DMS, "[^_]+")))),
  Dataset = c("Toothfish_2018_muscle_DMS",
          "Toothfish_2024_muscle_DMS", 
          "Toothfish_ALL_muscle_DMS",
          "Toothfish_2024_fin_DMS", 
          "Toothfish_ALL_fin_DMS", 
          "Pooled_2024_muscle_DMS", 
          "Pooled_2024_fin_DMS",
          "Toothfish_ALL_tissue_DMS", 
          "Toothfish_2018_muscle_RANDOM_DMS", 
          "Toothfish_2024_muscle_RANDOM_DMS",
          "Toothfish_ALL_muscle_RANDOM_DMS", 
          "Toothfish_2024_fin_RANDOM_DMS", 
          "Toothfish_ALL_fin_RANDOM_DMS", 
          "Pooled_2024_muscle_RANDOM_DMS", 
          "Pooled_2024_fin_RANDOM_DMS", 
          "Toothfish_ALL_tissue_RANDOM_DMS"
          ),
  dataset.colour = as.character(rep(1:8, 2))
  )

dataset.levels <- c("Toothfish_ALL_tissue_DMS",
          "Toothfish_ALL_tissue_RANDOM_DMS",
          "Toothfish_2018_muscle_DMS",
          "Toothfish_2018_muscle_RANDOM_DMS",
          "Toothfish_2024_muscle_DMS", 
          "Toothfish_2024_muscle_RANDOM_DMS",
          "Toothfish_ALL_muscle_DMS",
          "Toothfish_ALL_muscle_RANDOM_DMS", 
          "Toothfish_2024_fin_DMS", 
          "Toothfish_2024_fin_RANDOM_DMS",
          "Toothfish_ALL_fin_DMS",
          "Toothfish_ALL_fin_RANDOM_DMS", 
          "Pooled_2024_muscle_DMS", 
          "Pooled_2024_muscle_RANDOM_DMS", 
          "Pooled_2024_fin_DMS",
          "Pooled_2024_fin_RANDOM_DMS"
          )

DMS.df$Dataset <- factor(DMS.df$Dataset, levels = dataset.levels)
DMS.df <- DMS.df[order(DMS.df$Dataset),]

hist.plot <-  ggplot2::ggplot(DMS.df, ggplot2::aes(x = Dataset, y = DMS, fill = dataset.colour)) + 
  ggplot2::geom_col() +
  ggplot2::geom_text(ggplot2::aes(x = Dataset, y = DMS + 1000, label = CHR)) +
  ggplot2::scale_fill_manual(values = c(as.character(colours6m), "blue", "orange")) +
  ggplot2::theme_bw() + 
  ggplot2::theme(axis.title.x = ggplot2::element_blank(), 
                 legend.position = "none",
                 axis.text.y = ggplot2::element_text(size = 15),
                 axis.title.y = ggplot2::element_text(size = 15),
                 axis.text.x = ggplot2::element_text(size = 7, angle = 90, hjust = 1))
print(hist.plot)

ggplot2::ggsave(hist.plot,
                filename =
                  paste0(figure.path,"Toothfish_DMS_full_vs_random_histogram2.png"), 
                width = 35, height = 15, units = "cm")


hist.plot <-  ggplot2::ggplot(DMS.df[!DMS.df$dataset.colour == "8",], 
                              ggplot2::aes(x = Dataset, y = DMS, fill = dataset.colour)) + 
  ggplot2::geom_col() +
  ggplot2::geom_text(ggplot2::aes(x = Dataset, y = DMS + 100, label = CHR)) +
  ggplot2::scale_fill_manual(values = c(as.character(colours6m), "blue")) +
  ggplot2::theme_bw() + 
  ggplot2::ylab("Number of DMS") +
  ggplot2::theme(axis.title.x = ggplot2::element_blank(), 
                 legend.position = "none",
                 axis.text.y = ggplot2::element_text(size = 15),
                 axis.title.y = ggplot2::element_text(size = 15),
                 axis.text.x = ggplot2::element_text(size = 8, angle = 90, hjust = 1))
print(hist.plot)

ggplot2::ggsave(hist.plot,
                filename =
                  paste0(figure.path,"Toothfish_DMS_full_vs_random_histogram_no_tissue2.png"), 
                width = 35, height = 15, units = "cm")
```

## Heatmap CpG per Chromsosomes per dataset

### ALL datasets

```{r}
DMS.df <- data.frame(
  CHR = c(stringr::str_extract(Toothfish_2018_muscle_DMS, "[^_]+"),
          stringr::str_extract(Toothfish_2024_muscle_DMS, "[^_]+"),
          stringr::str_extract(Toothfish_ALL_muscle_DMS, "[^_]+"),
          stringr::str_extract(Toothfish_2024_fin_DMS, "[^_]+"),
          stringr::str_extract(Toothfish_ALL_fin_DMS, "[^_]+"),
          stringr::str_extract(Pooled_2024_muscle_DMS, "[^_]+"),
          stringr::str_extract(Pooled_2024_fin_DMS, "[^_]+")
          ),
  Dataset = c(rep("Toothfish_2018_muscle_DMS", length(Toothfish_2018_muscle_DMS)),
          rep("Toothfish_2024_muscle_DMS", length(Toothfish_2024_muscle_DMS)),
          rep("Toothfish_ALL_muscle_DMS", length(Toothfish_ALL_muscle_DMS)),
          rep("Toothfish_2024_fin_DMS", length(Toothfish_2024_fin_DMS)),
          rep("Toothfish_ALL_fin_DMS", length(Toothfish_ALL_fin_DMS)),
          rep("Pooled_2024_muscle_DMS", length(Pooled_2024_muscle_DMS)),
          rep("Pooled_2024_fin_DMS", length(Pooled_2024_fin_DMS))
          )
)
DMS.df <- DMS.df[as.integer(gsub(pattern = "JAAKFY01", replacement = "", x = DMS.df$CHR)) < 35,]


plot.data <- DMS.df %>%
      # dplyr::distinct(CHR, Dataset) %>%
      dplyr::mutate(
        n = rep(1, n()),
        Dataset = stringi::stri_join("DATASET_", Dataset)
        ) %>%
      data.table::as.data.table(.) %>%
      data.table::dcast.data.table(
        data = .,
        formula = CHR ~ Dataset,
        value.var = "n",
        fill = 0
      ) %>%
      tidyr::gather(key = "dataset", value = "n", -CHR,
                factor_key = FALSE) %>%
      data.frame(.) 


heat.plot <- ggplot2::ggplot(data = plot.data, ggplot2::aes(x = dataset )) + 
    ggplot2::geom_tile(data = plot.data, 
                       ggplot2::aes(x = dataset, y = CHR, fill = n)) +
    ggplot2::scale_fill_gradient2(low = "blue", mid = "yellow", high = "red", 
                                  midpoint = 100, guide = "colourbar") +
    ggplot2::scale_y_discrete(drop = FALSE) +
    ggplot2::theme(legend.position = "top",
                   axis.title.y = ggplot2::element_blank(),
                   axis.title.x = ggplot2::element_blank(),
                   axis.text.x = ggplot2::element_text(face = "bold", color = "black", 
                                        size = 10, angle = 90, hjust = 1, vjust = 0.5))
print(heat.plot)

ggplot2::ggsave(heat.plot, bg = "white",
                filename = paste0(figure.path,
                   "Toothfish_DMS_on_common_chromosomes_all_datasets.png"), 
                width = 60, height = 30, units = "cm")
```


### 2024

```{r}
DMS.df <- data.frame(
  CHR = c(stringr::str_extract(Toothfish_2018_muscle_DMS, "[^_]+"),
          stringr::str_extract(Toothfish_2024_muscle_DMS, "[^_]+"),
          stringr::str_extract(Toothfish_2024_fin_DMS, "[^_]+"),
          stringr::str_extract(Pooled_2024_muscle_DMS, "[^_]+"),
          stringr::str_extract(Pooled_2024_fin_DMS, "[^_]+")
          ),
  Dataset = c(rep("Toothfish_2018_muscle_DMS", length(Toothfish_2018_muscle_DMS)),
          rep("Toothfish_2024_muscle_DMS", length(Toothfish_2024_muscle_DMS)),
          rep("Toothfish_2024_fin_DMS", length(Toothfish_2024_fin_DMS)),
          rep("Pooled_2024_muscle_DMS", length(Pooled_2024_muscle_DMS)),
          rep("Pooled_2024_fin_DMS", length(Pooled_2024_fin_DMS))
          )
)
# DMS.df <- data.frame(
#   CHR = c(
#           stringr::str_extract(Toothfish_2024_muscle_DMS, "[^_]+"),
#           stringr::str_extract(Toothfish_2024_fin_DMS, "[^_]+"),
#           stringr::str_extract(Pooled_2024_muscle_DMS, "[^_]+"),
#           stringr::str_extract(Pooled_2024_fin_DMS, "[^_]+")
#           ),
#   Dataset = c(
#           rep("Toothfish_2024_muscle_DMS", length(Toothfish_2024_muscle_DMS)),
#           rep("Toothfish_2024_fin_DMS", length(Toothfish_2024_fin_DMS)),
#           rep("Pooled_2024_muscle_DMS", length(Pooled_2024_muscle_DMS)),
#           rep("Pooled_2024_fin_DMS", length(Pooled_2024_fin_DMS))
#           )
# )
DMS.df <- DMS.df[as.integer(gsub(pattern = "JAAKFY01", replacement = "", x = DMS.df$CHR)) < 35,]


plot.data <- DMS.df %>%
      # dplyr::distinct(CHR, Dataset) %>%
      dplyr::mutate(
        n = rep(1, n()),
        Dataset = stringi::stri_join("DATASET_", Dataset)
        ) %>%
      data.table::as.data.table(.) %>%
      data.table::dcast.data.table(
        data = .,
        formula = CHR ~ Dataset,
        value.var = "n",
        fill = 0
      ) %>%
      tidyr::gather(key = "dataset", value = "n", -CHR,
                factor_key = FALSE) %>%
      data.frame(.) 


heat.plot <- ggplot2::ggplot(data = plot.data, ggplot2::aes(x = dataset )) + 
    ggplot2::geom_tile(data = plot.data, 
                       ggplot2::aes(x = dataset, y = CHR, fill = n)) +
    ggplot2::scale_fill_gradient2(low = "blue", mid = "yellow", high = "red", 
                                  midpoint = 70, guide = "colourbar") +
    ggplot2::scale_y_discrete(drop = FALSE) +
    ggplot2::theme(legend.position = "top",
                   axis.title.y = ggplot2::element_blank(),
                   axis.title.x = ggplot2::element_blank(),
                   axis.text.x = ggplot2::element_text(face = "bold", color = "black", 
                                        size = 10, angle = 90, hjust = 1, vjust = 0.5))
print(heat.plot)

ggplot2::ggsave(heat.plot, bg = "white",
                filename = paste0(figure.path,
                   "Toothfish_DMS_on_common_chromosomes_2024.png"), 
                width = 60, height = 30, units = "cm")
```

### Top 100

```{r}
DMS.df <- data.frame(
  CHR = c(stringr::str_extract(Toothfish_2018_muscle_TOP100_DMS, "[^_]+"),
          stringr::str_extract(Toothfish_2024_muscle_TOP100_DMS, "[^_]+"),
          stringr::str_extract(Toothfish_ALL_muscle_TOP100_DMS, "[^_]+"),
          stringr::str_extract(Toothfish_2024_fin_TOP100_DMS, "[^_]+"),
          stringr::str_extract(Toothfish_ALL_fin_TOP100_DMS, "[^_]+"),
          stringr::str_extract(Pooled_2024_muscle_TOP100_DMS, "[^_]+"),
          stringr::str_extract(Pooled_2024_fin_TOP100_DMS, "[^_]+"),
          stringr::str_extract(Toothfish_ALL_tissue_TOP100_DMS, "[^_]+")
          ),
  Dataset = c(rep("Toothfish_2018_muscle_TOP100_DMS", length(Toothfish_2018_muscle_TOP100_DMS)),
          rep("Toothfish_2024_muscle_TOP100_DMS", length(Toothfish_2024_muscle_TOP100_DMS)),
          rep("Toothfish_ALL_muscle_TOP100_DMS", length(Toothfish_ALL_muscle_TOP100_DMS)),
          rep("Toothfish_2024_fin_TOP100_DMS", length(Toothfish_2024_fin_TOP100_DMS)),
          rep("Toothfish_ALL_fin_TOP100_DMS", length(Toothfish_ALL_fin_TOP100_DMS)),
          rep("Pooled_2024_muscle_TOP100_DMS", length(Pooled_2024_muscle_TOP100_DMS)),
          rep("Pooled_2024_fin_TOP100_DMS", length(Pooled_2024_fin_TOP100_DMS)),
          rep("Toothfish_ALL_tissue_TOP100_DMS", length(Toothfish_ALL_tissue_TOP100_DMS))
          )
)
DMS.df <- DMS.df[as.integer(gsub(pattern = "JAAKFY01", replacement = "", x = DMS.df$CHR)) < 35,]


plot.data <- DMS.df %>%
      # dplyr::distinct(CHR, Dataset) %>%
      dplyr::mutate(
        n = rep(1, n()),
        Dataset = stringi::stri_join("DATASET_", Dataset)
        ) %>%
      data.table::as.data.table(.) %>%
      data.table::dcast.data.table(
        data = .,
        formula = CHR ~ Dataset,
        value.var = "n",
        fill = 0
      ) %>%
      tidyr::gather(key = "dataset", value = "n", -CHR,
                factor_key = FALSE) %>%
      data.frame(.) 


heat.plot <- ggplot2::ggplot(data = plot.data, ggplot2::aes(x = dataset )) + 
    ggplot2::geom_tile(data = plot.data, 
                       ggplot2::aes(x = dataset, y = CHR, fill = n)) +
    ggplot2::scale_fill_gradient2(low = "blue", mid = "yellow", high = "red",
                                  midpoint = 25, guide = "colourbar") +
    ggplot2::scale_y_discrete(drop = FALSE) +
    ggplot2::theme(legend.position = "top",
                   axis.title.y = ggplot2::element_blank(),
                   axis.title.x = ggplot2::element_blank(),
                   axis.text.x = ggplot2::element_text(face = "bold", color = "black", 
                                        size = 10, angle = 90, hjust = 1, vjust = 0.5))
print(heat.plot)

ggplot2::ggsave(heat.plot, bg = "white",
                filename = paste0(figure.path,
                   "Toothfish_DMS_on_common_chromosomes_TOP100.png"), 
                width = 60, height = 30, units = "cm")
```


## ALL dataset
```{r}
DMS.df <- data.frame(
  DMS = c(Toothfish_2018_muscle_DMS,
          Toothfish_2024_muscle_DMS,
          Toothfish_ALL_muscle_DMS,
          Toothfish_2024_fin_DMS,
          Toothfish_ALL_fin_DMS,
          Pooled_2024_muscle_DMS,
          Pooled_2024_fin_DMS,
          Toothfish_ALL_tissue_DMS),
  Dataset = c(rep("Toothfish_2018_muscle_DMS", length(Toothfish_2018_muscle_DMS)),
          rep("Toothfish_2024_muscle_DMS", length(Toothfish_2024_muscle_DMS)),
          rep("Toothfish_ALL_muscle_DMS", length(Toothfish_ALL_muscle_DMS)),
          rep("Toothfish_2024_fin_DMS", length(Toothfish_2024_fin_DMS)),
          rep("Toothfish_ALL_fin_DMS", length(Toothfish_ALL_fin_DMS)),
          rep("Pooled_2024_muscle_DMS", length(Pooled_2024_muscle_DMS)),
          rep("Pooled_2024_fin_DMS", length(Pooled_2024_fin_DMS)),
          rep("Toothfish_ALL_tissue_DMS", length(Toothfish_ALL_tissue_DMS)))
)


plot.data <- DMS.df %>%
      dplyr::distinct(DMS, Dataset) %>%
      dplyr::mutate(
        n = rep(1, n()),
        Dataset = stringi::stri_join("DATASET_", Dataset)
        ) %>%
      data.table::as.data.table(.) %>%
      data.table::dcast.data.table(
        data = .,
        formula = DMS ~ Dataset,
        value.var = "n",
        fill = 0
      ) %>%
      data.frame(.)

n.pop = length(unique(DMS.df$Dataset))
Upsetplot <- UpSetR::upset(
      plot.data,
      nsets = n.pop,
      order.by = "freq",
      empty.intersections = NULL
    )
print(Upsetplot)

ggplot2::ggsave(ggplotify::as.ggplot(Upsetplot),
                filename =
                  paste0(figure.path,"Toothfish_UPSET_DMS_overlap_all.png"), 
                width = 35, height = 15, units = "cm")
```

## ALL dataset - no tissue

```{r}
DMS.df <- data.frame(
  DMS = c(Toothfish_2018_muscle_DMS,
          Toothfish_2024_muscle_DMS,
          Toothfish_ALL_muscle_DMS,
          Toothfish_2024_fin_DMS,
          Toothfish_ALL_fin_DMS,
          Pooled_2024_muscle_DMS,
          Pooled_2024_fin_DMS),
  Dataset = c(rep("Toothfish_2018_muscle_DMS", length(Toothfish_2018_muscle_DMS)),
          rep("Toothfish_2024_muscle_DMS", length(Toothfish_2024_muscle_DMS)),
          rep("Toothfish_ALL_muscle_DMS", length(Toothfish_ALL_muscle_DMS)),
          rep("Toothfish_2024_fin_DMS", length(Toothfish_2024_fin_DMS)),
          rep("Toothfish_ALL_fin_DMS", length(Toothfish_ALL_fin_DMS)),
          rep("Pooled_2024_muscle_DMS", length(Pooled_2024_muscle_DMS)),
          rep("Pooled_2024_fin_DMS", length(Pooled_2024_fin_DMS)))
)


plot.data <- DMS.df %>%
      dplyr::distinct(DMS, Dataset) %>%
      dplyr::mutate(
        n = rep(1, n()),
        Dataset = stringi::stri_join("DATASET_", Dataset)
        ) %>%
      data.table::as.data.table(.) %>%
      data.table::dcast.data.table(
        data = .,
        formula = DMS ~ Dataset,
        value.var = "n",
        fill = 0
      ) %>%
      data.frame(.)

n.pop = length(unique(DMS.df$Dataset))
Upsetplot <- UpSetR::upset(
      plot.data,
      nsets = n.pop,
      order.by = "freq",
      empty.intersections = NULL
    )
print(Upsetplot)

ggplot2::ggsave(ggplotify::as.ggplot(Upsetplot),
                filename =
                  paste0(figure.path,"Toothfish_UPSET_DMS_overlap_all_no_tissue.png"), 
                width = 35, height = 15, units = "cm")
```


## ALL dataset - no tissue or pool

```{r}
DMS.df <- data.frame(
  DMS = c(Toothfish_2024_muscle_DMS,
          Toothfish_ALL_muscle_DMS,
          Toothfish_2024_fin_DMS,
          Toothfish_ALL_fin_DMS),
  Dataset = c(rep("Toothfish_2024_muscle_DMS", length(Toothfish_2024_muscle_DMS)),
          rep("Toothfish_ALL_muscle_DMS", length(Toothfish_ALL_muscle_DMS)),
          rep("Toothfish_2024_fin_DMS", length(Toothfish_2024_fin_DMS)),
          rep("Toothfish_ALL_fin_DMS", length(Toothfish_ALL_fin_DMS)))
)


plot.data <- DMS.df %>%
      dplyr::distinct(DMS, Dataset) %>%
      dplyr::mutate(
        n = rep(1, n()),
        Dataset = stringi::stri_join("DATASET_", Dataset)
        ) %>%
      data.table::as.data.table(.) %>%
      data.table::dcast.data.table(
        data = .,
        formula = DMS ~ Dataset,
        value.var = "n",
        fill = 0
      ) %>%
      data.frame(.)

n.pop = length(unique(DMS.df$Dataset))
Upsetplot <- UpSetR::upset(
      plot.data,
      nsets = n.pop,
      order.by = "freq",
      empty.intersections = NULL
    )
print(Upsetplot)

ggplot2::ggsave(ggplotify::as.ggplot(Upsetplot),
                filename =
                  paste0(figure.path,"Toothfish_UPSET_DMS_overlap_all_no_pool.png"), 
                width = 35, height = 15, units = "cm")
```


## Muscle dataset
```{r}
DMS.df <- data.frame(
  DMS = c(Toothfish_2018_muscle_DMS,
          Toothfish_2024_muscle_DMS,
          Toothfish_ALL_muscle_DMS
          ),
  Dataset = c(
          rep("Toothfish_2018_muscle_DMS", length(Toothfish_2018_muscle_DMS)),
          rep("Toothfish_2024_muscle_DMS", length(Toothfish_2024_muscle_DMS)),
          rep("Toothfish_ALL_muscle_DMS", length(Toothfish_ALL_muscle_DMS))
          )
)


plot.data <- DMS.df %>%
      dplyr::distinct(DMS, Dataset) %>%
      dplyr::mutate(
        n = rep(1, n()),
        Dataset = stringi::stri_join("DATASET_", Dataset)
        ) %>%
      data.table::as.data.table(.) %>%
      data.table::dcast.data.table(
        data = .,
        formula = DMS ~ Dataset,
        value.var = "n",
        fill = 0
      ) %>%
      data.frame(.)

n.pop = length(unique(DMS.df$Dataset))
Upsetplot <- UpSetR::upset(
      plot.data,
      nsets = n.pop,
      order.by = "freq",
      empty.intersections = NULL
    )
print(Upsetplot)

ggplot2::ggsave(ggplotify::as.ggplot(Upsetplot),
                filename =
                  paste0(figure.path,"Toothfish_UPSET_DMS_overlap_Muscle.png"), 
                width = 35, height = 15, units = "cm")
```



## Fin dataset

```{r}
DMS.df <- data.frame(
  DMS = c(Toothfish_2024_fin_DMS,
          Toothfish_ALL_fin_DMS),
  Dataset = c(
          rep("Toothfish_2024_fin_DMS", length(Toothfish_2024_fin_DMS)),
          rep("Toothfish_ALL_fin_DMS", length(Toothfish_ALL_fin_DMS))
          )
  )


plot.data <- DMS.df %>%
      dplyr::distinct(DMS, Dataset) %>%
      dplyr::mutate(
        n = rep(1, n()),
        Dataset = stringi::stri_join("DATASET_", Dataset)
        ) %>%
      data.table::as.data.table(.) %>%
      data.table::dcast.data.table(
        data = .,
        formula = DMS ~ Dataset,
        value.var = "n",
        fill = 0
      ) %>%
      data.frame(.)

n.pop = length(unique(DMS.df$Dataset))
Upsetplot <- UpSetR::upset(
      plot.data,
      nsets = n.pop,
      order.by = "freq",
      empty.intersections = NULL
    )
print(Upsetplot)

ggplot2::ggsave(ggplotify::as.ggplot(Upsetplot),
                filename =
                  paste0(figure.path,"Toothfish_UPSET_DMS_overlap_FIn.png"), 
                width = 35, height = 15, units = "cm")
```



## 2024 datasets
```{r}
DMS.df <- data.frame(
  DMS = c(
          Toothfish_2024_muscle_DMS,
          Toothfish_2024_fin_DMS,
          Pooled_2024_muscle_DMS,
          Pooled_2024_fin_DMS),
  Dataset = c(
          rep("Toothfish_2024_muscle_DMS", length(Toothfish_2024_muscle_DMS)),
          rep("Toothfish_2024_fin_DMS", length(Toothfish_2024_fin_DMS)),
          rep("Pooled_2024_muscle_DMS", length(Pooled_2024_muscle_DMS)),
          rep("Pooled_2024_fin_DMS", length(Pooled_2024_fin_DMS)))
)


plot.data <- DMS.df %>%
      dplyr::distinct(DMS, Dataset) %>%
      dplyr::mutate(
        n = rep(1, n()),
        Dataset = stringi::stri_join("DATASET_", Dataset)
        ) %>%
      data.table::as.data.table(.) %>%
      data.table::dcast.data.table(
        data = .,
        formula = DMS ~ Dataset,
        value.var = "n",
        fill = 0
      ) %>%
      data.frame(.)

n.pop = length(unique(DMS.df$Dataset))
Upsetplot <- UpSetR::upset(
      plot.data,
      nsets = n.pop,
      order.by = "freq",
      empty.intersections = NULL
    )
print(Upsetplot)

ggplot2::ggsave(ggplotify::as.ggplot(Upsetplot),
                filename =
                  paste0(figure.path,"Toothfish_UPSET_DMS_overlap_all_2024.png"), 
                width = 35, height = 15, units = "cm")
```


## 2018/2024 ALL dataset

```{r}
DMS.df <- data.frame(
  DMS = c(Toothfish_2018_muscle_DMS,
          Toothfish_ALL_muscle_DMS,
          Toothfish_ALL_fin_DMS),
  Dataset = c(rep("Toothfish_2018_muscle_DMS", length(Toothfish_2018_muscle_DMS)),
          rep("Toothfish_ALL_muscle_DMS", length(Toothfish_ALL_muscle_DMS)),
          rep("Toothfish_ALL_fin_DMS", length(Toothfish_ALL_fin_DMS)))
)


plot.data <- DMS.df %>%
      dplyr::distinct(DMS, Dataset) %>%
      dplyr::mutate(
        n = rep(1, n()),
        Dataset = stringi::stri_join("DATASET_", Dataset)
        ) %>%
      data.table::as.data.table(.) %>%
      data.table::dcast.data.table(
        data = .,
        formula = DMS ~ Dataset,
        value.var = "n",
        fill = 0
      ) %>%
      data.frame(.)

n.pop = length(unique(DMS.df$Dataset))
Upsetplot <- UpSetR::upset(
      plot.data,
      nsets = n.pop,
      order.by = "freq",
      empty.intersections = NULL
    )
print(Upsetplot)

ggplot2::ggsave(ggplotify::as.ggplot(Upsetplot),
                filename =
                  paste0(figure.path,"Toothfish_UPSET_DMS_overlap_all_2018-2024.png"), 
                width = 35, height = 15, units = "cm")
```






## POOLED - Muscle dataset
```{r}
DMS.df <- data.frame(
  DMS = c(
          Toothfish_2024_muscle_DMS,
          Toothfish_ALL_muscle_DMS,
          Pooled_2024_muscle_DMS),
  Dataset = c(
          rep("Toothfish_2024_muscle_DMS", length(Toothfish_2024_muscle_DMS)),
          rep("Toothfish_ALL_muscle_DMS", length(Toothfish_ALL_muscle_DMS)),
          rep("Pooled_2024_muscle_DMS", length(Pooled_2024_muscle_DMS)))
)


plot.data <- DMS.df %>%
      dplyr::distinct(DMS, Dataset) %>%
      dplyr::mutate(
        n = rep(1, n()),
        Dataset = stringi::stri_join("DATASET_", Dataset)
        ) %>%
      data.table::as.data.table(.) %>%
      data.table::dcast.data.table(
        data = .,
        formula = DMS ~ Dataset,
        value.var = "n",
        fill = 0
      ) %>%
      data.frame(.)

n.pop = length(unique(DMS.df$Dataset))
Upsetplot <- UpSetR::upset(
      plot.data,
      nsets = n.pop,
      order.by = "freq",
      empty.intersections = NULL
    )
print(Upsetplot)

ggplot2::ggsave(ggplotify::as.ggplot(Upsetplot),
                filename =
                  paste0(figure.path,"Toothfish_UPSET_DMS_overlap_Pooled_muscle.png"), 
                width = 35, height = 15, units = "cm")
```





## POOLED - Fin dataset
```{r}
DMS.df <- data.frame(
  DMS = c(Toothfish_2024_fin_DMS,
          Toothfish_ALL_fin_DMS,
          Pooled_2024_fin_DMS),
  Dataset = c(
          rep("Toothfish_2024_fin_DMS", length(Toothfish_2024_fin_DMS)),
          rep("Toothfish_ALL_fin_DMS", length(Toothfish_ALL_fin_DMS)),
          rep("Pooled_2024_fin_DMS", length(Pooled_2024_fin_DMS)))
)


plot.data <- DMS.df %>%
      dplyr::distinct(DMS, Dataset) %>%
      dplyr::mutate(
        n = rep(1, n()),
        Dataset = stringi::stri_join("DATASET_", Dataset)
        ) %>%
      data.table::as.data.table(.) %>%
      data.table::dcast.data.table(
        data = .,
        formula = DMS ~ Dataset,
        value.var = "n",
        fill = 0
      ) %>%
      data.frame(.)

n.pop = length(unique(DMS.df$Dataset))
Upsetplot <- UpSetR::upset(
      plot.data,
      nsets = n.pop,
      order.by = "freq",
      empty.intersections = NULL
    )
print(Upsetplot)

ggplot2::ggsave(ggplotify::as.ggplot(Upsetplot),
                filename =
                  paste0(figure.path,"Toothfish_UPSET_DMS_overlap_Pooled_fin.png"), 
                width = 35, height = 15, units = "cm")
```



## Top 100 
```{r}
length(unique(stringr::str_extract(Toothfish_2018_muscle_TOP100_DMS, "[^_]+")))
length(unique(stringr::str_extract(Toothfish_2024_muscle_TOP100_DMS, "[^_]+")))
length(unique(stringr::str_extract(Toothfish_ALL_muscle_TOP100_DMS, "[^_]+")))
length(unique(stringr::str_extract(Toothfish_2024_fin_TOP100_DMS, "[^_]+")))
length(unique(stringr::str_extract(Toothfish_ALL_fin_TOP100_DMS, "[^_]+")))
length(unique(stringr::str_extract(Pooled_2024_muscle_TOP100_DMS, "[^_]+")))
length(unique(stringr::str_extract(Pooled_2024_fin_TOP100_DMS, "[^_]+")))
length(unique(stringr::str_extract(Toothfish_ALL_tissue_TOP100_DMS, "[^_]+")))

sort(table(factor(stringr::str_extract(Toothfish_2018_muscle_TOP100_DMS, "[^_]+"))), decreasing = TRUE)
sort(table(factor(stringr::str_extract(Toothfish_2024_muscle_TOP100_DMS, "[^_]+"))), decreasing = TRUE)
sort(table(factor(stringr::str_extract(Toothfish_ALL_muscle_TOP100_DMS, "[^_]+"))), decreasing = TRUE)
sort(table(factor(stringr::str_extract(Toothfish_2024_fin_TOP100_DMS, "[^_]+"))), decreasing = TRUE)
sort(table(factor(stringr::str_extract(Toothfish_ALL_fin_TOP100_DMS, "[^_]+"))), decreasing = TRUE)
sort(table(factor(stringr::str_extract(Pooled_2024_muscle_TOP100_DMS, "[^_]+"))), decreasing = TRUE)
sort(table(factor(stringr::str_extract(Pooled_2024_fin_TOP100_DMS, "[^_]+"))), decreasing = TRUE)
sort(table(factor(stringr::str_extract(Toothfish_ALL_tissue_TOP100_DMS, "[^_]+"))), decreasing = TRUE)


DMS.df <- data.frame(
  DMS = c(
    Toothfish_2018_muscle_TOP100_DMS,
    Toothfish_2024_muscle_TOP100_DMS,
    Toothfish_ALL_muscle_TOP100_DMS,
    Toothfish_2024_fin_TOP100_DMS,
    Toothfish_ALL_fin_TOP100_DMS,
    Pooled_2024_muscle_TOP100_DMS,
    Pooled_2024_fin_TOP100_DMS,
    Toothfish_ALL_tissue_TOP100_DMS
  ),
  Dataset = c(
          rep("Toothfish_2018_muscle_TOP100_DMS", length(Toothfish_2018_muscle_TOP100_DMS)),
          rep("Toothfish_2024_muscle_TOP100_DMS", length(Toothfish_2024_muscle_TOP100_DMS)),
          rep("Toothfish_ALL_muscle_TOP100_DMS", length(Toothfish_ALL_muscle_TOP100_DMS)),
          rep("Toothfish_2024_fin_TOP100_DMS", length(Toothfish_2024_fin_TOP100_DMS)),
          rep("Toothfish_ALL_fin_TOP100_DMS", length(Toothfish_ALL_fin_TOP100_DMS)),
          rep("Pooled_2024_muscle_TOP100_DMS", length(Pooled_2024_muscle_TOP100_DMS)),
          rep("Pooled_2024_fin_TOP100_DMS", length(Pooled_2024_fin_TOP100_DMS)),
          rep("Toothfish_ALL_tissue_TOP100_DMS", length(Toothfish_ALL_tissue_TOP100_DMS))
          )
)

# DMS.df <- data.frame(
#   DMS = c(
#     Toothfish_2024_muscle_TOP100_DMS,
#     Toothfish_2024_fin_TOP100_DMS,
#     Pooled_2024_muscle_TOP100_DMS,
#     Pooled_2024_fin_TOP100_DMS
#   ),
#   Dataset = c(
#           rep("Toothfish_2024_muscle_TOP100_DMS", length(Toothfish_2024_muscle_TOP100_DMS)),
#           rep("Toothfish_2024_fin_TOP100_DMS", length(Toothfish_2024_fin_TOP100_DMS)),
#           rep("Pooled_2024_muscle_TOP100_DMS", length(Pooled_2024_muscle_TOP100_DMS)),
#           rep("Pooled_2024_fin_TOP100_DMS", length(Pooled_2024_fin_TOP100_DMS))
#           )
# )




plot.data <- DMS.df %>%
      dplyr::distinct(DMS, Dataset) %>%
      dplyr::mutate(
        n = rep(1, n()),
        Dataset = stringi::stri_join("DATASET_", Dataset)
        ) %>%
      data.table::as.data.table(.) %>%
      data.table::dcast.data.table(
        data = .,
        formula = DMS ~ Dataset,
        value.var = "n",
        fill = 0
      ) %>%
      data.frame(.)

n.pop = length(unique(DMS.df$Dataset))
Upsetplot <- UpSetR::upset(
      plot.data,
      nsets = n.pop,
      order.by = "freq",
      empty.intersections = NULL
    )
print(Upsetplot)

sort(plot.data$DMS[rowSums(plot.data[-1]) > 1])
# JAAKFY010000002_21526720
# JAAKFY010000002_21526721
# JAAKFY010000002_21526744

ggplot2::ggsave(ggplotify::as.ggplot(Upsetplot),
                filename =
                  paste0(figure.path,"Toothfish_UPSET_DMS_overlap_TOP100.png"), 
                width = 35, height = 15, units = "cm")
```



## Randomised

```{r}
DMS.df <- data.frame(
  DMS = c(Toothfish_2018_muscle_RANDOM_DMS,
          Toothfish_2024_muscle_RANDOM_DMS,
          Toothfish_ALL_muscle_RANDOM_DMS,
          Toothfish_2024_fin_RANDOM_DMS,
          Toothfish_ALL_fin_RANDOM_DMS,
          Pooled_2024_muscle_RANDOM_DMS,
          Pooled_2024_fin_RANDOM_DMS,
          Toothfish_ALL_tissue_RANDOM_DMS),
  Dataset = c(rep("Toothfish_2018_muscle_DMS", length(Toothfish_2018_muscle_RANDOM_DMS)),
          rep("Toothfish_2024_muscle_DMS", length(Toothfish_2024_muscle_RANDOM_DMS)),
          rep("Toothfish_ALL_muscle_DMS", length(Toothfish_ALL_muscle_RANDOM_DMS)),
          rep("Toothfish_2024_fin_DMS", length(Toothfish_2024_fin_RANDOM_DMS)),
          rep("Toothfish_ALL_fin_DMS", length(Toothfish_ALL_fin_RANDOM_DMS)),
          rep("Pooled_2024_muscle_DMS", length(Pooled_2024_muscle_RANDOM_DMS)),
          rep("Pooled_2024_fin_DMS", length(Pooled_2024_fin_RANDOM_DMS)),
          rep("Toothfish_ALL_tissue_DMS", length(Toothfish_ALL_tissue_RANDOM_DMS)))
)


plot.data <- DMS.df %>%
      dplyr::distinct(DMS, Dataset) %>%
      dplyr::mutate(
        n = rep(1, n()),
        Dataset = stringi::stri_join("DATASET_", Dataset)
        ) %>%
      data.table::as.data.table(.) %>%
      data.table::dcast.data.table(
        data = .,
        formula = DMS ~ Dataset,
        value.var = "n",
        fill = 0
      ) %>%
      data.frame(.)

n.pop = length(unique(DMS.df$Dataset))
Upsetplot <- UpSetR::upset(
      plot.data,
      nsets = n.pop,
      order.by = "freq",
      empty.intersections = NULL
    )
print(Upsetplot)

ggplot2::ggsave(ggplotify::as.ggplot(Upsetplot),
                filename =
                  paste0(figure.path,"Toothfish_UPSET_DMS_overlap_all_RANDOM.png"), 
                width = 35, height = 15, units = "cm")
```



## All muscle with top100 DMS

```{r}
print(load(paste0(rdata.path,"Toothfish_ALL_muscle_methylkit_object.Rdata")))
print(load(paste0(rdata.path,"Toothfish_metadata.Rdata"))) 
metadata.new <- metadata[metadata$Genotype_names %in% meth.ALLM.norm15L@sample.ids,]
metadata.new <- metadata.new[order(match(metadata.new$Genotype_names,
                                         meth.ALLM.norm15L@sample.ids)),]

meth.cpg <- paste0(meth.ALLM.norm15L$chr, "_", meth.ALLM.norm15L$start) #283819 
DMS.cpg <- unique(c(
    Toothfish_2024_muscle_TOP100_DMS,
    Toothfish_ALL_muscle_TOP100_DMS,
    Toothfish_2024_fin_TOP100_DMS,
    Toothfish_ALL_fin_TOP100_DMS)) #336

sum(meth.cpg %in% DMS.cpg) #204
meth.DMS <- meth.ALLM.norm15L[meth.cpg %in% DMS.cpg,]
```



### Heatmap

```{r}
meth.diff.df <- as.data.frame(t(methylKit::percMethylation(meth.DMS)))
loc.names <- paste0(meth.DMS$chr,"_",meth.DMS$start)
colnames(meth.diff.df) <- loc.names

meth.diff.df$individual <- rownames(meth.diff.df)
pop <- metadata.new$pop
meth.diff.df$pop <- as.character(pop)


heat.plot.data <- meth.diff.df %>%
  tibble::as_tibble() %>%
  tidyr::gather(key = "CpG", value = "Methylation_percentage", -individual, -pop,
                factor_key = FALSE) %>%
  dplyr::group_by(pop, CpG) %>%
  dplyr::mutate(CpG_mean = mean(Methylation_percentage, na.rm = TRUE)) %>%
  dplyr::arrange(pop, individual, CpG) %>%
  dplyr::mutate(individual = factor(individual, levels = unique(individual))) %>%
  dplyr::mutate(CpG = factor(CpG, levels = unique(CpG)))
cpgs <- heat.plot.data$CpG[heat.plot.data$individual == unique(heat.plot.data$individual)[1]]
cpgs_mean <- heat.plot.data$CpG_mean[heat.plot.data$individual == unique(heat.plot.data$individual)[1]]

cpgs <- cpgs[order(cpgs_mean, decreasing = TRUE)]
heat.plot.data$CpG <- factor(heat.plot.data$CpG, levels = c("pop", unique(as.character(cpgs))))
heat.plot.data <- heat.plot.data[order(heat.plot.data$CpG),]

heat.plot <- ggplot2::ggplot(data = heat.plot.data, ggplot2::aes(x = individual)) + 
    ggplot2::geom_tile(data = heat.plot.data, 
                       ggplot2::aes(x = individual, y = CpG, fill = Methylation_percentage)) +
    ggplot2::scale_fill_gradient2(low = "blue", mid = "yellow", high = "red", 
                                  midpoint = 50, guide = "colourbar") +
    ggnewscale::new_scale_fill() +
    ggplot2::geom_tile(data = heat.plot.data, ggplot2::aes(x = individual, y = "pop", fill = pop,
                                    height = length(unique(heat.plot.data$CpG))*0.02)) +
    ggplot2::scale_fill_manual(values = colours6m) +
    ggplot2::scale_y_discrete(drop = FALSE) +
    ggplot2::theme(legend.position = "top",
                   axis.title.y = ggplot2::element_blank(),
                   axis.title.x = ggplot2::element_blank(),
                   axis.text.x = ggplot2::element_text(face = "bold", color = "black", 
                                        size = 10, angle = 90, hjust = 1, vjust = 0.5))
print(heat.plot)

ggplot2::ggsave(heat.plot, bg = "white",
                filename = paste0(figure.path,
                                  "Toothfish_ALL_muscle_BASED_ON_TOP100_DMS_HEATMAP.png"), 
                width = 30, height = 30, units = "cm")
```


### DMS Clustering
```{r}
methdif.perc <- methylKit::percMethylation(meth.DMS)
methdif.perc2 <- zoo::na.aggregate(methdif.perc, by = meth.DMS@treatment)
PCobj <- prcomp(methdif.perc2, retx = TRUE, scale. = TRUE)

var_frac <- PCobj$sdev/sum(PCobj$sdev)
print(paste0("Total variance explained: ",signif(sum(var_frac[1:3]) * 100, 3), "%")) 
barplot(var_frac[1:3], main = "variance explained", 
        col = heat.colors(length(var_frac[1:3])))


identical(metadata.new$Genotype_names, meth.DMS@sample.ids)
identical(metadata.new$Genotype_names, rownames(PCobj$rotation))
pop <- metadata.new$pop

plot.title <- "Toothfish methylation - ALL Muscle (204 TOP100 DMS)"

data <- data.frame(ind = rownames(PCobj$rotation),PCobj$rotation,  pop = pop)
data <- merge(data, aggregate(cbind(mean.x = PC1, mean.y = PC2) ~
                                pop, data, mean), by = "pop")

pca.plot <- ggplot2::ggplot(data, ggplot2::aes(PC1, PC2, color = pop)) +
  ggplot2::geom_point(size = 4) + 
  ggplot2::geom_point(ggplot2::aes(x = mean.x,
                                   y = mean.y, color = pop),
                      size = 7,
                      shape = 17) +
  ggplot2::scale_colour_manual(values = colours6m) +
  ggplot2::geom_hline(yintercept = 0) +
  ggplot2::geom_vline(xintercept = 0) + 
  ggplot2::labs(
    subtitle = paste0("Total variance explained: ",
                      signif(sum(var_frac[1:3]) * 100, 3), "%"),
    y = paste0("PC2; ", signif(var_frac[2] * 100, digits = 2), " %"),
    x = paste0("PC1; ", signif(var_frac[1] * 100, digits = 2), " %"),
    title = plot.title,
    caption = "") + 
  ggplot2::theme_minimal() +
  ggplot2::theme(
    axis.text = ggplot2::element_text(size = 15),
    axis.title.x = ggplot2::element_text(size = 20),
    axis.title.y = ggplot2::element_text(size = 20),
    legend.text = ggplot2::element_text(size = 15)
  )

print(pca.plot)
ggplot2::ggsave(pca.plot, bg = "white",
                filename = paste0(figure.path,"Toothfish_ALL_muscle_PCA_CpG_TOP100_DMS_PC12.png"), 
                width = 30, height = 15, units = "cm")



data <- data.frame(ind = rownames(PCobj$rotation),PCobj$rotation,  pop = pop)
data <- merge(data, aggregate(cbind(mean.x = PC1, mean.y = PC3) ~
                                pop, data, mean), by = "pop")

pca.plot <- ggplot2::ggplot(data, ggplot2::aes(PC1, PC3, color = pop)) +
  ggplot2::geom_point(size = 4) + 
  ggplot2::geom_point(ggplot2::aes(x = mean.x,
                                   y = mean.y, color = pop),
                      size = 7,
                      shape = 17) +
  ggplot2::scale_colour_manual(values = colours6m) +
  ggplot2::geom_hline(yintercept = 0) +
  ggplot2::geom_vline(xintercept = 0) + 
  ggplot2::labs(
    subtitle = paste0("Total variance explained: ",
                      signif(sum(var_frac[1:3]) * 100, 3), "%"),
    y = paste0("PC3; ", signif(var_frac[3] * 100, digits = 2), " %"),
    x = paste0("PC1; ", signif(var_frac[1] * 100, digits = 2), " %"),
    title = plot.title,
    caption = "") + 
  ggplot2::theme_minimal() +
  ggplot2::theme(
    axis.text = ggplot2::element_text(size = 15),
    axis.title.x = ggplot2::element_text(size = 20),
    axis.title.y = ggplot2::element_text(size = 20),
    legend.text = ggplot2::element_text(size = 15)
  )

print(pca.plot)
ggplot2::ggsave(pca.plot, bg = "white",
                filename = paste0(figure.path,
                                  "Toothfish_ALL_muscle_PCA_CpG_TOP100_DMS_PC13.png"), 
                width = 30, height = 15, units = "cm")




data <- data.frame(ind = rownames(PCobj$rotation),PCobj$rotation,  pop = pop)
data <- merge(data, aggregate(cbind(mean.x = PC2, mean.y = PC3) ~
                                pop, data, mean), by = "pop")

pca.plot <- ggplot2::ggplot(data, ggplot2::aes(PC2, PC3, color = pop)) +
  ggplot2::geom_point(size = 4) + 
  ggplot2::geom_point(ggplot2::aes(x = mean.x,
                                   y = mean.y, color = pop),
                      size = 7,
                      shape = 17) +
  ggplot2::scale_colour_manual(values = colours6m) +
  ggplot2::geom_hline(yintercept = 0) +
  ggplot2::geom_vline(xintercept = 0) + 
  ggplot2::labs(
    subtitle = paste0("Total variance explained: ",
                      signif(sum(var_frac[1:3]) * 100, 3), "%"),
    y = paste0("PC3; ", signif(var_frac[3] * 100, digits = 2), " %"),
    x = paste0("PC2; ", signif(var_frac[2] * 100, digits = 2), " %"),
    title = plot.title,
    caption = "") + 
  ggplot2::theme_minimal() +
  ggplot2::theme(
    axis.text = ggplot2::element_text(size = 15),
    axis.title.x = ggplot2::element_text(size = 20),
    axis.title.y = ggplot2::element_text(size = 20),
    legend.text = ggplot2::element_text(size = 15)
  )

print(pca.plot)
ggplot2::ggsave(pca.plot, bg = "white",
                filename = paste0(figure.path,"Toothfish_ALL_muscle_PCA_CpG_TOP100_DMS_PC23.png"), 
                width = 30, height = 15, units = "cm")



data <- data.frame(ind = rownames(PCobj$rotation),PCobj$rotation,  pop = pop)
data <- merge(data, aggregate(cbind(mean.x = PC4, mean.y = PC3) ~
                                pop, data, mean), by = "pop")

pca.plot <- ggplot2::ggplot(data, ggplot2::aes(PC4, PC3, color = pop)) +
  ggplot2::geom_point(size = 4) + 
  ggplot2::geom_point(ggplot2::aes(x = mean.x,
                                   y = mean.y, color = pop),
                      size = 7,
                      shape = 17) +
  ggplot2::scale_colour_manual(values = colours6m) +
  ggplot2::geom_hline(yintercept = 0) +
  ggplot2::geom_vline(xintercept = 0) + 
  ggplot2::labs(
    subtitle = paste0("Total variance explained: ",
                      signif(sum(var_frac[1:3]) * 100, 3), "%"),
    y = paste0("PC3; ", signif(var_frac[3] * 100, digits = 2), " %"),
    x = paste0("PC4; ", signif(var_frac[4] * 100, digits = 2), " %"),
    title = plot.title,
    caption = "") + 
  ggplot2::theme_minimal() +
  ggplot2::theme(
    axis.text = ggplot2::element_text(size = 15),
    axis.title.x = ggplot2::element_text(size = 20),
    axis.title.y = ggplot2::element_text(size = 20),
    legend.text = ggplot2::element_text(size = 15)
  )

print(pca.plot)
ggplot2::ggsave(pca.plot, bg = "white",
                filename = paste0(figure.path,"Toothfish_ALL_muscle_PCA_CpG_TOP100_DMS_PC34.png"), 
                width = 30, height = 15, units = "cm")
```





## All fin with top100 DMS

```{r}
print(load(paste0(rdata.path,"Toothfish_ALL_fin_methylkit_object.Rdata")))
print(load(paste0(rdata.path,"Toothfish_metadata.Rdata"))) 
metadata.new <- metadata[metadata$Genotype_names %in% meth.ALLF.norm10L@sample.ids,]
metadata.new <- metadata.new[order(match(metadata.new$Genotype_names,
                                         meth.ALLF.norm10L@sample.ids)),]

meth.cpg <- paste0(meth.ALLF.norm10L$chr, "_", meth.ALLF.norm10L$start) #283819 
DMS.cpg <- unique(c(
    Toothfish_2024_muscle_TOP100_DMS,
    Toothfish_ALL_muscle_TOP100_DMS,
    Toothfish_2024_fin_TOP100_DMS,
    Toothfish_ALL_fin_TOP100_DMS)) #336

sum(meth.cpg %in% DMS.cpg) #199
meth.DMS <- meth.ALLF.norm10L[meth.cpg %in% DMS.cpg,]
```



### Heatmap

```{r}
meth.diff.df <- as.data.frame(t(methylKit::percMethylation(meth.DMS)))
loc.names <- paste0(meth.DMS$chr,"_",meth.DMS$start)
colnames(meth.diff.df) <- loc.names

meth.diff.df$individual <- rownames(meth.diff.df)
pop <- metadata.new$pop
meth.diff.df$pop <- as.character(pop)


heat.plot.data <- meth.diff.df %>%
  tibble::as_tibble() %>%
  tidyr::gather(key = "CpG", value = "Methylation_percentage", -individual, -pop,
                factor_key = FALSE) %>%
  dplyr::group_by(pop, CpG) %>%
  dplyr::mutate(CpG_mean = mean(Methylation_percentage, na.rm = TRUE)) %>%
  dplyr::arrange(pop, individual, CpG) %>%
  dplyr::mutate(individual = factor(individual, levels = unique(individual))) %>%
  dplyr::mutate(CpG = factor(CpG, levels = unique(CpG)))
cpgs <- heat.plot.data$CpG[heat.plot.data$individual == unique(heat.plot.data$individual)[1]]
cpgs_mean <- heat.plot.data$CpG_mean[heat.plot.data$individual == unique(heat.plot.data$individual)[1]]

cpgs <- cpgs[order(cpgs_mean, decreasing = TRUE)]
heat.plot.data$CpG <- factor(heat.plot.data$CpG, levels = c("pop", unique(as.character(cpgs))))
heat.plot.data <- heat.plot.data[order(heat.plot.data$CpG),]

heat.plot <- ggplot2::ggplot(data = heat.plot.data, ggplot2::aes(x = individual)) + 
    ggplot2::geom_tile(data = heat.plot.data, 
                       ggplot2::aes(x = individual, y = CpG, fill = Methylation_percentage)) +
    ggplot2::scale_fill_gradient2(low = "blue", mid = "yellow", high = "red", 
                                  midpoint = 50, guide = "colourbar") +
    ggnewscale::new_scale_fill() +
    ggplot2::geom_tile(data = heat.plot.data, ggplot2::aes(x = individual, y = "pop", fill = pop,
                                    height = length(unique(heat.plot.data$CpG))*0.02)) +
    ggplot2::scale_fill_manual(values = colours3f) +
    ggplot2::scale_y_discrete(drop = FALSE) +
    ggplot2::theme(legend.position = "top",
                   axis.title.y = ggplot2::element_blank(),
                   axis.title.x = ggplot2::element_blank(),
                   axis.text.x = ggplot2::element_text(face = "bold", color = "black", 
                                        size = 10, angle = 90, hjust = 1, vjust = 0.5))
print(heat.plot)

ggplot2::ggsave(heat.plot, bg = "white",
                filename = paste0(figure.path,
                                  "Toothfish_ALL_fin_BASED_ON_TOP100_DMS_HEATMAP.png"), 
                width = 30, height = 30, units = "cm")
```


### DMS Clustering
```{r}
methdif.perc <- methylKit::percMethylation(meth.DMS)
methdif.perc2 <- zoo::na.aggregate(methdif.perc, by = meth.DMS@treatment)
PCobj <- prcomp(methdif.perc2, retx = TRUE, scale. = TRUE)

var_frac <- PCobj$sdev/sum(PCobj$sdev)
print(paste0("Total variance explained: ",signif(sum(var_frac[1:3]) * 100, 3), "%")) 
barplot(var_frac[1:3], main = "variance explained", 
        col = heat.colors(length(var_frac[1:3])))


identical(metadata.new$Genotype_names, meth.DMS@sample.ids)
identical(metadata.new$Genotype_names, rownames(PCobj$rotation))
pop <- metadata.new$pop

plot.title <- "Toothfish methylation - ALL Fin (199 TOP100 DMS)"

data <- data.frame(ind = rownames(PCobj$rotation),PCobj$rotation,  pop = pop)
data <- merge(data, aggregate(cbind(mean.x = PC1, mean.y = PC2) ~
                                pop, data, mean), by = "pop")

pca.plot <- ggplot2::ggplot(data, ggplot2::aes(PC1, PC2, color = pop)) +
  ggplot2::geom_point(size = 4) + 
  ggplot2::geom_point(ggplot2::aes(x = mean.x,
                                   y = mean.y, color = pop),
                      size = 7,
                      shape = 17) +
  ggplot2::scale_colour_manual(values = colours3f) +
  ggplot2::geom_hline(yintercept = 0) +
  ggplot2::geom_vline(xintercept = 0) + 
  ggplot2::labs(
    subtitle = paste0("Total variance explained: ",
                      signif(sum(var_frac[1:3]) * 100, 3), "%"),
    y = paste0("PC2; ", signif(var_frac[2] * 100, digits = 2), " %"),
    x = paste0("PC1; ", signif(var_frac[1] * 100, digits = 2), " %"),
    title = plot.title,
    caption = "") + 
  ggplot2::theme_minimal() +
  ggplot2::theme(
    axis.text = ggplot2::element_text(size = 15),
    axis.title.x = ggplot2::element_text(size = 20),
    axis.title.y = ggplot2::element_text(size = 20),
    legend.text = ggplot2::element_text(size = 15)
  )

print(pca.plot)
ggplot2::ggsave(pca.plot, bg = "white",
                filename = paste0(figure.path,"Toothfish_ALL_fin_PCA_CpG_TOP100_DMS_PC12.png"), 
                width = 30, height = 15, units = "cm")



data <- data.frame(ind = rownames(PCobj$rotation),PCobj$rotation,  pop = pop)
data <- merge(data, aggregate(cbind(mean.x = PC1, mean.y = PC3) ~
                                pop, data, mean), by = "pop")

pca.plot <- ggplot2::ggplot(data, ggplot2::aes(PC1, PC3, color = pop)) +
  ggplot2::geom_point(size = 4) + 
  ggplot2::geom_point(ggplot2::aes(x = mean.x,
                                   y = mean.y, color = pop),
                      size = 7,
                      shape = 17) +
  ggplot2::scale_colour_manual(values = colours3f) +
  ggplot2::geom_hline(yintercept = 0) +
  ggplot2::geom_vline(xintercept = 0) + 
  ggplot2::labs(
    subtitle = paste0("Total variance explained: ",
                      signif(sum(var_frac[1:3]) * 100, 3), "%"),
    y = paste0("PC3; ", signif(var_frac[3] * 100, digits = 2), " %"),
    x = paste0("PC1; ", signif(var_frac[1] * 100, digits = 2), " %"),
    title = plot.title,
    caption = "") + 
  ggplot2::theme_minimal() +
  ggplot2::theme(
    axis.text = ggplot2::element_text(size = 15),
    axis.title.x = ggplot2::element_text(size = 20),
    axis.title.y = ggplot2::element_text(size = 20),
    legend.text = ggplot2::element_text(size = 15)
  )

print(pca.plot)
ggplot2::ggsave(pca.plot, bg = "white",
                filename = paste0(figure.path,
                                  "Toothfish_ALL_fin_PCA_CpG_TOP100_DMS_PC13.png"), 
                width = 30, height = 15, units = "cm")




data <- data.frame(ind = rownames(PCobj$rotation),PCobj$rotation,  pop = pop)
data <- merge(data, aggregate(cbind(mean.x = PC2, mean.y = PC3) ~
                                pop, data, mean), by = "pop")

pca.plot <- ggplot2::ggplot(data, ggplot2::aes(PC2, PC3, color = pop)) +
  ggplot2::geom_point(size = 4) + 
  ggplot2::geom_point(ggplot2::aes(x = mean.x,
                                   y = mean.y, color = pop),
                      size = 7,
                      shape = 17) +
  ggplot2::scale_colour_manual(values = colours3f) +
  ggplot2::geom_hline(yintercept = 0) +
  ggplot2::geom_vline(xintercept = 0) + 
  ggplot2::labs(
    subtitle = paste0("Total variance explained: ",
                      signif(sum(var_frac[1:3]) * 100, 3), "%"),
    y = paste0("PC3; ", signif(var_frac[3] * 100, digits = 2), " %"),
    x = paste0("PC2; ", signif(var_frac[2] * 100, digits = 2), " %"),
    title = plot.title,
    caption = "") + 
  ggplot2::theme_minimal() +
  ggplot2::theme(
    axis.text = ggplot2::element_text(size = 15),
    axis.title.x = ggplot2::element_text(size = 20),
    axis.title.y = ggplot2::element_text(size = 20),
    legend.text = ggplot2::element_text(size = 15)
  )

print(pca.plot)
ggplot2::ggsave(pca.plot, bg = "white",
                filename = paste0(figure.path,"Toothfish_ALL_fin_PCA_CpG_TOP100_DMS_PC23.png"), 
                width = 30, height = 15, units = "cm")
```





#____________________________

# Intervals between CpG sites

```{r}
gr.sub <- as(meth,"GRanges")
dat.100 <- data.table::as.data.table(GenomicRanges::reduce(gr.sub, min.gapwidth = 100L))
gr.100 <- as(dat.100,"GRanges")

cnt <- GenomicRanges::countOverlaps(gr.100,gr.sub)
summary(cnt)
length(cnt) #7915
length(cnt[cnt > 4]) #2987

gr.sub <- as(meth,"GRanges")
my.raw <- as(myobjDB[[1]],"methylRaw")

mat <- methylKit::percMethylation(meth)
rowMeans(mat)
data2 <- data.frame(seqname = gr.sub@seqnames, range = gr.sub@ranges, mat )

data2$range.start <- as.numeric(data2$range.start)
data2$range.end <- as.numeric(data2$range.end)
df <- data.frame(scaffold = character(), interval = integer())
n <- 0
tot <- length(unique(data2$seqname))
for (i in unique(data2$seqname)) {
  n <- n + 1
  print(paste0(n, " out of ", tot))
  print(i)
  if (sum(data2$seqname == i) > 1) {
    subdata <- data2[data2$seqname == i, ]
    for (j in 2:nrow(subdata)) {
      res <- subdata$range.start[j] - subdata$range.end[j - 1]
      distr <- data.frame(scaffold = as.character(i), interval = as.integer(res))
      df <- rbind(df, distr)
    }
  }
}
save(df, file = paste0(rdata.path,"Toothfish_meth_intervals.Rdata"))
```

```{r}

print(load(paste0(rdata.path,"Toothfish_meth_intervals.Rdata")))

summary(df$interval)
df[df$interval == max(df$interval),]
# data2[2970:2980,]
chk <- summary(df$interval)
chk[5] #3rd quart

df1 <- df[df$interval < 1000,]
df2 <- df[df$interval < 100,]
df3 <- df[df$interval > 100,]

df.plot2 <- ggplot2::ggplot(df) + 
  ggplot2::geom_histogram(ggplot2::aes(interval)) +
  ggplot2::xlab("Distance between CpG sites") +
  ggplot2::ylab("Count") +
  ggplot2::scale_x_continuous(trans = 'log2') 

print(df.plot2)

ggplot2::ggsave(df.plot2, file = paste0(figure.path,"5a.Histogram_of_intervals_LOG.png"))


df.plot2b <- ggplot2::ggplot(df1) + 
  ggplot2::geom_histogram(ggplot2::aes(interval), binwidth = 10) +
  ggplot2::xlab("Distance between CpG sites") +
  ggplot2::ylab("Count") 
print(df.plot2b)

ggplot2::ggsave(df.plot2b, 
                file = paste0(figure.path,"5b.Histogram_of_intervals_smaller_than_1000bp.png"))

df.plot3 <- ggplot2::ggplot(df2) + 
  ggplot2::geom_histogram(ggplot2::aes(interval), binwidth = 1) +
  ggplot2::xlab("Distance between CpG sites") +
  ggplot2::ylab("Count") 
print(df.plot3)

ggplot2::ggsave(df.plot3, 
                file = paste0(figure.path,"5c.Histogram_of_intervals_smaller_than_100bp.png"))


df.plot4 <- ggplot2::ggplot(df3) + 
  ggplot2::geom_histogram(ggplot2::aes(interval), binwidth = 300) +
  ggplot2::xlab("Distance between CpG sites") +
  ggplot2::ylab("Count") 
print(df.plot4)

ggplot2::ggsave(df.plot4, 
                file = paste0(figure.path,"5d.Histogram_of_intervals_larger_than_100bp.png"))



summary(df3$interval)
chk2 <- summary(df3$interval)
# df4 <- df[df$interval > chk2[5],]
df4 <- df[df$interval > 15000,]

df.plot5 <- ggplot2::ggplot(df4) + 
  ggplot2::geom_histogram(ggplot2::aes(interval), binwidth = 300) +
  ggplot2::xlab("Distance between CpG sites") +
  ggplot2::ylab("Count") 
print(df.plot5)

ggplot2::ggsave(df.plot5, 
                file = paste0(figure.path,"5e.Histogram_of_intervals_larger_than_15000bp.png"))


length(df$interval[df$interval < 100]) / length(df$interval) #0.79
length(df$interval[df$interval < 25]) / length(df$interval) #0.59
```

## Methylation by features

```{r}
gff.file <- "/datasets/work/ncmi-toa-rrbs/work/1.Toothfish/Bismark_genome_index/D.mawsoni_2.gtf"
gff.file <- "/datasets/work/ncmi-toa-rrbs/work/1.Toothfish/Bismark_genome_index/D.mawsoni_CPGI.gff"

gff <- genomation::gffToGRanges(gff.file)
gfflhap2 <- as(split(gff, gff$type), "GRangesList")
CpGGranges <- as(myDiff25p,"GRanges")
anotaCpG <- annotateWithFeatures(CpGGranges, gfflhap2)
```


```{r}
chr.file <- "/datasets/work/ncmi-toa-rrbs/work/1.Toothfish/Bismark_genome_index/sizes.chr"

chrom.info <- read.table(file = chr.file, header = F)
chrom.info$V3 <- FALSE
colnames(chrom.info) <- c("chrom", "length", "is_circular")

gtf.file <- "/datasets/work/ncmi-toa-rrbs/work/1.Toothfish/Bismark_genome_index/D.mawsoni_2.gtf"

gencode <- GenomicFeatures::makeTxDbFromGFF(gtf.file,
                format = "gtf",
                dataSource = "Annotation From Chen et al. 2021",
                organism = "Dissostichus mawsoni",
                taxonomyId = NA,
                circ_seqs = NULL,
                chrominfo = chrom.info,
                miRBaseBuild = NA,
                metadata = NULL)

# saveDb(gencode, file = paste0(rdata.path, "gencode_D_mawsoni.sqlite"))
gencode <- loadDb(paste0(rdata.path, "gencode_D_mawsoni.sqlite"))

#### Create GRAnges objects for the each feature
genes.gencode <- sort(GenomicFeatures::genes(gencode))
intergenic.gencode <- GenomicRanges::gaps(genes.gencode)
transcript.gencode <- sort(GenomicFeatures::transcripts(gencode, use.names = TRUE))
cds.gencode <- sort(GenomicFeatures::cds(gencode))
exons.gencode <- sort(GenomicFeatures::exons(gencode))
introns.gencode <- sort(unlist(GenomicFeatures::intronsByTranscript(gencode)))
promoters.gencode <- sort(GenomicFeatures::promoters(gencode, upstream = 1000, downstream = 1000))
TSS.gencode <- transcript.gencode
TSS.gencode@ranges@width <- rep(1L, times = length(TSS.gencode@ranges@width))

annotation.all.features <-  GenomicRanges::GRangesList(
  promoters = promoters.gencode, 
  exons = exons.gencode,
  genes = genes.gencode,
  introns = introns.gencode,
  intergenics = intergenic.gencode,
  transcripts = transcript.gencode,
  cds = cds.gencode,
  TSSes = TSS.gencode)

TOA.gr <- as(meth, "GRanges")

feat.df <- data.frame(annot = NULL, target = NULL, feature = NULL)
ann <- genomation::annotateWithFeature(TOA.gr, genes.gencode, intersect.chr = TRUE) #48.79%, 11.05%
feat.df <- rbind(feat.df, data.frame(annot = ann@annotation[1], 
                                     target = ann@perc.of.OlapFeat, 
                                     feature = "Genes"))
ann <- genomation::annotateWithFeature(TOA.gr, transcript.gencode, intersect.chr = TRUE)
feat.df <- rbind(feat.df, data.frame(annot = ann@annotation[1], 
                                     target = ann@perc.of.OlapFeat, 
                                     feature = "Transcipts"))
ann <- genomation::annotateWithFeature(TOA.gr, cds.gencode, intersect.chr = TRUE) #17.97%, 0.6#
feat.df <- rbind(feat.df, data.frame(annot = ann@annotation[1], 
                                     target = ann@perc.of.OlapFeat, 
                                     feature = "CDS"))
ann <- genomation::annotateWithFeature(TOA.gr, intergenic.gencode, intersect.chr = TRUE) #97.83%, 23.19%
feat.df <- rbind(feat.df, data.frame(annot = ann@annotation[1], 
                                     target = ann@perc.of.OlapFeat, 
                                     feature = "Intergenic"))
ann <- genomation::annotateWithFeature(TOA.gr, exons.gencode, intersect.chr = TRUE) #20.57%, 0.65%
feat.df <- rbind(feat.df, data.frame(annot = ann@annotation[1], 
                                     target = ann@perc.of.OlapFeat, 
                                     feature = "Exons"))
ann <- genomation::annotateWithFeature(TOA.gr, introns.gencode, intersect.chr = TRUE) #29.23%, 1.23%
feat.df <- rbind(feat.df, data.frame(annot = ann@annotation[1], 
                                     target = ann@perc.of.OlapFeat, 
                                     feature = "Introns"))
ann <- genomation::annotateWithFeature(TOA.gr, promoters.gencode, intersect.chr = TRUE)# 7.61%, 1.75%
feat.df <- rbind(feat.df, data.frame(annot = ann@annotation[1], 
                                     target = ann@perc.of.OlapFeat, 
                                     feature = "Promoters"))
ann <- genomation::annotateWithFeature(TOA.gr, TSS.gencode, intersect.chr = TRUE)# 0.01%, 0.01%
feat.df <- rbind(feat.df, data.frame(annot = ann@annotation[1], 
                                     target = ann@perc.of.OlapFeat, 
                                     feature = "TSS"))

feat.df2 <- reshape2::melt(feat.df, id.vars='feature')
feat.df2$variable <- as.character(feat.df2$variable)
feat.df2$variable[feat.df2$variable == "annot"] <- "Other annotations"
feat.df2$variable[feat.df2$variable == "target"] <- "RRBS CpG sites"

annot.plot <- ggplot2::ggplot(data = feat.df2, 
                              ggplot2::aes(x =  feature, y = value, fill = variable)) +
  ggplot2::geom_bar(stat='identity', position='dodge') +
    ggplot2::labs(
    y = "Overlap %",
    x = "Annotation feature",
    # title = "Overlap between annotations and 36000 CpG sites",
    title = "Overlap between annotations and 1,316,799 CpG sites",
    caption = "") +
  ggplot2::theme_minimal() +
  ggplot2::theme(
    axis.text = ggplot2::element_text(size = 15),
    axis.text.x = ggplot2::element_text(size = 15, angle = 90),
    axis.title.x = ggplot2::element_text(size = 20),
    axis.title.y = ggplot2::element_text(size = 20),
    legend.text = ggplot2::element_text(size = 15)
  )
print(annot.plot)
ggplot2::ggsave(annot.plot, bg = "white",
                filename = paste0(figure.path,"Annotation_overlap_plot_Subset.png"), 
                width = 30, height = 15, units = "cm")
```

### Promotor methylation

```{r}
idx_overlaps <- S4Vectors::queryHits(GenomicRanges::findOverlaps(
  TOA.gr,
  annotation.all.features$promoters, 
  type = 'any'))
promotor.gr <- TOA.gr[unique(idx_overlaps),]
promotor.raw <- meth[unique(idx_overlaps),]
promotor.raw <- meth[unique(idx_overlaps),]

promoters <- methylkit::regionCounts(myobj,gene.obj$promoters)


promotor.prc <- methylKit::percMethylation(promotor.raw) # get percent methylation matrix
mds <- matrixStats::rowSds(promotor.prc) # calculate standard deviation of CpGs
head(promotor.raw[mds>20,])
hist(mds,col="cornflowerblue",xlab="Std. dev. per CpG")

promotor.prc <- as.data.frame(promotor.prc)
promotor.prc$seqs <- paste0(promotor.gr@seqnames,"_", promotor.gr@ranges@start)



promotor.prc.long <- tidyr::gather(data = promotor.prc, sample, methylation, -seqs)
strata <- data.frame(sample  = metadata$SAMPLE_ID, region = metadata$region)
promotor.prc.long <- dplyr::left_join(promotor.prc.long, strata, by = "sample" )
promotor.grp <- dplyr::group_by(promotor.prc.long, region, seqs) %>%
    dplyr::summarise(mean_meth  = mean(methylation))
promotor.grp$feat <- "Promoter"


methylKit::clusterSamples(promotor.raw, dist = "correlation", method = "ward.D2", plot = TRUE)
dev.print(file = paste0(figure.path,"cluster_plot_promoters_subset.png"),
          device = png, res = 300,width = 45,height = 15,units = "cm")
methylKit::PCASamples(promotor.raw)
dev.print(file = paste0(figure.path,"PCA_plot_promoters_subset.png"),
          device = png, res = 300,width = 30,height = 15,units = "cm")

```


```{r}
sm <- genomation::ScoreMatrix(target = promotor.gr, windows = annotation.all.features$promoters,
                              strand.aware = TRUE,
                               weight.col = NULL, is.noCovNA = FALSE, type = "auto", 
                              rpm = FALSE, unique = FALSE, extend = 0, param = NULL, 
                              bam.paired.end = FALSE, library.size = NULL)
genomation::heatMatrix(sm, xcoords = c(-1000, 1000))


sml.sub <- intersectScoreMatrixList(sml, reorder = FALSE) # because scoreMatrices have different dimensions, we need them to cover the same promoters for the heatmap
multiHeatMatrix(sml.sub, xcoords=c(-2000, 2000), matrix.main=names(targets), xlab="region around TSS")


```

### Exon methylation

```{r}
idx_overlaps <- S4Vectors::queryHits(GenomicRanges::findOverlaps(
  TOA.gr,
  annotation.all.features$exons, 
  type = 'any'))
exons.gr <- TOA.gr[unique(idx_overlaps),]
exons.raw <- meth[unique(idx_overlaps),]
exons.raw <- meth[unique(idx_overlaps),]



exons.prc <- methylKit::percMethylation(exons.raw) # get percent methylation matrix
mds <- matrixStats::rowSds(exons.prc) # calculate standard deviation of CpGs
head(exons.raw[mds>20,])
hist(mds,col="cornflowerblue",xlab="Std. dev. per CpG")
exons.prc <- as.data.frame(exons.prc)
exons.prc$seqs <- paste0(exons.gr@seqnames,"_", exons.gr@ranges@start)

exons.prc.long <- tidyr::gather(data = exons.prc, sample, methylation, -seqs)
strata <- data.frame(sample  = metadata$SAMPLE_ID, region = metadata$region)
exons.prc.long <- dplyr::left_join(exons.prc.long, strata, by = "sample" )
exons.grp <- dplyr::group_by(exons.prc.long, region, seqs) %>%
    dplyr::summarise(mean_meth  = mean(methylation))
exons.grp$feat <- "Exons"


methylKit::clusterSamples(exons.raw, dist = "correlation", method = "ward.D2", plot = TRUE)
dev.print(file = paste0(figure.path,"cluster_plot_exons_subset.png"),
          device = png, res = 300,width = 45,height = 15,units = "cm")
methylKit::PCASamples(exons.raw)
dev.print(file = paste0(figure.path,"PCA_plot_exons_subset.png"),
          device = png, res = 300,width = 30,height = 15,units = "cm")
```

### Intron methylation

```{r}
idx_overlaps <- S4Vectors::queryHits(GenomicRanges::findOverlaps(
  TOA.gr,
  annotation.all.features$introns, 
  type = 'any'))
introns.gr <- TOA.gr[unique(idx_overlaps),]
introns.raw <- meth[unique(idx_overlaps),]
introns.raw <- meth[unique(idx_overlaps),]

introns.prc <- methylKit::percMethylation(introns.raw) # get percent methylation matrix
mds <- matrixStats::rowSds(introns.prc) # calculate standard deviation of CpGs
head(introns.raw[mds>20,])
hist(mds,col="cornflowerblue",xlab="Std. dev. per CpG")
introns.prc <- as.data.frame(introns.prc)
introns.prc$seqs <- paste0(introns.gr@seqnames,"_", introns.gr@ranges@start)

introns.prc.long <- tidyr::gather(data = introns.prc, sample, methylation, -seqs)
strata <- data.frame(sample  = metadata$SAMPLE_ID, region = metadata$region)
introns.prc.long <- dplyr::left_join(introns.prc.long, strata, by = "sample" )
introns.grp <- dplyr::group_by(introns.prc.long, region, seqs) %>%
    dplyr::summarise(mean_meth  = mean(methylation))
introns.grp$feat <- "Introns"


methylKit::clusterSamples(introns.raw, dist = "correlation", method = "ward.D2", plot = TRUE)
dev.print(file = paste0(figure.path,"cluster_plot_introns_subset.png"),
          device = png, res = 300,width = 45,height = 15,units = "cm")
methylKit::PCASamples(introns.raw)
dev.print(file = paste0(figure.path,"PCA_plot_introns_subset.png"),
          device = png, res = 300,width = 30,height = 15,units = "cm")
```

### Intergenic methylation

```{r}
idx_overlaps <- S4Vectors::queryHits(GenomicRanges::findOverlaps(
  TOA.gr,
  annotation.all.features$intergenics, 
  type = 'any'))
intergenics.gr <- TOA.gr[unique(idx_overlaps),]
intergenics.raw <- meth[unique(idx_overlaps),]

intergenics.raw <- meth[unique(idx_overlaps),]


intergenics.prc <- methylKit::percMethylation(intergenics.raw) # get percent methylation matrix
mds <- matrixStats::rowSds(intergenics.prc) # calculate standard deviation of CpGs
head(intergenics.raw[mds>20,])
hist(mds,col="cornflowerblue",xlab="Std. dev. per CpG")
intergenics.prc <- as.data.frame(intergenics.prc)
intergenics.prc$seqs <- paste0(intergenics.gr@seqnames,"_", intergenics.gr@ranges@start)

intergenics.prc.long <- tidyr::gather(data = intergenics.prc, sample, methylation, -seqs)
strata <- data.frame(sample  = metadata$SAMPLE_ID, region = metadata$region)
intergenics.prc.long <- dplyr::left_join(intergenics.prc.long, strata, by = "sample" )
intergenics.grp <- dplyr::group_by(intergenics.prc.long, region, seqs) %>%
    dplyr::summarise(mean_meth  = mean(methylation))
intergenics.grp$feat <- "Intergenics"


methylKit::clusterSamples(intergenics.raw, dist = "correlation", method = "ward.D2", plot = TRUE)
dev.print(file = paste0(figure.path,"cluster_plot_intergenics_subset.png"),
          device = png, res = 300,width = 45,height = 15,units = "cm")
methylKit::PCASamples(intergenics.raw)
dev.print(file = paste0(figure.path,"PCA_plot_intergenics_subset.png"),
          device = png, res = 300,width = 30,height = 15,units = "cm")
```

### TSS methylation

```{r}
idx_overlaps <- S4Vectors::queryHits(GenomicRanges::findOverlaps(
  TOA.gr,
  annotation.all.features$TSSes, 
  type = 'any'))
TSS.gr <- TOA.gr[unique(idx_overlaps),] #only 2
TSS.raw <- meth[unique(idx_overlaps),]

TSS.prc <- methylKit::percMethylation(TSS.raw) # get percent methylation matrix
mds <- matrixStats::rowSds(TSS.prc) # calculate standard deviation of CpGs
head(TSS.raw[mds>20,])
hist(mds,col="cornflowerblue",xlab="Std. dev. per CpG")
TSS.prc <- as.data.frame(TSS.prc)
TSS.prc$seqs <- paste0(TSS.gr@seqnames,"_", TSS.gr@ranges@start)

TSS.prc.long <- tidyr::gather(data = TSS.prc, sample, methylation, -seqs)
strata <- data.frame(sample  = metadata$SAMPLE_ID, region = metadata$region)
TSS.prc.long <- dplyr::left_join(TSS.prc.long, strata, by = "sample" )
TSS.grp <- dplyr::group_by(TSS.prc.long, region, seqs) %>%
    dplyr::summarise(mean_meth  = mean(methylation))
TSS.grp$feat <- "TSS"


methylKit::clusterSamples(TSS.raw, dist = "correlation", method = "ward.D2", plot = TRUE)
dev.print(file = paste0(figure.path,"cluster_plot_TSS.png"),
          device = png, res = 300,width = 45,height = 15,units = "cm")
methylKit::PCASamples(TSS.raw)
dev.print(file = paste0(figure.path,"PCA_plot_TSS.png"),
          device = png, res = 300,width = 30,height = 15,units = "cm")
```


### Combined plot
```{r}
combo <- rbind(promotor.grp, exons.grp, introns.grp, intergenics.grp,TSS.grp)
combo <- rbind(promotor.grp, exons.grp, introns.grp, intergenics.grp)

feat.plot <- ggplot2::ggplot(data = combo,ggplot2::aes(x =  feat, y = mean_meth)) +
  ggplot2::geom_violin(width = 1.5) +
  # ggplot2::geom_jitter(size = 0.5, alpha = 0.4) +
    ggplot2::labs(
    y = "Mean methylation (%) per CpG site per region",
    x = "Annotation feature",
    title = "Each point is a CpG site",
    caption = "") +
  ggplot2::facet_wrap(~region) +
  ggplot2::theme_minimal() +
  ggplot2::theme(
    axis.text = ggplot2::element_text(size = 15),
    axis.text.x = ggplot2::element_text(size = 15, angle = 90),
    axis.title.x = ggplot2::element_text(size = 20),
    axis.title.y = ggplot2::element_text(size = 20),
    legend.text = ggplot2::element_text(size = 15)
  )
print(feat.plot)
ggplot2::ggsave(feat.plot, bg = "white",
                filename = paste0(figure.path,"Methylation_per_CpG_per_annotation2_subset.png"), 
                width = 30, height = 15, units = "cm")
```



### Population-specific missing data

```{r}
cov <- methylKit::getData(meth)
cov <- cov[,meth@coverage.index]
colnames(cov) <- meth@sample.ids
rownames(cov) <- paste0(meth$chr, "_", meth$start)
cov$CpG <- rownames(cov) 

# cov <- as.data.frame(t(cov))
# cov$pop <- as.character(meth@treatment)
miss.pop <- cov %>%
  tidyr::gather(key = "individual", value = "coverage", -CpG,
                factor_key = FALSE) %>%
  dplyr::left_join(data.frame(individual = meth@sample.ids, pop = meth@treatment),
                   by = "individual") %>%
  dplyr::mutate(pop = as.character(pop)) %>%
  dplyr::select(CpG, pop, coverage) %>%
  dplyr::group_by(pop,CpG) %>%
  dplyr::summarise(missing_prop = sum(is.na(coverage))/length(unique(CpG))) %>%
  dplyr::ungroup()


miss.pop$pop[miss.pop$pop == "1"] <- "California"
miss.pop$pop[miss.pop$pop == "2"] <- "New Zealand"
miss.pop$pop[miss.pop$pop == "3"] <- "South Australia"
miss.pop$pop[miss.pop$pop == "4"] <- "TAS-control (18C/33‰)"

colours.4 <- colours.8[names(colours.8) %in% c("TAS-control (18C/33‰)","California",
                                               "South Australia","New Zealand")]

miss.plot <- ggplot2::ggplot(miss.pop, ggplot2::aes(x = pop, y = missing_prop, colour = pop)) +
  ggplot2::geom_boxplot() +
  ggplot2::labs(y = "Proportion missing data per CpG") +
  ggplot2::scale_colour_manual(values = colours.4, 
                             na.value = "#999999") +
  ggplot2::theme(
    axis.title.x = ggplot2::element_blank(),
    axis.text.x = ggplot2::element_text(color = "black", 
                                        size = 10, angle = 90, 
                                        hjust = 1, vjust = 0.5),
    panel.background = ggplot2::element_rect(fill = "white"),
    panel.grid.major = ggplot2::element_line(linewidth = 0.5, linetype = 'solid',
                                             colour = "grey"), 
    panel.grid.minor = ggplot2::element_line(linewidth  = 0.1, linetype = 'solid',
                                             colour = "grey"),
    axis.line = ggplot2::element_line(colour = "black", linewidth = 0.5, 
                                      linetype = 1, lineend = "butt")
  )
print(miss.plot)
ggplot2::ggsave(plot = miss.plot, 
                filename = paste0(figure.path,"0_TOA_CpG_POPULATION_specific_missing_data.png"),
                width = 30, height = 15, units = "cm")
```

## sample correlation

```{r}
methylKit::getCorrelation(meth, plot = TRUE)
dev.print(file = paste0(figure.path,"Methylation_correlation.png"),
          device = png, res = 300,width = 15,height = 15,units = "cm")
```


#### PCA - no impute

```{r}
PCobj <- methylKit::PCASamples(meth, comp = c(1,2), adj.lim = c(0,0), obj.return = TRUE)

var_frac <- PCobj$sdev/sum(PCobj$sdev)
print(paste0("Total variance explained: ",signif(sum(var_frac[1:3]) * 100, 3), "%")) 
barplot(var_frac[1:20], main = "variance explained", 
        col = heat.colors(length(var_frac[1:20])))

identical(metadata.new$individual, meth.pop@sample.ids)
identical(metadata.new$individual, rownames(PCobj$x))

pop <- metadata.new$strata


plot.title <- "Toothfish methylation (1,601,561 CpG sites)"

data <- data.frame(ind = rownames(PCobj$x),PCobj$x,  pop = pop)
data <- merge(data, aggregate(cbind(mean.x = PC1, mean.y = PC2) ~
                                pop, data, mean), by = "pop")

pca.plot <- ggplot2::ggplot(data, ggplot2::aes(PC1, PC2, color = pop)) +
  ggplot2::geom_point(size = 4) + 
  ggplot2::geom_point(ggplot2::aes(x = mean.x,
                                   y = mean.y, color = pop),
                      size = 7,
                      shape = 17) +
  ggplot2::scale_colour_manual(values = colours.6) +
  ggplot2::scale_fill_manual(values = colours.6) + 
  ggplot2::geom_hline(yintercept = 0) +
  ggplot2::geom_vline(xintercept = 0) + 
  ggplot2::labs(
    subtitle = paste0("Total variance explained: ",
                      signif(sum(var_frac[1:4]) * 100, 3), "%"),
    y = paste0("PC2; ", signif(var_frac[2] * 100, digits = 2), " %"),
    x = paste0("PC1; ", signif(var_frac[1] * 100, digits = 2), " %"),
    title = plot.title,
    caption = "") + 
  ggplot2::theme_minimal() +
  ggplot2::theme(
    axis.text = ggplot2::element_text(size = 15),
    axis.title.x = ggplot2::element_text(size = 20),
    axis.title.y = ggplot2::element_text(size = 20),
    legend.text = ggplot2::element_text(size = 15)
  )

print(pca.plot)
ggplot2::ggsave(pca.plot, bg = "white",
                filename = paste0(figure.path,"TOA_PCA_CpG_POPULATION_PC12.png"), 
                width = 30, height = 15, units = "cm")



data <- data.frame(ind = rownames(PCobj$x),PCobj$x,  pop = pop)
data <- merge(data, aggregate(cbind(mean.x = PC2, mean.y = PC3) ~
                                pop, data, mean), by = "pop")

pca.plot <- ggplot2::ggplot(data, ggplot2::aes(PC2, PC3, color = pop)) +
  ggplot2::geom_point(size = 4) + 
  ggplot2::geom_point(ggplot2::aes(x = mean.x,
                                   y = mean.y, color = pop),
                      size = 7,
                      shape = 17) +
  ggplot2::scale_colour_manual(values = colours.6) +
  ggplot2::scale_fill_manual(values = colours.6) + 
  ggplot2::geom_hline(yintercept = 0) +
  ggplot2::geom_vline(xintercept = 0) + 
  ggplot2::labs(
    subtitle = paste0("Total variance explained: ",
                      signif(sum(var_frac[1:4]) * 100, 3), "%"),
    y = paste0("PC3; ", signif(var_frac[3] * 100, digits = 2), " %"),
    x = paste0("PC2; ", signif(var_frac[2] * 100, digits = 2), " %"),
    title = plot.title,
    caption = "") + 
  ggplot2::theme_minimal() +
  ggplot2::theme(
    axis.text = ggplot2::element_text(size = 15),
    axis.title.x = ggplot2::element_text(size = 20),
    axis.title.y = ggplot2::element_text(size = 20),
    legend.text = ggplot2::element_text(size = 15)
  )

print(pca.plot)
ggplot2::ggsave(pca.plot, bg = "white",
                filename = paste0(figure.path,"TOA_PCA_CpG_POPULATION_PC23.png"), 
                width = 30, height = 15, units = "cm")
```

## Low methylation regions

```{r}
print(load(paste0(rdata.path,"Toothfish_subset_methylation_objects_raw_NEW.Rdata")))


# Keep the sites with <5% methylation per group:  low frequency (0.0–0.3), intermediate frequency (0.3–0.7), and high frequency (0.7–1.0) 
# then look at spatial grouping on the genome

# meth.df <- methylKit::getData(meth)
meth.df <- methylKit::percMethylation(meth)
meth.df2 <- data.frame(indviduals = colnames(meth.df), Region = meth@treatment, t(meth.df))
colnames(meth.df2) <- c("indviduals","region", paste0(meth$chr, "_", meth$start))
meth.df2[1:10,1:10]
meth.df2$region <- as.character(meth.df2$region)

test <- meth.df2 %>% 
  tidyr::pivot_longer(
    cols = tidyr::starts_with("JAAKFY"),
    cols_vary = "fastest",
    names_to = "CpG",
    values_to = "methylation_percentage"
  ) %>%
  # dplyr::select(region, CpG, methylation_percentage) %>%
  dplyr::group_by(region, CpG) %>%
  dplyr::summarise(MEAN_METH = mean(methylation_percentage, na.rm = TRUE)) %>%
  # dplyr::select(region, CpG, MEAN_METH) %>%
  dplyr::filter(MEAN_METH < 5)

length(unique(test$CpG)) #146446
length(unique(test$indviduals)) #112

meth.df3 <- meth.df2[,colnames(meth.df2) %in% c("indviduals","region", unique(test$CpG))]
meth.df3[1:10,1:10]


test2 <- meth.df2 %>% 
  tidyr::pivot_longer(
    cols = tidyr::starts_with("JAAKFY"),
    cols_vary = "fastest",
    names_to = "CpG",
    values_to = "methylation_percentage"
  ) %>%
  dplyr::filter(CpG %in% unique(test$CpG)) %>%
  dplyr::group_by(region, CpG) %>%
  dplyr::summarise(MEAN_METH = mean(methylation_percentage, na.rm = TRUE))

test2 <- test2[order(test2$CpG),]

test2[test2$MEAN_METH > 25,] #3 CpG sites
```


## PCA

```{r}
print(load(paste0(rdata.path,"Toothfish_subset_methylation_objects_raw_NEW.Rdata")))

PCobj <- methylKit::PCASamples(meth, comp = c(1,2), 
                               # transpose = TRUE,
                               adj.lim = c(0,0), obj.return = TRUE)


methdif.perc <- methylKit::percMethylation(meth)
methdif.perc2 <- zoo::na.aggregate(methdif.perc, by = meth@treatment) # ISSUE Missing data only at CA
PCobj <- prcomp(methdif.perc2, retx = TRUE, scale. = TRUE)


var_frac <- PCobj$sdev/sum(PCobj$sdev)
print(paste0("Total variance explained: ",signif(sum(var_frac[1:3]) * 100, 3), "%")) 
barplot(var_frac[1:20], main = "variance explained", 
        col = heat.colors(length(var_frac[1:20])))

# metadata3 <- metadata2[metadata2$tissue %in% "M",]
metadata.new <- metadata2[metadata2$Genotype_names %in% meth@sample.ids,]
metadata.new <- metadata.new[order(match(metadata.new$Genotype_names, meth@sample.ids)),]
identical(metadata.new$Genotype_names, meth@sample.ids)
identical(metadata.new$Genotype_names, rownames(PCobj$rotation))

pop <- paste0(metadata.new$STRATA1, "-", metadata.new$Year)


# plot.title <- "Toothfish methylation (1,316,799 CpG sites)"
plot.title <- "Toothfish methylation (48,910 CpG sites)"


# data <- data.frame(ind = rownames(PCobj$x),PCobj$x,  pop = pop)
data <- data.frame(ind = rownames(PCobj$rotation),PCobj$rotation,  pop = pop)
data <- merge(data, aggregate(cbind(mean.x = PC1, mean.y = PC2) ~
                                pop, data, mean), by = "pop")

pca.plot <- ggplot2::ggplot(data, ggplot2::aes(PC1, PC2, color = pop)) +
  ggplot2::geom_point(size = 4) + 
  ggplot2::geom_point(ggplot2::aes(x = mean.x,
                                   y = mean.y, color = pop),
                      size = 7,
                      shape = 17) +
  ggplot2::scale_colour_manual(values = colours6m) +
  ggplot2::scale_fill_manual(values = colours6m) + 
  ggplot2::geom_hline(yintercept = 0) +
  ggplot2::geom_vline(xintercept = 0) + 
  ggplot2::labs(
    subtitle = paste0("Total variance explained: ",
                      signif(sum(var_frac[1:4]) * 100, 3), "%"),
    y = paste0("PC2; ", signif(var_frac[2] * 100, digits = 2), " %"),
    x = paste0("PC1; ", signif(var_frac[1] * 100, digits = 2), " %"),
    title = plot.title,
    caption = "") + 
  ggplot2::theme_minimal() +
  ggplot2::theme(
    axis.text = ggplot2::element_text(size = 15),
    axis.title.x = ggplot2::element_text(size = 20),
    axis.title.y = ggplot2::element_text(size = 20),
    legend.text = ggplot2::element_text(size = 15)
  )

print(pca.plot)
ggplot2::ggsave(pca.plot, bg = "white",
                filename = paste0(figure.path,"TOA_PCA_CpG_ALL_muscle_regions_2024_PC12.png"), 
                width = 30, height = 15, units = "cm")

# data[data$PC2 > 200, 1:5]
#       pop            ind        PC1      PC2        PC3
# 14 58-4-1 TOA_58-4-1_277  -47.45444 261.8010 -106.49835
# 17 58-4-1 TOA_58-4-1_289  -73.47030 328.2200 -104.75805
# 18 58-4-1 TOA_58-4-1_290 -161.06043 376.8552 -107.13636
# 20 58-4-1 TOA_58-4-1_295  -63.09566 290.4805  -94.24499
# data[data$PC1 > 450, 1:5]
#      pop            ind      PC1      PC2       PC3
# 3 58-4-1 TOA_58-4-1_118 513.1702 -2.60882 -26.78892
# 4 58-4-1 TOA_58-4-1_121 480.4375 19.32287 -77.53182
# 5 58-4-1 TOA_58-4-1_125 666.1397 70.07817  79.81629


# data <- data.frame(ind = rownames(PCobj$x),PCobj$x,  pop = pop)
data <- data.frame(ind = rownames(PCobj$rotation),PCobj$rotation,  pop = pop)
data <- merge(data, aggregate(cbind(mean.x = PC1, mean.y = PC3) ~
                                pop, data, mean), by = "pop")

pca.plot <- ggplot2::ggplot(data, ggplot2::aes(PC1, PC3, color = pop)) +
  ggplot2::geom_point(size = 4) + 
  ggplot2::geom_point(ggplot2::aes(x = mean.x,
                                   y = mean.y, color = pop),
                      size = 7,
                      shape = 17) +
  ggplot2::scale_colour_manual(values = colours6m) +
  ggplot2::scale_fill_manual(values = colours6m) + 
  ggplot2::geom_hline(yintercept = 0) +
  ggplot2::geom_vline(xintercept = 0) + 
  ggplot2::labs(
    subtitle = paste0("Total variance explained: ",
                      signif(sum(var_frac[1:4]) * 100, 3), "%"),
    y = paste0("PC3; ", signif(var_frac[3] * 100, digits = 2), " %"),
    x = paste0("PC1; ", signif(var_frac[1] * 100, digits = 2), " %"),
    title = plot.title,
    caption = "") + 
  ggplot2::theme_minimal() +
  ggplot2::theme(
    axis.text = ggplot2::element_text(size = 15),
    axis.title.x = ggplot2::element_text(size = 20),
    axis.title.y = ggplot2::element_text(size = 20),
    legend.text = ggplot2::element_text(size = 15)
  )

print(pca.plot)
ggplot2::ggsave(pca.plot, bg = "white",
                filename = paste0(figure.path,"TOA_PCA_CpG_ALL_muscle_regions_2024_PC13.png"), 
                width = 30, height = 15, units = "cm")




# data <- data.frame(ind = rownames(PCobj$x),PCobj$x,  pop = pop)
data <- data.frame(ind = rownames(PCobj$rotation),PCobj$rotation,  pop = pop)
data <- merge(data, aggregate(cbind(mean.x = PC2, mean.y = PC3) ~
                                pop, data, mean), by = "pop")

pca.plot <- ggplot2::ggplot(data, ggplot2::aes(PC2, PC3, color = pop)) +
  ggplot2::geom_point(size = 4) + 
  ggplot2::geom_point(ggplot2::aes(x = mean.x,
                                   y = mean.y, color = pop),
                      size = 7,
                      shape = 17) +
  ggplot2::scale_colour_manual(values = colours6m) +
  ggplot2::scale_fill_manual(values = colours6m) + 
  ggplot2::geom_hline(yintercept = 0) +
  ggplot2::geom_vline(xintercept = 0) + 
  ggplot2::labs(
    subtitle = paste0("Total variance explained: ",
                      signif(sum(var_frac[1:4]) * 100, 3), "%"),
    y = paste0("PC3; ", signif(var_frac[3] * 100, digits = 2), " %"),
    x = paste0("PC2; ", signif(var_frac[2] * 100, digits = 2), " %"),
    title = plot.title,
    caption = "") + 
  ggplot2::theme_minimal() +
  ggplot2::theme(
    axis.text = ggplot2::element_text(size = 15),
    axis.title.x = ggplot2::element_text(size = 20),
    axis.title.y = ggplot2::element_text(size = 20),
    legend.text = ggplot2::element_text(size = 15)
  )

print(pca.plot)
ggplot2::ggsave(pca.plot, bg = "white",
                filename = paste0(figure.path,"TOA_PCA_CpG_ALL_muscle_regions_2024_PC23.png"), 
                width = 30, height = 15, units = "cm")
```

### PCA remove outlier

```{r}
rm <- c("TOA_58-4-1_277", "TOA_58-4-1_289", "TOA_58-4-1_290", "TOA_58-4-1_295")
rm2 <- c("TOA_58-4-1_118", "TOA_58-4-1_121", "TOA_58-4-1_125")

metadata3 <- metadata2[!metadata2$SAMPLE_ID %in% c(rm, rm2),]
table(metadata3$region)

# meth2 <- methylKit::reorganize(
#   methylObj = meth,
#   sample.ids = as.character(metadata3$SAMPLE_ID),
#   treatment = as.integer(as.factor(metadata3$region)),
#   chunk.size = 1e+06,
#   save.db = FALSE,
# )

filtered.myobj3 <- methylKit::reorganize(
  methylObj = filtered.myobj2,
  sample.ids = as.character(metadata3$SAMPLE_ID),
  treatment = as.integer(as.factor(metadata3$region)),
  chunk.size = 1e+06,
  save.db = FALSE,
)

meth2 <- methylKit::unite(filtered.myobj3, destrand = FALSE,
                         min.per.group = 8L, #missing data allowed
                         mc.cores = 25)

# save(filtered.myobj3, filtered.myobj2,meth2,
#      file = paste0(rdata.path,"Toothfish_methylation_objects_All_pop_no_outliers.Rdata"))
print(load(paste0(rdata.path,"Toothfish_methylation_objects_All_pop_no_outliers.Rdata")))


PCobj <- methylKit::PCASamples(meth2, comp = c(1,2), adj.lim = c(0,0), obj.return = TRUE)


var_frac <- PCobj$sdev/sum(PCobj$sdev)
print(paste0("Total variance explained: ",signif(sum(var_frac[1:3]) * 100, 3), "%")) 
barplot(var_frac[1:20], main = "variance explained", 
        col = heat.colors(length(var_frac[1:20])))

# metadata3 <- metadata2[metadata2$tissue %in% "M",]
metadata.new <- metadata3[metadata3$SAMPLE_ID %in% meth2@sample.ids,]
metadata.new <- metadata.new[order(match(metadata.new$SAMPLE_ID, meth2@sample.ids)),]
identical(metadata.new$SAMPLE_ID, meth2@sample.ids)
identical(metadata.new$SAMPLE_ID, rownames(PCobj$x))

pop <- metadata.new$region

# pop <- metadata.new$`Date caught`

colours5 <- adegenet::funky(4)
plot.title <- "Toothfish methylation (1,399,077 CpG sites)"

data <- data.frame(ind = rownames(PCobj$x),PCobj$x,  pop = pop)
data <- merge(data, aggregate(cbind(mean.x = PC1, mean.y = PC2) ~
                                pop, data, mean), by = "pop")

pca.plot <- ggplot2::ggplot(data, ggplot2::aes(PC1, PC2, color = pop)) +
  ggplot2::geom_point(size = 4) + 
  ggplot2::geom_point(ggplot2::aes(x = mean.x,
                                   y = mean.y, color = pop),
                      size = 7,
                      shape = 17) +
  ggplot2::scale_colour_manual(values = colours5) +
  ggplot2::scale_fill_manual(values = colours5) + 
  ggplot2::geom_hline(yintercept = 0) +
  ggplot2::geom_vline(xintercept = 0) + 
  ggplot2::labs(
    subtitle = paste0("Total variance explained: ",
                      signif(sum(var_frac[1:4]) * 100, 3), "%"),
    y = paste0("PC2; ", signif(var_frac[2] * 100, digits = 2), " %"),
    x = paste0("PC1; ", signif(var_frac[1] * 100, digits = 2), " %"),
    title = plot.title,
    caption = "") + 
  ggplot2::theme_minimal() +
  ggplot2::theme(
    axis.text = ggplot2::element_text(size = 15),
    axis.title.x = ggplot2::element_text(size = 20),
    axis.title.y = ggplot2::element_text(size = 20),
    legend.text = ggplot2::element_text(size = 15)
  )

print(pca.plot)
ggplot2::ggsave(pca.plot, bg = "white",
                filename = paste0(figure.path,"TOA_PCA_CpG_SUB2_regions_PC12.png"), 
                width = 30, height = 15, units = "cm")




data <- data.frame(ind = rownames(PCobj$x),PCobj$x,  pop = pop)
data <- merge(data, aggregate(cbind(mean.x = PC1, mean.y = PC3) ~
                                pop, data, mean), by = "pop")

pca.plot <- ggplot2::ggplot(data, ggplot2::aes(PC1, PC3, color = pop)) +
  ggplot2::geom_point(size = 4) + 
  ggplot2::geom_point(ggplot2::aes(x = mean.x,
                                   y = mean.y, color = pop),
                      size = 7,
                      shape = 17) +
  ggplot2::scale_colour_manual(values = colours5) +
  ggplot2::scale_fill_manual(values = colours5) + 
  ggplot2::geom_hline(yintercept = 0) +
  ggplot2::geom_vline(xintercept = 0) + 
  ggplot2::labs(
    subtitle = paste0("Total variance explained: ",
                      signif(sum(var_frac[1:4]) * 100, 3), "%"),
    y = paste0("PC3; ", signif(var_frac[3] * 100, digits = 2), " %"),
    x = paste0("PC1; ", signif(var_frac[1] * 100, digits = 2), " %"),
    title = plot.title,
    caption = "") + 
  ggplot2::theme_minimal() +
  ggplot2::theme(
    axis.text = ggplot2::element_text(size = 15),
    axis.title.x = ggplot2::element_text(size = 20),
    axis.title.y = ggplot2::element_text(size = 20),
    legend.text = ggplot2::element_text(size = 15)
  )

print(pca.plot)
ggplot2::ggsave(pca.plot, bg = "white",
                filename = paste0(figure.path,"TOA_PCA_CpG_SUB2_regions_PC13.png"), 
                width = 30, height = 15, units = "cm")



data <- data.frame(ind = rownames(PCobj$x),PCobj$x,  pop = pop)
data <- merge(data, aggregate(cbind(mean.x = PC2, mean.y = PC3) ~
                                pop, data, mean), by = "pop")

pca.plot <- ggplot2::ggplot(data, ggplot2::aes(PC2, PC3, color = pop)) +
  ggplot2::geom_point(size = 4) + 
  ggplot2::geom_point(ggplot2::aes(x = mean.x,
                                   y = mean.y, color = pop),
                      size = 7,
                      shape = 17) +
  ggplot2::scale_colour_manual(values = colours5) +
  ggplot2::scale_fill_manual(values = colours5) + 
  ggplot2::geom_hline(yintercept = 0) +
  ggplot2::geom_vline(xintercept = 0) + 
  ggplot2::labs(
    subtitle = paste0("Total variance explained: ",
                      signif(sum(var_frac[1:4]) * 100, 3), "%"),
    y = paste0("PC3; ", signif(var_frac[3] * 100, digits = 2), " %"),
    x = paste0("PC2; ", signif(var_frac[2] * 100, digits = 2), " %"),
    title = plot.title,
    caption = "") + 
  ggplot2::theme_minimal() +
  ggplot2::theme(
    axis.text = ggplot2::element_text(size = 15),
    axis.title.x = ggplot2::element_text(size = 20),
    axis.title.y = ggplot2::element_text(size = 20),
    legend.text = ggplot2::element_text(size = 15)
  )

print(pca.plot)
ggplot2::ggsave(pca.plot, bg = "white",
                filename = paste0(figure.path,"TOA_PCA_CpG_SUB2_regions_PC23.png"), 
                width = 30, height = 15, units = "cm")
```

## Batch effect
```{r}
covariates <- as.data.frame(dplyr::select(metadata2, Sex, mat, Age, Tissue))
rownames(covariates) <- metadata2$SAMPLE_ID
as <- methylKit::assocComp(mBase = meth, sampleAnnotation = covariates)
as
as$pcs[,2]
newObj <- removeComp(meth,comp = 1)
mat <- percMethylation(meth)

```

## DMS

effect:

Method to calculate the mean methylation different between groups using read coverage as weights (default)

when set to "predicted", predicted means from the logistic regression model is used for calculating the effect.


test:
Fisher’s exact test

The Chisq-test is used by default for more than two groups, while the F-test can be chosen if overdispersion control is applied.

overdispersion:

if set to "none"(default), no overdispersion correction will be attempted.
If set to "MN", basic overdispersion correction, proposed by McCullagh and Nelder (1989) will be applied.This correction applies a scaling parameter to variance estimated by the model. 



### Region

Subset 3989 DMS CpG
RANDOMISED: 1564 CpG 
```{r}
print(load(paste0(rdata.path,"Toothfish_subset_methylation_objects_raw.Rdata")))
print(load(paste0(rdata.path,"Toothfish_subset_methylation_objects_raw_NEW.Rdata")))



metadata.new <- metadata[metadata$SAMPLE_ID %in% meth@sample.ids,]
metadata.new <- metadata.new[order(match(metadata.new$SAMPLE_ID, meth@sample.ids)),]
identical(metadata.new$SAMPLE_ID, meth@sample.ids)

# meth@treatment <- as.integer(factor(metadata.new$region))
# meth@treatment <- as.integer(factor(metadata.new$region))
# meth@treatment <- as.integer(factor(metadata.new$FAO2))

#RANDOMLY SET STRATA
# meth@treatment <- as.integer(sample(factor(metadata.new$region)))
# meth@treatment <- as.integer(sample(factor(metadata.new$FAO2)))

# covariates <- dplyr::select(metadata2, Sex, mat, Age, Tissue)
covariates <- dplyr::select(metadata.new, Sex, mat, TL)
# covariates <- dplyr::select(metadata3, Sex, mat, Age)
covariates$mat[is.na(covariates$mat)] <- "3"
meth <- as(meth, "methylBase")
myDiff <- methylKit::calculateDiffMeth(meth,
                                       covariates = covariates,
                                       adjust = "fdr",
                                       chunk.size = 1e+06,
                                       mc.cores = 10)

meth.diff.chr <- methylKit::diffMethPerChr(myDiff,plot = TRUE, 
                                           qvalue.cutoff = 0.01, meth.cutoff = 25)
myDiff25p.hyper <- methylKit::getMethylDiff(myDiff, difference = 25,qvalue = 0.01, type = "hyper")
#2058 
myDiff25p.hypo <- methylKit::getMethylDiff(myDiff, difference = 25,qvalue = 0.01, type = "hypo")
#0
myDiff25p <- methylKit::getMethylDiff(myDiff, difference = 25, qvalue = 0.01)


# save(myDiff,myDiff25p, myDiff25p.hyper, myDiff25p.hypo,
#      file = paste0(rdata.path,
#                    # "Toothfish_DMS_Region_subset1.Rdata"
#                    # "Toothfish_DMS_Region_subset2.Rdata"
#                    # "Toothfish_DMS_Region_subset_RANDOM_STRATA.Rdata"
#                    ))


volcano.plot <- ggplot2::ggplot() +
  ggplot2::geom_point(data = myDiff, ggplot2::aes(x = meth.diff, 
                                                  y = -log10(pvalue))) +
  ggplot2::geom_point(data = myDiff25p, ggplot2::aes(x = meth.diff, 
                                                  y = -log10(pvalue)),
                      colour = "red")
print(volcano.plot)

ggplot2::ggsave(volcano.plot, bg = "white",
                filename = paste0(figure.path,
                                  "TOA_ALL_DMS25_VOLCANO_methylkit_regions.png"
                                  # "TOA_ALL_DMS25_VOLCANO_methylkit_regions_RANDOM_STRATA.png"
                                  ), 
                width = 30, height = 15, units = "cm")
```


logistic regression, which can deal with data strictly bounded between 0 and 1 and with non-constant variance, such as methylation proportion/fraction values. In the logistic regression, it is assumed that fitted values have variation np(1−p), where p is the fitted methylation proportion for a given sample and n is the read coverage

If the observed variance is larger or smaller than assumed by the model, one speaks of under- or over-dispersion. This over/under-dispersion can be corrected by calculating a scaling factor

```{r}
# identical(metadata2$SAMPLE_ID, meth@sample.ids)
# meth@treatment <- as.integer(factor(metadata2$FAO2))
# covariates <- dplyr::select(metadata2, Sex, mat, Age, Tissue)

dm.lr <- methylKit::calculateDiffMeth(meth,
                                      overdispersion = "MN",
                                      test = "Chisq",
                                      covariates = covariates,
                                       adjust = "fdr",
                                       chunk.size = 1e+06,
                                       mc.cores = 10)
dm.lr.25p <- methylKit::getMethylDiff(dm.lr, difference = 25, qvalue = 0.01)

volcano.plot <- ggplot2::ggplot() +
  ggplot2::geom_point(data = dm.lr, ggplot2::aes(x = meth.diff,
                                                 y = -log10(pvalue))) +
  ggplot2::geom_point(data = dm.lr.25p, ggplot2::aes(x = meth.diff,
                                                 y = -log10(pvalue)),
                      colour = "red")
print(volcano.plot)
```


More complex regression models use beta binomial distribution and are particularly useful for better modeling the variance. Similar to logistic regression, their observation follows binomial distribution (number of reads), but methylation proportion itself can vary across samples, according to a beta distribution. It can deal with fitting values in the  [0,1] range and performs better when there is greater variance than expected by the simple logistic model.

```{r}
dm.dss <- calculateDiffMethDSS(meth)

```


#### clustering analysis

```{r}
print(load(paste0(rdata.path,"Toothfish_DMS_Region_subset1.Rdata")))
print(load(paste0(rdata.path,"Toothfish_DMS_Region_subset_RANDOM_STRATA.Rdata")))

meth@treatment <- as.integer(factor(metadata$region))
meth@treatment <- as.integer(factor(metadata$Sex, levels = c("F", "M", "U")))
meth@treatment <- as.integer(factor(metadata$Tissue))
meth@treatment <- as.integer(factor(metadata$Cohort2))
meth@treatment <- as.integer(factor(metadata$quality))


myDiff25p.hyper.gr <- as(myDiff25p.hyper, "GRanges")
meth.diff <- methylKit::selectByOverlap(meth, myDiff25p.hyper.gr) #3989 
# meth.diff <- methylKit::selectByOverlap(meth, myDiff25p.hyper.gr) #4859 #RANDOM

methylKit::clusterSamples(meth.diff, dist = "correlation", method = "ward.D2", plot = TRUE)
dev.print(file = paste0(figure.path,
                        # "cluster_plot_diffMeth_Region_subset.png"
                        "cluster_plot_diffMeth_Region_subset_RANDOM_STRATA.png"
                        ),
          device = png, res = 300,width = 45,height = 15,units = "cm")
```


#### PCA

```{r}
print(load(paste0(rdata.path,"Toothfish_DMS_Region_subset1.Rdata")))
print(load(paste0(rdata.path,"Toothfish_DMS_Region_subset_RANDOM_STRATA.Rdata")))


myDiff25p.hyper.gr <- as(myDiff25p.hyper, "GRanges")
meth.diff <- methylKit::selectByOverlap(meth2, myDiff25p.hyper.gr) #3989 


PCobj <- methylKit::PCASamples(meth, comp = c(1,2), adj.lim = c(0,0), obj.return = TRUE)
PCobj <- methylKit::PCASamples(meth.diff, comp = c(1,2), adj.lim = c(0,0), obj.return = TRUE)


var_frac <- PCobj$sdev/sum(PCobj$sdev)
print(paste0("Total variance explained: ",signif(sum(var_frac[1:3]) * 100, 3), "%")) 
barplot(var_frac[1:20], main = "variance explained", 
        col = heat.colors(length(var_frac[1:20])))

# metadata3 <- metadata2[metadata2$tissue %in% "M",]
metadata.new <- metadata[metadata$SAMPLE_ID %in% meth@sample.ids,]
metadata.new <- metadata.new[order(match(metadata.new$SAMPLE_ID, meth@sample.ids)),]
identical(metadata.new$SAMPLE_ID, meth@sample.ids)
identical(metadata.new$SAMPLE_ID, rownames(PCobj$x))

pop <- metadata.new$region

pop <- as.character(meth.diff@treatment)


# pop <- metadata.new$`Date caught`

# colours5 <- c("#fc9403","#4d6b40","#adcf9f", "#c5d6e6", "#08306B")
colours5 <- c("#4d6b40","#adcf9f", "#c5d6e6", "#08306B")

# plot.title <- "Toothfish methylation (1,316,799 CpG sites)"
# plot.title <- "Toothfish methylation (2,058  DMS methylkit)"
plot.title <- "Toothfish methylation - RANDOM STRATA (306 CpG sites)"

data <- data.frame(ind = rownames(PCobj$x),PCobj$x,  pop = pop)
data <- merge(data, aggregate(cbind(mean.x = PC1, mean.y = PC2) ~
                                pop, data, mean), by = "pop")

pca.plot <- ggplot2::ggplot(data, ggplot2::aes(PC1, PC2, color = pop)) +
  ggplot2::geom_point(size = 4) + 
  ggplot2::geom_point(ggplot2::aes(x = mean.x,
                                   y = mean.y, color = pop),
                      size = 7,
                      shape = 17) +
  ggplot2::scale_colour_manual(values = colours5) +
  ggplot2::scale_fill_manual(values = colours5) + 
  ggplot2::geom_hline(yintercept = 0) +
  ggplot2::geom_vline(xintercept = 0) + 
  ggplot2::labs(
    subtitle = paste0("Total variance explained: ",
                      signif(sum(var_frac[1:4]) * 100, 3), "%"),
    y = paste0("PC2; ", signif(var_frac[2] * 100, digits = 2), " %"),
    x = paste0("PC1; ", signif(var_frac[1] * 100, digits = 2), " %"),
    title = plot.title,
    caption = "") + 
  ggplot2::theme_minimal() +
  ggplot2::theme(
    axis.text = ggplot2::element_text(size = 15),
    axis.title.x = ggplot2::element_text(size = 20),
    axis.title.y = ggplot2::element_text(size = 20),
    legend.text = ggplot2::element_text(size = 15)
  )

print(pca.plot)
ggplot2::ggsave(pca.plot, bg = "white",
                filename = paste0(figure.path,
                                  "TOA_PCA_CpG_DMS25_methylkit_regions_subset_RANDOM_STRATA_PC12.png"
                                  # "TOA_PCA_CpG_DMS25_methylkit_regions_PC12.png"
                                  ), 
                width = 30, height = 15, units = "cm")



data <- data.frame(ind = rownames(PCobj$x),PCobj$x,  pop = pop)
data <- merge(data, aggregate(cbind(mean.x = PC1, mean.y = PC3) ~
                                pop, data, mean), by = "pop")

pca.plot <- ggplot2::ggplot(data, ggplot2::aes(PC1, PC3, color = pop)) +
  ggplot2::geom_point(size = 4) + 
  ggplot2::geom_point(ggplot2::aes(x = mean.x,
                                   y = mean.y, color = pop),
                      size = 7,
                      shape = 17) +
  ggplot2::scale_colour_manual(values = colours5) +
  ggplot2::scale_fill_manual(values = colours5) + 
  ggplot2::geom_hline(yintercept = 0) +
  ggplot2::geom_vline(xintercept = 0) + 
  ggplot2::labs(
    subtitle = paste0("Total variance explained: ",
                      signif(sum(var_frac[1:4]) * 100, 3), "%"),
    y = paste0("PC3; ", signif(var_frac[3] * 100, digits = 2), " %"),
    x = paste0("PC1; ", signif(var_frac[1] * 100, digits = 2), " %"),
    title = plot.title,
    caption = "") + 
  ggplot2::theme_minimal() +
  ggplot2::theme(
    axis.text = ggplot2::element_text(size = 15),
    axis.title.x = ggplot2::element_text(size = 20),
    axis.title.y = ggplot2::element_text(size = 20),
    legend.text = ggplot2::element_text(size = 15)
  )

print(pca.plot)
ggplot2::ggsave(pca.plot, bg = "white",
                filename = paste0(figure.path,
                                  "TOA_PCA_CpG_DMS25_methylkit_regions_subset_RANDOM_STRATA_PC12.png"
                                  # "TOA_PCA_CpG_DMS25_methylkit_regions_PC13.png"
                                  ), 
                width = 30, height = 15, units = "cm")




data <- data.frame(ind = rownames(PCobj$x),PCobj$x,  pop = pop)
data <- merge(data, aggregate(cbind(mean.x = PC2, mean.y = PC3) ~
                                pop, data, mean), by = "pop")

pca.plot <- ggplot2::ggplot(data, ggplot2::aes(PC2, PC3, color = pop)) +
  ggplot2::geom_point(size = 4) + 
  ggplot2::geom_point(ggplot2::aes(x = mean.x,
                                   y = mean.y, color = pop),
                      size = 7,
                      shape = 17) +
  ggplot2::scale_colour_manual(values = colours5) +
  ggplot2::scale_fill_manual(values = colours5) + 
  ggplot2::geom_hline(yintercept = 0) +
  ggplot2::geom_vline(xintercept = 0) + 
  ggplot2::labs(
    subtitle = paste0("Total variance explained: ",
                      signif(sum(var_frac[1:4]) * 100, 3), "%"),
    y = paste0("PC3; ", signif(var_frac[3] * 100, digits = 2), " %"),
    x = paste0("PC2; ", signif(var_frac[2] * 100, digits = 2), " %"),
    title = plot.title,
    caption = "") + 
  ggplot2::theme_minimal() +
  ggplot2::theme(
    axis.text = ggplot2::element_text(size = 15),
    axis.title.x = ggplot2::element_text(size = 20),
    axis.title.y = ggplot2::element_text(size = 20),
    legend.text = ggplot2::element_text(size = 15)
  )

print(pca.plot)
ggplot2::ggsave(pca.plot, bg = "white",
                filename = paste0(figure.path,
                                  "TOA_PCA_CpG_DMS25_methylkit_regions_subset_RANDOM_STRATA_PC12.png"
                                  # "TOA_PCA_CpG_DMS25_methylkit_regions_PC23.png"
                                  ), 
                width = 30, height = 15, units = "cm")
```

#### plot methyaltion difference

```{r}
print(load(paste0(rdata.path,"Toothfish_DMS_Region_subset1.Rdata")))

myDiff25p.hyper.gr@seqnames
length(unique(myDiff25p.hyper.gr@seqnames))

myDiff15.df <- as.data.frame(myDiff25p.hyper.gr)
hist(myDiff15.df$meth.diff)

meth.diff <- methylKit::selectByOverlap(meth, myDiff25p.hyper.gr)
meth.diff.df <- as.data.frame(t(methylKit::percMethylation(meth.diff)))
meth.diff.df[1:10]
# meth.diff.df %<>%
colnames(meth.diff.df) <- paste0(meth.diff$chr,"_",meth.diff$start)
meth.diff.df$individual <- rownames(meth.diff.df) 
identical(metadata2$SAMPLE_ID,meth.diff.df$individual)
meth.diff.df$pop <- metadata2$region

meth.diff.df.plot <- tidyr::gather(meth.diff.df, key = locus, 
                           value = perc_methylation, -individual, -pop,
                           factor_key=TRUE)

meth.diff.df.plot2 <- meth.diff.df.plot[
  meth.diff.df.plot$locus %in% unique(meth.diff.df.plot$locus)[1:100],]

meth.diff.df.plot2.centroid <- meth.diff.df.plot2 %>% dplyr::group_by(locus,pop) %>% 
  dplyr::summarise(mean_pop = mean(perc_methylation, na.rm = TRUE),
            .groups = 'drop')

ggplot2::ggplot() + 
  ggplot2::geom_violin(data = meth.diff.df.plot2, 
                       ggplot2::aes(x = locus, y = perc_methylation)) + 
  ggplot2::geom_point(data = meth.diff.df.plot2.centroid, 
                      ggplot2::aes(x = locus, y = mean_pop, colour = pop))

```


#### Heatmap

```{r}
print(load(paste0(rdata.path,"Toothfish_DMS_Region_subset1.Rdata")))
print(load(paste0(rdata.path,"Toothfish_DMS_Region_subset_RANDOM_STRATA.Rdata")))


myDiff25p.gr <- as(myDiff25p, "GRanges")
meth.diff <- methylKit::selectByOverlap(meth, myDiff25p.gr) #3989 #661

dm.lr.25p.gr <- as(dm.lr.25p, "GRanges")
meth.diff <- methylKit::selectByOverlap(meth, dm.lr.25p.gr) #3


meth.diff.df <- as.data.frame(t(methylKit::percMethylation(meth.diff)))
loc.names <- paste0(meth.diff$chr,"_",meth.diff$start)
colnames(meth.diff.df) <- loc.names

meth.diff.df$individual <- rownames(meth.diff.df) 
metadata.new <- metadata[metadata$SAMPLE_ID %in% meth.diff@sample.ids,]
metadata.new <- metadata.new[order(match(metadata.new$SAMPLE_ID, meth.diff@sample.ids)),]
identical(metadata.new$SAMPLE_ID, meth.diff@sample.ids)
identical(metadata.new$SAMPLE_ID,meth.diff.df$individual)
meth.diff.df$pop <- metadata.new$region

meth.diff.df$pop <- as.character(meth.diff@treatment)


heat.plot.data <- meth.diff.df %>%
  tibble::as_tibble() %>%
  tidyr::gather(key = "CpG", value = "Methylation_percentage", -individual, -pop,
                factor_key = FALSE) %>%
  dplyr::group_by(pop, CpG) %>%
  dplyr::mutate(CpG_mean = mean(Methylation_percentage, na.rm = TRUE)) %>%
  dplyr::arrange(pop, individual, CpG) %>%
  dplyr::mutate(individual = factor(individual, levels = unique(individual))) %>%
  dplyr::mutate(CpG = factor(CpG, levels = unique(CpG)))
cpgs <- heat.plot.data$CpG[heat.plot.data$individual == heat.plot.data$individual[1]]
cpgs_mean <- heat.plot.data$CpG_mean[heat.plot.data$individual == heat.plot.data$individual[1]]
# cpgs <- heat.plot.data$CpG[heat.plot.data$individual == unique(heat.plot.data$individual)[60]]
# cpgs_mean <- heat.plot.data$CpG_mean[heat.plot.data$individual == unique(heat.plot.data$individual)[60]]

cpgs <- cpgs[order(cpgs_mean)]
heat.plot.data$CpG <- factor(heat.plot.data$CpG, levels = unique(cpgs))
heat.plot.data <- heat.plot.data[order(heat.plot.data$CpG),]

heat.plot <- ggplot2::ggplot(data = heat.plot.data) + 
    ggplot2::geom_tile(ggplot2::aes(individual, y = CpG, fill = Methylation_percentage)) +
    ggplot2::scale_fill_gradient2(low = "blue", mid = "yellow", high = "red", midpoint = 50) +
    
    # ggnewscale::new_scale_fill() +
    # ggplot2::geom_tile(ggplot2::aes(individual, y = "pop", fill = pop, height = 10)) +
    # ggplot2::scale_fill_manual(values = c("#4d6b40","#adcf9f", "#c5d6e6", "#08306B")) +
    ggplot2::scale_y_discrete(drop = FALSE) +
    ggplot2::theme(legend.position = "top",
                   axis.title.y = ggplot2::element_blank(),
          axis.text.x = ggplot2::element_text(face = "bold", color = "black", 
                                        size = 10, angle = 90, hjust = 1, vjust = 0))
print(heat.plot)

ggplot2::ggsave(heat.plot, bg = "white",
                filename = paste0(figure.path,
                                  # "TOA_ALLcpg_methDiff_HEATMAP2.png"
                                  "TOA_ALLcpg_methDiff_RANDOM_STRATA_HEATMAP2.png"
                                  ), 
                width = 45, height = 30, units = "cm")
```


### Sex

```{r}
print(load(paste0(rdata.path,"Toothfish_subset_methylation_objects_raw.Rdata")))
print(load(paste0(rdata.path,"Toothfish_subset_methylation_objects_raw_NEW.Rdata")))

metadata.new <- metadata[metadata$SAMPLE_ID %in% meth@sample.ids,]
metadata.new <- metadata.new[order(match(metadata.new$SAMPLE_ID, meth@sample.ids)),]
identical(metadata.new$SAMPLE_ID, meth@sample.ids)

Sex <- metadata.new$Sex
Sex[Sex == "U"] <- "F"
# meth@treatment <- as.integer(factor(Sex, levels = c("F", "M")))
meth@treatment <- as.integer(factor(Sex, levels = c("M", "F")))

covariates <- dplyr::select(metadata.new, region, mat, TL)

myDiff <- methylKit::calculateDiffMeth(meth,
                                       covariates = covariates,
                                       suffix = "Sex",
                                       adjust = "fdr",
                                       chunk.size = 1e+06,
                                       mc.cores = 10)
meth.diff.chr <- methylKit::diffMethPerChr(myDiff,plot = TRUE, 
                                           qvalue.cutoff = 0.01, meth.cutoff = 20)
myDiff25p.hyper <- methylKit::getMethylDiff(myDiff, difference = 20,qvalue = 0.01, type = "hyper")
myDiff25p.hypo <- methylKit::getMethylDiff(myDiff, difference = 20,qvalue = 0.01, type = "hypo")
myDiff25p <- methylKit::getMethylDiff(myDiff, difference = 20, qvalue = 0.01)


save(myDiff,myDiff25p, myDiff25p.hyper, myDiff25p.hypo,
     file = paste0(rdata.path,
                   "Toothfish_DMS_Sex_subset1.Rdata"
                   # "Toothfish_DMS_Region_subset2.Rdata"
                   # "Toothfish_DMS_Region_subset_RANDOM_STRATA.Rdata"
                   ))


myDiff.df <- as.data.frame(as(myDiff, "GRanges"))
myDiff25p.df <- as.data.frame(as(myDiff25p, "GRanges"))
volcano.plot <- ggplot2::ggplot() +
  ggplot2::geom_point(data = myDiff.df, ggplot2::aes(x = meth.diff, 
                                                  y = -log10(pvalue))) +
  ggplot2::geom_point(data = myDiff25p.df, ggplot2::aes(x = meth.diff, 
                                                  y = -log10(pvalue)),
                      colour = "red")
print(volcano.plot)

ggplot2::ggsave(volcano.plot, bg = "white",
                filename = paste0(figure.path,
                                  "TOA_ALL_DMS20_VOLCANO_methylkit_Sex.png"
                                  # "TOA_ALL_DMS25_VOLCANO_methylkit_regions_RANDOM_STRATA.png"
                                  ), 
                width = 30, height = 15, units = "cm")
```


#### clustering analysis

```{r}
myDiff25p.hyper.gr <- as(myDiff25p.hyper, "GRanges")
meth.diff <- methylKit::selectByOverlap(meth, myDiff25p.hyper.gr) #160  

methylKit::clusterSamples(meth.diff, dist = "correlation", method = "ward.D2", plot = TRUE)
dev.print(file = paste0(figure.path,"cluster_plot_diffMeth_Sex_subset_hyper.png"),
          device = png, res = 300,width = 45,height = 15,units = "cm")


myDiff25p.hypo.gr <- as(myDiff25p.hypo, "GRanges")
meth.diff <- methylKit::selectByOverlap(meth, myDiff25p.hypo.gr) #169  

methylKit::clusterSamples(meth.diff, dist = "correlation", method = "ward.D2", plot = TRUE)
dev.print(file = paste0(figure.path,"cluster_plot_diffMeth_Sex_subset_hypo.png"),
          device = png, res = 300,width = 45,height = 15,units = "cm")
```


#### PCA

```{r}
myDiff25p.hyper.gr <- as(myDiff25p.hyper, "GRanges")
meth.diff <- methylKit::selectByOverlap(meth, myDiff25p.hyper.gr) #160  

PCobj <- methylKit::PCASamples(meth.diff, comp = c(1,2), adj.lim = c(0,0), obj.return = TRUE)


var_frac <- PCobj$sdev/sum(PCobj$sdev)
print(paste0("Total variance explained: ",signif(sum(var_frac[1:3]) * 100, 3), "%")) 
barplot(var_frac[1:20], main = "variance explained", 
        col = heat.colors(length(var_frac[1:20])))

# metadata3 <- metadata2[metadata2$tissue %in% "M",]
metadata.new <- metadata[metadata$SAMPLE_ID %in% meth@sample.ids,]
metadata.new <- metadata.new[order(match(metadata.new$SAMPLE_ID, meth@sample.ids)),]
identical(metadata.new$SAMPLE_ID, meth@sample.ids)
identical(metadata.new$SAMPLE_ID, rownames(PCobj$x))

pop <- metadata.new$region

pop <- as.character(meth.diff@treatment)


# pop <- metadata.new$`Date caught`

# colours5 <- c("#fc9403","#4d6b40","#adcf9f", "#c5d6e6", "#08306B")
colours5 <- c("#4d6b40","#adcf9f", "#c5d6e6", "#08306B")

# plot.title <- "Toothfish methylation (1,316,799 CpG sites)"
# plot.title <- "Toothfish methylation (2,058  DMS methylkit)"
plot.title <- "Toothfish methylation - RANDOM STRATA (306 CpG sites)"

data <- data.frame(ind = rownames(PCobj$x),PCobj$x,  pop = pop)
data <- merge(data, aggregate(cbind(mean.x = PC1, mean.y = PC2) ~
                                pop, data, mean), by = "pop")

pca.plot <- ggplot2::ggplot(data, ggplot2::aes(PC1, PC2, color = pop)) +
  ggplot2::geom_point(size = 4) + 
  ggplot2::geom_point(ggplot2::aes(x = mean.x,
                                   y = mean.y, color = pop),
                      size = 7,
                      shape = 17) +
  ggplot2::scale_colour_manual(values = colours5) +
  ggplot2::scale_fill_manual(values = colours5) + 
  ggplot2::geom_hline(yintercept = 0) +
  ggplot2::geom_vline(xintercept = 0) + 
  ggplot2::labs(
    subtitle = paste0("Total variance explained: ",
                      signif(sum(var_frac[1:4]) * 100, 3), "%"),
    y = paste0("PC2; ", signif(var_frac[2] * 100, digits = 2), " %"),
    x = paste0("PC1; ", signif(var_frac[1] * 100, digits = 2), " %"),
    title = plot.title,
    caption = "") + 
  ggplot2::theme_minimal() +
  ggplot2::theme(
    axis.text = ggplot2::element_text(size = 15),
    axis.title.x = ggplot2::element_text(size = 20),
    axis.title.y = ggplot2::element_text(size = 20),
    legend.text = ggplot2::element_text(size = 15)
  )

print(pca.plot)
ggplot2::ggsave(pca.plot, bg = "white",
                filename = paste0(figure.path,
                                  "TOA_PCA_CpG_DMS25_methylkit_Sex_subset_RANDOM_STRATA_PC12.png"
                                  # "TOA_PCA_CpG_DMS25_methylkit_regions_PC12.png"
                                  ), 
                width = 30, height = 15, units = "cm")



data <- data.frame(ind = rownames(PCobj$x),PCobj$x,  pop = pop)
data <- merge(data, aggregate(cbind(mean.x = PC1, mean.y = PC3) ~
                                pop, data, mean), by = "pop")

pca.plot <- ggplot2::ggplot(data, ggplot2::aes(PC1, PC3, color = pop)) +
  ggplot2::geom_point(size = 4) + 
  ggplot2::geom_point(ggplot2::aes(x = mean.x,
                                   y = mean.y, color = pop),
                      size = 7,
                      shape = 17) +
  ggplot2::scale_colour_manual(values = colours5) +
  ggplot2::scale_fill_manual(values = colours5) + 
  ggplot2::geom_hline(yintercept = 0) +
  ggplot2::geom_vline(xintercept = 0) + 
  ggplot2::labs(
    subtitle = paste0("Total variance explained: ",
                      signif(sum(var_frac[1:4]) * 100, 3), "%"),
    y = paste0("PC3; ", signif(var_frac[3] * 100, digits = 2), " %"),
    x = paste0("PC1; ", signif(var_frac[1] * 100, digits = 2), " %"),
    title = plot.title,
    caption = "") + 
  ggplot2::theme_minimal() +
  ggplot2::theme(
    axis.text = ggplot2::element_text(size = 15),
    axis.title.x = ggplot2::element_text(size = 20),
    axis.title.y = ggplot2::element_text(size = 20),
    legend.text = ggplot2::element_text(size = 15)
  )

print(pca.plot)
ggplot2::ggsave(pca.plot, bg = "white",
                filename = paste0(figure.path,
                                  "TOA_PCA_CpG_DMS25_methylkit_Sex_subset_RANDOM_STRATA_PC12.png"
                                  # "TOA_PCA_CpG_DMS25_methylkit_regions_PC13.png"
                                  ), 
                width = 30, height = 15, units = "cm")




data <- data.frame(ind = rownames(PCobj$x),PCobj$x,  pop = pop)
data <- merge(data, aggregate(cbind(mean.x = PC2, mean.y = PC3) ~
                                pop, data, mean), by = "pop")

pca.plot <- ggplot2::ggplot(data, ggplot2::aes(PC2, PC3, color = pop)) +
  ggplot2::geom_point(size = 4) + 
  ggplot2::geom_point(ggplot2::aes(x = mean.x,
                                   y = mean.y, color = pop),
                      size = 7,
                      shape = 17) +
  ggplot2::scale_colour_manual(values = colours5) +
  ggplot2::scale_fill_manual(values = colours5) + 
  ggplot2::geom_hline(yintercept = 0) +
  ggplot2::geom_vline(xintercept = 0) + 
  ggplot2::labs(
    subtitle = paste0("Total variance explained: ",
                      signif(sum(var_frac[1:4]) * 100, 3), "%"),
    y = paste0("PC3; ", signif(var_frac[3] * 100, digits = 2), " %"),
    x = paste0("PC2; ", signif(var_frac[2] * 100, digits = 2), " %"),
    title = plot.title,
    caption = "") + 
  ggplot2::theme_minimal() +
  ggplot2::theme(
    axis.text = ggplot2::element_text(size = 15),
    axis.title.x = ggplot2::element_text(size = 20),
    axis.title.y = ggplot2::element_text(size = 20),
    legend.text = ggplot2::element_text(size = 15)
  )

print(pca.plot)
ggplot2::ggsave(pca.plot, bg = "white",
                filename = paste0(figure.path,
                                  "TOA_PCA_CpG_DMS25_methylkit_Sex_subset_RANDOM_STRATA_PC12.png"
                                  # "TOA_PCA_CpG_DMS25_methylkit_regions_PC23.png"
                                  ), 
                width = 30, height = 15, units = "cm")
```

#### Heatmap

```{r}
print(load(paste0(rdata.path,"Toothfish_DMS_Sex_subset1.Rdata")))
# print(load(paste0(rdata.path,"Toothfish_DMS_Region_subset_RANDOM_STRATA.Rdata")))


myDiff25p.gr <- as(myDiff25p, "GRanges")
meth.diff <- methylKit::selectByOverlap(meth, myDiff25p.gr) #3989 #661

meth.diff.df <- as.data.frame(t(methylKit::percMethylation(meth.diff)))
loc.names <- paste0(meth.diff$chr,"_",meth.diff$start)
colnames(meth.diff.df) <- loc.names

meth.diff.df$individual <- rownames(meth.diff.df) 
metadata.new <- metadata[metadata$SAMPLE_ID %in% meth.diff@sample.ids,]
metadata.new <- metadata.new[order(match(metadata.new$SAMPLE_ID, meth.diff@sample.ids)),]
identical(metadata.new$SAMPLE_ID, meth.diff@sample.ids)
identical(metadata.new$SAMPLE_ID,meth.diff.df$individual)
meth.diff.df$pop <- metadata.new$Sex

# meth.diff.df$pop <- as.character(meth.diff@treatment)


heat.plot.data <- meth.diff.df %>%
  tibble::as_tibble() %>%
  tidyr::gather(key = "CpG", value = "Methylation_percentage", -individual, -pop,
                factor_key = FALSE) %>%
  dplyr::group_by(pop, CpG) %>%
  dplyr::mutate(CpG_mean = mean(Methylation_percentage, na.rm = TRUE)) %>%
  dplyr::arrange(pop, individual, CpG) %>%
  dplyr::mutate(individual = factor(individual, levels = unique(individual))) %>%
  dplyr::mutate(CpG = factor(CpG, levels = unique(CpG)))
cpgs <- heat.plot.data$CpG[heat.plot.data$individual == heat.plot.data$individual[1]]
cpgs_mean <- heat.plot.data$CpG_mean[heat.plot.data$individual == heat.plot.data$individual[1]]
# cpgs <- heat.plot.data$CpG[heat.plot.data$individual == unique(heat.plot.data$individual)[60]]
# cpgs_mean <- heat.plot.data$CpG_mean[heat.plot.data$individual == unique(heat.plot.data$individual)[60]]

cpgs <- cpgs[order(cpgs_mean)]
heat.plot.data$CpG <- factor(heat.plot.data$CpG, levels = unique(cpgs))
heat.plot.data <- heat.plot.data[order(heat.plot.data$CpG),]

heat.plot <- ggplot2::ggplot(data = heat.plot.data) + 
    ggplot2::geom_tile(ggplot2::aes(individual, y = CpG, fill = Methylation_percentage)) +
    ggplot2::scale_fill_gradient2(low = "blue", mid = "yellow", high = "red", midpoint = 50) +
    
    ggnewscale::new_scale_fill() +
    ggplot2::geom_tile(ggplot2::aes(individual, y = "pop", fill = pop, height = 2)) +
    ggplot2::scale_fill_manual(values = c("#4d6b40","#adcf9f", "#c5d6e6", "#08306B")) +
    ggplot2::scale_y_discrete(drop = FALSE) +
    ggplot2::theme(legend.position = "top",
                   axis.title.y = ggplot2::element_blank(),
          axis.text.x = ggplot2::element_text(face = "bold", color = "black", 
                                        size = 10, angle = 90, hjust = 1, vjust = 0))
print(heat.plot)

ggplot2::ggsave(heat.plot, bg = "white",
                filename = paste0(figure.path,
                                  "TOA_Sex_methDiff_HEATMAP1.png"
                                  # "TOA_ALLcpg_methDiff_RANDOM_STRATA_HEATMAP2.png"
                                  ), 
                width = 45, height = 30, units = "cm")
```


### Tissue

```{r}
Tissue <- metadata$Tissue
Tissue[Tissue == "Skin"] <- "fin"
meth@treatment <- as.integer(factor(Tissue))

covariates <- dplyr::select(metadata, region, mat, Age, Sex)

myDiff <- methylKit::calculateDiffMeth(meth,
                                       covariates = covariates,
                                      suffix = "tissue",
                                       adjust = "fdr",
                                       chunk.size = 1e+06,
                                       mc.cores = 10)
meth.diff.chr <- methylKit::diffMethPerChr(myDiff,plot = TRUE, 
                                           qvalue.cutoff = 0.01, meth.cutoff = 15)
myDiff25p.hyper <- methylKit::getMethylDiff(myDiff, difference = 15,qvalue = 0.01, type = "hyper")
myDiff25p.hypo <- methylKit::getMethylDiff(myDiff, difference = 15,qvalue = 0.01, type = "hypo")
myDiff25p <- methylKit::getMethylDiff(myDiff, difference = 15, qvalue = 0.01)

dm.dss <- calculateDiffMethDSS(meth)
```


#### clustering analysis

```{r}
myDiff25p.hyper.gr <- as(myDiff25p.hyper, "GRanges")
meth.diff <- methylKit::selectByOverlap(meth, myDiff25p.hyper.gr) #4631 

methylKit::clusterSamples(meth.diff, dist = "correlation", method = "ward.D2", plot = TRUE)
dev.print(file = paste0(figure.path,"cluster_plot_diffMeth_Tissue.png"),
          device = png, res = 300,width = 45,height = 15,units = "cm")
```


#### PCA

```{r}
methylKit::PCASamples(meth.diff, screeplot = TRUE)
dev.print(file = paste0(figure.path,"PCA_plot_diffMeth_Tissue_scree.png"),
          device = png, res = 300,width = 30,height = 15,units = "cm")
methylKit::PCASamples(meth.diff)
dev.print(file = paste0(figure.path,"PCA_plot_diffMeth_Tissue.png"),
          device = png, res = 300,width = 30,height = 15,units = "cm")
```

#### Features

```{r}
meth.diff.gr <- as(meth.diff, "GRanges")
ann <- genomation::annotateWithFeatures(meth.diff.gr, annotation.all.features, intersect.chr = TRUE)# 0.01%, 0.01%
ann
```


## DMR

```{r}
# This function can be used when differential methylation analysis is preferable to tilling windows instead of base pairs.
# cov.bases: minimum number of bases to be covered in a given window
tile.res <- methylKit::tileMethylCounts(meth, win.size = 100, step.size = 100, 
                                        cov.bases = 3, save.db = FALSE, mc.cores = 20)

methylKit::clusterSamples(tile.res, dist = "correlation", method = "ward.D2", plot = TRUE)
dev.print(file = paste0(figure.path,"cluster_plot_Tile3.png"),
          device = png, res = 300,width = 45,height = 15,units = "cm")
methylKit::PCASamples(tile.res)
dev.print(file = paste0(figure.path,"PCA_plot_Tile3.png"),
          device = png, res = 300,width = 45,height = 15,units = "cm")

tile.res7 <- methylKit::tileMethylCounts(meth, win.size = 1000, step.size = 1000, 
                                           cov.bases = 4, save.db = FALSE, mc.cores = 20)

methylKit::clusterSamples(tile.res7, dist = "correlation", method = "ward.D2", plot = TRUE)
dev.print(file = paste0(figure.path,"cluster_plot_Tile7.png"),
          device = png, res = 300,width = 45,height = 15,units = "cm")
methylKit::PCASamples(tile.res7)
dev.print(file = paste0(figure.path,"PCA_plot_Tile7.png"),
          device = png, res = 300,width = 45,height = 15,units = "cm")

tile.res50 <- methylKit::tileMethylCounts(meth, win.size = 10000, step.size = 10000, 
                                           cov.bases = 50, save.db = FALSE, mc.cores = 20)
methylKit::clusterSamples(tile.res500, dist = "correlation", method = "ward.D2", plot = TRUE)
dev.print(file = paste0(figure.path,"cluster_plot_Tile500.png"),
          device = png, res = 300,width = 45,height = 15,units = "cm")
methylKit::PCASamples(tile.res500)
dev.print(file = paste0(figure.path,"PCA_plot_Tile500.png"),
          device = png, res = 300,width = 45,height = 15,units = "cm")
```

```{r}
covariates <- dplyr::select(metadata2, Sex, mat, Age, Tissue)
myDiff <- methylKit::calculateDiffMeth(meth,
                                       covariates = covariates,
                                       adjust = "fdr",
                                       chunk.size = 1e+06,
                                       mc.cores = 10)
meth.diff.chr <- methylKit::diffMethPerChr(myDiff,plot = TRUE, 
                                           qvalue.cutoff = 0.01, meth.cutoff = 15)
myDiff25p.hyper <- methylKit::getMethylDiff(myDiff, difference = 15,qvalue = 0.01, type = "hyper")
myDiff25p.hypo <- methylKit::getMethylDiff(myDiff, difference = 15,qvalue = 0.01, type = "hypo")
myDiff25p <- methylKit::getMethylDiff(myDiff, difference = 15, qvalue = 0.01)

dm.dss <- calculateDiffMethDSS(meth)

#Hypomethylatated for bad DNA samples
#              chr   start     end strand       pvalue       qvalue meth.diff
# 1 JAAKFY010000021 5537317 5537317      * 5.592448e-39 6.829312e-35 -25.26441

```


### clustering analysis

```{r}
meth@treatment <- as.integer(factor(metadata$region))
meth@treatment <- as.integer(factor(metadata$Sex, levels = c("F", "M", "U")))
meth@treatment <- as.integer(factor(metadata$Tissue))
meth@treatment <- as.integer(factor(metadata$Cohort2))
meth@treatment <- as.integer(factor(metadata$quality))


methylKit::clusterSamples(meth, dist = "correlation", method = "ward.D2", plot = TRUE)
dev.print(file = paste0(figure.path,"cluster_plot02.png"),
          device = png, res = 300,width = 45,height = 15,units = "cm")
```


### PCA

```{r}
methylKit::PCASamples(meth, screeplot = TRUE)
methylKit::PCASamples(meth)
dev.print(file = paste0(figure.path,"PCA_plot01.png"),
          device = png, res = 300,width = 30,height = 15,units = "cm")
```

## Segment analysis

```{r}
pm <- methylkit::percMethylation(meth) # get percent methylation matrix
mds <- matrixStats::rowSds(pm) # calculate standard deviation of CpGs
head(meth[mds>20,])
hist(mds,col="cornflowerblue",xlab="Std. dev. per CpG")


my.gr <- as(myobjDB[[1]],"GRanges")
my.raw <- as(myobjDB[[1]],"methylRaw")

sm_bin = ScoreMatrixBin(target = my.gr, windows = promoter, bin.num = 50)
plotMeta(sm_bin, xcoords = c(-1000, 1000))



# it finds the optimal number of componets as 6
int <- methylKit::methSeg(my.raw, diagnostic.plot = TRUE,
                          maxInt = 100, minSeg = 10)
# however the BIC stabilizes after 4, we can also try 4 componets
res <- methylKit::methSeg(my.raw,diagnostic.plot=TRUE,maxInt=100,minSeg=10,G=1:4)
res1 <- methylKit::methSeg(my.raw, diagnostic.plot = TRUE, maxInt = 100, 
                          minSeg = 10, G = 1:4, 
                          join.neighbours = TRUE)

plot(res1$seg.mean,
     log10(width(res1)),pch=20,
     col=scales::alpha(rainbow(4)[as.numeric(res1$seg.group)], 0.2),
     ylab="log10(length)",
     xlab="methylation proportion")

# get segments to BED file
methSeg2bed(res,filename="H1.chr21.chr22.trial.seg.bed")
```


```{r}
chr.file <- "/datasets/work/ncmi-toa-rrbs/work/1.Toothfish/Bismark_genome_index/sizes.chr"
chrom.info <- read.table(file = chr.file, header = F)
chrom.info$V3 <- FALSE
colnames(chrom.info) <- c("chrom", "length", "is_circular")
chrom.info <- chrom.info[order(chrom.info$chrom),]
my.gr <- as(meth,"GRanges")

chrom.info2 <- chrom.info[chrom.info$chrom %in% paste0(my.gr@seqinfo@seqnames,".1"),] 
my.gr@seqinfo@seqlengths <- chrom.info2$length 
my.gr@seqinfo@is_circular <- rep(FALSE, length(my.gr@seqinfo@is_circular))
my.gr@seqinfo@genome <- rep("D.mawsoni", length(my.gr@seqinfo@genome))

gr.windows <- tileGenome(seqinfo(my.gr), tilewidth = 100, cut.last.tile.in.chrom = TRUE)

cov.names <- names(my.gr@elementMetadata)[grepl(pattern = "coverage*", names(my.gr@elementMetadata))]
my.gr2 <- my.gr[,cov.names]

#The coverage method quantifies the degree of overlap for all the ranges in a GRanges object.
gr.data.cov <- GenomicRanges::coverage(my.gr2, width = 10000)

seqlevels(gr.windows, pruning.mode="coarse") <- names(gr.data.cov)

gr.data.binnedAvg <- binnedAverage(gr.windows, gr.data.cov, "value")
gr.data.binnedAvg <- binnedAverage(promoters.gencode, gr.data.cov, "value")


GenomicRanges::intersect(my.gr, gr.windows)
o <-  findOverlaps(my.gr, gr.windows)
c <- countOverlaps(my.gr, gr.windows) # number of cpg per 100 bp
c <- countOverlaps(gr.windows, my.gr) # number of cpg per 100 bp
summary(c)

c <- countOverlaps(my.gr, promotor2) # number of cpg per 2kb promoter
summary(c)
c <- countOverlaps(promotor2, my.gr) # number of cpg per 100 bp
summary(c)
c <- countOverlaps(promoters.gencode, my.gr) # number of cpg per 100 bp

seqlevels(promoters.gencode)
seqlevels(my.gr)


# d <- disjoin(my.gr, gr.windows, with.revmap = TRUE, ignore.strand = TRUE)
r <- reduce(xyz, with.revmap = TRUE)


olaps <- findOverlaps(my.gr, gr.windows)
df <- DataFrame(seqnames=seqnames(gr.windows)[subjectHits(olaps)],
                 start=start(gr.windows)[subjectHits(olaps)],
                 Score=my.gr@elementMetadata[queryHits(olaps),])



getScores <- function(interval) {
    scores <- Rle(0, width(interval))
    bases <- GRanges(
        seqnames = seqnames(interval),
        ranges = IRanges(start(interval):end(interval), width = 1))
    overlaps <- findOverlaps(signal, bases)
    scores[start(bases)[subjectHits(overlaps)] - start(interval) + 1] <- score(signal)[queryHits(overlaps)]
    scores
}
Reduce('+', sapply(split(intervals, 1:length(intervals)), getScores)) / length(intervals)



regional.methylRaw <- methylKit::regionCounts(object=meth, regions=gr.windows,
                                cov.bases=0,strand.aware=FALSE, save.db = FALSE)
my.gr.100 <- as(regional.methylRaw,"GRanges")

```


```{r}

averagePerBin <- function(x, binsize, mcolnames=NULL)
{
     if (!is(x, "GenomicRanges"))
         stop("'x' must be a GenomicRanges object")
     if (any(is.na(seqlengths(x))))
         stop("'seqlengths(x)' contains NAs")
     bins <- IRangesList(lapply(seqlengths(x),
                                function(seqlen)
                                  IRanges::IRanges(IRanges::breakInChunks(totalsize = seqlen,
                                                                          chunksize = binsize))))
     ans <- as(bins, "GRanges")
     seqinfo(ans) <- seqinfo(x)
     if (is.null(mcolnames))
         return(ans)
     averageMCol <- function(colname)
     {
         cvg <- GenomicRanges::coverage(x, weight = colname)
         views_list <- RleViewsList(
                           lapply(names(cvg),
                               function(seqname)
                                   Views(cvg[[seqname]], bins[[seqname]])))
         unlist(viewMeans(views_list), use.names=FALSE)
     }
     mcols(ans) <- DataFrame(lapply(mcols(x)[mcolnames], averageMCol))
     ans
}

## Set the sequence lengths.
seqlengths(my.gr)

## Add the density metadata col.
mcols(my.gr)
cov.names <- names(my.gr@elementMetadata)[grepl(pattern = "coverage*", names(my.gr@elementMetadata))]

## Compute the average per bin for the specified metadata cols.
avg_per_bin <- averagePerBin(x = my.gr, binsize = 100, mcolnames = "coverage1")

avg_per_bin[avg_per_bin$density > 0]
```

## Annotate

```{r}
gff.file <- "/datasets/work/ncmi-toa-rrbs/work/1.Toothfish/Bismark_genome_index/D.mawsoni.gff"
gtf.file <- "/datasets/work/ncmi-toa-rrbs/work/1.Toothfish/Bismark_genome_index/D.mawsoni.gtf"
gtf.file <- "/datasets/work/ncmi-toa-rrbs/work/1.Toothfish/Bismark_genome_index/D.mawsoni_2.gtf"

bed.file <- "/datasets/work/ncmi-toa-rrbs/work/1.Toothfish/Bismark_genome_index/D.mawsoni.bed"

# GFF <- read.csv(gff.file, sep = '\t', header = F)
# BED <- GFF[ ,c(1,4,5,9,8,7)]
# BED$V8 <- 100
# BED$V9 <- gsub(".*GeneID:|;Name.*", "", BED$V9)
# write.table(BED, bed.file, row.names = F, col.names = F, quote = F, sep = "\t")

promotor.file <- "/datasets/work/ncmi-toa-rrbs/work/1.Toothfish/Bismark_genome_index/promoters.bed"
genes.file <- "/datasets/work/ncmi-toa-rrbs/work/1.Toothfish/Bismark_genome_index/genes.bed"
chr.file <- "/datasets/work/ncmi-toa-rrbs/work/1.Toothfish/Bismark_genome_index/sizes.chr"

chrom.info <- read.table(file = chr.file, header = F)
chrom.info$V3 <- FALSE
colnames(chrom.info) <- c("chrom", "length", "is_circular")
chrom.info$chrom <- paste0(chrom.info$chrom,".1")

gencode <- GenomicFeatures::makeTxDbFromGFF(gtf.file,
                format = "gtf",
                dataSource = "Annotation From Chen et al. 2021",
                organism = "Dissostichus mawsoni",
                taxonomyId = NA,
                circ_seqs = NULL,
                chrominfo = chrom.info,
                miRBaseBuild = NA,
                metadata = NULL)

# saveDb(gencode, file = paste0(rdata.path, "gencode_D_mawsoni.sqlite"))
loadDb(paste0(rdata.path, "gencode_D_mawsoni.sqlite"))

#### Create GRAnges objects for the each feature
genes.gencode <- sort(GenomicFeatures::genes(gencode))
intergenic.gencode <- GenomicRanges::gaps(genes.gencode)
transcript.gencode <- sort(GenomicFeatures::transcripts(gencode, use.names = TRUE))
cds.gencode <- sort(GenomicFeatures::cds(gencode))
exons.gencode <- sort(GenomicFeatures::exons(gencode))
introns.gencode <- sort(unlist(GenomicFeatures::intronsByTranscript(gencode)))
promoters.gencode <- sort(GenomicFeatures::promoters(gencode, upstream = 1000, downstream = 1000))
TSS.gencode <- promoters.gencode
TSS.gencode <- transcript.gencode
TSS.gencode@ranges@width <- rep(1L, times = length(TSS.gencode@ranges@width))


upstream <- GenomicFeatures::extractUpstreamSeqs(x, genes.gencode, width = 1000)


length(GenomicFeatures::intronsByTranscript(gencode)) #29240
introns <- GenomicFeatures::intronsByTranscript(gencode)
introns <- unlist(introns) #209109
summary(IRanges::width(introns))

transcriptsBy(x, by=c("gene", "exon", "cds"), ...)
exonsBy(x, by=c("tx", "gene"), ...)

length(GenomicFeatures::fiveUTRsByTranscript(gencode)) #7486
length(GenomicFeatures::threeUTRsByTranscript(txdb))

# save the combined obj for easy loading in future.

# save(genes.gencode,intergenic.gencode,transcript.gencode,cds.gencode,
#      exons.gencode,introns.gencode, 
#      file = paste0(rdata.path, "gencode_all_features.Rdata"))

# Load the database in future
load(paste0(rdata.path, "gencode_all_features.Rdata"))

# Alternatively save all the above objects into a single 'GenomicRangesList'
annotation.all.features <-  GenomicRanges::GRangesList(
  promoters = promoters.gencode, 
  exons = genes.gencode,
  introns = introns.gencode,
  intergenics = intergenic.gencode,
  transcripts = transcript.gencode,
  cds = cds.gencode,
  TSSes = TSS.gencode)

elementNROWS(annotation.all.features)
table(seqnames(annotation.all.features))
as(annotation.all.features, "IRangesList")


annotation.all.features@partitioning
promoters.gencode@seqnames

# save this list of features
save(annotation.all.features, file = paste0(rdata.path,"gencode_all_features_list.Rdata"))
# Load in future
load(paste0(rdata.path,"gencode_all_features_list.Rdata"))

my.gr <- as(meth,"GRanges")
GenomeInfoDb::seqlevels(my.gr) <- paste0(GenomeInfoDb::seqlevels(my.gr),".1")
my.gr@seqnames <- S4Vectors::Rle(factor(my.gr@seqnames, 
                                        levels = sort(levels(promoters.gencode@seqnames))))

genomation::annotateWithFeature(TOA.gr, genes.gencode, intersect.chr = TRUE) #48.79%, 11.05%
genomation::annotateWithFeature(TOA.gr, transcript.gencode, intersect.chr = TRUE)
genomation::annotateWithFeature(TOA.gr, cds.gencode, intersect.chr = TRUE) #17.97%, 0.6#
genomation::annotateWithFeature(TOA.gr, intergenic.gencode, intersect.chr = TRUE) #97.83%, 23.19%
genomation::annotateWithFeature(TOA.gr, exons.gencode, intersect.chr = TRUE) #20.57%, 0.65%
genomation::annotateWithFeature(TOA.gr, introns.gencode, intersect.chr = TRUE) #29.23%, 1.23%
genomation::annotateWithFeature(TOA.gr, promoters.gencode, intersect.chr = TRUE)# 7.61%, 1.75%
genomation::annotateWithFeature(TOA.gr, TSS.gencode, intersect.chr = TRUE)# 0.01%, 0.01%


gtf <-  genomation::gffToGRanges(gtf.file)
grl21 <- as(split(gtf, gtf$type), "GRangesList")
annot <- genomation::annotateWithFeatures(gr, grl21)
genomation::getTargetAnnotationStats(annot,percentage=TRUE,precedence=TRUE)
genomation::plotTargetAnnotation(annot,precedence=TRUE,
    main="differential methylation annotation")

genomation::calculateOverlapSignificance(target = my.gr, feature = gtf, 
                                         chrom.sizes = my.gr@seqinfo@seqlengths,
  stranded = TRUE, keep.strand.prop = TRUE, keep.chrom = TRUE,
  exclude = NULL, include = NULL, seed = NULL, nrand = 1)

elementMetadata(my.gr)

str(gr)
gr@seqnames <- S4Vectors::Rle(factor(paste0(gr@seqnames,".1")))

gtf.annot <- genomation::annotateWithGeneParts(target = TOA.gr, feature = annotation.all.features
                                               ,strand = FALSE,
                                               intersect.chr = TRUE)
gtf.annot <- genomation::annotateWithGeneParts(target = TOA.gr, feature = grl21
                                               ,strand = FALSE,
                                               intersect.chr = TRUE)
head(getAssociationWithTSS(annotation.all.features))


setwd("/datasets/work/ncmi-toa-rrbs/work/1.Toothfish/Bismark_genome_index/")
feats <- genomation::readTranscriptFeatures(bed.file)
gff <- genomation::readTranscriptFeatures(system.file("D.mawsoni.gtf", package = "methylKit"))


gff <- genomation::readTranscriptFeatures(gtf.file)

# gff.file = system.file("extdata/chr21.refseq.hg19.gtf", package = "genomation")
gff <-  genomation::gffToGRanges(gff.file)


gr <- as(meth,"GRanges")
gtf.annot <- genomation::annotateWithGeneParts(target = gr, feature = gtf,strand = FALSE, 
                                    intersect.chr = TRUE)

promotor <- genomation::readGeneric(promotor.file, header = FALSE, keep.all.metadata = TRUE)
genes <- genomation::readGeneric(genes.file, header = FALSE, keep.all.metadata = TRUE)
cpgi.flanks <- genomation::readFeatureFlank(gff.file)

prom.annot <- genomation::annotateWithGeneParts(target = my.gr, feature = promotor,strand = TRUE, 
                                    intersect.chr = TRUE)
head(getAssociationWithTSS(cage.annot))

feats <- genomation::readTranscriptFeatures(promotor.file, remove.unusual = TRUE,
                                            up.flank = 1000, down.flank = 1000, 
                                            unique.prom = TRUE)

names(feats)
sapply(feats, head)


```

```{r}

promotor2 <- sort(promotor[promotor@ranges@width == 2001,]) #5 shorter regions
promoters.gencode2 <- sort(promoters.gencode[promoters.gencode@ranges@width == 2000,]) #5 shorter regions
#  when 'width' has names, they must be identical to 'seqlevels(x)'


gr <- as(meth,"GRanges")
sum(my.gr@seqnames %in% promoters.gencode2@seqnames)
sum(seqlevels(my.gr) %in% seqlevels(promoters.gencode2))
promoters.gencode3 <- promoters.gencode2[seqlevels(promoters.gencode2) %in% seqlevels(my.gr),]
my.gr %over% promoters.gencode2
seqlengths(my.gr)
seqlevels(my.gr)

countOverlaps(promoters.gencode2, promotor2)

which(GenomicRanges::countOverlaps(promoters.gencode2, my.gr) > 0)
which(GenomicRanges::countOverlaps(promoters.gencode2, promotor2) > 0)

names(promoters.gencode2) <- NULL
names(promotor2)
names(my.gr)

sm <- genomation::ScoreMatrix(target = my.gr, windows = promoters.gencode2, strand.aware = TRUE,
                               weight.col = NULL, is.noCovNA = FALSE, type = "auto", 
                              rpm = FALSE, unique = FALSE, extend = 0, param = NULL, 
                              bam.paired.end = FALSE, library.size = NULL)
genomation::heatMatrix(sm, xcoords = c(-1000, 1000))

sm <- genomation::ScoreMatrix(target = gr, windows = promotor2, strand.aware = TRUE)
genomation::heatMatrix(sm, xcoords = c(-1000, 1000))


cpg.ind <- which(GenomicRanges::countOverlaps(promotor2, my.gr) > 0)
nocpg.ind <- which(GenomicRanges::countOverlaps(promotor2, my.gr) == 0)
genomation::heatMatrix(sm, xcoords = c(-1000, 1000), group = list(CpGi = cpg.ind, noCpGi = nocpg.ind))

sm.scaled <- genomation:: scaleScoreMatrix(sm)
genomation::heatMatrix(sm.scaled, xcoords = c(-1000, 1000), group = list(CpGi = cpg.ind, noCpGi = nocpg.ind))

heatMeta(mat, centralTend = "mean", profile.names = NULL, xcoords = NULL,
  col = NULL, meta.rescale = FALSE, winsorize = c(0, 100),
  legend.name = NULL, cex.legend = 1, xlab = NULL, main = "",
  cex.lab = 1, cex.axis = 1)

```

### TSS

```{r}
my.gr <- as(meth,"GRanges")
GenomeInfoDb::seqlevels(my.gr) <- paste0(GenomeInfoDb::seqlevels(my.gr),".1")
my.gr@seqnames <- S4Vectors::Rle(factor(my.gr@seqnames, 
                                        levels = sort(levels(promoters.gencode@seqnames))))
GenomeInfoDb::seqinfo(promoters.gencode)

promoters.gencode@seqnames <- S4Vectors::Rle(factor(promoters.gencode@seqnames, 
                                                    levels = sort(levels(promoters.gencode@seqnames))))

sum(levels(promoters.gencode@seqnames) == levels(my.gr@seqnames))
sum(seqlevels(promoters.gencode) %in% seqlevels(my.gr))

peak.annot <- genomation::annotateWithFeature(my.gr, promoters.gencode, intersect.chr = TRUE)
peak.annot <- genomation::annotateWithFeature(target = my.gr, feature = annotation.all.features, intersect.chr = TRUE)

promoters.list <- GenomicRanges::GRangesList(promoters.gencode)
annot <- genomation::annotateWithGeneParts(target = my.gr, feature = promoters.list, intersect.chr = TRUE)
  
annot <- genomation::annotateWithGeneParts(target = my.gr, feature = annotation.all.features, intersect.chr = TRUE)

annot <- genomation::annotateWithGeneParts(target = my.gr, feature = grl21, intersect.chr = TRUE)
test <- genomation::getAssociationWithTSS(peak.annot)
test <- genomation::getAssociationWithTSS(grl21)
```


### GeneDMR

```{r}
inputmethfile <- Methfile_read(paths = "C:/Users/Desktop/methdata", suffix = ".gz")
inputrefseqfile <- Bedfile_read(paths = "C:/Users/Desktop/methdata", bedfile = "refseq", suffix = ".txt", feature = FALSE)

allDMGs <- Quick_GeneDMRs(paths = paste(system.file(package = "GeneDMRs"), "/methdata", sep=""))
controls <- c("C:/Users/GeneDMRs/methdata/1_1.gz", "C:/Users/GeneDMRs/methdata/1_2.gz", "C:/Users/GeneDMRs/methdata/1_3.gz")
cases <- c("C:/Users/GeneDMRs/methdata/2_1.gz", "C:/Users/GeneDMRs/methdata/2_1.gz")
allDMGs <- Quick_GeneDMRs(paths = "C:/Users/GeneDMRs/methdata", control_paths = controls, case_paths = cases)
```



# DMS PERCENTAGE
## plot CpG difference
```{r}
# print(load(paste0(rdata.path,"Toothfish_subset_methylation_objects_raw.Rdata")))
print(load(paste0(rdata.path,"Toothfish_methylation_objects_All_pop_no_outliers.Rdata")))
meth <- as(meth2, "methylBase")

print(load(paste0(rdata.path,"Toothfish_metadata.Rdata")))
rm <- c("TOA_58-4-1_277", "TOA_58-4-1_289", "TOA_58-4-1_290", "TOA_58-4-1_295")
rm2 <- c("TOA_58-4-1_118", "TOA_58-4-1_121", "TOA_58-4-1_125")
metadata3 <- metadata2[!metadata2$SAMPLE_ID %in% c(rm, rm2),]


meth.diff.df <- as.data.frame(t(methylKit::percMethylation(meth)))
loc.names <- paste0(meth$chr,"_",meth$start)
colnames(meth.diff.df) <- loc.names

meth.diff.df$individual <- rownames(meth.diff.df) 
# identical(metadata2$SAMPLE_ID, meth.diff.df$individual)
# meth.diff.df$pop <- metadata2$region
identical(metadata3$SAMPLE_ID, meth.diff.df$individual)
meth.diff.df$pop <- metadata3$region



meth.diff.df.plot <- tidyr::gather(meth.diff.df, key = locus, 
                           value = perc_methylation, -individual, -pop,
                           factor_key = TRUE) %>%
  dplyr::mutate(chr = stringr::str_split_fixed(locus, 
                                               pattern = "_", 
                                               n = 2)[,1])
meth.diff.df.plot.locus <- meth.diff.df.plot %>% 
  dplyr::group_by(chr,locus,pop) %>% 
  dplyr::summarise(mean_pop = mean(perc_methylation, na.rm = TRUE),
                   sum_NA = sum(is.na(perc_methylation)),
            .groups = 'drop')

meth.diff.df.plot.locus2 <- dplyr::group_by(.data = meth.diff.df.plot.locus, locus) %>%
  
  dplyr::filter(  abs(mean_pop[pop == "58-4-1"] - mean_pop[pop == "58-4-2"]) > 25 |
                  abs(mean_pop[pop == "58-4-2"] - mean_pop[pop == "58-5-2"]) > 25 |
                  abs(mean_pop[pop == "58-5-2"] - mean_pop[pop == "88-1"  ]) > 25 |
                  abs(mean_pop[pop == "88-1"  ] - mean_pop[pop == "58-4-1"]) > 25 |
                  abs(mean_pop[pop == "88-1"  ] - mean_pop[pop == "58-4-2"]) > 25 |
                  abs(mean_pop[pop == "58-4-1"] - mean_pop[pop == "58-5-2"]) > 25)
  
  # dplyr::filter(  mean_pop[pop == "58-5-2"] > mean_pop[pop == "88-1"] &
  #                 mean_pop[pop == "58-5-2"] > mean_pop[pop == "58-4-2"] &
  #                 mean_pop[pop == "58-5-2"] > mean_pop[pop == "58-4-1"]) %>%
  # dplyr::filter(  mean_pop[pop == "58-5-2"] - mean_pop[pop == "88-1"] > 30 |
  #                 mean_pop[pop == "58-5-2"] - mean_pop[pop == "58-4-2"] > 30 |
  #                 mean_pop[pop == "58-5-2"] - mean_pop[pop == "58-4-1"] > 30) %>%
  # dplyr::filter(  abs(mean_pop[pop == "88-1"] - mean_pop[pop == "58-4-2"]) < 15 &
  #                 abs(mean_pop[pop == "88-1"] - mean_pop[pop == "58-4-1"]) < 15 &
  #                 abs(mean_pop[pop == "58-4-2"] - mean_pop[pop == "58-4-1"]) < 15)
  # dplyr::filter(  abs(mean_pop[pop == "88-1"] - mean_pop[pop == "58-4-2"]) > 10 &
  #                 abs(mean_pop[pop == "88-1"] - mean_pop[pop == "58-4-1"]) > 10 &
  #                 abs(mean_pop[pop == "58-4-2"] - mean_pop[pop == "58-4-1"]) > 10)

  # dplyr::filter(  mean_pop[pop == "88-1"] > mean_pop[pop == "58-5-2"] &
  #                 mean_pop[pop == "88-1"] > mean_pop[pop == "58-4-2"] &
  #                 mean_pop[pop == "88-1"] > mean_pop[pop == "58-4-1"]) %>%
  # dplyr::filter(  mean_pop[pop == "88-1"] - mean_pop[pop == "58-5-2"] > 25 |
  #                 mean_pop[pop == "88-1"] - mean_pop[pop == "58-4-2"] > 25 |
  #                 mean_pop[pop == "88-1"] - mean_pop[pop == "58-4-1"] > 25) %>%
  # dplyr::filter(  abs(mean_pop[pop == "58-5-2"] - mean_pop[pop == "58-4-2"]) < 15 &
  #                 abs(mean_pop[pop == "58-5-2"] - mean_pop[pop == "58-4-1"]) < 15 &
  #                 abs(mean_pop[pop == "58-4-2"] - mean_pop[pop == "58-4-1"]) < 15)
  # dplyr::filter(  abs(mean_pop[pop == "58-5-2"] - mean_pop[pop == "58-4-2"]) > 10 &
  #                 abs(mean_pop[pop == "58-5-2"] - mean_pop[pop == "58-4-1"]) > 10 &
  #                 abs(mean_pop[pop == "58-4-2"] - mean_pop[pop == "58-4-1"]) > 10)

  # dplyr::filter(  mean_pop[pop == "58-4-1"] > mean_pop[pop == "58-5-2"] &
  #                 mean_pop[pop == "58-4-1"] > mean_pop[pop == "58-4-2"] &
  #                 mean_pop[pop == "58-4-1"] > mean_pop[pop == "88-1"]) %>%
  # dplyr::filter(  mean_pop[pop == "58-4-1"] - mean_pop[pop == "58-5-2"] > 50 |
  #                 mean_pop[pop == "58-4-1"] - mean_pop[pop == "58-4-2"] > 50 |
  #                 mean_pop[pop == "58-4-1"] - mean_pop[pop == "88-1"] > 50) %>%
  # dplyr::filter(  abs(mean_pop[pop == "58-5-2"] - mean_pop[pop == "58-4-2"]) < 15 &
  #                 abs(mean_pop[pop == "58-4-2"] - mean_pop[pop == "88-1"]) < 15 &
  #                 abs(mean_pop[pop == "58-5-2"] - mean_pop[pop == "88-1"]) < 15)
  # dplyr::filter(  abs(mean_pop[pop == "58-5-2"] - mean_pop[pop == "58-4-2"]) > 10 &
  #                 abs(mean_pop[pop == "58-4-2"] - mean_pop[pop == "88-1"]) > 10 &
  #                 abs(mean_pop[pop == "58-5-2"] - mean_pop[pop == "88-1"]) > 10)

  # dplyr::filter(  mean_pop[pop == "58-4-2"] > mean_pop[pop == "58-5-2"] &
  #                 mean_pop[pop == "58-4-2"] > mean_pop[pop == "58-4-1"] &
  #                 mean_pop[pop == "58-4-2"] > mean_pop[pop == "88-1"]) %>%
  # dplyr::filter(  mean_pop[pop == "58-4-2"] - mean_pop[pop == "58-5-2"] > 50 |
  #                 mean_pop[pop == "58-4-2"] - mean_pop[pop == "58-4-1"] > 50 |
  #                 mean_pop[pop == "58-4-2"] - mean_pop[pop == "88-1"] > 50) %>%
  # dplyr::filter(  abs(mean_pop[pop == "58-5-2"] - mean_pop[pop == "58-4-1"]) < 15 &
  #                 abs(mean_pop[pop == "58-4-1"] - mean_pop[pop == "88-1"]) < 15 &
  #                 abs(mean_pop[pop == "58-5-2"] - mean_pop[pop == "88-1"]) < 15)
  # dplyr::filter(  abs(mean_pop[pop == "58-5-2"] - mean_pop[pop == "58-4-1"]) > 10 &
  #                 abs(mean_pop[pop == "58-4-1"] - mean_pop[pop == "88-1"]) > 10 &
  #                 abs(mean_pop[pop == "58-5-2"] - mean_pop[pop == "88-1"]) > 10)



length(unique(meth.diff.df.plot.locus2$locus))
#2719 (all25) #761 (all30) #1 (all50)
#1295 (HEARD25) #401 (HEARD30) #0 (HEARD50) #13 (HEARD45)
#127 (ROSS25) #48 (ROSS30) #0 (ROSS50)
#586 (EANT4.1-25) #143 (EANT4.1-30) #1 (EANT4.1-50)
#569 (EANT4.2-25) #139 (EANT4.2-30) #0 (EANT4.2-50)

##PEAK <15
#432 (HEARD25) #94 (HEARD30) #0 (HEARD50)
#49 (ROSS25) #14 (ROSS30) #0 (ROSS50)
#237 (EANT4.1-25) #45 (EANT4.1-30) #0 (EANT4.1-50)
#92 (EANT4.2-25) #12 (EANT4.2-30) #0 (EANT4.2-50)

##Gradient >10
#62 (HEARD25) #35 (HEARD30) #0 (HEARD50)
#21 (ROSS25) #7 (ROSS30) #0 (ROSS50)
#32 (EANT4.1-25) #15 (EANT4.1-30) #0 (EANT4.1-50)
#42 (EANT4.2-25) #22 (EANT4.2-30) #0 (EANT4.2-50)


HYPER.plot <- ggplot2::ggplot() +
  # ggplot2::geom_point(data = meth.diff.df.plot2,
  #                     ggplot2::aes(x = pop, y = perc_methylation, colour = locus)) +
  ggplot2::geom_line(data = meth.diff.df.plot.locus2,
                     ggplot2::aes(x = pop, y = mean_pop, 
                                  colour = locus, group = locus
                                  )) +
  ggplot2::geom_point(data = meth.diff.df.plot.locus2,
                     ggplot2::aes(x = pop, y = mean_pop, 
                                  colour = locus, group = locus,
                                  # size = sqrt(sum_NA)
                                  size = sum_NA
                                  )
                     # ,size = 5
                     ) +
  ggplot2::ylim(c(0,100)) +
  # ggplot2::facet_wrap(~chr) +
  ggplot2::theme(legend.position = "none")
  # ggplot2::guides(colour = ggplot2::guide_legend(ncol = 1))

print(HYPER.plot)

ggplot2::ggsave(plot = HYPER.plot, 
                filename = paste0(figure.path,
                                  "TOA_allcpg_methDiff_25-a.png"
                                  # "TOA_methDiff_all_increase_25-c.png"
                                  # "TOA_methDiff_HEARD_increase_25-e.png"
                                  # "TOA_methDiff_ROSS_increase_25-f.png"
                                  # "TOA_methDiff_EANT1_increase_30-f.png"
                                  # "TOA_methDiff_EANT2_increase_30-e.png"
                                  ),
                width = 60, height = 45, units = "cm", device = "png")




meth.diff.df.plot.locus2 <- dplyr::group_by(.data = meth.diff.df.plot.locus, locus) %>%
    dplyr::filter(mean_pop[pop == "TAS-warm"] < 50 & mean_pop[pop == "TAS-cold"] > 50) %>%
  dplyr::filter(mean_pop[pop == "TAS-cold"] > mean_pop[pop == "TAS-control"] &
                  mean_pop[pop == "TAS-control"] > mean_pop[pop == "TAS-warm"]) %>%
  dplyr::filter(mean_pop[pop == "TAS-cold"] - mean_pop[pop == "TAS-warm"] > 60)
  # dplyr::filter(mean_pop[pop == "TAS-cold"] - mean_pop[pop == "TAS-warm"] > 50 &
  #                 mean_pop[pop == "TAS-cold"] - mean_pop[pop == "TAS-control"] < 5)
  # dplyr::filter(mean_pop[pop == "TAS-cold"] - mean_pop[pop == "TAS-warm"] > 50 &
  #                 mean_pop[pop == "TAS-control"] - mean_pop[pop == "TAS-warm"] < 15)
  dplyr::filter(mean_pop[pop == "TAS-cold"] - mean_pop[pop == "TAS-warm"] > 50 &
                  mean_pop[pop == "TAS-cold"] - mean_pop[pop == "TAS-control"] > 25 &
                  mean_pop[pop == "TAS-control"] - mean_pop[pop == "TAS-warm"] > 25)
length(unique(meth.diff.df.plot.locus2$locus))#873 #51

HYPO.plot <- ggplot2::ggplot() +
  # ggplot2::geom_point(data = meth.diff.df.plot2,
  #                     ggplot2::aes(x = pop, y = perc_methylation, colour = locus)) +
  ggplot2::geom_line(data = meth.diff.df.plot.locus2,
                     ggplot2::aes(x = pop, y = mean_pop, colour = locus, group = locus)) +
  ggplot2::geom_point(data = meth.diff.df.plot.locus2,
                     ggplot2::aes(x = pop, y = mean_pop, colour = locus, group = locus),
                     size = 5) +
  # ggplot2::facet_wrap(~chr) +
    # ggplot2::theme(legend.position = "none")
  ggplot2::guides(colour = ggplot2::guide_legend(ncol = 2))

print(HYPO.plot)

ggplot2::ggsave(plot = HYPO.plot, 
                filename = paste0(figure.path,"SS_HYPERcpg_methDiff_TAS_exp_muscle_decrease_50-a.png"),
                width = 30, height = 15, units = "cm", device = "png")
```

### HEARD heatmap

```{r}
meth.diff.df.plot.locus2 <- dplyr::group_by(.data = meth.diff.df.plot.locus, locus) %>%
    dplyr::filter(mean_pop[pop == "TAS-warm"] > 50 & mean_pop[pop == "TAS-cold"] < 50) %>%
    dplyr::filter(mean_pop[pop == "TAS-cold"] < mean_pop[pop == "TAS-control"] &
                  mean_pop[pop == "TAS-control"] < mean_pop[pop == "TAS-warm"]) %>%
  dplyr::filter(mean_pop[pop == "TAS-warm"] - mean_pop[pop == "TAS-cold"] > 50)
    dplyr::filter(mean_pop[pop == "TAS-warm"] - mean_pop[pop == "TAS-cold"] > 50 &
                  mean_pop[pop == "TAS-warm"] - mean_pop[pop == "TAS-control"] > 25 &
                  mean_pop[pop == "TAS-control"] - mean_pop[pop == "TAS-cold"] > 25)
# 
#   dplyr::filter(mean_pop[pop == "TAS-cold"] < mean_pop[pop == "TAS-control"] &
#                   mean_pop[pop == "TAS-control"] < mean_pop[pop == "TAS-warm"])
length(unique(meth.diff.df.plot.locus2$locus))#1424 #144 $43

hyper.TAS.warm.gr <- as(data.frame(chr = stringr::str_split_fixed(
                                        string = as.character(unique(meth.diff.df.plot.locus2$locus)), 
                                        pattern =  "_", n = 2)[,1],
                                    start = 
                                      as.integer(stringr::str_split_fixed(
                                        string = as.character(unique(meth.diff.df.plot.locus2$locus)), 
                                        pattern =  "_", n = 2)[,2]),
                                    end = 
                                      as.integer(stringr::str_split_fixed(
                                        string = as.character(unique(meth.diff.df.plot.locus2$locus)), 
                                        pattern =  "_", n = 2)[,2])
                                    ), "GRanges")

meth.diff <- methylKit::selectByOverlap(meth.2018M.norm11L, hyper.TAS.warm.gr) #1424 

meth.diff.df <- as.data.frame(t(methylKit::percMethylation(meth.diff)))
loc.names <- paste0(meth.diff$chr,"_",meth.diff$start)
colnames(meth.diff.df) <- loc.names

meth.diff.df$individual <- rownames(meth.diff.df) 
identical(metadata.new$individual, meth.diff.df$individual)
meth.diff.df$pop <- metadata.new$strata



heat.plot.data <- meth.diff.df %>%
  # Data wrangling
  tibble::as_tibble() %>%
  # tibble::rowid_to_column(var="X") %>%
  # tibble::rownames_to_column(var="individuals") %>%
  # dplyr::mutate(individual = row.names(meth.diff.df),
  #               pop = metadata.new$strata) %>%
  tidyr::gather(key = "CpG", value = "Methylation_percentage", -individual, -pop,
                factor_key = FALSE) %>%
  # tidyr::gather(key="CpG", value = "Methylation_percentage", -individual) %>%
  dplyr::arrange(pop, individual, CpG) %>%
  dplyr::mutate(individual = factor(individual, levels = unique(individual))) %>%
  dplyr::mutate(CpG = factor(CpG, levels = unique(CpG)))


  # Viz
heat.plot <- ggplot2::ggplot(data = heat.plot.data) + 
    ggplot2::geom_tile(ggplot2::aes(individual, y = CpG, fill = Methylation_percentage)) +
    ggplot2::scale_fill_gradient2(low = "blue", mid = "yellow", high = "red", midpoint = 50) +
    ggnewscale::new_scale_fill() +
    ggplot2::geom_tile(ggplot2::aes(individual, y = "pop", fill = pop, height = 2)) +
    ggplot2::scale_fill_manual(values = c("#4B98C9", "#e1eaf2","#fc9403")) +
    ggplot2::theme(legend.position = "top",
          axis.text.x = ggplot2::element_text(face = "bold", color = "black", 
                                        size = 10, angle = 90, hjust = 1, vjust = 0))
print(heat.plot)
# ggplot2::ggsave(heat.plot, bg = "white",
#                 filename = paste0(figure.path,"SS_Heatmap_CpG_TAS_exp_muscle_DMS_increase_60.png"), 
#                 width = 45, height = 30, units = "cm")

ggplot2::ggsave(heat.plot, bg = "white",
                filename = paste0(figure.path,"SS_Heatmap_CpG_TAS_exp_muscle_DMS_methylkit.png"), 
                width = 45, height = 30, units = "cm")
```

### ROSS heatmap

```{r}
meth.diff.df.plot.locus2 <- dplyr::group_by(.data = meth.diff.df.plot.locus, locus) %>%
    dplyr::filter(mean_pop[pop == "TAS-warm"] < 50 & mean_pop[pop == "TAS-cold"] > 50) %>%
  dplyr::filter(mean_pop[pop == "TAS-cold"] > mean_pop[pop == "TAS-control"] &
                  mean_pop[pop == "TAS-control"] > mean_pop[pop == "TAS-warm"]) %>%
  dplyr::filter(mean_pop[pop == "TAS-cold"] - mean_pop[pop == "TAS-warm"] > 50)
  # dplyr::filter(mean_pop[pop == "TAS-cold"] - mean_pop[pop == "TAS-warm"] > 50 &
  #                 mean_pop[pop == "TAS-cold"] - mean_pop[pop == "TAS-control"] < 5)
  # dplyr::filter(mean_pop[pop == "TAS-cold"] - mean_pop[pop == "TAS-warm"] > 50 &
  #                 mean_pop[pop == "TAS-control"] - mean_pop[pop == "TAS-warm"] < 15)
  # dplyr::filter(mean_pop[pop == "TAS-cold"] - mean_pop[pop == "TAS-warm"] > 50 &
  #                 mean_pop[pop == "TAS-cold"] - mean_pop[pop == "TAS-control"] > 5 |
  #                 mean_pop[pop == "TAS-control"] - mean_pop[pop == "TAS-warm"] > 5)
# dplyr::filter(mean_pop[pop == "TAS-cold"] - mean_pop[pop == "TAS-warm"] > 50 &
#                   mean_pop[pop == "TAS-cold"] - mean_pop[pop == "TAS-control"] > 25 &
#                   mean_pop[pop == "TAS-control"] - mean_pop[pop == "TAS-warm"] > 25)

length(unique(meth.diff.df.plot.locus2$locus))#873 #661 #293 #146 #51 #15 #7

hyper.TAS.cold.gr <- as(data.frame(chr = stringr::str_split_fixed(
                                        string = as.character(unique(meth.diff.df.plot.locus2$locus)), 
                                        pattern =  "_", n = 2)[,1],
                                    start = 
                                      as.integer(stringr::str_split_fixed(
                                        string = as.character(unique(meth.diff.df.plot.locus2$locus)), 
                                        pattern =  "_", n = 2)[,2]),
                                    end = 
                                      as.integer(stringr::str_split_fixed(
                                        string = as.character(unique(meth.diff.df.plot.locus2$locus)), 
                                        pattern =  "_", n = 2)[,2])
                                    ), "GRanges")
meth.diff <- methylKit::selectByOverlap(meth.2018M.norm11L, hyper.TAS.cold.gr) 

meth.diff.df <- as.data.frame(t(methylKit::percMethylation(meth.diff)))
loc.names <- paste0(meth.diff$chr,"_",meth.diff$start)
colnames(meth.diff.df) <- loc.names

meth.diff.df$individual <- rownames(meth.diff.df) 
identical(metadata.new$individual, meth.diff.df$individual)
meth.diff.df$pop <- metadata.new$strata

heat.plot.data <- meth.diff.df %>%
  # Data wrangling
  tibble::as_tibble() %>%
  # tibble::rowid_to_column(var="X") %>%
  # tibble::rownames_to_column(var="individuals") %>%
  # dplyr::mutate(individual = row.names(meth.diff.df),
  #               pop = metadata.new$strata) %>%
  tidyr::gather(key = "CpG", value = "Methylation_percentage", -individual, -pop) %>%
  # tidyr::gather(key="CpG", value = "Methylation_percentage", -individual) %>%
  dplyr::arrange(pop, individual, CpG) %>%
  dplyr::mutate(individual = factor(individual, levels = unique(individual))) %>%
  dplyr::mutate(CpG = factor(CpG, levels = unique(CpG)))

  # Viz
heat.plot <- ggplot2::ggplot(data = heat.plot.data,
                             ggplot2::aes(individual, CpG, fill = Methylation_percentage)) + 
    ggplot2::geom_tile() +
    # ggplot2::scale_fill_gradient2(low = "red", mid = "yellow", high = "blue", midpoint = 50) +
    ggplot2::scale_fill_gradient2(low = "blue", mid = "yellow", high = "red", midpoint = 50) +
    ggnewscale::new_scale_fill() +
    ggplot2::geom_tile(ggplot2::aes(individual, y = "pop", fill = pop, height = 2)) +
    ggplot2::scale_fill_manual(values = c("#4B98C9", "#e1eaf2","#fc9403")) +
    ggplot2::theme(legend.position = "top",
          axis.text.x = ggplot2::element_text(face = "bold", color = "black", 
                                        size = 10, angle = 90, hjust = 1, vjust = 0))
print(heat.plot)
ggplot2::ggsave(heat.plot, bg = "white",
                filename = paste0(figure.path,"SS_Heatmap_CpG_TAS_exp_muscle_DMS_decrease_50.png"), 
                width = 45, height = 30, units = "cm")
```

### Save highly differentiating CpG

```{r}
meth.diff.df.plot.locus45 <- dplyr::group_by(.data = meth.diff.df.plot.locus, locus) %>%
  
  dplyr::filter(  abs(mean_pop[pop == "58-4-1"] - mean_pop[pop == "58-4-2"]) > 45 |
                  abs(mean_pop[pop == "58-4-2"] - mean_pop[pop == "58-5-2"]) > 45 |
                  abs(mean_pop[pop == "58-5-2"] - mean_pop[pop == "88-1"  ]) > 45 |
                  abs(mean_pop[pop == "88-1"  ] - mean_pop[pop == "58-4-1"]) > 45 |
                  abs(mean_pop[pop == "88-1"  ] - mean_pop[pop == "58-4-2"]) > 45 |
                  abs(mean_pop[pop == "58-4-1"] - mean_pop[pop == "58-5-2"]) > 45)

length(unique(meth.diff.df.plot.locus25$locus))#2719
length(unique(meth.diff.df.plot.locus30$locus))#761
length(unique(meth.diff.df.plot.locus45$locus))#20

# save(meth.diff.df.plot.locus25, meth.diff.df.plot.locus30, meth.diff.df.plot.locus45, 
#      file = paste0(rdata.path, "Toothfish_PERCENTAGE_DMS_CpGsites.Rdata"))
print(load(paste0(rdata.path, "Toothfish_PERCENTAGE_DMS_CpGsites.Rdata")))


sum(unique(DMS$DMS) %in% unique(meth.diff.df.plot.locus25$locus))# 631
sum(unique(DMS$DMS) %in% unique(meth.diff.df.plot.locus30$locus))# 288
sum(unique(DMS$DMS) %in% unique(meth.diff.df.plot.locus45$locus))# 10


sort(table(DMS.df$chr[!duplicated(DMS.df$locus)]), decreasing = TRUE)

print(load(paste0(rdata.path,"Toothfish_DMS_CpGsites.Rdata")))
```

#### DMS PCA

```{r}
print(load(paste0(rdata.path, "Toothfish_PERCENTAGE_DMS_CpGsites.Rdata")))
print(load(paste0(rdata.path,"Toothfish_methylation_objects_All_pop_no_outliers.Rdata")))

meth.diff.df.plot.locus3 <- meth.diff.df.plot.locus25
meth.diff.df.plot.locus3 <- meth.diff.df.plot.locus25[
  meth.diff.df.plot.locus25$locus %in% unique(DMS$DMS),]

meth.diff.df.plot.locus3 <- meth.diff.df.plot.locus30
meth.diff.df.plot.locus3 <- meth.diff.df.plot.locus45

TOA_DMS.gr <- as(data.frame(chr = stringr::str_split_fixed(
                                        string = as.character(unique(meth.diff.df.plot.locus3$locus)), 
                                        pattern =  "_", n = 2)[,1],
                                    start = 
                                      as.integer(stringr::str_split_fixed(
                                        string = as.character(unique(meth.diff.df.plot.locus3$locus)), 
                                        pattern =  "_", n = 2)[,2]),
                                    end = 
                                      as.integer(stringr::str_split_fixed(
                                        string = as.character(unique(meth.diff.df.plot.locus3$locus)), 
                                        pattern =  "_", n = 2)[,2])
                                    ), "GRanges")

meth.diff <- methylKit::selectByOverlap(meth2, TOA_DMS.gr) #2719 #761 #20 #631 


PCobj <- methylKit::PCASamples(meth.diff, comp = c(1,2), adj.lim = c(0,0), obj.return = TRUE)
var_frac <- PCobj$sdev/sum(PCobj$sdev)
print(paste0("Total variance explained: ",signif(sum(var_frac[1:3]) * 100, 3), "%")) 
barplot(var_frac[1:20], main = "variance explained", 
        col = heat.colors(length(var_frac[1:20])))

# metadata3 <- metadata2[metadata2$tissue %in% "M",]
# rm <- c("TOA_58-4-1_277", "TOA_58-4-1_289", "TOA_58-4-1_290", "TOA_58-4-1_295")
# rm2 <- c("TOA_58-4-1_118", "TOA_58-4-1_121", "TOA_58-4-1_125")
# metadata3 <- metadata2[!metadata2$SAMPLE_ID %in% c(rm, rm2),]
table(metadata3$region)
metadata.new <- metadata3[metadata3$SAMPLE_ID %in% meth2@sample.ids,]
metadata.new <- metadata.new[order(match(metadata.new$SAMPLE_ID, meth2@sample.ids)),]
identical(metadata.new$SAMPLE_ID, meth2@sample.ids)
identical(metadata.new$SAMPLE_ID, rownames(PCobj$x))

pop <- metadata.new$region


colours5 <- adegenet::funky(4)
# plot.title <- "School_shark methylation (2,719 DMS)"
plot.title <- "Toothfish methylation (631 DMS in common with pairwise)"

# plot.title <- "Toothfish methylation (761 DMS)"
# plot.title <- "Toothfish methylation (20 DMS)"


data <- data.frame(ind = rownames(PCobj$x),PCobj$x,  pop = pop)
data <- merge(data, aggregate(cbind(mean.x = PC1, mean.y = PC2) ~
                                pop, data, mean), by = "pop")

pca.plot <- ggplot2::ggplot(data, ggplot2::aes(PC1, PC2, color = pop)) +
  ggplot2::geom_point(size = 4) + 
  ggplot2::geom_point(ggplot2::aes(x = mean.x,
                                   y = mean.y, color = pop),
                      size = 7,
                      shape = 17) +
  ggplot2::scale_colour_manual(values = colours5) +
  ggplot2::scale_fill_manual(values = colours5) + 
  ggplot2::geom_hline(yintercept = 0) +
  ggplot2::geom_vline(xintercept = 0) + 
  ggplot2::labs(
    subtitle = paste0("Total variance explained: ",
                      signif(sum(var_frac[1:4]) * 100, 3), "%"),
    y = paste0("PC2; ", signif(var_frac[2] * 100, digits = 2), " %"),
    x = paste0("PC1; ", signif(var_frac[1] * 100, digits = 2), " %"),
    title = plot.title,
    caption = "") + 
  ggplot2::theme_minimal() +
  ggplot2::theme(
    axis.text = ggplot2::element_text(size = 15),
    axis.title.x = ggplot2::element_text(size = 20),
    axis.title.y = ggplot2::element_text(size = 20),
    legend.text = ggplot2::element_text(size = 15)
  )

print(pca.plot)
ggplot2::ggsave(pca.plot, bg = "white",
                filename = paste0(figure.path,
                                  # "TOA_PCA_CpG_DMS25_regions_PC12.png"
                                  "TOA_PCA_CpG_DMS25_commonPW_regions_PC12.png"
                                  # "TOA_PCA_CpG_DMS30_regions_PC12.png"
                                  # "TOA_PCA_CpG_DMS45_regions_PC12.png"
                                  ), 
                width = 30, height = 15, units = "cm")




data <- data.frame(ind = rownames(PCobj$x),PCobj$x,  pop = pop)
data <- merge(data, aggregate(cbind(mean.x = PC1, mean.y = PC3) ~
                                pop, data, mean), by = "pop")

pca.plot <- ggplot2::ggplot(data, ggplot2::aes(PC1, PC3, color = pop)) +
  ggplot2::geom_point(size = 4) + 
  ggplot2::geom_point(ggplot2::aes(x = mean.x,
                                   y = mean.y, color = pop),
                      size = 7,
                      shape = 17) +
  ggplot2::scale_colour_manual(values = colours5) +
  ggplot2::scale_fill_manual(values = colours5) + 
  ggplot2::geom_hline(yintercept = 0) +
  ggplot2::geom_vline(xintercept = 0) + 
  ggplot2::labs(
    subtitle = paste0("Total variance explained: ",
                      signif(sum(var_frac[1:4]) * 100, 3), "%"),
    y = paste0("PC3; ", signif(var_frac[3] * 100, digits = 2), " %"),
    x = paste0("PC1; ", signif(var_frac[1] * 100, digits = 2), " %"),
    title = plot.title,
    caption = "") + 
  ggplot2::theme_minimal() +
  ggplot2::theme(
    axis.text = ggplot2::element_text(size = 15),
    axis.title.x = ggplot2::element_text(size = 20),
    axis.title.y = ggplot2::element_text(size = 20),
    legend.text = ggplot2::element_text(size = 15)
  )

print(pca.plot)
ggplot2::ggsave(pca.plot, bg = "white",
                filename = paste0(figure.path,
                                  # "TOA_PCA_CpG_DMS25_regions_PC13.png"
                                  "TOA_PCA_CpG_DMS25_commonPW_regions_PC13.png"
                                  # "TOA_PCA_CpG_DMS30_regions_PC13.png"
                                  # "TOA_PCA_CpG_DMS45_regions_PC13.png"
                                  ), 
                width = 30, height = 15, units = "cm")



data <- data.frame(ind = rownames(PCobj$x),PCobj$x,  pop = pop)
data <- merge(data, aggregate(cbind(mean.x = PC2, mean.y = PC3) ~
                                pop, data, mean), by = "pop")

pca.plot <- ggplot2::ggplot(data, ggplot2::aes(PC2, PC3, color = pop)) +
  ggplot2::geom_point(size = 4) + 
  ggplot2::geom_point(ggplot2::aes(x = mean.x,
                                   y = mean.y, color = pop),
                      size = 7,
                      shape = 17) +
  ggplot2::scale_colour_manual(values = colours5) +
  ggplot2::scale_fill_manual(values = colours5) + 
  ggplot2::geom_hline(yintercept = 0) +
  ggplot2::geom_vline(xintercept = 0) + 
  ggplot2::labs(
    subtitle = paste0("Total variance explained: ",
                      signif(sum(var_frac[1:4]) * 100, 3), "%"),
    y = paste0("PC3; ", signif(var_frac[3] * 100, digits = 2), " %"),
    x = paste0("PC2; ", signif(var_frac[2] * 100, digits = 2), " %"),
    title = plot.title,
    caption = "") + 
  ggplot2::theme_minimal() +
  ggplot2::theme(
    axis.text = ggplot2::element_text(size = 15),
    axis.title.x = ggplot2::element_text(size = 20),
    axis.title.y = ggplot2::element_text(size = 20),
    legend.text = ggplot2::element_text(size = 15)
  )

print(pca.plot)
ggplot2::ggsave(pca.plot, bg = "white",
                filename = paste0(figure.path,
                                  # "TOA_PCA_CpG_DMS25_regions_PC23.png"
                                  "TOA_PCA_CpG_DMS25_commonPW_regions_PC23.png"
                                  # "TOA_PCA_CpG_DMS30_regions_PC23.png"
                                  # "TOA_PCA_CpG_DMS45_regions_PC23.png"
                                  ), 
                width = 30, height = 15, units = "cm")
```

#### DMA heatmap

```{r}

meth.diff.df <- as.data.frame(t(methylKit::percMethylation(meth.diff)))
loc.names <- paste0(meth.diff$chr,"_",meth.diff$start)
colnames(meth.diff.df) <- loc.names
rm <- c("TOA_58-4-1_277", "TOA_58-4-1_289", "TOA_58-4-1_290", "TOA_58-4-1_295")
rm2 <- c("TOA_58-4-1_118", "TOA_58-4-1_121", "TOA_58-4-1_125")

metadata3 <- metadata2[!metadata2$SAMPLE_ID %in% c(rm, rm2),]
table(metadata3$region)

meth.diff.df$individual <- rownames(meth.diff.df) 
identical(metadata3$individual, meth.diff.df$individual)
meth.diff.df$pop <- strata.random

```


## DMS - pariwise
### HEARD vs ROSS
#### Prepare data

```{r}
print(load(paste0(rdata.path,"Toothfish_metadata.Rdata")))
metadata3 <- metadata2[metadata2$region %in% c("58-5-2", "88-1"),]
table(metadata3$region)
print(load(paste0(rdata.path,"Toothfish_methylation_objects_raw.Rdata")))


filtered.myobj3 <- methylKit::reorganize(
  methylObj = filtered.myobj,
  sample.ids = metadata3$SAMPLE_ID,
  treatment = as.numeric(factor(metadata3$region)),
  chunk.size = 1e+06,
  save.db = TRUE,
  suffix = "subset_HEARD_ROSS_raw"
)

# filtered.myobj3@treatment <- as.numeric(factor(metadata3$region))
filtered.myobj3 <- methylKit::filterByCoverage(filtered.myobj3,lo.count = 10,lo.perc = NULL,
                                      hi.count = NULL, hi.perc = 99)

meth.HR <- methylKit::unite(filtered.myobj3, destrand = FALSE,
                             save.db = TRUE,
                             suffix = "subset_HEARD_ROSS",
                             min.per.group = 20L, #missing data allowed
                             mc.cores = 25)
#1,318,203 CpGs before coverage filter
#1,297,166 CpGs after coverage filter

# save(meth.HR, 
#      file = paste0(rdata.path,
#                    "Toothfish_subset_pairwise_methylation_objects_raw.Rdata"))
```

#### diffMethPerChr


```{r}
print(load(paste0(rdata.path,"Toothfish_subset_pairwise_methylation_objects_raw.Rdata")))
metadata3 <- metadata2[metadata2$region %in% c("58-5-2", "88-1"),]
covariates <- dplyr::select(metadata3, Sex, mat, Age)
myDiff <- methylKit::calculateDiffMeth(meth.HR,
                                       covariates = covariates,
                                       adjust = "fdr",
                                       save.db = TRUE,
                                       chunk.size = 1e+06,
                                       mc.cores = 1)
meth.diff.chr <- methylKit::diffMethPerChr(myDiff,plot = TRUE, 
                                           qvalue.cutoff = 0.01, meth.cutoff = 25)
myDiff25p.hyper <- methylKit::getMethylDiff(myDiff, difference = 25,
                                            qvalue = 0.01, 
                                            type = "hyper") #1,478 # 99
myDiff25p.hypo <- methylKit::getMethylDiff(myDiff, difference = 25,
                                           qvalue = 0.01, 
                                           type = "hypo") #1,227 #35
myDiff25p <- methylKit::getMethylDiff(myDiff, difference = 25, 
                                      qvalue = 0.01) #2,705 #134
# save(myDiff,myDiff25p, myDiff25p.hyper, myDiff25p.hypo, 
#      file = paste0(rdata.path,
#                    "Toothfish_DMS_Region_subset_HEARD_ROSS.Rdata"))

myDiff.df <- as.data.frame(myDiff)
myDiff25p.df <- as.data.frame(as(myDiff25p, "GRanges"))
volcano.plot <- ggplot2::ggplot() +
  ggplot2::geom_point(data = myDiff.df, ggplot2::aes(x = meth.diff, 
                                                  y = -log10(pvalue))) +
  ggplot2::geom_point(data = myDiff25p.df, ggplot2::aes(x = meth.diff, 
                                                  y = -log10(pvalue)),
                      colour = "red")
print(volcano.plot)

ggplot2::ggsave(volcano.plot, bg = "white",
                filename = paste0(figure.path,
                                  "TOA_ALL_DMS25_VOLCANO_methylkit_regions_subset_HEARD_ROSS.png"
                                  ), 
                width = 30, height = 15, units = "cm")
```

#### Load data

```{r}
print(load(paste0(rdata.path,"Toothfish_DMS_Region_subset_HEARD_ROSS.Rdata")))

myDiff25p.gr <- as(myDiff25p, "GRanges")
meth.diff <- methylKit::selectByOverlap(meth.HR, myDiff25p.gr)

# DMS <- data.frame(DMS = NULL, region_pw = NULL)
DMS.HR <- data.frame(DMS = paste0(meth.diff$chr,"_", meth.diff$start),
                  region_pw = "HEARD_ROSS")
DMS <- rbind(DMS, DMS.HR)

# save(DMS, DMS.HR,
#      file = paste0(rdata.path,"Toothfish_DMS_CpGsites.Rdata"))
```


#### PCA - ALL

```{r}
PCobj <- methylKit::PCASamples(meth.diff, comp = c(1,2), adj.lim = c(0,0), obj.return = TRUE)


var_frac <- PCobj$sdev/sum(PCobj$sdev)
print(paste0("Total variance explained: ",signif(sum(var_frac[1:3]) * 100, 3), "%")) 
barplot(var_frac[1:20], main = "variance explained", 
        col = heat.colors(length(var_frac[1:20])))

# metadata3 <- metadata2[metadata2$tissue %in% "M",]
metadata.new <- metadata2[metadata2$SAMPLE_ID %in% meth.diff@sample.ids,]
metadata.new <- metadata.new[order(match(metadata.new$SAMPLE_ID, meth.diff@sample.ids)),]
identical(metadata.new$SAMPLE_ID, meth.diff@sample.ids)
identical(metadata.new$SAMPLE_ID, rownames(PCobj$x))

pop <- metadata.new$region

# pop <- metadata.new$`Date caught`

# colours5 <- adegenet::funky(4)
# colours5 <- c("#4d6b40","#adcf9f", "#c5d6e6", "#08306B")
colours5 <- c("#c5d6e6", "#08306B")

# plot.title <- "Toothfish methylation (1,316,799 CpG sites)"
plot.title <- "Toothfish methylation - Heard vs Ross (134 CpG sites)"


data <- data.frame(ind = rownames(PCobj$x),PCobj$x,  pop = pop)
data <- merge(data, aggregate(cbind(mean.x = PC1, mean.y = PC2) ~
                                pop, data, mean), by = "pop")

pca.plot <- ggplot2::ggplot(data, ggplot2::aes(PC1, PC2, color = pop)) +
  ggplot2::geom_point(size = 4) + 
  ggplot2::geom_point(ggplot2::aes(x = mean.x,
                                   y = mean.y, color = pop),
                      size = 7,
                      shape = 17) +
  ggplot2::scale_colour_manual(values = colours5) +
  ggplot2::scale_fill_manual(values = colours5) + 
  ggplot2::geom_hline(yintercept = 0) +
  ggplot2::geom_vline(xintercept = 0) + 
  ggplot2::labs(
    subtitle = paste0("Total variance explained: ",
                      signif(sum(var_frac[1:4]) * 100, 3), "%"),
    y = paste0("PC2; ", signif(var_frac[2] * 100, digits = 2), " %"),
    x = paste0("PC1; ", signif(var_frac[1] * 100, digits = 2), " %"),
    title = plot.title,
    caption = "") + 
  ggplot2::theme_minimal() +
  ggplot2::theme(
    axis.text = ggplot2::element_text(size = 15),
    axis.title.x = ggplot2::element_text(size = 20),
    axis.title.y = ggplot2::element_text(size = 20),
    legend.text = ggplot2::element_text(size = 15)
  )

print(pca.plot)
ggplot2::ggsave(pca.plot, bg = "white",
                filename = paste0(figure.path,"PCA_plot_diffMeth_ALL_Region_subset_HEARD_ROSS_PC12.png"), 
                width = 30, height = 15, units = "cm")



data <- data.frame(ind = rownames(PCobj$x),PCobj$x,  pop = pop)
data <- merge(data, aggregate(cbind(mean.x = PC1, mean.y = PC3) ~
                                pop, data, mean), by = "pop")

pca.plot <- ggplot2::ggplot(data, ggplot2::aes(PC1, PC3, color = pop)) +
  ggplot2::geom_point(size = 4) + 
  ggplot2::geom_point(ggplot2::aes(x = mean.x,
                                   y = mean.y, color = pop),
                      size = 7,
                      shape = 17) +
  ggplot2::scale_colour_manual(values = colours5) +
  ggplot2::scale_fill_manual(values = colours5) + 
  ggplot2::geom_hline(yintercept = 0) +
  ggplot2::geom_vline(xintercept = 0) + 
  ggplot2::labs(
    subtitle = paste0("Total variance explained: ",
                      signif(sum(var_frac[1:4]) * 100, 3), "%"),
    y = paste0("PC3; ", signif(var_frac[3] * 100, digits = 2), " %"),
    x = paste0("PC1; ", signif(var_frac[1] * 100, digits = 2), " %"),
    title = plot.title,
    caption = "") + 
  ggplot2::theme_minimal() +
  ggplot2::theme(
    axis.text = ggplot2::element_text(size = 15),
    axis.title.x = ggplot2::element_text(size = 20),
    axis.title.y = ggplot2::element_text(size = 20),
    legend.text = ggplot2::element_text(size = 15)
  )

print(pca.plot)
ggplot2::ggsave(pca.plot, bg = "white",
                filename = paste0(figure.path,"PCA_plot_diffMeth_ALL_Region_subset_HEARD_ROSS_PC13.png"), 
                width = 30, height = 15, units = "cm")




data <- data.frame(ind = rownames(PCobj$x),PCobj$x,  pop = pop)
data <- merge(data, aggregate(cbind(mean.x = PC2, mean.y = PC3) ~
                                pop, data, mean), by = "pop")

pca.plot <- ggplot2::ggplot(data, ggplot2::aes(PC2, PC3, color = pop)) +
  ggplot2::geom_point(size = 4) + 
  ggplot2::geom_point(ggplot2::aes(x = mean.x,
                                   y = mean.y, color = pop),
                      size = 7,
                      shape = 17) +
  ggplot2::scale_colour_manual(values = colours5) +
  ggplot2::scale_fill_manual(values = colours5) + 
  ggplot2::geom_hline(yintercept = 0) +
  ggplot2::geom_vline(xintercept = 0) + 
  ggplot2::labs(
    subtitle = paste0("Total variance explained: ",
                      signif(sum(var_frac[1:4]) * 100, 3), "%"),
    y = paste0("PC3; ", signif(var_frac[3] * 100, digits = 2), " %"),
    x = paste0("PC2; ", signif(var_frac[2] * 100, digits = 2), " %"),
    title = plot.title,
    caption = "") + 
  ggplot2::theme_minimal() +
  ggplot2::theme(
    axis.text = ggplot2::element_text(size = 15),
    axis.title.x = ggplot2::element_text(size = 20),
    axis.title.y = ggplot2::element_text(size = 20),
    legend.text = ggplot2::element_text(size = 15)
  )

print(pca.plot)
ggplot2::ggsave(pca.plot, bg = "white",
                filename = paste0(figure.path,"PCA_plot_diffMeth_ALL_Region_subset_HEARD_ROSS_PC23.png"), 
                width = 30, height = 15, units = "cm")

```


##### Plot CpG difference

```{r}
myDiff25p.gr <- as(myDiff25p, "GRanges")
hist(myDiff25p.gr$meth.diff)
min(myDiff25p.gr$meth.diff)#-30.58783 %
max(myDiff25p.gr$meth.diff)#47.26297 %

meth.diff <- methylKit::selectByOverlap(meth.HR, myDiff25p.gr)
length(unique(meth.diff$chr)) #25
meth.diff.df <- as.data.frame(t(methylKit::percMethylation(meth.diff)))
loc.names <- paste0(meth.diff$chr,"_",meth.diff$start)
colnames(meth.diff.df) <- loc.names

meth.diff.df$individual <- rownames(meth.diff.df) 
metadata.new <- metadata[metadata$SAMPLE_ID %in% meth.diff@sample.ids,]
metadata.new <- metadata.new[order(match(metadata.new$SAMPLE_ID, meth.diff@sample.ids)),]
identical(metadata.new$SAMPLE_ID, meth.diff@sample.ids)
identical(metadata.new$SAMPLE_ID,meth.diff.df$individual)

meth.diff.df$pop <- metadata.new$region

meth.diff.df.plot <- tidyr::gather(meth.diff.df, key = locus, 
                           value = perc_methylation, -individual, -pop,
                           factor_key = TRUE) %>%
  dplyr::mutate(chr = stringr::str_split_fixed(locus, 
                                               pattern = "_", 
                                               n = 2)[,1])

meth.diff.df.plot.locus <- meth.diff.df.plot %>% dplyr::group_by(chr,locus,pop) %>% 
  dplyr::summarise(mean_pop = mean(perc_methylation, na.rm = TRUE),
            .groups = 'drop')

meth.diff.df.plot.centroid <- meth.diff.df.plot %>% dplyr::group_by(chr,pop) %>% 
  dplyr::summarise(mean_pop = mean(perc_methylation, na.rm = TRUE),
            .groups = 'drop')

meth.diff.df.plot.label <- meth.diff.df.plot %>% dplyr::group_by(chr) %>% 
  dplyr::summarise(nmarkers = length(unique(locus)),
            .groups = 'drop')


ALL.plot <- ggplot2::ggplot() + 
  ggplot2::geom_violin(data = meth.diff.df.plot, 
                       ggplot2::aes(x = chr, y = perc_methylation)) + 
  ggplot2::geom_point(data = meth.diff.df.plot.centroid, 
                      ggplot2::aes(x = chr, y = mean_pop, colour = pop),
                      size = 7, shape = 5) + 
  ggplot2::geom_point(data = meth.diff.df.plot,
                      ggplot2::aes(x = chr, y = perc_methylation, colour = pop),
                      size = 1) +
  ggplot2::geom_text(data = meth.diff.df.plot.label,
                     ggplot2::aes(x = chr, y = 110, label = nmarkers),
                     size = 4) +
  ggplot2::geom_point(data = meth.diff.df.plot.locus, 
                      ggplot2::aes(x = chr, y = mean_pop, colour = pop),
                      size = 3, shape = 15) +
    ggplot2::theme(
        panel.background = ggplot2::element_blank(),
        panel.border = ggplot2::element_blank(),
        panel.grid = ggplot2::element_blank(),
        axis.line = ggplot2::element_line("gray25"),
        text = ggplot2::element_text(size = 12),
        axis.text = ggplot2::element_text(angle = 90,size = 10, colour = "gray25"),
        axis.title = ggplot2::element_text(size = 14, colour = "gray25"),
        legend.title = ggplot2::element_text(size = 14),
        legend.text = ggplot2::element_text(size = 10),
        legend.key = ggplot2::element_blank())

print(ALL.plot)

ggplot2::ggsave(plot = ALL.plot, 
                filename = paste0(figure.path,"TOA_ALLcpg_methDiff_HEARD_ROSS.png"),
                width = 30, height = 15, units = "cm", device = "png")



diff.plot <- ggplot2::ggplot() +
  # ggplot2::geom_point(data = meth.diff.df.plot,
  #                     ggplot2::aes(x = pop, y = perc_methylation, colour = locus)) +
  ggplot2::geom_line(data = meth.diff.df.plot.locus,
                     ggplot2::aes(x = pop, y = mean_pop, colour = locus, group = locus)) +
  ggplot2::geom_point(data = meth.diff.df.plot.locus,
                     ggplot2::aes(x = pop, y = mean_pop, colour = locus, group = locus),
                     size = 5) +
  ggplot2::ylim(c(0,100)) +
  ggplot2::facet_wrap(~chr) +
  ggplot2::theme(legend.position = "none")
  # ggplot2::guides(colour = ggplot2::guide_legend(ncol = 1))
  # ggplot2::guides(colour = ggplot2::guide_legend(ncol = 5))

print(diff.plot)

ggplot2::ggsave(plot = diff.plot, 
                filename = paste0(figure.path,
                                  "TOA_ALLcpg_methDiff_HEARD_ROSS_pw.png"
                                  ),
                width = 30, height = 25, units = "cm", device = "png")


table(meth.diff.df.plot$chr[!duplicated(meth.diff.df.plot$locus)])
meth.diff.df.plot2 <- meth.diff.df.plot[meth.diff.df.plot$chr == "JAAKFY010000002",]
meth.diff.df.plot.locus2 <- meth.diff.df.plot.locus[meth.diff.df.plot.locus$chr == "JAAKFY010000002",]
diff.plot2 <- ggplot2::ggplot() +
  ggplot2::geom_point(data = meth.diff.df.plot2,
                      ggplot2::aes(x = pop, y = perc_methylation, colour = locus)) +
  ggplot2::geom_line(data = meth.diff.df.plot.locus2,
                     ggplot2::aes(x = pop, y = mean_pop, colour = locus, group = locus)) +
  ggplot2::geom_point(data = meth.diff.df.plot.locus2,
                     ggplot2::aes(x = pop, y = mean_pop, colour = locus, group = locus),
                     size = 5) +
  ggplot2::ylim(c(0,100)) +
  ggplot2::facet_wrap(~locus) +
  ggplot2::theme(legend.position = "none")
  # ggplot2::guides(colour = ggplot2::guide_legend(ncol = 1))
  # ggplot2::guides(colour = ggplot2::guide_legend(ncol = 5))

print(diff.plot2)

ggplot2::ggsave(plot = diff.plot2, 
                filename = paste0(figure.path,
                                  "TOA_ALLcpg_methDiff_HEARD_ROSS_pw_CHR02.png"
                                  ),
                width = 30, height = 25, units = "cm", device = "png")

meth.diff.df.plot2 <- meth.diff.df.plot[meth.diff.df.plot$chr == "JAAKFY010000014",]
meth.diff.df.plot.locus2 <- meth.diff.df.plot.locus[meth.diff.df.plot.locus$chr == "JAAKFY010000014",]
diff.plot14 <- ggplot2::ggplot() +
  ggplot2::geom_point(data = meth.diff.df.plot2,
                      ggplot2::aes(x = pop, y = perc_methylation, colour = locus)) +
  ggplot2::geom_line(data = meth.diff.df.plot.locus2,
                     ggplot2::aes(x = pop, y = mean_pop, colour = locus, group = locus)) +
  ggplot2::geom_point(data = meth.diff.df.plot.locus2,
                     ggplot2::aes(x = pop, y = mean_pop, colour = locus, group = locus),
                     size = 5) +
  ggplot2::ylim(c(0,100)) +
  ggplot2::facet_wrap(~locus) +
  ggplot2::theme(legend.position = "none")
  # ggplot2::guides(colour = ggplot2::guide_legend(ncol = 1))
  # ggplot2::guides(colour = ggplot2::guide_legend(ncol = 5))

print(diff.plot14)

ggplot2::ggsave(plot = diff.plot14, 
                filename = paste0(figure.path,
                                  "TOA_ALLcpg_methDiff_HEARD_ROSS_pw_CHR14.png"
                                  ),
                width = 30, height = 25, units = "cm", device = "png")


```

##### Heatmap

```{r}
meth.diff.df <- as.data.frame(t(methylKit::percMethylation(meth.diff)))
loc.names <- paste0(meth.diff$chr,"_",meth.diff$start)
colnames(meth.diff.df) <- loc.names

meth.diff.df$individual <- rownames(meth.diff.df) 
metadata.new <- metadata[metadata$SAMPLE_ID %in% meth.diff@sample.ids,]
metadata.new <- metadata.new[order(match(metadata.new$SAMPLE_ID, meth.diff@sample.ids)),]
identical(metadata.new$SAMPLE_ID, meth.diff@sample.ids)
identical(metadata.new$SAMPLE_ID,meth.diff.df$individual)
meth.diff.df$pop <- metadata.new$region


heat.plot.data <- meth.diff.df %>%
  tibble::as_tibble() %>%
  tidyr::gather(key = "CpG", value = "Methylation_percentage", -individual, -pop,
                factor_key = FALSE) %>%
  dplyr::group_by(pop, CpG) %>%
  dplyr::mutate(CpG_mean = mean(Methylation_percentage, na.rm = TRUE)) %>%
  dplyr::arrange(pop, individual, CpG) %>%
  dplyr::mutate(individual = factor(individual, levels = unique(individual))) %>%
  dplyr::mutate(CpG = factor(CpG, levels = unique(CpG)))
cpgs <- heat.plot.data$CpG[heat.plot.data$individual == heat.plot.data$individual[1]]
cpgs_mean <- heat.plot.data$CpG_mean[heat.plot.data$individual == heat.plot.data$individual[1]]
cpgs <- cpgs[order(cpgs_mean)]
heat.plot.data$CpG <- factor(heat.plot.data$CpG, levels = unique(cpgs))
heat.plot.data <- heat.plot.data[order(heat.plot.data$CpG),]

heat.plot <- ggplot2::ggplot(data = heat.plot.data) + 
    ggplot2::geom_tile(ggplot2::aes(individual, y = CpG, fill = Methylation_percentage)) +
    ggplot2::scale_fill_gradient2(low = "blue", mid = "yellow", high = "red", midpoint = 50) +
    
    ggnewscale::new_scale_fill() +
    # ggplot2::geom_tile(ggplot2::aes(individual, y = "pop", fill = pop, height = 10)) +
    # ggplot2::scale_fill_manual(values = c("#4d6b40","#adcf9f", "#c5d6e6", "#08306B")) +
    # ggplot2::scale_fill_manual(values = c( "#c5d6e6", "#08306B")) +
    ggplot2::scale_y_discrete(drop = FALSE) +
    ggplot2::theme(legend.position = "top",
                   axis.title.y = ggplot2::element_blank(),
          axis.text.x = ggplot2::element_text(face = "bold", color = "black", 
                                        size = 10, angle = 90, hjust = 1, vjust = 0))
print(heat.plot)

ggplot2::ggsave(heat.plot, bg = "white",
                filename = paste0(figure.path,"TOA_ALLcpg_methDiff_HEARD_ROSS_HEATMAP.png"), 
                width = 45, height = 30, units = "cm")
```


#### PCA - HYPER

```{r}
myDiff25p.hyper.gr <- as(myDiff25p.hyper, "GRanges")
meth.diff <- methylKit::selectByOverlap(meth.HR, myDiff25p.hyper.gr) 

methylKit::PCASamples(meth.diff, screeplot = TRUE)
dev.print(file = paste0(figure.path,"PCA_plot_diffMeth_HYPER_Region_subset_HEARD_ROSS_scree.png"),
          device = png, res = 300,width = 30,height = 15,units = "cm")

methylKit::PCASamples(meth.diff, comp = c(1,2))
dev.print(file = paste0(figure.path,
                        "PCA_plot_diffMeth_HYPER_Region_subset_HEARD_ROSS_PC12.png"
                        ),
          device = png, res = 300,width = 30,height = 15,units = "cm")

methylKit::PCASamples(meth.diff, comp = c(1,3))
dev.print(file = paste0(figure.path,
                        "PCA_plot_diffMeth_HYPER_Region_subset_HEARD_ROSS_PC13.png"
                        ),
          device = png, res = 300,width = 30,height = 15,units = "cm")

methylKit::PCASamples(meth.diff, comp = c(2,3))
dev.print(file = paste0(figure.path,
                        "PCA_plot_diffMeth_HYPER_Region_subset_HEARD_ROSS_PC23.png"
                        ),
          device = png, res = 300,width = 30,height = 15,units = "cm")
```


##### Plot CpG difference

```{r}
myDiff25p.hyper.gr <- as(myDiff25p.hyper, "GRanges")
hist(myDiff25p.hyper.gr$meth.diff)
min(myDiff25p.hyper.gr$meth.diff)#-30.58783 %
max(myDiff25p.hyper.gr$meth.diff)#47.26297d %

meth.diff <- methylKit::selectByOverlap(meth.HR, myDiff25p.hyper.gr)
length(unique(meth.diff$chr)) #25
meth.diff.df <- as.data.frame(t(methylKit::percMethylation(meth.diff)))
loc.names <- paste0(meth.diff$chr,"_",meth.diff$start)
colnames(meth.diff.df) <- loc.names

meth.diff.df$individual <- rownames(meth.diff.df) 
identical(metadata3$SAMPLE_ID,meth.diff.df$individual)
meth.diff.df$pop <- metadata3$region

meth.diff.df.plot <- tidyr::gather(meth.diff.df, key = locus, 
                           value = perc_methylation, -individual, -pop,
                           factor_key = TRUE) %>%
  dplyr::mutate(chr = stringr::str_split_fixed(locus, 
                                               pattern = "_", 
                                               n = 2)[,1])
meth.diff.df.plot.locus <- meth.diff.df.plot %>% dplyr::group_by(chr,locus,pop) %>% 
  dplyr::summarise(mean_pop = mean(perc_methylation, na.rm = TRUE),
            .groups = 'drop')

meth.diff.df.plot.centroid <- meth.diff.df.plot %>% dplyr::group_by(chr,pop) %>% 
  dplyr::summarise(mean_pop = mean(perc_methylation, na.rm = TRUE),
            .groups = 'drop')

meth.diff.df.plot.label <- meth.diff.df.plot %>% dplyr::group_by(chr) %>% 
  dplyr::summarise(nmarkers = length(unique(locus)),
            .groups = 'drop')


HYPER.plot <- ggplot2::ggplot() + 
  ggplot2::geom_violin(data = meth.diff.df.plot, 
                       ggplot2::aes(x = chr, y = perc_methylation)) + 
  ggplot2::geom_point(data = meth.diff.df.plot.centroid, 
                      ggplot2::aes(x = chr, y = mean_pop, colour = pop),
                      size = 7, shape = 5) + 
  ggplot2::geom_point(data = meth.diff.df.plot,
                      ggplot2::aes(x = chr, y = perc_methylation, colour = pop),
                      size = 1) +
  ggplot2::geom_text(data = meth.diff.df.plot.label,
                     ggplot2::aes(x = chr, y = 110, label = nmarkers),
                     size = 4) +
  ggplot2::geom_point(data = meth.diff.df.plot.locus, 
                      ggplot2::aes(x = chr, y = mean_pop, colour = pop),
                      size = 3, shape = 15) +
    ggplot2::theme(
        panel.background = ggplot2::element_blank(),
        panel.border = ggplot2::element_blank(),
        panel.grid = ggplot2::element_blank(),
        axis.line = ggplot2::element_line("gray25"),
        text = ggplot2::element_text(size = 12),
        axis.text = ggplot2::element_text(angle = 90,size = 10, colour = "gray25"),
        axis.title = ggplot2::element_text(size = 14, colour = "gray25"),
        legend.title = ggplot2::element_text(size = 14),
        legend.text = ggplot2::element_text(size = 10),
        legend.key = ggplot2::element_blank())

print(HYPER.plot)

ggplot2::ggsave(plot = HYPER.plot, 
                filename = paste0(figure.path,"TOA_HYPERcpg_methDiff_HEARD_ROSS.png"),
                width = 30, height = 15, units = "cm", device = "png")
```

#### PCA - HYPO

```{r}
myDiff25p.hypo.gr <- as(myDiff25p.hypo, "GRanges")
meth.diff <- methylKit::selectByOverlap(meth.HR, myDiff25p.hypo.gr) 

methylKit::PCASamples(meth.diff, screeplot = TRUE)
dev.print(file = paste0(figure.path,"
                        PCA_plot_diffMeth_HYPO_Region_subset_HEARD_ROSS_scree.png"),
          device = png, res = 300,width = 30,height = 15,units = "cm")

methylKit::PCASamples(meth.diff, comp = c(1,2))
dev.print(file = paste0(figure.path,
                        "PCA_plot_diffMeth_HYPO_Region_subset_HEARD_ROSS_PC12.png"
                        ),
          device = png, res = 300,width = 30,height = 15,units = "cm")

methylKit::PCASamples(meth.diff, comp = c(1,3))
dev.print(file = paste0(figure.path,
                        "PCA_plot_diffMeth_HYPO_Region_subset_HEARD_ROSS_PC13.png"
                        ),
          device = png, res = 300,width = 30,height = 15,units = "cm")

methylKit::PCASamples(meth.diff, comp = c(2,3))
dev.print(file = paste0(figure.path,
                        "PCA_plot_diffMeth_HYPO_Region_subset_HEARD_ROSS_PC23.png"
                        ),
          device = png, res = 300,width = 30,height = 15,units = "cm")
```


##### Plot CpG difference

```{r}
myDiff25p.hypo.gr <- as(myDiff25p.hypo, "GRanges")
hist(myDiff25p.hypo.gr$meth.diff)
min(myDiff25p.hypo.gr$meth.diff)
max(myDiff25p.hypo.gr$meth.diff)

meth.diff <- methylKit::selectByOverlap(meth.HR, myDiff25p.hypo.gr)
length(unique(meth.diff$chr)) #25
meth.diff.df <- as.data.frame(t(methylKit::percMethylation(meth.diff)))
loc.names <- paste0(meth.diff$chr,"_",meth.diff$start)
colnames(meth.diff.df) <- loc.names

meth.diff.df$individual <- rownames(meth.diff.df) 
identical(metadata3$SAMPLE_ID, meth.diff.df$individual)
meth.diff.df$pop <- metadata3$region

meth.diff.df.plot <- tidyr::gather(meth.diff.df, key = locus, 
                           value = perc_methylation, -individual, -pop,
                           factor_key = TRUE) %>%
  dplyr::mutate(chr = stringr::str_split_fixed(locus, 
                                               pattern = "_", 
                                               n = 2)[,1])
meth.diff.df.plot.locus <- meth.diff.df.plot %>% dplyr::group_by(chr,locus,pop) %>% 
  dplyr::summarise(mean_pop = mean(perc_methylation, na.rm = TRUE),
            .groups = 'drop')

meth.diff.df.plot.centroid <- meth.diff.df.plot %>% dplyr::group_by(chr,pop) %>% 
  dplyr::summarise(mean_pop = mean(perc_methylation, na.rm = TRUE),
            .groups = 'drop')

meth.diff.df.plot.label <- meth.diff.df.plot %>% dplyr::group_by(chr) %>% 
  dplyr::summarise(nmarkers = length(unique(locus)),
            .groups = 'drop')


HYPO.plot <- ggplot2::ggplot() + 
  ggplot2::geom_violin(data = meth.diff.df.plot, 
                       ggplot2::aes(x = chr, y = perc_methylation)) + 
  ggplot2::geom_point(data = meth.diff.df.plot.centroid, 
                      ggplot2::aes(x = chr, y = mean_pop, colour = pop),
                      size = 7, shape = 5) + 
  ggplot2::geom_point(data = meth.diff.df.plot,
                      ggplot2::aes(x = chr, y = perc_methylation, colour = pop),
                      size = 1) +
  ggplot2::geom_text(data = meth.diff.df.plot.label,
                     ggplot2::aes(x = chr, y = 110, label = nmarkers),
                     size = 4) +
  ggplot2::geom_point(data = meth.diff.df.plot.locus, 
                      ggplot2::aes(x = chr, y = mean_pop, colour = pop),
                      size = 3, shape = 15) +
    ggplot2::theme(
        panel.background = ggplot2::element_blank(),
        panel.border = ggplot2::element_blank(),
        panel.grid = ggplot2::element_blank(),
        axis.line = ggplot2::element_line("gray25"),
        text = ggplot2::element_text(size = 12),
        axis.text = ggplot2::element_text(angle = 90,size = 10, colour = "gray25"),
        axis.title = ggplot2::element_text(size = 14, colour = "gray25"),
        legend.title = ggplot2::element_text(size = 14),
        legend.text = ggplot2::element_text(size = 10),
        legend.key = ggplot2::element_blank())

print(HYPO.plot)

ggplot2::ggsave(plot = HYPO.plot, 
                filename = paste0(figure.path,"TOA_HYPOcpg_methDiff_HEARD_ROSS.png"),
                width = 30, height = 15, units = "cm", device = "png")
```


### EANT
#### Prepare data

```{r}
print(load(paste0(rdata.path,"Toothfish_metadata.Rdata")))

metadata3 <- metadata2[metadata2$region %in% c("58-4-1", "58-4-2"),]
table(metadata3$region)
print(load(paste0(rdata.path,"Toothfish_methylation_objects_raw.Rdata")))

filtered.myobj3 <- methylKit::reorganize(
  methylObj = filtered.myobj,
  sample.ids = metadata3$SAMPLE_ID,
  treatment = as.numeric(factor(metadata3$region)),
  chunk.size = 1e+06,
  save.db = TRUE,
  suffix = "subset_EANT_raw"
)

# filtered.myobj3@treatment <- as.numeric(factor(metadata3$region))
filtered.myobj3 <- methylKit::filterByCoverage(filtered.myobj3,lo.count = 10,lo.perc = NULL,
                                      hi.count = NULL, hi.perc = 99)

meth.EANT <- methylKit::unite(filtered.myobj3, destrand = FALSE,
                             save.db = TRUE,
                             suffix = "subset_EANT",
                             min.per.group = 20L, #missing data allowed
                             mc.cores = 25)
#716,400  CpGs after coverage filter

# save(meth.HR,meth.EANT,
#      file = paste0(rdata.path,
#                    "Toothfish_subset_pairwise_methylation_objects_raw.Rdata"))
```


#### diffMethPerChr

```{r}
print(load(paste0(rdata.path,"Toothfish_subset_pairwise_methylation_objects_raw.Rdata")))
metadata3 <- metadata2[metadata2$region %in% c("58-4-1", "58-4-2"),]
covariates <- dplyr::select(metadata3, Sex, mat, Age)
myDiff <- methylKit::calculateDiffMeth(meth.EANT,
                                       covariates = covariates,
                                       adjust = "fdr",
                                       chunk.size = 1e+06,
                                       mc.cores = 10)
meth.diff.chr <- methylKit::diffMethPerChr(myDiff,plot = TRUE, 
                                           qvalue.cutoff = 0.01, meth.cutoff = 25)
myDiff25p.hyper <- methylKit::getMethylDiff(myDiff, difference = 25,qvalue = 0.01, type = "hyper") #35 
myDiff25p.hypo <- methylKit::getMethylDiff(myDiff, difference = 25,qvalue = 0.01, type = "hypo") #24
myDiff25p <- methylKit::getMethylDiff(myDiff, difference = 25, qvalue = 0.01)

# save(myDiff,myDiff25p, myDiff25p.hyper, myDiff25p.hypo,
#      file = paste0(rdata.path,
#                    "Toothfish_DMS_Region_subset_EANT.Rdata"))


myDiff.df <- as.data.frame(as(myDiff, "GRanges"))
myDiff25p.df <- as.data.frame(as(myDiff25p, "GRanges"))
volcano.plot <- ggplot2::ggplot() +
  ggplot2::geom_point(data = myDiff.df, ggplot2::aes(x = meth.diff, 
                                                  y = -log10(pvalue))) +
  ggplot2::geom_point(data = myDiff25p.df, ggplot2::aes(x = meth.diff, 
                                                  y = -log10(pvalue)),
                      colour = "red")
print(volcano.plot)

ggplot2::ggsave(volcano.plot, bg = "white",
                filename = paste0(figure.path,
                                  "TOA_ALL_DMS25_VOLCANO_methylkit_regions_subset_EANT.png"
                                  ), 
                width = 30, height = 15, units = "cm")
```

#### Load data

```{r}
print(load(paste0(rdata.path,"Toothfish_DMS_Region_subset_EANT.Rdata")))

myDiff25p.gr <- as(myDiff25p, "GRanges")
meth.diff <- methylKit::selectByOverlap(meth.EANT, myDiff25p.gr) 

# DMS <- data.frame(DMS = NULL, region_pw = NULL)
DMS.EANT <- data.frame(DMS = paste0(meth.diff$chr,"_", meth.diff$start),
                  region_pw = "EANT4.1_EANT4.2")
DMS <- rbind(DMS, DMS.EANT)

# save(DMS, DMS.HR,DMS.EANT,
#      file = paste0(rdata.path,"Toothfish_DMS_CpGsites.Rdata"))
```

#### PCA - ALL

```{r}
PCobj <- methylKit::PCASamples(meth.diff, comp = c(1,2), adj.lim = c(0,0), obj.return = TRUE)

var_frac <- PCobj$sdev/sum(PCobj$sdev)
print(paste0("Total variance explained: ",signif(sum(var_frac[1:3]) * 100, 3), "%")) 
barplot(var_frac[1:20], main = "variance explained", 
        col = heat.colors(length(var_frac[1:20])))

# metadata3 <- metadata2[metadata2$tissue %in% "M",]
metadata.new <- metadata[metadata$SAMPLE_ID %in% meth.diff@sample.ids,]
metadata.new <- metadata.new[order(match(metadata.new$SAMPLE_ID, meth.diff@sample.ids)),]
identical(metadata.new$SAMPLE_ID, meth.diff@sample.ids)
identical(metadata.new$SAMPLE_ID, rownames(PCobj$x))

pop <- metadata.new$region

# pop <- metadata.new$`Date caught`

# colours5 <- adegenet::funky(4)
# colours5 <- c("#4d6b40","#adcf9f", "#c5d6e6", "#08306B")
colours5 <- c("#4d6b40","#adcf9f")

# plot.title <- "Toothfish methylation (1,316,799 CpG sites)"
plot.title <- "Toothfish methylation - EANT1 vs EANT2 (59 CpG sites)"


data <- data.frame(ind = rownames(PCobj$x),PCobj$x,  pop = pop)
data <- merge(data, aggregate(cbind(mean.x = PC1, mean.y = PC2) ~
                                pop, data, mean), by = "pop")

pca.plot <- ggplot2::ggplot(data, ggplot2::aes(PC1, PC2, color = pop)) +
  ggplot2::geom_point(size = 4) + 
  ggplot2::geom_point(ggplot2::aes(x = mean.x,
                                   y = mean.y, color = pop),
                      size = 7,
                      shape = 17) +
  ggplot2::scale_colour_manual(values = colours5) +
  ggplot2::scale_fill_manual(values = colours5) + 
  ggplot2::geom_hline(yintercept = 0) +
  ggplot2::geom_vline(xintercept = 0) + 
  ggplot2::labs(
    subtitle = paste0("Total variance explained: ",
                      signif(sum(var_frac[1:4]) * 100, 3), "%"),
    y = paste0("PC2; ", signif(var_frac[2] * 100, digits = 2), " %"),
    x = paste0("PC1; ", signif(var_frac[1] * 100, digits = 2), " %"),
    title = plot.title,
    caption = "") + 
  ggplot2::theme_minimal() +
  ggplot2::theme(
    axis.text = ggplot2::element_text(size = 15),
    axis.title.x = ggplot2::element_text(size = 20),
    axis.title.y = ggplot2::element_text(size = 20),
    legend.text = ggplot2::element_text(size = 15)
  )

print(pca.plot)
ggplot2::ggsave(pca.plot, bg = "white",
                filename = paste0(figure.path,"PCA_plot_diffMeth_ALL_Region_subset_EANT_PC12.png"), 
                width = 30, height = 15, units = "cm")



data <- data.frame(ind = rownames(PCobj$x),PCobj$x,  pop = pop)
data <- merge(data, aggregate(cbind(mean.x = PC1, mean.y = PC3) ~
                                pop, data, mean), by = "pop")

pca.plot <- ggplot2::ggplot(data, ggplot2::aes(PC1, PC3, color = pop)) +
  ggplot2::geom_point(size = 4) + 
  ggplot2::geom_point(ggplot2::aes(x = mean.x,
                                   y = mean.y, color = pop),
                      size = 7,
                      shape = 17) +
  ggplot2::scale_colour_manual(values = colours5) +
  ggplot2::scale_fill_manual(values = colours5) + 
  ggplot2::geom_hline(yintercept = 0) +
  ggplot2::geom_vline(xintercept = 0) + 
  ggplot2::labs(
    subtitle = paste0("Total variance explained: ",
                      signif(sum(var_frac[1:4]) * 100, 3), "%"),
    y = paste0("PC3; ", signif(var_frac[3] * 100, digits = 2), " %"),
    x = paste0("PC1; ", signif(var_frac[1] * 100, digits = 2), " %"),
    title = plot.title,
    caption = "") + 
  ggplot2::theme_minimal() +
  ggplot2::theme(
    axis.text = ggplot2::element_text(size = 15),
    axis.title.x = ggplot2::element_text(size = 20),
    axis.title.y = ggplot2::element_text(size = 20),
    legend.text = ggplot2::element_text(size = 15)
  )

print(pca.plot)
ggplot2::ggsave(pca.plot, bg = "white",
                filename = paste0(figure.path,"PCA_plot_diffMeth_ALL_Region_subset_EANT_PC13.png"), 
                width = 30, height = 15, units = "cm")




data <- data.frame(ind = rownames(PCobj$x),PCobj$x,  pop = pop)
data <- merge(data, aggregate(cbind(mean.x = PC2, mean.y = PC3) ~
                                pop, data, mean), by = "pop")

pca.plot <- ggplot2::ggplot(data, ggplot2::aes(PC2, PC3, color = pop)) +
  ggplot2::geom_point(size = 4) + 
  ggplot2::geom_point(ggplot2::aes(x = mean.x,
                                   y = mean.y, color = pop),
                      size = 7,
                      shape = 17) +
  ggplot2::scale_colour_manual(values = colours5) +
  ggplot2::scale_fill_manual(values = colours5) + 
  ggplot2::geom_hline(yintercept = 0) +
  ggplot2::geom_vline(xintercept = 0) + 
  ggplot2::labs(
    subtitle = paste0("Total variance explained: ",
                      signif(sum(var_frac[1:4]) * 100, 3), "%"),
    y = paste0("PC3; ", signif(var_frac[3] * 100, digits = 2), " %"),
    x = paste0("PC2; ", signif(var_frac[2] * 100, digits = 2), " %"),
    title = plot.title,
    caption = "") + 
  ggplot2::theme_minimal() +
  ggplot2::theme(
    axis.text = ggplot2::element_text(size = 15),
    axis.title.x = ggplot2::element_text(size = 20),
    axis.title.y = ggplot2::element_text(size = 20),
    legend.text = ggplot2::element_text(size = 15)
  )

print(pca.plot)
ggplot2::ggsave(pca.plot, bg = "white",
                filename = paste0(figure.path,"PCA_plot_diffMeth_ALL_Region_subset_EANT_PC23.png"), 
                width = 30, height = 15, units = "cm")

```


##### Plot CpG difference

```{r}
myDiff25p.gr <- as(myDiff25p, "GRanges")
hist(myDiff25p.gr$meth.diff)
min(myDiff25p.gr$meth.diff)#-30.58783 %
max(myDiff25p.gr$meth.diff)#35.30028 %

meth.diff <- methylKit::selectByOverlap(meth.EANT, myDiff25p.gr)
length(unique(meth.diff$chr)) #19
meth.diff.df <- as.data.frame(t(methylKit::percMethylation(meth.diff)))
loc.names <- paste0(meth.diff$chr,"_",meth.diff$start)
colnames(meth.diff.df) <- loc.names

meth.diff.df$individual <- rownames(meth.diff.df) 
metadata.new <- metadata[metadata$SAMPLE_ID %in% meth.diff@sample.ids,]
metadata.new <- metadata.new[order(match(metadata.new$SAMPLE_ID, meth.diff@sample.ids)),]
identical(metadata.new$SAMPLE_ID, meth.diff@sample.ids)
identical(metadata.new$SAMPLE_ID,meth.diff.df$individual)
meth.diff.df$pop <- metadata.new$region

meth.diff.df.plot <- tidyr::gather(meth.diff.df, key = locus, 
                           value = perc_methylation, -individual, -pop,
                           factor_key = TRUE) %>%
  dplyr::mutate(chr = stringr::str_split_fixed(locus, 
                                               pattern = "_", 
                                               n = 2)[,1])

meth.diff.df.plot.locus <- meth.diff.df.plot %>% dplyr::group_by(chr,locus,pop) %>% 
  dplyr::summarise(mean_pop = mean(perc_methylation, na.rm = TRUE),
            .groups = 'drop')

meth.diff.df.plot.centroid <- meth.diff.df.plot %>% dplyr::group_by(chr,pop) %>% 
  dplyr::summarise(mean_pop = mean(perc_methylation, na.rm = TRUE),
            .groups = 'drop')

meth.diff.df.plot.label <- meth.diff.df.plot %>% dplyr::group_by(chr) %>% 
  dplyr::summarise(nmarkers = length(unique(locus)),
            .groups = 'drop')


ALL.plot <- ggplot2::ggplot() + 
  ggplot2::geom_violin(data = meth.diff.df.plot, 
                       ggplot2::aes(x = chr, y = perc_methylation)) + 
  ggplot2::geom_point(data = meth.diff.df.plot.centroid, 
                      ggplot2::aes(x = chr, y = mean_pop, colour = pop),
                      size = 7, shape = 5) + 
  ggplot2::geom_point(data = meth.diff.df.plot,
                      ggplot2::aes(x = chr, y = perc_methylation, colour = pop),
                      size = 1) +
  ggplot2::geom_text(data = meth.diff.df.plot.label,
                     ggplot2::aes(x = chr, y = 110, label = nmarkers),
                     size = 4) +
  ggplot2::geom_point(data = meth.diff.df.plot.locus, 
                      ggplot2::aes(x = chr, y = mean_pop, colour = pop),
                      size = 3, shape = 15) +
    ggplot2::theme(
        panel.background = ggplot2::element_blank(),
        panel.border = ggplot2::element_blank(),
        panel.grid = ggplot2::element_blank(),
        axis.line = ggplot2::element_line("gray25"),
        text = ggplot2::element_text(size = 12),
        axis.text = ggplot2::element_text(angle = 90,size = 10, colour = "gray25"),
        axis.title = ggplot2::element_text(size = 14, colour = "gray25"),
        legend.title = ggplot2::element_text(size = 14),
        legend.text = ggplot2::element_text(size = 10),
        legend.key = ggplot2::element_blank())

print(ALL.plot)

ggplot2::ggsave(plot = ALL.plot, 
                filename = paste0(figure.path,"TOA_ALLcpg_methDiff_EANT.png"),
                width = 30, height = 15, units = "cm", device = "png")



diff.plot <- ggplot2::ggplot() +
  # ggplot2::geom_point(data = meth.diff.df.plot,
  #                     ggplot2::aes(x = pop, y = perc_methylation, colour = locus)) +
  ggplot2::geom_line(data = meth.diff.df.plot.locus,
                     ggplot2::aes(x = pop, y = mean_pop, colour = locus, group = locus)) +
  ggplot2::geom_point(data = meth.diff.df.plot.locus,
                     ggplot2::aes(x = pop, y = mean_pop, colour = locus, group = locus),
                     size = 5) +
  ggplot2::ylim(c(0,100)) +
  ggplot2::facet_wrap(~chr) +
  ggplot2::theme(legend.position = "none")
  # ggplot2::guides(colour = ggplot2::guide_legend(ncol = 1))
  # ggplot2::guides(colour = ggplot2::guide_legend(ncol = 5))

print(diff.plot)

ggplot2::ggsave(plot = diff.plot, 
                filename = paste0(figure.path,
                                  "TOA_ALLcpg_methDiff_EANT_pw.png"
                                  ),
                width = 30, height = 25, units = "cm", device = "png")


table(meth.diff.df.plot$chr[!duplicated(meth.diff.df.plot$locus)])
meth.diff.df.plot2 <- meth.diff.df.plot[meth.diff.df.plot$chr == "JAAKFY010000017",]
meth.diff.df.plot.locus2 <- meth.diff.df.plot.locus[meth.diff.df.plot.locus$chr == "JAAKFY010000017",]
diff.plot17 <- ggplot2::ggplot() +
  ggplot2::geom_point(data = meth.diff.df.plot2,
                      ggplot2::aes(x = pop, y = perc_methylation, colour = locus)) +
  ggplot2::geom_line(data = meth.diff.df.plot.locus2,
                     ggplot2::aes(x = pop, y = mean_pop, colour = locus, group = locus)) +
  ggplot2::geom_point(data = meth.diff.df.plot.locus2,
                     ggplot2::aes(x = pop, y = mean_pop, colour = locus, group = locus),
                     size = 5) +
  ggplot2::ylim(c(0,100)) +
  ggplot2::facet_wrap(~locus) +
  ggplot2::theme(legend.position = "none")
  # ggplot2::guides(colour = ggplot2::guide_legend(ncol = 1))
  # ggplot2::guides(colour = ggplot2::guide_legend(ncol = 5))

print(diff.plot17)

ggplot2::ggsave(plot = diff.plot17, 
                filename = paste0(figure.path,
                                  "TOA_ALLcpg_methDiff_EANT_pw_CHR17.png"
                                  ),
                width = 30, height = 25, units = "cm", device = "png")

meth.diff.df.plot2 <- meth.diff.df.plot[meth.diff.df.plot$chr == "JAAKFY010000026",]
meth.diff.df.plot.locus2 <- meth.diff.df.plot.locus[meth.diff.df.plot.locus$chr == "JAAKFY010000026",]
diff.plot26 <- ggplot2::ggplot() +
  ggplot2::geom_point(data = meth.diff.df.plot2,
                      ggplot2::aes(x = pop, y = perc_methylation, colour = locus)) +
  ggplot2::geom_line(data = meth.diff.df.plot.locus2,
                     ggplot2::aes(x = pop, y = mean_pop, colour = locus, group = locus)) +
  ggplot2::geom_point(data = meth.diff.df.plot.locus2,
                     ggplot2::aes(x = pop, y = mean_pop, colour = locus, group = locus),
                     size = 5) +
  ggplot2::ylim(c(0,100)) +
  ggplot2::facet_wrap(~locus) +
  ggplot2::theme(legend.position = "none")
  # ggplot2::guides(colour = ggplot2::guide_legend(ncol = 1))
  # ggplot2::guides(colour = ggplot2::guide_legend(ncol = 5))

print(diff.plot26)

ggplot2::ggsave(plot = diff.plot26, 
                filename = paste0(figure.path,
                                  "TOA_ALLcpg_methDiff_EANT_pw_CHR26.png"
                                  ),
                width = 30, height = 25, units = "cm", device = "png")

```


##### Heatmap

```{r}
meth.diff.df <- as.data.frame(t(methylKit::percMethylation(meth.diff)))
loc.names <- paste0(meth.diff$chr,"_",meth.diff$start)
colnames(meth.diff.df) <- loc.names

meth.diff.df$individual <- rownames(meth.diff.df) 
metadata.new <- metadata[metadata$SAMPLE_ID %in% meth.diff@sample.ids,]
metadata.new <- metadata.new[order(match(metadata.new$SAMPLE_ID, meth.diff@sample.ids)),]
identical(metadata.new$SAMPLE_ID, meth.diff@sample.ids)
identical(metadata.new$SAMPLE_ID,meth.diff.df$individual)
meth.diff.df$pop <- metadata.new$region


heat.plot.data <- meth.diff.df %>%
  tibble::as_tibble() %>%
  tidyr::gather(key = "CpG", value = "Methylation_percentage", -individual, -pop,
                factor_key = FALSE) %>%
  dplyr::group_by(pop, CpG) %>%
  dplyr::mutate(CpG_mean = mean(Methylation_percentage, na.rm = TRUE)) %>%
  dplyr::arrange(pop, individual, CpG) %>%
  dplyr::mutate(individual = factor(individual, levels = unique(individual))) %>%
  dplyr::mutate(CpG = factor(CpG, levels = unique(CpG)))
cpgs <- heat.plot.data$CpG[heat.plot.data$individual == heat.plot.data$individual[1]]
cpgs_mean <- heat.plot.data$CpG_mean[heat.plot.data$individual == heat.plot.data$individual[1]]
cpgs <- cpgs[order(cpgs_mean)]
heat.plot.data$CpG <- factor(heat.plot.data$CpG, levels = unique(cpgs))
heat.plot.data <- heat.plot.data[order(heat.plot.data$CpG),]

heat.plot <- ggplot2::ggplot(data = heat.plot.data) + 
    ggplot2::geom_tile(ggplot2::aes(individual, y = CpG, fill = Methylation_percentage)) +
    ggplot2::scale_fill_gradient2(low = "blue", mid = "yellow", high = "red", midpoint = 50) +
    
    ggnewscale::new_scale_fill() +
    # ggplot2::geom_tile(ggplot2::aes(individual, y = "pop", fill = pop, height = 10)) +
    # ggplot2::scale_fill_manual(values = c("#4d6b40","#adcf9f", "#c5d6e6", "#08306B")) +
    # ggplot2::scale_fill_manual(values = c( "#c5d6e6", "#08306B")) +
    ggplot2::scale_y_discrete(drop = FALSE) +
    ggplot2::theme(legend.position = "top",
                   axis.title.y = ggplot2::element_blank(),
          axis.text.x = ggplot2::element_text(face = "bold", color = "black", 
                                        size = 10, angle = 90, hjust = 1, vjust = 0))
print(heat.plot)

ggplot2::ggsave(heat.plot, bg = "white",
                filename = paste0(figure.path,"TOA_ALLcpg_methDiff_EANT_HEATMAP.png"), 
                width = 45, height = 30, units = "cm")
```


#### PCA - HYPER

```{r}
myDiff25p.hyper.gr <- as(myDiff25p.hyper, "GRanges")
meth.diff <- methylKit::selectByOverlap(meth.EANT, myDiff25p.hyper.gr) 

methylKit::PCASamples(meth.diff, screeplot = TRUE)
dev.print(file = paste0(figure.path,
                        "PCA_plot_diffMeth_HYPER_Region_subset_EANT_scree.png"),
          device = png, res = 300,width = 30,height = 15,units = "cm")

methylKit::PCASamples(meth.diff, comp = c(1,2))
dev.print(file = paste0(figure.path,
                        "PCA_plot_diffMeth_HYPER_Region_subset_EANT_PC12.png"
                        ),
          device = png, res = 300,width = 30,height = 15,units = "cm")

methylKit::PCASamples(meth.diff, comp = c(1,3))
dev.print(file = paste0(figure.path,
                        "PCA_plot_diffMeth_HYPER_Region_subset_EANT_PC13.png"
                        ),
          device = png, res = 300,width = 30,height = 15,units = "cm")

methylKit::PCASamples(meth.diff, comp = c(2,3))
dev.print(file = paste0(figure.path,
                        "PCA_plot_diffMeth_HYPER_Region_subset_EANT_PC23.png"
                        ),
          device = png, res = 300,width = 30,height = 15,units = "cm")
```


##### Plot CpG difference

```{r}
myDiff25p.hyper.gr <- as(myDiff25p.hyper, "GRanges")
hist(myDiff25p.hyper.gr$meth.diff)
min(myDiff25p.hyper.gr$meth.diff)#-30.58783 %
max(myDiff25p.hyper.gr$meth.diff)#47.26297 %

meth.diff <- methylKit::selectByOverlap(meth.EANT, myDiff25p.hyper.gr)
length(unique(meth.diff$chr)) #25
meth.diff.df <- as.data.frame(t(methylKit::percMethylation(meth.diff)))
loc.names <- paste0(meth.diff$chr,"_",meth.diff$start)
colnames(meth.diff.df) <- loc.names

meth.diff.df$individual <- rownames(meth.diff.df) 
identical(metadata3$SAMPLE_ID,meth.diff.df$individual)
meth.diff.df$pop <- metadata3$region

meth.diff.df.plot <- tidyr::gather(meth.diff.df, key = locus, 
                           value = perc_methylation, -individual, -pop,
                           factor_key = TRUE) %>%
  dplyr::mutate(chr = stringr::str_split_fixed(locus, 
                                               pattern = "_", 
                                               n = 2)[,1])
meth.diff.df.plot.locus <- meth.diff.df.plot %>% dplyr::group_by(chr,locus,pop) %>% 
  dplyr::summarise(mean_pop = mean(perc_methylation, na.rm = TRUE),
            .groups = 'drop')

meth.diff.df.plot.centroid <- meth.diff.df.plot %>% dplyr::group_by(chr,pop) %>% 
  dplyr::summarise(mean_pop = mean(perc_methylation, na.rm = TRUE),
            .groups = 'drop')

meth.diff.df.plot.label <- meth.diff.df.plot %>% dplyr::group_by(chr) %>% 
  dplyr::summarise(nmarkers = length(unique(locus)),
            .groups = 'drop')


HYPER.plot <- ggplot2::ggplot() + 
  ggplot2::geom_violin(data = meth.diff.df.plot, 
                       ggplot2::aes(x = chr, y = perc_methylation)) + 
  ggplot2::geom_point(data = meth.diff.df.plot.centroid, 
                      ggplot2::aes(x = chr, y = mean_pop, colour = pop),
                      size = 7, shape = 5) + 
  ggplot2::geom_point(data = meth.diff.df.plot,
                      ggplot2::aes(x = chr, y = perc_methylation, colour = pop),
                      size = 1) +
  ggplot2::geom_text(data = meth.diff.df.plot.label,
                     ggplot2::aes(x = chr, y = 110, label = nmarkers),
                     size = 4) +
  ggplot2::geom_point(data = meth.diff.df.plot.locus, 
                      ggplot2::aes(x = chr, y = mean_pop, colour = pop),
                      size = 3, shape = 15) +
    ggplot2::theme(
        panel.background = ggplot2::element_blank(),
        panel.border = ggplot2::element_blank(),
        panel.grid = ggplot2::element_blank(),
        axis.line = ggplot2::element_line("gray25"),
        text = ggplot2::element_text(size = 12),
        axis.text = ggplot2::element_text(angle = 90,size = 10, colour = "gray25"),
        axis.title = ggplot2::element_text(size = 14, colour = "gray25"),
        legend.title = ggplot2::element_text(size = 14),
        legend.text = ggplot2::element_text(size = 10),
        legend.key = ggplot2::element_blank())

print(HYPER.plot)

ggplot2::ggsave(plot = HYPER.plot, 
                filename = paste0(figure.path,"TOA_HYPERcpg_methDiff_EANT.png"),
                width = 30, height = 15, units = "cm", device = "png")
```

#### PCA - HYPO

```{r}
myDiff25p.hypo.gr <- as(myDiff25p.hypo, "GRanges")
meth.diff <- methylKit::selectByOverlap(meth.EANT, myDiff25p.hypo.gr) 

methylKit::PCASamples(meth.diff, screeplot = TRUE)
dev.print(file = paste0(figure.path,
                        "PCA_plot_diffMeth_HYPO_Region_subset_EANT_scree.png"),
          device = png, res = 300,width = 30,height = 15,units = "cm")

methylKit::PCASamples(meth.diff, comp = c(1,2))
dev.print(file = paste0(figure.path,
                        "PCA_plot_diffMeth_HYPO_Region_subset_EANT_PC12.png"
                        ),
          device = png, res = 300,width = 30,height = 15,units = "cm")

methylKit::PCASamples(meth.diff, comp = c(1,3))
dev.print(file = paste0(figure.path,
                        "PCA_plot_diffMeth_HYPO_Region_subset_EANT_PC13.png"
                        ),
          device = png, res = 300,width = 30,height = 15,units = "cm")

methylKit::PCASamples(meth.diff, comp = c(2,3))
dev.print(file = paste0(figure.path,
                        "PCA_plot_diffMeth_HYPO_Region_subset_EANT_PC23.png"
                        ),
          device = png, res = 300,width = 30,height = 15,units = "cm")
```


##### Plot CpG difference

```{r}
myDiff25p.hypo.gr <- as(myDiff25p.hypo, "GRanges")
hist(myDiff25p.hypo.gr$meth.diff)
min(myDiff25p.hypo.gr$meth.diff)
max(myDiff25p.hypo.gr$meth.diff)

meth.diff <- methylKit::selectByOverlap(meth.EANT, myDiff25p.hypo.gr)
length(unique(meth.diff$chr)) #25
meth.diff.df <- as.data.frame(t(methylKit::percMethylation(meth.diff)))
loc.names <- paste0(meth.diff$chr,"_",meth.diff$start)
colnames(meth.diff.df) <- loc.names

meth.diff.df$individual <- rownames(meth.diff.df) 
identical(metadata3$SAMPLE_ID, meth.diff.df$individual)
meth.diff.df$pop <- metadata3$region

meth.diff.df.plot <- tidyr::gather(meth.diff.df, key = locus, 
                           value = perc_methylation, -individual, -pop,
                           factor_key = TRUE) %>%
  dplyr::mutate(chr = stringr::str_split_fixed(locus, 
                                               pattern = "_", 
                                               n = 2)[,1])
meth.diff.df.plot.locus <- meth.diff.df.plot %>% dplyr::group_by(chr,locus,pop) %>% 
  dplyr::summarise(mean_pop = mean(perc_methylation, na.rm = TRUE),
            .groups = 'drop')

meth.diff.df.plot.centroid <- meth.diff.df.plot %>% dplyr::group_by(chr,pop) %>% 
  dplyr::summarise(mean_pop = mean(perc_methylation, na.rm = TRUE),
            .groups = 'drop')

meth.diff.df.plot.label <- meth.diff.df.plot %>% dplyr::group_by(chr) %>% 
  dplyr::summarise(nmarkers = length(unique(locus)),
            .groups = 'drop')


HYPO.plot <- ggplot2::ggplot() + 
  ggplot2::geom_violin(data = meth.diff.df.plot, 
                       ggplot2::aes(x = chr, y = perc_methylation)) + 
  ggplot2::geom_point(data = meth.diff.df.plot.centroid, 
                      ggplot2::aes(x = chr, y = mean_pop, colour = pop),
                      size = 7, shape = 5) + 
  ggplot2::geom_point(data = meth.diff.df.plot,
                      ggplot2::aes(x = chr, y = perc_methylation, colour = pop),
                      size = 1) +
  ggplot2::geom_text(data = meth.diff.df.plot.label,
                     ggplot2::aes(x = chr, y = 110, label = nmarkers),
                     size = 4) +
  ggplot2::geom_point(data = meth.diff.df.plot.locus, 
                      ggplot2::aes(x = chr, y = mean_pop, colour = pop),
                      size = 3, shape = 15) +
    ggplot2::theme(
        panel.background = ggplot2::element_blank(),
        panel.border = ggplot2::element_blank(),
        panel.grid = ggplot2::element_blank(),
        axis.line = ggplot2::element_line("gray25"),
        text = ggplot2::element_text(size = 12),
        axis.text = ggplot2::element_text(angle = 90,size = 10, colour = "gray25"),
        axis.title = ggplot2::element_text(size = 14, colour = "gray25"),
        legend.title = ggplot2::element_text(size = 14),
        legend.text = ggplot2::element_text(size = 10),
        legend.key = ggplot2::element_blank())

print(HYPO.plot)

ggplot2::ggsave(plot = HYPO.plot, 
                filename = paste0(figure.path,"TOA_HYPOcpg_methDiff_EANT.png"),
                width = 30, height = 15, units = "cm", device = "png")
```



### EANT - HEARD
#### Prepare data

```{r}
print(load(paste0(rdata.path,"Toothfish_metadata.Rdata")))

metadata3 <- metadata2[metadata2$FAO2 %in% c("58.4", "58.5"),]
table(metadata3$FAO2)
print(load(paste0(rdata.path,"Toothfish_methylation_objects_raw.Rdata")))

filtered.myobj3 <- methylKit::reorganize(
  methylObj = filtered.myobj,
  sample.ids = metadata3$SAMPLE_ID,
  treatment = as.numeric(factor(metadata3$FAO2)),
  chunk.size = 1e+06,
  save.db = TRUE,
  suffix = "subset_EANT-HEARD_raw"
)

# filtered.myobj3@treatment <- as.numeric(factor(metadata3$region))
filtered.myobj3 <- methylKit::filterByCoverage(filtered.myobj3,lo.count = 10,lo.perc = NULL,
                                      hi.count = NULL, hi.perc = 99)

meth.EH <- methylKit::unite(filtered.myobj3, destrand = FALSE,
                             save.db = TRUE,
                             suffix = "subset_EANT-HEARD",
                             min.per.group = 20L, #missing data allowed
                             mc.cores = 25)
#1,321,644  CpGs after coverage filter

# save(meth.HR,meth.EANT,meth.EH,
#      file = paste0(rdata.path,
#                    "Toothfish_subset_pairwise_methylation_objects_raw.Rdata"))
```

#### diffMethPerChr

```{r}
print(load(paste0(rdata.path,"Toothfish_subset_pairwise_methylation_objects_raw.Rdata")))
metadata3 <- metadata2[metadata2$region %in% c("58-4-1", "58-4-2"),]
covariates <- dplyr::select(metadata3, Sex, mat, Age)
myDiff <- methylKit::calculateDiffMeth(meth.EH,
                                       covariates = covariates,
                                       adjust = "fdr",
                                       chunk.size = 1e+06,
                                       mc.cores = 10)
meth.diff.chr <- methylKit::diffMethPerChr(myDiff,plot = TRUE, 
                                           qvalue.cutoff = 0.01, meth.cutoff = 25)
myDiff25p.hyper <- methylKit::getMethylDiff(myDiff, difference = 25,qvalue = 0.01, type = "hyper") #218  
myDiff25p.hypo <- methylKit::getMethylDiff(myDiff, difference = 25,qvalue = 0.01, type = "hypo") #220 
myDiff25p <- methylKit::getMethylDiff(myDiff, difference = 25, qvalue = 0.01)

# save(myDiff,myDiff25p, myDiff25p.hyper, myDiff25p.hypo, 
#      file = paste0(rdata.path,
#                    "Toothfish_DMS_Region_subset_EANT-HEARD.Rdata"))


myDiff.df <- as.data.frame(as(myDiff, "GRanges"))
myDiff25p.df <- as.data.frame(as(myDiff25p, "GRanges"))
volcano.plot <- ggplot2::ggplot() +
  ggplot2::geom_point(data = myDiff.df, ggplot2::aes(x = meth.diff, 
                                                  y = -log10(pvalue))) +
  ggplot2::geom_point(data = myDiff25p.df, ggplot2::aes(x = meth.diff, 
                                                  y = -log10(pvalue)),
                      colour = "red")
print(volcano.plot)

ggplot2::ggsave(volcano.plot, bg = "white",
                filename = paste0(figure.path,
                                  "TOA_ALL_DMS25_VOLCANO_methylkit_regions_subset_EANT-HEARD.png"
                                  ), 
                width = 30, height = 15, units = "cm")
```

#### Load data

```{r}
print(load(paste0(rdata.path,"Toothfish_DMS_Region_subset_EANT-HEARD.Rdata")))

myDiff25p.gr <- as(myDiff25p, "GRanges")
meth.diff <- methylKit::selectByOverlap(meth.EH, myDiff25p.gr) 

# DMS <- data.frame(DMS = NULL, region_pw = NULL)
DMS.EH <- data.frame(DMS = paste0(meth.diff$chr,"_", meth.diff$start),
                  region_pw = "EANT_HEARD")
DMS <- rbind(DMS, DMS.EH)

# save(DMS, DMS.HR,DMS.EANT,DMS.EH,
#      file = paste0(rdata.path,"Toothfish_DMS_CpGsites.Rdata"))
```

#### PCA - ALL

```{r}
print(load(paste0(rdata.path,"Toothfish_DMS_Region_subset_EANT-HEARD.Rdata")))

myDiff25p.gr <- as(myDiff25p, "GRanges")
meth.diff <- methylKit::selectByOverlap(meth.EH, myDiff25p.gr) 
PCobj <- methylKit::PCASamples(meth.diff, comp = c(1,2), adj.lim = c(0,0), obj.return = TRUE)

var_frac <- PCobj$sdev/sum(PCobj$sdev)
print(paste0("Total variance explained: ",signif(sum(var_frac[1:3]) * 100, 3), "%")) 
barplot(var_frac[1:20], main = "variance explained", 
        col = heat.colors(length(var_frac[1:20])))

# metadata3 <- metadata2[metadata2$tissue %in% "M",]
metadata.new <- metadata[metadata$SAMPLE_ID %in% meth.diff@sample.ids,]
metadata.new <- metadata.new[order(match(metadata.new$SAMPLE_ID, meth.diff@sample.ids)),]
identical(metadata.new$SAMPLE_ID, meth.diff@sample.ids)
identical(metadata.new$SAMPLE_ID, rownames(PCobj$x))

pop <- metadata.new$region

# pop <- metadata.new$`Date caught`

# colours5 <- adegenet::funky(4)
# colours5 <- c("#4d6b40","#adcf9f", "#c5d6e6", "#08306B")
colours5 <- c("#4d6b40","#adcf9f","#c5d6e6")


# plot.title <- "Toothfish methylation (1,316,799 CpG sites)"
plot.title <- "Toothfish methylation - EANt vs Heard (438 CpG sites)"


data <- data.frame(ind = rownames(PCobj$x),PCobj$x,  pop = pop)
data <- merge(data, aggregate(cbind(mean.x = PC1, mean.y = PC2) ~
                                pop, data, mean), by = "pop")

pca.plot <- ggplot2::ggplot(data, ggplot2::aes(PC1, PC2, color = pop)) +
  ggplot2::geom_point(size = 4) + 
  ggplot2::geom_point(ggplot2::aes(x = mean.x,
                                   y = mean.y, color = pop),
                      size = 7,
                      shape = 17) +
  ggplot2::scale_colour_manual(values = colours5) +
  ggplot2::scale_fill_manual(values = colours5) + 
  ggplot2::geom_hline(yintercept = 0) +
  ggplot2::geom_vline(xintercept = 0) + 
  ggplot2::labs(
    subtitle = paste0("Total variance explained: ",
                      signif(sum(var_frac[1:4]) * 100, 3), "%"),
    y = paste0("PC2; ", signif(var_frac[2] * 100, digits = 2), " %"),
    x = paste0("PC1; ", signif(var_frac[1] * 100, digits = 2), " %"),
    title = plot.title,
    caption = "") + 
  ggplot2::theme_minimal() +
  ggplot2::theme(
    axis.text = ggplot2::element_text(size = 15),
    axis.title.x = ggplot2::element_text(size = 20),
    axis.title.y = ggplot2::element_text(size = 20),
    legend.text = ggplot2::element_text(size = 15)
  )

print(pca.plot)
ggplot2::ggsave(pca.plot, bg = "white",
                filename = paste0(figure.path,"PCA_plot_diffMeth_ALL_Region_subset_EANT-HEARD_P12.png"), 
                width = 30, height = 15, units = "cm")



data <- data.frame(ind = rownames(PCobj$x),PCobj$x,  pop = pop)
data <- merge(data, aggregate(cbind(mean.x = PC1, mean.y = PC3) ~
                                pop, data, mean), by = "pop")

pca.plot <- ggplot2::ggplot(data, ggplot2::aes(PC1, PC3, color = pop)) +
  ggplot2::geom_point(size = 4) + 
  ggplot2::geom_point(ggplot2::aes(x = mean.x,
                                   y = mean.y, color = pop),
                      size = 7,
                      shape = 17) +
  ggplot2::scale_colour_manual(values = colours5) +
  ggplot2::scale_fill_manual(values = colours5) + 
  ggplot2::geom_hline(yintercept = 0) +
  ggplot2::geom_vline(xintercept = 0) + 
  ggplot2::labs(
    subtitle = paste0("Total variance explained: ",
                      signif(sum(var_frac[1:4]) * 100, 3), "%"),
    y = paste0("PC3; ", signif(var_frac[3] * 100, digits = 2), " %"),
    x = paste0("PC1; ", signif(var_frac[1] * 100, digits = 2), " %"),
    title = plot.title,
    caption = "") + 
  ggplot2::theme_minimal() +
  ggplot2::theme(
    axis.text = ggplot2::element_text(size = 15),
    axis.title.x = ggplot2::element_text(size = 20),
    axis.title.y = ggplot2::element_text(size = 20),
    legend.text = ggplot2::element_text(size = 15)
  )

print(pca.plot)
ggplot2::ggsave(pca.plot, bg = "white",
                filename = paste0(figure.path,"PCA_plot_diffMeth_ALL_Region_subset_EANT-HEARD_P13.png"), 
                width = 30, height = 15, units = "cm")




data <- data.frame(ind = rownames(PCobj$x),PCobj$x,  pop = pop)
data <- merge(data, aggregate(cbind(mean.x = PC2, mean.y = PC3) ~
                                pop, data, mean), by = "pop")

pca.plot <- ggplot2::ggplot(data, ggplot2::aes(PC2, PC3, color = pop)) +
  ggplot2::geom_point(size = 4) + 
  ggplot2::geom_point(ggplot2::aes(x = mean.x,
                                   y = mean.y, color = pop),
                      size = 7,
                      shape = 17) +
  ggplot2::scale_colour_manual(values = colours5) +
  ggplot2::scale_fill_manual(values = colours5) + 
  ggplot2::geom_hline(yintercept = 0) +
  ggplot2::geom_vline(xintercept = 0) + 
  ggplot2::labs(
    subtitle = paste0("Total variance explained: ",
                      signif(sum(var_frac[1:4]) * 100, 3), "%"),
    y = paste0("PC3; ", signif(var_frac[3] * 100, digits = 2), " %"),
    x = paste0("PC2; ", signif(var_frac[2] * 100, digits = 2), " %"),
    title = plot.title,
    caption = "") + 
  ggplot2::theme_minimal() +
  ggplot2::theme(
    axis.text = ggplot2::element_text(size = 15),
    axis.title.x = ggplot2::element_text(size = 20),
    axis.title.y = ggplot2::element_text(size = 20),
    legend.text = ggplot2::element_text(size = 15)
  )

print(pca.plot)
ggplot2::ggsave(pca.plot, bg = "white",
                filename = paste0(figure.path,"PCA_plot_diffMeth_ALL_Region_subset_EANT-HEARD_P23.png"), 
                width = 30, height = 15, units = "cm")
```


##### Plot CpG difference

```{r}
myDiff25p.gr <- as(myDiff25p, "GRanges")
hist(myDiff25p.gr$meth.diff)
min(myDiff25p.gr$meth.diff)#-48.01865 %
max(myDiff25p.gr$meth.diff)#42.58486 %

meth.diff <- methylKit::selectByOverlap(meth.EH, myDiff25p.gr)
length(unique(meth.diff$chr)) #27
meth.diff.df <- as.data.frame(t(methylKit::percMethylation(meth.diff)))
loc.names <- paste0(meth.diff$chr,"_",meth.diff$start)
colnames(meth.diff.df) <- loc.names

meth.diff.df$individual <- rownames(meth.diff.df) 
metadata.new <- metadata[metadata$SAMPLE_ID %in% meth.diff@sample.ids,]
metadata.new <- metadata.new[order(match(metadata.new$SAMPLE_ID, meth.diff@sample.ids)),]
identical(metadata.new$SAMPLE_ID, meth.diff@sample.ids)
identical(metadata.new$SAMPLE_ID,meth.diff.df$individual)
meth.diff.df$pop <- metadata.new$region

meth.diff.df.plot <- tidyr::gather(meth.diff.df, key = locus, 
                           value = perc_methylation, -individual, -pop,
                           factor_key = TRUE) %>%
  dplyr::mutate(chr = stringr::str_split_fixed(locus, 
                                               pattern = "_", 
                                               n = 2)[,1])

meth.diff.df.plot.locus <- meth.diff.df.plot %>% dplyr::group_by(chr,locus,pop) %>% 
  dplyr::summarise(mean_pop = mean(perc_methylation, na.rm = TRUE),
            .groups = 'drop')

meth.diff.df.plot.centroid <- meth.diff.df.plot %>% dplyr::group_by(chr,pop) %>% 
  dplyr::summarise(mean_pop = mean(perc_methylation, na.rm = TRUE),
            .groups = 'drop')

meth.diff.df.plot.label <- meth.diff.df.plot %>% dplyr::group_by(chr) %>% 
  dplyr::summarise(nmarkers = length(unique(locus)),
            .groups = 'drop')


ALL.plot <- ggplot2::ggplot() + 
  ggplot2::geom_violin(data = meth.diff.df.plot, 
                       ggplot2::aes(x = chr, y = perc_methylation)) + 
  ggplot2::geom_point(data = meth.diff.df.plot.centroid, 
                      ggplot2::aes(x = chr, y = mean_pop, colour = pop),
                      size = 7, shape = 5) + 
  ggplot2::geom_point(data = meth.diff.df.plot,
                      ggplot2::aes(x = chr, y = perc_methylation, colour = pop),
                      size = 1) +
  ggplot2::geom_text(data = meth.diff.df.plot.label,
                     ggplot2::aes(x = chr, y = 110, label = nmarkers),
                     size = 4) +
  ggplot2::geom_point(data = meth.diff.df.plot.locus, 
                      ggplot2::aes(x = chr, y = mean_pop, colour = pop),
                      size = 3, shape = 15) +
    ggplot2::theme(
        panel.background = ggplot2::element_blank(),
        panel.border = ggplot2::element_blank(),
        panel.grid = ggplot2::element_blank(),
        axis.line = ggplot2::element_line("gray25"),
        text = ggplot2::element_text(size = 12),
        axis.text = ggplot2::element_text(angle = 90,size = 10, colour = "gray25"),
        axis.title = ggplot2::element_text(size = 14, colour = "gray25"),
        legend.title = ggplot2::element_text(size = 14),
        legend.text = ggplot2::element_text(size = 10),
        legend.key = ggplot2::element_blank())

print(ALL.plot)

ggplot2::ggsave(plot = ALL.plot, 
                filename = paste0(figure.path,"TOA_ALLcpg_methDiff_EANT-HEARD.png"),
                width = 30, height = 15, units = "cm", device = "png")


diff.plot <- ggplot2::ggplot() +
  # ggplot2::geom_point(data = meth.diff.df.plot,
  #                     ggplot2::aes(x = pop, y = perc_methylation, colour = locus)) +
  ggplot2::geom_line(data = meth.diff.df.plot.locus,
                     ggplot2::aes(x = pop, y = mean_pop, colour = locus, group = locus)) +
  ggplot2::geom_point(data = meth.diff.df.plot.locus,
                     ggplot2::aes(x = pop, y = mean_pop, colour = locus, group = locus),
                     size = 5) +
  ggplot2::ylim(c(0,100)) +
  ggplot2::facet_wrap(~chr) +
  ggplot2::theme(legend.position = "none")
  # ggplot2::guides(colour = ggplot2::guide_legend(ncol = 1))
  # ggplot2::guides(colour = ggplot2::guide_legend(ncol = 5))

print(diff.plot)

ggplot2::ggsave(plot = diff.plot, 
                filename = paste0(figure.path,
                                  "TOA_ALLcpg_methDiff_EANT-HEARD_pw.png"
                                  ),
                width = 30, height = 25, units = "cm", device = "png")


table(meth.diff.df.plot$chr[!duplicated(meth.diff.df.plot$locus)])
meth.diff.df.plot2 <- meth.diff.df.plot[meth.diff.df.plot$chr == "JAAKFY010000002",]
meth.diff.df.plot.locus2 <- meth.diff.df.plot.locus[meth.diff.df.plot.locus$chr == "JAAKFY010000002",]
diff.plot2 <- ggplot2::ggplot() +
  ggplot2::geom_point(data = meth.diff.df.plot2,
                      ggplot2::aes(x = pop, y = perc_methylation, colour = locus)) +
  ggplot2::geom_line(data = meth.diff.df.plot.locus2,
                     ggplot2::aes(x = pop, y = mean_pop, colour = locus, group = locus)) +
  ggplot2::geom_point(data = meth.diff.df.plot.locus2,
                     ggplot2::aes(x = pop, y = mean_pop, colour = locus, group = locus),
                     size = 5) +
  ggplot2::ylim(c(0,100)) +
  ggplot2::facet_wrap(~locus) +
  ggplot2::theme(legend.position = "none")
  # ggplot2::guides(colour = ggplot2::guide_legend(ncol = 1))
  # ggplot2::guides(colour = ggplot2::guide_legend(ncol = 5))

print(diff.plot2)
ggplot2::ggsave(plot = diff.plot2, 
                filename = paste0(figure.path,
                                  "TOA_ALLcpg_methDiff_EANT-HEARD_pw_CHR02.png"
                                  ),
                width = 30, height = 25, units = "cm", device = "png")


meth.diff.df.plot2 <- meth.diff.df.plot[meth.diff.df.plot$chr == "JAAKFY010000014",]
meth.diff.df.plot.locus2 <- meth.diff.df.plot.locus[meth.diff.df.plot.locus$chr == "JAAKFY010000014",]
diff.plot14 <- ggplot2::ggplot() +
  ggplot2::geom_point(data = meth.diff.df.plot2,
                      ggplot2::aes(x = pop, y = perc_methylation, colour = locus)) +
  ggplot2::geom_line(data = meth.diff.df.plot.locus2,
                     ggplot2::aes(x = pop, y = mean_pop, colour = locus, group = locus)) +
  ggplot2::geom_point(data = meth.diff.df.plot.locus2,
                     ggplot2::aes(x = pop, y = mean_pop, colour = locus, group = locus),
                     size = 5) +
  ggplot2::ylim(c(0,100)) +
  ggplot2::facet_wrap(~locus) +
  ggplot2::theme(legend.position = "none")
  # ggplot2::guides(colour = ggplot2::guide_legend(ncol = 1))
  # ggplot2::guides(colour = ggplot2::guide_legend(ncol = 5))

print(diff.plot14)
ggplot2::ggsave(plot = diff.plot14, 
                filename = paste0(figure.path,
                                  "TOA_ALLcpg_methDiff_EANT-HEARD_pw_CHR14.png"
                                  ),
                width = 60, height = 45, units = "cm", device = "png")

meth.diff.df.plot2 <- meth.diff.df.plot[meth.diff.df.plot$chr == "JAAKFY010000018",]
meth.diff.df.plot.locus2 <- meth.diff.df.plot.locus[meth.diff.df.plot.locus$chr == "JAAKFY010000018",]
diff.plot18 <- ggplot2::ggplot() +
  ggplot2::geom_point(data = meth.diff.df.plot2,
                      ggplot2::aes(x = pop, y = perc_methylation, colour = locus)) +
  ggplot2::geom_line(data = meth.diff.df.plot.locus2,
                     ggplot2::aes(x = pop, y = mean_pop, colour = locus, group = locus)) +
  ggplot2::geom_point(data = meth.diff.df.plot.locus2,
                     ggplot2::aes(x = pop, y = mean_pop, colour = locus, group = locus),
                     size = 5) +
  ggplot2::ylim(c(0,100)) +
  ggplot2::facet_wrap(~locus) +
  ggplot2::theme(legend.position = "none")
  # ggplot2::guides(colour = ggplot2::guide_legend(ncol = 1))
  # ggplot2::guides(colour = ggplot2::guide_legend(ncol = 5))

print(diff.plot18)

ggplot2::ggsave(plot = diff.plot18, 
                filename = paste0(figure.path,
                                  "TOA_ALLcpg_methDiff_EANT-HEARD_pw_CHR18.png"
                                  ),
                width = 30, height = 25, units = "cm", device = "png")

meth.diff.df.plot2 <- meth.diff.df.plot[meth.diff.df.plot$chr == "JAAKFY010000026",]
meth.diff.df.plot.locus2 <- meth.diff.df.plot.locus[meth.diff.df.plot.locus$chr == "JAAKFY010000026",]
diff.plot26 <- ggplot2::ggplot() +
  ggplot2::geom_point(data = meth.diff.df.plot2,
                      ggplot2::aes(x = pop, y = perc_methylation, colour = locus)) +
  ggplot2::geom_line(data = meth.diff.df.plot.locus2,
                     ggplot2::aes(x = pop, y = mean_pop, colour = locus, group = locus)) +
  ggplot2::geom_point(data = meth.diff.df.plot.locus2,
                     ggplot2::aes(x = pop, y = mean_pop, colour = locus, group = locus),
                     size = 5) +
  ggplot2::ylim(c(0,100)) +
  ggplot2::facet_wrap(~locus) +
  ggplot2::theme(legend.position = "none")
  # ggplot2::guides(colour = ggplot2::guide_legend(ncol = 1))
  # ggplot2::guides(colour = ggplot2::guide_legend(ncol = 5))

print(diff.plot26)

ggplot2::ggsave(plot = diff.plot26, 
                filename = paste0(figure.path,
                                  "TOA_ALLcpg_methDiff_EANT-HEARD_pw_CHR26.png"
                                  ),
                width = 30, height = 25, units = "cm", device = "png")

```


##### Heatmap

```{r}
meth.diff.df <- as.data.frame(t(methylKit::percMethylation(meth.diff)))
loc.names <- paste0(meth.diff$chr,"_",meth.diff$start)
colnames(meth.diff.df) <- loc.names

meth.diff.df$individual <- rownames(meth.diff.df) 
metadata.new <- metadata[metadata$SAMPLE_ID %in% meth.diff@sample.ids,]
metadata.new <- metadata.new[order(match(metadata.new$SAMPLE_ID, meth.diff@sample.ids)),]
identical(metadata.new$SAMPLE_ID, meth.diff@sample.ids)
identical(metadata.new$SAMPLE_ID,meth.diff.df$individual)
meth.diff.df$pop <- metadata.new$region


heat.plot.data <- meth.diff.df %>%
  tibble::as_tibble() %>%
  tidyr::gather(key = "CpG", value = "Methylation_percentage", -individual, -pop,
                factor_key = FALSE) %>%
  dplyr::group_by(pop, CpG) %>%
  dplyr::mutate(CpG_mean = mean(Methylation_percentage, na.rm = TRUE)) %>%
  dplyr::arrange(pop, individual, CpG) %>%
  dplyr::mutate(individual = factor(individual, levels = unique(individual))) %>%
  dplyr::mutate(CpG = factor(CpG, levels = unique(CpG)))
cpgs <- heat.plot.data$CpG[heat.plot.data$individual == heat.plot.data$individual[1]]
cpgs_mean <- heat.plot.data$CpG_mean[heat.plot.data$individual == heat.plot.data$individual[1]]
cpgs <- cpgs[order(cpgs_mean)]
heat.plot.data$CpG <- factor(heat.plot.data$CpG, levels = unique(cpgs))
heat.plot.data <- heat.plot.data[order(heat.plot.data$CpG),]

heat.plot <- ggplot2::ggplot(data = heat.plot.data) + 
    ggplot2::geom_tile(ggplot2::aes(individual, y = CpG, fill = Methylation_percentage)) +
    ggplot2::scale_fill_gradient2(low = "blue", mid = "yellow", high = "red", midpoint = 50) +
    
    ggnewscale::new_scale_fill() +
    # ggplot2::geom_tile(ggplot2::aes(individual, y = "pop", fill = pop, height = 10)) +
    # ggplot2::scale_fill_manual(values = c("#4d6b40","#adcf9f", "#c5d6e6", "#08306B")) +
    # ggplot2::scale_fill_manual(values = c( "#c5d6e6", "#08306B")) +
    ggplot2::scale_y_discrete(drop = FALSE) +
    ggplot2::theme(legend.position = "top",
                   axis.title.y = ggplot2::element_blank(),
          axis.text.x = ggplot2::element_text(face = "bold", color = "black", 
                                        size = 10, angle = 90, hjust = 1, vjust = 0))
print(heat.plot)

ggplot2::ggsave(heat.plot, bg = "white",
                filename = paste0(figure.path,"TOA_ALLcpg_methDiff_EANT_HEARD_HEATMAP.png"), 
                width = 45, height = 30, units = "cm")
```


#### PCA - HYPER

```{r}
myDiff25p.hyper.gr <- as(myDiff25p.hyper, "GRanges")
meth.diff <- methylKit::selectByOverlap(meth.EH, myDiff25p.hyper.gr) 

methylKit::PCASamples(meth.diff, screeplot = TRUE)
dev.print(file = paste0(figure.path,
                        "PCA_plot_diffMeth_HYPER_Region_subset_EANT-HEARD_scree.png"),
          device = png, res = 300,width = 30,height = 15,units = "cm")

methylKit::PCASamples(meth.diff, comp = c(1,2))
dev.print(file = paste0(figure.path,
                        "PCA_plot_diffMeth_HYPER_Region_subset_EANT-HEARD_PC12.png"
                        ),
          device = png, res = 300,width = 30,height = 15,units = "cm")

methylKit::PCASamples(meth.diff, comp = c(1,3))
dev.print(file = paste0(figure.path,
                        "PCA_plot_diffMeth_HYPER_Region_subset_EANT-HEARD_PC13.png"
                        ),
          device = png, res = 300,width = 30,height = 15,units = "cm")

methylKit::PCASamples(meth.diff, comp = c(2,3))
dev.print(file = paste0(figure.path,
                        "PCA_plot_diffMeth_HYPER_Region_subset_EANT-HEARD_PC23.png"
                        ),
          device = png, res = 300,width = 30,height = 15,units = "cm")
```


##### Plot CpG difference

```{r}
myDiff25p.hyper.gr <- as(myDiff25p.hyper, "GRanges")
hist(myDiff25p.hyper.gr$meth.diff)
max(myDiff25p.hyper.gr$meth.diff) #42.5 %

meth.diff <- methylKit::selectByOverlap(meth.EH, myDiff25p.hyper.gr)
length(unique(meth.diff$chr)) #25
meth.diff.df <- as.data.frame(t(methylKit::percMethylation(meth.diff)))
loc.names <- paste0(meth.diff$chr,"_",meth.diff$start)
colnames(meth.diff.df) <- loc.names

meth.diff.df$individual <- rownames(meth.diff.df) 
identical(metadata3$SAMPLE_ID,meth.diff.df$individual)
meth.diff.df$pop <- metadata3$region

meth.diff.df.plot <- tidyr::gather(meth.diff.df, key = locus, 
                           value = perc_methylation, -individual, -pop,
                           factor_key = TRUE) %>%
  dplyr::mutate(chr = stringr::str_split_fixed(locus, 
                                               pattern = "_", 
                                               n = 2)[,1])
meth.diff.df.plot.locus <- meth.diff.df.plot %>% dplyr::group_by(chr,locus,pop) %>% 
  dplyr::summarise(mean_pop = mean(perc_methylation, na.rm = TRUE),
            .groups = 'drop')

meth.diff.df.plot.centroid <- meth.diff.df.plot %>% dplyr::group_by(chr,pop) %>% 
  dplyr::summarise(mean_pop = mean(perc_methylation, na.rm = TRUE),
            .groups = 'drop')

meth.diff.df.plot.label <- meth.diff.df.plot %>% dplyr::group_by(chr) %>% 
  dplyr::summarise(nmarkers = length(unique(locus)),
            .groups = 'drop')


HYPER.plot <- ggplot2::ggplot() + 
  ggplot2::geom_violin(data = meth.diff.df.plot, 
                       ggplot2::aes(x = chr, y = perc_methylation)) + 
  ggplot2::geom_point(data = meth.diff.df.plot.centroid, 
                      ggplot2::aes(x = chr, y = mean_pop, colour = pop),
                      size = 7, shape = 5) + 
  ggplot2::geom_point(data = meth.diff.df.plot,
                      ggplot2::aes(x = chr, y = perc_methylation, colour = pop),
                      size = 1) +
  ggplot2::geom_text(data = meth.diff.df.plot.label,
                     ggplot2::aes(x = chr, y = 110, label = nmarkers),
                     size = 4) +
  ggplot2::geom_point(data = meth.diff.df.plot.locus, 
                      ggplot2::aes(x = chr, y = mean_pop, colour = pop),
                      size = 3, shape = 15) +
    ggplot2::theme(
        panel.background = ggplot2::element_blank(),
        panel.border = ggplot2::element_blank(),
        panel.grid = ggplot2::element_blank(),
        axis.line = ggplot2::element_line("gray25"),
        text = ggplot2::element_text(size = 12),
        axis.text = ggplot2::element_text(angle = 90,size = 10, colour = "gray25"),
        axis.title = ggplot2::element_text(size = 14, colour = "gray25"),
        legend.title = ggplot2::element_text(size = 14),
        legend.text = ggplot2::element_text(size = 10),
        legend.key = ggplot2::element_blank())

print(HYPER.plot)

ggplot2::ggsave(plot = HYPER.plot, 
                filename = paste0(figure.path,"TOA_HYPERcpg_methDiff_EANT-HEARD.png"),
                width = 30, height = 15, units = "cm", device = "png")
```

#### PCA - HYPO

```{r}
myDiff25p.hypo.gr <- as(myDiff25p.hypo, "GRanges")
meth.diff <- methylKit::selectByOverlap(meth.EH, myDiff25p.hypo.gr) 

methylKit::PCASamples(meth.diff, screeplot = TRUE)
dev.print(file = paste0(figure.path,
                        "PCA_plot_diffMeth_HYPO_Region_subset_EANT-HEARD_scree.png"),
          device = png, res = 300,width = 30,height = 15,units = "cm")

methylKit::PCASamples(meth.diff, comp = c(1,2))
dev.print(file = paste0(figure.path,
                        "PCA_plot_diffMeth_HYPO_Region_subset_EANT-HEARD_PC12.png"
                        ),
          device = png, res = 300,width = 30,height = 15,units = "cm")

methylKit::PCASamples(meth.diff, comp = c(1,3))
dev.print(file = paste0(figure.path,
                        "PCA_plot_diffMeth_HYPO_Region_subset_EANT-HEARD_PC13.png"
                        ),
          device = png, res = 300,width = 30,height = 15,units = "cm")

methylKit::PCASamples(meth.diff, comp = c(2,3))
dev.print(file = paste0(figure.path,
                        "PCA_plot_diffMeth_HYPO_Region_subset_EANT-HEARD_PC23.png"
                        ),
          device = png, res = 300,width = 30,height = 15,units = "cm")
```


##### Plot CpG difference

```{r}
myDiff25p.hypo.gr <- as(myDiff25p.hypo, "GRanges")
hist(myDiff25p.hypo.gr$meth.diff)
min(myDiff25p.hypo.gr$meth.diff)
max(myDiff25p.hypo.gr$meth.diff)

meth.diff <- methylKit::selectByOverlap(meth.EH, myDiff25p.hypo.gr)
length(unique(meth.diff$chr)) #25
meth.diff.df <- as.data.frame(t(methylKit::percMethylation(meth.diff)))
loc.names <- paste0(meth.diff$chr,"_",meth.diff$start)
colnames(meth.diff.df) <- loc.names

meth.diff.df$individual <- rownames(meth.diff.df) 
identical(metadata3$SAMPLE_ID, meth.diff.df$individual)
meth.diff.df$pop <- metadata3$region

meth.diff.df.plot <- tidyr::gather(meth.diff.df, key = locus, 
                           value = perc_methylation, -individual, -pop,
                           factor_key = TRUE) %>%
  dplyr::mutate(chr = stringr::str_split_fixed(locus, 
                                               pattern = "_", 
                                               n = 2)[,1])
meth.diff.df.plot.locus <- meth.diff.df.plot %>% dplyr::group_by(chr,locus,pop) %>% 
  dplyr::summarise(mean_pop = mean(perc_methylation, na.rm = TRUE),
            .groups = 'drop')

meth.diff.df.plot.centroid <- meth.diff.df.plot %>% dplyr::group_by(chr,pop) %>% 
  dplyr::summarise(mean_pop = mean(perc_methylation, na.rm = TRUE),
            .groups = 'drop')

meth.diff.df.plot.label <- meth.diff.df.plot %>% dplyr::group_by(chr) %>% 
  dplyr::summarise(nmarkers = length(unique(locus)),
            .groups = 'drop')


HYPO.plot <- ggplot2::ggplot() + 
  ggplot2::geom_violin(data = meth.diff.df.plot, 
                       ggplot2::aes(x = chr, y = perc_methylation)) + 
  ggplot2::geom_point(data = meth.diff.df.plot.centroid, 
                      ggplot2::aes(x = chr, y = mean_pop, colour = pop),
                      size = 7, shape = 5) + 
  ggplot2::geom_point(data = meth.diff.df.plot,
                      ggplot2::aes(x = chr, y = perc_methylation, colour = pop),
                      size = 1) +
  ggplot2::geom_text(data = meth.diff.df.plot.label,
                     ggplot2::aes(x = chr, y = 110, label = nmarkers),
                     size = 4) +
  ggplot2::geom_point(data = meth.diff.df.plot.locus, 
                      ggplot2::aes(x = chr, y = mean_pop, colour = pop),
                      size = 3, shape = 15) +
    ggplot2::theme(
        panel.background = ggplot2::element_blank(),
        panel.border = ggplot2::element_blank(),
        panel.grid = ggplot2::element_blank(),
        axis.line = ggplot2::element_line("gray25"),
        text = ggplot2::element_text(size = 12),
        axis.text = ggplot2::element_text(angle = 90,size = 10, colour = "gray25"),
        axis.title = ggplot2::element_text(size = 14, colour = "gray25"),
        legend.title = ggplot2::element_text(size = 14),
        legend.text = ggplot2::element_text(size = 10),
        legend.key = ggplot2::element_blank())

print(HYPO.plot)

ggplot2::ggsave(plot = HYPO.plot, 
                filename = paste0(figure.path,"TOA_HYPOcpg_methDiff_EANT-HEARD.png"),
                width = 30, height = 15, units = "cm", device = "png")
```


### EANT4.1 - HEARD
#### Prepare data

```{r}
print(load(paste0(rdata.path,"Toothfish_metadata.Rdata")))

metadata3 <- metadata2[metadata2$region %in% c("58-4-1", "58-5-2"),]
table(metadata3$region)
print(load(paste0(rdata.path,"Toothfish_methylation_objects_raw.Rdata")))

filtered.myobj3 <- methylKit::reorganize(
  methylObj = filtered.myobj,
  sample.ids = metadata3$SAMPLE_ID,
  treatment = as.numeric(factor(metadata3$region)),
  chunk.size = 1e+06,
  save.db = TRUE,
  suffix = "subset_EANT1-HEARD_raw"
)

# filtered.myobj3@treatment <- as.numeric(factor(metadata3$region))
filtered.myobj3 <- methylKit::filterByCoverage(filtered.myobj3,lo.count = 10,lo.perc = NULL,
                                      hi.count = NULL, hi.perc = 99)

meth.EH1 <- methylKit::unite(filtered.myobj3, destrand = FALSE,
                             save.db = TRUE,
                             suffix = "subset_EANT1-HEARD",
                             min.per.group = 16L, #missing data allowed
                             mc.cores = 25)
#1,087,960   CpGs after coverage filter

# save(meth.HR,meth.EANT,meth.EH,meth.EH1
#      file = paste0(rdata.path,
#                    "Toothfish_subset_pairwise_methylation_objects_raw.Rdata"))
```

#### diffMethPerChr

```{r}
print(load(paste0(rdata.path,"Toothfish_subset_pairwise_methylation_objects_raw.Rdata")))
metadata3 <- metadata2[metadata2$region %in% c("58-4-1", "58-5-2"),]
covariates <- dplyr::select(metadata3, Sex, mat, Age)
myDiff <- methylKit::calculateDiffMeth(meth.EH1,
                                       covariates = covariates,
                                       adjust = "fdr",
                                       chunk.size = 1e+06,
                                       mc.cores = 10)
meth.diff.chr <- methylKit::diffMethPerChr(myDiff,plot = TRUE, 
                                           qvalue.cutoff = 0.01, meth.cutoff = 25)
myDiff25p.hyper <- methylKit::getMethylDiff(myDiff, difference = 25,qvalue = 0.01, type = "hyper") #483   
myDiff25p.hypo <- methylKit::getMethylDiff(myDiff, difference = 25,qvalue = 0.01, type = "hypo") #338  
myDiff25p <- methylKit::getMethylDiff(myDiff, difference = 25, qvalue = 0.01)

# save(myDiff,myDiff25p, myDiff25p.hyper, myDiff25p.hypo, 
#      file = paste0(rdata.path,
#                    "Toothfish_DMS_Region_subset_EANT1-HEARD.Rdata"))

myDiff.df <- as.data.frame(as(myDiff, "GRanges"))
myDiff25p.df <- as.data.frame(as(myDiff25p, "GRanges"))
volcano.plot <- ggplot2::ggplot() +
  ggplot2::geom_point(data = myDiff.df, ggplot2::aes(x = meth.diff, 
                                                  y = -log10(pvalue))) +
  ggplot2::geom_point(data = myDiff25p.df, ggplot2::aes(x = meth.diff, 
                                                  y = -log10(pvalue)),
                      colour = "red")
print(volcano.plot)

ggplot2::ggsave(volcano.plot, bg = "white",
                filename = paste0(figure.path,
                                  "TOA_ALL_DMS25_VOLCANO_methylkit_regions_subset_EANT1-HEARD.png"
                                  ), 
                width = 30, height = 15, units = "cm")
```

#### Load data

```{r}
print(load(paste0(rdata.path,"Toothfish_DMS_Region_subset_EANT1-HEARD.Rdata")))

myDiff25p.gr <- as(myDiff25p, "GRanges")
meth.diff <- methylKit::selectByOverlap(meth.EH1, myDiff25p.gr) 

# DMS <- data.frame(DMS = NULL, region_pw = NULL)
DMS.EH1 <- data.frame(DMS = paste0(meth.diff$chr,"_", meth.diff$start),
                  region_pw = "EANT4.1_HEARD")
DMS <- rbind(DMS, DMS.EH1)

# save(DMS, DMS.HR,DMS.EANT,DMS.EH,DMS.EH1,
#      file = paste0(rdata.path,"Toothfish_DMS_CpGsites.Rdata"))
```

#### PCA - ALL

```{r}
print(load(paste0(rdata.path,"Toothfish_DMS_Region_subset_EANT1-HEARD.Rdata")))

myDiff25p.gr <- as(myDiff25p, "GRanges")
meth.diff <- methylKit::selectByOverlap(meth.EH1, myDiff25p.gr) 

PCobj <- methylKit::PCASamples(meth.diff, comp = c(1,2), adj.lim = c(0,0), obj.return = TRUE)

var_frac <- PCobj$sdev/sum(PCobj$sdev)
print(paste0("Total variance explained: ",signif(sum(var_frac[1:3]) * 100, 3), "%")) 
barplot(var_frac[1:20], main = "variance explained", 
        col = heat.colors(length(var_frac[1:20])))

# metadata3 <- metadata2[metadata2$tissue %in% "M",]
metadata.new <- metadata[metadata$SAMPLE_ID %in% meth.diff@sample.ids,]
metadata.new <- metadata.new[order(match(metadata.new$SAMPLE_ID, meth.diff@sample.ids)),]
identical(metadata.new$SAMPLE_ID, meth.diff@sample.ids)
identical(metadata.new$SAMPLE_ID, rownames(PCobj$x))

pop <- metadata.new$region

# pop <- metadata.new$`Date caught`

# colours5 <- adegenet::funky(4)
# colours5 <- c("#4d6b40","#adcf9f", "#c5d6e6", "#08306B")
colours5 <- c("#4d6b40","#c5d6e6")


# plot.title <- "Toothfish methylation (1,316,799 CpG sites)"
plot.title <- "Toothfish methylation - EANT1 vs Heard (804 CpG sites)"


data <- data.frame(ind = rownames(PCobj$x),PCobj$x,  pop = pop)
data <- merge(data, aggregate(cbind(mean.x = PC1, mean.y = PC2) ~
                                pop, data, mean), by = "pop")

pca.plot <- ggplot2::ggplot(data, ggplot2::aes(PC1, PC2, color = pop)) +
  ggplot2::geom_point(size = 4) + 
  ggplot2::geom_point(ggplot2::aes(x = mean.x,
                                   y = mean.y, color = pop),
                      size = 7,
                      shape = 17) +
  ggplot2::scale_colour_manual(values = colours5) +
  ggplot2::scale_fill_manual(values = colours5) + 
  ggplot2::geom_hline(yintercept = 0) +
  ggplot2::geom_vline(xintercept = 0) + 
  ggplot2::labs(
    subtitle = paste0("Total variance explained: ",
                      signif(sum(var_frac[1:4]) * 100, 3), "%"),
    y = paste0("PC2; ", signif(var_frac[2] * 100, digits = 2), " %"),
    x = paste0("PC1; ", signif(var_frac[1] * 100, digits = 2), " %"),
    title = plot.title,
    caption = "") + 
  ggplot2::theme_minimal() +
  ggplot2::theme(
    axis.text = ggplot2::element_text(size = 15),
    axis.title.x = ggplot2::element_text(size = 20),
    axis.title.y = ggplot2::element_text(size = 20),
    legend.text = ggplot2::element_text(size = 15)
  )

print(pca.plot)
ggplot2::ggsave(pca.plot, bg = "white",
                filename = paste0(figure.path,"PCA_plot_diffMeth_ALL_Region_subset_EANT1-HEARD_P12.png"), 
                width = 30, height = 15, units = "cm")



data <- data.frame(ind = rownames(PCobj$x),PCobj$x,  pop = pop)
data <- merge(data, aggregate(cbind(mean.x = PC1, mean.y = PC3) ~
                                pop, data, mean), by = "pop")

pca.plot <- ggplot2::ggplot(data, ggplot2::aes(PC1, PC3, color = pop)) +
  ggplot2::geom_point(size = 4) + 
  ggplot2::geom_point(ggplot2::aes(x = mean.x,
                                   y = mean.y, color = pop),
                      size = 7,
                      shape = 17) +
  ggplot2::scale_colour_manual(values = colours5) +
  ggplot2::scale_fill_manual(values = colours5) + 
  ggplot2::geom_hline(yintercept = 0) +
  ggplot2::geom_vline(xintercept = 0) + 
  ggplot2::labs(
    subtitle = paste0("Total variance explained: ",
                      signif(sum(var_frac[1:4]) * 100, 3), "%"),
    y = paste0("PC3; ", signif(var_frac[3] * 100, digits = 2), " %"),
    x = paste0("PC1; ", signif(var_frac[1] * 100, digits = 2), " %"),
    title = plot.title,
    caption = "") + 
  ggplot2::theme_minimal() +
  ggplot2::theme(
    axis.text = ggplot2::element_text(size = 15),
    axis.title.x = ggplot2::element_text(size = 20),
    axis.title.y = ggplot2::element_text(size = 20),
    legend.text = ggplot2::element_text(size = 15)
  )

print(pca.plot)
ggplot2::ggsave(pca.plot, bg = "white",
                filename = paste0(figure.path,"PCA_plot_diffMeth_ALL_Region_subset_EANT1-HEARD_P13.png"), 
                width = 30, height = 15, units = "cm")




data <- data.frame(ind = rownames(PCobj$x),PCobj$x,  pop = pop)
data <- merge(data, aggregate(cbind(mean.x = PC2, mean.y = PC3) ~
                                pop, data, mean), by = "pop")

pca.plot <- ggplot2::ggplot(data, ggplot2::aes(PC2, PC3, color = pop)) +
  ggplot2::geom_point(size = 4) + 
  ggplot2::geom_point(ggplot2::aes(x = mean.x,
                                   y = mean.y, color = pop),
                      size = 7,
                      shape = 17) +
  ggplot2::scale_colour_manual(values = colours5) +
  ggplot2::scale_fill_manual(values = colours5) + 
  ggplot2::geom_hline(yintercept = 0) +
  ggplot2::geom_vline(xintercept = 0) + 
  ggplot2::labs(
    subtitle = paste0("Total variance explained: ",
                      signif(sum(var_frac[1:4]) * 100, 3), "%"),
    y = paste0("PC3; ", signif(var_frac[3] * 100, digits = 2), " %"),
    x = paste0("PC2; ", signif(var_frac[2] * 100, digits = 2), " %"),
    title = plot.title,
    caption = "") + 
  ggplot2::theme_minimal() +
  ggplot2::theme(
    axis.text = ggplot2::element_text(size = 15),
    axis.title.x = ggplot2::element_text(size = 20),
    axis.title.y = ggplot2::element_text(size = 20),
    legend.text = ggplot2::element_text(size = 15)
  )

print(pca.plot)
ggplot2::ggsave(pca.plot, bg = "white",
                filename = paste0(figure.path,"PCA_plot_diffMeth_ALL_Region_subset_EANT1-HEARD_P23.png"), 
                width = 30, height = 15, units = "cm")
```


##### Plot CpG difference

```{r}
myDiff25p.gr <- as(myDiff25p, "GRanges")
hist(myDiff25p.gr$meth.diff)
min(myDiff25p.gr$meth.diff)#-43.27445 %
max(myDiff25p.gr$meth.diff)#51.8423 %


meth.diff <- methylKit::selectByOverlap(meth.EH1, myDiff25p.gr)
length(unique(meth.diff$chr)) #29
meth.diff.df <- as.data.frame(t(methylKit::percMethylation(meth.diff)))
loc.names <- paste0(meth.diff$chr,"_",meth.diff$start)
colnames(meth.diff.df) <- loc.names

meth.diff.df$individual <- rownames(meth.diff.df) 
metadata.new <- metadata[metadata$SAMPLE_ID %in% meth.diff@sample.ids,]
metadata.new <- metadata.new[order(match(metadata.new$SAMPLE_ID, meth.diff@sample.ids)),]
identical(metadata.new$SAMPLE_ID, meth.diff@sample.ids)
identical(metadata.new$SAMPLE_ID,meth.diff.df$individual)
meth.diff.df$pop <- metadata.new$region

meth.diff.df.plot <- tidyr::gather(meth.diff.df, key = locus, 
                           value = perc_methylation, -individual, -pop,
                           factor_key = TRUE) %>%
  dplyr::mutate(chr = stringr::str_split_fixed(locus, 
                                               pattern = "_", 
                                               n = 2)[,1])

meth.diff.df.plot.locus <- meth.diff.df.plot %>% dplyr::group_by(chr,locus,pop) %>% 
  dplyr::summarise(mean_pop = mean(perc_methylation, na.rm = TRUE),
            .groups = 'drop')

meth.diff.df.plot.centroid <- meth.diff.df.plot %>% dplyr::group_by(chr,pop) %>% 
  dplyr::summarise(mean_pop = mean(perc_methylation, na.rm = TRUE),
            .groups = 'drop')

meth.diff.df.plot.label <- meth.diff.df.plot %>% dplyr::group_by(chr) %>% 
  dplyr::summarise(nmarkers = length(unique(locus)),
            .groups = 'drop')


ALL.plot <- ggplot2::ggplot() + 
  ggplot2::geom_violin(data = meth.diff.df.plot, 
                       ggplot2::aes(x = chr, y = perc_methylation)) + 
  ggplot2::geom_point(data = meth.diff.df.plot.centroid, 
                      ggplot2::aes(x = chr, y = mean_pop, colour = pop),
                      size = 7, shape = 5) + 
  ggplot2::geom_point(data = meth.diff.df.plot,
                      ggplot2::aes(x = chr, y = perc_methylation, colour = pop),
                      size = 1) +
  ggplot2::geom_text(data = meth.diff.df.plot.label,
                     ggplot2::aes(x = chr, y = 110, label = nmarkers),
                     size = 4) +
  ggplot2::geom_point(data = meth.diff.df.plot.locus, 
                      ggplot2::aes(x = chr, y = mean_pop, colour = pop),
                      size = 3, shape = 15) +
    ggplot2::theme(
        panel.background = ggplot2::element_blank(),
        panel.border = ggplot2::element_blank(),
        panel.grid = ggplot2::element_blank(),
        axis.line = ggplot2::element_line("gray25"),
        text = ggplot2::element_text(size = 12),
        axis.text = ggplot2::element_text(angle = 90,size = 10, colour = "gray25"),
        axis.title = ggplot2::element_text(size = 14, colour = "gray25"),
        legend.title = ggplot2::element_text(size = 14),
        legend.text = ggplot2::element_text(size = 10),
        legend.key = ggplot2::element_blank())

print(ALL.plot)

ggplot2::ggsave(plot = ALL.plot, 
                filename = paste0(figure.path,"TOA_ALLcpg_methDiff_EANT1-HEARD.png"),
                width = 30, height = 15, units = "cm", device = "png")



diff.plot <- ggplot2::ggplot() +
  # ggplot2::geom_point(data = meth.diff.df.plot,
  #                     ggplot2::aes(x = pop, y = perc_methylation, colour = locus)) +
  ggplot2::geom_line(data = meth.diff.df.plot.locus,
                     ggplot2::aes(x = pop, y = mean_pop, colour = locus, group = locus)) +
  ggplot2::geom_point(data = meth.diff.df.plot.locus,
                     ggplot2::aes(x = pop, y = mean_pop, colour = locus, group = locus),
                     size = 5) +
  ggplot2::ylim(c(0,100)) +
  ggplot2::facet_wrap(~chr) +
  ggplot2::theme(legend.position = "none")
  # ggplot2::guides(colour = ggplot2::guide_legend(ncol = 1))
  # ggplot2::guides(colour = ggplot2::guide_legend(ncol = 5))

print(diff.plot)

ggplot2::ggsave(plot = diff.plot, 
                filename = paste0(figure.path,
                                  "TOA_ALLcpg_methDiff_EANT1-HEARD_pw.png"
                                  ),
                width = 30, height = 25, units = "cm", device = "png")


table(meth.diff.df.plot$chr[!duplicated(meth.diff.df.plot$locus)])
meth.diff.df.plot2 <- meth.diff.df.plot[meth.diff.df.plot$chr == "JAAKFY010000002",]
meth.diff.df.plot.locus2 <- meth.diff.df.plot.locus[meth.diff.df.plot.locus$chr == "JAAKFY010000002",]
diff.plot2 <- ggplot2::ggplot() +
  ggplot2::geom_point(data = meth.diff.df.plot2,
                      ggplot2::aes(x = pop, y = perc_methylation, colour = locus)) +
  ggplot2::geom_line(data = meth.diff.df.plot.locus2,
                     ggplot2::aes(x = pop, y = mean_pop, colour = locus, group = locus)) +
  ggplot2::geom_point(data = meth.diff.df.plot.locus2,
                     ggplot2::aes(x = pop, y = mean_pop, colour = locus, group = locus),
                     size = 5) +
  ggplot2::ylim(c(0,100)) +
  ggplot2::facet_wrap(~locus) +
  ggplot2::theme(legend.position = "none")
  # ggplot2::guides(colour = ggplot2::guide_legend(ncol = 1))
  # ggplot2::guides(colour = ggplot2::guide_legend(ncol = 5))

print(diff.plot2)
ggplot2::ggsave(plot = diff.plot2, 
                filename = paste0(figure.path,
                                  "TOA_ALLcpg_methDiff_EANT1-HEARD_pw_CHR02.png"
                                  ),
                width = 30, height = 25, units = "cm", device = "png")


meth.diff.df.plot2 <- meth.diff.df.plot[meth.diff.df.plot$chr == "JAAKFY010000014",]
meth.diff.df.plot.locus2 <- meth.diff.df.plot.locus[meth.diff.df.plot.locus$chr == "JAAKFY010000014",]
diff.plot14 <- ggplot2::ggplot() +
  ggplot2::geom_point(data = meth.diff.df.plot2,
                      ggplot2::aes(x = pop, y = perc_methylation, colour = locus)) +
  ggplot2::geom_line(data = meth.diff.df.plot.locus2,
                     ggplot2::aes(x = pop, y = mean_pop, colour = locus, group = locus)) +
  ggplot2::geom_point(data = meth.diff.df.plot.locus2,
                     ggplot2::aes(x = pop, y = mean_pop, colour = locus, group = locus),
                     size = 5) +
  ggplot2::ylim(c(0,100)) +
  ggplot2::facet_wrap(~locus) +
  ggplot2::theme(legend.position = "none")
  # ggplot2::guides(colour = ggplot2::guide_legend(ncol = 1))
  # ggplot2::guides(colour = ggplot2::guide_legend(ncol = 5))

print(diff.plot14)
ggplot2::ggsave(plot = diff.plot14, 
                filename = paste0(figure.path,
                                  "TOA_ALLcpg_methDiff_EANT1-HEARD_pw_CHR14.png"
                                  ),
                width = 60, height = 45, units = "cm", device = "png")

meth.diff.df.plot2 <- meth.diff.df.plot[meth.diff.df.plot$chr == "JAAKFY010000018",]
meth.diff.df.plot.locus2 <- meth.diff.df.plot.locus[meth.diff.df.plot.locus$chr == "JAAKFY010000018",]
diff.plot18 <- ggplot2::ggplot() +
  ggplot2::geom_point(data = meth.diff.df.plot2,
                      ggplot2::aes(x = pop, y = perc_methylation, colour = locus)) +
  ggplot2::geom_line(data = meth.diff.df.plot.locus2,
                     ggplot2::aes(x = pop, y = mean_pop, colour = locus, group = locus)) +
  ggplot2::geom_point(data = meth.diff.df.plot.locus2,
                     ggplot2::aes(x = pop, y = mean_pop, colour = locus, group = locus),
                     size = 5) +
  ggplot2::ylim(c(0,100)) +
  ggplot2::facet_wrap(~locus) +
  ggplot2::theme(legend.position = "none")
  # ggplot2::guides(colour = ggplot2::guide_legend(ncol = 1))
  # ggplot2::guides(colour = ggplot2::guide_legend(ncol = 5))

print(diff.plot18)

ggplot2::ggsave(plot = diff.plot18, 
                filename = paste0(figure.path,
                                  "TOA_ALLcpg_methDiff_EANT1-HEARD_pw_CHR18.png"
                                  ),
                width = 30, height = 25, units = "cm", device = "png")

meth.diff.df.plot2 <- meth.diff.df.plot[meth.diff.df.plot$chr == "JAAKFY010000026",]
meth.diff.df.plot.locus2 <- meth.diff.df.plot.locus[meth.diff.df.plot.locus$chr == "JAAKFY010000026",]
diff.plot26 <- ggplot2::ggplot() +
  ggplot2::geom_point(data = meth.diff.df.plot2,
                      ggplot2::aes(x = pop, y = perc_methylation, colour = locus)) +
  ggplot2::geom_line(data = meth.diff.df.plot.locus2,
                     ggplot2::aes(x = pop, y = mean_pop, colour = locus, group = locus)) +
  ggplot2::geom_point(data = meth.diff.df.plot.locus2,
                     ggplot2::aes(x = pop, y = mean_pop, colour = locus, group = locus),
                     size = 5) +
  ggplot2::ylim(c(0,100)) +
  ggplot2::facet_wrap(~locus) +
  ggplot2::theme(legend.position = "none")
  # ggplot2::guides(colour = ggplot2::guide_legend(ncol = 1))
  # ggplot2::guides(colour = ggplot2::guide_legend(ncol = 5))

print(diff.plot26)

ggplot2::ggsave(plot = diff.plot26, 
                filename = paste0(figure.path,
                                  "TOA_ALLcpg_methDiff_EANT1-HEARD_pw_CHR26.png"
                                  ),
                width = 30, height = 25, units = "cm", device = "png")
```



##### Heatmap

```{r}
meth.diff.df <- as.data.frame(t(methylKit::percMethylation(meth.diff)))
loc.names <- paste0(meth.diff$chr,"_",meth.diff$start)
colnames(meth.diff.df) <- loc.names

meth.diff.df$individual <- rownames(meth.diff.df) 
metadata.new <- metadata[metadata$SAMPLE_ID %in% meth.diff@sample.ids,]
metadata.new <- metadata.new[order(match(metadata.new$SAMPLE_ID, meth.diff@sample.ids)),]
identical(metadata.new$SAMPLE_ID, meth.diff@sample.ids)
identical(metadata.new$SAMPLE_ID,meth.diff.df$individual)
meth.diff.df$pop <- metadata.new$region


heat.plot.data <- meth.diff.df %>%
  tibble::as_tibble() %>%
  tidyr::gather(key = "CpG", value = "Methylation_percentage", -individual, -pop,
                factor_key = FALSE) %>%
  dplyr::group_by(pop, CpG) %>%
  dplyr::mutate(CpG_mean = mean(Methylation_percentage, na.rm = TRUE)) %>%
  dplyr::arrange(pop, individual, CpG) %>%
  dplyr::mutate(individual = factor(individual, levels = unique(individual))) %>%
  dplyr::mutate(CpG = factor(CpG, levels = unique(CpG)))
cpgs <- heat.plot.data$CpG[heat.plot.data$individual == heat.plot.data$individual[1]]
cpgs_mean <- heat.plot.data$CpG_mean[heat.plot.data$individual == heat.plot.data$individual[1]]
cpgs <- cpgs[order(cpgs_mean)]
heat.plot.data$CpG <- factor(heat.plot.data$CpG, levels = unique(cpgs))
heat.plot.data <- heat.plot.data[order(heat.plot.data$CpG),]

heat.plot <- ggplot2::ggplot(data = heat.plot.data) + 
    ggplot2::geom_tile(ggplot2::aes(individual, y = CpG, fill = Methylation_percentage)) +
    ggplot2::scale_fill_gradient2(low = "blue", mid = "yellow", high = "red", midpoint = 50) +
    
    ggnewscale::new_scale_fill() +
    # ggplot2::geom_tile(ggplot2::aes(individual, y = "pop", fill = pop, height = 10)) +
    # ggplot2::scale_fill_manual(values = c("#4d6b40","#adcf9f", "#c5d6e6", "#08306B")) +
    # ggplot2::scale_fill_manual(values = c( "#c5d6e6", "#08306B")) +
    ggplot2::scale_y_discrete(drop = FALSE) +
    ggplot2::theme(legend.position = "top",
                   axis.title.y = ggplot2::element_blank(),
          axis.text.x = ggplot2::element_text(face = "bold", color = "black", 
                                        size = 10, angle = 90, hjust = 1, vjust = 0))
print(heat.plot)

ggplot2::ggsave(heat.plot, bg = "white",
                filename = paste0(figure.path,"TOA_ALLcpg_methDiff_EANT1_HEARD_HEATMAP.png"), 
                width = 45, height = 30, units = "cm")
```


#### PCA - HYPER

```{r}
myDiff25p.hyper.gr <- as(myDiff25p.hyper, "GRanges")
meth.diff <- methylKit::selectByOverlap(meth.EH1, myDiff25p.hyper.gr) 

methylKit::PCASamples(meth.diff, screeplot = TRUE)
dev.print(file = paste0(figure.path,
                        "PCA_plot_diffMeth_HYPER_Region_subset_EANT1-HEARD_scree.png"),
          device = png, res = 300,width = 30,height = 15,units = "cm")

methylKit::PCASamples(meth.diff, comp = c(1,2))
dev.print(file = paste0(figure.path,
                        "PCA_plot_diffMeth_HYPER_Region_subset_EANT1-HEARD_PC12.png"
                        ),
          device = png, res = 300,width = 30,height = 15,units = "cm")

methylKit::PCASamples(meth.diff, comp = c(1,3))
dev.print(file = paste0(figure.path,
                        "PCA_plot_diffMeth_HYPER_Region_subset_EANT1-HEARD_PC13.png"
                        ),
          device = png, res = 300,width = 30,height = 15,units = "cm")

methylKit::PCASamples(meth.diff, comp = c(2,3))
dev.print(file = paste0(figure.path,
                        "PCA_plot_diffMeth_HYPER_Region_subset_EANT1-HEARD_PC23.png"
                        ),
          device = png, res = 300,width = 30,height = 15,units = "cm")
```


##### Plot CpG difference

```{r}
myDiff25p.hyper.gr <- as(myDiff25p.hyper, "GRanges")
hist(myDiff25p.hyper.gr$meth.diff)
max(myDiff25p.hyper.gr$meth.diff) #51 %

meth.diff <- methylKit::selectByOverlap(meth.EH1, myDiff25p.hyper.gr)
length(unique(meth.diff$chr)) #25
meth.diff.df <- as.data.frame(t(methylKit::percMethylation(meth.diff)))
loc.names <- paste0(meth.diff$chr,"_",meth.diff$start)
colnames(meth.diff.df) <- loc.names

meth.diff.df$individual <- rownames(meth.diff.df) 
identical(metadata3$SAMPLE_ID,meth.diff.df$individual)
meth.diff.df$pop <- metadata3$region

meth.diff.df.plot <- tidyr::gather(meth.diff.df, key = locus, 
                           value = perc_methylation, -individual, -pop,
                           factor_key = TRUE) %>%
  dplyr::mutate(chr = stringr::str_split_fixed(locus, 
                                               pattern = "_", 
                                               n = 2)[,1])
meth.diff.df.plot.locus <- meth.diff.df.plot %>% dplyr::group_by(chr,locus,pop) %>% 
  dplyr::summarise(mean_pop = mean(perc_methylation, na.rm = TRUE),
            .groups = 'drop')

meth.diff.df.plot.centroid <- meth.diff.df.plot %>% dplyr::group_by(chr,pop) %>% 
  dplyr::summarise(mean_pop = mean(perc_methylation, na.rm = TRUE),
            .groups = 'drop')

meth.diff.df.plot.label <- meth.diff.df.plot %>% dplyr::group_by(chr) %>% 
  dplyr::summarise(nmarkers = length(unique(locus)),
            .groups = 'drop')


HYPER.plot <- ggplot2::ggplot() + 
  ggplot2::geom_violin(data = meth.diff.df.plot, 
                       ggplot2::aes(x = chr, y = perc_methylation)) + 
  ggplot2::geom_point(data = meth.diff.df.plot.centroid, 
                      ggplot2::aes(x = chr, y = mean_pop, colour = pop),
                      size = 7, shape = 5) + 
  ggplot2::geom_point(data = meth.diff.df.plot,
                      ggplot2::aes(x = chr, y = perc_methylation, colour = pop),
                      size = 1) +
  ggplot2::geom_text(data = meth.diff.df.plot.label,
                     ggplot2::aes(x = chr, y = 110, label = nmarkers),
                     size = 4) +
  ggplot2::geom_point(data = meth.diff.df.plot.locus, 
                      ggplot2::aes(x = chr, y = mean_pop, colour = pop),
                      size = 3, shape = 15) +
    ggplot2::theme(
        panel.background = ggplot2::element_blank(),
        panel.border = ggplot2::element_blank(),
        panel.grid = ggplot2::element_blank(),
        axis.line = ggplot2::element_line("gray25"),
        text = ggplot2::element_text(size = 12),
        axis.text = ggplot2::element_text(angle = 90,size = 10, colour = "gray25"),
        axis.title = ggplot2::element_text(size = 14, colour = "gray25"),
        legend.title = ggplot2::element_text(size = 14),
        legend.text = ggplot2::element_text(size = 10),
        legend.key = ggplot2::element_blank())

print(HYPER.plot)

ggplot2::ggsave(plot = HYPER.plot, 
                filename = paste0(figure.path,"TOA_HYPERcpg_methDiff_EANT1-HEARD.png"),
                width = 30, height = 15, units = "cm", device = "png")
```

#### PCA - HYPO

```{r}
myDiff25p.hypo.gr <- as(myDiff25p.hypo, "GRanges")
meth.diff <- methylKit::selectByOverlap(meth.EH1, myDiff25p.hypo.gr) 

methylKit::PCASamples(meth.diff, screeplot = TRUE)
dev.print(file = paste0(figure.path,
                        "PCA_plot_diffMeth_HYPO_Region_subset_EANT1-HEARD_scree.png"),
          device = png, res = 300,width = 30,height = 15,units = "cm")

methylKit::PCASamples(meth.diff, comp = c(1,2))
dev.print(file = paste0(figure.path,
                        "PCA_plot_diffMeth_HYPO_Region_subset_EANT1-HEARD_PC12.png"
                        ),
          device = png, res = 300,width = 30,height = 15,units = "cm")

methylKit::PCASamples(meth.diff, comp = c(1,3))
dev.print(file = paste0(figure.path,
                        "PCA_plot_diffMeth_HYPO_Region_subset_EANT1-HEARD_PC13.png"
                        ),
          device = png, res = 300,width = 30,height = 15,units = "cm")

methylKit::PCASamples(meth.diff, comp = c(2,3))
dev.print(file = paste0(figure.path,
                        "PCA_plot_diffMeth_HYPO_Region_subset_EANT1-HEARD_PC23.png"
                        ),
          device = png, res = 300,width = 30,height = 15,units = "cm")
```


##### Plot CpG difference

```{r}
myDiff25p.hypo.gr <- as(myDiff25p.hypo, "GRanges")
hist(myDiff25p.hypo.gr$meth.diff)
min(myDiff25p.hypo.gr$meth.diff)
max(myDiff25p.hypo.gr$meth.diff)

meth.diff <- methylKit::selectByOverlap(meth.EH1, myDiff25p.hypo.gr)
length(unique(meth.diff$chr)) #25
meth.diff.df <- as.data.frame(t(methylKit::percMethylation(meth.diff)))
loc.names <- paste0(meth.diff$chr,"_",meth.diff$start)
colnames(meth.diff.df) <- loc.names

meth.diff.df$individual <- rownames(meth.diff.df) 
identical(metadata3$SAMPLE_ID, meth.diff.df$individual)
meth.diff.df$pop <- metadata3$region

meth.diff.df.plot <- tidyr::gather(meth.diff.df, key = locus, 
                           value = perc_methylation, -individual, -pop,
                           factor_key = TRUE) %>%
  dplyr::mutate(chr = stringr::str_split_fixed(locus, 
                                               pattern = "_", 
                                               n = 2)[,1])
meth.diff.df.plot.locus <- meth.diff.df.plot %>% dplyr::group_by(chr,locus,pop) %>% 
  dplyr::summarise(mean_pop = mean(perc_methylation, na.rm = TRUE),
            .groups = 'drop')

meth.diff.df.plot.centroid <- meth.diff.df.plot %>% dplyr::group_by(chr,pop) %>% 
  dplyr::summarise(mean_pop = mean(perc_methylation, na.rm = TRUE),
            .groups = 'drop')

meth.diff.df.plot.label <- meth.diff.df.plot %>% dplyr::group_by(chr) %>% 
  dplyr::summarise(nmarkers = length(unique(locus)),
            .groups = 'drop')


HYPO.plot <- ggplot2::ggplot() + 
  ggplot2::geom_violin(data = meth.diff.df.plot, 
                       ggplot2::aes(x = chr, y = perc_methylation)) + 
  ggplot2::geom_point(data = meth.diff.df.plot.centroid, 
                      ggplot2::aes(x = chr, y = mean_pop, colour = pop),
                      size = 7, shape = 5) + 
  ggplot2::geom_point(data = meth.diff.df.plot,
                      ggplot2::aes(x = chr, y = perc_methylation, colour = pop),
                      size = 1) +
  ggplot2::geom_text(data = meth.diff.df.plot.label,
                     ggplot2::aes(x = chr, y = 110, label = nmarkers),
                     size = 4) +
  ggplot2::geom_point(data = meth.diff.df.plot.locus, 
                      ggplot2::aes(x = chr, y = mean_pop, colour = pop),
                      size = 3, shape = 15) +
    ggplot2::theme(
        panel.background = ggplot2::element_blank(),
        panel.border = ggplot2::element_blank(),
        panel.grid = ggplot2::element_blank(),
        axis.line = ggplot2::element_line("gray25"),
        text = ggplot2::element_text(size = 12),
        axis.text = ggplot2::element_text(angle = 90,size = 10, colour = "gray25"),
        axis.title = ggplot2::element_text(size = 14, colour = "gray25"),
        legend.title = ggplot2::element_text(size = 14),
        legend.text = ggplot2::element_text(size = 10),
        legend.key = ggplot2::element_blank())

print(HYPO.plot)

ggplot2::ggsave(plot = HYPO.plot, 
                filename = paste0(figure.path,"TOA_HYPOcpg_methDiff_EANT1-HEARD.png"),
                width = 30, height = 15, units = "cm", device = "png")
```


### EANT4.2 - HEARD
#### Prepare data

```{r}
print(load(paste0(rdata.path,"Toothfish_metadata.Rdata")))

metadata3 <- metadata2[metadata2$region %in% c("58-4-2", "58-5-2"),]
table(metadata3$region)
print(load(paste0(rdata.path,"Toothfish_methylation_objects_raw.Rdata")))

filtered.myobj3 <- methylKit::reorganize(
  methylObj = filtered.myobj,
  sample.ids = metadata3$SAMPLE_ID,
  treatment = as.numeric(factor(metadata3$FAO2)),
  chunk.size = 1e+06,
  save.db = TRUE,
  suffix = "subset_EANT2-HEARD_raw"
)

# filtered.myobj3@treatment <- as.numeric(factor(metadata3$region))
filtered.myobj3 <- methylKit::filterByCoverage(filtered.myobj3,lo.count = 10,lo.perc = NULL,
                                      hi.count = NULL, hi.perc = 99)

meth.EH2 <- methylKit::unite(filtered.myobj3, destrand = FALSE,
                             save.db = TRUE,
                             suffix = "subset_EANT2-HEARD",
                             min.per.group = 20L, #missing data allowed
                             mc.cores = 25)
#1,113,136   CpGs after coverage filter

# save(meth.HR,meth.EANT,meth.EH,meth.EH1, meth.EH2,
#      file = paste0(rdata.path,
#                    "Toothfish_subset_pairwise_methylation_objects_raw.Rdata"))
```

#### diffMethPerChr

```{r}
print(load(paste0(rdata.path,"Toothfish_subset_pairwise_methylation_objects_raw.Rdata")))
metadata3 <- metadata2[metadata2$region %in% c("58-4-2", "58-5-2"),]
covariates <- dplyr::select(metadata3, Sex, mat, Age)
myDiff <- methylKit::calculateDiffMeth(meth.EH2,
                                       covariates = covariates,
                                       adjust = "fdr",
                                       chunk.size = 1e+06,
                                       mc.cores = 10)
meth.diff.chr <- methylKit::diffMethPerChr(myDiff,plot = TRUE, 
                                           qvalue.cutoff = 0.01, meth.cutoff = 25)
myDiff25p.hyper <- methylKit::getMethylDiff(myDiff, difference = 25,qvalue = 0.01, type = "hyper") #137  
myDiff25p.hypo <- methylKit::getMethylDiff(myDiff, difference = 25,qvalue = 0.01, type = "hypo") #165 
myDiff25p <- methylKit::getMethylDiff(myDiff, difference = 25, qvalue = 0.01)

# save(myDiff,myDiff25p, myDiff25p.hyper, myDiff25p.hypo, 
#      file = paste0(rdata.path,
#                    "Toothfish_DMS_Region_subset_EANT2-HEARD.Rdata"))

myDiff.df <- as.data.frame(as(myDiff, "GRanges"))
myDiff25p.df <- as.data.frame(as(myDiff25p, "GRanges"))
volcano.plot <- ggplot2::ggplot() +
  ggplot2::geom_point(data = myDiff.df, ggplot2::aes(x = meth.diff, 
                                                  y = -log10(pvalue))) +
  ggplot2::geom_point(data = myDiff25p.df, ggplot2::aes(x = meth.diff, 
                                                  y = -log10(pvalue)),
                      colour = "red")
print(volcano.plot)

ggplot2::ggsave(volcano.plot, bg = "white",
                filename = paste0(figure.path,
                                  "TOA_ALL_DMS25_VOLCANO_methylkit_regions_subset_EANT2-HEARD.png"
                                  ), 
                width = 30, height = 15, units = "cm")
```

#### Load data

```{r}
print(load(paste0(rdata.path,"Toothfish_DMS_Region_subset_EANT2-HEARD.Rdata")))

myDiff25p.gr <- as(myDiff25p, "GRanges")
meth.diff <- methylKit::selectByOverlap(meth.EH2, myDiff25p.gr) 

# DMS <- data.frame(DMS = NULL, region_pw = NULL)
DMS.EH2 <- data.frame(DMS = paste0(meth.diff$chr,"_", meth.diff$start),
                  region_pw = "EANT4.2_HEARD")
DMS <- rbind(DMS, DMS.EH2)

# save(DMS, DMS.HR,DMS.EANT,DMS.EH,DMS.EH1, DMS.EH2,
#      file = paste0(rdata.path,"Toothfish_DMS_CpGsites.Rdata"))
```

#### PCA - ALL

```{r}
print(load(paste0(rdata.path,"Toothfish_DMS_Region_subset_EANT2-HEARD.Rdata")))

myDiff25p.gr <- as(myDiff25p, "GRanges")
meth.diff <- methylKit::selectByOverlap(meth.EH2, myDiff25p.gr) 

PCobj <- methylKit::PCASamples(meth.diff, comp = c(1,2), adj.lim = c(0,0), obj.return = TRUE)

var_frac <- PCobj$sdev/sum(PCobj$sdev)
print(paste0("Total variance explained: ",signif(sum(var_frac[1:3]) * 100, 3), "%")) 
barplot(var_frac[1:20], main = "variance explained", 
        col = heat.colors(length(var_frac[1:20])))

# metadata3 <- metadata2[metadata2$tissue %in% "M",]
metadata.new <- metadata[metadata$SAMPLE_ID %in% meth.diff@sample.ids,]
metadata.new <- metadata.new[order(match(metadata.new$SAMPLE_ID, meth.diff@sample.ids)),]
identical(metadata.new$SAMPLE_ID, meth.diff@sample.ids)
identical(metadata.new$SAMPLE_ID, rownames(PCobj$x))

pop <- metadata.new$region

# pop <- metadata.new$`Date caught`

# colours5 <- adegenet::funky(4)
# colours5 <- c("#4d6b40","#adcf9f", "#c5d6e6", "#08306B")
colours5 <- c("#adcf9f","#c5d6e6")


# plot.title <- "Toothfish methylation (1,316,799 CpG sites)"
plot.title <- "Toothfish methylation - EANT2 vs Heard (302 CpG sites)"


data <- data.frame(ind = rownames(PCobj$x),PCobj$x,  pop = pop)
data <- merge(data, aggregate(cbind(mean.x = PC1, mean.y = PC2) ~
                                pop, data, mean), by = "pop")

pca.plot <- ggplot2::ggplot(data, ggplot2::aes(PC1, PC2, color = pop)) +
  ggplot2::geom_point(size = 4) + 
  ggplot2::geom_point(ggplot2::aes(x = mean.x,
                                   y = mean.y, color = pop),
                      size = 7,
                      shape = 17) +
  ggplot2::scale_colour_manual(values = colours5) +
  ggplot2::scale_fill_manual(values = colours5) + 
  ggplot2::geom_hline(yintercept = 0) +
  ggplot2::geom_vline(xintercept = 0) + 
  ggplot2::labs(
    subtitle = paste0("Total variance explained: ",
                      signif(sum(var_frac[1:4]) * 100, 3), "%"),
    y = paste0("PC2; ", signif(var_frac[2] * 100, digits = 2), " %"),
    x = paste0("PC1; ", signif(var_frac[1] * 100, digits = 2), " %"),
    title = plot.title,
    caption = "") + 
  ggplot2::theme_minimal() +
  ggplot2::theme(
    axis.text = ggplot2::element_text(size = 15),
    axis.title.x = ggplot2::element_text(size = 20),
    axis.title.y = ggplot2::element_text(size = 20),
    legend.text = ggplot2::element_text(size = 15)
  )

print(pca.plot)
ggplot2::ggsave(pca.plot, bg = "white",
                filename = paste0(figure.path,"PCA_plot_diffMeth_ALL_Region_subset_EANT2-HEARD_P12.png"), 
                width = 30, height = 15, units = "cm")



data <- data.frame(ind = rownames(PCobj$x),PCobj$x,  pop = pop)
data <- merge(data, aggregate(cbind(mean.x = PC1, mean.y = PC3) ~
                                pop, data, mean), by = "pop")

pca.plot <- ggplot2::ggplot(data, ggplot2::aes(PC1, PC3, color = pop)) +
  ggplot2::geom_point(size = 4) + 
  ggplot2::geom_point(ggplot2::aes(x = mean.x,
                                   y = mean.y, color = pop),
                      size = 7,
                      shape = 17) +
  ggplot2::scale_colour_manual(values = colours5) +
  ggplot2::scale_fill_manual(values = colours5) + 
  ggplot2::geom_hline(yintercept = 0) +
  ggplot2::geom_vline(xintercept = 0) + 
  ggplot2::labs(
    subtitle = paste0("Total variance explained: ",
                      signif(sum(var_frac[1:4]) * 100, 3), "%"),
    y = paste0("PC3; ", signif(var_frac[3] * 100, digits = 2), " %"),
    x = paste0("PC1; ", signif(var_frac[1] * 100, digits = 2), " %"),
    title = plot.title,
    caption = "") + 
  ggplot2::theme_minimal() +
  ggplot2::theme(
    axis.text = ggplot2::element_text(size = 15),
    axis.title.x = ggplot2::element_text(size = 20),
    axis.title.y = ggplot2::element_text(size = 20),
    legend.text = ggplot2::element_text(size = 15)
  )

print(pca.plot)
ggplot2::ggsave(pca.plot, bg = "white",
                filename = paste0(figure.path,"PCA_plot_diffMeth_ALL_Region_subset_EANT2-HEARD_P13.png"), 
                width = 30, height = 15, units = "cm")




data <- data.frame(ind = rownames(PCobj$x),PCobj$x,  pop = pop)
data <- merge(data, aggregate(cbind(mean.x = PC2, mean.y = PC3) ~
                                pop, data, mean), by = "pop")

pca.plot <- ggplot2::ggplot(data, ggplot2::aes(PC2, PC3, color = pop)) +
  ggplot2::geom_point(size = 4) + 
  ggplot2::geom_point(ggplot2::aes(x = mean.x,
                                   y = mean.y, color = pop),
                      size = 7,
                      shape = 17) +
  ggplot2::scale_colour_manual(values = colours5) +
  ggplot2::scale_fill_manual(values = colours5) + 
  ggplot2::geom_hline(yintercept = 0) +
  ggplot2::geom_vline(xintercept = 0) + 
  ggplot2::labs(
    subtitle = paste0("Total variance explained: ",
                      signif(sum(var_frac[1:4]) * 100, 3), "%"),
    y = paste0("PC3; ", signif(var_frac[3] * 100, digits = 2), " %"),
    x = paste0("PC2; ", signif(var_frac[2] * 100, digits = 2), " %"),
    title = plot.title,
    caption = "") + 
  ggplot2::theme_minimal() +
  ggplot2::theme(
    axis.text = ggplot2::element_text(size = 15),
    axis.title.x = ggplot2::element_text(size = 20),
    axis.title.y = ggplot2::element_text(size = 20),
    legend.text = ggplot2::element_text(size = 15)
  )

print(pca.plot)
ggplot2::ggsave(pca.plot, bg = "white",
                filename = paste0(figure.path,"PCA_plot_diffMeth_ALL_Region_subset_EANT2-HEARD_P23.png"), 
                width = 30, height = 15, units = "cm")
```


##### Plot CpG difference

```{r}
myDiff25p.gr <- as(myDiff25p, "GRanges")
hist(myDiff25p.gr$meth.diff)
min(myDiff25p.gr$meth.diff)#-40.32129 %
max(myDiff25p.gr$meth.diff)#38.63991 %

meth.diff <- methylKit::selectByOverlap(meth.EH2, myDiff25p.gr)
length(unique(meth.diff$chr)) #49
meth.diff.df <- as.data.frame(t(methylKit::percMethylation(meth.diff)))
loc.names <- paste0(meth.diff$chr,"_",meth.diff$start)
colnames(meth.diff.df) <- loc.names

meth.diff.df$individual <- rownames(meth.diff.df) 
metadata.new <- metadata[metadata$SAMPLE_ID %in% meth.diff@sample.ids,]
metadata.new <- metadata.new[order(match(metadata.new$SAMPLE_ID, meth.diff@sample.ids)),]
identical(metadata.new$SAMPLE_ID, meth.diff@sample.ids)
identical(metadata.new$SAMPLE_ID,meth.diff.df$individual)
meth.diff.df$pop <- metadata.new$region

meth.diff.df.plot <- tidyr::gather(meth.diff.df, key = locus, 
                           value = perc_methylation, -individual, -pop,
                           factor_key = TRUE) %>%
  dplyr::mutate(chr = stringr::str_split_fixed(locus, 
                                               pattern = "_", 
                                               n = 2)[,1])

meth.diff.df.plot.locus <- meth.diff.df.plot %>% dplyr::group_by(chr,locus,pop) %>% 
  dplyr::summarise(mean_pop = mean(perc_methylation, na.rm = TRUE),
            .groups = 'drop')

meth.diff.df.plot.centroid <- meth.diff.df.plot %>% dplyr::group_by(chr,pop) %>% 
  dplyr::summarise(mean_pop = mean(perc_methylation, na.rm = TRUE),
            .groups = 'drop')

meth.diff.df.plot.label <- meth.diff.df.plot %>% dplyr::group_by(chr) %>% 
  dplyr::summarise(nmarkers = length(unique(locus)),
            .groups = 'drop')


ALL.plot <- ggplot2::ggplot() + 
  ggplot2::geom_violin(data = meth.diff.df.plot, 
                       ggplot2::aes(x = chr, y = perc_methylation)) + 
  ggplot2::geom_point(data = meth.diff.df.plot.centroid, 
                      ggplot2::aes(x = chr, y = mean_pop, colour = pop),
                      size = 7, shape = 5) + 
  ggplot2::geom_point(data = meth.diff.df.plot,
                      ggplot2::aes(x = chr, y = perc_methylation, colour = pop),
                      size = 1) +
  ggplot2::geom_text(data = meth.diff.df.plot.label,
                     ggplot2::aes(x = chr, y = 110, label = nmarkers),
                     size = 4) +
  ggplot2::geom_point(data = meth.diff.df.plot.locus, 
                      ggplot2::aes(x = chr, y = mean_pop, colour = pop),
                      size = 3, shape = 15) +
    ggplot2::theme(
        panel.background = ggplot2::element_blank(),
        panel.border = ggplot2::element_blank(),
        panel.grid = ggplot2::element_blank(),
        axis.line = ggplot2::element_line("gray25"),
        text = ggplot2::element_text(size = 12),
        axis.text = ggplot2::element_text(angle = 90,size = 10, colour = "gray25"),
        axis.title = ggplot2::element_text(size = 14, colour = "gray25"),
        legend.title = ggplot2::element_text(size = 14),
        legend.text = ggplot2::element_text(size = 10),
        legend.key = ggplot2::element_blank())

print(ALL.plot)

ggplot2::ggsave(plot = ALL.plot, 
                filename = paste0(figure.path,"TOA_ALLcpg_methDiff_EANT2-HEARD.png"),
                width = 30, height = 15, units = "cm", device = "png")



diff.plot <- ggplot2::ggplot() +
  # ggplot2::geom_point(data = meth.diff.df.plot,
  #                     ggplot2::aes(x = pop, y = perc_methylation, colour = locus)) +
  ggplot2::geom_line(data = meth.diff.df.plot.locus,
                     ggplot2::aes(x = pop, y = mean_pop, colour = locus, group = locus)) +
  ggplot2::geom_point(data = meth.diff.df.plot.locus,
                     ggplot2::aes(x = pop, y = mean_pop, colour = locus, group = locus),
                     size = 5) +
  ggplot2::ylim(c(0,100)) +
  ggplot2::facet_wrap(~chr) +
  ggplot2::theme(legend.position = "none")
  # ggplot2::guides(colour = ggplot2::guide_legend(ncol = 1))
  # ggplot2::guides(colour = ggplot2::guide_legend(ncol = 5))

print(diff.plot)

ggplot2::ggsave(plot = diff.plot, 
                filename = paste0(figure.path,
                                  "TOA_ALLcpg_methDiff_EANT2-HEARD_pw.png"
                                  ),
                width = 30, height = 25, units = "cm", device = "png")


table(meth.diff.df.plot$chr[!duplicated(meth.diff.df.plot$locus)])
meth.diff.df.plot2 <- meth.diff.df.plot[meth.diff.df.plot$chr == "JAAKFY010000002",]
meth.diff.df.plot.locus2 <- meth.diff.df.plot.locus[meth.diff.df.plot.locus$chr == "JAAKFY010000002",]
diff.plot2 <- ggplot2::ggplot() +
  ggplot2::geom_point(data = meth.diff.df.plot2,
                      ggplot2::aes(x = pop, y = perc_methylation, colour = locus)) +
  ggplot2::geom_line(data = meth.diff.df.plot.locus2,
                     ggplot2::aes(x = pop, y = mean_pop, colour = locus, group = locus)) +
  ggplot2::geom_point(data = meth.diff.df.plot.locus2,
                     ggplot2::aes(x = pop, y = mean_pop, colour = locus, group = locus),
                     size = 5) +
  ggplot2::ylim(c(0,100)) +
  ggplot2::facet_wrap(~locus) +
  ggplot2::theme(legend.position = "none")
  # ggplot2::guides(colour = ggplot2::guide_legend(ncol = 1))
  # ggplot2::guides(colour = ggplot2::guide_legend(ncol = 5))

print(diff.plot2)
ggplot2::ggsave(plot = diff.plot2, 
                filename = paste0(figure.path,
                                  "TOA_ALLcpg_methDiff_EANT2-HEARD_pw_CHR02.png"
                                  ),
                width = 30, height = 25, units = "cm", device = "png")


meth.diff.df.plot2 <- meth.diff.df.plot[meth.diff.df.plot$chr == "JAAKFY010000014",]
meth.diff.df.plot.locus2 <- meth.diff.df.plot.locus[meth.diff.df.plot.locus$chr == "JAAKFY010000014",]
diff.plot14 <- ggplot2::ggplot() +
  ggplot2::geom_point(data = meth.diff.df.plot2,
                      ggplot2::aes(x = pop, y = perc_methylation, colour = locus)) +
  ggplot2::geom_line(data = meth.diff.df.plot.locus2,
                     ggplot2::aes(x = pop, y = mean_pop, colour = locus, group = locus)) +
  ggplot2::geom_point(data = meth.diff.df.plot.locus2,
                     ggplot2::aes(x = pop, y = mean_pop, colour = locus, group = locus),
                     size = 5) +
  ggplot2::ylim(c(0,100)) +
  ggplot2::facet_wrap(~locus) +
  ggplot2::theme(legend.position = "none")
  # ggplot2::guides(colour = ggplot2::guide_legend(ncol = 1))
  # ggplot2::guides(colour = ggplot2::guide_legend(ncol = 5))

print(diff.plot14)
ggplot2::ggsave(plot = diff.plot14, 
                filename = paste0(figure.path,
                                  "TOA_ALLcpg_methDiff_EANT2-HEARD_pw_CHR14.png"
                                  ),
                width = 60, height = 45, units = "cm", device = "png")

meth.diff.df.plot2 <- meth.diff.df.plot[meth.diff.df.plot$chr == "JAAKFY010000018",]
meth.diff.df.plot.locus2 <- meth.diff.df.plot.locus[meth.diff.df.plot.locus$chr == "JAAKFY010000018",]
diff.plot18 <- ggplot2::ggplot() +
  ggplot2::geom_point(data = meth.diff.df.plot2,
                      ggplot2::aes(x = pop, y = perc_methylation, colour = locus)) +
  ggplot2::geom_line(data = meth.diff.df.plot.locus2,
                     ggplot2::aes(x = pop, y = mean_pop, colour = locus, group = locus)) +
  ggplot2::geom_point(data = meth.diff.df.plot.locus2,
                     ggplot2::aes(x = pop, y = mean_pop, colour = locus, group = locus),
                     size = 5) +
  ggplot2::ylim(c(0,100)) +
  ggplot2::facet_wrap(~locus) +
  ggplot2::theme(legend.position = "none")
  # ggplot2::guides(colour = ggplot2::guide_legend(ncol = 1))
  # ggplot2::guides(colour = ggplot2::guide_legend(ncol = 5))

print(diff.plot18)

ggplot2::ggsave(plot = diff.plot18, 
                filename = paste0(figure.path,
                                  "TOA_ALLcpg_methDiff_EANT2-HEARD_pw_CHR18.png"
                                  ),
                width = 30, height = 25, units = "cm", device = "png")

meth.diff.df.plot2 <- meth.diff.df.plot[meth.diff.df.plot$chr == "JAAKFY010000026",]
meth.diff.df.plot.locus2 <- meth.diff.df.plot.locus[meth.diff.df.plot.locus$chr == "JAAKFY010000026",]
diff.plot26 <- ggplot2::ggplot() +
  ggplot2::geom_point(data = meth.diff.df.plot2,
                      ggplot2::aes(x = pop, y = perc_methylation, colour = locus)) +
  ggplot2::geom_line(data = meth.diff.df.plot.locus2,
                     ggplot2::aes(x = pop, y = mean_pop, colour = locus, group = locus)) +
  ggplot2::geom_point(data = meth.diff.df.plot.locus2,
                     ggplot2::aes(x = pop, y = mean_pop, colour = locus, group = locus),
                     size = 5) +
  ggplot2::ylim(c(0,100)) +
  ggplot2::facet_wrap(~locus) +
  ggplot2::theme(legend.position = "none")
  # ggplot2::guides(colour = ggplot2::guide_legend(ncol = 1))
  # ggplot2::guides(colour = ggplot2::guide_legend(ncol = 5))

print(diff.plot26)

ggplot2::ggsave(plot = diff.plot26, 
                filename = paste0(figure.path,
                                  "TOA_ALLcpg_methDiff_EANT2-HEARD_pw_CHR26.png"
                                  ),
                width = 30, height = 25, units = "cm", device = "png")
```



##### Heatmap

```{r}
meth.diff.df <- as.data.frame(t(methylKit::percMethylation(meth.diff)))
loc.names <- paste0(meth.diff$chr,"_",meth.diff$start)
colnames(meth.diff.df) <- loc.names

meth.diff.df$individual <- rownames(meth.diff.df) 
metadata.new <- metadata[metadata$SAMPLE_ID %in% meth.diff@sample.ids,]
metadata.new <- metadata.new[order(match(metadata.new$SAMPLE_ID, meth.diff@sample.ids)),]
identical(metadata.new$SAMPLE_ID, meth.diff@sample.ids)
identical(metadata.new$SAMPLE_ID,meth.diff.df$individual)
meth.diff.df$pop <- metadata.new$region


heat.plot.data <- meth.diff.df %>%
  tibble::as_tibble() %>%
  tidyr::gather(key = "CpG", value = "Methylation_percentage", -individual, -pop,
                factor_key = FALSE) %>%
  dplyr::group_by(pop, CpG) %>%
  dplyr::mutate(CpG_mean = mean(Methylation_percentage, na.rm = TRUE)) %>%
  dplyr::arrange(pop, individual, CpG) %>%
  dplyr::mutate(individual = factor(individual, levels = unique(individual))) %>%
  dplyr::mutate(CpG = factor(CpG, levels = unique(CpG)))
cpgs <- heat.plot.data$CpG[heat.plot.data$individual == heat.plot.data$individual[1]]
cpgs_mean <- heat.plot.data$CpG_mean[heat.plot.data$individual == heat.plot.data$individual[1]]
cpgs <- cpgs[order(cpgs_mean)]
heat.plot.data$CpG <- factor(heat.plot.data$CpG, levels = unique(cpgs))
heat.plot.data <- heat.plot.data[order(heat.plot.data$CpG),]

heat.plot <- ggplot2::ggplot(data = heat.plot.data) + 
    ggplot2::geom_tile(ggplot2::aes(individual, y = CpG, fill = Methylation_percentage)) +
    ggplot2::scale_fill_gradient2(low = "blue", mid = "yellow", high = "red", midpoint = 50) +
    
    ggnewscale::new_scale_fill() +
    # ggplot2::geom_tile(ggplot2::aes(individual, y = "pop", fill = pop, height = 10)) +
    # ggplot2::scale_fill_manual(values = c("#4d6b40","#adcf9f", "#c5d6e6", "#08306B")) +
    # ggplot2::scale_fill_manual(values = c( "#c5d6e6", "#08306B")) +
    ggplot2::scale_y_discrete(drop = FALSE) +
    ggplot2::theme(legend.position = "top",
                   axis.title.y = ggplot2::element_blank(),
          axis.text.x = ggplot2::element_text(face = "bold", color = "black", 
                                        size = 10, angle = 90, hjust = 1, vjust = 0))
print(heat.plot)

ggplot2::ggsave(heat.plot, bg = "white",
                filename = paste0(figure.path,"TOA_ALLcpg_methDiff_EANT2_HEARD_HEATMAP.png"), 
                width = 45, height = 30, units = "cm")
```


#### PCA - HYPER

```{r}
myDiff25p.hyper.gr <- as(myDiff25p.hyper, "GRanges")
meth.diff <- methylKit::selectByOverlap(meth.EH2, myDiff25p.hyper.gr) 

methylKit::PCASamples(meth.diff, screeplot = TRUE)
dev.print(file = paste0(figure.path,
                        "PCA_plot_diffMeth_HYPER_Region_subset_EANT2-HEARD_scree.png"),
          device = png, res = 300,width = 30,height = 15,units = "cm")

methylKit::PCASamples(meth.diff, comp = c(1,2))
dev.print(file = paste0(figure.path,
                        "PCA_plot_diffMeth_HYPER_Region_subset_EANT2-HEARD_PC12.png"
                        ),
          device = png, res = 300,width = 30,height = 15,units = "cm")

methylKit::PCASamples(meth.diff, comp = c(1,3))
dev.print(file = paste0(figure.path,
                        "PCA_plot_diffMeth_HYPER_Region_subset_EANT2-HEARD_PC13.png"
                        ),
          device = png, res = 300,width = 30,height = 15,units = "cm")

methylKit::PCASamples(meth.diff, comp = c(2,3))
dev.print(file = paste0(figure.path,
                        "PCA_plot_diffMeth_HYPER_Region_subset_EANT2-HEARD_PC23.png"
                        ),
          device = png, res = 300,width = 30,height = 15,units = "cm")
```


##### Plot CpG difference

```{r}
myDiff25p.hyper.gr <- as(myDiff25p.hyper, "GRanges")
hist(myDiff25p.hyper.gr$meth.diff)
max(myDiff25p.hyper.gr$meth.diff) #42.5 %

meth.diff <- methylKit::selectByOverlap(meth.EH2, myDiff25p.hyper.gr)
length(unique(meth.diff$chr)) #25
meth.diff.df <- as.data.frame(t(methylKit::percMethylation(meth.diff)))
loc.names <- paste0(meth.diff$chr,"_",meth.diff$start)
colnames(meth.diff.df) <- loc.names

meth.diff.df$individual <- rownames(meth.diff.df) 
identical(metadata3$SAMPLE_ID,meth.diff.df$individual)
meth.diff.df$pop <- metadata3$region

meth.diff.df.plot <- tidyr::gather(meth.diff.df, key = locus, 
                           value = perc_methylation, -individual, -pop,
                           factor_key = TRUE) %>%
  dplyr::mutate(chr = stringr::str_split_fixed(locus, 
                                               pattern = "_", 
                                               n = 2)[,1])
meth.diff.df.plot.locus <- meth.diff.df.plot %>% dplyr::group_by(chr,locus,pop) %>% 
  dplyr::summarise(mean_pop = mean(perc_methylation, na.rm = TRUE),
            .groups = 'drop')

meth.diff.df.plot.centroid <- meth.diff.df.plot %>% dplyr::group_by(chr,pop) %>% 
  dplyr::summarise(mean_pop = mean(perc_methylation, na.rm = TRUE),
            .groups = 'drop')

meth.diff.df.plot.label <- meth.diff.df.plot %>% dplyr::group_by(chr) %>% 
  dplyr::summarise(nmarkers = length(unique(locus)),
            .groups = 'drop')


HYPER.plot <- ggplot2::ggplot() + 
  ggplot2::geom_violin(data = meth.diff.df.plot, 
                       ggplot2::aes(x = chr, y = perc_methylation)) + 
  ggplot2::geom_point(data = meth.diff.df.plot.centroid, 
                      ggplot2::aes(x = chr, y = mean_pop, colour = pop),
                      size = 7, shape = 5) + 
  ggplot2::geom_point(data = meth.diff.df.plot,
                      ggplot2::aes(x = chr, y = perc_methylation, colour = pop),
                      size = 1) +
  ggplot2::geom_text(data = meth.diff.df.plot.label,
                     ggplot2::aes(x = chr, y = 110, label = nmarkers),
                     size = 4) +
  ggplot2::geom_point(data = meth.diff.df.plot.locus, 
                      ggplot2::aes(x = chr, y = mean_pop, colour = pop),
                      size = 3, shape = 15) +
    ggplot2::theme(
        panel.background = ggplot2::element_blank(),
        panel.border = ggplot2::element_blank(),
        panel.grid = ggplot2::element_blank(),
        axis.line = ggplot2::element_line("gray25"),
        text = ggplot2::element_text(size = 12),
        axis.text = ggplot2::element_text(angle = 90,size = 10, colour = "gray25"),
        axis.title = ggplot2::element_text(size = 14, colour = "gray25"),
        legend.title = ggplot2::element_text(size = 14),
        legend.text = ggplot2::element_text(size = 10),
        legend.key = ggplot2::element_blank())

print(HYPER.plot)

ggplot2::ggsave(plot = HYPER.plot, 
                filename = paste0(figure.path,"TOA_HYPERcpg_methDiff_EANT2-HEARD.png"),
                width = 30, height = 15, units = "cm", device = "png")
```

#### PCA - HYPO

```{r}
myDiff25p.hypo.gr <- as(myDiff25p.hypo, "GRanges")
meth.diff <- methylKit::selectByOverlap(meth.EH2, myDiff25p.hypo.gr) 

methylKit::PCASamples(meth.diff, screeplot = TRUE)
dev.print(file = paste0(figure.path,
                        "PCA_plot_diffMeth_HYPO_Region_subset_EANT2-HEARD_scree.png"),
          device = png, res = 300,width = 30,height = 15,units = "cm")

methylKit::PCASamples(meth.diff, comp = c(1,2))
dev.print(file = paste0(figure.path,
                        "PCA_plot_diffMeth_HYPO_Region_subset_EANT2-HEARD_PC12.png"
                        ),
          device = png, res = 300,width = 30,height = 15,units = "cm")

methylKit::PCASamples(meth.diff, comp = c(1,3))
dev.print(file = paste0(figure.path,
                        "PCA_plot_diffMeth_HYPO_Region_subset_EANT2-HEARD_PC13.png"
                        ),
          device = png, res = 300,width = 30,height = 15,units = "cm")

methylKit::PCASamples(meth.diff, comp = c(2,3))
dev.print(file = paste0(figure.path,
                        "PCA_plot_diffMeth_HYPO_Region_subset_EANT2-HEARD_PC23.png"
                        ),
          device = png, res = 300,width = 30,height = 15,units = "cm")
```


##### Plot CpG difference

```{r}
myDiff25p.hypo.gr <- as(myDiff25p.hypo, "GRanges")
hist(myDiff25p.hypo.gr$meth.diff)
min(myDiff25p.hypo.gr$meth.diff)
max(myDiff25p.hypo.gr$meth.diff)

meth.diff <- methylKit::selectByOverlap(meth.EH2, myDiff25p.hypo.gr)
length(unique(meth.diff$chr)) #25
meth.diff.df <- as.data.frame(t(methylKit::percMethylation(meth.diff)))
loc.names <- paste0(meth.diff$chr,"_",meth.diff$start)
colnames(meth.diff.df) <- loc.names

meth.diff.df$individual <- rownames(meth.diff.df) 
identical(metadata3$SAMPLE_ID, meth.diff.df$individual)
meth.diff.df$pop <- metadata3$region

meth.diff.df.plot <- tidyr::gather(meth.diff.df, key = locus, 
                           value = perc_methylation, -individual, -pop,
                           factor_key = TRUE) %>%
  dplyr::mutate(chr = stringr::str_split_fixed(locus, 
                                               pattern = "_", 
                                               n = 2)[,1])
meth.diff.df.plot.locus <- meth.diff.df.plot %>% dplyr::group_by(chr,locus,pop) %>% 
  dplyr::summarise(mean_pop = mean(perc_methylation, na.rm = TRUE),
            .groups = 'drop')

meth.diff.df.plot.centroid <- meth.diff.df.plot %>% dplyr::group_by(chr,pop) %>% 
  dplyr::summarise(mean_pop = mean(perc_methylation, na.rm = TRUE),
            .groups = 'drop')

meth.diff.df.plot.label <- meth.diff.df.plot %>% dplyr::group_by(chr) %>% 
  dplyr::summarise(nmarkers = length(unique(locus)),
            .groups = 'drop')


HYPO.plot <- ggplot2::ggplot() + 
  ggplot2::geom_violin(data = meth.diff.df.plot, 
                       ggplot2::aes(x = chr, y = perc_methylation)) + 
  ggplot2::geom_point(data = meth.diff.df.plot.centroid, 
                      ggplot2::aes(x = chr, y = mean_pop, colour = pop),
                      size = 7, shape = 5) + 
  ggplot2::geom_point(data = meth.diff.df.plot,
                      ggplot2::aes(x = chr, y = perc_methylation, colour = pop),
                      size = 1) +
  ggplot2::geom_text(data = meth.diff.df.plot.label,
                     ggplot2::aes(x = chr, y = 110, label = nmarkers),
                     size = 4) +
  ggplot2::geom_point(data = meth.diff.df.plot.locus, 
                      ggplot2::aes(x = chr, y = mean_pop, colour = pop),
                      size = 3, shape = 15) +
    ggplot2::theme(
        panel.background = ggplot2::element_blank(),
        panel.border = ggplot2::element_blank(),
        panel.grid = ggplot2::element_blank(),
        axis.line = ggplot2::element_line("gray25"),
        text = ggplot2::element_text(size = 12),
        axis.text = ggplot2::element_text(angle = 90,size = 10, colour = "gray25"),
        axis.title = ggplot2::element_text(size = 14, colour = "gray25"),
        legend.title = ggplot2::element_text(size = 14),
        legend.text = ggplot2::element_text(size = 10),
        legend.key = ggplot2::element_blank())

print(HYPO.plot)

ggplot2::ggsave(plot = HYPO.plot, 
                filename = paste0(figure.path,"TOA_HYPOcpg_methDiff_EANT2-HEARD.png"),
                width = 30, height = 15, units = "cm", device = "png")
```


### EANT - ROSS
#### Prepare data

```{r}
print(load(paste0(rdata.path,"Toothfish_metadata.Rdata")))

metadata3 <- metadata2[metadata2$FAO2 %in% c("58.4", "88.1"),]
table(metadata3$FAO2)
print(load(paste0(rdata.path,"Toothfish_methylation_objects_raw.Rdata")))

filtered.myobj3 <- methylKit::reorganize(
  methylObj = filtered.myobj,
  sample.ids = metadata3$SAMPLE_ID,
  treatment = as.numeric(factor(metadata3$FAO2)),
  chunk.size = 1e+06,
  save.db = TRUE,
  suffix = "subset_EANT-ROSS_raw"
)

# filtered.myobj3@treatment <- as.numeric(factor(metadata3$region))
filtered.myobj3 <- methylKit::filterByCoverage(filtered.myobj3,lo.count = 10,lo.perc = NULL,
                                      hi.count = NULL, hi.perc = 99)

meth.ER <- methylKit::unite(filtered.myobj3, destrand = FALSE,
                             save.db = TRUE,
                             suffix = "subset_EANT-ROSS",
                             min.per.group = 36L, #missing data allowed - 80%
                             mc.cores = 25)
#969,125   CpGs after coverage filter

# save(meth.HR,meth.EANT,meth.EH,meth.ER,
#      file = paste0(rdata.path,
#                    "Toothfish_subset_pairwise_methylation_objects_raw.Rdata"))
```

#### diffMethPerChr

```{r}
print(load(paste0(rdata.path,"Toothfish_subset_pairwise_methylation_objects_raw.Rdata")))
metadata3 <- metadata2[metadata2$FAO2 %in% c("58.4", "88.1"),]
covariates <- dplyr::select(metadata3, Sex, mat, Age)
myDiff <- methylKit::calculateDiffMeth(meth.ER,
                                       covariates = covariates,
                                       adjust = "fdr",
                                       chunk.size = 1e+06,
                                       mc.cores = 10)
meth.diff.chr <- methylKit::diffMethPerChr(myDiff,plot = TRUE, 
                                           qvalue.cutoff = 0.01, meth.cutoff = 25)
myDiff25p.hyper <- methylKit::getMethylDiff(myDiff, difference = 25,qvalue = 0.01, type = "hyper") #10  
myDiff25p.hypo <- methylKit::getMethylDiff(myDiff, difference = 25,qvalue = 0.01, type = "hypo") #5 
myDiff25p <- methylKit::getMethylDiff(myDiff, difference = 25, qvalue = 0.01)

# save(myDiff,myDiff25p, myDiff25p.hyper, myDiff25p.hypo, 
#      file = paste0(rdata.path,
#                    "Toothfish_DMS_Region_subset_EANT-ROSS.Rdata"))


myDiff.df <- as.data.frame(as(myDiff, "GRanges"))
myDiff25p.df <- as.data.frame(as(myDiff25p, "GRanges"))
volcano.plot <- ggplot2::ggplot() +
  ggplot2::geom_point(data = myDiff.df, ggplot2::aes(x = meth.diff, 
                                                  y = -log10(pvalue))) +
  ggplot2::geom_point(data = myDiff25p.df, ggplot2::aes(x = meth.diff, 
                                                  y = -log10(pvalue)),
                      colour = "red")
print(volcano.plot)

ggplot2::ggsave(volcano.plot, bg = "white",
                filename = paste0(figure.path,
                                  "TOA_ALL_DMS25_VOLCANO_methylkit_regions_subset_EANT-ROSS.png"
                                  ), 
                width = 30, height = 15, units = "cm")
```

#### Load data

```{r}
print(load(paste0(rdata.path,"Toothfish_DMS_Region_subset_EANT-ROSS.Rdata")))

myDiff25p.gr <- as(myDiff25p, "GRanges")
meth.diff <- methylKit::selectByOverlap(meth.ER, myDiff25p.gr) 

# DMS <- data.frame(DMS = NULL, region_pw = NULL)
DMS.ER <- data.frame(DMS = paste0(meth.diff$chr,"_", meth.diff$start),
                  region_pw = "EANT_ROSS")
DMS <- rbind(DMS, DMS.ER)

# save(DMS, DMS.HR,DMS.EANT,DMS.EH,DMS.EH1, DMS.EH2,DMS.ER,
#      file = paste0(rdata.path,"Toothfish_DMS_CpGsites.Rdata"))
```


#### PCA - ALL

```{r}
print(load(paste0(rdata.path,"Toothfish_DMS_Region_subset_EANT-ROSS.Rdata")))


myDiff25p.gr <- as(myDiff25p, "GRanges")
meth.diff <- methylKit::selectByOverlap(meth.ER, myDiff25p.gr) 

methylKit::PCASamples(meth.diff, screeplot = TRUE)
dev.print(file = paste0(figure.path,
                        "PCA_plot_diffMeth_ALL_Region_subset_EANT-ROSS_scree.png"),
          device = png, res = 300,width = 30,height = 15,units = "cm")

methylKit::PCASamples(meth.diff, comp = c(1,2))
dev.print(file = paste0(figure.path,
                        "PCA_plot_diffMeth_ALL_Region_subset_EANT-ROSS_PC12.png"
                        ),
          device = png, res = 300,width = 30,height = 15,units = "cm")

methylKit::PCASamples(meth.diff, comp = c(1,3))
dev.print(file = paste0(figure.path,
                        "PCA_plot_diffMeth_ALL_Region_subset_EANT-ROSS_PC13.png"
                        ),
          device = png, res = 300,width = 30,height = 15,units = "cm")

methylKit::PCASamples(meth.diff, comp = c(2,3))
dev.print(file = paste0(figure.path,
                        "PCA_plot_diffMeth_ALL_Region_subset_EANT-ROSS_PC23.png"
                        ),
          device = png, res = 300,width = 30,height = 15,units = "cm")
```


##### Plot CpG difference

```{r}
myDiff25p.gr <- as(myDiff25p, "GRanges")
hist(myDiff25p.gr$meth.diff)
min(myDiff25p.gr$meth.diff)#-26 %
max(myDiff25p.gr$meth.diff)#33 %

meth.diff <- methylKit::selectByOverlap(meth.ER, myDiff25p.gr)
length(unique(meth.diff$chr)) #49
meth.diff.df <- as.data.frame(t(methylKit::percMethylation(meth.diff)))
loc.names <- paste0(meth.diff$chr,"_",meth.diff$start)
colnames(meth.diff.df) <- loc.names

meth.diff.df$individual <- rownames(meth.diff.df) 
metadata.new <- metadata[metadata$SAMPLE_ID %in% meth.diff@sample.ids,]
metadata.new <- metadata.new[order(match(metadata.new$SAMPLE_ID, meth.diff@sample.ids)),]
identical(metadata.new$SAMPLE_ID, meth.diff@sample.ids)
identical(metadata.new$SAMPLE_ID,meth.diff.df$individual)
meth.diff.df$pop <- metadata.new$region

meth.diff.df.plot <- tidyr::gather(meth.diff.df, key = locus, 
                           value = perc_methylation, -individual, -pop,
                           factor_key = TRUE) %>%
  dplyr::mutate(chr = stringr::str_split_fixed(locus, 
                                               pattern = "_", 
                                               n = 2)[,1])

meth.diff.df.plot.locus <- meth.diff.df.plot %>% dplyr::group_by(chr,locus,pop) %>% 
  dplyr::summarise(mean_pop = mean(perc_methylation, na.rm = TRUE),
            .groups = 'drop')

meth.diff.df.plot.centroid <- meth.diff.df.plot %>% dplyr::group_by(chr,pop) %>% 
  dplyr::summarise(mean_pop = mean(perc_methylation, na.rm = TRUE),
            .groups = 'drop')

meth.diff.df.plot.label <- meth.diff.df.plot %>% dplyr::group_by(chr) %>% 
  dplyr::summarise(nmarkers = length(unique(locus)),
            .groups = 'drop')


ALL.plot <- ggplot2::ggplot() + 
  ggplot2::geom_violin(data = meth.diff.df.plot, 
                       ggplot2::aes(x = chr, y = perc_methylation)) + 
  ggplot2::geom_point(data = meth.diff.df.plot.centroid, 
                      ggplot2::aes(x = chr, y = mean_pop, colour = pop),
                      size = 7, shape = 5) + 
  ggplot2::geom_point(data = meth.diff.df.plot,
                      ggplot2::aes(x = chr, y = perc_methylation, colour = pop),
                      size = 1) +
  ggplot2::geom_text(data = meth.diff.df.plot.label,
                     ggplot2::aes(x = chr, y = 110, label = nmarkers),
                     size = 4) +
  ggplot2::geom_point(data = meth.diff.df.plot.locus, 
                      ggplot2::aes(x = chr, y = mean_pop, colour = pop),
                      size = 3, shape = 15) +
    ggplot2::theme(
        panel.background = ggplot2::element_blank(),
        panel.border = ggplot2::element_blank(),
        panel.grid = ggplot2::element_blank(),
        axis.line = ggplot2::element_line("gray25"),
        text = ggplot2::element_text(size = 12),
        axis.text = ggplot2::element_text(angle = 90,size = 10, colour = "gray25"),
        axis.title = ggplot2::element_text(size = 14, colour = "gray25"),
        legend.title = ggplot2::element_text(size = 14),
        legend.text = ggplot2::element_text(size = 10),
        legend.key = ggplot2::element_blank())

print(ALL.plot)

ggplot2::ggsave(plot = ALL.plot, 
                filename = paste0(figure.path,"TOA_ALLcpg_methDiff_EANT-ROSS.png"),
                width = 30, height = 15, units = "cm", device = "png")
```


##### Heatmap

```{r}
meth.diff.df <- as.data.frame(t(methylKit::percMethylation(meth.diff)))
loc.names <- paste0(meth.diff$chr,"_",meth.diff$start)
colnames(meth.diff.df) <- loc.names

meth.diff.df$individual <- rownames(meth.diff.df) 
metadata.new <- metadata[metadata$SAMPLE_ID %in% meth.diff@sample.ids,]
metadata.new <- metadata.new[order(match(metadata.new$SAMPLE_ID, meth.diff@sample.ids)),]
identical(metadata.new$SAMPLE_ID, meth.diff@sample.ids)
identical(metadata.new$SAMPLE_ID,meth.diff.df$individual)
meth.diff.df$pop <- metadata.new$region


heat.plot.data <- meth.diff.df %>%
  tibble::as_tibble() %>%
  tidyr::gather(key = "CpG", value = "Methylation_percentage", -individual, -pop,
                factor_key = FALSE) %>%
  dplyr::group_by(pop, CpG) %>%
  dplyr::mutate(CpG_mean = mean(Methylation_percentage, na.rm = TRUE)) %>%
  dplyr::arrange(pop, individual, CpG) %>%
  dplyr::mutate(individual = factor(individual, levels = unique(individual))) %>%
  dplyr::mutate(CpG = factor(CpG, levels = unique(CpG)))
cpgs <- heat.plot.data$CpG[heat.plot.data$individual == heat.plot.data$individual[1]]
cpgs_mean <- heat.plot.data$CpG_mean[heat.plot.data$individual == heat.plot.data$individual[1]]
cpgs <- cpgs[order(cpgs_mean)]
heat.plot.data$CpG <- factor(heat.plot.data$CpG, levels = unique(cpgs))
heat.plot.data <- heat.plot.data[order(heat.plot.data$CpG),]

heat.plot <- ggplot2::ggplot(data = heat.plot.data) + 
    ggplot2::geom_tile(ggplot2::aes(individual, y = CpG, fill = Methylation_percentage)) +
    ggplot2::scale_fill_gradient2(low = "blue", mid = "yellow", high = "red", midpoint = 50) +
    
    ggnewscale::new_scale_fill() +
    # ggplot2::geom_tile(ggplot2::aes(individual, y = "pop", fill = pop, height = 10)) +
    # ggplot2::scale_fill_manual(values = c("#4d6b40","#adcf9f", "#c5d6e6", "#08306B")) +
    # ggplot2::scale_fill_manual(values = c( "#c5d6e6", "#08306B")) +
    ggplot2::scale_y_discrete(drop = FALSE) +
    ggplot2::theme(legend.position = "top",
                   axis.title.y = ggplot2::element_blank(),
          axis.text.x = ggplot2::element_text(face = "bold", color = "black", 
                                        size = 10, angle = 90, hjust = 1, vjust = 0))
print(heat.plot)

ggplot2::ggsave(heat.plot, bg = "white",
                filename = paste0(figure.path,"TOA_ALLcpg_methDiff_EANT_ROSS_HEATMAP.png"), 
                width = 45, height = 30, units = "cm")
```


#### PCA - HYPER

```{r}
myDiff25p.hyper.gr <- as(myDiff25p.hyper, "GRanges")
meth.diff <- methylKit::selectByOverlap(meth.ER, myDiff25p.hyper.gr) 

methylKit::PCASamples(meth.diff, screeplot = TRUE)
dev.print(file = paste0(figure.path,
                        "PCA_plot_diffMeth_HYPER_Region_subset_EANT-ROSS_scree.png"),
          device = png, res = 300,width = 30,height = 15,units = "cm")


methylKit::PCASamples(meth.diff, comp = c(1,2))
dev.print(file = paste0(figure.path,
                        "PCA_plot_diffMeth_HYPER_Region_subset_EANT-ROSS_PC12.png"
                        ),
          device = png, res = 300,width = 30,height = 15,units = "cm")

methylKit::PCASamples(meth.diff, comp = c(1,3))
dev.print(file = paste0(figure.path,
                        "PCA_plot_diffMeth_HYPER_Region_subset_EANT-ROSS_PC13.png"
                        ),
          device = png, res = 300,width = 30,height = 15,units = "cm")

methylKit::PCASamples(meth.diff, comp = c(2,3))
dev.print(file = paste0(figure.path,
                        "PCA_plot_diffMeth_HYPER_Region_subset_EANT-ROSS_PC23.png"
                        ),
          device = png, res = 300,width = 30,height = 15,units = "cm")
```


##### Plot CpG difference

```{r}
myDiff25p.hyper.gr <- as(myDiff25p.hyper, "GRanges")
hist(myDiff25p.hyper.gr$meth.diff)
max(myDiff25p.hyper.gr$meth.diff) #33.5 %

meth.diff <- methylKit::selectByOverlap(meth.ER, myDiff25p.hyper.gr)
length(unique(meth.diff$chr)) #25
meth.diff.df <- as.data.frame(t(methylKit::percMethylation(meth.diff)))
loc.names <- paste0(meth.diff$chr,"_",meth.diff$start)
colnames(meth.diff.df) <- loc.names

meth.diff.df$individual <- rownames(meth.diff.df) 
identical(metadata3$SAMPLE_ID,meth.diff.df$individual)
meth.diff.df$pop <- metadata3$region

meth.diff.df.plot <- tidyr::gather(meth.diff.df, key = locus, 
                           value = perc_methylation, -individual, -pop,
                           factor_key = TRUE) %>%
  dplyr::mutate(chr = stringr::str_split_fixed(locus, 
                                               pattern = "_", 
                                               n = 2)[,1])
meth.diff.df.plot.locus <- meth.diff.df.plot %>% dplyr::group_by(chr,locus,pop) %>% 
  dplyr::summarise(mean_pop = mean(perc_methylation, na.rm = TRUE),
            .groups = 'drop')

meth.diff.df.plot.centroid <- meth.diff.df.plot %>% dplyr::group_by(chr,pop) %>% 
  dplyr::summarise(mean_pop = mean(perc_methylation, na.rm = TRUE),
            .groups = 'drop')

meth.diff.df.plot.label <- meth.diff.df.plot %>% dplyr::group_by(chr) %>% 
  dplyr::summarise(nmarkers = length(unique(locus)),
            .groups = 'drop')


HYPER.plot <- ggplot2::ggplot() + 
  ggplot2::geom_violin(data = meth.diff.df.plot, 
                       ggplot2::aes(x = chr, y = perc_methylation)) + 
  ggplot2::geom_point(data = meth.diff.df.plot.centroid, 
                      ggplot2::aes(x = chr, y = mean_pop, colour = pop),
                      size = 7, shape = 5) + 
  ggplot2::geom_point(data = meth.diff.df.plot,
                      ggplot2::aes(x = chr, y = perc_methylation, colour = pop),
                      size = 1) +
  ggplot2::geom_text(data = meth.diff.df.plot.label,
                     ggplot2::aes(x = chr, y = 110, label = nmarkers),
                     size = 4) +
  ggplot2::geom_point(data = meth.diff.df.plot.locus, 
                      ggplot2::aes(x = chr, y = mean_pop, colour = pop),
                      size = 3, shape = 15) +
    ggplot2::theme(
        panel.background = ggplot2::element_blank(),
        panel.border = ggplot2::element_blank(),
        panel.grid = ggplot2::element_blank(),
        axis.line = ggplot2::element_line("gray25"),
        text = ggplot2::element_text(size = 12),
        axis.text = ggplot2::element_text(angle = 90,size = 10, colour = "gray25"),
        axis.title = ggplot2::element_text(size = 14, colour = "gray25"),
        legend.title = ggplot2::element_text(size = 14),
        legend.text = ggplot2::element_text(size = 10),
        legend.key = ggplot2::element_blank())

print(HYPER.plot)

ggplot2::ggsave(plot = HYPER.plot, 
                filename = paste0(figure.path,"TOA_HYPERcpg_methDiff_EANT-ROSS.png"),
                width = 30, height = 15, units = "cm", device = "png")
```

#### PCA - HYPO

```{r}
myDiff25p.hypo.gr <- as(myDiff25p.hypo, "GRanges")
meth.diff <- methylKit::selectByOverlap(meth.ER, myDiff25p.hypo.gr) 

methylKit::PCASamples(meth.diff, screeplot = TRUE)
dev.print(file = paste0(figure.path,
                        "PCA_plot_diffMeth_HYPO_Region_subset_EANT-ROSS_scree.png"),
          device = png, res = 300,width = 30,height = 15,units = "cm")

methylKit::PCASamples(meth.diff, comp = c(1,2))
dev.print(file = paste0(figure.path,
                        "PCA_plot_diffMeth_HYPO_Region_subset_EANT-ROSS_PC12.png"
                        ),
          device = png, res = 300,width = 30,height = 15,units = "cm")

methylKit::PCASamples(meth.diff, comp = c(1,3))
dev.print(file = paste0(figure.path,
                        "PCA_plot_diffMeth_HYPO_Region_subset_EANT-ROSS_PC13.png"
                        ),
          device = png, res = 300,width = 30,height = 15,units = "cm")

methylKit::PCASamples(meth.diff, comp = c(2,3))
dev.print(file = paste0(figure.path,
                        "PCA_plot_diffMeth_HYPO_Region_subset_EANT-ROSS_PC23.png"
                        ),
          device = png, res = 300,width = 30,height = 15,units = "cm")
```


##### Plot CpG difference

```{r}
myDiff25p.hypo.gr <- as(myDiff25p.hypo, "GRanges")
hist(myDiff25p.hypo.gr$meth.diff)
min(myDiff25p.hypo.gr$meth.diff)
max(myDiff25p.hypo.gr$meth.diff)

meth.diff <- methylKit::selectByOverlap(meth.ER, myDiff25p.hypo.gr)
length(unique(meth.diff$chr)) #25
meth.diff.df <- as.data.frame(t(methylKit::percMethylation(meth.diff)))
loc.names <- paste0(meth.diff$chr,"_",meth.diff$start)
colnames(meth.diff.df) <- loc.names

meth.diff.df$individual <- rownames(meth.diff.df) 
identical(metadata3$SAMPLE_ID, meth.diff.df$individual)
meth.diff.df$pop <- metadata3$region

meth.diff.df.plot <- tidyr::gather(meth.diff.df, key = locus, 
                           value = perc_methylation, -individual, -pop,
                           factor_key = TRUE) %>%
  dplyr::mutate(chr = stringr::str_split_fixed(locus, 
                                               pattern = "_", 
                                               n = 2)[,1])
meth.diff.df.plot.locus <- meth.diff.df.plot %>% dplyr::group_by(chr,locus,pop) %>% 
  dplyr::summarise(mean_pop = mean(perc_methylation, na.rm = TRUE),
            .groups = 'drop')

meth.diff.df.plot.centroid <- meth.diff.df.plot %>% dplyr::group_by(chr,pop) %>% 
  dplyr::summarise(mean_pop = mean(perc_methylation, na.rm = TRUE),
            .groups = 'drop')

meth.diff.df.plot.label <- meth.diff.df.plot %>% dplyr::group_by(chr) %>% 
  dplyr::summarise(nmarkers = length(unique(locus)),
            .groups = 'drop')


HYPO.plot <- ggplot2::ggplot() + 
  ggplot2::geom_violin(data = meth.diff.df.plot, 
                       ggplot2::aes(x = chr, y = perc_methylation)) + 
  ggplot2::geom_point(data = meth.diff.df.plot.centroid, 
                      ggplot2::aes(x = chr, y = mean_pop, colour = pop),
                      size = 7, shape = 5) + 
  ggplot2::geom_point(data = meth.diff.df.plot,
                      ggplot2::aes(x = chr, y = perc_methylation, colour = pop),
                      size = 1) +
  ggplot2::geom_text(data = meth.diff.df.plot.label,
                     ggplot2::aes(x = chr, y = 110, label = nmarkers),
                     size = 4) +
  ggplot2::geom_point(data = meth.diff.df.plot.locus, 
                      ggplot2::aes(x = chr, y = mean_pop, colour = pop),
                      size = 3, shape = 15) +
    ggplot2::theme(
        panel.background = ggplot2::element_blank(),
        panel.border = ggplot2::element_blank(),
        panel.grid = ggplot2::element_blank(),
        axis.line = ggplot2::element_line("gray25"),
        text = ggplot2::element_text(size = 12),
        axis.text = ggplot2::element_text(angle = 90,size = 10, colour = "gray25"),
        axis.title = ggplot2::element_text(size = 14, colour = "gray25"),
        legend.title = ggplot2::element_text(size = 14),
        legend.text = ggplot2::element_text(size = 10),
        legend.key = ggplot2::element_blank())

print(HYPO.plot)

ggplot2::ggsave(plot = HYPO.plot, 
                filename = paste0(figure.path,"TOA_HYPOcpg_methDiff_EANT-ROSS.png"),
                width = 30, height = 15, units = "cm", device = "png")
```



### EANT4.1 - ROSS
#### Prepare data

```{r}
print(load(paste0(rdata.path,"Toothfish_metadata.Rdata")))

metadata3 <- metadata2[metadata2$region %in% c("58-4-1", "88-1"),]
table(metadata3$FAO2)
print(load(paste0(rdata.path,"Toothfish_methylation_objects_raw.Rdata")))

filtered.myobj3 <- methylKit::reorganize(
  methylObj = filtered.myobj,
  sample.ids = metadata3$SAMPLE_ID,
  treatment = as.numeric(factor(metadata3$FAO2)),
  chunk.size = 1e+06,
  save.db = TRUE,
  suffix = "subset_EANT1-ROSS_raw"
)

# filtered.myobj3@treatment <- as.numeric(factor(metadata3$region))
filtered.myobj3 <- methylKit::filterByCoverage(filtered.myobj3,lo.count = 10,lo.perc = NULL,
                                      hi.count = NULL, hi.perc = 99)

meth.ER1 <- methylKit::unite(filtered.myobj3, destrand = FALSE,
                             save.db = TRUE,
                             suffix = "subset_EANT1-ROSS",
                             min.per.group = 18L, #missing data allowed
                             mc.cores = 25)
#940,601    CpGs after coverage filter

# save(meth.HR,meth.EANT,meth.EH,meth.ER,meth.ER1,
#      file = paste0(rdata.path,
#                    "Toothfish_subset_pairwise_methylation_objects_raw.Rdata"))
```

#### diffMethPerChr

```{r}
print(load(paste0(rdata.path,"Toothfish_subset_pairwise_methylation_objects_raw.Rdata")))
metadata3 <- metadata2[metadata2$region %in% c("58-4-1", "88-1"),]
covariates <- dplyr::select(metadata3, Sex, mat, Age)
myDiff <- methylKit::calculateDiffMeth(meth.ER1,
                                       covariates = covariates,
                                       adjust = "fdr",
                                       chunk.size = 1e+06,
                                       mc.cores = 10)
meth.diff.chr <- methylKit::diffMethPerChr(myDiff,plot = TRUE, 
                                           qvalue.cutoff = 0.01, meth.cutoff = 25)
myDiff25p.hyper <- methylKit::getMethylDiff(myDiff, difference = 25,qvalue = 0.01, type = "hyper") #79  
myDiff25p.hypo <- methylKit::getMethylDiff(myDiff, difference = 25,qvalue = 0.01, type = "hypo") #15 
myDiff25p <- methylKit::getMethylDiff(myDiff, difference = 25, qvalue = 0.01)

# save(myDiff,myDiff25p, myDiff25p.hyper, myDiff25p.hypo, 
#      file = paste0(rdata.path,
#                    "Toothfish_DMS_Region_subset_EANT1-ROSS.Rdata"))


myDiff.df <- as.data.frame(as(myDiff, "GRanges"))
myDiff25p.df <- as.data.frame(as(myDiff25p, "GRanges"))
volcano.plot <- ggplot2::ggplot() +
  ggplot2::geom_point(data = myDiff.df, ggplot2::aes(x = meth.diff, 
                                                  y = -log10(pvalue))) +
  ggplot2::geom_point(data = myDiff25p.df, ggplot2::aes(x = meth.diff, 
                                                  y = -log10(pvalue)),
                      colour = "red")
print(volcano.plot)

ggplot2::ggsave(volcano.plot, bg = "white",
                filename = paste0(figure.path,
                                  "TOA_ALL_DMS25_VOLCANO_methylkit_regions_subset_EANT1-ROSS.png"
                                  ), 
                width = 30, height = 15, units = "cm")
```

#### Load data

```{r}
print(load(paste0(rdata.path,"Toothfish_DMS_Region_subset_EANT1-ROSS.Rdata")))

myDiff25p.gr <- as(myDiff25p, "GRanges")
meth.diff <- methylKit::selectByOverlap(meth.ER1, myDiff25p.gr) 

# DMS <- data.frame(DMS = NULL, region_pw = NULL)
DMS.ER1 <- data.frame(DMS = paste0(meth.diff$chr,"_", meth.diff$start),
                  region_pw = "EANT4.1_ROSS")
DMS <- rbind(DMS, DMS.ER1)

# save(DMS, DMS.HR,DMS.EANT,DMS.EH,DMS.EH1, DMS.EH2,DMS.ER,DMS.ER1,
#      file = paste0(rdata.path,"Toothfish_DMS_CpGsites.Rdata"))
```


#### PCA - ALL

```{r}
print(load(paste0(rdata.path,"Toothfish_DMS_Region_subset_EANT1-ROSS.Rdata")))

myDiff25p.gr <- as(myDiff25p, "GRanges")
meth.diff <- methylKit::selectByOverlap(meth.ER1, myDiff25p.gr) 

methylKit::PCASamples(meth.diff, screeplot = TRUE)
dev.print(file = paste0(figure.path,
                        "PCA_plot_diffMeth_ALL_Region_subset_EANT1-ROSS_scree.png"),
          device = png, res = 300,width = 30,height = 15,units = "cm")

methylKit::PCASamples(meth.diff, comp = c(1,2))
dev.print(file = paste0(figure.path,
                        "PCA_plot_diffMeth_ALL_Region_subset_EANT1-ROSS_PC12.png"
                        ),
          device = png, res = 300,width = 30,height = 15,units = "cm")

methylKit::PCASamples(meth.diff, comp = c(1,3))
dev.print(file = paste0(figure.path,
                        "PCA_plot_diffMeth_ALL_Region_subset_EANT1-ROSS_PC13.png"
                        ),
          device = png, res = 300,width = 30,height = 15,units = "cm")
methylKit::PCASamples(meth.diff, comp = c(2,3))
dev.print(file = paste0(figure.path,
                        "PCA_plot_diffMeth_ALL_Region_subset_EANT1-ROSS_PC23.png"
                        ),
          device = png, res = 300,width = 30,height = 15,units = "cm")
```


##### Plot CpG difference

```{r}
myDiff25p.gr <- as(myDiff25p, "GRanges")
hist(myDiff25p.gr$meth.diff)
min(myDiff25p.gr$meth.diff)#-33 %
max(myDiff25p.gr$meth.diff)#34 %

meth.diff <- methylKit::selectByOverlap(meth.ER1, myDiff25p.gr)
length(unique(meth.diff$chr)) #49
meth.diff.df <- as.data.frame(t(methylKit::percMethylation(meth.diff)))
loc.names <- paste0(meth.diff$chr,"_",meth.diff$start)
colnames(meth.diff.df) <- loc.names

meth.diff.df$individual <- rownames(meth.diff.df) 
metadata.new <- metadata[metadata$SAMPLE_ID %in% meth.diff@sample.ids,]
metadata.new <- metadata.new[order(match(metadata.new$SAMPLE_ID, meth.diff@sample.ids)),]
identical(metadata.new$SAMPLE_ID, meth.diff@sample.ids)
identical(metadata.new$SAMPLE_ID,meth.diff.df$individual)
meth.diff.df$pop <- metadata.new$region

meth.diff.df.plot <- tidyr::gather(meth.diff.df, key = locus, 
                           value = perc_methylation, -individual, -pop,
                           factor_key = TRUE) %>%
  dplyr::mutate(chr = stringr::str_split_fixed(locus, 
                                               pattern = "_", 
                                               n = 2)[,1])

meth.diff.df.plot.locus <- meth.diff.df.plot %>% dplyr::group_by(chr,locus,pop) %>% 
  dplyr::summarise(mean_pop = mean(perc_methylation, na.rm = TRUE),
            .groups = 'drop')

meth.diff.df.plot.centroid <- meth.diff.df.plot %>% dplyr::group_by(chr,pop) %>% 
  dplyr::summarise(mean_pop = mean(perc_methylation, na.rm = TRUE),
            .groups = 'drop')

meth.diff.df.plot.label <- meth.diff.df.plot %>% dplyr::group_by(chr) %>% 
  dplyr::summarise(nmarkers = length(unique(locus)),
            .groups = 'drop')


ALL.plot <- ggplot2::ggplot() + 
  ggplot2::geom_violin(data = meth.diff.df.plot, 
                       ggplot2::aes(x = chr, y = perc_methylation)) + 
  ggplot2::geom_point(data = meth.diff.df.plot.centroid, 
                      ggplot2::aes(x = chr, y = mean_pop, colour = pop),
                      size = 7, shape = 5) + 
  ggplot2::geom_point(data = meth.diff.df.plot,
                      ggplot2::aes(x = chr, y = perc_methylation, colour = pop),
                      size = 1) +
  ggplot2::geom_text(data = meth.diff.df.plot.label,
                     ggplot2::aes(x = chr, y = 110, label = nmarkers),
                     size = 4) +
  ggplot2::geom_point(data = meth.diff.df.plot.locus, 
                      ggplot2::aes(x = chr, y = mean_pop, colour = pop),
                      size = 3, shape = 15) +
    ggplot2::theme(
        panel.background = ggplot2::element_blank(),
        panel.border = ggplot2::element_blank(),
        panel.grid = ggplot2::element_blank(),
        axis.line = ggplot2::element_line("gray25"),
        text = ggplot2::element_text(size = 12),
        axis.text = ggplot2::element_text(angle = 90,size = 10, colour = "gray25"),
        axis.title = ggplot2::element_text(size = 14, colour = "gray25"),
        legend.title = ggplot2::element_text(size = 14),
        legend.text = ggplot2::element_text(size = 10),
        legend.key = ggplot2::element_blank())

print(ALL.plot)

ggplot2::ggsave(plot = ALL.plot, 
                filename = paste0(figure.path,"TOA_ALLcpg_methDiff_EANT1-ROSS.png"),
                width = 30, height = 15, units = "cm", device = "png")
```


##### Heatmap

```{r}
meth.diff.df <- as.data.frame(t(methylKit::percMethylation(meth.diff)))
loc.names <- paste0(meth.diff$chr,"_",meth.diff$start)
colnames(meth.diff.df) <- loc.names

meth.diff.df$individual <- rownames(meth.diff.df) 
metadata.new <- metadata[metadata$SAMPLE_ID %in% meth.diff@sample.ids,]
metadata.new <- metadata.new[order(match(metadata.new$SAMPLE_ID, meth.diff@sample.ids)),]
identical(metadata.new$SAMPLE_ID, meth.diff@sample.ids)
identical(metadata.new$SAMPLE_ID,meth.diff.df$individual)
meth.diff.df$pop <- metadata.new$region


heat.plot.data <- meth.diff.df %>%
  tibble::as_tibble() %>%
  tidyr::gather(key = "CpG", value = "Methylation_percentage", -individual, -pop,
                factor_key = FALSE) %>%
  dplyr::group_by(pop, CpG) %>%
  dplyr::mutate(CpG_mean = mean(Methylation_percentage, na.rm = TRUE)) %>%
  dplyr::arrange(pop, individual, CpG) %>%
  dplyr::mutate(individual = factor(individual, levels = unique(individual))) %>%
  dplyr::mutate(CpG = factor(CpG, levels = unique(CpG)))
cpgs <- heat.plot.data$CpG[heat.plot.data$individual == heat.plot.data$individual[1]]
cpgs_mean <- heat.plot.data$CpG_mean[heat.plot.data$individual == heat.plot.data$individual[1]]
cpgs <- cpgs[order(cpgs_mean)]
heat.plot.data$CpG <- factor(heat.plot.data$CpG, levels = unique(cpgs))
heat.plot.data <- heat.plot.data[order(heat.plot.data$CpG),]

heat.plot <- ggplot2::ggplot(data = heat.plot.data) + 
    ggplot2::geom_tile(ggplot2::aes(individual, y = CpG, fill = Methylation_percentage)) +
    ggplot2::scale_fill_gradient2(low = "blue", mid = "yellow", high = "red", midpoint = 50) +
    
    ggnewscale::new_scale_fill() +
    # ggplot2::geom_tile(ggplot2::aes(individual, y = "pop", fill = pop, height = 10)) +
    # ggplot2::scale_fill_manual(values = c("#4d6b40","#adcf9f", "#c5d6e6", "#08306B")) +
    # ggplot2::scale_fill_manual(values = c( "#c5d6e6", "#08306B")) +
    ggplot2::scale_y_discrete(drop = FALSE) +
    ggplot2::theme(legend.position = "top",
                   axis.title.y = ggplot2::element_blank(),
          axis.text.x = ggplot2::element_text(face = "bold", color = "black", 
                                        size = 10, angle = 90, hjust = 1, vjust = 0))
print(heat.plot)

ggplot2::ggsave(heat.plot, bg = "white",
                filename = paste0(figure.path,"TOA_ALLcpg_methDiff_EANT1_ROSS_HEATMAP.png"), 
                width = 45, height = 30, units = "cm")
```


#### PCA - HYPER

```{r}
myDiff25p.hyper.gr <- as(myDiff25p.hyper, "GRanges")
meth.diff <- methylKit::selectByOverlap(meth.ER1, myDiff25p.hyper.gr) 

methylKit::PCASamples(meth.diff, screeplot = TRUE)
dev.print(file = paste0(figure.path,
                        "PCA_plot_diffMeth_HYPER_Region_subset_EANT1-ROSS_scree.png"),
          device = png, res = 300,width = 30,height = 15,units = "cm")

methylKit::PCASamples(meth.diff, comp = c(1,2))
dev.print(file = paste0(figure.path,
                        "PCA_plot_diffMeth_HYPER_Region_subset_EANT1-ROSS_PC12.png"
                        ),
          device = png, res = 300,width = 30,height = 15,units = "cm")

methylKit::PCASamples(meth.diff, comp = c(1,3))
dev.print(file = paste0(figure.path,
                        "PCA_plot_diffMeth_HYPER_Region_subset_EANT1-ROSS_PC13.png"
                        ),
          device = png, res = 300,width = 30,height = 15,units = "cm")

methylKit::PCASamples(meth.diff, comp = c(2,3))
dev.print(file = paste0(figure.path,
                        "PCA_plot_diffMeth_HYPER_Region_subset_EANT1-ROSS_PC23.png"
                        ),
          device = png, res = 300,width = 30,height = 15,units = "cm")
```


##### Plot CpG difference

```{r}
myDiff25p.hyper.gr <- as(myDiff25p.hyper, "GRanges")
hist(myDiff25p.hyper.gr$meth.diff)
max(myDiff25p.hyper.gr$meth.diff) #33.5 %

meth.diff <- methylKit::selectByOverlap(meth.ER1, myDiff25p.hyper.gr)
length(unique(meth.diff$chr)) #25
meth.diff.df <- as.data.frame(t(methylKit::percMethylation(meth.diff)))
loc.names <- paste0(meth.diff$chr,"_",meth.diff$start)
colnames(meth.diff.df) <- loc.names

meth.diff.df$individual <- rownames(meth.diff.df) 
identical(metadata3$SAMPLE_ID,meth.diff.df$individual)
meth.diff.df$pop <- metadata3$region

meth.diff.df.plot <- tidyr::gather(meth.diff.df, key = locus, 
                           value = perc_methylation, -individual, -pop,
                           factor_key = TRUE) %>%
  dplyr::mutate(chr = stringr::str_split_fixed(locus, 
                                               pattern = "_", 
                                               n = 2)[,1])
meth.diff.df.plot.locus <- meth.diff.df.plot %>% dplyr::group_by(chr,locus,pop) %>% 
  dplyr::summarise(mean_pop = mean(perc_methylation, na.rm = TRUE),
            .groups = 'drop')

meth.diff.df.plot.centroid <- meth.diff.df.plot %>% dplyr::group_by(chr,pop) %>% 
  dplyr::summarise(mean_pop = mean(perc_methylation, na.rm = TRUE),
            .groups = 'drop')

meth.diff.df.plot.label <- meth.diff.df.plot %>% dplyr::group_by(chr) %>% 
  dplyr::summarise(nmarkers = length(unique(locus)),
            .groups = 'drop')


HYPER.plot <- ggplot2::ggplot() + 
  ggplot2::geom_violin(data = meth.diff.df.plot, 
                       ggplot2::aes(x = chr, y = perc_methylation)) + 
  ggplot2::geom_point(data = meth.diff.df.plot.centroid, 
                      ggplot2::aes(x = chr, y = mean_pop, colour = pop),
                      size = 7, shape = 5) + 
  ggplot2::geom_point(data = meth.diff.df.plot,
                      ggplot2::aes(x = chr, y = perc_methylation, colour = pop),
                      size = 1) +
  ggplot2::geom_text(data = meth.diff.df.plot.label,
                     ggplot2::aes(x = chr, y = 110, label = nmarkers),
                     size = 4) +
  ggplot2::geom_point(data = meth.diff.df.plot.locus, 
                      ggplot2::aes(x = chr, y = mean_pop, colour = pop),
                      size = 3, shape = 15) +
    ggplot2::theme(
        panel.background = ggplot2::element_blank(),
        panel.border = ggplot2::element_blank(),
        panel.grid = ggplot2::element_blank(),
        axis.line = ggplot2::element_line("gray25"),
        text = ggplot2::element_text(size = 12),
        axis.text = ggplot2::element_text(angle = 90,size = 10, colour = "gray25"),
        axis.title = ggplot2::element_text(size = 14, colour = "gray25"),
        legend.title = ggplot2::element_text(size = 14),
        legend.text = ggplot2::element_text(size = 10),
        legend.key = ggplot2::element_blank())

print(HYPER.plot)

ggplot2::ggsave(plot = HYPER.plot, 
                filename = paste0(figure.path,"TOA_HYPERcpg_methDiff_EANT1-ROSS.png"),
                width = 30, height = 15, units = "cm", device = "png")
```

#### PCA - HYPO

```{r}
myDiff25p.hypo.gr <- as(myDiff25p.hypo, "GRanges")
meth.diff <- methylKit::selectByOverlap(meth.ER1, myDiff25p.hypo.gr) 

methylKit::PCASamples(meth.diff, screeplot = TRUE)
dev.print(file = paste0(figure.path,
                        "PCA_plot_diffMeth_HYPO_Region_subset_EANT1-ROSS_scree.png"),
          device = png, res = 300,width = 30,height = 15,units = "cm")

methylKit::PCASamples(meth.diff, comp = c(1,2))
dev.print(file = paste0(figure.path,
                        "PCA_plot_diffMeth_HYPO_Region_subset_EANT1-ROSS_PC12.png"
                        ),
          device = png, res = 300,width = 30,height = 15,units = "cm")

methylKit::PCASamples(meth.diff, comp = c(1,3))
dev.print(file = paste0(figure.path,
                        "PCA_plot_diffMeth_HYPO_Region_subset_EANT1-ROSS_PC13.png"
                        ),
          device = png, res = 300,width = 30,height = 15,units = "cm")

methylKit::PCASamples(meth.diff, comp = c(2,3))
dev.print(file = paste0(figure.path,
                        "PCA_plot_diffMeth_HYPO_Region_subset_EANT1-ROSS_PC23.png"
                        ),
          device = png, res = 300,width = 30,height = 15,units = "cm")
```


##### Plot CpG difference

```{r}
myDiff25p.hypo.gr <- as(myDiff25p.hypo, "GRanges")
hist(myDiff25p.hypo.gr$meth.diff)
min(myDiff25p.hypo.gr$meth.diff)
max(myDiff25p.hypo.gr$meth.diff)

meth.diff <- methylKit::selectByOverlap(meth.ER1, myDiff25p.hypo.gr)
length(unique(meth.diff$chr)) #25
meth.diff.df <- as.data.frame(t(methylKit::percMethylation(meth.diff)))
loc.names <- paste0(meth.diff$chr,"_",meth.diff$start)
colnames(meth.diff.df) <- loc.names

meth.diff.df$individual <- rownames(meth.diff.df) 
identical(metadata3$SAMPLE_ID, meth.diff.df$individual)
meth.diff.df$pop <- metadata3$region

meth.diff.df.plot <- tidyr::gather(meth.diff.df, key = locus, 
                           value = perc_methylation, -individual, -pop,
                           factor_key = TRUE) %>%
  dplyr::mutate(chr = stringr::str_split_fixed(locus, 
                                               pattern = "_", 
                                               n = 2)[,1])
meth.diff.df.plot.locus <- meth.diff.df.plot %>% dplyr::group_by(chr,locus,pop) %>% 
  dplyr::summarise(mean_pop = mean(perc_methylation, na.rm = TRUE),
            .groups = 'drop')

meth.diff.df.plot.centroid <- meth.diff.df.plot %>% dplyr::group_by(chr,pop) %>% 
  dplyr::summarise(mean_pop = mean(perc_methylation, na.rm = TRUE),
            .groups = 'drop')

meth.diff.df.plot.label <- meth.diff.df.plot %>% dplyr::group_by(chr) %>% 
  dplyr::summarise(nmarkers = length(unique(locus)),
            .groups = 'drop')


HYPO.plot <- ggplot2::ggplot() + 
  ggplot2::geom_violin(data = meth.diff.df.plot, 
                       ggplot2::aes(x = chr, y = perc_methylation)) + 
  ggplot2::geom_point(data = meth.diff.df.plot.centroid, 
                      ggplot2::aes(x = chr, y = mean_pop, colour = pop),
                      size = 7, shape = 5) + 
  ggplot2::geom_point(data = meth.diff.df.plot,
                      ggplot2::aes(x = chr, y = perc_methylation, colour = pop),
                      size = 1) +
  ggplot2::geom_text(data = meth.diff.df.plot.label,
                     ggplot2::aes(x = chr, y = 110, label = nmarkers),
                     size = 4) +
  ggplot2::geom_point(data = meth.diff.df.plot.locus, 
                      ggplot2::aes(x = chr, y = mean_pop, colour = pop),
                      size = 3, shape = 15) +
    ggplot2::theme(
        panel.background = ggplot2::element_blank(),
        panel.border = ggplot2::element_blank(),
        panel.grid = ggplot2::element_blank(),
        axis.line = ggplot2::element_line("gray25"),
        text = ggplot2::element_text(size = 12),
        axis.text = ggplot2::element_text(angle = 90,size = 10, colour = "gray25"),
        axis.title = ggplot2::element_text(size = 14, colour = "gray25"),
        legend.title = ggplot2::element_text(size = 14),
        legend.text = ggplot2::element_text(size = 10),
        legend.key = ggplot2::element_blank())

print(HYPO.plot)

ggplot2::ggsave(plot = HYPO.plot, 
                filename = paste0(figure.path,"TOA_HYPOcpg_methDiff_EANT1-ROSS.png"),
                width = 30, height = 15, units = "cm", device = "png")
```






### EANT4.2 - ROSS
#### Prepare data

```{r}
print(load(paste0(rdata.path,"Toothfish_metadata.Rdata")))

metadata3 <- metadata2[metadata2$region %in% c("58-4-2", "88-1"),]
table(metadata3$FAO2)
print(load(paste0(rdata.path,"Toothfish_methylation_objects_raw.Rdata")))

filtered.myobj3 <- methylKit::reorganize(
  methylObj = filtered.myobj,
  sample.ids = metadata3$SAMPLE_ID,
  treatment = as.numeric(factor(metadata3$region)),
  chunk.size = 1e+06,
  save.db = TRUE,
  suffix = "subset_EANT2-ROSS_raw"
)

# filtered.myobj3@treatment <- as.numeric(factor(metadata3$region))
filtered.myobj3 <- methylKit::filterByCoverage(filtered.myobj3,lo.count = 10,
                                               lo.perc = NULL,
                                      hi.count = NULL, hi.perc = 99)

meth.ER2 <- methylKit::unite(filtered.myobj3, destrand = FALSE,
                             save.db = TRUE,
                             suffix = "subset_EANT2-ROSS",
                             min.per.group = 18L, #missing data allowed
                             mc.cores = 25)
#1,251,372 CpGs after coverage filter

save(meth.HR,meth.EANT,meth.EH,
     meth.EH1, meth.EH2,
     meth.ER,meth.ER1,meth.ER2,
     file = paste0(rdata.path,
                   "Toothfish_subset_pairwise_methylation_objects_raw.Rdata"))
```

#### diffMethPerChr

```{r}
print(load(paste0(rdata.path,"Toothfish_subset_pairwise_methylation_objects_raw.Rdata")))
metadata3 <- metadata2[metadata2$region %in% c("58-4-2", "88-1"),]
covariates <- dplyr::select(metadata3, Sex, mat, Age)
myDiff <- methylKit::calculateDiffMeth(meth.ER2,
                                       covariates = covariates,
                                       adjust = "fdr",
                                       chunk.size = 1e+06,
                                       mc.cores = 10)
meth.diff.chr <- methylKit::diffMethPerChr(myDiff,plot = TRUE, 
                                           qvalue.cutoff = 0.01, meth.cutoff = 25)
myDiff25p.hyper <- methylKit::getMethylDiff(myDiff, difference = 25,qvalue = 0.01, type = "hyper") #34  
myDiff25p.hypo <- methylKit::getMethylDiff(myDiff, difference = 25,qvalue = 0.01, type = "hypo") #22 
myDiff25p <- methylKit::getMethylDiff(myDiff, difference = 25, qvalue = 0.01)

# save(myDiff,myDiff25p, myDiff25p.hyper, myDiff25p.hypo, 
#      file = paste0(rdata.path,
#                    "Toothfish_DMS_Region_subset_EANT2-ROSS.Rdata"))


myDiff.df <- as.data.frame(as(myDiff, "GRanges"))
myDiff25p.df <- as.data.frame(as(myDiff25p, "GRanges"))
volcano.plot <- ggplot2::ggplot() +
  ggplot2::geom_point(data = myDiff.df, ggplot2::aes(x = meth.diff, 
                                                  y = -log10(pvalue))) +
  ggplot2::geom_point(data = myDiff25p.df, ggplot2::aes(x = meth.diff, 
                                                  y = -log10(pvalue)),
                      colour = "red")
print(volcano.plot)

ggplot2::ggsave(volcano.plot, bg = "white",
                filename = paste0(figure.path,
                                  "TOA_ALL_DMS25_VOLCANO_methylkit_regions_subset_EANT2-ROSS.png"
                                  ), 
                width = 30, height = 15, units = "cm")
```


#### Load data

```{r}
print(load(paste0(rdata.path,"Toothfish_DMS_Region_subset_EANT2-ROSS.Rdata")))

myDiff25p.gr <- as(myDiff25p, "GRanges")
meth.diff <- methylKit::selectByOverlap(meth.ER2, myDiff25p.gr) 

# DMS <- data.frame(DMS = NULL, region_pw = NULL)
DMS.ER2 <- data.frame(DMS = paste0(meth.diff$chr,"_", meth.diff$start),
                  region_pw = "EANT4.2_ROSS")
DMS <- rbind(DMS, DMS.ER2)

# save(DMS, DMS.HR,DMS.EANT,DMS.EH,DMS.EH1, DMS.EH2,DMS.ER,DMS.ER1,DMS.ER2,
#      file = paste0(rdata.path,"Toothfish_DMS_CpGsites.Rdata"))
```


#### PCA - ALL

```{r}
print(load(paste0(rdata.path,"Toothfish_DMS_Region_subset_EANT2-ROSS.Rdata")))

myDiff25p.gr <- as(myDiff25p, "GRanges")
meth.diff <- methylKit::selectByOverlap(meth.ER2, myDiff25p.gr) 

methylKit::PCASamples(meth.diff, screeplot = TRUE)
dev.print(file = paste0(figure.path,
                        "PCA_plot_diffMeth_ALL_Region_subset_EANT2-ROSS_scree.png"),
          device = png, res = 300,width = 30,height = 15,units = "cm")

methylKit::PCASamples(meth.diff, comp = c(1,2))
dev.print(file = paste0(figure.path,
                        "PCA_plot_diffMeth_ALL_Region_subset_EANT2-ROSS_PC12.png"
                        ),
          device = png, res = 300,width = 30,height = 15,units = "cm")

methylKit::PCASamples(meth.diff, comp = c(1,3))
dev.print(file = paste0(figure.path,
                        "PCA_plot_diffMeth_ALL_Region_subset_EANT2-ROSS_PC13.png"
                        ),
          device = png, res = 300,width = 30,height = 15,units = "cm")

methylKit::PCASamples(meth.diff, comp = c(2,3))
dev.print(file = paste0(figure.path,
                        "PCA_plot_diffMeth_ALL_Region_subset_EANT2-ROSS_PC23.png"
                        ),
          device = png, res = 300,width = 30,height = 15,units = "cm")
```


##### Plot CpG difference

```{r}
myDiff25p.gr <- as(myDiff25p, "GRanges")
hist(myDiff25p.gr$meth.diff)
min(myDiff25p.gr$meth.diff)#-32 %
max(myDiff25p.gr$meth.diff)#34 %

meth.diff <- methylKit::selectByOverlap(meth.ER2, myDiff25p.gr)
length(unique(meth.diff$chr)) #49
meth.diff.df <- as.data.frame(t(methylKit::percMethylation(meth.diff)))
loc.names <- paste0(meth.diff$chr,"_",meth.diff$start)
colnames(meth.diff.df) <- loc.names

meth.diff.df$individual <- rownames(meth.diff.df) 
metadata.new <- metadata[metadata$SAMPLE_ID %in% meth.diff@sample.ids,]
metadata.new <- metadata.new[order(match(metadata.new$SAMPLE_ID, meth.diff@sample.ids)),]
identical(metadata.new$SAMPLE_ID, meth.diff@sample.ids)
identical(metadata.new$SAMPLE_ID,meth.diff.df$individual)
meth.diff.df$pop <- metadata.new$region

meth.diff.df.plot <- tidyr::gather(meth.diff.df, key = locus, 
                           value = perc_methylation, -individual, -pop,
                           factor_key = TRUE) %>%
  dplyr::mutate(chr = stringr::str_split_fixed(locus, 
                                               pattern = "_", 
                                               n = 2)[,1])

meth.diff.df.plot.locus <- meth.diff.df.plot %>% dplyr::group_by(chr,locus,pop) %>% 
  dplyr::summarise(mean_pop = mean(perc_methylation, na.rm = TRUE),
            .groups = 'drop')

meth.diff.df.plot.centroid <- meth.diff.df.plot %>% dplyr::group_by(chr,pop) %>% 
  dplyr::summarise(mean_pop = mean(perc_methylation, na.rm = TRUE),
            .groups = 'drop')

meth.diff.df.plot.label <- meth.diff.df.plot %>% dplyr::group_by(chr) %>% 
  dplyr::summarise(nmarkers = length(unique(locus)),
            .groups = 'drop')


ALL.plot <- ggplot2::ggplot() + 
  ggplot2::geom_violin(data = meth.diff.df.plot, 
                       ggplot2::aes(x = chr, y = perc_methylation)) + 
  ggplot2::geom_point(data = meth.diff.df.plot.centroid, 
                      ggplot2::aes(x = chr, y = mean_pop, colour = pop),
                      size = 7, shape = 5) + 
  ggplot2::geom_point(data = meth.diff.df.plot,
                      ggplot2::aes(x = chr, y = perc_methylation, colour = pop),
                      size = 1) +
  ggplot2::geom_text(data = meth.diff.df.plot.label,
                     ggplot2::aes(x = chr, y = 110, label = nmarkers),
                     size = 4) +
  ggplot2::geom_point(data = meth.diff.df.plot.locus, 
                      ggplot2::aes(x = chr, y = mean_pop, colour = pop),
                      size = 3, shape = 15) +
    ggplot2::theme(
        panel.background = ggplot2::element_blank(),
        panel.border = ggplot2::element_blank(),
        panel.grid = ggplot2::element_blank(),
        axis.line = ggplot2::element_line("gray25"),
        text = ggplot2::element_text(size = 12),
        axis.text = ggplot2::element_text(angle = 90,size = 10, colour = "gray25"),
        axis.title = ggplot2::element_text(size = 14, colour = "gray25"),
        legend.title = ggplot2::element_text(size = 14),
        legend.text = ggplot2::element_text(size = 10),
        legend.key = ggplot2::element_blank())

print(ALL.plot)

ggplot2::ggsave(plot = ALL.plot, 
                filename = paste0(figure.path,"TOA_ALLcpg_methDiff_EANT2-ROSS.png"),
                width = 30, height = 15, units = "cm", device = "png")
```



##### Heatmap

```{r}
meth.diff.df <- as.data.frame(t(methylKit::percMethylation(meth.diff)))
loc.names <- paste0(meth.diff$chr,"_",meth.diff$start)
colnames(meth.diff.df) <- loc.names

meth.diff.df$individual <- rownames(meth.diff.df) 
metadata.new <- metadata[metadata$SAMPLE_ID %in% meth.diff@sample.ids,]
metadata.new <- metadata.new[order(match(metadata.new$SAMPLE_ID, meth.diff@sample.ids)),]
identical(metadata.new$SAMPLE_ID, meth.diff@sample.ids)
identical(metadata.new$SAMPLE_ID,meth.diff.df$individual)
meth.diff.df$pop <- metadata.new$region


heat.plot.data <- meth.diff.df %>%
  tibble::as_tibble() %>%
  tidyr::gather(key = "CpG", value = "Methylation_percentage", -individual, -pop,
                factor_key = FALSE) %>%
  dplyr::group_by(pop, CpG) %>%
  dplyr::mutate(CpG_mean = mean(Methylation_percentage, na.rm = TRUE)) %>%
  dplyr::arrange(pop, individual, CpG) %>%
  dplyr::mutate(individual = factor(individual, levels = unique(individual))) %>%
  dplyr::mutate(CpG = factor(CpG, levels = unique(CpG)))
cpgs <- heat.plot.data$CpG[heat.plot.data$individual == heat.plot.data$individual[1]]
cpgs_mean <- heat.plot.data$CpG_mean[heat.plot.data$individual == heat.plot.data$individual[1]]
cpgs <- cpgs[order(cpgs_mean)]
heat.plot.data$CpG <- factor(heat.plot.data$CpG, levels = unique(cpgs))
heat.plot.data <- heat.plot.data[order(heat.plot.data$CpG),]

heat.plot <- ggplot2::ggplot(data = heat.plot.data) + 
    ggplot2::geom_tile(ggplot2::aes(individual, y = CpG, fill = Methylation_percentage)) +
    ggplot2::scale_fill_gradient2(low = "blue", mid = "yellow", high = "red", midpoint = 50) +
    
    ggnewscale::new_scale_fill() +
    # ggplot2::geom_tile(ggplot2::aes(individual, y = "pop", fill = pop, height = 10)) +
    # ggplot2::scale_fill_manual(values = c("#4d6b40","#adcf9f", "#c5d6e6", "#08306B")) +
    # ggplot2::scale_fill_manual(values = c( "#c5d6e6", "#08306B")) +
    ggplot2::scale_y_discrete(drop = FALSE) +
    ggplot2::theme(legend.position = "top",
                   axis.title.y = ggplot2::element_blank(),
          axis.text.x = ggplot2::element_text(face = "bold", color = "black", 
                                        size = 10, angle = 90, hjust = 1, vjust = 0))
print(heat.plot)

ggplot2::ggsave(heat.plot, bg = "white",
                filename = paste0(figure.path,"TOA_ALLcpg_methDiff_EANT2_ROSS_HEATMAP.png"), 
                width = 45, height = 30, units = "cm")
```


#### PCA - HYPER

```{r}
myDiff25p.hyper.gr <- as(myDiff25p.hyper, "GRanges")
meth.diff <- methylKit::selectByOverlap(meth.ER2, myDiff25p.hyper.gr) 

methylKit::PCASamples(meth.diff, screeplot = TRUE)
dev.print(file = paste0(figure.path,
                        "PCA_plot_diffMeth_HYPER_Region_subset_EANT2-ROSS_scree.png"),
          device = png, res = 300,width = 30,height = 15,units = "cm")

methylKit::PCASamples(meth.diff, comp = c(1,2))
dev.print(file = paste0(figure.path,
                        "PCA_plot_diffMeth_HYPER_Region_subset_EANT2-ROSS_PC12.png"
                        ),
          device = png, res = 300,width = 30,height = 15,units = "cm")

methylKit::PCASamples(meth.diff, comp = c(1,3))
dev.print(file = paste0(figure.path,
                        "PCA_plot_diffMeth_HYPER_Region_subset_EANT2-ROSS_PC13.png"
                        ),
          device = png, res = 300,width = 30,height = 15,units = "cm")

methylKit::PCASamples(meth.diff, comp = c(2,3))
dev.print(file = paste0(figure.path,
                        "PCA_plot_diffMeth_HYPER_Region_subset_EANT2-ROSS_PC23.png"
                        ),
          device = png, res = 300,width = 30,height = 15,units = "cm")
```


##### Plot CpG difference

```{r}
myDiff25p.hyper.gr <- as(myDiff25p.hyper, "GRanges")
hist(myDiff25p.hyper.gr$meth.diff)
max(myDiff25p.hyper.gr$meth.diff) #33.5 %

meth.diff <- methylKit::selectByOverlap(meth.ER2, myDiff25p.hyper.gr)
length(unique(meth.diff$chr)) #25
meth.diff.df <- as.data.frame(t(methylKit::percMethylation(meth.diff)))
loc.names <- paste0(meth.diff$chr,"_",meth.diff$start)
colnames(meth.diff.df) <- loc.names

meth.diff.df$individual <- rownames(meth.diff.df) 
identical(metadata3$SAMPLE_ID,meth.diff.df$individual)
meth.diff.df$pop <- metadata3$region

meth.diff.df.plot <- tidyr::gather(meth.diff.df, key = locus, 
                           value = perc_methylation, -individual, -pop,
                           factor_key = TRUE) %>%
  dplyr::mutate(chr = stringr::str_split_fixed(locus, 
                                               pattern = "_", 
                                               n = 2)[,1])
meth.diff.df.plot.locus <- meth.diff.df.plot %>% dplyr::group_by(chr,locus,pop) %>% 
  dplyr::summarise(mean_pop = mean(perc_methylation, na.rm = TRUE),
            .groups = 'drop')

meth.diff.df.plot.centroid <- meth.diff.df.plot %>% dplyr::group_by(chr,pop) %>% 
  dplyr::summarise(mean_pop = mean(perc_methylation, na.rm = TRUE),
            .groups = 'drop')

meth.diff.df.plot.label <- meth.diff.df.plot %>% dplyr::group_by(chr) %>% 
  dplyr::summarise(nmarkers = length(unique(locus)),
            .groups = 'drop')


HYPER.plot <- ggplot2::ggplot() + 
  ggplot2::geom_violin(data = meth.diff.df.plot, 
                       ggplot2::aes(x = chr, y = perc_methylation)) + 
  ggplot2::geom_point(data = meth.diff.df.plot.centroid, 
                      ggplot2::aes(x = chr, y = mean_pop, colour = pop),
                      size = 7, shape = 5) + 
  ggplot2::geom_point(data = meth.diff.df.plot,
                      ggplot2::aes(x = chr, y = perc_methylation, colour = pop),
                      size = 1) +
  ggplot2::geom_text(data = meth.diff.df.plot.label,
                     ggplot2::aes(x = chr, y = 110, label = nmarkers),
                     size = 4) +
  ggplot2::geom_point(data = meth.diff.df.plot.locus, 
                      ggplot2::aes(x = chr, y = mean_pop, colour = pop),
                      size = 3, shape = 15) +
    ggplot2::theme(
        panel.background = ggplot2::element_blank(),
        panel.border = ggplot2::element_blank(),
        panel.grid = ggplot2::element_blank(),
        axis.line = ggplot2::element_line("gray25"),
        text = ggplot2::element_text(size = 12),
        axis.text = ggplot2::element_text(angle = 90,size = 10, colour = "gray25"),
        axis.title = ggplot2::element_text(size = 14, colour = "gray25"),
        legend.title = ggplot2::element_text(size = 14),
        legend.text = ggplot2::element_text(size = 10),
        legend.key = ggplot2::element_blank())

print(HYPER.plot)

ggplot2::ggsave(plot = HYPER.plot, 
                filename = paste0(figure.path,"TOA_HYPERcpg_methDiff_EANT2-ROSS.png"),
                width = 30, height = 15, units = "cm", device = "png")
```

#### PCA - HYPO

```{r}
myDiff25p.hypo.gr <- as(myDiff25p.hypo, "GRanges")
meth.diff <- methylKit::selectByOverlap(meth.ER2, myDiff25p.hypo.gr) 

methylKit::PCASamples(meth.diff, screeplot = TRUE)
dev.print(file = paste0(figure.path,
                        "PCA_plot_diffMeth_HYPO_Region_subset_EANT2-ROSS_scree.png"),
          device = png, res = 300,width = 30,height = 15,units = "cm")

methylKit::PCASamples(meth.diff, comp = c(1,2))
dev.print(file = paste0(figure.path,
                        "PCA_plot_diffMeth_HYPO_Region_subset_EANT2-ROSS_PC12.png"
                        ),
          device = png, res = 300,width = 30,height = 15,units = "cm")

methylKit::PCASamples(meth.diff, comp = c(1,3))
dev.print(file = paste0(figure.path,
                        "PCA_plot_diffMeth_HYPO_Region_subset_EANT2-ROSS_PC13.png"
                        ),
          device = png, res = 300,width = 30,height = 15,units = "cm")

methylKit::PCASamples(meth.diff, comp = c(2,3))
dev.print(file = paste0(figure.path,
                        "PCA_plot_diffMeth_HYPO_Region_subset_EANT2-ROSS_PC23.png"
                        ),
          device = png, res = 300,width = 30,height = 15,units = "cm")
```


##### Plot CpG difference

```{r}
myDiff25p.hypo.gr <- as(myDiff25p.hypo, "GRanges")
hist(myDiff25p.hypo.gr$meth.diff)
min(myDiff25p.hypo.gr$meth.diff)
max(myDiff25p.hypo.gr$meth.diff)

meth.diff <- methylKit::selectByOverlap(meth.ER2, myDiff25p.hypo.gr)
length(unique(meth.diff$chr)) #25
meth.diff.df <- as.data.frame(t(methylKit::percMethylation(meth.diff)))
loc.names <- paste0(meth.diff$chr,"_",meth.diff$start)
colnames(meth.diff.df) <- loc.names

meth.diff.df$individual <- rownames(meth.diff.df) 
identical(metadata3$SAMPLE_ID, meth.diff.df$individual)
meth.diff.df$pop <- metadata3$region

meth.diff.df.plot <- tidyr::gather(meth.diff.df, key = locus, 
                           value = perc_methylation, -individual, -pop,
                           factor_key = TRUE) %>%
  dplyr::mutate(chr = stringr::str_split_fixed(locus, 
                                               pattern = "_", 
                                               n = 2)[,1])
meth.diff.df.plot.locus <- meth.diff.df.plot %>% dplyr::group_by(chr,locus,pop) %>% 
  dplyr::summarise(mean_pop = mean(perc_methylation, na.rm = TRUE),
            .groups = 'drop')

meth.diff.df.plot.centroid <- meth.diff.df.plot %>% dplyr::group_by(chr,pop) %>% 
  dplyr::summarise(mean_pop = mean(perc_methylation, na.rm = TRUE),
            .groups = 'drop')

meth.diff.df.plot.label <- meth.diff.df.plot %>% dplyr::group_by(chr) %>% 
  dplyr::summarise(nmarkers = length(unique(locus)),
            .groups = 'drop')


HYPO.plot <- ggplot2::ggplot() + 
  ggplot2::geom_violin(data = meth.diff.df.plot, 
                       ggplot2::aes(x = chr, y = perc_methylation)) + 
  ggplot2::geom_point(data = meth.diff.df.plot.centroid, 
                      ggplot2::aes(x = chr, y = mean_pop, colour = pop),
                      size = 7, shape = 5) + 
  ggplot2::geom_point(data = meth.diff.df.plot,
                      ggplot2::aes(x = chr, y = perc_methylation, colour = pop),
                      size = 1) +
  ggplot2::geom_text(data = meth.diff.df.plot.label,
                     ggplot2::aes(x = chr, y = 110, label = nmarkers),
                     size = 4) +
  ggplot2::geom_point(data = meth.diff.df.plot.locus, 
                      ggplot2::aes(x = chr, y = mean_pop, colour = pop),
                      size = 3, shape = 15) +
    ggplot2::theme(
        panel.background = ggplot2::element_blank(),
        panel.border = ggplot2::element_blank(),
        panel.grid = ggplot2::element_blank(),
        axis.line = ggplot2::element_line("gray25"),
        text = ggplot2::element_text(size = 12),
        axis.text = ggplot2::element_text(angle = 90,size = 10, colour = "gray25"),
        axis.title = ggplot2::element_text(size = 14, colour = "gray25"),
        legend.title = ggplot2::element_text(size = 14),
        legend.text = ggplot2::element_text(size = 10),
        legend.key = ggplot2::element_blank())

print(HYPO.plot)

ggplot2::ggsave(plot = HYPO.plot, 
                filename = paste0(figure.path,"TOA_HYPOcpg_methDiff_EANT2-ROSS.png"),
                width = 30, height = 15, units = "cm", device = "png")
```

## Most differentiating CpGs

```{r}
print(load(paste0(rdata.path,"Toothfish_DMS_CpGsites.Rdata")))

DMS.data <- DMS %>%
  dplyr::count(DMS, region_pw) %>%
  tidyr::spread(region_pw, n, fill = 0)
dim(DMS.data)
common.cpg <- rowSums(DMS.data[,-1])
hist(common.cpg)
common.cpg.names <- DMS.data$DMS[common.cpg > 1]

DMS.data[common.cpg > 3,
         colnames(DMS.data) %in% c("HEARD_ROSS","EANT_HEARD","EANT4.1_ROSS","EANT4.2_ROSS")]
```

### PCA
```{r}
print(load(paste0(rdata.path,"Toothfish_subset_methylation_objects_raw.Rdata")))
print(load(paste0(rdata.path,"Toothfish_metadata.Rdata")))



meth@treatment <- as.integer(as.factor(metadata2$region))
common.cpg.names

meth.gr <- as(meth, "GRanges")
meth.df <- as.data.frame(meth.gr)
common.cpg.gr <- meth.gr[
  paste0(meth.df$seqnames,"_",meth.df$start) 
  %in% common.cpg.names]

meth.diff <- methylKit::selectByOverlap(meth, common.cpg.gr) 


PCobj <- methylKit::PCASamples(meth.diff, comp = c(1,2), adj.lim = c(0,0), obj.return = TRUE)

var_frac <- PCobj$sdev/sum(PCobj$sdev)
print(paste0("Total variance explained: ",signif(sum(var_frac[1:3]) * 100, 3), "%")) 
barplot(var_frac[1:20], main = "variance explained", 
        col = heat.colors(length(var_frac[1:20])))

# metadata3 <- metadata2[metadata2$tissue %in% "M",]
metadata.new <- metadata[metadata$SAMPLE_ID %in% meth.diff@sample.ids,]
metadata.new <- metadata.new[order(match(metadata.new$SAMPLE_ID, meth.diff@sample.ids)),]
identical(metadata.new$SAMPLE_ID, meth.diff@sample.ids)
identical(metadata.new$SAMPLE_ID, rownames(PCobj$x))

pop <- metadata.new$region
# pop <- metadata.new$`Life stage`

# pop <- metadata.new$`Date caught`

# colours5 <- adegenet::funky(8)
colours5 <- c("#4d6b40","#adcf9f", "#c5d6e6", "#08306B")
# colours5 <- c("#fc9403","#4d6b40","#adcf9f", "#c5d6e6", "#08306B")

# plot.title <- "Toothfish methylation (1,236,968 CpG sites)"
# plot.title <- "Toothfish methylation (559,543 CpG sites)"
plot.title <- "433 common CpG sites between region pairs"


data <- data.frame(ind = rownames(PCobj$x),PCobj$x,  pop = pop)
data <- merge(data, aggregate(cbind(mean.x = PC1, mean.y = PC2) ~
                                pop, data, mean), by = "pop")

pca.plot <- ggplot2::ggplot(data, ggplot2::aes(PC1, PC2, color = pop)) +
  ggplot2::geom_point(size = 4) + 
  ggplot2::geom_point(ggplot2::aes(x = mean.x,
                                   y = mean.y, color = pop),
                      size = 7,
                      shape = 17) +
  ggplot2::scale_colour_manual(values = colours5) +
  ggplot2::scale_fill_manual(values = colours5) + 
  ggplot2::geom_hline(yintercept = 0) +
  ggplot2::geom_vline(xintercept = 0) + 
  ggplot2::labs(
    subtitle = paste0("Total variance explained: ",
                      signif(sum(var_frac[1:4]) * 100, 3), "%"),
    y = paste0("PC2; ", signif(var_frac[2] * 100, digits = 2), " %"),
    x = paste0("PC1; ", signif(var_frac[1] * 100, digits = 2), " %"),
    title = plot.title,
    caption = "") + 
  ggplot2::theme_minimal() +
  ggplot2::theme(
    axis.text = ggplot2::element_text(size = 15),
    axis.title.x = ggplot2::element_text(size = 20),
    axis.title.y = ggplot2::element_text(size = 20),
    legend.text = ggplot2::element_text(size = 15)
  )

print(pca.plot)
ggplot2::ggsave(pca.plot, bg = "white",
                filename = paste0(figure.path,
                                  "PCA_plot_diffMeth_commonCpG_PC12.png"
                                  ), 
                width = 30, height = 15, units = "cm")



data <- data.frame(ind = rownames(PCobj$x),PCobj$x,  pop = pop)
data <- merge(data, aggregate(cbind(mean.x = PC1, mean.y = PC3) ~
                                pop, data, mean), by = "pop")

pca.plot <- ggplot2::ggplot(data, ggplot2::aes(PC1, PC3, color = pop)) +
  ggplot2::geom_point(size = 4) + 
  ggplot2::geom_point(ggplot2::aes(x = mean.x,
                                   y = mean.y, color = pop),
                      size = 7,
                      shape = 17) +
  ggplot2::scale_colour_manual(values = colours5) +
  ggplot2::scale_fill_manual(values = colours5) + 
  ggplot2::geom_hline(yintercept = 0) +
  ggplot2::geom_vline(xintercept = 0) + 
  ggplot2::labs(
    subtitle = paste0("Total variance explained: ",
                      signif(sum(var_frac[1:4]) * 100, 3), "%"),
    y = paste0("PC3; ", signif(var_frac[3] * 100, digits = 2), " %"),
    x = paste0("PC1; ", signif(var_frac[1] * 100, digits = 2), " %"),
    title = plot.title,
    caption = "") + 
  ggplot2::theme_minimal() +
  ggplot2::theme(
    axis.text = ggplot2::element_text(size = 15),
    axis.title.x = ggplot2::element_text(size = 20),
    axis.title.y = ggplot2::element_text(size = 20),
    legend.text = ggplot2::element_text(size = 15)
  )

print(pca.plot)
ggplot2::ggsave(pca.plot, bg = "white",
                filename = paste0(figure.path,
                                  "PCA_plot_diffMeth_commonCpG_PC13.png"
                                  ), 
                width = 30, height = 15, units = "cm")




data <- data.frame(ind = rownames(PCobj$x),PCobj$x,  pop = pop)
data <- merge(data, aggregate(cbind(mean.x = PC2, mean.y = PC3) ~
                                pop, data, mean), by = "pop")

pca.plot <- ggplot2::ggplot(data, ggplot2::aes(PC2, PC3, color = pop)) +
  ggplot2::geom_point(size = 4) + 
  ggplot2::geom_point(ggplot2::aes(x = mean.x,
                                   y = mean.y, color = pop),
                      size = 7,
                      shape = 17) +
  ggplot2::scale_colour_manual(values = colours5) +
  ggplot2::scale_fill_manual(values = colours5) + 
  ggplot2::geom_hline(yintercept = 0) +
  ggplot2::geom_vline(xintercept = 0) + 
  ggplot2::labs(
    subtitle = paste0("Total variance explained: ",
                      signif(sum(var_frac[1:4]) * 100, 3), "%"),
    y = paste0("PC3; ", signif(var_frac[3] * 100, digits = 2), " %"),
    x = paste0("PC2; ", signif(var_frac[2] * 100, digits = 2), " %"),
    title = plot.title,
    caption = "") + 
  ggplot2::theme_minimal() +
  ggplot2::theme(
    axis.text = ggplot2::element_text(size = 15),
    axis.title.x = ggplot2::element_text(size = 20),
    axis.title.y = ggplot2::element_text(size = 20),
    legend.text = ggplot2::element_text(size = 15)
  )

print(pca.plot)
ggplot2::ggsave(pca.plot, bg = "white",
                filename = paste0(figure.path,
                                  "PCA_plot_diffMeth_commonCpG_PC23.png"
                                  ), 
                width = 30, height = 15, units = "cm")
```


### Plot CpG distance

```{r}
split <- as.data.frame(stringr::str_split_fixed(string = common.cpg.names, 
                                                pattern = "_", n = 2))
split$V2 <- as.integer(split$V2)
table(split$V1)
hist(split$V2)
split$V2 <- as.integer(split$V2)

split2 <- split[split$V1 %in% c("JAAKFY010000014"),]
split2 <- split2[-c(1:3),]

split.plot <- ggplot2::ggplot(split2, aes(x = V2, y = V1)) +
  ggplot2::geom_jitter() +
  ggplot2::theme(
    panel.background = ggplot2::element_blank(),
    panel.border = ggplot2::element_blank(),
    panel.grid = ggplot2::element_blank(),
    axis.line = ggplot2::element_line("gray25"),
    text = ggplot2::element_text(size = 12),
    axis.text.x = ggplot2::element_text(
      angle = 90,
      size = 10,
      colour = "gray25"
    ),
    axis.title = ggplot2::element_text(size = 14, colour = "gray25"),
    legend.title = ggplot2::element_text(size = 14),
    legend.text = ggplot2::element_text(size = 10),
    legend.key = ggplot2::element_blank()
  )
print(split.plot)
```

```{r}
dat.100 <- data.table::as.data.table(GenomicRanges::reduce(common.cpg.gr, 
                                                           min.gapwidth = 10L,
                                                           drop.empty.ranges=FALSE))

range(common.cpg.gr)

gr.100 <- as(dat.100,"GRanges")
```


### Plot CpG difference

```{r}
meth.diff.df <- as.data.frame(t(methylKit::percMethylation(meth.diff)))
loc.names <- paste0(meth.diff$chr,"_",meth.diff$start)
colnames(meth.diff.df) <- loc.names

meth.diff.df$individual <- rownames(meth.diff.df) 
identical(metadata2$SAMPLE_ID, meth.diff.df$individual)
meth.diff.df$pop <- metadata2$region

meth.diff.df.plot <- tidyr::gather(meth.diff.df, key = locus, 
                           value = perc_methylation, -individual, -pop,
                           factor_key = TRUE) %>%
  dplyr::mutate(chr = stringr::str_split_fixed(locus, 
                                               pattern = "_", 
                                               n = 2)[,1])
meth.diff.df.plot.locus <- meth.diff.df.plot %>% dplyr::group_by(chr,locus,pop) %>% 
  dplyr::summarise(mean_pop = mean(perc_methylation, na.rm = TRUE),
            .groups = 'drop')

meth.diff.df.plot.centroid <- meth.diff.df.plot %>% dplyr::group_by(chr,pop) %>% 
  dplyr::summarise(mean_pop = mean(perc_methylation, na.rm = TRUE),
            .groups = 'drop')

meth.diff.df.plot.label <- meth.diff.df.plot %>% dplyr::group_by(chr) %>% 
  dplyr::summarise(nmarkers = length(unique(locus)),
            .groups = 'drop')


cpg.plot <- ggplot2::ggplot() + 
  ggplot2::geom_violin(data = meth.diff.df.plot, 
                       ggplot2::aes(x = chr, y = perc_methylation)) + 
  ggplot2::geom_point(data = meth.diff.df.plot.centroid, 
                      ggplot2::aes(x = chr, y = mean_pop, colour = pop),
                      size = 7, shape = 5) + 
  ggplot2::geom_point(data = meth.diff.df.plot,
                      ggplot2::aes(x = chr, y = perc_methylation, colour = pop),
                      size = 1) +
  ggplot2::geom_text(data = meth.diff.df.plot.label,
                     ggplot2::aes(x = chr, y = 110, label = nmarkers),
                     size = 4) +
  ggplot2::geom_point(data = meth.diff.df.plot.locus, 
                      ggplot2::aes(x = chr, y = mean_pop, colour = pop),
                      size = 3, shape = 15) +
    ggplot2::theme(
        panel.background = ggplot2::element_blank(),
        panel.border = ggplot2::element_blank(),
        panel.grid = ggplot2::element_blank(),
        axis.line = ggplot2::element_line("gray25"),
        text = ggplot2::element_text(size = 12),
        axis.text = ggplot2::element_text(angle = 90,size = 10, colour = "gray25"),
        axis.title = ggplot2::element_text(size = 14, colour = "gray25"),
        legend.title = ggplot2::element_text(size = 14),
        legend.text = ggplot2::element_text(size = 10),
        legend.key = ggplot2::element_blank())

print(cpg.plot)

ggplot2::ggsave(plot = HYPO.plot, 
                filename = paste0(figure.path,"TOA_commonCpG_methDiff.png"),
                width = 30, height = 15, units = "cm", device = "png")


res <- as.data.frame(table(meth.diff.df.plot$chr))
plot(res)
meth.diff.df.plot2 <- meth.diff.df.plot[meth.diff.df.plot$chr %in% c("JAAKFY010000014"),]

cpg.plot2 <- ggplot2::ggplot(data = meth.diff.df.plot2, 
                       ggplot2::aes(x = locus, y = perc_methylation)) + 
  ggplot2::geom_violin(data = meth.diff.df.plot2, 
                       ggplot2::aes(x = locus, y = perc_methylation)) + 
  # ggplot2::geom_point(data = meth.diff.df.plot.centroid, 
  #                     ggplot2::aes(x = locus, y = mean_pop, colour = pop),
  #                     size = 7, shape = 5) + 
  ggplot2::geom_point(data = meth.diff.df.plot2,
                      ggplot2::aes(x = locus, y = perc_methylation, colour = pop),
                      size = 1) +
  # ggplot2::geom_text(data = meth.diff.df.plot.label,
  #                    ggplot2::aes(x = locus, y = 110, label = nmarkers),
  #                    size = 4) +
  # ggplot2::geom_point(data = meth.diff.df.plot.locus,
  #                     ggplot2::aes(x = locus, y = mean_pop, colour = pop),
  #                     size = 3, shape = 15) +
  # ggplot2::facet_grid(~chr) +
    ggplot2::theme(
        panel.background = ggplot2::element_blank(),
        panel.border = ggplot2::element_blank(),
        panel.grid = ggplot2::element_blank(),
        axis.line = ggplot2::element_line("gray25"),
        text = ggplot2::element_text(size = 12),
        axis.text = ggplot2::element_text(angle = 90,size = 10, colour = "gray25"),
        axis.title = ggplot2::element_text(size = 14, colour = "gray25"),
        legend.title = ggplot2::element_text(size = 14),
        legend.text = ggplot2::element_text(size = 10),
        legend.key = ggplot2::element_blank())

print(cpg.plot2)

ggplot2::ggsave(plot = HYPO.plot, 
                filename = paste0(figure.path,"TOA_commonCpG_methDiff.png"),
                width = 30, height = 15, units = "cm", device = "png")


diff.plot <- ggplot2::ggplot() +
  # ggplot2::geom_point(data = meth.diff.df.plot2,
  #                     ggplot2::aes(x = pop, y = perc_methylation, colour = locus)) +
  ggplot2::geom_line(data = meth.diff.df.plot.locus,
                     ggplot2::aes(x = pop, y = mean_pop, colour = locus, group = locus)) +
  ggplot2::geom_point(data = meth.diff.df.plot.locus,
                     ggplot2::aes(x = pop, y = mean_pop, colour = locus, group = locus),
                     size = 5) +
  ggplot2::ylim(c(0,100)) +
  ggplot2::facet_wrap(~chr) +
  ggplot2::theme(legend.position = "none")
  # ggplot2::guides(colour = ggplot2::guide_legend(ncol = 1))
  # ggplot2::guides(colour = ggplot2::guide_legend(ncol = 5))

print(diff.plot)

ggplot2::ggsave(plot = diff.plot, 
                filename = paste0(figure.path,
                                  # "0.PRESS_SS_PERCENTAGE_methDiff_TAS_exp_muscle_DMS25_methylkit.png"
                                  # "0.PRESS_SS_PERCENTAGE_methDiff_TAS_exp_muscle_DMS50_methylkit.png"
                                  "0.PRESS_TOA_PERCENTAGE_methDiff_commonDMS.png"
                                  # "0.PRESS_SS_PERCENTAGE_methDiff_TAS_exp_liver_DMS50_methylkit.png"
                                  ),
                width = 60, height = 50, units = "cm", device = "png")

```

### Heatmap
```{r}
print(load(paste0(rdata.path,"Toothfish_DMS_CpGsites.Rdata")))
DMS.data <- DMS %>%
  dplyr::count(DMS, region_pw) %>%
  tidyr::spread(region_pw, n, fill = 0)
dim(DMS.data)
common.cpg <- rowSums(DMS.data[,-1])
hist(common.cpg)
common.cpg.names <- DMS.data$DMS[common.cpg > 1]

print(load(paste0(rdata.path,"Toothfish_subset_methylation_objects_raw.Rdata")))


length(unique(common.cpg.names))#433

common.cpg.gr <- as(data.frame(chr = stringr::str_split_fixed(
                                        string = as.character(unique(common.cpg.names)), 
                                        pattern =  "_", n = 2)[,1],
                                    start = 
                                      as.integer(stringr::str_split_fixed(
                                        string = as.character(unique(common.cpg.names)), 
                                        pattern =  "_", n = 2)[,2]),
                                    end = 
                                      as.integer(stringr::str_split_fixed(
                                        string = as.character(unique(common.cpg.names)), 
                                        pattern =  "_", n = 2)[,2])
                                    ), "GRanges")

meth.diff <- methylKit::selectByOverlap(meth, common.cpg.gr) #433 

meth.diff.df <- as.data.frame(t(methylKit::percMethylation(meth.diff)))
loc.names <- paste0(meth.diff$chr,"_",meth.diff$start)
colnames(meth.diff.df) <- loc.names

meth.diff.df$individual <- rownames(meth.diff.df) 
metadata.new <- metadata[metadata$SAMPLE_ID %in% meth.diff@sample.ids,]
metadata.new <- metadata.new[order(match(metadata.new$SAMPLE_ID, meth.diff@sample.ids)),]
identical(metadata.new$SAMPLE_ID, meth.diff@sample.ids)
identical(metadata.new$SAMPLE_ID,meth.diff.df$individual)
meth.diff.df$pop <- metadata.new$region


heat.plot.data <- meth.diff.df %>%
  tibble::as_tibble() %>%
  tidyr::gather(key = "CpG", value = "Methylation_percentage", -individual, -pop,
                factor_key = FALSE) %>%
  dplyr::group_by(pop, CpG) %>%
  dplyr::mutate(CpG_mean = mean(Methylation_percentage, na.rm = TRUE)) %>%
  dplyr::arrange(pop, individual, CpG) %>%
  dplyr::mutate(individual = factor(individual, levels = unique(individual))) %>%
  dplyr::mutate(CpG = factor(CpG, levels = unique(CpG)))
# cpgs <- heat.plot.data$CpG[heat.plot.data$individual == heat.plot.data$individual[1]]
# cpgs_mean <- heat.plot.data$CpG_mean[heat.plot.data$individual == heat.plot.data$individual[1]]
cpgs <- heat.plot.data$CpG[heat.plot.data$individual == unique(heat.plot.data$individual)[60]]
cpgs_mean <- heat.plot.data$CpG_mean[heat.plot.data$individual == unique(heat.plot.data$individual)[60]]
cpgs <- cpgs[order(cpgs_mean)]
heat.plot.data$CpG <- factor(heat.plot.data$CpG, levels = unique(cpgs))
heat.plot.data <- heat.plot.data[order(heat.plot.data$CpG),]


heat.plot <- ggplot2::ggplot(data = heat.plot.data) + 
    ggplot2::geom_tile(ggplot2::aes(individual, y = CpG, fill = Methylation_percentage)) +
    ggplot2::scale_fill_gradient2(low = "blue", mid = "yellow", high = "red", midpoint = 50) +
    # ggnewscale::new_scale_fill() +
    # ggplot2::geom_tile(ggplot2::aes(individual, y = "pop", fill = pop, height = 10)) +
    # ggplot2::scale_fill_manual(values = c("#4d6b40","#adcf9f", "#c5d6e6", "#08306B")) +
    ggplot2::scale_y_discrete(drop = FALSE) +
    ggplot2::theme(legend.position = "top",
                   axis.title.y = ggplot2::element_blank(),
          axis.text.x = ggplot2::element_text(face = "bold", color = "black", 
                                        size = 10, angle = 90, hjust = 1, vjust = 0))
print(heat.plot)

ggplot2::ggsave(heat.plot, bg = "white",
                filename = paste0(figure.path,"TOA_commonCpG_methDiff_HEATMAP2.png"), 
                width = 45, height = 30, units = "cm")
```


### Vendiagram

```{r}
ggVennDiagram::process_region_data(ggVennDiagram::Venn(DMS.list))

DMS.list <- list(DMS.HR[,1],DMS.EANT[,1],DMS.EH1[,1],
                 DMS.EH2[,1],DMS.ER1[,1],DMS.ER2[,1])

DMS.plot <- ggVennDiagram::ggVennDiagram(
  x = DMS.list, label_alpha = 0, category.names =
    c("HEARD_ROSS","EANT1_EANT2","EANT1_HEARD","EANT2_HEARD","EANT1_ROSS","EANT2_ROSS")
    # c("HEARD_ROSS","EANT_HEARD","EANT1_ROSS","EANT2_ROSS")
    # c("HEARD_ROSS","EANT1_EANT2","EANT_HEARD","EANT_ROSS")

  ) +
  ggplot2::scale_fill_gradient2(low = "blue", mid = "yellow",high = "red", midpoint = 400)
print(DMS.plot)
ggsave(plot = DMS.plot, filename = paste0(figure.path,
                                          "TOA_commonCpGs_between_regions1_VENN.png"),
       width = 30, height = 30, device = "png", units = "cm")


DMS.list <- list(DMS.HR[,1],DMS.EH[,1],DMS.ER1[,1], DMS.ER2[,1])

DMS.plot <- ggVennDiagram::ggVennDiagram(
  x = DMS.list, label_alpha = 0, category.names =
    c("HEARD_ROSS","EANT_HEARD","EANT1_ROSS","EANT2_ROSS")
  ) +
  ggplot2::scale_fill_gradient2(low = "blue", mid = "yellow",high = "red", midpoint = 200)
print(DMS.plot)
ggsave(plot = DMS.plot, filename = paste0(figure.path,
                                          "TOA_commonCpGs_between_regions2_VENN.png"),
       width = 30, height = 30, device = "png", units = "cm")



DMS.list <- list(DMS.HR[,1],DMS.EANT[,1],DMS.EH[,1],DMS.ER[,1])

DMS.plot <- ggVennDiagram::ggVennDiagram(
  x = DMS.list, label_alpha = 0, category.names =
    c("HEARD_ROSS","EANT1_EANT2","EANT_HEARD","EANT_ROSS")

  ) +
  ggplot2::scale_fill_gradient2(low = "blue", mid = "yellow",high = "red", midpoint = 200)
print(DMS.plot)
ggsave(plot = DMS.plot, filename = paste0(figure.path,
                                          "TOA_commonCpGs_between_regions3_VENN.png"),
       width = 30, height = 30, device = "png", units = "cm")
```

#### Venndiagram methylkit DMS pairwise vs together

```{r}
print(load(paste0(rdata.path,"Toothfish_DMS_Region_subset1.Rdata")))
print(load(paste0(rdata.path,"Toothfish_DMS_CpGsites.Rdata")))

DMS.pw <- unique(DMS$DMS)
DMS.kit <- paste0(myDiff25p$chr, "_", myDiff25p$start)
DMS.list <- list(DMS.pw, DMS.kit)

DMS.plot <- ggVennDiagram::ggVennDiagram(
  x = DMS.list, label_alpha = 0, category.names =
    c("DMS_pairwise_regions","DMS_methylkit_together")
  ) +
  ggplot2::scale_fill_gradient2(low = "blue", mid = "yellow",high = "red", midpoint = 200)
print(DMS.plot)
ggsave(plot = DMS.plot, filename = paste0(figure.path,
                                          "TOA_VENN_DMS_methylkit_pairwise_vs_together.png"),
       width = 30, height = 30, device = "png", units = "cm")



```

#### Heatmap intersecting DMS

```{r}
pw.cpg.gr <- as(data.frame(chr = stringr::str_split_fixed(
                                        string = as.character(unique(DMS$DMS)), 
                                        pattern =  "_", n = 2)[,1],
                                    start = 
                                      as.integer(stringr::str_split_fixed(
                                        string = as.character(unique(DMS$DMS)), 
                                        pattern =  "_", n = 2)[,2]),
                                    end = 
                                      as.integer(stringr::str_split_fixed(
                                        string = as.character(unique(DMS$DMS)), 
                                        pattern =  "_", n = 2)[,2])
                                    ), "GRanges")


intersect.cpg <- methylKit::selectByOverlap(myDiff25p, pw.cpg.gr) #334 
intersect.cpg.gr <- as(intersect.cpg, "GRanges")

meth.diff <- methylKit::selectByOverlap(meth, intersect.cpg.gr) #334 

meth.diff.df <- as.data.frame(t(methylKit::percMethylation(meth.diff)))
loc.names <- paste0(meth.diff$chr,"_",meth.diff$start)
colnames(meth.diff.df) <- loc.names

meth.diff.df$individual <- rownames(meth.diff.df) 
metadata.new <- metadata[metadata$SAMPLE_ID %in% meth.diff@sample.ids,]
metadata.new <- metadata.new[order(match(metadata.new$SAMPLE_ID, meth.diff@sample.ids)),]
identical(metadata.new$SAMPLE_ID, meth.diff@sample.ids)
identical(metadata.new$SAMPLE_ID,meth.diff.df$individual)
meth.diff.df$pop <- metadata.new$region


heat.plot.data <- meth.diff.df %>%
  tibble::as_tibble() %>%
  tidyr::gather(key = "CpG", value = "Methylation_percentage", -individual, -pop,
                factor_key = FALSE) %>%
  dplyr::group_by(pop, CpG) %>%
  dplyr::mutate(CpG_mean = mean(Methylation_percentage, na.rm = TRUE)) %>%
  dplyr::arrange(pop, individual, CpG) %>%
  dplyr::mutate(individual = factor(individual, levels = unique(individual))) %>%
  dplyr::mutate(CpG = factor(CpG, levels = unique(CpG)))
# cpgs <- heat.plot.data$CpG[heat.plot.data$individual == heat.plot.data$individual[1]]
# cpgs_mean <- heat.plot.data$CpG_mean[heat.plot.data$individual == heat.plot.data$individual[1]]
cpgs <- heat.plot.data$CpG[heat.plot.data$individual == unique(heat.plot.data$individual)[60]]
cpgs_mean <- heat.plot.data$CpG_mean[heat.plot.data$individual == unique(heat.plot.data$individual)[60]]
cpgs <- cpgs[order(cpgs_mean)]
heat.plot.data$CpG <- factor(heat.plot.data$CpG, levels = unique(cpgs))
heat.plot.data <- heat.plot.data[order(heat.plot.data$CpG),]


heat.plot <- ggplot2::ggplot(data = heat.plot.data) + 
    ggplot2::geom_tile(ggplot2::aes(individual, y = CpG, fill = Methylation_percentage)) +
    ggplot2::scale_fill_gradient2(low = "blue", mid = "yellow", high = "red", midpoint = 50) +
    # ggnewscale::new_scale_fill() +
    # ggplot2::geom_tile(ggplot2::aes(individual, y = "pop", fill = pop, height = 10)) +
    # ggplot2::scale_fill_manual(values = c("#4d6b40","#adcf9f", "#c5d6e6", "#08306B")) +
    ggplot2::scale_y_discrete(drop = FALSE) +
    ggplot2::theme(legend.position = "top",
                   axis.title.y = ggplot2::element_blank(),
          axis.text.x = ggplot2::element_text(face = "bold", color = "black", 
                                        size = 10, angle = 90, hjust = 1, vjust = 0))
print(heat.plot)

ggplot2::ggsave(heat.plot, bg = "white",
                filename = paste0(figure.path,"TOA_intersectCpG_methDiff_HEATMAP.png"), 
                width = 45, height = 30, units = "cm")
```


### Upset plot

```{r}
n.pop = length(unique(DMS$region_pw))
plot.data <- DMS %>%
  dplyr::distinct(DMS, region_pw) %>%
  dplyr::mutate(n = rep(1, n()),
                region_pw = stringi::stri_join("", region_pw)) %>%
  data.table::as.data.table(.) %>%
  data.table::dcast.data.table(
    data = .,
    formula = DMS ~ region_pw,
    value.var = "n",
    fill = 0
  ) %>%
  data.frame(.)

Upsetplot <- UpSetR::upset(
  plot.data,
  nsets = n.pop,
  # keep.order = TRUE,
  # order.by = "freq",
  order.by = "degree",
  empty.intersections = NULL
)
print(Upsetplot)

ggsave(plot = Upsetplot, 
       filename = paste0(figure.path,"TOA_commonCpGs_between_regions_Upsetplot1.png"),
       width = 30, height = 30, device = "png", units = "cm")


pop.names <- c("HEARD_ROSS","EANT_HEARD","EANT4.1_ROSS","EANT4.2_ROSS")
plot.data2 <- plot.data[,colnames(plot.data) %in% pop.names]


Upsetplot <- UpSetR::upset(
  plot.data2,
  nsets = length(pop.names),
  # keep.order = TRUE,
  # order.by = "freq",
  order.by = "degree",
  empty.intersections = NULL
)
print(Upsetplot)

ggsave(plot = Upsetplot, 
       filename = paste0(figure.path,"TOA_commonCpGs_between_regions_Upsetplot1.png"),
       width = 30, height = 30, device = "png", units = "cm")
```


### Overlap with features

```{r}
gencode <- AnnotationDbi::loadDb(paste0(rdata.path, "gencode_D_mawsoni.sqlite"))

genes.gencode <- sort(GenomicFeatures::genes(gencode))
intergenic.gencode <- GenomicRanges::gaps(genes.gencode)
transcript.gencode <- sort(GenomicFeatures::transcripts(gencode, use.names = TRUE))
cds.gencode <- sort(GenomicFeatures::cds(gencode))
exons.gencode <- sort(GenomicFeatures::exons(gencode))
introns.gencode <- sort(unlist(GenomicFeatures::intronsByTranscript(gencode)))
promoters.gencode <- sort(GenomicFeatures::promoters(gencode, upstream = 1000, downstream = 1000))
TSS.gencode <- transcript.gencode
TSS.gencode@ranges@width <- rep(1L, times = length(TSS.gencode@ranges@width))

annotation.all.features <-  GenomicRanges::GRangesList(
  promoters = promoters.gencode, 
  exons = exons.gencode,
  genes = genes.gencode,
  introns = introns.gencode,
  intergenics = intergenic.gencode,
  transcripts = transcript.gencode,
  cds = cds.gencode,
  TSSes = TSS.gencode)


feat.df <- data.frame(annot = NULL, target = NULL, feature = NULL)
ann <- genomation::annotateWithFeature(common.cpg.gr, genes.gencode, intersect.chr = TRUE) #59.4%, 50.6%
feat.df <- rbind(feat.df, data.frame(annot = ann@num.annotation[1], 
                                     target = ann@num.annotation[2], 
                                     feature = "Genes"))
ann <- genomation::annotateWithFeature(common.cpg.gr, transcript.gencode, intersect.chr = TRUE)
feat.df <- rbind(feat.df, data.frame(annot = ann@num.annotation[1], 
                                     target = ann@num.annotation[2], 
                                     feature = "Transcipts"))
ann <- genomation::annotateWithFeature(common.cpg.gr, cds.gencode, intersect.chr = TRUE) #16.24%, 83.76#
feat.df <- rbind(feat.df, data.frame(annot = ann@num.annotation[1], 
                                     target = ann@num.annotation[2], 
                                     feature = "CDS"))
ann <- genomation::annotateWithFeature(common.cpg.gr, intergenic.gencode, intersect.chr = TRUE) #100%, 0%
feat.df <- rbind(feat.df, data.frame(annot = ann@num.annotation[1], 
                                     target = ann@num.annotation[2], 
                                     feature = "Intergenic"))
ann <- genomation::annotateWithFeature(common.cpg.gr, exons.gencode, intersect.chr = TRUE) #16.24%, 83.76%
feat.df <- rbind(feat.df, data.frame(annot = ann@num.annotation[1], 
                                     target = ann@num.annotation[2], 
                                     feature = "Exons"))
ann <- genomation::annotateWithFeature(common.cpg.gr, introns.gencode, intersect.chr = TRUE) #4339%, 46.61%
feat.df <- rbind(feat.df, data.frame(annot = ann@num.annotation[1], 
                                     target = ann@num.annotation[2], 
                                     feature = "Introns"))
ann <- genomation::annotateWithFeature(common.cpg.gr, promoters.gencode, intersect.chr = TRUE)# 11.37%, 88.63%
feat.df <- rbind(feat.df, data.frame(annot = ann@num.annotation[1], 
                                     target = ann@num.annotation[2], 
                                     feature = "Promoters"))
ann <- genomation::annotateWithFeature(common.cpg.gr, TSS.gencode, intersect.chr = TRUE)# 0%, 100%
feat.df <- rbind(feat.df, data.frame(annot = ann@num.annotation[1], 
                                     target = ann@num.annotation[2], 
                                     feature = "TSS"))

feat.df2 <- reshape2::melt(feat.df, id.vars='feature')
feat.df2$variable <- as.character(feat.df2$variable)
feat.df2$variable[feat.df2$variable == "annot"] <- "Other annotations"
feat.df2$variable[feat.df2$variable == "target"] <- "RRBS CpG sites"

feat.df2 <- feat.df2[feat.df2$variable == "RRBS CpG sites",]
feat.df2 <- feat.df2[feat.df2$variable == "Other annotations",]

annot.plot <- ggplot2::ggplot(data = feat.df2, 
                              ggplot2::aes(x =  feature, y = value, fill = variable)) +
  ggplot2::geom_bar(stat='identity', position = 'dodge') +
    ggplot2::labs(
    y = "Number of CpG",
    x = "Annotation feature",
    # title = "Overlap between annotations and 36000 CpG sites",
    title = "Overlap between annotations and 433 CpG sites",
    caption = "") +
  # ggplot2::scale_fill_manual(values = "darkgreen") +
  ggplot2::theme_minimal() +
  ggplot2::theme(
    axis.text = ggplot2::element_text(size = 15),
    axis.text.x = ggplot2::element_text(size = 15, angle = 90, vjust = 0.5),
    axis.title.x = ggplot2::element_text(size = 20),
    axis.title.y = ggplot2::element_text(size = 20),
    legend.text = ggplot2::element_text(size = 15)
  )
print(annot.plot)
ggplot2::ggsave(annot.plot, bg = "white",
                filename = paste0(figure.path,"Annotation_overlap_plot_commonCpG2.png"), 
                width = 30, height = 15, units = "cm")
```

```{r}
sm <- genomation::ScoreMatrix(target = common.cpg.gr, 
                             windows = annotation.all.features$promoters, 
                             strand.aware = TRUE)

count <- GenomicRanges::countOverlaps(query = common.cpg.gr, 
                             subject = annotation.all.features$promoters)
hist(count)

count <- GenomicRanges::findOverlaps(query = common.cpg.gr, 
                             subject = annotation.all.features$promoters)
count <- GenomicRanges::subsetByOverlaps(common.cpg.gr, annotation.all.features$promoters, type = "all")

meth.gr <- as(meth,"GRanges")
count <- GenomicRanges::countOverlaps(query = meth.gr, 
                             subject = annotation.all.features$promoters)
hist(count)

count <- GenomicRanges::countOverlaps(query = meth.gr, 
                             subject = annotation.all.features$genes)
hist(count)

cpg.ind <- which(GenomicRanges::countOverlaps(
  annotation.all.features$promoters, common.cpg.gr) > 0)

promotors.commoncpg.gr <- annotation.all.features$promoters[
  names(annotation.all.features$promoters@ranges) %in% names(cpg.ind),]

nocpg.ind <- which(GenomicRanges::countOverlaps(
  annotation.all.features$promoters, common.cpg.gr) == 0)
genomation::heatMatrix(sm, xcoords = c(-1000, 1000), 
                       group = list(CpGi = cpg.ind))
```


```{r}
gtf.file <- "/datasets/work/ncmi-toa-rrbs/work/1.Toothfish/Bismark_genome_index/D.mawsoni_2.gtf"
gtf <-  genomation::gffToGRanges(gtf.file)
grl21 <- as(split(gtf, gtf$type), "GRangesList")
gr.l <- split(gtf, gtf$type)
annot <- genomation::annotateWithFeatures(common.cpg.gr, grl21, strand.aware = TRUE)

annot@num.annotation

genomation::heatTargetAnnotation(annot,cluster = FALSE,
                                 col = c("white","blue"), precedence = FALSE, 
                                 plot = TRUE)

genomation::getFeatsWithTargetsStats(x = annot, percentage = FALSE)

genomation::getTargetAnnotationStats(annot,percentage=TRUE,precedence=TRUE)
genomation::plotTargetAnnotation(annot,precedence=TRUE,
    main="differential methylation annotation")

genomation::calculateOverlapSignificance(target = common.cpg.gr, feature = gtf, 
                                         chrom.sizes = common.cpg.gr@seqinfo@seqlengths,
  stranded = TRUE, keep.strand.prop = TRUE, keep.chrom = TRUE,
  exclude = NULL, include = NULL, seed = NULL, nrand = 1)
```


```{r}
idx_overlaps <- S4Vectors::queryHits(GenomicRanges::findOverlaps(
  query = common.cpg.gr,subject = annotation.all.features$promoters, 
  type = 'any'))
promoters.common.cpg.gr <- common.cpg.gr[unique(idx_overlaps),]
promoters.common.cpg.raw <- meth.diff[unique(idx_overlaps),]

idx_overlaps <- S4Vectors::queryHits(GenomicRanges::findOverlaps(
  query = common.cpg.gr,subject = gtf, 
  type = 'any'))
featured.common.cpg.gr <- common.cpg.gr[unique(idx_overlaps),]


idx_overlaps2 <- S4Vectors::queryHits(GenomicRanges::findOverlaps(
  query = annotation.all.features,subject = common.cpg.gr, 
  type = 'any'))
features.gr <- annotation.all.features[unique(idx_overlaps2),]
features.gr <- annotation.all.features[unique(idx_overlaps2),]

idx_overlaps2 <- S4Vectors::queryHits(GenomicRanges::findOverlaps(
  query = gtf,subject = common.cpg.gr, 
  type = 'any'))

features.gr <- gtf[unique(idx_overlaps2),]


common.cpg.prc <- methylKit::percMethylation(promoters.common.cpg.raw) # get percent methylation matrix
mds <- matrixStats::rowSds(common.cpg.prc) # calculate standard deviation of CpGs
head(common.cpg.prc[mds>20,])
hist(mds,col="cornflowerblue",xlab="Std. dev. per CpG")
common.cpg.prc <- as.data.frame(common.cpg.prc)
TSS.prc$seqs <- paste0(TSS.gr@seqnames,"_", TSS.gr@ranges@start)

TSS.prc.long <- tidyr::gather(data = TSS.prc, sample, methylation, -seqs)
strata <- data.frame(sample  = metadata$SAMPLE_ID, region = metadata$region)
TSS.prc.long <- dplyr::left_join(TSS.prc.long, strata, by = "sample" )
TSS.grp <- dplyr::group_by(TSS.prc.long, region, seqs) %>%
    dplyr::summarise(mean_meth  = mean(methylation))
TSS.grp$feat <- "TSS"
```


### Plot CpG difference - promotor

```{r}
meth.diff.df <- as.data.frame(t(methylKit::percMethylation(promoters.common.cpg.raw)))
loc.names <- paste0(promoters.common.cpg.raw$chr,"_",promoters.common.cpg.raw$start)
colnames(meth.diff.df) <- loc.names

meth.diff.df$individual <- rownames(meth.diff.df) 
identical(metadata2$SAMPLE_ID, meth.diff.df$individual)
meth.diff.df$pop <- metadata2$region

meth.diff.df.plot <- tidyr::gather(meth.diff.df, key = locus, 
                           value = perc_methylation, -individual, -pop,
                           factor_key = TRUE) %>%
  dplyr::mutate(chr = stringr::str_split_fixed(locus, 
                                               pattern = "_", 
                                               n = 2)[,1])
meth.diff.df.plot.locus <- meth.diff.df.plot %>% dplyr::group_by(chr,locus,pop) %>% 
  dplyr::summarise(mean_pop = mean(perc_methylation, na.rm = TRUE),
            .groups = 'drop')

meth.diff.df.plot.centroid <- meth.diff.df.plot %>% dplyr::group_by(chr,pop) %>% 
  dplyr::summarise(mean_pop = mean(perc_methylation, na.rm = TRUE),
            .groups = 'drop')

meth.diff.df.plot.label <- meth.diff.df.plot %>% dplyr::group_by(chr) %>% 
  dplyr::summarise(nmarkers = length(unique(locus)),
            .groups = 'drop')


cpg.plot <- ggplot2::ggplot() + 
  ggplot2::geom_violin(data = meth.diff.df.plot, 
                       ggplot2::aes(x = chr, y = perc_methylation)) + 
  ggplot2::geom_point(data = meth.diff.df.plot.centroid, 
                      ggplot2::aes(x = chr, y = mean_pop, colour = pop),
                      size = 7, shape = 5) + 
  ggplot2::geom_point(data = meth.diff.df.plot,
                      ggplot2::aes(x = chr, y = perc_methylation, colour = pop),
                      size = 1) +
  ggplot2::geom_text(data = meth.diff.df.plot.label,
                     ggplot2::aes(x = chr, y = 110, label = nmarkers),
                     size = 4) +
  ggplot2::geom_point(data = meth.diff.df.plot.locus, 
                      ggplot2::aes(x = chr, y = mean_pop, colour = pop),
                      size = 3, shape = 15) +
    ggplot2::theme(
        panel.background = ggplot2::element_blank(),
        panel.border = ggplot2::element_blank(),
        panel.grid = ggplot2::element_blank(),
        axis.line = ggplot2::element_line("gray25"),
        text = ggplot2::element_text(size = 12),
        axis.text = ggplot2::element_text(angle = 90,size = 10, colour = "gray25"),
        axis.title = ggplot2::element_text(size = 14, colour = "gray25"),
        legend.title = ggplot2::element_text(size = 14),
        legend.text = ggplot2::element_text(size = 10),
        legend.key = ggplot2::element_blank())

print(cpg.plot)

# ggplot2::ggsave(plot = HYPO.plot, 
#                 filename = paste0(figure.path,"TOA_commonCpG_methDiff.png"),
#                 width = 30, height = 15, units = "cm", device = "png")


res <- as.data.frame(table(meth.diff.df.plot$chr))
plot(res)
meth.diff.df.plot2 <- meth.diff.df.plot[meth.diff.df.plot$chr %in% c("JAAKFY010000014"),]

cpg.plot2 <- ggplot2::ggplot(data = meth.diff.df.plot2, 
                       ggplot2::aes(x = locus, y = perc_methylation)) + 
  ggplot2::geom_violin(data = meth.diff.df.plot2, 
                       ggplot2::aes(x = locus, y = perc_methylation)) + 
  # ggplot2::geom_point(data = meth.diff.df.plot.centroid, 
  #                     ggplot2::aes(x = locus, y = mean_pop, colour = pop),
  #                     size = 7, shape = 5) + 
  ggplot2::geom_point(data = meth.diff.df.plot2,
                      ggplot2::aes(x = locus, y = perc_methylation, colour = pop),
                      size = 1) +
  # ggplot2::geom_text(data = meth.diff.df.plot.label,
  #                    ggplot2::aes(x = locus, y = 110, label = nmarkers),
  #                    size = 4) +
  # ggplot2::geom_point(data = meth.diff.df.plot.locus,
  #                     ggplot2::aes(x = locus, y = mean_pop, colour = pop),
  #                     size = 3, shape = 15) +
  # ggplot2::facet_grid(~chr) +
    ggplot2::theme(
        panel.background = ggplot2::element_blank(),
        panel.border = ggplot2::element_blank(),
        panel.grid = ggplot2::element_blank(),
        axis.line = ggplot2::element_line("gray25"),
        text = ggplot2::element_text(size = 12),
        axis.text = ggplot2::element_text(angle = 90,size = 10, colour = "gray25"),
        axis.title = ggplot2::element_text(size = 14, colour = "gray25"),
        legend.title = ggplot2::element_text(size = 14),
        legend.text = ggplot2::element_text(size = 10),
        legend.key = ggplot2::element_blank())

print(cpg.plot2)

ggplot2::ggsave(plot = HYPO.plot, 
                filename = paste0(figure.path,"TOA_commonCpG_methDiff.png"),
                width = 30, height = 15, units = "cm", device = "png")
```

## Glmnet models
### Train models
```{r}
print(load(paste0(rdata.path, "TOA_aging_markers.rdata")))
print(load(paste0(rdata.path, "TOA_aging_test_data.Rdata")))
print(load(paste0(rdata.path,"Toothfish_metadata.Rdata")))

metadata3 <- metadata2[metadata2$SAMPLE_ID %in% rownames(trainTransformed3),]
metadata3 <- metadata3[order(match(metadata3$SAMPLE_ID, rownames(trainTransformed3))),]
identical(rownames(trainTransformed3),metadata3$SAMPLE_ID)

trainTransformed3$age <- metadata3$region


fitControl <- caret::trainControl(method = "repeatedcv", number = 10, repeats = 10)


ridge_model <-
  caret::train(x = trainTransformed3[,-ncol(trainTransformed3)],
               y = trainTransformed3$age,
    method = "glmnet",
    trControl = fitControl,
    tuneGrid = data.frame(alpha = 0, lambda = lambda),
    tuneLength = 10
  )

# save(ridge_model,
#      file = paste0(rdata.path,"Toothfish_regional_models.Rdata"))


# LASSO

set.seed(123)

lasso_model <-
  caret::train(x = trainTransformed3[,-ncol(trainTransformed3)],
               y = trainTransformed3$age,
    method = "glmnet",
    trControl = fitControl,
    tuneGrid = data.frame(alpha = 1, lambda = lambda),
    tuneLength = 10
  )
# save(ridge_model,lasso_model,
#      file = paste0(rdata.path,"Toothfish_regional_models.Rdata"))



# In Elastic net best tuning of both lambda and alpha will be automatically selected

set.seed(123)

elastic_model <-
  caret::train(x = trainTransformed3[,-ncol(trainTransformed3)],
               y = trainTransformed3$age,
    method = "glmnet",
    trControl = fitControl,
    tuneLength = 10
  )

# save(ridge_model,lasso_model,elastic_model,
#      file = paste0(rdata.path,"Toothfish_regional_models.Rdata"))



# Elastic net with alpha set to 0.5 and best tuning of lambda will be automatically selected

set.seed(123)

elastic_model.05 <-
  caret::train(x = trainTransformed3[,-ncol(trainTransformed3)],
               y = trainTransformed3$age,
    method = "glmnet",
    trControl = fitControl,
    tuneGrid = data.frame(alpha = 0.5, lambda = lambda),
    tuneLength = 10
  )

# save(ridge_model,lasso_model,elastic_model,elastic_model.05,
#      file = paste0(rdata.path,"Toothfish_regional_models.Rdata"))



```

### Compare models

```{r}
print(load(paste0(rdata.path,"Toothfish_regional_models.Rdata")))
print(load(paste0(rdata.path, "TOA_aging_markers.rdata")))
print(load(paste0(rdata.path, "TOA_aging_test_data.Rdata")))
print(load(paste0(rdata.path,"Toothfish_metadata.Rdata")))

metadata3 <- metadata2[metadata2$SAMPLE_ID %in% rownames(trainTransformed3),]
metadata3 <- metadata3[order(match(metadata3$SAMPLE_ID, rownames(trainTransformed3))),]
identical(rownames(trainTransformed3),metadata3$SAMPLE_ID)
trainTransformed3$age <- metadata3$region

metadata3 <- metadata2[metadata2$SAMPLE_ID %in% rownames(testTransformed),]
identical(metadata3$SAMPLE_ID, rownames(testTransformed))
testTransformed$age <- metadata3$region


predicted.class <- caret::predict.train(ridge_model)
predicted.class <- as.character(predicted.class)
# caret::postResample(pred = predicted.class, trainTransformed3$age)
chisq.test(predicted.class, trainTransformed3$age) #X-squared = 252, df = 9, p-value < 2.2e-16
predicted.class == trainTransformed3$age # all TRUE
# heatmap(as.character(predicted.class), trainTransformed3$age) 
predict.ridge.test <- predict(ridge_model, testTransformed)
predict.ridge.test <- as.character(predict.ridge.test)
# caret::postResample(pred = predict.ridge.test, testTransformed$age)
# plot(testTransformed$age, predict.ridge.test)
chisq.test(predict.ridge.test, testTransformed$age) #X-squared = 66.861, df = 9, p-value = 6.26e-11
predict.ridge.test == testTransformed$age 
sum(predict.ridge.test != testTransformed$age)# 5 FALSE


predicted.class <- caret::predict.train(lasso_model)
predicted.class <- as.character(predicted.class)
# caret::postResample(pred = predicted.class, trainTransformed3$age)
chisq.test(predicted.class, trainTransformed3$age) #X-squared = 245.52, df = 9, p-value < 2.2e-16
predicted.class == trainTransformed3$age # all TRUE
# heatmap(as.character(predicted.class), trainTransformed3$age) 
predict.lasso.test <- predict(lasso_model, testTransformed)
predict.lasso.test <- as.character(predict.lasso.test)
# caret::postResample(pred = predict.lasso.test, testTransformed$age)
# plot(testTransformed$age, predict.lasso.test)
chisq.test(predict.lasso.test, testTransformed$age) #X-squared = 61.189, df = 9, p-value = 7.905e-10
predict.lasso.test == testTransformed$age 
sum(predict.lasso.test != testTransformed$age)# 4 FALSE



predicted.class <- caret::predict.train(elastic_model)
predicted.class <- as.character(predicted.class)
# caret::postResample(pred = predicted.class, trainTransformed3$age)
chisq.test(predicted.class, trainTransformed3$age) #X-squared = 252, df = 9, p-value < 2.2e-16
predicted.class == trainTransformed3$age # all TRUE
# heatmap(as.character(predicted.class), trainTransformed3$age) 
predict.enet.test <- predict(elastic_model, testTransformed)
predict.enet.test <- as.character(predict.enet.test)
# caret::postResample(pred = predict.enet.test, testTransformed$age)
# plot(testTransformed$age, predict.enet.test)
chisq.test(predict.enet.test, testTransformed$age) #X-squared = 64.813, df = 9, p-value = 1.569e-10
predict.enet.test == testTransformed$age 
sum(predict.enet.test != testTransformed$age)# 4 FALSE




predicted.class <- caret::predict.train(elastic_model.05)
predicted.class <- as.character(predicted.class)
# caret::postResample(pred = predicted.class, trainTransformed3$age)
chisq.test(predicted.class, trainTransformed3$age) #X-squared = 252, df = 9, p-value < 2.2e-16
predicted.class == trainTransformed3$age # all TRUE
# heatmap(as.character(predicted.class), trainTransformed3$age) 
predict.enet05.test <- predict(elastic_model.05, testTransformed)
predict.enet05.test <- as.character(predict.enet05.test)
# caret::postResample(pred = predict.enet05.test, testTransformed$age)
# plot(testTransformed$age, predict.enet05.test)
chisq.test(predict.enet05.test, testTransformed$age) #X-squared = 64.813, df = 9, p-value = 1.569e-10
predict.enet05.test == testTransformed$age 
sum(predict.enet05.test != testTransformed$age)# 4 FALSE


combo <- data.frame(individuals = rownames(testTransformed), 
                    region = testTransformed$age,
                    predict.ridge.test, predict.lasso.test,
                    predict.enet.test, predict.enet05.test)
```

### Variable contributions
#### Ridge model
```{r}
print(load(paste0(rdata.path,"Toothfish_regional_models.Rdata")))
print(load(paste0(rdata.path,"TOA_data_for_RDA_meth_percent.Rdata"))) #347,614 SMPs

varImp.ridge <- caret::varImp(ridge_model, scale = FALSE)
plot(varImp.ridge, top = 1000)
varImp.ridge.df <- varImp.ridge$importance
varImp.ridge.df$locus <- rownames(varImp.ridge.df)
varImp.ridge.names.58.5.2 <- varImp.ridge.df$locus[varImp.ridge.df$`58-5-2` > 4e-04] #14
varImp.ridge.names.88.1 <- varImp.ridge.df$locus[varImp.ridge.df$`88-1` > 4e-04] #14
varImp.ridge.names.58.4.1 <- varImp.ridge.df$locus[varImp.ridge.df$`58-4-1` > 3e-04] #14
varImp.ridge.names.58.4.2 <- varImp.ridge.df$locus[varImp.ridge.df$`58-4-2` > 3.5e-04] #14

all.model.names <- c(varImp.ridge.names.58.5.2, varImp.ridge.names.88.1,
                     varImp.ridge.names.58.4.1, varImp.ridge.names.58.4.2)
length(unique(all.model.names))#76
sum(duplicated(all.model.names))# 2


meth.RDA.gr <- as(meth.RDA, "GRanges")
x <- as.integer(gsub(pattern = "V", replacement = "", x = all.model.names))
x <- unique(x )
print(data.frame(loc.number = x, 
                 loc.name = paste0(meth.RDA$chr[x],"_",meth.RDA$start[x])))
meth.regional.gr <- meth.RDA.gr[x,]
meth.regional.raw <- meth.RDA[x,]
# meth.aging.prc <- methylKit::percMethylation(meth.regional.raw)
meth.diff.df <- as.data.frame(t(methylKit::percMethylation(meth.regional.raw)))
loc.names <- paste0(meth.regional.raw$chr,"_",meth.regional.raw$start)
colnames(meth.diff.df) <- loc.names

meth.diff.df$individual <- rownames(meth.diff.df)
print(load(paste0(rdata.path,"Toothfish_metadata.Rdata")))
identical(metadata2$SAMPLE_ID, meth.diff.df$individual)
meth.diff.df$pop <- metadata2$region

meth.diff.df.plot <- tidyr::gather(meth.diff.df, key = locus, 
                                   value = perc_methylation, -individual, -pop,
                                   factor_key = TRUE) %>%
  dplyr::mutate(chr = stringr::str_split_fixed(locus, 
                                               pattern = "_", 
                                               n = 2)[,1])
meth.diff.df.plot.locus <- meth.diff.df.plot %>% dplyr::group_by(chr,locus,pop) %>% 
  dplyr::summarise(mean_pop = mean(perc_methylation, na.rm = TRUE),
            .groups = 'drop')

# meth.diff.df.plot.centroid <- meth.diff.df.plot %>% dplyr::group_by(chr,pop) %>% 
#   dplyr::summarise(mean_pop = mean(perc_methylation, na.rm = TRUE),
#             .groups = 'drop')
# 
# meth.diff.df.plot.label <- meth.diff.df.plot %>% dplyr::group_by(chr) %>% 
#   dplyr::summarise(nmarkers = length(unique(locus)),
#             .groups = 'drop')



region.meth.plot <- ggplot2::ggplot() +
  ggplot2::geom_point(data = meth.diff.df.plot[meth.diff.df.plot$locus 
                                               %in%
                                                 unique(meth.diff.df.plot$locus)
                                               [1:5],],
                      ggplot2::aes(x = pop, y = perc_methylation, colour = locus)) +
  ggplot2::geom_line(data = meth.diff.df.plot.locus[meth.diff.df.plot.locus$locus 
                                             %in% 
                                               unique(meth.diff.df.plot$locus)
                                             [1:5],],
                     ggplot2::aes(x = pop, y = mean_pop, colour = locus, group = locus)) +
  ggplot2::geom_point(data = meth.diff.df.plot.locus[meth.diff.df.plot.locus$locus 
                                             %in% 
                                               unique(meth.diff.df.plot$locus)[1:5],],
                     ggplot2::aes(x = pop, y = mean_pop, colour = locus, group = locus),
                     size = 5)
print(region.meth.plot)


region.meth.plot2 <- ggplot2::ggplot() +
  ggplot2::geom_point(data = meth.diff.df.plot,
                      ggplot2::aes(x = pop, y = perc_methylation, colour = locus)) +
  ggplot2::geom_line(data = meth.diff.df.plot.locus,
                     ggplot2::aes(x = pop, y = mean_pop, colour = locus, group = locus)) +
  ggplot2::geom_point(data = meth.diff.df.plot.locus,
                     ggplot2::aes(x = pop, y = mean_pop, colour = locus, group = locus),
                     size = 5) +
  ggplot2::facet_wrap(~chr) + 
  ggplot2::guides(colour = ggplot2::guide_legend(ncol = 2))
print(region.meth.plot2)


ggplot2::ggsave(plot = region.meth.plot2,
                filename = paste0(figure.path,
                                  # "TOA_aging_methylation_percentage_enet05.png"
                                  "TOA_Regional_methylation_percentage_ridge.png"
                                  # "TOA_aging_methylation_percentage_all.models.png"
                                  ),
                                  width = 45, height = 30,
                                  units = "cm", device = "png")
```

##### PCA
```{r}
print(load(paste0(rdata.path,"Toothfish_metadata.Rdata")))

meth.regional.raw@treatment <- as.integer(as.factor(metadata2$region))

methylKit::PCASamples(meth.regional.raw, screeplot = TRUE)
dev.print(file = paste0(figure.path,
                        "PCA_plot_diffMeth_commonCpG_RidgeModel_scree.png"),
          device = png, res = 300,width = 30,height = 15,units = "cm")

methylKit::PCASamples(meth.regional.raw, comp = c(1,2),
                      adj.lim = c(4e-04, 0.1),
                      transpose = TRUE,
                      obj.return = FALSE,
                      # color = c("blue", "red", "black", "purple")
                      )
dev.print(file = paste0(figure.path,
                        "PCA_plot_diffMeth_commonCpG_RidgeModel_PC12.png"
                        ),
          device = png, res = 300,width = 30,height = 15,units = "cm")

methylKit::PCASamples(meth.regional.raw, comp = c(1,3))
dev.print(file = paste0(figure.path,
                        "PCA_plot_diffMeth_commonCpG_RidgeModel_PC13.png"
                        ),
          device = png, res = 300,width = 30,height = 15,units = "cm")

methylKit::PCASamples(meth.regional.raw, comp = c(2,3))
dev.print(file = paste0(figure.path,
                        "PCA_plot_diffMeth_commonCpG_RidgeModel_PC23.png"
                        ),
          device = png, res = 300,width = 30,height = 15,units = "cm")
```


#### Lasso model
```{r}
print(load(paste0(rdata.path,"Toothfish_regional_models.Rdata")))
print(load(paste0(rdata.path,"TOA_data_for_RDA_meth_percent.Rdata"))) #347,614 SMPs

varImp.lasso <- caret::varImp(lasso_model, scale = FALSE)
plot(varImp.lasso, top = 60)
varImp.lasso.df <- varImp.lasso$importance
varImp.lasso.df$locus <- rownames(varImp.lasso.df)
varImp.lasso.names.58.5.2 <- varImp.lasso.df$locus[varImp.lasso.df$`58-5-2` > 0] #7
varImp.lasso.names.88.1 <- varImp.lasso.df$locus[varImp.lasso.df$`88-1` > 0] #21
varImp.lasso.names.58.4.1 <- varImp.lasso.df$locus[varImp.lasso.df$`58-4-1` > 0] #11
varImp.lasso.names.58.4.2 <- varImp.lasso.df$locus[varImp.lasso.df$`58-4-2` > 0] #17

all.model.names <- c(varImp.lasso.names.58.5.2, varImp.lasso.names.88.1,
                     varImp.lasso.names.58.4.1, varImp.lasso.names.58.4.2)
length(unique(all.model.names))#56
sum(duplicated(all.model.names))# 0


meth.RDA.gr <- as(meth.RDA, "GRanges")
x <- as.integer(gsub(pattern = "V", replacement = "", x = all.model.names))
x <- unique(x )
print(data.frame(loc.number = x, 
                 loc.name = paste0(meth.RDA$chr[x],"_",meth.RDA$start[x])))
meth.regional.gr <- meth.RDA.gr[x,]
meth.regional.raw <- meth.RDA[x,]
# meth.aging.prc <- methylKit::percMethylation(meth.regional.raw)
meth.diff.df <- as.data.frame(t(methylKit::percMethylation(meth.regional.raw)))
loc.names <- paste0(meth.regional.raw$chr,"_",meth.regional.raw$start)
colnames(meth.diff.df) <- loc.names

meth.diff.df$individual <- rownames(meth.diff.df)
print(load(paste0(rdata.path,"Toothfish_metadata.Rdata")))
identical(metadata2$SAMPLE_ID, meth.diff.df$individual)
meth.diff.df$pop <- metadata2$region

meth.diff.df.plot <- tidyr::gather(meth.diff.df, key = locus, 
                                   value = perc_methylation, -individual, -pop,
                                   factor_key = TRUE) %>%
  dplyr::mutate(chr = stringr::str_split_fixed(locus, 
                                               pattern = "_", 
                                               n = 2)[,1])
meth.diff.df.plot.locus <- meth.diff.df.plot %>% dplyr::group_by(chr,locus,pop) %>% 
  dplyr::summarise(mean_pop = mean(perc_methylation, na.rm = TRUE),
            .groups = 'drop')

# meth.diff.df.plot.centroid <- meth.diff.df.plot %>% dplyr::group_by(chr,pop) %>% 
#   dplyr::summarise(mean_pop = mean(perc_methylation, na.rm = TRUE),
#             .groups = 'drop')
# 
# meth.diff.df.plot.label <- meth.diff.df.plot %>% dplyr::group_by(chr) %>% 
#   dplyr::summarise(nmarkers = length(unique(locus)),
#             .groups = 'drop')



region.meth.plot <- ggplot2::ggplot() +
  ggplot2::geom_point(data = meth.diff.df.plot[meth.diff.df.plot$locus 
                                               %in%
                                                 unique(meth.diff.df.plot$locus)
                                               [1:5],],
                      ggplot2::aes(x = pop, y = perc_methylation, colour = locus)) +
  ggplot2::geom_line(data = meth.diff.df.plot.locus[meth.diff.df.plot.locus$locus 
                                             %in% 
                                               unique(meth.diff.df.plot$locus)
                                             [1:5],],
                     ggplot2::aes(x = pop, y = mean_pop, colour = locus, group = locus)) +
  ggplot2::geom_point(data = meth.diff.df.plot.locus[meth.diff.df.plot.locus$locus 
                                             %in% 
                                               unique(meth.diff.df.plot$locus)[1:5],],
                     ggplot2::aes(x = pop, y = mean_pop, colour = locus, group = locus),
                     size = 5)
print(region.meth.plot)



region.meth.plot2 <- ggplot2::ggplot() +
  ggplot2::geom_point(data = meth.diff.df.plot,
                      ggplot2::aes(x = pop, y = perc_methylation, colour = locus)) +
  ggplot2::geom_line(data = meth.diff.df.plot.locus,
                     ggplot2::aes(x = pop, y = mean_pop, colour = locus, group = locus)) +
  ggplot2::geom_point(data = meth.diff.df.plot.locus,
                     ggplot2::aes(x = pop, y = mean_pop, colour = locus, group = locus),
                     size = 5) +
  ggplot2::facet_wrap(~chr) + 
  ggplot2::guides(colour = ggplot2::guide_legend(ncol = 2))
print(region.meth.plot2)

ggplot2::ggsave(plot = region.meth.plot2,
                filename = paste0(figure.path,
                                  # "TOA_aging_methylation_percentage_enet05.png"
                                  # "TOA_aging_methylation_percentage_ridge.png"
                                  "TOA_Regional_methylation_percentage_lasso.png"
                                  # "TOA_aging_methylation_percentage_all.models.png"
                                  ),
                                  width = 45, height = 30,
                                  units = "cm", device = "png")
```

##### PCA
```{r}
print(load(paste0(rdata.path,"Toothfish_metadata.Rdata")))

meth.regional.raw@treatment <- as.integer(as.factor(metadata2$region))

methylKit::PCASamples(meth.regional.raw, screeplot = TRUE)
dev.print(file = paste0(figure.path,
                        "PCA_plot_diffMeth_commonCpG_LassoModel_scree.png"),
          device = png, res = 300,width = 30,height = 15,units = "cm")

methylKit::PCASamples(meth.regional.raw, comp = c(1,2),
                      adj.lim = c(4e-04, 0.1),
                      transpose = TRUE,
                      obj.return = FALSE,
                      # color = c("blue", "red", "black", "purple")
                      )
dev.print(file = paste0(figure.path,
                        "PCA_plot_diffMeth_commonCpG_LassoModel_PC12.png"
                        ),
          device = png, res = 300,width = 30,height = 15,units = "cm")

methylKit::PCASamples(meth.regional.raw, comp = c(1,3))
dev.print(file = paste0(figure.path,
                        "PCA_plot_diffMeth_commonCpG_LassoModel_PC13.png"
                        ),
          device = png, res = 300,width = 30,height = 15,units = "cm")

methylKit::PCASamples(meth.regional.raw, comp = c(2,3))
dev.print(file = paste0(figure.path,
                        "PCA_plot_diffMeth_commonCpG_LassoModel_PC23.png"
                        ),
          device = png, res = 300,width = 30,height = 15,units = "cm")
```




#### elastic network model
```{r}
print(load(paste0(rdata.path,"Toothfish_regional_models.Rdata")))
print(load(paste0(rdata.path,"TOA_data_for_RDA_meth_percent.Rdata"))) #347,614 SMPs

varImp.enet <- caret::varImp(elastic_model, scale = FALSE)
plot(varImp.enet, top = 500)
varImp.enet.df <- varImp.enet$importance
varImp.enet.df$locus <- rownames(varImp.enet.df)
varImp.enet.names.58.5.2 <- varImp.enet.df$locus[varImp.enet.df$`58-5-2` > 0.025] #25
varImp.enet.names.88.1 <- varImp.enet.df$locus[varImp.enet.df$`88-1` > 0.025] #27
varImp.enet.names.58.4.1 <- varImp.enet.df$locus[varImp.enet.df$`58-4-1` > 0.025] #28
varImp.enet.names.58.4.2 <- varImp.enet.df$locus[varImp.enet.df$`58-4-2` > 0.025] #23

all.model.names <- c(varImp.enet.names.58.5.2, varImp.enet.names.88.1,
                     varImp.enet.names.58.4.1, varImp.enet.names.58.4.2)
length(unique(all.model.names))#104
sum(duplicated(all.model.names))# 1


meth.RDA.gr <- as(meth.RDA, "GRanges")
x <- as.integer(gsub(pattern = "V", replacement = "", x = all.model.names))
x <- unique(x )
print(data.frame(loc.number = x, 
                 loc.name = paste0(meth.RDA$chr[x],"_",meth.RDA$start[x])))
meth.regional.gr <- meth.RDA.gr[x,]
meth.regional.raw <- meth.RDA[x,]
# meth.aging.prc <- methylKit::percMethylation(meth.regional.raw)
meth.diff.df <- as.data.frame(t(methylKit::percMethylation(meth.regional.raw)))
loc.names <- paste0(meth.regional.raw$chr,"_",meth.regional.raw$start)
colnames(meth.diff.df) <- loc.names

meth.diff.df$individual <- rownames(meth.diff.df)
print(load(paste0(rdata.path,"Toothfish_metadata.Rdata")))
identical(metadata2$SAMPLE_ID, meth.diff.df$individual)
meth.diff.df$pop <- metadata2$region

meth.diff.df.plot <- tidyr::gather(meth.diff.df, key = locus, 
                                   value = perc_methylation, -individual, -pop,
                                   factor_key = TRUE) %>%
  dplyr::mutate(chr = stringr::str_split_fixed(locus, 
                                               pattern = "_", 
                                               n = 2)[,1])
meth.diff.df.plot.locus <- meth.diff.df.plot %>% dplyr::group_by(chr,locus,pop) %>% 
  dplyr::summarise(mean_pop = mean(perc_methylation, na.rm = TRUE),
            .groups = 'drop')

# meth.diff.df.plot.centroid <- meth.diff.df.plot %>% dplyr::group_by(chr,pop) %>% 
#   dplyr::summarise(mean_pop = mean(perc_methylation, na.rm = TRUE),
#             .groups = 'drop')
# 
# meth.diff.df.plot.label <- meth.diff.df.plot %>% dplyr::group_by(chr) %>% 
#   dplyr::summarise(nmarkers = length(unique(locus)),
#             .groups = 'drop')



region.meth.plot <- ggplot2::ggplot() +
  ggplot2::geom_point(data = meth.diff.df.plot[meth.diff.df.plot$locus 
                                               %in%
                                                 unique(meth.diff.df.plot$locus)
                                               [1:5],],
                      ggplot2::aes(x = pop, y = perc_methylation, colour = locus)) +
  ggplot2::geom_line(data = meth.diff.df.plot.locus[meth.diff.df.plot.locus$locus 
                                             %in% 
                                               unique(meth.diff.df.plot$locus)
                                             [1:5],],
                     ggplot2::aes(x = pop, y = mean_pop, colour = locus, group = locus)) +
  ggplot2::geom_point(data = meth.diff.df.plot.locus[meth.diff.df.plot.locus$locus 
                                             %in% 
                                               unique(meth.diff.df.plot$locus)[1:5],],
                     ggplot2::aes(x = pop, y = mean_pop, colour = locus, group = locus),
                     size = 5)
print(region.meth.plot)



region.meth.plot2 <- ggplot2::ggplot() +
  ggplot2::geom_point(data = meth.diff.df.plot,
                      ggplot2::aes(x = pop, y = perc_methylation, colour = locus)) +
  ggplot2::geom_line(data = meth.diff.df.plot.locus,
                     ggplot2::aes(x = pop, y = mean_pop, colour = locus, group = locus)) +
  ggplot2::geom_point(data = meth.diff.df.plot.locus,
                     ggplot2::aes(x = pop, y = mean_pop, colour = locus, group = locus),
                     size = 5) +
  ggplot2::facet_wrap(~chr) + 
  ggplot2::guides(colour = ggplot2::guide_legend(ncol = 3))
  
print(region.meth.plot2)

ggplot2::ggsave(plot = region.meth.plot2,
                filename = paste0(figure.path,
                                  # "TOA_aging_methylation_percentage_enet05.png"
                                  # "TOA_aging_methylation_percentage_ridge.png"
                                  # "TOA_aging_methylation_percentage_lasso.png"
                                  "TOA_Regional_methylation_percentage_enet.png"
                                  # "TOA_aging_methylation_percentage_all.models.png"
                                  ),
                                  width = 45, height = 30,
                                  units = "cm", device = "png")
```

##### PCA
```{r}
print(load(paste0(rdata.path,"Toothfish_metadata.Rdata")))

meth.regional.raw@treatment <- as.integer(as.factor(metadata2$region))

methylKit::PCASamples(meth.regional.raw, screeplot = TRUE)
dev.print(file = paste0(figure.path,
                        "PCA_plot_diffMeth_commonCpG_enetModel_scree.png"),
          device = png, res = 300,width = 30,height = 15,units = "cm")

methylKit::PCASamples(meth.regional.raw, comp = c(1,2),
                      adj.lim = c(4e-04, 0.1),
                      transpose = TRUE,
                      obj.return = FALSE,
                      # color = c("blue", "red", "black", "purple")
                      )
dev.print(file = paste0(figure.path,
                        "PCA_plot_diffMeth_commonCpG_enetModel_PC12.png"
                        ),
          device = png, res = 300,width = 30,height = 15,units = "cm")

methylKit::PCASamples(meth.regional.raw, comp = c(1,3))
dev.print(file = paste0(figure.path,
                        "PCA_plot_diffMeth_commonCpG_enetModel_PC13.png"
                        ),
          device = png, res = 300,width = 30,height = 15,units = "cm")

methylKit::PCASamples(meth.regional.raw, comp = c(2,3))
dev.print(file = paste0(figure.path,
                        "PCA_plot_diffMeth_commonCpG_enetModel_PC23.png"
                        ),
          device = png, res = 300,width = 30,height = 15,units = "cm")
```




#### elastic network model
```{r}
print(load(paste0(rdata.path,"Toothfish_regional_models.Rdata")))
print(load(paste0(rdata.path,"TOA_data_for_RDA_meth_percent.Rdata"))) #347,614 SMPs

varImp.enet <- caret::varImp(elastic_model, scale = FALSE)
plot(varImp.enet, top = 500)
varImp.enet.df <- varImp.enet$importance
varImp.enet.df$locus <- rownames(varImp.enet.df)
varImp.enet.names.58.5.2 <- varImp.enet.df$locus[varImp.enet.df$`58-5-2` > 0.025] #25
varImp.enet.names.88.1 <- varImp.enet.df$locus[varImp.enet.df$`88-1` > 0.025] #27
varImp.enet.names.58.4.1 <- varImp.enet.df$locus[varImp.enet.df$`58-4-1` > 0.025] #28
varImp.enet.names.58.4.2 <- varImp.enet.df$locus[varImp.enet.df$`58-4-2` > 0.025] #23

all.model.names <- c(varImp.enet.names.58.5.2, varImp.enet.names.88.1,
                     varImp.enet.names.58.4.1, varImp.enet.names.58.4.2)
length(unique(all.model.names))#104
sum(duplicated(all.model.names))# 1


meth.RDA.gr <- as(meth.RDA, "GRanges")
x <- as.integer(gsub(pattern = "V", replacement = "", x = all.model.names))
x <- unique(x )
print(data.frame(loc.number = x, 
                 loc.name = paste0(meth.RDA$chr[x],"_",meth.RDA$start[x])))
meth.regional.gr <- meth.RDA.gr[x,]
meth.regional.raw <- meth.RDA[x,]
# meth.aging.prc <- methylKit::percMethylation(meth.regional.raw)
meth.diff.df <- as.data.frame(t(methylKit::percMethylation(meth.regional.raw)))
loc.names <- paste0(meth.regional.raw$chr,"_",meth.regional.raw$start)
colnames(meth.diff.df) <- loc.names

meth.diff.df$individual <- rownames(meth.diff.df)
print(load(paste0(rdata.path,"Toothfish_metadata.Rdata")))
identical(metadata2$SAMPLE_ID, meth.diff.df$individual)
meth.diff.df$pop <- metadata2$region

meth.diff.df.plot <- tidyr::gather(meth.diff.df, key = locus, 
                                   value = perc_methylation, -individual, -pop,
                                   factor_key = TRUE) %>%
  dplyr::mutate(chr = stringr::str_split_fixed(locus, 
                                               pattern = "_", 
                                               n = 2)[,1])
meth.diff.df.plot.locus <- meth.diff.df.plot %>% dplyr::group_by(chr,locus,pop) %>% 
  dplyr::summarise(mean_pop = mean(perc_methylation, na.rm = TRUE),
            .groups = 'drop')

# meth.diff.df.plot.centroid <- meth.diff.df.plot %>% dplyr::group_by(chr,pop) %>% 
#   dplyr::summarise(mean_pop = mean(perc_methylation, na.rm = TRUE),
#             .groups = 'drop')
# 
# meth.diff.df.plot.label <- meth.diff.df.plot %>% dplyr::group_by(chr) %>% 
#   dplyr::summarise(nmarkers = length(unique(locus)),
#             .groups = 'drop')



region.meth.plot <- ggplot2::ggplot() +
  ggplot2::geom_point(data = meth.diff.df.plot[meth.diff.df.plot$locus 
                                               %in%
                                                 unique(meth.diff.df.plot$locus)
                                               [1:5],],
                      ggplot2::aes(x = pop, y = perc_methylation, colour = locus)) +
  ggplot2::geom_line(data = meth.diff.df.plot.locus[meth.diff.df.plot.locus$locus 
                                             %in% 
                                               unique(meth.diff.df.plot$locus)
                                             [1:5],],
                     ggplot2::aes(x = pop, y = mean_pop, colour = locus, group = locus)) +
  ggplot2::geom_point(data = meth.diff.df.plot.locus[meth.diff.df.plot.locus$locus 
                                             %in% 
                                               unique(meth.diff.df.plot$locus)[1:5],],
                     ggplot2::aes(x = pop, y = mean_pop, colour = locus, group = locus),
                     size = 5)
print(region.meth.plot)



region.meth.plot2 <- ggplot2::ggplot() +
  ggplot2::geom_point(data = meth.diff.df.plot,
                      ggplot2::aes(x = pop, y = perc_methylation, colour = locus)) +
  ggplot2::geom_line(data = meth.diff.df.plot.locus,
                     ggplot2::aes(x = pop, y = mean_pop, colour = locus, group = locus)) +
  ggplot2::geom_point(data = meth.diff.df.plot.locus,
                     ggplot2::aes(x = pop, y = mean_pop, colour = locus, group = locus),
                     size = 5) +
  ggplot2::facet_wrap(~chr) + 
  ggplot2::guides(colour = ggplot2::guide_legend(ncol = 3))
  
print(region.meth.plot2)

ggplot2::ggsave(plot = region.meth.plot2,
                filename = paste0(figure.path,
                                  # "TOA_aging_methylation_percentage_enet05.png"
                                  # "TOA_aging_methylation_percentage_ridge.png"
                                  # "TOA_aging_methylation_percentage_lasso.png"
                                  "TOA_Regional_methylation_percentage_enet.png"
                                  # "TOA_aging_methylation_percentage_all.models.png"
                                  ),
                                  width = 45, height = 30,
                                  units = "cm", device = "png")
```

##### PCA
```{r}
print(load(paste0(rdata.path,"Toothfish_metadata.Rdata")))

meth.regional.raw@treatment <- as.integer(as.factor(metadata2$region))

methylKit::PCASamples(meth.regional.raw, screeplot = TRUE)
dev.print(file = paste0(figure.path,
                        "PCA_plot_diffMeth_commonCpG_enetModel_scree.png"),
          device = png, res = 300,width = 30,height = 15,units = "cm")

methylKit::PCASamples(meth.regional.raw, comp = c(1,2),
                      adj.lim = c(4e-04, 0.1),
                      transpose = TRUE,
                      obj.return = FALSE,
                      # color = c("blue", "red", "black", "purple")
                      )
dev.print(file = paste0(figure.path,
                        "PCA_plot_diffMeth_commonCpG_enetModel_PC12.png"
                        ),
          device = png, res = 300,width = 30,height = 15,units = "cm")

methylKit::PCASamples(meth.regional.raw, comp = c(1,3))
dev.print(file = paste0(figure.path,
                        "PCA_plot_diffMeth_commonCpG_enetModel_PC13.png"
                        ),
          device = png, res = 300,width = 30,height = 15,units = "cm")

methylKit::PCASamples(meth.regional.raw, comp = c(2,3))
dev.print(file = paste0(figure.path,
                        "PCA_plot_diffMeth_commonCpG_enetModel_PC23.png"
                        ),
          device = png, res = 300,width = 30,height = 15,units = "cm")
```




#### elastic network model, Alpha = 0.5

```{r}
print(load(paste0(rdata.path,"Toothfish_regional_models.Rdata")))
print(load(paste0(rdata.path,"TOA_data_for_RDA_meth_percent.Rdata"))) #347,614 SMPs

varImp.enet05 <- caret::varImp(elastic_model.05, scale = FALSE)
plot(varImp.enet05, top = 200)
varImp.enet05.df <- varImp.enet05$importance
varImp.enet05.df$locus <- rownames(varImp.enet05.df)
varImp.enet05.names.58.5.2 <- varImp.enet05.df$locus[varImp.enet05.df$`58-5-2` > 0.025] #18
varImp.enet05.names.88.1 <- varImp.enet05.df$locus[varImp.enet05.df$`88-1` > 0.025] #34
varImp.enet05.names.58.4.1 <- varImp.enet05.df$locus[varImp.enet05.df$`58-4-1` > 0.025] #28
varImp.enet05.names.58.4.2 <- varImp.enet05.df$locus[varImp.enet05.df$`58-4-2` > 0.025] #28


all.model.names <- c(varImp.enet05.names.58.5.2, varImp.enet05.names.88.1,
                     varImp.enet05.names.58.4.1, varImp.enet05.names.58.4.2)
length(unique(all.model.names))#104
sum(duplicated(all.model.names))# 1


meth.RDA.gr <- as(meth.RDA, "GRanges")
x <- as.integer(gsub(pattern = "V", replacement = "", x = all.model.names))
x <- unique(x )
print(data.frame(loc.number = x, 
                 loc.name = paste0(meth.RDA$chr[x],"_",meth.RDA$start[x])))
meth.regional.gr <- meth.RDA.gr[x,]
meth.regional.raw <- meth.RDA[x,]
# meth.aging.prc <- methylKit::percMethylation(meth.regional.raw)
meth.diff.df <- as.data.frame(t(methylKit::percMethylation(meth.regional.raw)))
loc.names <- paste0(meth.regional.raw$chr,"_",meth.regional.raw$start)
colnames(meth.diff.df) <- loc.names

meth.diff.df$individual <- rownames(meth.diff.df)
print(load(paste0(rdata.path,"Toothfish_metadata.Rdata")))
identical(metadata2$SAMPLE_ID, meth.diff.df$individual)
meth.diff.df$pop <- metadata2$region

meth.diff.df.plot <- tidyr::gather(meth.diff.df, key = locus, 
                                   value = perc_methylation, -individual, -pop,
                                   factor_key = TRUE) %>%
  dplyr::mutate(chr = stringr::str_split_fixed(locus, 
                                               pattern = "_", 
                                               n = 2)[,1])
meth.diff.df.plot.locus <- meth.diff.df.plot %>% dplyr::group_by(chr,locus,pop) %>% 
  dplyr::summarise(mean_pop = mean(perc_methylation, na.rm = TRUE),
            .groups = 'drop')

# meth.diff.df.plot.centroid <- meth.diff.df.plot %>% dplyr::group_by(chr,pop) %>% 
#   dplyr::summarise(mean_pop = mean(perc_methylation, na.rm = TRUE),
#             .groups = 'drop')
# 
# meth.diff.df.plot.label <- meth.diff.df.plot %>% dplyr::group_by(chr) %>% 
#   dplyr::summarise(nmarkers = length(unique(locus)),
#             .groups = 'drop')



region.meth.plot <- ggplot2::ggplot() +
  ggplot2::geom_point(data = meth.diff.df.plot[meth.diff.df.plot$locus 
                                               %in%
                                                 unique(meth.diff.df.plot$locus)
                                               [1:5],],
                      ggplot2::aes(x = pop, y = perc_methylation, colour = locus)) +
  ggplot2::geom_line(data = meth.diff.df.plot.locus[meth.diff.df.plot.locus$locus 
                                             %in% 
                                               unique(meth.diff.df.plot$locus)
                                             [1:5],],
                     ggplot2::aes(x = pop, y = mean_pop, colour = locus, group = locus)) +
  ggplot2::geom_point(data = meth.diff.df.plot.locus[meth.diff.df.plot.locus$locus 
                                             %in% 
                                               unique(meth.diff.df.plot$locus)[1:5],],
                     ggplot2::aes(x = pop, y = mean_pop, colour = locus, group = locus),
                     size = 5)
print(region.meth.plot)



region.meth.plot2 <- ggplot2::ggplot() +
  ggplot2::geom_point(data = meth.diff.df.plot,
                      ggplot2::aes(x = pop, y = perc_methylation, colour = locus)) +
  ggplot2::geom_line(data = meth.diff.df.plot.locus,
                     ggplot2::aes(x = pop, y = mean_pop, colour = locus, group = locus)) +
  ggplot2::geom_point(data = meth.diff.df.plot.locus,
                     ggplot2::aes(x = pop, y = mean_pop, colour = locus, group = locus),
                     size = 5) +
  ggplot2::facet_wrap(~chr) + 
  ggplot2::guides(colour = ggplot2::guide_legend(ncol = 1))
  
print(region.meth.plot2)

ggplot2::ggsave(plot = region.meth.plot2,
                filename = paste0(figure.path,
                                  # "TOA_aging_methylation_percentage_enet05.png"
                                  # "TOA_aging_methylation_percentage_ridge.png"
                                  # "TOA_aging_methylation_percentage_lasso.png"
                                  "TOA_Regional_methylation_percentage_enet05.png"
                                  # "TOA_aging_methylation_percentage_all.models.png"
                                  ),
                                  width = 45, height = 30,
                                  units = "cm", device = "png")
```

##### PCA
```{r}
print(load(paste0(rdata.path,"Toothfish_metadata.Rdata")))

meth.regional.raw@treatment <- as.integer(as.factor(metadata2$region))

methylKit::PCASamples(meth.regional.raw, screeplot = TRUE)
dev.print(file = paste0(figure.path,
                        "PCA_plot_diffMeth_commonCpG_enet05Model_scree.png"),
          device = png, res = 300,width = 30,height = 15,units = "cm")

methylKit::PCASamples(meth.regional.raw, comp = c(1,2),
                      adj.lim = c(4e-04, 0.1),
                      transpose = TRUE,
                      obj.return = FALSE,
                      # color = c("blue", "red", "black", "purple")
                      )
dev.print(file = paste0(figure.path,
                        "PCA_plot_diffMeth_commonCpG_enet05Model_PC12.png"
                        ),
          device = png, res = 300,width = 30,height = 15,units = "cm")

methylKit::PCASamples(meth.regional.raw, comp = c(1,3))
dev.print(file = paste0(figure.path,
                        "PCA_plot_diffMeth_commonCpG_enet05Model_PC13.png"
                        ),
          device = png, res = 300,width = 30,height = 15,units = "cm")

methylKit::PCASamples(meth.regional.raw, comp = c(2,3))
dev.print(file = paste0(figure.path,
                        "PCA_plot_diffMeth_commonCpG_enet05Model_PC23.png"
                        ),
          device = png, res = 300,width = 30,height = 15,units = "cm")
```

# NEW SAMPLES, OLD MARKERS

```{r}
print(load(paste0(rdata.path,"Toothfish_methylkit_All_muscle_objects_raw.Rdata")))
print(load(paste0(rdata.path,"Toothfish_DMS_CpGsites.Rdata")))
meth <- meth.2018M.norm11L
meth.cpg <- paste0(meth$chr, "_", meth$start)
DMS.cpg <- unique(c(DMS$DMS)) #1243
meth.DMS <- meth[meth.cpg %in% DMS.cpg,] #1144
```
### Combined heatmap

```{r}
metadata.new <- metadata[metadata$Genotype_names %in% attributes(meth)$sample.ids,]
attributes(meth)$sample.ids == metadata.new$Genotype_names

pop <- paste0(metadata.new$STRATA1,"-", metadata.new$Year)
pop.levels <- c("58.4.1-2018","58.4.2-2018", "58.4.2-2024", 
                "58.5.2-2017", "88.1-2018", "88.2-2024" )
pop <- factor(pop, levels = pop.levels)
pop <- as.numeric(pop)

meth.cpg <- paste0(meth$chr, "_", meth$start)
DMS.cpg <- unique(c(DMS$DMS)) #1243

meth.DMS <- meth[meth.cpg %in% DMS.cpg,] #1144
length(unique(meth.DMS$chr)) #26
sort(table(meth.DMS$chr), decreasing = TRUE)[1:10]



meth.diff.df <- as.data.frame(t(methylKit::percMethylation(meth.DMS)))
loc.names <- paste0(meth.DMS$chr,"_",meth.DMS$start)
colnames(meth.diff.df) <- loc.names

meth.diff.df$individual <- rownames(meth.diff.df) 
identical(metadata.new$Genotype_names,meth.diff.df$individual)
pop <- paste0(metadata.new$STRATA1,"-", metadata.new$Year)
pop.levels <- c("58.4.1-2018","58.4.2-2018", "58.4.2-2024", 
                "58.5.2-2017", "88.1-2018", "88.2-2024" )
pop <- factor(pop, levels = pop.levels)
meth.diff.df$pop <- pop
reference.ind <- c("TOA_58-4-1_113_AN176969")


#### overlap levels
heat.plot.data <- meth.diff.df %>%
  tibble::as_tibble() %>%
  tidyr::gather(key = "CpG", value = "Methylation_percentage", -individual, -pop,
                factor_key = FALSE) %>%
  # dplyr::group_by(pop, CpG) %>%
  dplyr::group_by(CpG) %>%
  # dplyr::filter(CpG %in% overlap.cpg) %>%
  dplyr::mutate(CpG_mean = mean(Methylation_percentage, na.rm = TRUE)) %>%
  dplyr::arrange(pop, individual, CpG) %>%
  dplyr::mutate(individual = factor(individual, levels = unique(individual)))

# cpgs <- heat.plot.data$CpG[heat.plot.data$individual == reference.ind]
# cpgs_mean <- heat.plot.data$CpG_mean[heat.plot.data$individual == reference.ind]

# cpgs_mean <- heat.plot.data[heat.plot.data$individual %in%
#                               reference.ind,] %>%
#   dplyr::group_by(CpG) %>%
#   dplyr::summarise(CpG_mean = mean(Methylation_percentage , na.rm = TRUE))

# overlap.cpg <- cpgs[order(cpgs_mean$CpG_mean, decreasing = TRUE)]
# overlap.cpg <- cpgs[order(cpgs_mean, decreasing = TRUE)]


# cpg.levels <- c("pop", meth.cpg.L, "Liver_DMS", meth.cpg.H, "Heart_DMS",
#                 meth.cpg.M, "Muscle_DMS", overlap.cpg, "Common_DMS" )


heat.plot.data <- meth.diff.df %>%
  tibble::as_tibble() %>%
  tidyr::gather(key = "CpG", value = "Methylation_percentage", -individual, -pop,
                factor_key = FALSE) %>%
  dplyr::group_by(pop, CpG) %>%
  dplyr::mutate(CpG_mean = mean(Methylation_percentage, na.rm = TRUE)) %>%
  dplyr::arrange(pop, individual, CpG) %>%
  dplyr::mutate(individual = factor(individual, levels = unique(individual))) %>%
  # dplyr::mutate(CpG = factor(CpG, levels = cpg.levels)) %>%
  dplyr::ungroup() %>%
  dplyr::group_by(individual) %>%
  dplyr::mutate(ind.mean = mean(Methylation_percentage, na.rm = TRUE))

# meta <- metadata.new[order(match(metadata.new$Genotype_names, unique(heat.plot.data$individual))),]
# 
# ord1 <- order(heat.plot.data$pop[heat.plot.data$CpG == "ptg001232l_375643"],
#               meta$POP,
#       heat.plot.data$ind.mean[heat.plot.data$CpG == "ptg001232l_375643"])

# ind.levels <- as.character(unique(heat.plot.data$individual)[ord1])
# meta <- metadata.new[order(match(metadata.new$individual, ind.levels)),]
# meta2 <- meta[meta$strata %in% c("TAS-control (18C/33‰)","TAS-cold-fresh (14C/25‰)",
#                                    "TAS-warm-salty (22C/40‰)"),]
# ind.levels[meta$strata %in% c("TAS-control (18C/33‰)","TAS-cold-fresh (14C/25‰)",
#                                    "TAS-warm-salty (22C/40‰)")] <- meta2$individual[order(meta2$strata, meta2$individual)]

# heat.plot.data$individual <- factor(heat.plot.data$individual, levels = ind.levels)
# heat.plot.data <- heat.plot.data[order(heat.plot.data$CpG, heat.plot.data$individual),]


heat.plot <- ggplot2::ggplot(data = heat.plot.data, ggplot2::aes(x = individual)) + 
    ggplot2::geom_tile(data = heat.plot.data, 
                       ggplot2::aes(x = individual, y = CpG, fill = Methylation_percentage)) +
    ggplot2::scale_fill_gradient2(low = "blue", mid = "yellow", high = "red", 
                                  midpoint = 50, guide = "colourbar") +
    # ggnewscale::new_scale_fill() +
    # ggplot2::geom_tile(data = heat.plot.data, ggplot2::aes(x = individual, y = "Common_DMS",
    #                                                        height = 5), fill = "white") +
    # ggplot2::geom_text(ggplot2::aes(x = 62, y = "Common_DMS",
    #                        label = "Common DMS between muscle, heart and liver tissues")) +
    # ggnewscale::new_scale_fill() +
    # ggplot2::geom_tile(data = heat.plot.data, ggplot2::aes(x = individual, y = "Muscle_DMS",
    #                                                        height = 5), fill = "white") +
    # ggplot2::geom_text(ggplot2::aes(x = 62,y = "Muscle_DMS",label = "Experiment DMS unique from muscle tissue")) +
    # ggnewscale::new_scale_fill() +
    # ggplot2::geom_tile(data = heat.plot.data, ggplot2::aes(x = individual, y = "Heart_DMS",
    #                                                        height = 5), fill = "white") +
    # ggplot2::geom_text(ggplot2::aes(x = 62,y = "Heart_DMS",label = "Experiment DMS unique from heart tissue")) +
    # ggnewscale::new_scale_fill() +
    # ggplot2::geom_tile(data = heat.plot.data, ggplot2::aes(x = individual, y = "Liver_DMS",
    #                                                        height = 5), fill = "white") +
    # ggplot2::geom_text(ggplot2::aes(x = 62,y = "Liver_DMS",label = "Experiment DMS unique from liver tissue")) +
    ggnewscale::new_scale_fill() +
    ggplot2::geom_tile(data = heat.plot.data, ggplot2::aes(x = individual, 
                                                           y = c("pop"), fill = pop,
                                    height = 5)) +
    ggplot2::scale_fill_manual(values = colours6m) +
    ggplot2::scale_y_discrete(drop = FALSE) +
    ggplot2::theme(legend.position = "top",
                   axis.title.y = ggplot2::element_blank(),
                   axis.title.x = ggplot2::element_blank(),
                   axis.text.x = ggplot2::element_text(face = "bold", color = "black", 
                                        size = 10, angle = 90, hjust = 1, vjust = 0.5))
# print(heat.plot)

ggplot2::ggsave(heat.plot, bg = "white",
                filename = paste0(figure.path,
                                  "TOA_commonCpG_ALLFish_muscle_2024_methDiff_HEATMAP2.png.png"
                                  ), 
                width = 60, height = 30, units = "cm")
```




# CHECK BAD COVERAGE, range and strand
```{r, eval=FALSE}
print(load(paste0(rdata.path,"Toothfish_2018_muscle_methylkit_object.Rdata")))

meth.df <- methylKit::getData(meth.2018M.norm11L)
cov <- meth.df[,meth.2018M.norm11L@coverage.index]
colnames(cov) <- meth.2018M.norm11L@sample.ids
BOOLEAN <- rep(FALSE, nrow(cov))
for (i in 1:ncol(cov)) {
  BOOLEAN1 <- (cov[,i] > 100 | cov[,i] < 0) & !is.na(cov[,i])
  print(sum(BOOLEAN1))
  BOOLEAN <- BOOLEAN | BOOLEAN1
  print(sum(BOOLEAN))
}
meth.2018M.norm11L <- methylKit::select(meth.2018M.norm11L, !BOOLEAN) #81387 removed

BOOLEAN3 <- (meth.2018M.norm11L$start > meth.2018M.norm11L$end) | #629
  (meth.2018M.norm11L$start < meth.2018M.norm11L$end)#643
sum(BOOLEAN3)
meth.2018M.norm11L <- methylKit::select(meth.2018M.norm11L, !BOOLEAN3) #1272 removed

BOOLEAN4 <-  meth.2018M.norm11L$strand!= "*"
sum(BOOLEAN4)
meth.2018M.norm11L <- methylKit::select(meth.2018M.norm11L, !BOOLEAN4) #416960 removed


save(meth.2018M, meth.2018M15L, meth.2018M.norm11L,
     file = paste0(rdata.path,"Toothfish_2018_muscle_methylkit_object.Rdata"))
```


# Session Info

```{r}
sessionInfo() 
```

